// @tamagui/form/dist/cjs/Form.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Form_exports = {};
__export(Form_exports, {
  Form: () => Form2,
  FormContext: () => FormContext,
  FormFrame: () => FormFrame,
  FormProvider: () => FormProvider,
  FormTrigger: () => FormTrigger,
  useFormContext: () => useFormContext
});
module.exports = __toCommonJS(Form_exports);
var import_core = require("@tamagui/core");
var import_helpers = require("@tamagui/helpers");
var import_jsx_runtime = require("react/jsx-runtime");
const FORM_NAME = "Form";
const FormFrame = (0, import_core.styled)(import_core.View, {
  name: FORM_NAME,
  render: "form"
});
const FormContext = (0, import_core.createStyledContext)({
  onSubmit: void 0
});
const {
  useStyledContext: useFormContext,
  Provider: FormProvider
} = FormContext;
const FormTriggerFrame = (0, import_core.styled)(import_core.View, {
  name: "FormTrigger"
});
const FormTrigger = FormTriggerFrame.styleable((props, forwardedRef) => {
  const {
    scope,
    children,
    onPress,
    ...triggerProps
  } = props;
  const context = useFormContext(scope);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(FormTriggerFrame, {
    role: "button",
    ...triggerProps,
    ref: forwardedRef,
    onPress: (0, import_helpers.composeEventHandlers)(onPress, context.onSubmit),
    children
  });
});
const FormComponent = FormFrame.styleable(function Form({
  scope,
  onSubmit,
  ...props
}, ref) {
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(FormProvider, {
    scope,
    onSubmit,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(FormFrame, {
      ref,
      ...props,
      onSubmit: e => {
        e.preventDefault();
        onSubmit?.();
      }
    })
  });
});
const Form2 = (0, import_helpers.withStaticProperties)(FormComponent, {
  Trigger: FormTrigger
});
