// @tamagui/web/dist/cjs/hooks/useMedia.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useMedia_exports = {};
__export(useMedia_exports, {
  _disableMediaTouch: () => _disableMediaTouch,
  configureMedia: () => configureMedia,
  getMediaImportanceIfMoreImportant: () => getMediaImportanceIfMoreImportant,
  getMediaKey: () => getMediaKey,
  getMediaKeyImportance: () => getMediaKeyImportance,
  getMediaState: () => getMediaState,
  isMediaKey: () => isMediaKey,
  mediaKeyMatch: () => mediaKeyMatch,
  mediaKeyToQuery: () => mediaKeyToQuery,
  setMediaShouldUpdate: () => setMediaShouldUpdate,
  setupMediaListeners: () => setupMediaListeners,
  updateMediaListeners: () => updateMediaListeners,
  useMedia: () => useMedia
});
module.exports = __toCommonJS(useMedia_exports);
var import_constants = require("@tamagui/constants");
var import_react = require("react");
var import_config = require("../config.native.js");
var import_createMediaStyle = require("../helpers/createMediaStyle.native.js");
var import_matchMedia = require("../helpers/matchMedia.native.js");
var import_mediaObjectToString = require("../helpers/mediaObjectToString.native.js");
var import_mediaState = require("../helpers/mediaState.native.js");
var import_pseudoDescriptors = require("../helpers/pseudoDescriptors.native.js");
var mediaKeyRegex = /\$(platform|theme|group)-/;
var isMediaKey = function (key) {
  if (key[0] !== "$") return false;
  if (import_mediaState.mediaKeys.has(key)) return true;
  if (mediaKeyRegex.test(key)) return true;
  return false;
};
var getMediaKey = function (key) {
  if (key[0] !== "$") return false;
  if (import_mediaState.mediaKeys.has(key)) return true;
  var match = key.match(mediaKeyRegex);
  if (match) return match[1];
  return false;
};
var initState;
var mediaKeysOrdered;
var getMediaKeyImportance = function (key) {
  if (process.env.NODE_ENV === "development" && key[0] === "$") {
    throw new Error("use short key");
  }
  return mediaKeysOrdered.indexOf(key) + 100;
};
var dispose = /* @__PURE__ */new Set();
var mediaVersion = 0;
var configureMedia = function (config) {
  var {
    media
  } = config;
  var mediaQueryDefaultActive = (0, import_config.getSetting)("mediaQueryDefaultActive");
  if (!media) return;
  mediaVersion++;
  (0, import_createMediaStyle.resetMediaStyleCache)();
  for (var key in media) {
    (0, import_mediaState.getMedia)()[key] = (mediaQueryDefaultActive === null || mediaQueryDefaultActive === void 0 ? void 0 : mediaQueryDefaultActive[key]) || false;
    import_mediaState.mediaKeys.add(`$${key}`);
  }
  Object.assign(import_mediaState.mediaQueryConfig, media);
  initState = {
    ...(0, import_mediaState.getMedia)()
  };
  mediaKeysOrdered = Object.keys(media);
  setupMediaListeners();
};
function unlisten() {
  dispose.forEach(function (cb) {
    return cb();
  });
  dispose.clear();
}
var setupVersion = -1;
function setupMediaListeners() {
  var _loop = function (key2) {
    var str = (0, import_mediaObjectToString.mediaObjectToString)(import_mediaState.mediaQueryConfig[key2]);
    var getMatch = function () {
      return (0, import_matchMedia.matchMedia)(str);
    };
    var match = getMatch();
    if (!match) {
      throw new Error("\u26A0\uFE0F No match");
    }
    match.addListener(update);
    dispose.add(function () {
      match.removeListener(update);
    });
    function update() {
      var next = !!getMatch().matches;
      if (next === (0, import_mediaState.getMedia)()[key2]) return;
      (0, import_mediaState.setMediaState)({
        ...(0, import_mediaState.getMedia)(),
        [key2]: next
      });
      updateMediaListeners();
    }
    update();
  };
  if (import_constants.isWeb && import_constants.isServer) return;
  if (process.env.IS_STATIC) return;
  if (setupVersion === mediaVersion) return;
  setupVersion = mediaVersion;
  unlisten();
  for (var key in import_mediaState.mediaQueryConfig) _loop(key);
}
var listeners = /* @__PURE__ */new Set();
function updateMediaListeners() {
  listeners.forEach(function (cb) {
    return cb((0, import_mediaState.getMedia)());
  });
}
var States = /* @__PURE__ */new WeakMap();
function setMediaShouldUpdate(ref, enabled, keys) {
  var cur = States.get(ref);
  if (!cur || cur.enabled !== enabled || keys) {
    States.set(ref, {
      ...cur,
      enabled,
      keys
    });
  }
}
function subscribe(subscriber) {
  listeners.add(subscriber);
  return function () {
    listeners.delete(subscriber);
  };
}
function useMedia(componentContext, debug) {
  "use no memo";

  var componentState = componentContext ? States.get(componentContext) : null;
  var internalRef = (0, import_react.useRef)(null);
  if (!internalRef.current) {
    internalRef.current = {
      keys: /* @__PURE__ */new Set(),
      lastState: (0, import_mediaState.getMedia)()
    };
  }
  if (internalRef.current.pendingState) {
    internalRef.current.lastState = internalRef.current.pendingState;
    internalRef.current.pendingState = void 0;
  }
  var {
    keys
  } = internalRef.current;
  if (keys.size) {
    keys.clear();
  }
  var state = (0, import_react.useSyncExternalStore)(subscribe, function () {
    var curKeys = (componentState === null || componentState === void 0 ? void 0 : componentState.keys) || keys;
    var {
      lastState,
      pendingState
    } = internalRef.current;
    if (!curKeys.size) {
      return lastState;
    }
    var ms = (0, import_mediaState.getMedia)();
    var _iteratorNormalCompletion = true,
      _didIteratorError = false,
      _iteratorError = void 0;
    try {
      for (var _iterator = curKeys[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var key = _step.value;
        if (ms[key] !== (pendingState || lastState)[key]) {
          if (process.env.NODE_ENV === "development" && debug) {
            console.warn(`useMedia() \u270D\uFE0F`, key, lastState[key], "=>", ms[key]);
          }
          if (componentContext === null || componentContext === void 0 ? void 0 : componentContext.mediaEmit) {
            componentContext.mediaEmit(ms);
            internalRef.current.pendingState = ms;
            return lastState;
          }
          internalRef.current.lastState = ms;
          return ms;
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
    return lastState;
  }, getServerSnapshot);
  return new Proxy(state, {
    get(_, key) {
      if (!disableMediaTouch && typeof key === "string") {
        keys.add(key);
      }
      return Reflect.get(state, key);
    }
  });
}
var getServerSnapshot = function () {
  return initState;
};
var disableMediaTouch = false;
function _disableMediaTouch(val) {
  disableMediaTouch = val;
}
function getMediaState(mediaGroups, layout) {
  disableMediaTouch = true;
  var res;
  try {
    res = Object.fromEntries([...mediaGroups].map(function (mediaKey) {
      return [mediaKey, mediaKeyMatch(mediaKey, layout)];
    }));
  } finally {
    disableMediaTouch = false;
  }
  return res;
}
var getMediaImportanceIfMoreImportant = function (mediaKey, key, styleState, isSizeMedia) {
  var importance = isSizeMedia ? getMediaKeyImportance(mediaKey) : import_pseudoDescriptors.defaultMediaImportance;
  var usedKeys = styleState.usedKeys;
  return !usedKeys[key] || importance > usedKeys[key] ? importance : null;
};
var cachedMediaKeyToQuery = {};
function mediaKeyToQuery(key) {
  return cachedMediaKeyToQuery[key] || (cachedMediaKeyToQuery[key] = (0, import_mediaObjectToString.mediaObjectToString)(import_mediaState.mediaQueryConfig[key]));
}
function mediaKeyMatch(key, dimensions) {
  var mediaQueries = import_mediaState.mediaQueryConfig[key];
  var result = Object.keys(mediaQueries).every(function (query) {
    var expectedVal = +mediaQueries[query];
    var isMax = query.startsWith("max");
    var isWidth = query.endsWith("Width");
    var givenVal = dimensions[isWidth ? "width" : "height"];
    return isMax ? givenVal < expectedVal : givenVal > expectedVal;
  });
  return result;
}
