// @tamagui/web/dist/cjs/hooks/useProps.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useProps_exports = {};
__export(useProps_exports, {
  useProps: () => useProps,
  usePropsAndStyle: () => usePropsAndStyle,
  useStyle: () => useStyle
});
module.exports = __toCommonJS(useProps_exports);
var import_constants = require("@tamagui/constants");
var import_react = __toESM(require("react"), 1);
var import_config = require("../config.native.js");
var import_ComponentContext = require("../contexts/ComponentContext.native.js");
var import_GroupContext = require("../contexts/GroupContext.native.js");
var import_getSplitStyles = require("../helpers/getSplitStyles.native.js");
var import_subscribeToContextGroup = require("../helpers/subscribeToContextGroup.native.js");
var import_View = require("../views/View.native.js");
var import_useComponentState = require("./useComponentState.native.js");
var import_mediaState = require("../helpers/mediaState.native.js");
var import_useMedia = require("./useMedia.native.js");
var import_useTheme = require("./useTheme.native.js");
function useProps(props, opts) {
  var [propsOut, styleOut] = usePropsAndStyle(props, {
    ...opts,
    noExpand: true,
    noNormalize: true,
    resolveValues: "none"
  });
  return {
    ...propsOut,
    ...styleOut
  };
}
function useStyle(props, opts) {
  return usePropsAndStyle(props, opts)[1] || {};
}
function usePropsAndStyle(props, opts) {
  var _opts_forComponent;
  var _opts_forComponent_staticConfig;
  var staticConfig = (_opts_forComponent_staticConfig = opts === null || opts === void 0 ? void 0 : (_opts_forComponent = opts.forComponent) === null || _opts_forComponent === void 0 ? void 0 : _opts_forComponent.staticConfig) !== null && _opts_forComponent_staticConfig !== void 0 ? _opts_forComponent_staticConfig : import_View.View.staticConfig;
  var [theme, themeState] = (0, import_useTheme.useThemeWithState)({
    componentName: staticConfig.componentName,
    name: "theme" in props ? props.theme : void 0,
    needsUpdate() {
      return true;
    }
  });
  var componentContext = import_react.default.useContext(import_ComponentContext.ComponentContext);
  var groupContext = import_react.default.useContext(import_GroupContext.GroupContext);
  var {
    state,
    disabled,
    setStateShallow
  } = (0, import_useComponentState.useComponentState)(props, componentContext.animationDriver, staticConfig, (0, import_config.getConfig)());
  var mediaStateNow = (opts === null || opts === void 0 ? void 0 : opts.noMedia) ?
  // not safe to use mediaState but really marginal to hit this
  import_mediaState.mediaState : (0, import_useMedia.useMedia)();
  var splitStyles = (0, import_getSplitStyles.useSplitStyles)(props, staticConfig, theme, (themeState === null || themeState === void 0 ? void 0 : themeState.name) || "", state, {
    isAnimated: false,
    mediaState: mediaStateNow,
    noSkip: true,
    noMergeStyle: true,
    noClass: true,
    resolveValues: "auto",
    ...opts
  }, null, componentContext, groupContext);
  var {
    mediaGroups,
    pseudoGroups
  } = splitStyles || {};
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    if (disabled) {
      return;
    }
    if (state.unmounted) {
      setStateShallow({
        unmounted: false
      });
      return;
    }
    if (groupContext) {
      return (0, import_subscribeToContextGroup.subscribeToContextGroup)({
        groupContext,
        setStateShallow,
        mediaGroups,
        pseudoGroups
      });
    }
  }, [disabled, groupContext, pseudoGroups ? Object.keys([...pseudoGroups]).join("") : 0, mediaGroups ? Object.keys([...mediaGroups]).join("") : 0]);
  return [(splitStyles === null || splitStyles === void 0 ? void 0 : splitStyles.viewProps) || {}, (splitStyles === null || splitStyles === void 0 ? void 0 : splitStyles.style) || {}, theme, import_mediaState.mediaState];
}
