// @tamagui/button/dist/jsx/Button.mjs
import { getFontSize } from "@tamagui/font-size";
import { getButtonSized } from "@tamagui/get-button-sized";
import { getIcon, useCurrentColor } from "@tamagui/helpers-tamagui";
import { ButtonNestingContext, getElevation, themeableVariants } from "@tamagui/stacks";
import { SizableText, wrapChildrenInText } from "@tamagui/text";
import { createStyledContext, getTokenValue, styled, useProps, View, withStaticProperties } from "@tamagui/web";
import { useContext } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
const context = createStyledContext({
  size: void 0,
  variant: void 0,
  color: void 0,
  elevation: void 0
});
const Frame = styled(View, {
  context,
  name: "Button",
  role: "button",
  render: /* @__PURE__ */jsx("button", {
    type: "button"
  }),
  tabIndex: 0,
  variants: {
    unstyled: {
      false: {
        size: "$true",
        justifyContent: "center",
        alignItems: "center",
        flexWrap: "nowrap",
        flexDirection: "row",
        cursor: "pointer",
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "transparent",
        hoverStyle: {
          backgroundColor: "$backgroundHover",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress",
          borderColor: "$borderColorHover"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineStyle: "solid",
          outlineWidth: 2
        }
      }
    },
    variant: {
      outlined: process.env.TAMAGUI_HEADLESS === "1" ? {} : {
        backgroundColor: "transparent",
        borderWidth: 1,
        borderColor: "$borderColor",
        hoverStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorPress"
        }
      }
    },
    circular: themeableVariants.circular,
    chromeless: themeableVariants.chromeless,
    size: {
      "...size": (val, extras) => {
        const buttonStyle = getButtonSized(val, extras);
        const gap = getTokenValue(val);
        return {
          ...buttonStyle,
          gap
        };
      },
      ":number": (val, extras) => {
        const buttonStyle = getButtonSized(val, extras);
        const gap = val * 0.4;
        return {
          ...buttonStyle,
          gap
        };
      }
    },
    elevation: {
      "...size": getElevation,
      ":number": getElevation
    },
    disabled: {
      true: {
        pointerEvents: "none",
        // @ts-ignore
        "aria-disabled": true
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const Text = styled(SizableText, {
  context,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        cursor: "pointer",
        // flexGrow 1 leads to inconsistent native style where text pushes to start of view
        flexGrow: 0,
        flexShrink: 1,
        ellipsis: true,
        color: "$color"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const Icon = props => {
  const {
    children,
    scaleIcon = 1,
    size
  } = props;
  const styledContext = context.useStyledContext();
  if (!styledContext) {
    throw new Error("Button.Icon must be used within a Button");
  }
  const sizeToken = size ?? styledContext.size;
  const iconColor = useCurrentColor(styledContext.color);
  const iconSize = (typeof sizeToken === "number" ? sizeToken * 0.5 : getFontSize(sizeToken)) * scaleIcon;
  return getIcon(children, {
    size: iconSize,
    color: iconColor
  });
};
const ButtonContext = createStyledContext({
  size: void 0,
  variant: void 0,
  color: void 0
});
const ButtonComponent = Frame.styleable((propsIn, ref) => {
  const isNested = useContext(ButtonNestingContext);
  const processedProps = useProps(propsIn, {
    noNormalize: true,
    noExpand: true
  });
  const {
    children,
    iconSize,
    icon,
    iconAfter,
    scaleIcon = 1,
    ...props
  } = processedProps;
  const size = propsIn.size || (propsIn.unstyled ? void 0 : "$true");
  const styledContext = context.useStyledContext();
  const iconColor = useCurrentColor(styledContext?.color);
  const finalSize = iconSize ?? size ?? styledContext?.size;
  const iconSizeNumber = (typeof finalSize === "number" ? finalSize * 0.5 : getFontSize(finalSize)) * scaleIcon;
  const [themedIcon, themedIconAfter] = [icon, iconAfter].map(icon2 => {
    if (!icon2) return null;
    return getIcon(icon2, {
      size: iconSizeNumber,
      color: iconColor
      // No marginLeft or marginRight needed - spacing is handled by the gap property in Frame's size variants
    });
  });
  const wrappedChildren = wrapChildrenInText(Text, {
    children
  }, {
    unstyled: process.env.TAMAGUI_HEADLESS === "1",
    size: finalSize ?? styledContext?.size
  });
  return /* @__PURE__ */jsx(ButtonNestingContext.Provider, {
    value: true,
    children: /* @__PURE__ */jsxs(Frame, {
      ref,
      ...props,
      ...(isNested && {
        render: "span"
      }),
      ...(props.circular && !propsIn.size && {
        size
      }),
      tabIndex: 0,
      children: [themedIcon, wrappedChildren, themedIconAfter]
    })
  });
});
const Button = withStaticProperties(ButtonComponent, {
  Apply: context.Provider,
  Frame,
  Text,
  Icon
});
export { Button, ButtonContext };
