// @tamagui/switch-headless/dist/esm/useSwitch.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { composeEventHandlers } from "@tamagui/helpers";
import { useLabelContext } from "@tamagui/label";
import { usePrevious } from "@tamagui/use-previous";
import * as React from "react";
import { jsx } from "react/jsx-runtime";
function getState(checked) {
  return checked ? "checked" : "unchecked";
}
const BubbleInput = props => {
  const {
    control,
    checked,
    bubbles = true,
    ...inputProps
  } = props;
  const ref = React.useRef(null);
  const prevChecked = usePrevious(checked);
  React.useEffect(() => {
    const input = ref.current;
    const inputProto = window.HTMLInputElement.prototype;
    const setChecked = Object.getOwnPropertyDescriptor(inputProto, "checked").set;
    if (prevChecked !== checked && setChecked) {
      const event = new Event("click", {
        bubbles
      });
      setChecked.call(input, checked);
      input.dispatchEvent(event);
    }
  }, [prevChecked, checked, bubbles]);
  return /* @__PURE__ */jsx("input", {
    type: "checkbox",
    "aria-hidden": true,
    defaultChecked: checked,
    ...inputProps,
    tabIndex: -1,
    ref,
    style: {
      ...props.style,
      position: "absolute",
      pointerEvents: "none",
      opacity: 0,
      margin: 0
    }
  });
};
function useSwitch(props, [checked, setChecked], ref) {
  {
    const {
      disabled,
      name,
      value,
      required
    } = props;
    const hasConsumerStoppedPropagationRef = React.useRef(false);
    const [button, setButton] = React.useState(null);
    const composedRefs = useComposedRefs(ref, setButton);
    const isFormControl = isWeb ? button ? Boolean(button.closest("form")) : true : false;
    const labelId = useLabelContext(button);
    const ariaLabelledBy = props["aria-labelledby"] || props.labeledBy || labelId;
    return {
      switchProps: {
        role: "switch",
        "aria-checked": checked,
        ...(isWeb ? {
          tabIndex: disabled ? void 0 : 0,
          "data-state": getState(checked),
          "data-disabled": disabled ? "" : void 0,
          disabled
        } : {}),
        "aria-labelledby": ariaLabelledBy,
        onPress: composeEventHandlers(props.onPress, event => {
          setChecked(prevChecked => !prevChecked);
          if (isWeb && isFormControl) {
            hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
            if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
          }
        })
      },
      switchRef: composedRefs,
      bubbleInput: isWeb && isFormControl ? /* @__PURE__ */jsx(BubbleInput, {
        control: button,
        bubbles: !hasConsumerStoppedPropagationRef.current,
        name,
        value,
        checked,
        required,
        disabled,
        style: {
          transform: "translateX(-100%)"
        }
      }) : null
    };
  }
}
export { useSwitch };
