// @tamagui/input/dist/esm/v1/Input.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import React from "react";
import { styled } from "@tamagui/core";
import { registerFocusable } from "@tamagui/focusable";
import { useNativeInputRef } from "@tamagui/element";
import { TextInput } from "react-native";
import { styledBody } from "../shared.native.js";
var StyledInput = styled(TextInput, styledBody[0], styledBody[1]);
var Input = StyledInput.styleable(function (inProps, forwardedRef) {
  var {
    // some of destructed props are just to avoid passing them to ...rest because they are not in native.
    type,
    //@ts-ignore
    dirname,
    max,
    min,
    minLength,
    multiple,
    name,
    required,
    step,
    disabled,
    id,
    // @ts-ignore
    caretColor,
    onChange,
    onInput,
    rows,
    enterKeyHint,
    returnKeyType,
    onKeyDown,
    inputMode,
    render,
    ...rest
  } = inProps;
  var {
    ref,
    composedRef
  } = useNativeInputRef(forwardedRef);
  var secureTextEntry = false;
  var cursorColor = caretColor;
  var _returnKeyType = returnKeyType;
  var _enterKeyHint = enterKeyHint;
  if (enterKeyHint === "go") {
    _returnKeyType = "go";
    _enterKeyHint = void 0;
  }
  var _inputMode = inputMode;
  if (type === "email") {
    _inputMode = "email";
  } else if (type === "tel") {
    _inputMode = "tel";
  } else if (type === "search") {
    _inputMode = "search";
  } else if (type === "url") {
    _inputMode = "url";
  } else if (type === "password") {
    secureTextEntry = true;
    _inputMode = "text";
  } else if (type === "number") {
    _inputMode = "numeric";
  } else {
    _inputMode = "text";
  }
  var showSoftInputOnFocus = true;
  if (inputMode === "none") {
    showSoftInputOnFocus = false;
  }
  var finalProps = {
    ...rest,
    inputMode: _inputMode,
    showSoftInputOnFocus,
    disabled,
    id,
    cursorColor,
    enterKeyHint: _enterKeyHint,
    returnKeyType: _returnKeyType,
    secureTextEntry,
    numberOfLines: rows || rest.numberOfLines
  };
  if (render === "textarea") {
    finalProps.multiline = true;
  }
  if (onKeyDown) {
    finalProps.onKeyPress = function (e) {
      var {
        key
      } = e.nativeEvent;
      if (key === "Backspace" || render === "textarea" && key === "Enter" || key.length === 1) {
        onKeyDown({
          key,
          type: "keydown"
        });
      }
    };
    finalProps.onSubmitEditing = function (e) {
      onKeyDown({
        key: "Enter",
        type: "keydown"
      });
    };
  }
  if (onChange || onInput) {
    finalProps.onChange = function (e) {
      var {
        text
      } = e.nativeEvent;
      if (onChange) {
        onChange({
          target: {
            value: text
          },
          type: "change"
        });
      }
      if (onInput != null) {
        onInput({
          target: {
            value: text
          },
          type: "input"
        });
      }
    };
  }
  React.useEffect(function () {
    if (!id) return;
    if (disabled) return;
    return registerFocusable(id, {
      focusAndSelect: function () {
        var _ref_current;
        (_ref_current = ref.current) === null || _ref_current === void 0 ? void 0 : _ref_current.focus();
      },
      focus: function () {}
    });
  }, [id, disabled]);
  return /* @__PURE__ */_jsx(StyledInput, {
    onChange: function (e) {},
    ref: composedRef,
    ...finalProps
  });
});
export { Input };
