// @tamagui/use-debounce/dist/cjs/index.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  debounce: () => debounce,
  useDebounce: () => useDebounce,
  useDebounceValue: () => useDebounceValue
});
module.exports = __toCommonJS(index_exports);
var React = __toESM(require("react"), 1);
function debounce(func, wait, leading) {
  var timeout;
  var isCancelled = false;
  function debounced() {
    var _this = this;
    isCancelled = false;
    var args = arguments;
    if (leading && !timeout) {
      func.apply(this, args);
    }
    clearTimeout(timeout);
    timeout = setTimeout(function () {
      timeout = null;
      if (!(leading || isCancelled)) {
        func.apply(_this, args);
      }
      isCancelled = false;
    }, wait);
  }
  debounced.cancel = function () {
    isCancelled = true;
  };
  return debounced;
}
var defaultOpts = {
  leading: false
};
function useDebounce(fn, wait) {
  var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : defaultOpts,
    mountArgs = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : [fn];
  var dbEffect = React.useRef(null);
  React.useEffect(function () {
    return function () {
      var _dbEffect_current;
      (_dbEffect_current = dbEffect.current) === null || _dbEffect_current === void 0 ? void 0 : _dbEffect_current.cancel();
    };
  }, []);
  return React.useMemo(function () {
    dbEffect.current = debounce(fn, wait, options.leading);
    return dbEffect.current;
  }, [options.leading, ...mountArgs]);
}
function useDebounceValue(val) {
  var amt = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 0;
  var [state, setState] = React.useState(val);
  React.useEffect(function () {
    var tm = setTimeout(function () {
      setState(function (prev) {
        if (prev === val) return prev;
        return val;
      });
    }, amt);
    return function () {
      clearTimeout(tm);
    };
  }, [val]);
  return state;
}
