// @tamagui/checkbox-headless/dist/esm/useCheckbox.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { composeEventHandlers } from "@tamagui/helpers";
import { useLabelContext } from "@tamagui/label";
import React, { useMemo } from "react";
import { BubbleInput } from "./BubbleInput.native.js";
import { getState, isIndeterminate } from "./utils.native.js";
function useCheckbox(props, param, ref) {
  var [checked, setChecked] = param;
  var {
    labelledBy: ariaLabelledby,
    name,
    required,
    disabled,
    value = "on",
    onCheckedChange,
    ...checkboxProps
  } = props;
  var [button, setButton] = React.useState(null);
  var composedRefs = useComposedRefs(ref, setButton);
  var hasConsumerStoppedPropagationRef = React.useRef(false);
  var isFormControl = isWeb ? button ? Boolean(button.closest("form")) : true : false;
  var labelId = useLabelContext(button);
  var labelledBy = ariaLabelledby || labelId;
  var parentKeyDown = props.onKeyDown;
  var handleKeyDown = useMemo(function () {
    return composeEventHandlers(parentKeyDown, function (event) {
      if (event.key === "Enter") event.preventDefault();
    });
  }, [parentKeyDown]);
  var handlePress = useMemo(function () {
    return composeEventHandlers(props.onPress, function (event) {
      setChecked(function (prevChecked) {
        return isIndeterminate(prevChecked) ? true : !prevChecked;
      });
      if (isFormControl && "isPropagationStopped" in event) {
        hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
        if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
      }
    });
  }, [isFormControl]);
  return {
    bubbleInput: isWeb && isFormControl ? /* @__PURE__ */_jsx(BubbleInput, {
      isHidden: true,
      control: button,
      bubbles: !hasConsumerStoppedPropagationRef.current,
      name,
      value,
      checked,
      required,
      disabled
    }) : null,
    checkboxRef: composedRefs,
    checkboxProps: {
      role: "checkbox",
      "aria-labelledby": labelledBy,
      "aria-checked": isIndeterminate(checked) ? "mixed" : checked,
      ...checkboxProps,
      ...(isWeb && {
        type: "button",
        value,
        "data-state": getState(checked),
        "data-disabled": disabled ? "" : void 0,
        disabled,
        onKeyDown: disabled ? void 0 : handleKeyDown
      }),
      onPress: disabled ? void 0 : handlePress
    }
  };
}
export { useCheckbox };
