// @tamagui/checkbox-headless/dist/esm/useCheckbox.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { composeEventHandlers } from "@tamagui/helpers";
import { useLabelContext } from "@tamagui/label";
import React, { useMemo } from "react";
import { BubbleInput } from "./BubbleInput.mjs";
import { getState, isIndeterminate } from "./utils.mjs";
import { jsx } from "react/jsx-runtime";
function useCheckbox(props, [checked, setChecked], ref) {
  const {
    labelledBy: ariaLabelledby,
    name,
    required,
    disabled,
    value = "on",
    onCheckedChange,
    ...checkboxProps
  } = props;
  const [button, setButton] = React.useState(null);
  const composedRefs = useComposedRefs(ref, setButton);
  const hasConsumerStoppedPropagationRef = React.useRef(false);
  const isFormControl = isWeb ? button ? Boolean(button.closest("form")) : true : false;
  const labelId = useLabelContext(button);
  const labelledBy = ariaLabelledby || labelId;
  const parentKeyDown = props.onKeyDown;
  const handleKeyDown = useMemo(() => composeEventHandlers(parentKeyDown, event => {
    if (event.key === "Enter") event.preventDefault();
  }), [parentKeyDown]);
  const handlePress = useMemo(() => composeEventHandlers(props.onPress, event => {
    setChecked(prevChecked => isIndeterminate(prevChecked) ? true : !prevChecked);
    if (isFormControl && "isPropagationStopped" in event) {
      hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
      if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
    }
  }), [isFormControl]);
  return {
    bubbleInput: isWeb && isFormControl ? /* @__PURE__ */jsx(BubbleInput, {
      isHidden: true,
      control: button,
      bubbles: !hasConsumerStoppedPropagationRef.current,
      name,
      value,
      checked,
      required,
      disabled
    }) : null,
    checkboxRef: composedRefs,
    checkboxProps: {
      role: "checkbox",
      "aria-labelledby": labelledBy,
      "aria-checked": isIndeterminate(checked) ? "mixed" : checked,
      ...checkboxProps,
      ...(isWeb && {
        type: "button",
        value,
        "data-state": getState(checked),
        "data-disabled": disabled ? "" : void 0,
        disabled,
        onKeyDown: disabled ? void 0 : handleKeyDown
      }),
      onPress: disabled ? void 0 : handlePress
    }
  };
}
export { useCheckbox };
