// @tamagui/text/dist/cjs/SizableText.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SizableText_exports = {};
__export(SizableText_exports, {
  SizableText: () => SizableText
});
module.exports = __toCommonJS(SizableText_exports);
var import_get_font_sized = require("@tamagui/get-font-sized");
var import_web = require("@tamagui/web");
const SizableText = (0, import_web.styled)(import_web.Text, {
  name: "SizableText",
  fontFamily: "$body",
  variants: {
    unstyled: {
      false: {
        size: "$true",
        color: "$color"
      }
    },
    size: import_get_font_sized.getFontSized
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
SizableText.staticConfig.variants.fontFamily = {
  "...": (val, extras) => {
    if (val === "inherit") {
      return {
        fontFamily: "inherit"
      };
    }
    const sizeProp = extras.props["size"];
    const fontSizeProp = extras.props["fontSize"];
    const size = sizeProp === "$true" && fontSizeProp ? fontSizeProp : extras.props["size"] || "$true";
    return (0, import_get_font_sized.getFontSized)(size, extras);
  }
};
