// @tamagui/web/dist/esm/hooks/useMedia.mjs
import { isServer, isWeb } from "@tamagui/constants";
import { useRef, useSyncExternalStore } from "react";
import { getSetting } from "../config.mjs";
import { resetMediaStyleCache } from "../helpers/createMediaStyle.mjs";
import { matchMedia } from "../helpers/matchMedia.mjs";
import { mediaObjectToString } from "../helpers/mediaObjectToString.mjs";
import { getMedia, mediaKeys, mediaQueryConfig, setMediaState } from "../helpers/mediaState.mjs";
import { defaultMediaImportance } from "../helpers/pseudoDescriptors.mjs";
const mediaKeyRegex = /\$(platform|theme|group)-/;
const isMediaKey = key => {
  if (key[0] !== "$") return false;
  if (mediaKeys.has(key)) return true;
  if (mediaKeyRegex.test(key)) return true;
  return false;
};
const getMediaKey = key => {
  if (key[0] !== "$") return false;
  if (mediaKeys.has(key)) return true;
  const match = key.match(mediaKeyRegex);
  if (match) return match[1];
  return false;
};
let initState;
let mediaKeysOrdered;
const getMediaKeyImportance = key => {
  if (process.env.NODE_ENV === "development" && key[0] === "$") {
    throw new Error("use short key");
  }
  return mediaKeysOrdered.indexOf(key) + 100;
};
const dispose = /* @__PURE__ */new Set();
let mediaVersion = 0;
const configureMedia = config => {
  const {
    media
  } = config;
  const mediaQueryDefaultActive = getSetting("mediaQueryDefaultActive");
  if (!media) return;
  mediaVersion++;
  resetMediaStyleCache();
  for (const key in media) {
    getMedia()[key] = mediaQueryDefaultActive?.[key] || false;
    mediaKeys.add(`$${key}`);
  }
  Object.assign(mediaQueryConfig, media);
  initState = {
    ...getMedia()
  };
  mediaKeysOrdered = Object.keys(media);
  setupMediaListeners();
};
function unlisten() {
  dispose.forEach(cb => cb());
  dispose.clear();
}
let setupVersion = -1;
function setupMediaListeners() {
  if (isWeb && isServer) return;
  if (process.env.IS_STATIC) return;
  if (setupVersion === mediaVersion) return;
  setupVersion = mediaVersion;
  unlisten();
  for (const key in mediaQueryConfig) {
    let update = function () {
      const next = !!getMatch().matches;
      if (next === getMedia()[key]) return;
      setMediaState({
        ...getMedia(),
        [key]: next
      });
      updateMediaListeners();
    };
    const str = mediaObjectToString(mediaQueryConfig[key]);
    const getMatch = () => matchMedia(str);
    const match = getMatch();
    if (!match) {
      throw new Error("\u26A0\uFE0F No match");
    }
    match.addListener(update);
    dispose.add(() => {
      match.removeListener(update);
    });
    update();
  }
}
const listeners = /* @__PURE__ */new Set();
function updateMediaListeners() {
  listeners.forEach(cb => cb(getMedia()));
}
const States = /* @__PURE__ */new WeakMap();
function setMediaShouldUpdate(ref, enabled, keys) {
  const cur = States.get(ref);
  if (!cur || cur.enabled !== enabled || keys) {
    States.set(ref, {
      ...cur,
      enabled,
      keys
    });
  }
}
function subscribe(subscriber) {
  listeners.add(subscriber);
  return () => {
    listeners.delete(subscriber);
  };
}
function useMedia(componentContext, debug) {
  "use no memo";

  const componentState = componentContext ? States.get(componentContext) : null;
  const internalRef = useRef(null);
  if (!internalRef.current) {
    internalRef.current = {
      keys: /* @__PURE__ */new Set(),
      lastState: getMedia()
    };
  }
  if (internalRef.current.pendingState) {
    internalRef.current.lastState = internalRef.current.pendingState;
    internalRef.current.pendingState = void 0;
  }
  const {
    keys
  } = internalRef.current;
  if (keys.size) {
    keys.clear();
  }
  const state = useSyncExternalStore(subscribe, () => {
    const curKeys = componentState?.keys || keys;
    const {
      lastState,
      pendingState
    } = internalRef.current;
    if (!curKeys.size) {
      return lastState;
    }
    const ms = getMedia();
    for (const key of curKeys) {
      if (ms[key] !== (pendingState || lastState)[key]) {
        if (process.env.NODE_ENV === "development" && debug) {
          console.warn(`useMedia() \u270D\uFE0F`, key, lastState[key], "=>", ms[key]);
        }
        if (componentContext?.mediaEmit) {
          componentContext.mediaEmit(ms);
          internalRef.current.pendingState = ms;
          return lastState;
        }
        internalRef.current.lastState = ms;
        return ms;
      }
    }
    return lastState;
  }, getServerSnapshot);
  return new Proxy(state, {
    get(_, key) {
      if (!disableMediaTouch && typeof key === "string") {
        keys.add(key);
      }
      return Reflect.get(state, key);
    }
  });
}
const getServerSnapshot = () => initState;
let disableMediaTouch = false;
function _disableMediaTouch(val) {
  disableMediaTouch = val;
}
function getMediaState(mediaGroups, layout) {
  disableMediaTouch = true;
  let res;
  try {
    res = Object.fromEntries([...mediaGroups].map(mediaKey => {
      return [mediaKey, mediaKeyMatch(mediaKey, layout)];
    }));
  } finally {
    disableMediaTouch = false;
  }
  return res;
}
const getMediaImportanceIfMoreImportant = (mediaKey, key, styleState, isSizeMedia) => {
  const importance = isSizeMedia ? getMediaKeyImportance(mediaKey) : defaultMediaImportance;
  const usedKeys = styleState.usedKeys;
  return !usedKeys[key] || importance > usedKeys[key] ? importance : null;
};
const cachedMediaKeyToQuery = {};
function mediaKeyToQuery(key) {
  return cachedMediaKeyToQuery[key] || (cachedMediaKeyToQuery[key] = mediaObjectToString(mediaQueryConfig[key]));
}
function mediaKeyMatch(key, dimensions) {
  const mediaQueries = mediaQueryConfig[key];
  const result = Object.keys(mediaQueries).every(query => {
    const expectedVal = +mediaQueries[query];
    const isMax = query.startsWith("max");
    const isWidth = query.endsWith("Width");
    const givenVal = dimensions[isWidth ? "width" : "height"];
    return isMax ? givenVal < expectedVal : givenVal > expectedVal;
  });
  return result;
}
export { _disableMediaTouch, configureMedia, getMediaImportanceIfMoreImportant, getMediaKey, getMediaKeyImportance, getMediaState, isMediaKey, mediaKeyMatch, mediaKeyToQuery, setMediaShouldUpdate, setupMediaListeners, updateMediaListeners, useMedia };
