// @tamagui/web/dist/esm/hooks/useProps.native.js
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import React from "react";
import { getConfig } from "../config.native.js";
import { ComponentContext } from "../contexts/ComponentContext.native.js";
import { GroupContext } from "../contexts/GroupContext.native.js";
import { useSplitStyles } from "../helpers/getSplitStyles.native.js";
import { subscribeToContextGroup } from "../helpers/subscribeToContextGroup.native.js";
import { View } from "../views/View.native.js";
import { useComponentState } from "./useComponentState.native.js";
import { mediaState } from "../helpers/mediaState.native.js";
import { useMedia } from "./useMedia.native.js";
import { useThemeWithState } from "./useTheme.native.js";
function useProps(props, opts) {
  var [propsOut, styleOut] = usePropsAndStyle(props, {
    ...opts,
    noExpand: true,
    noNormalize: true,
    resolveValues: "none"
  });
  return {
    ...propsOut,
    ...styleOut
  };
}
function useStyle(props, opts) {
  return usePropsAndStyle(props, opts)[1] || {};
}
function usePropsAndStyle(props, opts) {
  var _opts_forComponent;
  var _opts_forComponent_staticConfig;
  var staticConfig = (_opts_forComponent_staticConfig = opts === null || opts === void 0 ? void 0 : (_opts_forComponent = opts.forComponent) === null || _opts_forComponent === void 0 ? void 0 : _opts_forComponent.staticConfig) !== null && _opts_forComponent_staticConfig !== void 0 ? _opts_forComponent_staticConfig : View.staticConfig;
  var [theme, themeState] = useThemeWithState({
    componentName: staticConfig.componentName,
    name: "theme" in props ? props.theme : void 0,
    needsUpdate() {
      return true;
    }
  });
  var componentContext = React.useContext(ComponentContext);
  var groupContext = React.useContext(GroupContext);
  var {
    state,
    disabled,
    setStateShallow
  } = useComponentState(props, componentContext.animationDriver, staticConfig, getConfig());
  var mediaStateNow = (opts === null || opts === void 0 ? void 0 : opts.noMedia) ?
  // not safe to use mediaState but really marginal to hit this
  mediaState : useMedia();
  var splitStyles = useSplitStyles(props, staticConfig, theme, (themeState === null || themeState === void 0 ? void 0 : themeState.name) || "", state, {
    isAnimated: false,
    mediaState: mediaStateNow,
    noSkip: true,
    noMergeStyle: true,
    noClass: true,
    resolveValues: "auto",
    ...opts
  }, null, componentContext, groupContext);
  var {
    mediaGroups,
    pseudoGroups
  } = splitStyles || {};
  useIsomorphicLayoutEffect(function () {
    if (disabled) {
      return;
    }
    if (state.unmounted) {
      setStateShallow({
        unmounted: false
      });
      return;
    }
    if (groupContext) {
      return subscribeToContextGroup({
        groupContext,
        setStateShallow,
        mediaGroups,
        pseudoGroups
      });
    }
  }, [disabled, groupContext, pseudoGroups ? Object.keys([...pseudoGroups]).join("") : 0, mediaGroups ? Object.keys([...mediaGroups]).join("") : 0]);
  return [(splitStyles === null || splitStyles === void 0 ? void 0 : splitStyles.viewProps) || {}, (splitStyles === null || splitStyles === void 0 ? void 0 : splitStyles.style) || {}, theme, mediaState];
}
export { useProps, usePropsAndStyle, useStyle };
