// @tamagui/web/dist/esm/hooks/useProps.mjs
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import React from "react";
import { getConfig } from "../config.mjs";
import { ComponentContext } from "../contexts/ComponentContext.mjs";
import { GroupContext } from "../contexts/GroupContext.mjs";
import { useSplitStyles } from "../helpers/getSplitStyles.mjs";
import { subscribeToContextGroup } from "../helpers/subscribeToContextGroup.mjs";
import { View } from "../views/View.mjs";
import { useComponentState } from "./useComponentState.mjs";
import { mediaState } from "../helpers/mediaState.mjs";
import { useMedia } from "./useMedia.mjs";
import { useThemeWithState } from "./useTheme.mjs";
function useProps(props, opts) {
  const [propsOut, styleOut] = usePropsAndStyle(props, {
    ...opts,
    noExpand: true,
    noNormalize: true,
    resolveValues: "none"
  });
  return {
    ...propsOut,
    ...styleOut
  };
}
function useStyle(props, opts) {
  return usePropsAndStyle(props, opts)[1] || {};
}
function usePropsAndStyle(props, opts) {
  const staticConfig = opts?.forComponent?.staticConfig ?? View.staticConfig;
  const [theme, themeState] = useThemeWithState({
    componentName: staticConfig.componentName,
    name: "theme" in props ? props.theme : void 0,
    needsUpdate() {
      return true;
    }
  });
  const componentContext = React.useContext(ComponentContext);
  const groupContext = React.useContext(GroupContext);
  const {
    state,
    disabled,
    setStateShallow
  } = useComponentState(props, componentContext.animationDriver, staticConfig, getConfig());
  const mediaStateNow = opts?.noMedia ?
  // not safe to use mediaState but really marginal to hit this
  mediaState : useMedia();
  const splitStyles = useSplitStyles(props, staticConfig, theme, themeState?.name || "", state, {
    isAnimated: false,
    mediaState: mediaStateNow,
    noSkip: true,
    noMergeStyle: true,
    noClass: true,
    resolveValues: "auto",
    ...opts
  }, null, componentContext, groupContext);
  const {
    mediaGroups,
    pseudoGroups
  } = splitStyles || {};
  useIsomorphicLayoutEffect(() => {
    if (disabled) {
      return;
    }
    if (state.unmounted) {
      setStateShallow({
        unmounted: false
      });
      return;
    }
    if (groupContext) {
      return subscribeToContextGroup({
        groupContext,
        setStateShallow,
        mediaGroups,
        pseudoGroups
      });
    }
  }, [disabled, groupContext, pseudoGroups ? Object.keys([...pseudoGroups]).join("") : 0, mediaGroups ? Object.keys([...mediaGroups]).join("") : 0]);
  return [splitStyles?.viewProps || {}, splitStyles?.style || {}, theme, mediaState];
}
export { useProps, usePropsAndStyle, useStyle };
