// @tamagui/web/dist/esm/hooks/useThemeState.native.js
import { isIos } from "@tamagui/constants";
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { createContext, useCallback, useContext, useId, useSyncExternalStore } from "react";
import { getConfig, getSetting } from "../config.native.js";
import { MISSING_THEME_MESSAGE } from "../constants/constants.native.js";
var ThemeStateContext = createContext("");
var allListeners = /* @__PURE__ */new Map();
var listenersByParent = {};
var HasRenderedOnce = /* @__PURE__ */new WeakMap();
var HadTheme = /* @__PURE__ */new WeakMap();
var PendingUpdate = /* @__PURE__ */new Map();
var states = /* @__PURE__ */new Map();
var localStates = /* @__PURE__ */new Map();
var shouldForce = false;
var forceUpdateThemes = function () {
  cacheVersion++;
  shouldForce = true;
  allListeners.forEach(function (cb) {
    return cb();
  });
  shouldForce = false;
};
var getThemeState = function (id) {
  return states.get(id);
};
var cacheVersion = 0;
var themeNameCache = /* @__PURE__ */new Map();
var themeNameCacheVer = -1;
var themes = null;
var rootThemeState = null;
var getRootThemeState = function () {
  return rootThemeState;
};
var getThemeBaseName = function (name) {
  return name.replace(/^(light|dark)_/, "");
};
var useThemeState = function (props) {
  "use no memo";

  var isRoot = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : false,
    keys = arguments.length > 2 ? arguments[2] : void 0,
    schemeKeys = arguments.length > 3 ? arguments[3] : void 0;
  var {
    disable
  } = props;
  var parentId = useContext(ThemeStateContext);
  if (!parentId && !isRoot) throw new Error(process.env.NODE_ENV === "development" ? `${MISSING_THEME_MESSAGE}

Looked for theme${props.name ? ` "${props.name}"` : ""}${props.componentName ? ` (component: ${props.componentName})` : ""}, but no parent theme context was found (parentId: ${parentId}).` : MISSING_THEME_MESSAGE);
  if (disable) return states.get(parentId) || {
    id: "",
    name: "light",
    theme: getConfig().themes.light
  };
  var id = useId();
  var subscribe = useCallback(function (cb) {
    listenersByParent[parentId] = listenersByParent[parentId] || /* @__PURE__ */new Set();
    listenersByParent[parentId].add(id);
    allListeners.set(id, function () {
      PendingUpdate.set(id, shouldForce ? "force" : true);
      cb();
    });
    return function () {
      allListeners.delete(id);
      listenersByParent[parentId].delete(id);
      localStates.delete(id);
      states.delete(id);
      PendingUpdate.delete(id);
    };
  }, [id, parentId]);
  var propsKey = getPropsKey(props);
  var getSnapshot = function () {
    var _keys_current, _schemeKeys_current, _keys_current1, _props_needsUpdate;
    var local = localStates.get(id);
    var parentState = states.get(parentId);
    if (local && !PendingUpdate.has(id)) {
      if (parentState && local._parentName === parentState.name && local._propsKey === propsKey) return local;
    }
    var isSchemeOnlyChange = isIos && getSetting("fastSchemeChange") && local && parentState && local.scheme !== parentState.scheme && getThemeBaseName(local.name) === getThemeBaseName(parentState.name);
    var _keys_current_size;
    var keysSize = (_keys_current_size = keys === null || keys === void 0 ? void 0 : (_keys_current = keys.current) === null || _keys_current === void 0 ? void 0 : _keys_current.size) !== null && _keys_current_size !== void 0 ? _keys_current_size : 0;
    var _schemeKeys_current_size;
    var allKeysSchemeOptimized = ((_schemeKeys_current_size = schemeKeys === null || schemeKeys === void 0 ? void 0 : (_schemeKeys_current = schemeKeys.current) === null || _schemeKeys_current === void 0 ? void 0 : _schemeKeys_current.size) !== null && _schemeKeys_current_size !== void 0 ? _schemeKeys_current_size : 0) === keysSize && keysSize > 0;
    var canSkipForSchemeChange = isSchemeOnlyChange && allKeysSchemeOptimized;
    var needsUpdate = props.passThrough ? false : isRoot || props.name === "light" || props.name === "dark" || props.name === null ? true : !HasRenderedOnce.get(keys) ? true : canSkipForSchemeChange ? false : (keys === null || keys === void 0 ? void 0 : (_keys_current1 = keys.current) === null || _keys_current1 === void 0 ? void 0 : _keys_current1.size) ? true : (_props_needsUpdate = props.needsUpdate) === null || _props_needsUpdate === void 0 ? void 0 : _props_needsUpdate.call(props);
    var [rerender, next] = getNextState(local, props, propsKey, isRoot, id, parentId, needsUpdate, PendingUpdate.get(id));
    PendingUpdate.delete(id);
    if (!local || rerender) {
      local = {
        ...next
      };
      localStates.set(id, local);
    }
    if (process.env.NODE_ENV === "development" && props.debug === "verbose") {
      console.groupCollapsed(` ${id} getSnapshot ${rerender}`, local.name, ">", next.name);
      console.info({
        props,
        propsKey,
        isRoot,
        parentId,
        local,
        next,
        needsUpdate,
        isSchemeOnlyChange,
        allKeysSchemeOptimized,
        canSkipForSchemeChange
      });
      console.groupEnd();
    }
    if (next !== local) {
      Object.assign(local, next);
      local.id = id;
    }
    local._parentName = parentState === null || parentState === void 0 ? void 0 : parentState.name;
    local._propsKey = propsKey;
    states.set(id, next);
    return local;
  };
  if (process.env.NODE_ENV === "development" && globalThis.time) globalThis.time`theme-prep-uses`;
  var state = useSyncExternalStore(subscribe, getSnapshot, getSnapshot);
  useIsomorphicLayoutEffect(function () {
    if (!HasRenderedOnce.get(keys)) {
      HasRenderedOnce.set(keys, true);
      return;
    }
    if (!propsKey) {
      if (HadTheme.get(keys)) scheduleUpdate(id);
      HadTheme.set(keys, false);
      return;
    }
    if (process.env.NODE_ENV === "development" && props.debug === "verbose") {
      var _states_get;
      console.warn(` \xB7 useTheme(${id}) scheduleUpdate`, propsKey, (_states_get = states.get(id)) === null || _states_get === void 0 ? void 0 : _states_get.name);
    }
    scheduleUpdate(id);
    HadTheme.set(keys, true);
  }, [keys, propsKey]);
  return state;
};
var getNextState = function (lastState, props, propsKey) {
  var isRoot = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : false,
    id = arguments.length > 4 ? arguments[4] : void 0,
    parentId = arguments.length > 5 ? arguments[5] : void 0,
    needsUpdate = arguments.length > 6 ? arguments[6] : void 0,
    pendingUpdate = arguments.length > 7 ? arguments[7] : void 0;
  var {
    debug
  } = props;
  var parentState = states.get(parentId);
  if (props.passThrough) return [false, lastState || parentState || {
    name: ""
  }];
  if (!themes) themes = getConfig().themes;
  var name = !propsKey && (!lastState || !(lastState === null || lastState === void 0 ? void 0 : lastState.isNew)) ? null : getNewThemeName(parentState === null || parentState === void 0 ? void 0 : parentState.name, props, pendingUpdate === "force" ? true : !!needsUpdate);
  var isSameAsParent = parentState && (!name || name === parentState.name);
  var shouldRerender = Boolean(pendingUpdate === "force" || needsUpdate && (pendingUpdate || (lastState === null || lastState === void 0 ? void 0 : lastState.name) !== (parentState === null || parentState === void 0 ? void 0 : parentState.name)));
  if (process.env.NODE_ENV === "development" && debug === "verbose") {
    var message = ` \xB7 useTheme(${id}) getNextState => ${name} needsUpdate ${needsUpdate} shouldRerender ${shouldRerender}`;
    console.info(message);
  }
  if (isSameAsParent) {
    if (!shouldRerender && lastState && lastState.name === parentState.name) return [false, lastState];
    return [shouldRerender, {
      ...parentState,
      isNew: false
    }];
  }
  if (!name) {
    var next = lastState !== null && lastState !== void 0 ? lastState : parentState;
    if (!next) throw new Error(process.env.NODE_ENV === "development" ? `${MISSING_THEME_MESSAGE}

Looked for theme${props.name ? ` "${props.name}"` : ""}${props.componentName ? ` (component: ${props.componentName})` : ""}, but no theme state was resolved (parentId: ${parentId}, id: ${id}).` : MISSING_THEME_MESSAGE);
    if (shouldRerender) return [true, {
      ...(parentState || lastState)
    }];
    return [false, next];
  }
  var scheme = getScheme(name);
  var isInverse = parentState && scheme !== parentState.scheme;
  var nextState = {
    id,
    name,
    theme: themes[name],
    scheme,
    parentId,
    parentName: parentState === null || parentState === void 0 ? void 0 : parentState.name,
    isInverse,
    isNew: true
  };
  if (isRoot) rootThemeState = nextState;
  if (pendingUpdate !== "force" && lastState && lastState.name === name) return [false, nextState];
  var shouldAvoidRerender = pendingUpdate !== "force" && lastState && !needsUpdate && nextState.name === lastState.name;
  if (process.env.NODE_ENV === "development" && debug === "verbose") {
    console.groupCollapsed(` \xB7 useTheme(${id}) \u23ED\uFE0F ${name} shouldAvoidRerender: ${shouldAvoidRerender}`);
    console.info({
      lastState,
      needsUpdate,
      nextState,
      pendingUpdate
    });
    console.groupEnd();
  }
  if (shouldAvoidRerender) return [false, nextState];
  return [true, nextState];
};
function scheduleUpdate(id) {
  var queue = [id];
  var visited = /* @__PURE__ */new Set();
  while (queue.length) {
    var children = listenersByParent[queue.shift()];
    if (children) {
      var _iteratorNormalCompletion = true,
        _didIteratorError = false,
        _iteratorError = void 0;
      try {
        for (var _iterator = children[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var childId = _step.value;
          if (!visited.has(childId)) {
            visited.add(childId);
            queue.push(childId);
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) _iterator.return();
        } finally {
          if (_didIteratorError) throw _iteratorError;
        }
      }
    }
  }
  visited.forEach(function (childId2) {
    var cb = allListeners.get(childId2);
    cb === null || cb === void 0 || cb();
  });
}
var validSchemes = {
  light: "light",
  dark: "dark"
};
function getScheme(name) {
  return validSchemes[name.split("_")[0]];
}
function getNewThemeName() {
  var parentName = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "",
    props = arguments.length > 1 ? arguments[1] : void 0,
    forceUpdate = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : false;
  var {
    name,
    reset
  } = props;
  var componentName = props.unstyled ? void 0 : props.componentName;
  if (name && reset) throw new Error(process.env.NODE_ENV === "production" ? `\u274C004` : "Cannot reset and set a new name at the same time.");
  var cacheKey = `${parentName}|${name || ""}|${componentName || ""}|${reset ? 1 : 0}|${forceUpdate ? 1 : 0}`;
  if (themeNameCacheVer !== cacheVersion) {
    themeNameCache.clear();
    themeNameCacheVer = cacheVersion;
  } else {
    var cached = themeNameCache.get(cacheKey);
    if (cached !== void 0) return cached;
  }
  var {
    themes: themes2
  } = getConfig();
  if (reset) {
    if (parentName === "light" || parentName === "dark") {
      var result = parentName === "light" ? "dark" : "light";
      themeNameCache.set(cacheKey, result);
      return result;
    }
    var lastPartIndex = parentName.lastIndexOf("_");
    var name1 = lastPartIndex <= 0 ? parentName : parentName.slice(lastPartIndex);
    var scheme = parentName.slice(0, lastPartIndex);
    var result1 = themes2[name1] ? name1 : scheme;
    themeNameCache.set(cacheKey, result1);
    return result1;
  }
  var parentParts = parentName.split("_");
  var lastName = parentParts[parentParts.length - 1];
  if (lastName && lastName[0].toLowerCase() !== lastName[0]) parentParts.pop();
  var subNames = [name && componentName ? `${name}_${componentName}` : void 0, name, componentName].filter(Boolean);
  var found = null;
  if (name) {
    var nameHasScheme = getScheme(name);
    if (nameHasScheme) {
      var _iteratorNormalCompletion = true,
        _didIteratorError = false,
        _iteratorError = void 0;
      try {
        for (var _iterator = subNames[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var subName = _step.value;
          if (subName in themes2) {
            found = subName;
            break;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) _iterator.return();
        } finally {
          if (_didIteratorError) throw _iteratorError;
        }
      }
    }
    if (!found && !nameHasScheme) {
      if (getScheme(parentName)) {
        var potentialBases = [];
        for (var i = parentParts.length; i >= 1; i--) potentialBases.push(parentParts.slice(0, i).join("_"));
        var _iteratorNormalCompletion1 = true,
          _didIteratorError1 = false,
          _iteratorError1 = void 0;
        try {
          outer: for (var _iterator1 = potentialBases[Symbol.iterator](), _step1; !(_iteratorNormalCompletion1 = (_step1 = _iterator1.next()).done); _iteratorNormalCompletion1 = true) {
            var base = _step1.value;
            var candidates = [componentName ? `${base}_${name}_${componentName}` : void 0, `${base}_${name}`].filter(Boolean);
            var _iteratorNormalCompletion2 = true,
              _didIteratorError2 = false,
              _iteratorError2 = void 0;
            try {
              for (var _iterator2 = candidates[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                var potential = _step2.value;
                if (potential in themes2) {
                  found = potential;
                  break outer;
                }
              }
            } catch (err) {
              _didIteratorError2 = true;
              _iteratorError2 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion2 && _iterator2.return != null) _iterator2.return();
              } finally {
                if (_didIteratorError2) throw _iteratorError2;
              }
            }
          }
        } catch (err) {
          _didIteratorError1 = true;
          _iteratorError1 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion1 && _iterator1.return != null) _iterator1.return();
          } finally {
            if (_didIteratorError1) throw _iteratorError1;
          }
        }
      }
    }
  }
  if (!found) if (!name && componentName) {
    var potential1 = `${parentParts.join("_")}_${componentName}`;
    if (potential1 in themes2) found = potential1;
  } else {
    var max = parentParts.length;
    for (var i1 = 0; i1 <= max; i1++) {
      var base1 = (i1 === 0 ? parentParts : parentParts.slice(0, -i1)).join("_");
      var _iteratorNormalCompletion3 = true,
        _didIteratorError3 = false,
        _iteratorError3 = void 0;
      try {
        for (var _iterator3 = subNames[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
          var subName1 = _step3.value;
          var potential2 = base1 ? `${base1}_${subName1}` : subName1;
          if (potential2 in themes2) {
            found = potential2;
            break;
          }
        }
      } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion3 && _iterator3.return != null) _iterator3.return();
        } finally {
          if (_didIteratorError3) throw _iteratorError3;
        }
      }
      if (found) break;
    }
  }
  if (!forceUpdate && found === parentName && !validSchemes[found]) {
    themeNameCache.set(cacheKey, null);
    return null;
  }
  themeNameCache.set(cacheKey, found);
  return found;
}
var getPropsKey = function (param) {
  var {
    name,
    reset,
    forceClassName,
    componentName
  } = param;
  return `${name || ""}${reset || ""}${forceClassName || ""}${componentName || ""}`;
};
var hasThemeUpdatingProps = function (props) {
  return "name" in props || "reset" in props || "forceClassName" in props;
};
export { ThemeStateContext, forceUpdateThemes, getRootThemeState, getThemeState, hasThemeUpdatingProps, useThemeState };
