// @tamagui/get-font-sized/dist/cjs/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  SizableText: () => SizableText,
  getFontSized: () => getFontSized
});
module.exports = __toCommonJS(index_exports);
var import_constants = require("@tamagui/constants");
var import_web = require("@tamagui/web");
const getFontSized = (sizeTokenIn = "$true", {
  font,
  fontFamily,
  props
}) => {
  if (!font) {
    return {
      fontSize: sizeTokenIn
    };
  }
  const sizeToken = sizeTokenIn === "$true" ? getDefaultSizeToken(font) : sizeTokenIn;
  const style = {};
  const fontSize = font.size[sizeToken];
  const lineHeight = font.lineHeight?.[sizeToken];
  const fontWeight = font.weight?.[sizeToken];
  const letterSpacing = font.letterSpacing?.[sizeToken];
  const textTransform = font.transform?.[sizeToken];
  const fontStyle = props.fontStyle ?? font.style?.[sizeToken];
  const color = props.color ?? font.color?.[sizeToken];
  if (fontStyle) style.fontStyle = fontStyle;
  if (textTransform) style.textTransform = textTransform;
  if (fontFamily) style.fontFamily = fontFamily;
  if (fontWeight) style.fontWeight = fontWeight;
  if (letterSpacing) style.letterSpacing = letterSpacing;
  if (fontSize) style.fontSize = fontSize;
  if (lineHeight) style.lineHeight = lineHeight;
  if (color) style.color = color;
  if (process.env.NODE_ENV === "development") {
    if (props["debug"] && props["debug"] === "verbose") {
      console.groupCollapsed("  \u{1F539} getFontSized", sizeTokenIn, sizeToken);
      if (import_constants.isClient) {
        console.info({
          style,
          props,
          font
        });
      }
      console.groupEnd();
    }
  }
  return style;
};
const SizableText = (0, import_web.styled)(import_web.Text, {
  name: "SizableText",
  fontFamily: "$body",
  variants: {
    size: {
      "...fontSize": getFontSized
    }
  },
  defaultVariants: {
    size: "$true"
  }
});
const cache = /* @__PURE__ */new WeakMap();
function getDefaultSizeToken(font) {
  if (typeof font === "object" && cache.has(font)) {
    return cache.get(font);
  }
  const tokens = (0, import_web.getTokens)();
  const sizeTokens = "$true" in font.size ? font.size : tokens?.size;
  if (!sizeTokens) {
    return Object.keys(font.size)[3];
  }
  const sizeDefault = sizeTokens["$true"];
  const sizeDefaultSpecific = sizeDefault ? Object.keys(sizeTokens).find(x => x !== "$true" && sizeTokens[x]["val"] === sizeDefault["val"]) : null;
  if (!sizeDefault || !sizeDefaultSpecific) {
    if (process.env.NODE_ENV === "development") {
      console.warn(`No default size is set in your tokens for the "true" key, fonts will be inconsistent.

      Fix this by having consistent tokens across fonts and sizes and setting a true key for your size tokens, or
      set true keys for all your font tokens: "size", "lineHeight", "fontStyle", etc.`);
    }
    return Object.keys(font.size)[3];
  }
  cache.set(font, sizeDefaultSpecific);
  return sizeDefaultSpecific;
}
