// @tamagui/get-font-sized/dist/cjs/index.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  SizableText: () => SizableText,
  getFontSized: () => getFontSized
});
module.exports = __toCommonJS(index_exports);
var import_constants = require("@tamagui/constants");
var import_web = require("@tamagui/web");
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var getFontSized = function () {
  var sizeTokenIn = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "$true",
    {
      font,
      fontFamily,
      props
    } = arguments.length > 1 ? arguments[1] : void 0;
  var _font_lineHeight, _font_weight, _font_letterSpacing, _font_transform, _font_style, _font_color;
  if (!font) {
    return {
      fontSize: sizeTokenIn
    };
  }
  var sizeToken = sizeTokenIn === "$true" ? getDefaultSizeToken(font) : sizeTokenIn;
  var style = {};
  var fontSize = font.size[sizeToken];
  var lineHeight = (_font_lineHeight = font.lineHeight) === null || _font_lineHeight === void 0 ? void 0 : _font_lineHeight[sizeToken];
  var fontWeight = (_font_weight = font.weight) === null || _font_weight === void 0 ? void 0 : _font_weight[sizeToken];
  var letterSpacing = (_font_letterSpacing = font.letterSpacing) === null || _font_letterSpacing === void 0 ? void 0 : _font_letterSpacing[sizeToken];
  var textTransform = (_font_transform = font.transform) === null || _font_transform === void 0 ? void 0 : _font_transform[sizeToken];
  var _props_fontStyle;
  var fontStyle = (_props_fontStyle = props.fontStyle) !== null && _props_fontStyle !== void 0 ? _props_fontStyle : (_font_style = font.style) === null || _font_style === void 0 ? void 0 : _font_style[sizeToken];
  var _props_color;
  var color = (_props_color = props.color) !== null && _props_color !== void 0 ? _props_color : (_font_color = font.color) === null || _font_color === void 0 ? void 0 : _font_color[sizeToken];
  if (fontStyle) style.fontStyle = fontStyle;
  if (textTransform) style.textTransform = textTransform;
  if (fontFamily) style.fontFamily = fontFamily;
  if (fontWeight) style.fontWeight = fontWeight;
  if (letterSpacing) style.letterSpacing = letterSpacing;
  if (fontSize) style.fontSize = fontSize;
  if (lineHeight) style.lineHeight = lineHeight;
  if (color) style.color = color;
  if (process.env.NODE_ENV === "development") {
    if (props["debug"] && props["debug"] === "verbose") {
      console.groupCollapsed("  \u{1F539} getFontSized", sizeTokenIn, sizeToken);
      if (import_constants.isClient) {
        console.info({
          style,
          props,
          font
        });
      }
      console.groupEnd();
    }
  }
  return style;
};
var SizableText = (0, import_web.styled)(import_web.Text, {
  name: "SizableText",
  fontFamily: "$body",
  variants: {
    size: {
      "...fontSize": getFontSized
    }
  },
  defaultVariants: {
    size: "$true"
  }
});
var cache = /* @__PURE__ */new WeakMap();
function getDefaultSizeToken(font) {
  if ((typeof font === "undefined" ? "undefined" : _type_of(font)) === "object" && cache.has(font)) {
    return cache.get(font);
  }
  var tokens = (0, import_web.getTokens)();
  var sizeTokens = "$true" in font.size ? font.size : tokens === null || tokens === void 0 ? void 0 : tokens.size;
  if (!sizeTokens) {
    return Object.keys(font.size)[3];
  }
  var sizeDefault = sizeTokens["$true"];
  var sizeDefaultSpecific = sizeDefault ? Object.keys(sizeTokens).find(function (x) {
    return x !== "$true" && sizeTokens[x]["val"] === sizeDefault["val"];
  }) : null;
  if (!sizeDefault || !sizeDefaultSpecific) {
    if (process.env.NODE_ENV === "development") {
      console.warn(`No default size is set in your tokens for the "true" key, fonts will be inconsistent.

      Fix this by having consistent tokens across fonts and sizes and setting a true key for your size tokens, or
      set true keys for all your font tokens: "size", "lineHeight", "fontStyle", etc.`);
    }
    return Object.keys(font.size)[3];
  }
  cache.set(font, sizeDefaultSpecific);
  return sizeDefaultSpecific;
}
