// @tamagui/avatar/dist/esm/Avatar.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { getTokens, getVariableValue, styled } from "@tamagui/core";
import { createContextScope } from "@tamagui/create-context";
import { withStaticProperties } from "@tamagui/helpers";
import { Image } from "@tamagui/image";
import { Square, getShapeSize } from "@tamagui/shapes";
import { YStack } from "@tamagui/stacks";
import * as React from "react";
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var AVATAR_NAME = "Avatar";
var [createAvatarContext, createAvatarScope] = createContextScope(AVATAR_NAME);
var [AvatarProvider, useAvatarContext] = createAvatarContext(AVATAR_NAME);
var IMAGE_NAME = "AvatarImage";
var AvatarImage = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var _getShapeSize;
  var {
    __scopeAvatar,
    src,
    source,
    onLoadingStatusChange = function () {},
    ...imageProps
  } = props;
  var context = useAvatarContext(IMAGE_NAME, __scopeAvatar);
  var [status, setStatus] = React.useState("idle");
  var shapeSize = getVariableValue((_getShapeSize = getShapeSize(context.size,
  // @ts-expect-error
  {
    tokens: getTokens()
  })) === null || _getShapeSize === void 0 ? void 0 : _getShapeSize.width);
  var resolvedSrc = src || (source && (typeof source === "undefined" ? "undefined" : _type_of(source)) === "object" && "uri" in source ? source.uri : source);
  React.useEffect(function () {
    if (!resolvedSrc) {
      setStatus("error");
    } else {
      setStatus("idle");
    }
  }, [resolvedSrc]);
  React.useEffect(function () {
    onLoadingStatusChange(status);
    context.onImageLoadingStatusChange(status);
  }, [status]);
  if (!resolvedSrc) {
    return null;
  }
  return /* @__PURE__ */_jsx(YStack, {
    fullscreen: true,
    zIndex: 1,
    children: /* @__PURE__ */_jsx(Image, {
      position: "absolute",
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      objectFit: "cover",
      ...(typeof shapeSize === "number" && !Number.isNaN(shapeSize) && {
        width: shapeSize,
        height: shapeSize
      }),
      ...imageProps,
      // @ts-ignore
      ref: forwardedRef,
      // @ts-ignore
      src: resolvedSrc,
      // onLoadStart={() => {
      //   // setStatus('loading')
      // }}
      onError: function () {
        setStatus("error");
      },
      onLoad: function () {
        setStatus("loaded");
      }
    })
  });
});
AvatarImage.displayName = IMAGE_NAME;
var FALLBACK_NAME = "AvatarFallback";
var AvatarFallbackFrame = styled(YStack, {
  name: FALLBACK_NAME,
  position: "absolute",
  fullscreen: true,
  zIndex: 0
});
var AvatarFallback = AvatarFallbackFrame.styleable(function (props, forwardedRef) {
  var {
    __scopeAvatar,
    delayMs,
    ...fallbackProps
  } = props;
  var context = useAvatarContext(FALLBACK_NAME, __scopeAvatar);
  var [canRender, setCanRender] = React.useState(delayMs === void 0);
  React.useEffect(function () {
    if (delayMs !== void 0) {
      var timerId = setTimeout(function () {
        return setCanRender(true);
      }, delayMs);
      return function () {
        return clearTimeout(timerId);
      };
    }
  }, [delayMs]);
  return canRender && context.imageLoadingStatus !== "loaded" ? /* @__PURE__ */_jsx(AvatarFallbackFrame, {
    ...fallbackProps,
    ref: forwardedRef
  }) : null;
});
AvatarFallback.displayName = FALLBACK_NAME;
var AvatarFrame = styled(Square, {
  name: AVATAR_NAME,
  position: "relative",
  overflow: "hidden"
});
var Avatar = withStaticProperties(/* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var {
    __scopeAvatar,
    size = "$true",
    ...avatarProps
  } = props;
  var [imageLoadingStatus, setImageLoadingStatus] = React.useState("idle");
  return /* @__PURE__ */_jsx(AvatarProvider, {
    size,
    scope: __scopeAvatar,
    imageLoadingStatus,
    onImageLoadingStatusChange: setImageLoadingStatus,
    children: /* @__PURE__ */_jsx(AvatarFrame, {
      size,
      ...avatarProps,
      ref: forwardedRef
    })
  });
}), {
  Image: AvatarImage,
  Fallback: AvatarFallback
});
Avatar.displayName = AVATAR_NAME;
export { Avatar, AvatarFallback, AvatarFallbackFrame, AvatarFrame, AvatarImage, createAvatarScope };
