// @tamagui/accordion/dist/esm/Accordion.mjs
import { Collapsible } from "@tamagui/collapsible";
import { createCollection } from "@tamagui/collection";
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { View, createStyledContext, styled } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { YStack } from "@tamagui/stacks";
import { H1 } from "@tamagui/text";
import { useControllableState } from "@tamagui/use-controllable-state";
import { useDirection } from "@tamagui/use-direction";
import * as React from "react";
import { jsx } from "react/jsx-runtime";
const ACCORDION_NAME = "Accordion";
const ACCORDION_KEYS = ["Home", "End", "ArrowDown", "ArrowUp", "ArrowLeft", "ArrowRight"];
const [Collection, useCollection] = createCollection(ACCORDION_NAME);
const ACCORDION_CONTEXT = "Accordion";
const AccordionComponent = React.forwardRef((props, forwardedRef) => {
  const {
    type,
    ...accordionProps
  } = props;
  const singleProps = accordionProps;
  const multipleProps = accordionProps;
  return /* @__PURE__ */jsx(Collection.Provider, {
    scope: props.__scopeAccordion || ACCORDION_CONTEXT,
    children: type === "multiple" ? /* @__PURE__ */jsx(AccordionImplMultiple, {
      ...multipleProps,
      ref: forwardedRef
    }) : /* @__PURE__ */jsx(AccordionImplSingle, {
      ...singleProps,
      ref: forwardedRef
    })
  });
});
AccordionComponent.displayName = ACCORDION_NAME;
AccordionComponent.propTypes = {
  type(props) {
    const value = props.value || props.defaultValue;
    if (props.type && !["single", "multiple"].includes(props.type)) {
      return new Error("Invalid prop `type` supplied to `Accordion`. Expected one of `single | multiple`.");
    }
    if (props.type === "multiple" && typeof value === "string") {
      return new Error("Invalid prop `type` supplied to `Accordion`. Expected `single` when `defaultValue` or `value` is type `string`.");
    }
    if (props.type === "single" && Array.isArray(value)) {
      return new Error("Invalid prop `type` supplied to `Accordion`. Expected `multiple` when `defaultValue` or `value` is type `string[]`.");
    }
    return null;
  }
};
const {
  Provider: AccordionValueProvider,
  useStyledContext: useAccordionValueContext
} = createStyledContext();
const {
  Provider: AccordionCollapsibleProvider,
  useStyledContext: useAccordionCollapsibleContext
} = createStyledContext();
const AccordionImplSingle = React.forwardRef((props, forwardedRef) => {
  const {
    value: valueProp,
    defaultValue,
    control,
    onValueChange = () => {},
    collapsible = false,
    ...accordionSingleProps
  } = props;
  const [value, setValue] = useControllableState({
    prop: valueProp,
    defaultProp: defaultValue || "",
    onChange: onValueChange
  });
  return /* @__PURE__ */jsx(AccordionValueProvider, {
    scope: props.__scopeAccordion,
    value: value ? [value] : [],
    onItemOpen: setValue,
    onItemClose: React.useCallback(() => collapsible && setValue(""), [setValue, collapsible]),
    children: /* @__PURE__ */jsx(AccordionCollapsibleProvider, {
      scope: props.__scopeAccordion,
      collapsible,
      children: /* @__PURE__ */jsx(AccordionImpl, {
        ...accordionSingleProps,
        ref: forwardedRef
      })
    })
  });
});
const AccordionImplMultiple = React.forwardRef((props, forwardedRef) => {
  const {
    value: valueProp,
    defaultValue,
    onValueChange = () => {},
    ...accordionMultipleProps
  } = props;
  const [value, setValue] = useControllableState({
    prop: valueProp,
    defaultProp: defaultValue || [],
    onChange: onValueChange
  });
  const handleItemOpen = React.useCallback(itemValue => setValue((prevValue = []) => [...prevValue, itemValue]), [setValue]);
  const handleItemClose = React.useCallback(itemValue => setValue((prevValue = []) => prevValue.filter(value2 => value2 !== itemValue)), [setValue]);
  return /* @__PURE__ */jsx(AccordionValueProvider, {
    scope: props.__scopeAccordion,
    value: value || [],
    onItemOpen: handleItemOpen,
    onItemClose: handleItemClose,
    children: /* @__PURE__ */jsx(AccordionCollapsibleProvider, {
      scope: props.__scopeAccordion,
      collapsible: true,
      children: /* @__PURE__ */jsx(AccordionImpl, {
        ...accordionMultipleProps,
        ref: forwardedRef
      })
    })
  });
});
const {
  Provider: AccordionImplProvider,
  useStyledContext: useAccordionContext
} = createStyledContext();
const AccordionImpl = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeAccordion,
    disabled,
    dir,
    orientation = "vertical",
    ...accordionProps
  } = props;
  const accordionRef = React.useRef(null);
  const composedRef = useComposedRefs(accordionRef, forwardedRef);
  const getItems = useCollection(__scopeAccordion || ACCORDION_CONTEXT);
  const direction = useDirection(dir);
  const isDirectionLTR = direction === "ltr";
  const handleKeyDown = composeEventHandlers(props.onKeyDown, event => {
    if (!ACCORDION_KEYS.includes(event.key)) return;
    const target = event.target;
    const triggerCollection = getItems().filter(item => {
      const el = item.ref.current;
      return !el?.disabled;
    });
    const triggerIndex = triggerCollection.findIndex(item => item.ref.current === target);
    const triggerCount = triggerCollection.length;
    if (triggerIndex === -1) return;
    event.preventDefault();
    let nextIndex = triggerIndex;
    const homeIndex = 0;
    const endIndex = triggerCount - 1;
    const moveNext = () => {
      nextIndex = triggerIndex + 1;
      if (nextIndex > endIndex) {
        nextIndex = homeIndex;
      }
    };
    const movePrev = () => {
      nextIndex = triggerIndex - 1;
      if (nextIndex < homeIndex) {
        nextIndex = endIndex;
      }
    };
    switch (event.key) {
      case "Home":
        nextIndex = homeIndex;
        break;
      case "End":
        nextIndex = endIndex;
        break;
      case "ArrowRight":
        if (orientation === "horizontal") {
          if (isDirectionLTR) {
            moveNext();
          } else {
            movePrev();
          }
        }
        break;
      case "ArrowDown":
        if (orientation === "vertical") {
          moveNext();
        }
        break;
      case "ArrowLeft":
        if (orientation === "horizontal") {
          if (isDirectionLTR) {
            movePrev();
          } else {
            moveNext();
          }
        }
        break;
      case "ArrowUp":
        if (orientation === "vertical") {
          movePrev();
        }
        break;
    }
    const clampedIndex = nextIndex % triggerCount;
    triggerCollection[clampedIndex].ref.current?.focus();
  });
  return /* @__PURE__ */jsx(AccordionImplProvider, {
    scope: __scopeAccordion,
    disabled,
    direction: dir,
    orientation,
    children: /* @__PURE__ */jsx(Collection.Slot, {
      scope: __scopeAccordion || ACCORDION_CONTEXT,
      children: /* @__PURE__ */jsx(YStack, {
        "data-orientation": orientation,
        ref: composedRef,
        ...accordionProps,
        ...(isWeb && {
          onKeyDown: handleKeyDown
        })
      })
    })
  });
});
const ITEM_NAME = "AccordionItem";
const {
  Provider: AccordionItemProvider,
  useStyledContext: useAccordionItemContext
} = createStyledContext();
const AccordionItem = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeAccordion,
    value,
    ...accordionItemProps
  } = props;
  const accordionContext = useAccordionContext(__scopeAccordion);
  const valueContext = useAccordionValueContext(__scopeAccordion);
  const triggerId = React.useId();
  const open = value && valueContext.value.includes(value) || false;
  const disabled = accordionContext.disabled || props.disabled;
  return /* @__PURE__ */jsx(AccordionItemProvider, {
    scope: __scopeAccordion,
    open,
    disabled,
    triggerId,
    children: /* @__PURE__ */jsx(Collapsible, {
      "data-orientation": accordionContext.orientation,
      "data-state": open ? "open" : "closed",
      __scopeCollapsible: __scopeAccordion || ACCORDION_CONTEXT,
      ...accordionItemProps,
      ref: forwardedRef,
      disabled,
      open,
      onOpenChange: open2 => {
        if (open2) {
          valueContext.onItemOpen(value);
        } else {
          valueContext.onItemClose(value);
        }
      }
    })
  });
});
AccordionItem.displayName = ITEM_NAME;
const HEADER_NAME = "AccordionHeader";
const AccordionHeader = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeAccordion,
    ...headerProps
  } = props;
  const accordionContext = useAccordionContext(__scopeAccordion);
  const itemContext = useAccordionItemContext(__scopeAccordion);
  return /* @__PURE__ */jsx(H1, {
    "data-orientation": accordionContext.orientation,
    "data-state": getState(itemContext.open),
    "data-disabled": itemContext.disabled ? "" : void 0,
    ...headerProps,
    ref: forwardedRef
  });
});
AccordionHeader.displayName = HEADER_NAME;
const AccordionTriggerFrame = styled(Collapsible.Trigger, {
  variants: {
    unstyled: {
      false: {
        cursor: "pointer",
        backgroundColor: "$background",
        padding: "$true",
        hoverStyle: {
          backgroundColor: "$backgroundHover"
        },
        focusStyle: {
          backgroundColor: "$backgroundFocus"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress"
        }
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const AccordionTrigger = AccordionTriggerFrame.styleable(function AccordionTrigger2(props, forwardedRef) {
  const {
    __scopeAccordion,
    ...triggerProps
  } = props;
  const accordionContext = useAccordionContext(__scopeAccordion);
  const itemContext = useAccordionItemContext(__scopeAccordion);
  const collapsibleContext = useAccordionCollapsibleContext(__scopeAccordion);
  return /* @__PURE__ */jsx(Collection.ItemSlot, {
    scope: __scopeAccordion || ACCORDION_CONTEXT,
    children: /* @__PURE__ */jsx(AccordionTriggerFrame, {
      __scopeCollapsible: __scopeAccordion || ACCORDION_CONTEXT,
      "aria-disabled": itemContext.open && !collapsibleContext.collapsible || void 0,
      "data-orientation": accordionContext.orientation,
      id: itemContext.triggerId,
      ...triggerProps,
      ref: forwardedRef
    })
  });
});
const AccordionContentFrame = styled(Collapsible.Content, {
  variants: {
    unstyled: {
      false: {
        padding: "$true",
        backgroundColor: "$background"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const AccordionContent = AccordionContentFrame.styleable(function AccordionContent2(props, forwardedRef) {
  const {
    __scopeAccordion,
    ...contentProps
  } = props;
  const accordionContext = useAccordionContext(__scopeAccordion);
  const itemContext = useAccordionItemContext(__scopeAccordion);
  return /* @__PURE__ */jsx(AccordionContentFrame, {
    role: "region",
    "aria-labelledby": itemContext.triggerId,
    "data-orientation": accordionContext.orientation,
    __scopeCollapsible: __scopeAccordion || ACCORDION_CONTEXT,
    ...contentProps,
    ref: forwardedRef
  });
});
const HeightAnimator = View.styleable((props, ref) => {
  const itemContext = useAccordionItemContext();
  const {
    children,
    ...rest
  } = props;
  const [measuredHeight, setMeasuredHeight] = React.useState(0);
  const hasMeasured = measuredHeight > 0;
  const height = itemContext.open ? hasMeasured ? measuredHeight : "auto" : 0;
  const shouldAbsolutePosition = hasMeasured || !itemContext.open;
  return /* @__PURE__ */jsx(View, {
    ref,
    height,
    position: "relative",
    ...rest,
    children: /* @__PURE__ */jsx(View, {
      position: shouldAbsolutePosition ? "absolute" : "relative",
      top: shouldAbsolutePosition ? 0 : void 0,
      left: shouldAbsolutePosition ? 0 : void 0,
      right: shouldAbsolutePosition ? 0 : void 0,
      onLayout: ({
        nativeEvent
      }) => {
        if (nativeEvent.layout.height && nativeEvent.layout.height !== measuredHeight) {
          setMeasuredHeight(nativeEvent.layout.height);
        }
      },
      children
    })
  });
});
function getState(open) {
  return open ? "open" : "closed";
}
const Accordion = withStaticProperties(AccordionComponent, {
  Trigger: AccordionTrigger,
  Header: AccordionHeader,
  Content: AccordionContent,
  Item: AccordionItem,
  HeightAnimator
});
export { Accordion };
