// @tamagui/animate/dist/esm/Animate.native.js
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { AnimatePresence, PresenceChild } from "@tamagui/animate-presence";
import { startTransition, useEffect, useState } from "react";
function Animate(param) {
  var {
    children,
    lazyMount,
    type,
    present,
    passThrough,
    ...props
  } = param;
  var [lazyMounted, setLazyMounted] = useState(lazyMount ? false : present);
  useEffect(function () {
    if (passThrough) return;
    if (!lazyMount) return;
    if (!present) return;
    startTransition(function () {
      setLazyMounted(present);
    });
  }, [lazyMount, present]);
  var mounted = !present ? false : lazyMount ? lazyMounted : present;
  if (type === "presence") {
    if (props.keepChildrenMounted) {
      return /* @__PURE__ */_jsx(PresenceChild, {
        isPresent: true,
        ...(!passThrough && {
          initial: props.initial ? void 0 : false,
          onExitComplete: props.onExitComplete,
          enterVariant: props.enterVariant,
          exitVariant: props.exitVariant,
          enterExitVariant: props.enterExitVariant,
          // BUGFIX: this causes continous re-renders if keepChildrenMounted is true, see HeaderMenu
          // but since we always re-render this component on open changes this should be fine to leave off?
          presenceAffectsLayout: false,
          isPresent: present,
          custom: props.custom
        }),
        children
      });
    }
    return /* @__PURE__ */_jsx(AnimatePresence, {
      passThrough,
      ...props,
      children: mounted || passThrough ? children : null
    });
  }
  return /* @__PURE__ */_jsx(_Fragment, {
    children
  });
}
export { Animate };
