// @tamagui/animate/dist/esm/Animate.mjs
import { AnimatePresence, PresenceChild } from "@tamagui/animate-presence";
import { startTransition, useEffect, useState } from "react";
import { Fragment, jsx } from "react/jsx-runtime";
function Animate({
  children,
  lazyMount,
  type,
  present,
  passThrough,
  ...props
}) {
  const [lazyMounted, setLazyMounted] = useState(lazyMount ? false : present);
  useEffect(() => {
    if (passThrough) return;
    if (!lazyMount) return;
    if (!present) return;
    startTransition(() => {
      setLazyMounted(present);
    });
  }, [lazyMount, present]);
  const mounted = !present ? false : lazyMount ? lazyMounted : present;
  if (type === "presence") {
    if (props.keepChildrenMounted) {
      return /* @__PURE__ */jsx(PresenceChild, {
        isPresent: true,
        ...(!passThrough && {
          initial: props.initial ? void 0 : false,
          onExitComplete: props.onExitComplete,
          enterVariant: props.enterVariant,
          exitVariant: props.exitVariant,
          enterExitVariant: props.enterExitVariant,
          // BUGFIX: this causes continous re-renders if keepChildrenMounted is true, see HeaderMenu
          // but since we always re-render this component on open changes this should be fine to leave off?
          presenceAffectsLayout: false,
          isPresent: present,
          custom: props.custom
        }),
        children
      });
    }
    return /* @__PURE__ */jsx(AnimatePresence, {
      passThrough,
      ...props,
      children: mounted || passThrough ? children : null
    });
  }
  return /* @__PURE__ */jsx(Fragment, {
    children
  });
}
export { Animate };
