// @tamagui/animation-helpers/dist/cjs/normalizeTransition.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var normalizeTransition_exports = {};
__export(normalizeTransition_exports, {
  getAnimatedProperties: () => getAnimatedProperties,
  getAnimationConfigsForKeys: () => getAnimationConfigsForKeys,
  getAnimationForProperty: () => getAnimationForProperty,
  getEffectiveAnimation: () => getEffectiveAnimation,
  hasAnimation: () => hasAnimation,
  normalizeTransition: () => normalizeTransition
});
module.exports = __toCommonJS(normalizeTransition_exports);
const SPRING_CONFIG_KEYS = /* @__PURE__ */new Set(["stiffness", "damping", "mass", "tension", "friction", "velocity", "overshootClamping", "duration", "bounciness", "speed"]);
function isSpringConfigKey(key) {
  return SPRING_CONFIG_KEYS.has(key);
}
function normalizeTransition(transition) {
  if (!transition) {
    return {
      default: null,
      enter: null,
      exit: null,
      delay: void 0,
      properties: {}
    };
  }
  if (typeof transition === "string") {
    return {
      default: transition,
      enter: null,
      exit: null,
      delay: void 0,
      properties: {}
    };
  }
  if (Array.isArray(transition)) {
    const [defaultAnimation, configObj] = transition;
    const properties = {};
    const springConfig = {};
    let delay;
    let enter = null;
    let exit = null;
    if (configObj && typeof configObj === "object") {
      for (const [key, value] of Object.entries(configObj)) {
        if (key === "delay" && typeof value === "number") {
          delay = value;
        } else if (key === "enter" && typeof value === "string") {
          enter = value;
        } else if (key === "exit" && typeof value === "string") {
          exit = value;
        } else if (isSpringConfigKey(key) && value !== void 0) {
          springConfig[key] = value;
        } else if (value !== void 0) {
          properties[key] = value;
        }
      }
    }
    return {
      default: defaultAnimation,
      enter,
      exit,
      delay,
      properties,
      config: Object.keys(springConfig).length > 0 ? springConfig : void 0
    };
  }
  if (typeof transition === "object") {
    const properties = {};
    const springConfig = {};
    let defaultAnimation = null;
    let enter = null;
    let exit = null;
    let delay;
    for (const [key, value] of Object.entries(transition)) {
      if (key === "default" && typeof value === "string") {
        defaultAnimation = value;
      } else if (key === "enter" && typeof value === "string") {
        enter = value;
      } else if (key === "exit" && typeof value === "string") {
        exit = value;
      } else if (key === "delay" && typeof value === "number") {
        delay = value;
      } else if (isSpringConfigKey(key) && value !== void 0) {
        springConfig[key] = value;
      } else if (value !== void 0) {
        properties[key] = value;
      }
    }
    return {
      default: defaultAnimation,
      enter,
      exit,
      delay,
      properties,
      config: Object.keys(springConfig).length > 0 ? springConfig : void 0
    };
  }
  return {
    default: null,
    enter: null,
    exit: null,
    delay: void 0,
    properties: {}
  };
}
function getAnimationForProperty(normalized, property) {
  const propertyAnimation = normalized.properties[property];
  if (propertyAnimation !== void 0) {
    return propertyAnimation;
  }
  return normalized.default;
}
function hasAnimation(normalized) {
  return normalized.default !== null || normalized.enter !== null || normalized.exit !== null || Object.keys(normalized.properties).length > 0;
}
function getAnimatedProperties(normalized) {
  return Object.keys(normalized.properties);
}
function getEffectiveAnimation(normalized, state) {
  if (state === "enter" && normalized.enter) {
    return normalized.enter;
  }
  if (state === "exit" && normalized.exit) {
    return normalized.exit;
  }
  return normalized.default;
}
function getAnimationConfigsForKeys(normalized, animations, keys, defaultAnimation) {
  const result = /* @__PURE__ */new Map();
  for (const key of keys) {
    const propAnimation = normalized.properties[key];
    let animationValue = null;
    if (typeof propAnimation === "string") {
      animationValue = animations[propAnimation] ?? null;
    } else if (propAnimation && typeof propAnimation === "object" && propAnimation.type) {
      animationValue = animations[propAnimation.type] ?? null;
    }
    if (animationValue === null) {
      animationValue = defaultAnimation;
    }
    result.set(key, animationValue);
  }
  return result;
}
