// @tamagui/switch/dist/jsx/Switch.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Switch_exports = {};
__export(Switch_exports, {
  SwitchFrame: () => SwitchFrame,
  SwitchThumb: () => SwitchThumb
});
module.exports = __toCommonJS(Switch_exports);
var import_core = require("@tamagui/core");
var import_get_token = require("@tamagui/get-token");
var import_stacks = require("@tamagui/stacks");
var SwitchThumb = (0, import_core.styled)(import_stacks.YStack, {
  name: "SwitchThumb",
  variants: {
    unstyled: {
      false: {
        size: "$true",
        backgroundColor: "$background",
        borderRadius: 1e3
      }
    },
    size: {
      "...size": function (val) {
        var size = getSwitchHeight(val);
        return {
          height: size,
          width: size
        };
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
}, {
  accept: {
    activeStyle: "style"
  }
});
var getSwitchHeight = function (val) {
  return Math.round((0, import_core.getVariableValue)((0, import_get_token.getSize)(val)) * 0.65);
};
var getSwitchWidth = function (val) {
  return getSwitchHeight(val) * 2;
};
var SwitchFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: "Switch",
  render: "button",
  tabIndex: 0,
  variants: {
    unstyled: {
      false: {
        borderRadius: 1e3,
        backgroundColor: "$background",
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineStyle: "solid",
          outlineWidth: 2
        }
      }
    },
    size: {
      "...size": function (val, param) {
        var {
          props
        } = param;
        if (props["unstyled"]) return;
        var height = getSwitchHeight(val);
        var width = getSwitchWidth(val);
        return {
          height,
          minHeight: height,
          width
        };
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
}, {
  accept: {
    activeStyle: "style"
  }
});
