// @tamagui/v2-toast/dist/cjs/Toast.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Toast_exports = {};
__export(Toast_exports, {
  Toast: () => Toast2,
  ToastProvider: () => import_ToastProvider.ToastProvider,
  ToastViewport: () => import_ToastViewport.ToastViewport,
  useToast: () => import_ToastImperative.useToast,
  useToastController: () => import_ToastImperative.useToastController,
  useToastState: () => import_ToastImperative.useToastState
});
module.exports = __toCommonJS(Toast_exports);
var import_animate_presence = require("@tamagui/animate-presence");
var import_core = require("@tamagui/core");
var import_helpers = require("@tamagui/helpers");
var import_stacks = require("@tamagui/stacks");
var import_text = require("@tamagui/text");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var React = __toESM(require("react"), 1);
var import_ToastAnnounce = require("./ToastAnnounce.cjs");
var import_ToastImperative = require("./ToastImperative.cjs");
var import_ToastImpl = require("./ToastImpl.cjs");
var import_ToastProvider = require("./ToastProvider.cjs");
var import_ToastViewport = require("./ToastViewport.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
const TITLE_NAME = "ToastTitle";
const ToastTitle = (0, import_core.styled)(import_text.SizableText, {
  name: TITLE_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color",
        size: "$4"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const DESCRIPTION_NAME = "ToastDescription";
const ToastDescription = (0, import_core.styled)(import_text.SizableText, {
  name: DESCRIPTION_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color11",
        size: "$1"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const ACTION_NAME = "ToastAction";
const ToastAction = React.forwardRef(function ToastAction2(props, forwardedRef) {
  const {
    altText,
    ...actionProps
  } = props;
  if (!altText) return null;
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ToastAnnounce.ToastAnnounceExclude, {
    altText,
    asChild: true,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(ToastClose, {
      ...actionProps,
      ref: forwardedRef
    })
  });
});
ToastAction.propTypes = {
  altText(props) {
    if (!props.altText) {
      return new Error(`Missing prop \`altText\` expected on \`${ACTION_NAME}\``);
    }
    return null;
  }
};
const CLOSE_NAME = "ToastClose";
const ToastCloseFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: CLOSE_NAME,
  render: "button"
});
const ToastClose = React.forwardRef(function ToastClose2(props, forwardedRef) {
  const {
    scope,
    ...closeProps
  } = props;
  const interactiveContext = (0, import_ToastImpl.useToastInteractiveContext)(scope);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ToastAnnounce.ToastAnnounceExclude, {
    asChild: true,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(ToastCloseFrame, {
      "aria-label": "Close",
      ...closeProps,
      ref: forwardedRef,
      onPress: (0, import_helpers.composeEventHandlers)(props.onPress, interactiveContext.onClose)
    })
  });
});
const ToastComponent = import_ToastImpl.ToastImplFrame.styleable(function Toast(props, forwardedRef) {
  const {
    forceMount,
    open: openProp,
    defaultOpen,
    onOpenChange,
    ...toastProps
  } = props;
  const [open, setOpen] = (0, import_use_controllable_state.useControllableState)({
    prop: openProp,
    defaultProp: defaultOpen ?? true,
    onChange: onOpenChange,
    strategy: "most-recent-wins"
  });
  const currentToast = (0, import_ToastImperative.useToastState)();
  const {
    hide
  } = (0, import_ToastImperative.useToastController)();
  const id = React.useId();
  const onPause = (0, import_core.useEvent)(props.onPause);
  const onResume = (0, import_core.useEvent)(props.onResume);
  const isHide = currentToast?.hide === true;
  const shouldShow = (forceMount || open) && !isHide;
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate_presence.AnimatePresence, {
    children: shouldShow ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ToastImpl.ToastImpl, {
      id,
      open,
      ...toastProps,
      ref: forwardedRef,
      onClose: () => {
        setOpen(false);
        hide();
      },
      onPause,
      onResume,
      onSwipeEnd: (0, import_helpers.composeEventHandlers)(props.onSwipeEnd, event => {
        setOpen(false);
      })
    }) : null
  }, id);
});
const Toast2 = (0, import_helpers.withStaticProperties)(ToastComponent, {
  Title: ToastTitle,
  Description: ToastDescription,
  Action: ToastAction,
  Close: ToastClose
});
