// @tamagui/v2-toast/dist/cjs/Toaster.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Toaster_exports = {};
__export(Toaster_exports, {
  Toaster: () => Toaster
});
module.exports = __toCommonJS(Toaster_exports);
var import_animate_presence = require("@tamagui/animate-presence");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_portal = require("@tamagui/portal");
var React = __toESM(require("react"), 1);
var import_ToastItem = require("./ToastItem.cjs");
var import_ToastState = require("./ToastState.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
const VISIBLE_TOASTS_AMOUNT = 4;
const VIEWPORT_OFFSET = 16;
const TOAST_GAP = 14;
const TOAST_LIFETIME = 4e3;
const TOAST_WIDTH = 356;
const ToasterFrame = (0, import_core.styled)(import_core.View, {
  name: "Toaster",
  variants: {
    unstyled: {
      false: {
        position: "fixed",
        zIndex: 1e5,
        pointerEvents: "box-none",
        maxWidth: "100%",
        // need min-height to contain absolutely positioned toasts
        // toasts will overflow upward/downward from their anchor position
        minHeight: 1
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const Toaster = React.forwardRef(function Toaster2(props, _ref) {
  const {
    position = "bottom-right",
    width = TOAST_WIDTH,
    expand = false,
    visibleToasts = VISIBLE_TOASTS_AMOUNT,
    gap = TOAST_GAP,
    duration = TOAST_LIFETIME,
    offset = VIEWPORT_OFFSET,
    hotkey = ["altKey", "KeyT"],
    swipeDirection = "right",
    swipeThreshold = 50,
    closeButton = false,
    theme: themeProp,
    icons,
    toastOptions,
    containerAriaLabel = "Notifications",
    disableNative = false,
    burntOptions,
    notificationOptions,
    className,
    style
  } = props;
  const [toasts, setToasts] = React.useState([]);
  const [heights, setHeights] = React.useState([]);
  const [expanded, setExpanded] = React.useState(false);
  const [interacting, setInteracting] = React.useState(false);
  const listRef = React.useRef(null);
  const lastFocusedElementRef = React.useRef(null);
  const isFocusWithinRef = React.useRef(false);
  React.useEffect(() => {
    return import_ToastState.ToastState.subscribe(toast => {
      if (toast.dismiss) {
        setToasts(toasts2 => toasts2.map(t => t.id === toast.id ? {
          ...t,
          delete: true
        } : t));
        return;
      }
      setToasts(toasts2 => {
        const indexOfExistingToast = toasts2.findIndex(t => t.id === toast.id);
        if (indexOfExistingToast !== -1) {
          return [...toasts2.slice(0, indexOfExistingToast), {
            ...toasts2[indexOfExistingToast],
            ...toast
          }, ...toasts2.slice(indexOfExistingToast + 1)];
        }
        return [toast, ...toasts2];
      });
    });
  }, []);
  React.useEffect(() => {
    if (toasts.length <= 1) {
      setExpanded(false);
    }
  }, [toasts.length]);
  React.useEffect(() => {
    if (!import_constants.isWeb) return;
    const handleKeyDown = event => {
      const isHotkeyPressed = hotkey.length > 0 && hotkey.every(key => event[key] || event.code === key);
      if (isHotkeyPressed) {
        setExpanded(true);
        listRef.current?.focus();
      }
      if (event.code === "Escape" && (document.activeElement === listRef.current || listRef.current?.contains(document.activeElement))) {
        setExpanded(false);
      }
    };
    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [hotkey]);
  React.useEffect(() => {
    if (!import_constants.isWeb || !listRef.current) return;
    return () => {
      if (lastFocusedElementRef.current) {
        lastFocusedElementRef.current.focus({
          preventScroll: true
        });
        lastFocusedElementRef.current = null;
        isFocusWithinRef.current = false;
      }
    };
  }, []);
  const removeToast = React.useCallback(toastToRemove => {
    setToasts(toasts2 => {
      if (!toasts2.find(toast => toast.id === toastToRemove.id)?.delete) {
        import_ToastState.ToastState.dismiss(toastToRemove.id);
      }
      return toasts2.filter(({
        id
      }) => id !== toastToRemove.id);
    });
  }, []);
  const [yPosition, xPosition] = position.split("-");
  const offsetStyles = React.useMemo(() => {
    const styles = {};
    const defaultOffset = typeof offset === "number" ? offset : VIEWPORT_OFFSET;
    const offsetObj = typeof offset === "object" ? offset : {
      top: defaultOffset,
      right: defaultOffset,
      bottom: defaultOffset,
      left: defaultOffset
    };
    if (yPosition === "top") {
      styles.top = offsetObj.top ?? defaultOffset;
    } else {
      styles.bottom = offsetObj.bottom ?? defaultOffset;
    }
    if (xPosition === "left") {
      styles.left = offsetObj.left ?? defaultOffset;
    } else if (xPosition === "right") {
      styles.right = offsetObj.right ?? defaultOffset;
    } else {
      styles.left = "50%";
      styles.transform = "translateX(-50%)";
    }
    return styles;
  }, [offset, yPosition, xPosition]);
  const currentTheme = (0, import_core.useThemeName)();
  const resolvedTheme = themeProp === "system" || !themeProp ? currentTheme?.includes("dark") ? "dark" : "light" : themeProp;
  const hotkeyLabel = hotkey.join("+").replace(/Key/g, "").replace(/Digit/g, "");
  if (toasts.length === 0) {
    return null;
  }
  const content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(ToasterFrame, {
    ref: listRef,
    width,
    "aria-label": `${containerAriaLabel} ${hotkeyLabel}`,
    tabIndex: -1,
    "aria-live": "polite",
    "aria-relevant": "additions text",
    "aria-atomic": false,
    style: {
      ...offsetStyles,
      ...style
    },
    className,
    "data-y-position": yPosition,
    "data-x-position": xPosition,
    onMouseEnter: () => setExpanded(true),
    onMouseMove: () => setExpanded(true),
    onMouseLeave: () => {
      if (!interacting) {
        setExpanded(false);
      }
    },
    onPointerDown: () => setInteracting(true),
    onPointerUp: () => setInteracting(false),
    ...(import_constants.isWeb && {
      onBlur: event => {
        if (isFocusWithinRef.current && !event.currentTarget.contains(event.relatedTarget)) {
          isFocusWithinRef.current = false;
          if (lastFocusedElementRef.current) {
            lastFocusedElementRef.current.focus({
              preventScroll: true
            });
            lastFocusedElementRef.current = null;
          }
        }
      },
      onFocus: event => {
        if (!isFocusWithinRef.current) {
          isFocusWithinRef.current = true;
          lastFocusedElementRef.current = event.relatedTarget;
        }
      }
    }),
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate_presence.AnimatePresence, {
      children: toasts.map((toast, index) => {
        const isVisible = index < visibleToasts;
        const isFront = index === 0;
        return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ToastItem.ToastItem, {
          toast,
          index,
          expanded: expanded || expand,
          interacting,
          position,
          visibleToasts,
          removeToast,
          heights,
          setHeights,
          duration: toast.duration ?? toastOptions?.duration ?? duration,
          gap,
          swipeDirection,
          swipeThreshold,
          closeButton: toast.closeButton ?? closeButton,
          icons,
          disableNative,
          burntOptions,
          notificationOptions
        }, toast.id);
      })
    })
  });
  if (import_constants.isWeb) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_portal.Portal, {
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.Theme, {
        name: resolvedTheme,
        children: content
      })
    });
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.Theme, {
    name: resolvedTheme,
    children: content
  });
});
Toaster.displayName = "Toaster";
