// @tamagui/v2-toast/dist/cjs/useDragGesture.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useDragGesture_exports = {};
__export(useDragGesture_exports, {
  useDragGesture: () => useDragGesture
});
module.exports = __toCommonJS(useDragGesture_exports);
var React = __toESM(require("react"), 1);
const VELOCITY_THRESHOLD = 0.11;
function resisted(delta, maxResist = 25) {
  if (delta >= 0) return delta;
  const pastBoundary = Math.abs(delta);
  const resistedDistance = Math.sqrt(pastBoundary) * 2;
  return -Math.min(resistedDistance, maxResist);
}
function useDragGesture(options) {
  const {
    direction,
    threshold,
    onDragStart,
    onDragEnd,
    onDragCancel,
    disabled
  } = options;
  const [dragState, setDragState] = React.useState({
    isDragging: false,
    offsetX: 0,
    offsetY: 0,
    velocityX: 0,
    velocityY: 0
  });
  const dragStartRef = React.useRef(null);
  const lockedDirectionRef = React.useRef(null);
  const isHorizontal = direction === "left" || direction === "right" || direction === "horizontal";
  const isVertical = direction === "up" || direction === "down" || direction === "vertical";
  const handlePointerDown = React.useCallback(event => {
    if (disabled) return;
    if (event.button !== 0) return;
    event.target.setPointerCapture(event.pointerId);
    dragStartRef.current = {
      startX: event.clientX,
      startY: event.clientY,
      startTime: Date.now()
    };
    setDragState(prev => ({
      ...prev,
      isDragging: true
    }));
    onDragStart?.();
  }, [disabled, onDragStart]);
  const handlePointerMove = React.useCallback(event => {
    if (!dragStartRef.current || disabled) return;
    const deltaX = event.clientX - dragStartRef.current.startX;
    const deltaY = event.clientY - dragStartRef.current.startY;
    if (!lockedDirectionRef.current && (Math.abs(deltaX) > 1 || Math.abs(deltaY) > 1)) {
      lockedDirectionRef.current = Math.abs(deltaX) > Math.abs(deltaY) ? "x" : "y";
    }
    let offsetX = 0;
    let offsetY = 0;
    if (lockedDirectionRef.current === "x" && isHorizontal) {
      if (direction === "right") {
        offsetX = deltaX > 0 ? deltaX : resisted(deltaX);
      } else if (direction === "left") {
        offsetX = deltaX < 0 ? deltaX : -resisted(-deltaX);
      } else {
        offsetX = deltaX;
      }
    } else if (lockedDirectionRef.current === "y" && isVertical) {
      if (direction === "down") {
        offsetY = deltaY > 0 ? deltaY : resisted(deltaY);
      } else if (direction === "up") {
        offsetY = deltaY < 0 ? deltaY : -resisted(-deltaY);
      } else {
        offsetY = deltaY;
      }
    }
    setDragState(prev => ({
      ...prev,
      offsetX,
      offsetY
    }));
  }, [disabled, direction, isHorizontal, isVertical]);
  const handlePointerUp = React.useCallback(event => {
    if (!dragStartRef.current || disabled) return;
    const deltaX = event.clientX - dragStartRef.current.startX;
    const deltaY = event.clientY - dragStartRef.current.startY;
    const timeTaken = Date.now() - dragStartRef.current.startTime;
    const velocityX = Math.abs(deltaX) / timeTaken;
    const velocityY = Math.abs(deltaY) / timeTaken;
    const relevantDelta = isHorizontal ? deltaX : deltaY;
    const relevantVelocity = isHorizontal ? velocityX : velocityY;
    const passedThreshold = Math.abs(relevantDelta) >= threshold;
    const hasVelocity = relevantVelocity > VELOCITY_THRESHOLD;
    let isCorrectDirection = false;
    if (direction === "right" && deltaX > 0) isCorrectDirection = true;else if (direction === "left" && deltaX < 0) isCorrectDirection = true;else if (direction === "horizontal" && Math.abs(deltaX) > Math.abs(deltaY)) isCorrectDirection = true;else if (direction === "down" && deltaY > 0) isCorrectDirection = true;else if (direction === "up" && deltaY < 0) isCorrectDirection = true;else if (direction === "vertical" && Math.abs(deltaY) > Math.abs(deltaX)) isCorrectDirection = true;
    const shouldDismiss = isCorrectDirection && (passedThreshold || hasVelocity);
    dragStartRef.current = null;
    lockedDirectionRef.current = null;
    if (shouldDismiss) {
      onDragEnd?.(true);
    } else {
      setDragState({
        isDragging: false,
        offsetX: 0,
        offsetY: 0,
        velocityX: 0,
        velocityY: 0
      });
      onDragCancel?.();
    }
  }, [disabled, direction, threshold, isHorizontal, onDragEnd, onDragCancel]);
  const handlePointerCancel = React.useCallback(() => {
    dragStartRef.current = null;
    lockedDirectionRef.current = null;
    setDragState({
      isDragging: false,
      offsetX: 0,
      offsetY: 0,
      velocityX: 0,
      velocityY: 0
    });
    onDragCancel?.();
  }, [onDragCancel]);
  const gestureHandlers = {
    onPointerDown: handlePointerDown,
    onPointerMove: handlePointerMove,
    onPointerUp: handlePointerUp,
    onPointerCancel: handlePointerCancel
  };
  return {
    dragState,
    gestureHandlers,
    resetDrag: handlePointerCancel
  };
}
