// @tamagui/animate-presence/dist/cjs/AnimatePresence.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var AnimatePresence_exports = {};
__export(AnimatePresence_exports, {
  AnimatePresence: () => AnimatePresence
});
module.exports = __toCommonJS(AnimatePresence_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = require("react");
var import_use_constant = require("@tamagui/use-constant");
var import_use_force_update = require("@tamagui/use-force-update");
var import_react2 = require("react");
var import_LayoutGroupContext = require("./LayoutGroupContext.native.js");
var import_PresenceChild = require("./PresenceChild.native.js");
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var getChildKey = function (child) {
  return child.key || function () {
    var ct = child.type;
    var defaultName = ct["displayName"] || ct["name"] || "";
    if (ct && (typeof ct === "undefined" ? "undefined" : _type_of(ct)) === "object" && "staticConfig" in ct) {
      return ct.staticConfig.componentName || defaultName;
    }
    return defaultName;
  }();
};
function onlyElements(children) {
  var filtered = [];
  import_react2.Children.forEach(children, function (child) {
    if (/* @__PURE__ */(0, import_react2.isValidElement)(child)) filtered.push(child);
  });
  return filtered;
}
var AnimatePresence = function (param) {
  var {
    children,
    enterVariant,
    exitVariant,
    enterExitVariant,
    initial = true,
    onExitComplete,
    exitBeforeEnter,
    mode,
    presenceAffectsLayout = true,
    custom,
    passThrough
  } = param;
  var effectiveMode = mode !== null && mode !== void 0 ? mode : exitBeforeEnter ? "wait" : "sync";
  var presentChildren = (0, import_react2.useMemo)(function () {
    return onlyElements(children);
  }, [children]);
  var presentKeys = presentChildren.map(getChildKey);
  var isInitialRender = (0, import_react2.useRef)(true);
  var frozenCustomRef = (0, import_react2.useRef)(/* @__PURE__ */new Map());
  var pendingPresentChildren = (0, import_react2.useRef)(presentChildren);
  var exitComplete = (0, import_use_constant.useConstant)(function () {
    return /* @__PURE__ */new Map();
  });
  var [diffedChildren, setDiffedChildren] = (0, import_react2.useState)(presentChildren);
  var [renderedChildren, setRenderedChildren] = (0, import_react2.useState)(presentChildren);
  var _useContext_forceRender;
  var forceRender = (_useContext_forceRender = (0, import_react2.useContext)(import_LayoutGroupContext.LayoutGroupContext).forceRender) !== null && _useContext_forceRender !== void 0 ? _useContext_forceRender : (0, import_use_force_update.useForceUpdate)();
  if (passThrough) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
      children
    });
  }
  (0, import_react.useInsertionEffect)(function () {
    isInitialRender.current = false;
    pendingPresentChildren.current = presentChildren;
    for (var i2 = 0; i2 < renderedChildren.length; i2++) {
      var key2 = getChildKey(renderedChildren[i2]);
      if (!presentKeys.includes(key2)) {
        if (exitComplete.get(key2) !== true) {
          exitComplete.set(key2, false);
        }
      } else {
        exitComplete.delete(key2);
        frozenCustomRef.current.delete(key2);
      }
    }
  }, [renderedChildren, presentKeys.length, presentKeys.join("-")]);
  if (presentChildren !== diffedChildren) {
    var nextChildren = [...presentChildren];
    for (var i = 0; i < renderedChildren.length; i++) {
      var child = renderedChildren[i];
      var key = getChildKey(child);
      if (!presentKeys.includes(key)) {
        nextChildren.splice(i, 0, child);
        if (!frozenCustomRef.current.has(key)) {
          frozenCustomRef.current.set(key, custom);
        }
      }
    }
    var exitingChildren = renderedChildren.filter(function (child2) {
      return !presentKeys.includes(getChildKey(child2));
    });
    if (effectiveMode === "wait" && exitingChildren.length) {
      nextChildren = exitingChildren;
    }
    setRenderedChildren(onlyElements(nextChildren));
    setDiffedChildren(presentChildren);
    return null;
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children: renderedChildren.map(function (child2) {
      var key2 = getChildKey(child2);
      var isPresent = presentChildren === renderedChildren || presentKeys.includes(key2);
      var onExit = function () {
        if (exitComplete.has(key2)) {
          exitComplete.set(key2, true);
        } else {
          return;
        }
        var isEveryExitComplete = true;
        exitComplete.forEach(function (isExitComplete) {
          if (!isExitComplete) isEveryExitComplete = false;
        });
        if (isEveryExitComplete) {
          forceRender === null || forceRender === void 0 ? void 0 : forceRender();
          setRenderedChildren(pendingPresentChildren.current);
          onExitComplete === null || onExitComplete === void 0 ? void 0 : onExitComplete();
        }
      };
      var _frozenCustomRef_current_get;
      return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_PresenceChild.PresenceChild, {
        isPresent,
        initial: !isInitialRender.current || initial ? void 0 : false,
        custom: isPresent ? custom : (_frozenCustomRef_current_get = frozenCustomRef.current.get(key2)) !== null && _frozenCustomRef_current_get !== void 0 ? _frozenCustomRef_current_get : custom,
        presenceAffectsLayout,
        enterExitVariant,
        enterVariant,
        exitVariant,
        onExitComplete: isPresent ? void 0 : onExit,
        children: child2
      }, key2);
    })
  });
};
AnimatePresence.displayName = "AnimatePresence";
