// @tamagui/linear-gradient/dist/esm/LinearGradient.mjs
import { normalizeColor, styled, useProps, useTheme } from "@tamagui/core";
import { YStack } from "@tamagui/stacks";
import { LinearGradient as ExpoLinearGradient } from "./linear-gradient.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
const LinearGradientFrame = styled(YStack, {
  name: "LinearGradient",
  overflow: "hidden",
  position: "relative"
});
const LinearGradient = LinearGradientFrame.styleable((propsIn, ref) => {
  const props = useProps(propsIn);
  const {
    start,
    end,
    colors: colorsProp,
    locations,
    children,
    ...stackProps
  } = props;
  const theme = useTheme();
  let colors = props.colors?.map(c => {
    return theme[c]?.get("web") ?? c;
  }) || [];
  if (process.env.NODE_ENV !== "production") {
    if (colors.some(c => {
      const normalized = normalizeColor(c);
      if (!normalized || normalized.startsWith("$")) {
        return true;
      }
    })) {
      console.error(`LinearGradient: "colors" prop contains invalid color tokens: ${colors} fallback to default colors: ["#000", "#fff"]`);
      colors = ["#000", "#fff"];
    }
  }
  return /* @__PURE__ */jsxs(LinearGradientFrame, {
    ref,
    ...stackProps,
    children: [/* @__PURE__ */jsx(ExpoLinearGradient, {
      start,
      end,
      colors,
      locations,
      style: gradientStyle
    }), children]
  });
});
const gradientStyle = {
  position: "absolute",
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  zIndex: 0
};
export { LinearGradient };
