// @tamagui/list-item/dist/jsx/ListItem.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var ListItem_exports = {};
__export(ListItem_exports, {
  ListItem: () => ListItem2
});
module.exports = __toCommonJS(ListItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_font_size = require("@tamagui/font-size");
var import_get_font_sized = require("@tamagui/get-font-sized");
var import_get_token = require("@tamagui/get-token");
var import_helpers = require("@tamagui/helpers");
var import_helpers_tamagui = require("@tamagui/helpers-tamagui");
var import_stacks = require("@tamagui/stacks");
var import_text = require("@tamagui/text");
var import_web = require("@tamagui/web");
var NAME = "ListItem";
var context = (0, import_web.createStyledContext)({
  size: void 0,
  variant: void 0,
  color: void 0
});
var ListItemFrame = (0, import_web.styled)(import_web.View, {
  context,
  name: NAME,
  render: "li",
  role: "listitem",
  variants: {
    unstyled: {
      false: {
        size: "$true",
        alignItems: "center",
        justifyContent: "space-between",
        flexWrap: "nowrap",
        borderColor: "$borderColor",
        width: "100%",
        maxWidth: "100%",
        overflow: "hidden",
        flexDirection: "row",
        backgroundColor: "$background",
        cursor: "default",
        hoverStyle: {
          backgroundColor: "$backgroundHover",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress",
          borderColor: "$borderColorPress"
        }
      }
    },
    variant: {
      outlined: process.env.TAMAGUI_HEADLESS === "1" ? {} : {
        backgroundColor: "transparent",
        borderWidth: 1,
        borderColor: "$borderColor",
        hoverStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorPress"
        }
      }
    },
    size: {
      "...size": function (val, param) {
        var {
          tokens
        } = param;
        return {
          minHeight: tokens.size[val],
          paddingHorizontal: tokens.space[val],
          paddingVertical: (0, import_get_token.getSpace)(tokens.space[val], {
            shift: -4
          })
        };
      }
    },
    active: {
      true: {
        hoverStyle: {
          backgroundColor: "$background"
        }
      }
    },
    disabled: {
      true: {
        opacity: 0.5,
        pointerEvents: "none"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var ListItemText = (0, import_web.styled)(import_text.SizableText, {
  context,
  name: "ListItemText",
  variants: {
    unstyled: {
      false: {
        color: "$color",
        size: "$true",
        flexGrow: 1,
        flexShrink: 1,
        ellipsis: true,
        cursor: "inherit"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var ListItemSubtitle = (0, import_web.styled)(ListItemText, {
  name: "ListItemSubtitle",
  context,
  variants: {
    unstyled: {
      false: {
        opacity: 0.6,
        maxWidth: "100%",
        color: "$color"
      }
    },
    size: {
      "...size": function (val, extras) {
        var oneSmaller = (0, import_get_token.getSize)(val, {
          shift: -1,
          excludeHalfSteps: true
        });
        var fontStyle = (0, import_get_font_sized.getFontSized)(oneSmaller.key, extras);
        return fontStyle;
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var ListItemTitle = (0, import_web.styled)(ListItemText, {
  name: "ListItemTitle",
  context,
  variants: {
    unstyled: {
      false: {}
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var ListItemIcon = function (props) {
  var {
    children,
    size,
    scaleIcon = 1
  } = props;
  var styledContext = context.useStyledContext();
  if (!styledContext) {
    throw new Error("ListItem.Icon must be used within a ListItem");
  }
  var _ref;
  var sizeToken = (_ref = size !== null && size !== void 0 ? size : styledContext.size) !== null && _ref !== void 0 ? _ref : "$true";
  var iconColor = (0, import_helpers_tamagui.useCurrentColor)(styledContext.color);
  var iconSize = (0, import_font_size.getFontSize)(sizeToken) * scaleIcon;
  return (0, import_helpers_tamagui.getIcon)(children, {
    size: iconSize,
    color: iconColor
  });
};
var ListItemComponent = ListItemFrame.styleable(function ListItem(propsIn, ref) {
  var {
    children,
    icon,
    iconAfter,
    scaleIcon = 1,
    unstyled = false,
    subTitle,
    title,
    iconSize,
    ...rest
  } = propsIn;
  var size = propsIn.size || "$true";
  var styledContext = context.useStyledContext();
  var iconColor = (0, import_helpers_tamagui.useCurrentColor)(styledContext === null || styledContext === void 0 ? void 0 : styledContext.color);
  var iconSizeNumber = (0, import_font_size.getFontSize)(iconSize || size) * scaleIcon;
  var [themedIcon, themedIconAfter] = [icon, iconAfter].map(function (icon2, i) {
    if (!icon2) return null;
    var isBefore = i === 0;
    return (0, import_helpers_tamagui.getIcon)(icon2, {
      size: iconSizeNumber,
      color: iconColor,
      style: {
        [isBefore ? "marginRight" : "marginLeft"]: `${iconSizeNumber * 0.4}%`
      }
    });
  });
  var wrappedChildren = (0, import_text.wrapChildrenInText)(ListItemText, {
    children
  }, propsIn.unstyled !== true ? {
    unstyled: process.env.TAMAGUI_HEADLESS === "1",
    fontSize: propsIn.size
  } : void 0);
  return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(ListItemFrame, {
    ref,
    ...rest,
    children: [themedIcon, title || subTitle ? /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_stacks.YStack, {
      flex: 1,
      children: [title ? typeof title === "string" ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(ListItemTitle, {
        unstyled,
        size,
        children: title
      }) : title : null, subTitle ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
        children: typeof subTitle === "string" ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(ListItemSubtitle, {
          unstyled,
          size,
          children: subTitle
        }) : subTitle
      }) : null, wrappedChildren]
    }) : wrappedChildren, themedIconAfter]
  });
});
var ListItem2 = (0, import_helpers.withStaticProperties)(ListItemComponent, {
  Apply: context.Provider,
  Frame: ListItemFrame,
  Text: ListItemText,
  Subtitle: ListItemSubtitle,
  Icon: ListItemIcon,
  Title: ListItemTitle
});
