// @tamagui/stacks/dist/jsx/getElevation.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var getElevation_exports = {};
__export(getElevation_exports, {
  getElevation: () => getElevation,
  getSizedElevation: () => getSizedElevation
});
module.exports = __toCommonJS(getElevation_exports);
var import_core = require("@tamagui/core");
var getElevation = function (size, extras) {
  if (!size) return;
  var {
    tokens
  } = extras;
  var token = tokens.size[size];
  var sizeNum = (0, import_core.isVariable)(token) ? +token.val : size;
  return getSizedElevation(sizeNum, extras);
};
var getSizedElevation = function (val, param) {
  var {
    theme,
    tokens
  } = param;
  var num = 0;
  if (val === true) {
    var _$val = (0, import_core.getVariableValue)(tokens.size["true"]);
    if (typeof _$val === "number") {
      num = _$val;
    } else {
      num = 10;
    }
  } else {
    num = +val;
  }
  if (num === 0) {
    return;
  }
  var [height, shadowRadius] = [Math.round(num / 4 + 1), Math.round(num / 2 + 2)];
  var shadow = {
    shadowColor: theme.shadowColor,
    shadowRadius,
    shadowOffset: {
      height,
      width: 0
    },
    ...(import_core.isAndroid ? {
      elevationAndroid: 2 * height
    } : {})
  };
  return shadow;
};
