// @tamagui/stacks/dist/jsx/SizableStack.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SizableStack_exports = {};
__export(SizableStack_exports, {
  SizableStack: () => SizableStack
});
module.exports = __toCommonJS(SizableStack_exports);
var import_core = require("@tamagui/core");
var import_get_button_sized = require("@tamagui/get-button-sized");
var import_ThemeableStack = require("./ThemeableStack.native.js");
var import_variants = require("./variants.native.js");
var SizableStack = (0, import_core.styled)(import_ThemeableStack.ThemeableStack, {
  name: "SizableStack",
  variants: {
    unstyled: {
      true: {
        elevate: false,
        bordered: false
      }
    },
    circular: import_variants.circular,
    elevate: import_variants.elevate,
    bordered: {
      true: import_variants.bordered
    },
    size: {
      "...size": function (val, extras) {
        return (0, import_get_button_sized.getButtonSized)(val, extras);
      }
    }
  }
});
