// @tamagui/tooltip/dist/esm/Tooltip.mjs
import { FloatingDelayGroup, useDelayGroupContext } from "@tamagui/floating";
import { useEvent } from "@tamagui/core";
import { FloatingOverrideContext } from "@tamagui/floating";
import { getSize } from "@tamagui/get-token";
import { withStaticProperties } from "@tamagui/helpers";
import { PopoverAnchor, PopoverArrow, PopoverContent, PopoverContextProvider, PopoverTrigger, useFloatingContext } from "@tamagui/popover";
import { Popper, PopperContentFrame } from "@tamagui/popper";
import { useControllableState } from "@tamagui/use-controllable-state";
import * as React from "react";
import { jsx } from "react/jsx-runtime";
const TOOLTIP_SCOPE = "";
const ALWAYS_DISABLE_TOOLTIP = {
  focus: true,
  "remove-scroll": true
  // it's nice to hit escape to hide a tooltip
  // dismiss: true
};
const TooltipContent = PopperContentFrame.styleable((props, ref) => {
  const preventAnimation = React.useContext(PreventTooltipAnimationContext);
  const zIndexFromContext = React.useContext(TooltipZIndexContext);
  return /* @__PURE__ */jsx(PopoverContent, {
    scope: props.scope || TOOLTIP_SCOPE,
    alwaysDisable: ALWAYS_DISABLE_TOOLTIP,
    ...(!props.unstyled && {
      backgroundColor: "$background",
      alignItems: "center",
      pointerEvents: "none",
      size: "$true"
    }),
    ref,
    ...(zIndexFromContext !== void 0 && {
      zIndex: zIndexFromContext
    }),
    ...props,
    ...(preventAnimation && {
      transition: null
    })
  });
}, {
  staticConfig: {
    componentName: "Tooltip"
  }
});
const TooltipArrow = React.forwardRef((props, ref) => {
  return /* @__PURE__ */jsx(PopoverArrow, {
    scope: props.scope || TOOLTIP_SCOPE,
    componentName: "Tooltip",
    ref,
    ...props
  });
});
const PreventTooltipAnimationContext = React.createContext(false);
const TooltipZIndexContext = React.createContext(void 0);
const TooltipGroup = ({
  children,
  delay,
  preventAnimation = false,
  timeoutMs
}) => {
  return /* @__PURE__ */jsx(PreventTooltipAnimationContext.Provider, {
    value: preventAnimation,
    children: /* @__PURE__ */jsx(FloatingDelayGroup, {
      timeoutMs,
      delay: React.useMemo(() => delay, [JSON.stringify(delay)]),
      children
    })
  });
};
const setOpens = /* @__PURE__ */new Set();
const closeOpenTooltips = () => {
  setOpens.forEach(x => x(false));
};
const TooltipComponent = React.forwardRef(function Tooltip(props, ref) {
  "use no memo";

  const {
    children,
    delay: delayProp,
    restMs: restMsProp,
    onOpenChange: onOpenChangeProp,
    focus,
    open: openProp,
    disableAutoCloseOnScroll,
    zIndex,
    scope = TOOLTIP_SCOPE,
    ...restProps
  } = props;
  const triggerRef = React.useRef(null);
  const [hasCustomAnchor, setHasCustomAnchor] = React.useState(false);
  const {
    delay: delayGroup,
    setCurrentId
  } = useDelayGroupContext();
  const delay = delayProp !== void 0 ? delayProp : delayGroup ?? 400;
  const restMs = restMsProp ?? (typeof delay === "number" ? delay : 0);
  const [open, setOpen] = useControllableState({
    prop: openProp,
    defaultProp: false,
    onChange: onOpenChangeProp
  });
  const id = props.groupId;
  const onOpenChange = useEvent(open2 => {
    if (open2) {
      setCurrentId(id);
    }
    setOpen(open2);
  });
  React.useEffect(() => {
    if (!open) return;
    if (disableAutoCloseOnScroll) return;
    if (typeof document === "undefined") return;
    const closeIt = () => {
      setOpen(false);
    };
    setOpens.add(setOpen);
    document.documentElement.addEventListener("scroll", closeIt);
    return () => {
      setOpens.delete(setOpen);
      document.documentElement.removeEventListener("scroll", closeIt);
    };
  }, [open, disableAutoCloseOnScroll]);
  const floatingContext = useFloatingContext({
    open,
    setOpen: onOpenChange,
    disable: false,
    disableFocus: false,
    hoverable: true,
    role: "tooltip",
    focus,
    groupId: id,
    delay,
    restMs
  });
  const onCustomAnchorAdd = React.useCallback(() => setHasCustomAnchor(true), []);
  const onCustomAnchorRemove = React.useCallback(() => setHasCustomAnchor(false), []);
  const contentId = React.useId();
  const smallerSize = props.unstyled ? null : getSize("$true", {
    shift: -2,
    bounds: [0]
  });
  const content = /* @__PURE__ */jsx(FloatingOverrideContext.Provider, {
    value: floatingContext,
    children: /* @__PURE__ */jsx(Popper, {
      scope,
      size: smallerSize?.key,
      allowFlip: true,
      stayInFrame: true,
      open,
      ...restProps,
      children: /* @__PURE__ */jsx(PopoverContextProvider, {
        scope,
        contentId,
        triggerRef,
        open,
        onOpenChange: setOpen,
        onOpenToggle: voidFn,
        hasCustomAnchor,
        onCustomAnchorAdd,
        onCustomAnchorRemove,
        children
      })
    })
  });
  if (zIndex !== void 0) {
    return /* @__PURE__ */jsx(TooltipZIndexContext.Provider, {
      value: zIndex,
      children: content
    });
  }
  return content;
});
const TooltipTrigger = React.forwardRef(function TooltipTrigger2(props, ref) {
  const {
    scope,
    ...rest
  } = props;
  return /* @__PURE__ */jsx(PopoverTrigger, {
    ...rest,
    scope: scope || TOOLTIP_SCOPE,
    ref
  });
});
const TooltipAnchor = React.forwardRef(function TooltipAnchor2(props, ref) {
  const {
    scope,
    ...rest
  } = props;
  return /* @__PURE__ */jsx(PopoverAnchor, {
    ...rest,
    scope: scope || TOOLTIP_SCOPE,
    ref
  });
});
const Tooltip2 = withStaticProperties(TooltipComponent, {
  Anchor: TooltipAnchor,
  Arrow: TooltipArrow,
  Content: TooltipContent,
  Trigger: TooltipTrigger
});
const voidFn = () => {};
export { Tooltip2 as Tooltip, TooltipGroup, closeOpenTooltips };
