// @tamagui/sheet/dist/esm/createSheet.native.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useComposedRefs } from "@tamagui/compose-refs";
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { View } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { resolveViewZIndex } from "@tamagui/portal";
import { RemoveScroll } from "@tamagui/remove-scroll";
import { useDidFinishSSR } from "@tamagui/use-did-finish-ssr";
import { StackZIndexContext } from "@tamagui/z-index-stack";
import { forwardRef, useMemo, useEffect, useRef } from "react";
import { Platform } from "react-native";
import { SHEET_HANDLE_NAME, SHEET_NAME, SHEET_OVERLAY_NAME } from "./constants.native.js";
import { getNativeSheet } from "./nativeSheet.native.js";
import { useSheetContext } from "./SheetContext.native.js";
import { SheetImplementationCustom } from "./SheetImplementationCustom.native.js";
import { SheetScrollView } from "./SheetScrollView.native.js";
import { useSheetController } from "./useSheetController.native.js";
import { useSheetOffscreenSize } from "./useSheetOffscreenSize.native.js";
function createSheet(param) {
  var {
    Handle,
    Frame,
    Overlay
  } = param;
  var SheetHandle = Handle.styleable(function (param2, forwardedRef) {
    var {
      __scopeSheet,
      ...props
    } = param2;
    var context = useSheetContext(SHEET_HANDLE_NAME, __scopeSheet);
    var composedRef = useComposedRefs(context.handleRef, forwardedRef);
    var wasDraggingRef = useRef(false);
    useEffect(function () {
      if (!context.scrollBridge) return;
      return context.scrollBridge.onParentDragging(function (isDragging) {
        if (isDragging) wasDraggingRef.current = true;
      });
    }, [context.scrollBridge]);
    if (context.onlyShowFrame) return null;
    return /* @__PURE__ */_jsx(Handle, {
      ref: composedRef,
      onPressIn: function () {
        wasDraggingRef.current = false;
      },
      onPress: function () {
        if (wasDraggingRef.current) {
          wasDraggingRef.current = false;
          return;
        }
        var max = context.snapPoints.length + (context.dismissOnSnapToBottom ? -1 : 0);
        var nextPos = (context.position + 1) % max;
        context.setPosition(nextPos);
      },
      open: context.open,
      ...props
    });
  });
  var SheetOverlay = Overlay.styleable(function (propsIn, ref) {
    var {
      __scopeSheet,
      ...props
    } = propsIn;
    var context = useSheetContext(SHEET_OVERLAY_NAME, __scopeSheet);
    var element = useMemo(function () {
      return /* @__PURE__ */_jsx(Overlay, {
        ...props,
        onPress: composeEventHandlers(props.onPress, context.dismissOnOverlayPress ? function () {
          context.setOpen(false);
        } : void 0)
      });
    }, [props.onPress, props.opacity, context.dismissOnOverlayPress]);
    useIsomorphicLayoutEffect(function () {
      var _context_onOverlayComponent;
      (_context_onOverlayComponent = context.onOverlayComponent) === null || _context_onOverlayComponent === void 0 || _context_onOverlayComponent.call(context, element);
    }, [element]);
    if (context.onlyShowFrame) return null;
    return null;
  });
  var SheetFrame = Frame.styleable(function (param2, forwardedRef) {
    var {
      __scopeSheet,
      adjustPaddingForOffscreenContent,
      disableHideBottomOverflow,
      children,
      ...props
    } = param2;
    var context = useSheetContext(SHEET_NAME, __scopeSheet);
    var {
      hasFit,
      disableRemoveScroll,
      frameSize,
      contentRef,
      open
    } = context;
    var composedContentRef = useComposedRefs(forwardedRef, contentRef);
    var offscreenSize = useSheetOffscreenSize(context);
    var stableFrameSize = useRef(frameSize);
    useEffect(function () {
      if (open && frameSize) stableFrameSize.current = frameSize;
    }, [open, frameSize]);
    var sheetContents = useMemo(function () {
      var shouldUseFixedHeight = hasFit && !open && stableFrameSize.current;
      return /* @__PURE__ */_jsxs(Frame, {
        ref: composedContentRef,
        flex: hasFit && open ? 0 : 1,
        flexBasis: hasFit ? "auto" : void 0,
        height: shouldUseFixedHeight ? stableFrameSize.current : hasFit ? void 0 : frameSize,
        pointerEvents: open ? "auto" : "none",
        "data-state": open ? "open" : "closed",
        ...props,
        children: [/* @__PURE__ */_jsx(StackZIndexContext, {
          zIndex: resolveViewZIndex(props.zIndex),
          children
        }), adjustPaddingForOffscreenContent && /* @__PURE__ */_jsx(View, {
          "data-sheet-offscreen-pad": true,
          height: offscreenSize,
          width: "100%"
        })]
      });
    }, [open, props, frameSize, offscreenSize, adjustPaddingForOffscreenContent, hasFit]);
    return /* @__PURE__ */_jsxs(_Fragment, {
      children: [/* @__PURE__ */_jsx(RemoveScroll, {
        enabled: !disableRemoveScroll && context.open,
        children: sheetContents
      }), !disableHideBottomOverflow && /* @__PURE__ */_jsx(Frame, {
        ...props,
        componentName: "SheetCover",
        children: null,
        testID: void 0,
        id: void 0,
        position: "absolute",
        bottom: "-100%",
        zIndex: -1,
        height: context.frameSize,
        left: 0,
        right: 0,
        borderWidth: 0,
        borderRadius: 0,
        shadowOpacity: 0
      })]
    });
  });
  var Sheet = /* @__PURE__ */forwardRef(function Sheet2(props, ref) {
    var hydrated = useDidFinishSSR();
    var {
      isShowingNonSheet
    } = useSheetController();
    var SheetImplementation = SheetImplementationCustom;
    if (props.native && Platform.OS === "ios") {
      var impl = getNativeSheet("ios");
      if (impl) SheetImplementation = impl;
    }
    if (isShowingNonSheet || !hydrated) return null;
    return /* @__PURE__ */_jsx(SheetImplementation, {
      ref,
      ...props
    });
  });
  var components = {
    Frame: SheetFrame,
    Overlay: SheetOverlay,
    Handle: SheetHandle,
    ScrollView: SheetScrollView
  };
  var Controlled = withStaticProperties(Sheet, components);
  return withStaticProperties(Sheet, {
    ...components,
    Controlled
  });
}
export { createSheet };
