// @tamagui/sheet/dist/esm/SheetScrollView.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { composeRefs } from "@tamagui/compose-refs";
import { View } from "@tamagui/core";
import { ScrollView } from "@tamagui/scroll-view";
import { useControllableState } from "@tamagui/use-controllable-state";
import React, { useEffect, useRef, useState } from "react";
import { useGestureSheetContext } from "./GestureSheetContext.native.js";
import { getGestureHandlerState, isGestureHandlerEnabled } from "./gestureState.native.js";
import { useSheetContext } from "./SheetContext.native.js";
import { useSheetScrollViewGestures } from "./useSheetScrollViewGestures.native.js";
var SHEET_SCROLL_VIEW_NAME = "SheetScrollView";
var SheetScrollView = /* @__PURE__ */React.forwardRef(function (param, ref) {
  var {
    __scopeSheet,
    children,
    onScroll,
    scrollEnabled: scrollEnabledProp,
    ...props
  } = param;
  var context = useSheetContext(SHEET_SCROLL_VIEW_NAME, __scopeSheet);
  var gestureContext = useGestureSheetContext();
  var {
    scrollBridge,
    setHasScrollView
  } = context;
  var [scrollEnabled] = useControllableState({
    prop: scrollEnabledProp,
    defaultProp: true
  });
  var scrollRef = React.useRef(null);
  var panGestureRef = gestureContext === null || gestureContext === void 0 ? void 0 : gestureContext.panGestureRef;
  var {
    ScrollView: RNGHScrollView
  } = getGestureHandlerState();
  var useRNGHScrollView = isGestureHandlerEnabled() && RNGHScrollView && panGestureRef;
  var currentScrollOffset = useRef(0);
  var lockedScrollY = useRef(0);
  var setScrollEnabled = function (next, lockTo) {
    if (!next) {
      var _scrollRef_current_scrollTo, _scrollRef_current;
      var lockY = lockTo !== null && lockTo !== void 0 ? lockTo : currentScrollOffset.current;
      lockedScrollY.current = lockY;
      scrollBridge.scrollLockY = lockY;
      (_scrollRef_current = scrollRef.current) === null || _scrollRef_current === void 0 ? void 0 : (_scrollRef_current_scrollTo = _scrollRef_current.scrollTo) === null || _scrollRef_current_scrollTo === void 0 ? void 0 : _scrollRef_current_scrollTo.call(_scrollRef_current, {
        x: 0,
        y: lockY,
        animated: false
      });
    } else {
      lockedScrollY.current = currentScrollOffset.current;
      scrollBridge.scrollLockY = void 0;
    }
  };
  var forceScrollTo = function (y) {
    var _scrollRef_current_scrollTo, _scrollRef_current;
    (_scrollRef_current = scrollRef.current) === null || _scrollRef_current === void 0 ? void 0 : (_scrollRef_current_scrollTo = _scrollRef_current.scrollTo) === null || _scrollRef_current_scrollTo === void 0 ? void 0 : _scrollRef_current_scrollTo.call(_scrollRef_current, {
      x: 0,
      y,
      animated: false
    });
  };
  useEffect(function () {
    setHasScrollView(true);
    if (isGestureHandlerEnabled()) {
      scrollBridge.setScrollEnabled = setScrollEnabled;
      scrollBridge.forceScrollTo = forceScrollTo;
    }
    return function () {
      setHasScrollView(false);
      scrollBridge.setScrollEnabled = void 0;
      scrollBridge.forceScrollTo = void 0;
    };
  }, []);
  var [hasScrollableContent, setHasScrollableContent] = useState(true);
  var parentHeight = useRef(0);
  var contentHeight = useRef(0);
  var updateScrollable = function () {
    if (parentHeight.current && contentHeight.current) {
      setHasScrollableContent(contentHeight.current > parentHeight.current);
    }
  };
  useEffect(function () {
    scrollBridge.hasScrollableContent = hasScrollableContent;
  }, [hasScrollableContent]);
  var gestureProps = useSheetScrollViewGestures({
    scrollRef,
    scrollBridge,
    hasScrollableContent,
    scrollEnabled,
    setScrollEnabled
  });
  var contentWrapper = /* @__PURE__ */_jsx(View, {
    onLayout: function (e) {
      var height = Math.floor(e.nativeEvent.layout.height);
      if (height !== contentHeight.current) {
        contentHeight.current = height;
        updateScrollable();
      }
    },
    children
  });
  if (useRNGHScrollView && RNGHScrollView && panGestureRef) {
    var RNGHComponent = RNGHScrollView;
    return /* @__PURE__ */_jsx(RNGHComponent, {
      ref: composeRefs(scrollRef, ref),
      style: {
        flex: 1
      },
      scrollEventThrottle: 1,
      scrollEnabled,
      simultaneousHandlers: [panGestureRef],
      onLayout: function (e) {
        parentHeight.current = Math.ceil(e.nativeEvent.layout.height);
        updateScrollable();
      },
      onScroll: function (e) {
        var {
          y
        } = e.nativeEvent.contentOffset;
        currentScrollOffset.current = y;
        if (scrollBridge.scrollLockY !== void 0) {
          if (y !== scrollBridge.scrollLockY) {
            var _scrollRef_current_scrollTo, _scrollRef_current;
            (_scrollRef_current = scrollRef.current) === null || _scrollRef_current === void 0 ? void 0 : (_scrollRef_current_scrollTo = _scrollRef_current.scrollTo) === null || _scrollRef_current_scrollTo === void 0 ? void 0 : _scrollRef_current_scrollTo.call(_scrollRef_current, {
              x: 0,
              y: scrollBridge.scrollLockY,
              animated: false
            });
          }
          scrollBridge.y = scrollBridge.scrollLockY;
          onScroll === null || onScroll === void 0 ? void 0 : onScroll({
            ...e,
            nativeEvent: {
              ...e.nativeEvent,
              contentOffset: {
                ...e.nativeEvent.contentOffset,
                y: scrollBridge.scrollLockY
              }
            }
          });
          return;
        }
        scrollBridge.y = y;
        if (y > 0) scrollBridge.scrollStartY = -1;
        onScroll === null || onScroll === void 0 ? void 0 : onScroll(e);
      },
      contentContainerStyle: {
        minHeight: "100%"
      },
      bounces: false,
      keyboardShouldPersistTaps: "always",
      keyboardDismissMode: "none",
      ...props,
      children: contentWrapper
    });
  }
  return /* @__PURE__ */_jsx(ScrollView, {
    onLayout: function (e) {
      parentHeight.current = Math.ceil(e.nativeEvent.layout.height);
      updateScrollable();
    },
    ref: composeRefs(scrollRef, ref),
    flex: 1,
    scrollEventThrottle: 1,
    className: "_ovs-contain",
    scrollEnabled,
    onScroll: function (e) {
      var {
        y
      } = e.nativeEvent.contentOffset;
      scrollBridge.y = y;
      if (y > 0) scrollBridge.scrollStartY = -1;
      onScroll === null || onScroll === void 0 ? void 0 : onScroll(e);
    },
    contentContainerStyle: {
      minHeight: "100%"
    },
    ...gestureProps,
    ...props,
    children: contentWrapper
  });
});
export { SheetScrollView };
