// @tamagui/sheet/dist/esm/useGestureHandlerPan.native.js
import { useCallback, useMemo, useRef } from "react";
import { getGestureHandlerState, isGestureHandlerEnabled } from "./gestureState.native.js";
var AT_TOP_THRESHOLD = 5;
function useGestureHandlerPan(config) {
  var {
    positions,
    frameSize,
    setPosition,
    animateTo,
    stopSpring,
    scrollBridge,
    setIsDragging,
    getCurrentPosition,
    resisted,
    disableDrag,
    isShowingInnerSheet,
    setAnimatedPosition,
    scrollGestureRef
  } = config;
  var gestureHandlerEnabled = isGestureHandlerEnabled();
  var panGestureRef = useRef(null);
  var gestureStateRef = useRef({
    startY: 0,
    // track last translation when pan was active (for position calculation after handoff)
    lastPanTranslationY: 0,
    // accumulated position offset from all pan movements
    accumulatedOffset: 0,
    // track previous translation for direction detection (like actions-sheet)
    prevTranslationY: 0,
    // track if scroll was engaged (scrollY > 0) at some point
    scrollEngaged: false,
    // positions frozen at gesture start — keyboard may dismiss during drag (input blur),
    // causing positions to revert. Frozen positions ensure stable snap calculation.
    frozenPositions: [],
    frozenMinY: 0,
    // whether pan gesture actually started (vs just a tap in onBegin)
    panStarted: false
  });
  var onStart = useCallback(function () {
    stopSpring();
  }, [stopSpring]);
  var onEnd = useCallback(function (closestPoint, animationOverride) {
    var _scrollBridge_setScrollEnabled;
    setIsDragging(false);
    scrollBridge.setParentDragging(false);
    (_scrollBridge_setScrollEnabled = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled === void 0 ? void 0 : _scrollBridge_setScrollEnabled.call(scrollBridge, true);
    setPosition(closestPoint);
    animateTo(closestPoint, animationOverride);
  }, [setIsDragging, scrollBridge, setPosition, animateTo]);
  var panGesture = useMemo(function () {
    if (!gestureHandlerEnabled || disableDrag || isShowingInnerSheet || !frameSize) {
      return null;
    }
    var {
      Gesture
    } = getGestureHandlerState();
    if (!Gesture) {
      return null;
    }
    var minY = positions[0];
    var gs = gestureStateRef.current;
    var gesture = Gesture.Pan().withRef(panGestureRef).activeOffsetY([-10, 10]).failOffsetX([-20, 20]).shouldCancelWhenOutside(false).onBegin(function () {
      gs.panStarted = false;
      if (config.pauseKeyboardHandler) {
        config.pauseKeyboardHandler.current = true;
      }
      var pos = getCurrentPosition();
      var atTop = pos <= minY + AT_TOP_THRESHOLD;
      var currentScrollY = scrollBridge.y;
      gs.startY = pos;
      gs.lastPanTranslationY = 0;
      gs.accumulatedOffset = 0;
      gs.prevTranslationY = 0;
      gs.scrollEngaged = currentScrollY > 0;
      gs.frozenPositions = [...positions];
      gs.frozenMinY = minY;
      if (!atTop) {
        var _scrollBridge_setScrollEnabled;
        (_scrollBridge_setScrollEnabled = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled === void 0 ? void 0 : _scrollBridge_setScrollEnabled.call(scrollBridge, false, 0);
      }
    }).onStart(function () {
      gs.panStarted = true;
      setIsDragging(true);
      scrollBridge.initialPosition = gs.startY;
      onStart();
    }).onChange(function (event) {
      var {
        translationY
      } = event;
      var isSwipingDown = gs.prevTranslationY < translationY;
      var deltaY = translationY - gs.prevTranslationY;
      gs.prevTranslationY = translationY;
      var scrollY = scrollBridge.y;
      if (scrollY > 0) {
        gs.scrollEngaged = true;
      }
      var currentPos = gs.startY + gs.accumulatedOffset;
      var isCurrentlyAtTop = currentPos <= minY + AT_TOP_THRESHOLD;
      var nodeIsScrolling = scrollY > 0;
      var panHandles = false;
      var hasScrollableContent = scrollBridge.hasScrollableContent !== false;
      if (!isCurrentlyAtTop) {
        if (isSwipingDown) {
          panHandles = !nodeIsScrolling || !hasScrollableContent;
        } else {
          panHandles = true;
        }
      } else {
        if (isSwipingDown) {
          if (nodeIsScrolling && hasScrollableContent) {
            panHandles = false;
          } else if (gs.scrollEngaged && hasScrollableContent) {
            panHandles = true;
          } else {
            panHandles = true;
          }
        } else {
          if (hasScrollableContent) {
            panHandles = false;
          } else {
            panHandles = true;
          }
        }
      }
      if (panHandles) {
        var _scrollBridge_setScrollEnabled;
        var lockTo = isCurrentlyAtTop && isSwipingDown && gs.scrollEngaged ? void 0 : 0;
        (_scrollBridge_setScrollEnabled = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled === void 0 ? void 0 : _scrollBridge_setScrollEnabled.call(scrollBridge, false, lockTo);
        gs.accumulatedOffset += deltaY;
        var newPosition = resisted(gs.startY + gs.accumulatedOffset, minY);
        scrollBridge.paneY = newPosition;
        setAnimatedPosition(newPosition);
        scrollBridge.setParentDragging(newPosition > minY);
      } else {
        var _scrollBridge_setScrollEnabled1;
        (_scrollBridge_setScrollEnabled1 = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled1 === void 0 ? void 0 : _scrollBridge_setScrollEnabled1.call(scrollBridge, true);
      }
    }).onEnd(function (event) {
      var {
        velocityY
      } = event;
      var currentPos = gs.startY + gs.accumulatedOffset;
      scrollBridge.scrollLockY = void 0;
      var snapPositions = gs.frozenPositions.length > 0 ? gs.frozenPositions : positions;
      var snapMinY = gs.frozenPositions.length > 0 ? gs.frozenMinY : minY;
      if (currentPos <= snapMinY + AT_TOP_THRESHOLD && scrollBridge.y > 0) {
        onEnd(0);
        return;
      }
      var velocity = velocityY / 1e3;
      var projectedEnd = currentPos + frameSize * velocity * 0.2;
      var closestPoint = 0;
      var minDist = Number.POSITIVE_INFINITY;
      for (var i = 0; i < snapPositions.length; i++) {
        var pos = snapPositions[i];
        var dist = Math.abs(projectedEnd - pos);
        if (dist < minDist) {
          minDist = dist;
          closestPoint = i;
        }
      }
      onEnd(closestPoint);
    }).onFinalize(function () {
      scrollBridge.scrollLockY = void 0;
      if (gs.panStarted) {
        setIsDragging(false);
      } else {
        if (config.pauseKeyboardHandler) {
          config.pauseKeyboardHandler.current = false;
        }
      }
    }).runOnJS(true);
    if (scrollGestureRef === null || scrollGestureRef === void 0 ? void 0 : scrollGestureRef.current) {
      return gesture.simultaneousWithExternalGesture(scrollGestureRef.current);
    }
    return gesture;
  }, [gestureHandlerEnabled, disableDrag, isShowingInnerSheet, frameSize, positions, scrollBridge, getCurrentPosition, resisted, onStart, onEnd, setIsDragging, setAnimatedPosition]);
  return {
    panGesture,
    panGestureRef,
    gestureHandlerEnabled
  };
}
export { useGestureHandlerPan };
