// @tamagui/sheet/dist/esm/useSheetScrollViewGestures.native.js
import { useRef } from "react";
function useSheetScrollViewGestures(param) {
  var {
    scrollBridge,
    hasScrollableContent,
    scrollEnabled,
    setScrollEnabled
  } = param;
  var state = useRef({
    lastPageY: 0,
    dragAt: 0,
    dys: [],
    isScrolling: false,
    isDraggingScrollArea: false,
    scrollEngaged: false,
    prevScrollY: 0,
    handoffOccurred: false,
    handoffDragOffset: 0
  });
  var release = function () {
    var s = state.current;
    if (!s.isDraggingScrollArea) return;
    s.isDraggingScrollArea = false;
    scrollBridge.scrollStartY = -1;
    scrollBridge.scrollLock = false;
    s.isScrolling = false;
    s.scrollEngaged = false;
    s.prevScrollY = 0;
    s.handoffOccurred = false;
    s.handoffDragOffset = 0;
    setScrollEnabled(true);
    var vy = 0;
    if (s.dys.length) {
      var recentDys = s.dys.slice(-10);
      var dist = recentDys.reduce(function (a, b) {
        return a + b;
      }, 0);
      vy = dist / recentDys.length * 0.04;
    }
    s.dys = [];
    scrollBridge.release({
      dragAt: s.dragAt,
      vy
    });
  };
  var onStartShouldSetResponder = function () {
    var s = state.current;
    scrollBridge.scrollStartY = -1;
    s.isDraggingScrollArea = true;
    s.scrollEngaged = scrollBridge.y > 0;
    s.prevScrollY = scrollBridge.y;
    s.handoffOccurred = false;
    s.handoffDragOffset = 0;
    return false;
  };
  var onMoveShouldSetResponder = function (e) {
    if (!scrollEnabled) return false;
    var s = state.current;
    var {
      pageY
    } = e.nativeEvent;
    if (s.lastPageY === 0) {
      s.lastPageY = pageY;
      return false;
    }
    return Math.abs(pageY - s.lastPageY) > 10;
  };
  var onResponderMove = function (e) {
    var s = state.current;
    var {
      pageY
    } = e.nativeEvent;
    if (!s.isScrolling && scrollBridge.scrollStartY === -1) {
      scrollBridge.scrollStartY = pageY;
      s.lastPageY = pageY;
    }
    var dragAt = pageY - scrollBridge.scrollStartY;
    var dy = pageY - s.lastPageY;
    s.lastPageY = pageY;
    var isDraggingDown = dy > 0;
    var isPaneAtTop = scrollBridge.paneY <= scrollBridge.paneMinY + 5;
    var currentScrollY = scrollBridge.y;
    if (currentScrollY > 0) s.scrollEngaged = true;
    var wasScrolledNowAtZero = s.scrollEngaged && s.prevScrollY > 0 && currentScrollY <= 0;
    if (wasScrolledNowAtZero && !s.handoffOccurred) {
      s.handoffOccurred = true;
      s.handoffDragOffset = 0;
    }
    s.prevScrollY = currentScrollY;
    var panHandles = false;
    if (!isPaneAtTop) {
      panHandles = isDraggingDown ? currentScrollY <= 0 || !hasScrollableContent : true;
    } else if (isDraggingDown) {
      if (currentScrollY > 0 && hasScrollableContent && !s.handoffOccurred) {
        panHandles = false;
      } else {
        panHandles = s.handoffOccurred || wasScrolledNowAtZero || currentScrollY <= 0;
      }
    } else {
      panHandles = !hasScrollableContent;
      if (!panHandles) {
        s.isScrolling = true;
        scrollBridge.scrollLock = true;
        setScrollEnabled(true);
        return;
      }
    }
    if (panHandles) {
      setScrollEnabled(false);
      var effectiveDragAt = dragAt;
      if (s.handoffOccurred) {
        s.handoffDragOffset += dy;
        effectiveDragAt = s.handoffDragOffset;
      }
      scrollBridge.drag(effectiveDragAt);
      s.dragAt = effectiveDragAt;
      s.dys.push(dy);
      if (s.dys.length > 100) s.dys = s.dys.slice(-10);
    } else {
      setScrollEnabled(true);
    }
  };
  return {
    onResponderRelease: release,
    onStartShouldSetResponder,
    onMoveShouldSetResponder,
    onResponderMove
  };
}
export { useSheetScrollViewGestures };
