// @tamagui/button/dist/esm/Button.native.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { getFontSize } from "@tamagui/font-size";
import { getButtonSized } from "@tamagui/get-button-sized";
import { getIcon, useCurrentColor } from "@tamagui/helpers-tamagui";
import { ButtonNestingContext, getElevation, themeableVariants } from "@tamagui/stacks";
import { SizableText, wrapChildrenInText } from "@tamagui/text";
import { createStyledContext, getTokenValue, styled, useProps, View, withStaticProperties } from "@tamagui/web";
import { useContext } from "react";
var context = createStyledContext({
  size: void 0,
  variant: void 0,
  color: void 0,
  elevation: void 0
});
var Frame = styled(View, {
  context,
  name: "Button",
  role: "button",
  render: /* @__PURE__ */_jsx("button", {
    type: "button"
  }),
  tabIndex: 0,
  variants: {
    unstyled: {
      false: {
        size: "$true",
        justifyContent: "center",
        alignItems: "center",
        flexWrap: "nowrap",
        flexDirection: "row",
        cursor: "pointer",
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "transparent",
        hoverStyle: {
          backgroundColor: "$backgroundHover",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress",
          borderColor: "$borderColorHover"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineStyle: "solid",
          outlineWidth: 2
        }
      }
    },
    variant: {
      outlined: process.env.TAMAGUI_HEADLESS === "1" ? {} : {
        backgroundColor: "transparent",
        borderWidth: 1,
        borderColor: "$borderColor",
        hoverStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorPress"
        }
      }
    },
    circular: themeableVariants.circular,
    chromeless: themeableVariants.chromeless,
    size: {
      "...size": function (val, extras) {
        var buttonStyle = getButtonSized(val, extras);
        var gap = getTokenValue(val);
        return {
          ...buttonStyle,
          gap
        };
      },
      ":number": function (val, extras) {
        var buttonStyle = getButtonSized(val, extras);
        var gap = val * 0.4;
        return {
          ...buttonStyle,
          gap
        };
      }
    },
    elevation: {
      "...size": getElevation,
      ":number": getElevation
    },
    disabled: {
      true: {
        pointerEvents: "none",
        // @ts-ignore
        "aria-disabled": true
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var Text = styled(SizableText, {
  context,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        cursor: "pointer",
        // flexGrow 1 leads to inconsistent native style where text pushes to start of view
        flexGrow: 0,
        flexShrink: 1,
        ellipsis: true,
        color: "$color"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var Icon = function (props) {
  var {
    children,
    scaleIcon = 1,
    size
  } = props;
  var styledContext = context.useStyledContext();
  if (!styledContext) {
    throw new Error("Button.Icon must be used within a Button");
  }
  var sizeToken = size !== null && size !== void 0 ? size : styledContext.size;
  var iconColor = useCurrentColor(styledContext.color);
  var iconSize = (typeof sizeToken === "number" ? sizeToken * 0.5 : getFontSize(sizeToken)) * scaleIcon;
  return getIcon(children, {
    size: iconSize,
    color: iconColor
  });
};
var ButtonContext = createStyledContext({
  size: void 0,
  variant: void 0,
  color: void 0
});
var ButtonComponent = Frame.styleable(function (propsIn, ref) {
  var isNested = useContext(ButtonNestingContext);
  var processedProps = useProps(propsIn, {
    noNormalize: true,
    noExpand: true
  });
  var {
    children,
    iconSize,
    icon,
    iconAfter,
    scaleIcon = 1,
    ...props
  } = processedProps;
  var size = propsIn.size || (propsIn.unstyled ? void 0 : "$true");
  var styledContext = context.useStyledContext();
  var iconColor = useCurrentColor(styledContext === null || styledContext === void 0 ? void 0 : styledContext.color);
  var _ref;
  var finalSize = (_ref = iconSize !== null && iconSize !== void 0 ? iconSize : size) !== null && _ref !== void 0 ? _ref : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size;
  var iconSizeNumber = (typeof finalSize === "number" ? finalSize * 0.5 : getFontSize(finalSize)) * scaleIcon;
  var [themedIcon, themedIconAfter] = [icon, iconAfter].map(function (icon2) {
    if (!icon2) return null;
    return getIcon(icon2, {
      size: iconSizeNumber,
      color: iconColor
    });
  });
  var wrappedChildren = wrapChildrenInText(Text, {
    children
  }, {
    unstyled: process.env.TAMAGUI_HEADLESS === "1",
    size: finalSize !== null && finalSize !== void 0 ? finalSize : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size
  });
  return /* @__PURE__ */_jsx(ButtonNestingContext.Provider, {
    value: true,
    children: /* @__PURE__ */_jsxs(Frame, {
      ref,
      ...props,
      ...(isNested && {
        render: "span"
      }),
      // Pass resolved size to circular variant when no explicit size provided
      ...(props.circular && !propsIn.size && {
        size
      }),
      tabIndex: 0,
      children: [themedIcon, wrappedChildren, themedIconAfter]
    })
  });
});
var Button = withStaticProperties(ButtonComponent, {
  Apply: context.Provider,
  Frame,
  Text,
  Icon
});
export { Button, ButtonContext };
