// @tamagui/select/dist/jsx/SelectTrigger.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { ListItem } from "@tamagui/list-item";
import * as React from "react";
import { useSelectContext, useSelectItemParentContext } from "./context.mjs";
import { jsx } from "react/jsx-runtime";
const TRIGGER_NAME = "SelectTrigger";
const isPointerCoarse = typeof window !== "undefined" && true ? window.matchMedia("(pointer:coarse)").matches : true;
const SelectTrigger = React.forwardRef(function SelectTrigger2(props, forwardedRef) {
  const {
    scope,
    disabled = false,
    unstyled = false,
    ...triggerProps
  } = props;
  const context = useSelectContext(scope);
  const itemParentContext = useSelectItemParentContext(scope);
  const composedRefs = useComposedRefs(forwardedRef, context.floatingContext?.refs.setReference);
  if (itemParentContext.shouldRenderWebNative) return null;
  return /* @__PURE__ */jsx(ListItem, {
    componentName: TRIGGER_NAME,
    unstyled,
    render: "button",
    type: "button",
    id: itemParentContext.id,
    ...(!unstyled && {
      focusVisibleStyle: {
        outlineStyle: "solid",
        outlineWidth: 2,
        outlineColor: "$outlineColor"
      },
      borderWidth: 1,
      size: itemParentContext.size
    }),
    role: "combobox",
    "aria-haspopup": "listbox",
    "aria-expanded": context.open,
    "aria-autocomplete": "none",
    dir: context.dir,
    disabled,
    "data-disabled": disabled ? "" : void 0,
    ...triggerProps,
    ref: composedRefs,
    ...(itemParentContext.interactions ? {
      ...itemParentContext.interactions.getReferenceProps(),
      ...(isPointerCoarse ? {
        onPress() {
          itemParentContext.setOpen(!context.open);
        }
      } : {
        onMouseDown() {
          context.floatingContext?.update?.();
          itemParentContext.setOpen(!context.open);
        }
      })
    } : {
      onPress() {
        itemParentContext.setOpen(!context.open);
      }
    })
  });
});
export { SelectTrigger };
