// @tamagui/select/dist/jsx/SelectItemText.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { styled, useIsomorphicLayoutEffect } from "@tamagui/core";
import { SizableText } from "@tamagui/text";
import * as React from "react";
import { useSelectItemParentContext } from "./context.mjs";
import { useSelectItemContext } from "./SelectItem.mjs";
import { Fragment, jsx } from "react/jsx-runtime";
const ITEM_TEXT_NAME = "SelectItemText";
const SelectItemTextFrame = styled(SizableText, {
  name: ITEM_TEXT_NAME,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        color: "$color",
        ellipsis: true
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const SelectItemText = SelectItemTextFrame.styleable(function SelectItemText2(props, forwardedRef) {
  const {
    scope,
    className,
    ...itemTextProps
  } = props;
  const itemParentContext = useSelectItemParentContext(scope);
  const ref = React.useRef(null);
  const composedRefs = useComposedRefs(forwardedRef, ref);
  const itemContext = useSelectItemContext(scope);
  const contents = React.useRef(null);
  contents.current = /* @__PURE__ */jsx(SelectItemTextFrame, {
    className,
    size: itemParentContext.size,
    id: itemContext.textId,
    ...itemTextProps,
    ref: composedRefs
  });
  useIsomorphicLayoutEffect(() => {
    if (itemParentContext.initialValue === itemContext.value) {
      itemParentContext.setSelectedItem(contents.current);
    }
  }, []);
  useIsomorphicLayoutEffect(() => {
    return itemParentContext.valueSubscribe(val => {
      if (val === itemContext.value) {
        itemParentContext.setSelectedItem(contents.current);
      }
    });
  }, [itemContext.value]);
  if (itemParentContext.shouldRenderWebNative) {
    return /* @__PURE__ */jsx(Fragment, {
      children: props.children
    });
  }
  return /* @__PURE__ */jsx(Fragment, {
    children: contents.current
  });
});
export { ITEM_TEXT_NAME, SelectItemText, SelectItemTextFrame };
