// @tamagui/select/dist/jsx/SelectImpl.mjs
import { autoUpdate, inner, offset, size, useClick, useFloatingRaw as useFloatingDom, useInteractions, useInnerOffset, useListNavigation, useRole, useTypeahead } from "@tamagui/floating";
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { useEvent, useIsTouchDevice } from "@tamagui/core";
import * as React from "react";
import { flushSync } from "react-dom";
import { SCROLL_ARROW_THRESHOLD } from "./constants.mjs";
import { SelectItemParentProvider, SelectProvider, useSelectContext, useSelectItemParentContext } from "./context.mjs";
import { jsx } from "react/jsx-runtime";
const SelectInlineImpl = props => {
  const {
    scope,
    children,
    open = false,
    listContentRef,
    setActiveIndexFast
  } = props;
  const selectContext = useSelectContext(scope);
  const selectItemParentContext = useSelectItemParentContext(scope);
  const {
    setActiveIndex,
    selectedIndex,
    activeIndexRef
  } = selectContext;
  const {
    setOpen,
    setSelectedIndex
  } = selectItemParentContext;
  const [scrollTop, setScrollTop] = React.useState(0);
  const touch = useIsTouchDevice();
  const listItemsRef = React.useRef([]);
  const overflowRef = React.useRef(null);
  const upArrowRef = React.useRef(null);
  const downArrowRef = React.useRef(null);
  const allowSelectRef = React.useRef(false);
  const allowMouseUpRef = React.useRef(true);
  const selectTimeoutRef = React.useRef(null);
  const state = React.useRef({
    isMouseOutside: false,
    isTyping: false
  });
  const [controlledScrolling, setControlledScrolling] = React.useState(false);
  const [fallback, setFallback] = React.useState(false);
  const [innerOffset, setInnerOffset] = React.useState(0);
  const [blockSelection, setBlockSelection] = React.useState(false);
  const floatingStyle = React.useRef({});
  React.useEffect(() => {
    if (open) setActiveIndexFast(selectedIndex ?? 0);else {
      setScrollTop(0);
      setFallback(false);
      setActiveIndexFast(null);
      setControlledScrolling(false);
    }
  }, [open, selectedIndex, setActiveIndexFast]);
  useIsomorphicLayoutEffect(() => {
    if (!open) return;
    const mouseUp = e => {
      if (state.current.isMouseOutside) setOpen(false);
    };
    document.addEventListener("mouseup", mouseUp);
    return () => {
      document.removeEventListener("mouseup", mouseUp);
    };
  }, [open]);
  const {
    x,
    y,
    strategy,
    refs,
    update,
    placement: computedPlacement
  } = useFloatingDom({
    open,
    placement: "bottom-start",
    whileElementsMounted: autoUpdate,
    middleware: [size({
      apply({
        rects: {
          reference: {
            width
          }
        }
      }) {
        Object.assign(floatingStyle.current, {
          minWidth: width + 8
        });
        if (refs.floating.current) Object.assign(refs.floating.current.style, floatingStyle.current);
      }
    }), inner({
      listRef: listItemsRef,
      overflowRef,
      index: selectedIndex,
      offset: innerOffset,
      onFallbackChange: setFallback,
      padding: 10,
      minItemsVisible: touch ? 10 : 4,
      referenceOverflowThreshold: 20
    }), offset({
      crossAxis: -5
    })]
  });
  const floatingRef = refs.floating;
  const showUpArrow = open && scrollTop > SCROLL_ARROW_THRESHOLD;
  const showDownArrow = open && floatingRef.current && scrollTop < floatingRef.current.scrollHeight - floatingRef.current.clientHeight - SCROLL_ARROW_THRESHOLD;
  const isScrollable = showDownArrow || showUpArrow;
  useIsomorphicLayoutEffect(() => {
    if (typeof window === "undefined") return;
    window.addEventListener("resize", update);
    if (open) update();
    return () => window.removeEventListener("resize", update);
  }, [update, open]);
  const onMatch = useEvent(index => {
    return (open ? setActiveIndex : setSelectedIndex)(index);
  });
  const dataRef = React.useRef({});
  dataRef.current.placement = computedPlacement;
  const interactionContext = {
    open,
    onOpenChange: val => setOpen(val),
    refs: {
      reference: refs.reference,
      floating: refs.floating,
      domReference: refs.reference
    },
    elements: {
      reference: refs.reference?.current || null,
      floating: refs.floating?.current || null,
      domReference: refs.reference?.current || null
    },
    dataRef
  };
  const interactionsProps = [useClick(interactionContext, {
    event: "mousedown",
    keyboardHandlers: false
  }), useRole(interactionContext, {
    role: "listbox"
  }), useInnerOffset(interactionContext, {
    enabled: !fallback && isScrollable,
    onChange: setInnerOffset,
    overflowRef,
    scrollRef: refs.floating
  }), useListNavigation(interactionContext, {
    listRef: listItemsRef,
    activeIndex: selectContext.activeIndex ?? 0,
    selectedIndex,
    onNavigate: index => {
      if (index !== null) setActiveIndex(index);
    },
    scrollItemIntoView: false
  }), useTypeahead(interactionContext, {
    listRef: listContentRef,
    onMatch,
    selectedIndex,
    activeIndex: selectContext.activeIndex,
    onTypingChange: e => {
      state.current.isTyping = e;
    }
  })];
  const interactions = useInteractions(React.useMemo(() => {
    return interactionsProps;
  }, interactionsProps));
  const interactionsContext = React.useMemo(() => {
    return {
      ...interactions,
      getReferenceProps() {
        return interactions.getReferenceProps({
          ref: refs.reference,
          className: "SelectTrigger",
          onKeyDown(event) {
            if (event.key === "Enter" || event.code === "Space" || event.key === " " && !state.current.isTyping) {
              event.preventDefault();
              setOpen(true);
            }
          }
        });
      },
      getFloatingProps(props2) {
        return interactions.getFloatingProps({
          ref: refs.floating,
          className: "Select",
          ...props2,
          style: {
            position: strategy,
            top: y ?? "",
            left: x ?? "",
            outline: 0,
            scrollbarWidth: "none",
            ...floatingStyle.current,
            ...props2?.style
          },
          onPointerEnter() {
            setControlledScrolling(false);
            state.current.isMouseOutside = false;
          },
          onPointerLeave() {
            state.current.isMouseOutside = true;
          },
          onPointerMove() {
            state.current.isMouseOutside = false;
            setControlledScrolling(false);
          },
          onKeyDown() {
            setControlledScrolling(true);
          },
          onContextMenu(e) {
            e.preventDefault();
          },
          onScroll(event) {
            flushSync(() => {
              setScrollTop(event.currentTarget.scrollTop);
            });
          }
        });
      }
    };
  }, [refs.reference.current, x, y, refs.floating.current, interactions]);
  useIsomorphicLayoutEffect(() => {
    if (open) {
      allowMouseUpRef.current = false;
      selectTimeoutRef.current = setTimeout(() => {
        allowSelectRef.current = true;
        allowMouseUpRef.current = true;
      }, 300);
      return () => {
        clearTimeout(selectTimeoutRef.current);
      };
    }
    allowSelectRef.current = false;
    allowMouseUpRef.current = true;
    setInnerOffset(0);
    setFallback(false);
    setBlockSelection(false);
  }, [open]);
  useIsomorphicLayoutEffect(() => {
    if (!open && state.current.isMouseOutside) state.current.isMouseOutside = false;
  }, [open]);
  useIsomorphicLayoutEffect(() => {
    function onPointerDown(e) {
      const target = e.target;
      if (!(refs.floating.current?.contains(target) || upArrowRef.current?.contains(target) || downArrowRef.current?.contains(target))) {
        setOpen(false);
        setControlledScrolling(false);
      }
    }
    if (open) {
      document.addEventListener("pointerdown", onPointerDown);
      return () => {
        document.removeEventListener("pointerdown", onPointerDown);
      };
    }
  }, [open, refs, setOpen]);
  React.useEffect(() => {
    if (!open) return;
    const scrollActiveIntoView = index => {
      if (controlledScrolling && index != null) listItemsRef.current[index]?.scrollIntoView({
        block: "nearest"
      });
      setScrollTop(refs.floating.current?.scrollTop ?? 0);
    };
    scrollActiveIntoView(activeIndexRef.current);
    return selectItemParentContext.activeIndexSubscribe(scrollActiveIntoView);
  }, [open, refs, controlledScrolling, selectItemParentContext.activeIndexSubscribe]);
  React.useEffect(() => {
    if (open && fallback) {
      if (selectedIndex != null) listItemsRef.current[selectedIndex]?.scrollIntoView({
        block: "nearest"
      });
    }
  }, [open, fallback, selectedIndex]);
  useIsomorphicLayoutEffect(() => {
    if (refs.floating.current && fallback) refs.floating.current.style.maxHeight = "";
  }, [refs, fallback]);
  const floatingContext = React.useMemo(() => ({
    refs,
    dataRef
  }), [refs]);
  return /* @__PURE__ */jsx(SelectProvider, {
    scope,
    ...selectContext,
    setScrollTop,
    setInnerOffset,
    fallback,
    floatingContext,
    canScrollDown: !!showDownArrow,
    canScrollUp: !!showUpArrow,
    controlledScrolling,
    blockSelection,
    upArrowRef,
    downArrowRef,
    update,
    children: /* @__PURE__ */jsx(SelectItemParentProvider, {
      scope,
      ...selectItemParentContext,
      allowMouseUpRef,
      allowSelectRef,
      dataRef,
      interactions: interactionsContext,
      listRef: listItemsRef,
      selectTimeoutRef,
      children
    })
  });
};
export { SelectInlineImpl };
