// @tamagui/select/dist/jsx/SelectContent.mjs
import { isWeb } from "@tamagui/core";
import { Dismissable } from "@tamagui/dismissable";
import { FocusScope } from "@tamagui/focus-scope";
import { Portal } from "@tamagui/portal";
import { RemoveScroll } from "@tamagui/remove-scroll";
import { useContext } from "react";
import { SelectZIndexContext, useSelectContext, useSelectItemParentContext } from "./context.mjs";
import { useShowSelectSheet } from "./useSelectBreakpointActive.mjs";
import { Fragment, jsx } from "react/jsx-runtime";
const SelectContent = ({
  children,
  scope,
  ...focusScopeProps
}) => {
  const context = useSelectContext(scope);
  const itemParentContext = useSelectItemParentContext(scope);
  const zIndex = useContext(SelectZIndexContext);
  const showSheet = useShowSelectSheet(context);
  const contents = children;
  if (itemParentContext.shouldRenderWebNative) {
    return /* @__PURE__ */jsx(Fragment, {
      children
    });
  }
  if (showSheet) {
    if (!context.open) {
      return null;
    }
    return /* @__PURE__ */jsx(Fragment, {
      children: contents
    });
  }
  return /* @__PURE__ */jsx(Portal, {
    open: context.open,
    zIndex,
    stackZIndex: 1e5,
    children: /* @__PURE__ */jsx(RemoveScroll, {
      enabled: context.open && !context.disablePreventBodyScroll,
      children: /* @__PURE__ */jsx(Dismissable, {
        asChild: true,
        forceUnmount: !context.open,
        onDismiss: () => itemParentContext.setOpen(false),
        onFocusOutside: e => e.preventDefault(),
        onPointerDownOutside: e => e.preventDefault(),
        children: /* @__PURE__ */jsx(FocusScope, {
          ...focusScopeProps,
          enabled: !!context.open,
          trapped: true,
          onMountAutoFocus: e => {
            e.preventDefault();
          },
          onUnmountAutoFocus: e => {
            e.preventDefault();
            const trigger = context.floatingContext?.refs?.reference?.current;
            if (trigger instanceof HTMLElement) {
              trigger.focus();
            }
          },
          children: isWeb ? /* @__PURE__ */jsx("div", {
            style: {
              display: "contents"
            },
            children: contents
          }) : contents
        })
      })
    })
  });
};
export { SelectContent };
