// @tamagui/animation-helpers/dist/esm/normalizeTransition.native.js
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var SPRING_CONFIG_KEYS = /* @__PURE__ */new Set(["stiffness", "damping", "mass", "tension", "friction", "velocity", "overshootClamping", "duration", "bounciness", "speed"]);
function isSpringConfigKey(key) {
  return SPRING_CONFIG_KEYS.has(key);
}
function normalizeTransition(transition) {
  if (!transition) {
    return {
      default: null,
      enter: null,
      exit: null,
      delay: void 0,
      properties: {}
    };
  }
  if (typeof transition === "string") {
    return {
      default: transition,
      enter: null,
      exit: null,
      delay: void 0,
      properties: {}
    };
  }
  if (Array.isArray(transition)) {
    var [defaultAnimation, configObj] = transition;
    var properties = {};
    var springConfig = {};
    var delay;
    var enter = null;
    var exit = null;
    if (configObj && (typeof configObj === "undefined" ? "undefined" : _type_of(configObj)) === "object") {
      var _iteratorNormalCompletion = true,
        _didIteratorError = false,
        _iteratorError = void 0;
      try {
        for (var _iterator = Object.entries(configObj)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var [key, value] = _step.value;
          if (key === "delay" && typeof value === "number") {
            delay = value;
          } else if (key === "enter" && typeof value === "string") {
            enter = value;
          } else if (key === "exit" && typeof value === "string") {
            exit = value;
          } else if (isSpringConfigKey(key) && value !== void 0) {
            springConfig[key] = value;
          } else if (value !== void 0) {
            properties[key] = value;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    }
    return {
      default: defaultAnimation,
      enter,
      exit,
      delay,
      properties,
      config: Object.keys(springConfig).length > 0 ? springConfig : void 0
    };
  }
  if ((typeof transition === "undefined" ? "undefined" : _type_of(transition)) === "object") {
    var properties1 = {};
    var springConfig1 = {};
    var defaultAnimation1 = null;
    var enter1 = null;
    var exit1 = null;
    var delay1;
    var _iteratorNormalCompletion1 = true,
      _didIteratorError1 = false,
      _iteratorError1 = void 0;
    try {
      for (var _iterator1 = Object.entries(transition)[Symbol.iterator](), _step1; !(_iteratorNormalCompletion1 = (_step1 = _iterator1.next()).done); _iteratorNormalCompletion1 = true) {
        var [key1, value1] = _step1.value;
        if (key1 === "default" && typeof value1 === "string") {
          defaultAnimation1 = value1;
        } else if (key1 === "enter" && typeof value1 === "string") {
          enter1 = value1;
        } else if (key1 === "exit" && typeof value1 === "string") {
          exit1 = value1;
        } else if (key1 === "delay" && typeof value1 === "number") {
          delay1 = value1;
        } else if (isSpringConfigKey(key1) && value1 !== void 0) {
          springConfig1[key1] = value1;
        } else if (value1 !== void 0) {
          properties1[key1] = value1;
        }
      }
    } catch (err) {
      _didIteratorError1 = true;
      _iteratorError1 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion1 && _iterator1.return != null) {
          _iterator1.return();
        }
      } finally {
        if (_didIteratorError1) {
          throw _iteratorError1;
        }
      }
    }
    return {
      default: defaultAnimation1,
      enter: enter1,
      exit: exit1,
      delay: delay1,
      properties: properties1,
      config: Object.keys(springConfig1).length > 0 ? springConfig1 : void 0
    };
  }
  return {
    default: null,
    enter: null,
    exit: null,
    delay: void 0,
    properties: {}
  };
}
function getAnimationForProperty(normalized, property) {
  var propertyAnimation = normalized.properties[property];
  if (propertyAnimation !== void 0) {
    return propertyAnimation;
  }
  return normalized.default;
}
function hasAnimation(normalized) {
  return normalized.default !== null || normalized.enter !== null || normalized.exit !== null || Object.keys(normalized.properties).length > 0;
}
function getAnimatedProperties(normalized) {
  return Object.keys(normalized.properties);
}
function getEffectiveAnimation(normalized, state) {
  if (state === "enter" && normalized.enter) {
    return normalized.enter;
  }
  if (state === "exit" && normalized.exit) {
    return normalized.exit;
  }
  return normalized.default;
}
function getAnimationConfigsForKeys(normalized, animations, keys, defaultAnimation) {
  var result = /* @__PURE__ */new Map();
  var _iteratorNormalCompletion = true,
    _didIteratorError = false,
    _iteratorError = void 0;
  try {
    for (var _iterator = keys[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var key = _step.value;
      var propAnimation = normalized.properties[key];
      var animationValue = null;
      if (typeof propAnimation === "string") {
        var _animations_propAnimation;
        animationValue = (_animations_propAnimation = animations[propAnimation]) !== null && _animations_propAnimation !== void 0 ? _animations_propAnimation : null;
      } else if (propAnimation && (typeof propAnimation === "undefined" ? "undefined" : _type_of(propAnimation)) === "object" && propAnimation.type) {
        var _animations_propAnimation_type;
        animationValue = (_animations_propAnimation_type = animations[propAnimation.type]) !== null && _animations_propAnimation_type !== void 0 ? _animations_propAnimation_type : null;
      }
      if (animationValue === null) {
        animationValue = defaultAnimation;
      }
      result.set(key, animationValue);
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return != null) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
  return result;
}
export { getAnimatedProperties, getAnimationConfigsForKeys, getAnimationForProperty, getEffectiveAnimation, hasAnimation, normalizeTransition };
