// @tamagui/get-token/dist/cjs/index.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  getRadius: () => getRadius,
  getSize: () => getSize,
  getSpace: () => getSpace,
  getTokenRelative: () => getTokenRelative,
  stepTokenUpOrDown: () => stepTokenUpOrDown
});
module.exports = __toCommonJS(index_exports);
var import_web = require("@tamagui/web");
var defaultOptions = {
  shift: 0,
  bounds: [0]
};
var getSize = function (size, options) {
  return getTokenRelative("size", size, options);
};
var getSpace = function (space, options) {
  return getTokenRelative("space", space, options);
};
var getRadius = function (radius, options) {
  return getTokenRelative("radius", radius, options);
};
var cacheVariables = {};
var cacheWholeVariables = {};
var cacheKeys = {};
var cacheWholeKeys = {};
var stepTokenUpOrDown = function (type, current) {
  var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : defaultOptions;
  var _options_bounds, _options_bounds1;
  var tokens = (0, import_web.getTokens)({
    prefixed: true
  })[type];
  if (!(type in cacheVariables)) {
    cacheKeys[type] = [];
    cacheVariables[type] = [];
    cacheWholeKeys[type] = [];
    cacheWholeVariables[type] = [];
    var sorted = Object.keys(tokens).map(function (k) {
      return tokens[k];
    }).sort(function (a, b) {
      return a.val - b.val;
    });
    var _iteratorNormalCompletion = true,
      _didIteratorError = false,
      _iteratorError = void 0;
    try {
      for (var _iterator = sorted[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var token = _step.value;
        cacheKeys[type].push(token.key);
        cacheVariables[type].push(token);
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
    var sortedExcludingHalfSteps = sorted.filter(function (x) {
      return !x.key.endsWith(".5");
    });
    var _iteratorNormalCompletion1 = true,
      _didIteratorError1 = false,
      _iteratorError1 = void 0;
    try {
      for (var _iterator1 = sortedExcludingHalfSteps[Symbol.iterator](), _step1; !(_iteratorNormalCompletion1 = (_step1 = _iterator1.next()).done); _iteratorNormalCompletion1 = true) {
        var token1 = _step1.value;
        cacheWholeKeys[type].push(token1.key);
        cacheWholeVariables[type].push(token1);
      }
    } catch (err) {
      _didIteratorError1 = true;
      _iteratorError1 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion1 && _iterator1.return != null) {
          _iterator1.return();
        }
      } finally {
        if (_didIteratorError1) {
          throw _iteratorError1;
        }
      }
    }
  }
  var isString = typeof current === "string";
  var cache = options.excludeHalfSteps ? isString ? cacheWholeKeys : cacheWholeVariables : isString ? cacheKeys : cacheVariables;
  var tokensOrdered = cache[type];
  var _options_bounds_;
  var min = (_options_bounds_ = (_options_bounds = options.bounds) === null || _options_bounds === void 0 ? void 0 : _options_bounds[0]) !== null && _options_bounds_ !== void 0 ? _options_bounds_ : 0;
  var _options_bounds_1;
  var max = (_options_bounds_1 = (_options_bounds1 = options.bounds) === null || _options_bounds1 === void 0 ? void 0 : _options_bounds1[1]) !== null && _options_bounds_1 !== void 0 ? _options_bounds_1 : tokensOrdered.length - 1;
  var currentIndex = tokensOrdered.indexOf(current);
  var shift = options.shift || 0;
  if (shift) {
    if (current === "$true" || (0, import_web.isVariable)(current) && current.name === "true") {
      shift += shift > 0 ? 1 : -1;
    }
  }
  var index = Math.min(max, Math.max(min, currentIndex + shift));
  var found = tokensOrdered[index];
  var result = (typeof found === "string" ? tokens[found] : found) || tokens["$true"];
  return result;
};
var getTokenRelative = stepTokenUpOrDown;
