// @tamagui/create-menu/dist/esm/createNativeMenu/utils.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import React, { Children, cloneElement, isValidElement } from "react";
function flattenChildrenKeyless(children) {
  var childrenArray = React.Children.toArray(children);
  return childrenArray.reduce(function (flatChildren, child) {
    if (child.type === React.Fragment) {
      return flatChildren.concat(flattenChildren(child.props.children));
    }
    flatChildren.push(child);
    return flatChildren;
  }, []);
}
function flattenChildren(children, componentNamesToIgnore) {
  var depth = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : 0,
    keys = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : [];
  return Children.toArray(children).flatMap(function (elem) {
    if (/* @__PURE__ */isValidElement(elem)) {
      if (typeof (elem === null || elem === void 0 ? void 0 : elem.type) == "function" && (componentNamesToIgnore === null || componentNamesToIgnore === void 0 ? void 0 : componentNamesToIgnore.some(function (skipComponentName) {
        return elem.type.displayName.includes(skipComponentName);
      }))) {
        return elem.props.children;
      }
    }
    return elem;
  }).reduce(function (acc, node, nodeIndex) {
    if (node.type === React.Fragment) {
      acc.push.apply(acc, flattenChildren(node.props.children, componentNamesToIgnore, depth + 1, keys.concat(node.key || nodeIndex)));
    } else {
      if (/* @__PURE__ */isValidElement(node)) {
        acc.push(/* @__PURE__ */cloneElement(node, {
          key: keys.concat(String(node.key)).join(".")
        }));
      } else if (typeof node === "string" || typeof node === "number") {
        acc.push(node);
      }
    }
    return acc;
  }, []);
}
var pickChildren = function (_children, targetChild, componentNamesToIgnore) {
  var children = flattenChildren(_children, componentNamesToIgnore);
  var target = [];
  var withoutTargetChildren = React.Children.map(children, function (item) {
    if (! /* @__PURE__ */isValidElement(item)) return item;
    if (isInstanceOfComponent(item, targetChild)) {
      target.push(/* @__PURE__ */cloneElement(item));
      return null;
    }
    return item;
  });
  var targetChildren = target.length >= 0 ? target : void 0;
  return {
    targetChildren,
    withoutTargetChildren
  };
};
var isInstanceOfComponent = function (element, targetElement) {
  var _element_type, _element_type1;
  var matches = (element === null || element === void 0 ? void 0 : element.type) === targetElement || typeof (element === null || element === void 0 ? void 0 : element.type) == "function" && ((element === null || element === void 0 ? void 0 : (_element_type = element.type) === null || _element_type === void 0 ? void 0 : _element_type.displayName) === targetElement.displayName || (element === null || element === void 0 ? void 0 : (_element_type1 = element.type) === null || _element_type1 === void 0 ? void 0 : _element_type1.displayName) === targetElement.displayName + "Wrapper");
  return matches;
};
var filterNull = function (t) {
  return t != null;
};
var create = function (Component, displayName) {
  var MenuComponent = function (props) {
    return /* @__PURE__ */_jsx(Component, {
      ...props
    });
  };
  MenuComponent.displayName = displayName;
  return MenuComponent;
};
export { create, filterNull, flattenChildren, flattenChildrenKeyless, isInstanceOfComponent, pickChildren };
