// @tamagui/context-menu/dist/esm/ContextMenu.native.js
import { createNativeMenu, withNativeMenu } from "@tamagui/create-menu";
import { withStaticProperties } from "@tamagui/web";
import React from "react";
import { CONTEXTMENU_CONTEXT, createNonNativeContextMenu } from "./createNonNativeContextMenu.native.js";
var COMMON_PARAMS = {
  isRoot: false,
  scope: CONTEXTMENU_CONTEXT
};
function createContextMenu(param) {
  var {
    Menu: NativeMenuRoot
  } = createNativeMenu("ContextMenu");
  var NonNativeContextMenu = createNonNativeContextMenu(param);
  var ContextMenuComp = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Root,
    NativeComponent: NativeMenuRoot,
    isRoot: true
  });
  var Trigger = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Trigger,
    NativeComponent: NativeMenuRoot.Trigger
  });
  var Portal = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Portal,
    NativeComponent: React.Fragment
  });
  var Content = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Content,
    NativeComponent: NativeMenuRoot.Content
  });
  var Preview = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Preview,
    NativeComponent: NativeMenuRoot.Preview
  });
  var Group = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Group,
    NativeComponent: NativeMenuRoot.Group
  });
  var Label = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Label,
    NativeComponent: NativeMenuRoot.Label
  });
  var Item = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Item,
    NativeComponent: NativeMenuRoot.Item
  });
  var ItemTitle = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemTitle,
    NativeComponent: NativeMenuRoot.ItemTitle
  });
  var ItemSubtitle = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemSubtitle,
    NativeComponent: NativeMenuRoot.ItemSubtitle
  });
  var ItemIcon = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemIcon,
    NativeComponent: NativeMenuRoot.ItemIcon
  });
  var ItemImage = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemImage,
    NativeComponent: NativeMenuRoot.ItemImage
  });
  var CheckboxItem = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.CheckboxItem,
    NativeComponent: NativeMenuRoot.CheckboxItem
  });
  var RadioGroup = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.RadioGroup,
    NativeComponent: function () {
      return null;
    }
  });
  var RadioItem = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.RadioItem,
    NativeComponent: function (param2) {
      var {
        children
      } = param2;
      return children;
    }
  });
  var ItemIndicator = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemIndicator,
    NativeComponent: NativeMenuRoot.ItemIndicator
  });
  var Separator = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Separator,
    NativeComponent: NativeMenuRoot.Separator
  });
  var Arrow = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Arrow,
    NativeComponent: NativeMenuRoot.Arrow
  });
  var Sub = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Sub,
    NativeComponent: NativeMenuRoot.Sub
  });
  var SubTrigger = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.SubTrigger,
    NativeComponent: NativeMenuRoot.SubTrigger
  });
  var SubContent = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.SubContent,
    NativeComponent: NativeMenuRoot.SubContent
  });
  var ContextMenu = withStaticProperties(ContextMenuComp, {
    Trigger,
    Portal,
    Content,
    Group,
    Label,
    Item,
    CheckboxItem,
    RadioGroup,
    RadioItem,
    ItemIndicator,
    Separator,
    Arrow,
    Sub,
    // cast to React.FC to avoid TS error
    SubTrigger,
    SubContent,
    ItemTitle,
    ItemSubtitle,
    // cast to React.FC to avoid TS error
    ItemIcon,
    ItemImage,
    Preview
  });
  return ContextMenu;
}
export { createContextMenu };
