// @tamagui/web/dist/cjs/createTamagui.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createTamagui_exports = {};
__export(createTamagui_exports, {
  createTamagui: () => createTamagui
});
module.exports = __toCommonJS(createTamagui_exports);
var import_config = require("./config.cjs");
var import_createVariables = require("./createVariables.cjs");
var import_defaultAnimationDriver = require("./helpers/defaultAnimationDriver.cjs");
var import_resolveAnimationDriver = require("./helpers/resolveAnimationDriver.cjs");
var import_createDesignSystem = require("./helpers/createDesignSystem.cjs");
var import_insertStyleRule = require("./helpers/insertStyleRule.cjs");
var import_proxyThemeToParents = require("./helpers/proxyThemeToParents.cjs");
var import_themes = require("./helpers/themes.cjs");
var import_useMedia = require("./hooks/useMedia.cjs");
var import_insertFont = require("./insertFont.cjs");
var import_Tamagui = require("./Tamagui.cjs");
function shouldTokenCategoryHaveUnits(category) {
  return (/* @__PURE__ */new Set(["size", "space", "radius"])).has(category);
}
function initializeTamaguiConfig(config) {
  (0, import_config.setConfig)(config);
  (0, import_useMedia.configureMedia)(config);
}
function createTamagui(configIn) {
  const existingConfig = (0, import_config.getConfigMaybe)();
  if (existingConfig) configIn = {
    ...existingConfig,
    ...configIn
  };
  const tokensParsed = {};
  const tokens = (0, import_createVariables.createVariables)(configIn.tokens || {});
  if (configIn.tokens) {
    const tokensMerged = {};
    for (const cat in tokens) {
      tokensParsed[cat] = {};
      tokensMerged[cat] = {};
      const tokenCat = tokens[cat];
      for (const key in tokenCat) {
        const val = tokenCat[key];
        const prefixedKey = `$${key}`;
        tokensParsed[cat][prefixedKey] = val;
        tokensMerged[cat][prefixedKey] = val;
        tokensMerged[cat][key] = val;
      }
    }
    (0, import_config.setTokens)(tokensMerged);
  }
  let foundThemes;
  if (configIn.themes) {
    const noThemes = Object.keys(configIn.themes).length === 0;
    if (noThemes && !process.env.TAMAGUI_DID_OUTPUT_CSS) foundThemes = (0, import_insertStyleRule.scanAllSheets)(noThemes, tokensParsed);
  }
  let fontSizeTokens = null;
  let fontsParsed;
  if (configIn.fonts) {
    const fontTokens = Object.fromEntries(Object.entries(configIn.fonts).map(([k, v]) => {
      return [k, (0, import_createVariables.createVariables)(v, "f", true)];
    }));
    fontsParsed = (() => {
      const res = {};
      for (const familyName in fontTokens) {
        const font = fontTokens[familyName];
        const fontParsed = (0, import_insertFont.parseFont)(font);
        res[`$${familyName}`] = fontParsed;
        if (!fontSizeTokens && fontParsed.size) fontSizeTokens = new Set(Object.keys(fontParsed.size));
      }
      return res;
    })();
  }
  const specificTokens = {};
  const themeConfig = (() => {
    const sortedTokenKeys = Object.keys(tokens).sort();
    for (const key of sortedTokenKeys) {
      const sortedSubKeys = Object.keys(tokens[key]).sort();
      for (const skey of sortedSubKeys) {
        const variable = tokens[key][skey];
        specificTokens[`$${key}.${skey}`] = variable;
        if (process.env.NODE_ENV === "development") {
          if (typeof variable === "undefined") throw new Error(`No value for tokens.${key}.${skey}:
${JSON.stringify(variable, null, 2)}`);
        }
      }
    }
    const declarations = (0, import_createDesignSystem.createTokenCSS)(tokens, shouldTokenCategoryHaveUnits);
    const fontDeclarations = (0, import_createDesignSystem.createFontCSS)(fontsParsed, import_insertFont.registerFontVariables);
    const cssRuleSets = (0, import_createDesignSystem.buildCSSRuleSets)(declarations, fontDeclarations);
    const themesIn = configIn.themes;
    const dedupedThemes = foundThemes ?? getThemesDeduped(themesIn, tokens.color);
    return {
      themes: (0, import_proxyThemeToParents.proxyThemesToParents)(dedupedThemes),
      cssRuleSets,
      getThemeRulesSets() {
        return (0, import_createDesignSystem.createThemeCSS)(dedupedThemes, configIn);
      }
    };
  })();
  const userShorthands = configIn.shorthands || {};
  const shorthands = {
    ...builtinShorthands,
    ...userShorthands
  };
  const lastCSSIndex = {
    value: -1
  };
  const getCSS = (opts = {}) => {
    return (0, import_createDesignSystem.getCSS)(themeConfig, opts, lastCSSIndex);
  };
  const getNewCSS = opts => getCSS({
    ...opts,
    sinceLastCall: true
  });
  const defaultFontSetting = configIn.settings?.defaultFont;
  const defaultFont = (() => {
    let val = defaultFontSetting;
    if (val?.[0] === "$") val = val.slice(1);
    return val;
  })();
  const defaultPositionSetting = configIn.settings?.defaultPosition || "static";
  const defaultProps = configIn.defaultProps || {};
  if (defaultPositionSetting !== "static") defaultProps.View = {
    ...defaultProps.View,
    position: defaultPositionSetting
  };
  const defaultFontToken = defaultFont ? `$${defaultFont}` : "";
  const inputAnimations = configIn.animations;
  const resolvedDriver = (0, import_resolveAnimationDriver.resolveAnimationDriver)(inputAnimations);
  const isMultiDriver = resolvedDriver !== null && resolvedDriver !== inputAnimations;
  const resolvedAnimations = resolvedDriver ?? inputAnimations;
  const animationDrivers = isMultiDriver ? inputAnimations : void 0;
  const config = {
    fonts: {},
    onlyAllowShorthands: false,
    fontLanguages: [],
    media: {},
    ...configIn,
    animations: resolvedAnimations ?? import_defaultAnimationDriver.defaultAnimationDriver,
    animationDrivers,
    defaultProps,
    settings: {
      webContainerType: "inline-size",
      ...configIn.settings
    },
    tokens,
    shorthands,
    userShorthands,
    inverseShorthands: shorthands ? Object.fromEntries(Object.entries(shorthands).map(([k, v]) => [v, k])) : {},
    themes: themeConfig.themes,
    fontsParsed: fontsParsed || {},
    themeConfig,
    tokensParsed,
    parsed: true,
    getNewCSS,
    getCSS,
    defaultFont,
    fontSizeTokens: fontSizeTokens || /* @__PURE__ */new Set(),
    specificTokens,
    defaultFontToken
  };
  initializeTamaguiConfig(config);
  if (process.env.NODE_ENV !== "development") return config;
  if (process.env.DEBUG?.startsWith("tamagui")) console.info("Tamagui config:", config);
  if (!globalThis["Tamagui"]) globalThis["Tamagui"] = import_Tamagui.Tamagui;
  return config;
}
function getThemesDeduped(themes, colorTokens) {
  const dedupedThemes = [];
  const existing = /* @__PURE__ */new Map();
  const sortedThemeNames = Object.keys(themes).sort();
  for (const themeName of sortedThemeNames) {
    const darkOrLightSpecificPrefix = themeName.startsWith("dark") ? "dark" : themeName.startsWith("light") ? "light" : "";
    const rawTheme = themes[themeName];
    const key = darkOrLightSpecificPrefix + JSON.stringify(rawTheme);
    if (existing.has(key)) {
      existing.get(key).names.push(themeName);
      continue;
    }
    const theme = {
      ...colorTokens,
      ...rawTheme
    };
    for (const key2 in theme) (0, import_themes.ensureThemeVariable)(theme, key2);
    const deduped = {
      names: [themeName],
      theme
    };
    dedupedThemes.push(deduped);
    existing.set(key, deduped);
  }
  return dedupedThemes;
}
const builtinShorthands = {
  bblr: "borderBottomLeftRadius",
  bbrr: "borderBottomRightRadius",
  bbs: "borderBottomStyle",
  bls: "borderLeftStyle",
  brc: "borderRightColor",
  brs: "borderRightStyle",
  brw: "borderRightWidth",
  bs: "borderStyle",
  btc: "borderTopColor",
  btlr: "borderTopLeftRadius",
  btrr: "borderTopRightRadius",
  bts: "borderTopStyle",
  btw: "borderTopWidth",
  bw: "borderWidth",
  bxs: "boxSizing",
  bxsh: "boxShadow",
  col: "color",
  cur: "cursor",
  dsp: "display",
  fb: "flexBasis",
  fd: "flexDirection",
  ff: "fontFamily",
  fs: "fontSize",
  fst: "fontStyle",
  fw: "fontWeight",
  fwr: "flexWrap",
  lh: "lineHeight",
  ls: "letterSpacing",
  o: "opacity",
  ov: "overflow",
  ox: "overflowX",
  oy: "overflowY",
  pe: "pointerEvents",
  pos: "position",
  td: "textDecorationLine",
  tr: "transform",
  tt: "textTransform",
  va: "verticalAlign",
  wb: "wordBreak",
  ws: "whiteSpace",
  ww: "wordWrap"
};
