// @tamagui/web/dist/cjs/createTamagui.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createTamagui_exports = {};
__export(createTamagui_exports, {
  createTamagui: () => createTamagui
});
module.exports = __toCommonJS(createTamagui_exports);
var import_config = require("./config.native.js");
var import_createVariables = require("./createVariables.native.js");
var import_defaultAnimationDriver = require("./helpers/defaultAnimationDriver.native.js");
var import_resolveAnimationDriver = require("./helpers/resolveAnimationDriver.native.js");
var import_createDesignSystem = require("./helpers/createDesignSystem.native.js");
var import_insertStyleRule = require("./helpers/insertStyleRule.native.js");
var import_proxyThemeToParents = require("./helpers/proxyThemeToParents.native.js");
var import_themes = require("./helpers/themes.native.js");
var import_useMedia = require("./hooks/useMedia.native.js");
var import_insertFont = require("./insertFont.native.js");
var import_Tamagui = require("./Tamagui.native.js");
function shouldTokenCategoryHaveUnits(category) {
  return (/* @__PURE__ */new Set(["size", "space", "radius"])).has(category);
}
function initializeTamaguiConfig(config) {
  (0, import_config.setConfig)(config);
  (0, import_useMedia.configureMedia)(config);
}
function createTamagui(configIn) {
  var _configIn_settings, _configIn_settings1, _process_env_DEBUG;
  var existingConfig = (0, import_config.getConfigMaybe)();
  if (existingConfig) configIn = {
    ...existingConfig,
    ...configIn
  };
  var tokensParsed = {};
  var tokens = (0, import_createVariables.createVariables)(configIn.tokens || {});
  if (configIn.tokens) {
    var tokensMerged = {};
    for (var cat in tokens) {
      tokensParsed[cat] = {};
      tokensMerged[cat] = {};
      var tokenCat = tokens[cat];
      for (var key in tokenCat) {
        var val = tokenCat[key];
        var prefixedKey = `$${key}`;
        tokensParsed[cat][prefixedKey] = val;
        tokensMerged[cat][prefixedKey] = val;
        tokensMerged[cat][key] = val;
      }
    }
    (0, import_config.setTokens)(tokensMerged);
  }
  var foundThemes;
  if (configIn.themes) {
    var noThemes = Object.keys(configIn.themes).length === 0;
    if (noThemes && !process.env.TAMAGUI_DID_OUTPUT_CSS) foundThemes = (0, import_insertStyleRule.scanAllSheets)(noThemes, tokensParsed);
  }
  var fontSizeTokens = null;
  var fontsParsed;
  if (configIn.fonts) {
    var fontTokens = Object.fromEntries(Object.entries(configIn.fonts).map(function (param) {
      var [k, v] = param;
      return [k, (0, import_createVariables.createVariables)(v, "f", true)];
    }));
    fontsParsed = function () {
      var res = {};
      for (var familyName in fontTokens) {
        var font = fontTokens[familyName];
        var fontParsed = (0, import_insertFont.parseFont)(font);
        res[`$${familyName}`] = fontParsed;
        if (!fontSizeTokens && fontParsed.size) fontSizeTokens = new Set(Object.keys(fontParsed.size));
      }
      return res;
    }();
  }
  var specificTokens = {};
  var themeConfig = function () {
    var sortedTokenKeys = Object.keys(tokens).sort();
    var _iteratorNormalCompletion = true,
      _didIteratorError = false,
      _iteratorError = void 0;
    try {
      for (var _iterator = sortedTokenKeys[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var key2 = _step.value;
        var sortedSubKeys = Object.keys(tokens[key2]).sort();
        var _iteratorNormalCompletion1 = true,
          _didIteratorError1 = false,
          _iteratorError1 = void 0;
        try {
          for (var _iterator1 = sortedSubKeys[Symbol.iterator](), _step1; !(_iteratorNormalCompletion1 = (_step1 = _iterator1.next()).done); _iteratorNormalCompletion1 = true) {
            var skey = _step1.value;
            var variable = tokens[key2][skey];
            specificTokens[`$${key2}.${skey}`] = variable;
            if (process.env.NODE_ENV === "development") {
              if (typeof variable === "undefined") throw new Error(`No value for tokens.${key2}.${skey}:
${JSON.stringify(variable, null, 2)}`);
            }
          }
        } catch (err) {
          _didIteratorError1 = true;
          _iteratorError1 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion1 && _iterator1.return != null) _iterator1.return();
          } finally {
            if (_didIteratorError1) throw _iteratorError1;
          }
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) _iterator.return();
      } finally {
        if (_didIteratorError) throw _iteratorError;
      }
    }
    var declarations = (0, import_createDesignSystem.createTokenCSS)(tokens, shouldTokenCategoryHaveUnits);
    var fontDeclarations = (0, import_createDesignSystem.createFontCSS)(fontsParsed, import_insertFont.registerFontVariables);
    var cssRuleSets = (0, import_createDesignSystem.buildCSSRuleSets)(declarations, fontDeclarations);
    var themesIn = configIn.themes;
    var dedupedThemes = foundThemes !== null && foundThemes !== void 0 ? foundThemes : getThemesDeduped(themesIn, tokens.color);
    return {
      themes: (0, import_proxyThemeToParents.proxyThemesToParents)(dedupedThemes),
      cssRuleSets,
      getThemeRulesSets() {
        return (0, import_createDesignSystem.createThemeCSS)(dedupedThemes, configIn);
      }
    };
  }();
  var userShorthands = configIn.shorthands || {};
  var shorthands = {
    ...builtinShorthands,
    ...userShorthands
  };
  var lastCSSIndex = {
    value: -1
  };
  var getCSS = function () {
    var opts = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {};
    return (0, import_createDesignSystem.getCSS)(themeConfig, opts, lastCSSIndex);
  };
  var getNewCSS = function (opts) {
    return getCSS({
      ...opts,
      sinceLastCall: true
    });
  };
  var defaultFontSetting = (_configIn_settings = configIn.settings) === null || _configIn_settings === void 0 ? void 0 : _configIn_settings.defaultFont;
  var defaultFont = function () {
    var val2 = defaultFontSetting;
    if ((val2 === null || val2 === void 0 ? void 0 : val2[0]) === "$") val2 = val2.slice(1);
    return val2;
  }();
  !((_configIn_settings1 = configIn.settings) === null || _configIn_settings1 === void 0) && _configIn_settings1.defaultPosition;
  var defaultProps = configIn.defaultProps || {};
  var defaultFontToken = defaultFont ? `$${defaultFont}` : "";
  var inputAnimations = configIn.animations;
  var resolvedDriver = (0, import_resolveAnimationDriver.resolveAnimationDriver)(inputAnimations);
  var isMultiDriver = resolvedDriver !== null && resolvedDriver !== inputAnimations;
  var resolvedAnimations = resolvedDriver !== null && resolvedDriver !== void 0 ? resolvedDriver : inputAnimations;
  var animationDrivers = isMultiDriver ? inputAnimations : void 0;
  var config = {
    fonts: {},
    onlyAllowShorthands: false,
    fontLanguages: [],
    media: {},
    ...configIn,
    animations: resolvedAnimations !== null && resolvedAnimations !== void 0 ? resolvedAnimations : import_defaultAnimationDriver.defaultAnimationDriver,
    animationDrivers,
    defaultProps,
    settings: {
      webContainerType: "inline-size",
      ...configIn.settings
    },
    tokens,
    shorthands,
    userShorthands,
    inverseShorthands: shorthands ? Object.fromEntries(Object.entries(shorthands).map(function (param) {
      var [k, v] = param;
      return [v, k];
    })) : {},
    themes: themeConfig.themes,
    fontsParsed: fontsParsed || {},
    themeConfig,
    tokensParsed,
    parsed: true,
    getNewCSS,
    getCSS,
    defaultFont,
    fontSizeTokens: fontSizeTokens || /* @__PURE__ */new Set(),
    specificTokens,
    defaultFontToken
  };
  initializeTamaguiConfig(config);
  if (process.env.NODE_ENV !== "development") return config;
  if ((_process_env_DEBUG = process.env.DEBUG) === null || _process_env_DEBUG === void 0 ? void 0 : _process_env_DEBUG.startsWith("tamagui")) console.info("Tamagui config:", config);
  if (!globalThis["Tamagui"]) globalThis["Tamagui"] = import_Tamagui.Tamagui;
  return config;
}
function getThemesDeduped(themes, colorTokens) {
  var dedupedThemes = [];
  var existing = /* @__PURE__ */new Map();
  var sortedThemeNames = Object.keys(themes).sort();
  var _iteratorNormalCompletion = true,
    _didIteratorError = false,
    _iteratorError = void 0;
  try {
    for (var _iterator = sortedThemeNames[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var themeName = _step.value;
      var darkOrLightSpecificPrefix = themeName.startsWith("dark") ? "dark" : themeName.startsWith("light") ? "light" : "";
      var rawTheme = themes[themeName];
      var key = darkOrLightSpecificPrefix + JSON.stringify(rawTheme);
      if (existing.has(key)) {
        existing.get(key).names.push(themeName);
        continue;
      }
      var theme = {
        ...colorTokens,
        ...rawTheme
      };
      for (var key1 in theme) (0, import_themes.ensureThemeVariable)(theme, key1);
      var deduped = {
        names: [themeName],
        theme
      };
      dedupedThemes.push(deduped);
      existing.set(key, deduped);
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return != null) _iterator.return();
    } finally {
      if (_didIteratorError) throw _iteratorError;
    }
  }
  return dedupedThemes;
}
var builtinShorthands = {
  bblr: "borderBottomLeftRadius",
  bbrr: "borderBottomRightRadius",
  bbs: "borderBottomStyle",
  bls: "borderLeftStyle",
  brc: "borderRightColor",
  brs: "borderRightStyle",
  brw: "borderRightWidth",
  bs: "borderStyle",
  btc: "borderTopColor",
  btlr: "borderTopLeftRadius",
  btrr: "borderTopRightRadius",
  bts: "borderTopStyle",
  btw: "borderTopWidth",
  bw: "borderWidth",
  bxs: "boxSizing",
  bxsh: "boxShadow",
  col: "color",
  cur: "cursor",
  dsp: "display",
  fb: "flexBasis",
  fd: "flexDirection",
  ff: "fontFamily",
  fs: "fontSize",
  fst: "fontStyle",
  fw: "fontWeight",
  fwr: "flexWrap",
  lh: "lineHeight",
  ls: "letterSpacing",
  o: "opacity",
  ov: "overflow",
  ox: "overflowX",
  oy: "overflowY",
  pe: "pointerEvents",
  pos: "position",
  td: "textDecorationLine",
  tr: "transform",
  tt: "textTransform",
  va: "verticalAlign",
  wb: "wordBreak",
  ws: "whiteSpace",
  ww: "wordWrap"
};
