// @tamagui/web/dist/cjs/eventHandling.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var eventHandling_native_exports = {};
__export(eventHandling_native_exports, {
  getWebEvents: () => getWebEvents,
  useEvents: () => useEvents,
  wrapWithGestureDetector: () => wrapWithGestureDetector
});
module.exports = __toCommonJS(eventHandling_native_exports);
var import_helpers = require("@tamagui/helpers");
var import_native = require("@tamagui/native");
var import_react = __toESM(require("react"), 1);
var import_mainThreadPressEvents = require("./helpers/mainThreadPressEvents.native.js");
function getWebEvents() {
  return {};
}
function useEvents(events, viewProps, stateRef, staticConfig, isHOC, isInsideNativeMenu) {
  if (events) {
    if (events.onFocus) {
      viewProps["onFocus"] = events.onFocus;
    }
    if (events.onBlur) {
      viewProps["onBlur"] = events.onBlur;
    }
  }
  var hasPressEvents =
  // its stable and always on if you have in/out/regular
  events === null || events === void 0 ? void 0 : events.onPress;
  var gh = (0, import_native.getGestureHandler)();
  if (hasPressEvents) {
    stateRef.current.hasHadEvents = true;
  }
  var everEnabled = Boolean(hasPressEvents || stateRef.current.hasHadEvents);
  var isUsingRNGH = gh.isEnabled;
  if (staticConfig.isInput) {
    if (events) {
      var {
        onPressIn,
        onPressOut,
        onPress
      } = events;
      var inputEvents = {
        onPressIn,
        onPressOut: onPressOut || onPress
      };
      if (onPressOut && onPress) {
        inputEvents.onPressOut = (0, import_helpers.composeEventHandlers)(onPress, onPressOut);
      }
      Object.assign(viewProps, inputEvents);
    }
    return null;
  }
  var isCompositeComponent = !isHOC && staticConfig.Component && typeof staticConfig.Component !== "string";
  if (isHOC || isCompositeComponent) {
    if (events) {
      var {
        onPressIn: onPressIn1,
        onPressOut: onPressOut1,
        onPress: onPress1,
        onLongPress,
        delayLongPress
      } = events;
      Object.assign(viewProps, {
        onPressIn: onPressIn1,
        onPressOut: onPressOut1,
        onPress: onPress1,
        onLongPress,
        delayLongPress
      });
    }
    return null;
  }
  if (isUsingRNGH) {
    var callbacksRef = (0, import_react.useRef)(isUsingRNGH ? {} : null);
    var gestureRef = (0, import_react.useRef)(null);
    if (everEnabled) {
      callbacksRef.current = hasPressEvents ? {
        onPressIn: events.onPressIn,
        onPressOut: events.onPressOut,
        onPress: events.onPress,
        onLongPress: events.onLongPress
      } : {};
      if (!gestureRef.current) {
        if (isInsideNativeMenu) {
          var {
            Gesture
          } = gh.state;
          var manual = Gesture.Manual().runOnJS(true).manualActivation(true).onTouchesDown(function () {
            var _callbacksRef_current_onPressIn, _callbacksRef_current;
            (_callbacksRef_current_onPressIn = (_callbacksRef_current = callbacksRef.current).onPressIn) === null || _callbacksRef_current_onPressIn === void 0 ? void 0 : _callbacksRef_current_onPressIn.call(_callbacksRef_current, {});
          }).onTouchesUp(function () {
            var _callbacksRef_current_onPress, _callbacksRef_current, _callbacksRef_current_onPressOut, _callbacksRef_current1;
            (_callbacksRef_current_onPress = (_callbacksRef_current = callbacksRef.current).onPress) === null || _callbacksRef_current_onPress === void 0 ? void 0 : _callbacksRef_current_onPress.call(_callbacksRef_current, {});
            (_callbacksRef_current_onPressOut = (_callbacksRef_current1 = callbacksRef.current).onPressOut) === null || _callbacksRef_current_onPressOut === void 0 ? void 0 : _callbacksRef_current_onPressOut.call(_callbacksRef_current1, {});
          }).onTouchesCancelled(function () {
            var _callbacksRef_current_onPressOut, _callbacksRef_current;
            (_callbacksRef_current_onPressOut = (_callbacksRef_current = callbacksRef.current).onPressOut) === null || _callbacksRef_current_onPressOut === void 0 ? void 0 : _callbacksRef_current_onPressOut.call(_callbacksRef_current, {});
          });
          gestureRef.current = manual;
        } else {
          gestureRef.current = gh.createPressGesture({
            onPressIn: function (e) {
              var _callbacksRef_current_onPressIn, _callbacksRef_current;
              return (_callbacksRef_current_onPressIn = (_callbacksRef_current = callbacksRef.current).onPressIn) === null || _callbacksRef_current_onPressIn === void 0 ? void 0 : _callbacksRef_current_onPressIn.call(_callbacksRef_current, e);
            },
            onPressOut: function (e) {
              var _callbacksRef_current_onPressOut, _callbacksRef_current;
              return (_callbacksRef_current_onPressOut = (_callbacksRef_current = callbacksRef.current).onPressOut) === null || _callbacksRef_current_onPressOut === void 0 ? void 0 : _callbacksRef_current_onPressOut.call(_callbacksRef_current, e);
            },
            onPress: function (e) {
              var _callbacksRef_current_onPress, _callbacksRef_current;
              return (_callbacksRef_current_onPress = (_callbacksRef_current = callbacksRef.current).onPress) === null || _callbacksRef_current_onPress === void 0 ? void 0 : _callbacksRef_current_onPress.call(_callbacksRef_current, e);
            },
            onLongPress: function (e) {
              var _callbacksRef_current_onLongPress, _callbacksRef_current;
              return (_callbacksRef_current_onLongPress = (_callbacksRef_current = callbacksRef.current).onLongPress) === null || _callbacksRef_current_onLongPress === void 0 ? void 0 : _callbacksRef_current_onLongPress.call(_callbacksRef_current, e);
            },
            delayLongPress: events === null || events === void 0 ? void 0 : events.delayLongPress,
            hitSlop: viewProps.hitSlop
          });
        }
      }
      return gestureRef.current;
    }
    return null;
  }
  (0, import_mainThreadPressEvents.useMainThreadPressEvents)(events, viewProps, hasPressEvents);
  return null;
}
function wrapWithGestureDetector(content, gesture, stateRef, isHOC, isCompositeComponent) {
  if (isHOC || isCompositeComponent) {
    return content;
  }
  var gh = (0, import_native.getGestureHandler)();
  var {
    GestureDetector,
    Gesture
  } = gh.state;
  var shouldWrap = stateRef.current.hasHadEvents;
  if (!GestureDetector || !shouldWrap) {
    return content;
  }
  var gestureToUse = gesture || (Gesture === null || Gesture === void 0 ? void 0 : Gesture.Manual());
  if (!gestureToUse) {
    return content;
  }
  return import_react.default.createElement(GestureDetector, {
    gesture: gestureToUse
  }, content);
}
