// @tamagui/web/dist/cjs/helpers/createDesignSystem.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createDesignSystem_exports = {};
__export(createDesignSystem_exports, {
  buildCSSRuleSets: () => buildCSSRuleSets,
  createFontCSS: () => createFontCSS,
  createThemeCSS: () => createThemeCSS,
  createTokenCSS: () => createTokenCSS,
  getCSS: () => getCSS,
  getFontPropertyDeclarations: () => getFontPropertyDeclarations
});
module.exports = __toCommonJS(createDesignSystem_exports);
var import_constants = require("@tamagui/constants");
var import_createVariable = require("../createVariable.cjs");
var import_registerCSSVariable = require("./registerCSSVariable.cjs");
var import_getThemeCSSRules = require("./getThemeCSSRules.cjs");
var import_insertStyleRule = require("./insertStyleRule.cjs");
function getFontPropertyDeclarations(fontParsed, tokenKey = "$true") {
  const props = ["font-family: var(--f-family)"];
  const getVarRef = obj => {
    const val = obj?.[tokenKey];
    if ((0, import_createVariable.isVariable)(val)) return (0, import_createVariable.getVariableVariable)(val);
  };
  const letterSpacing = getVarRef(fontParsed.letterSpacing);
  if (letterSpacing) props.push(`letter-spacing: ${letterSpacing}`);
  const lineHeight = getVarRef(fontParsed.lineHeight);
  if (lineHeight) props.push(`line-height: ${lineHeight}`);
  const fontStyle = getVarRef(fontParsed.style);
  if (fontStyle) props.push(`font-style: ${fontStyle}`);
  const fontWeight = getVarRef(fontParsed.weight);
  if (fontWeight) props.push(`font-weight: ${fontWeight}`);
  return props;
}
function createTokenCSS(tokens, shouldTokenCategoryHaveUnits) {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    const declarations = [];
    const sortedTokenKeys = Object.keys(tokens).sort();
    for (const key of sortedTokenKeys) {
      const sortedSubKeys = Object.keys(tokens[key]).sort();
      for (const skey of sortedSubKeys) {
        const variable = tokens[key][skey];
        if (import_constants.isWeb) {
          (0, import_registerCSSVariable.registerCSSVariable)(variable);
          const variableNeedsPx = variable.needsPx === true;
          const categoryNeedsPx = shouldTokenCategoryHaveUnits(key);
          const shouldBeUnitless = !(variableNeedsPx || categoryNeedsPx);
          declarations.push((0, import_registerCSSVariable.variableToCSS)(variable, shouldBeUnitless));
        }
      }
    }
    return declarations;
  }
  return [];
}
function createFontCSS(fontsParsed, registerFontVariables) {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    const fontDeclarations = {};
    if (!fontsParsed) return fontDeclarations;
    const sortedFontKeys = Object.keys(fontsParsed).sort();
    for (const key of sortedFontKeys) {
      const fontParsed = fontsParsed[key];
      const [name, language] = key.includes("_") ? key.split("_") : [key];
      const fontVars = registerFontVariables(fontParsed);
      fontDeclarations[key] = {
        name: name.slice(1),
        declarations: fontVars,
        language,
        fontParsed
      };
    }
    return fontDeclarations;
  }
  return {};
}
function buildCSSRuleSets(declarations, fontDeclarations, defaultFontToken = "$true") {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    let declarationsToRuleSet = function (decs, selector = "") {
      return `:root${selector} {${sep}${[...decs].join(`;${sep}`)}
}`;
    };
    const cssRuleSets = [];
    const sep = " ";
    if (declarations.length) cssRuleSets.push(declarationsToRuleSet(declarations));
    const fontSelectors = [];
    const sortedFontDeclarationKeys = Object.keys(fontDeclarations).sort();
    for (const key of sortedFontDeclarationKeys) {
      const {
        name,
        declarations: declarations2,
        language = "default"
      } = fontDeclarations[key];
      const fontSelector = `.font_${name}`;
      fontSelectors.push(fontSelector);
      const langSelector = `:root .t_lang-${name}-${language} ${fontSelector}`;
      const specificRuleSet = declarationsToRuleSet(declarations2, language === "default" ? ` ${fontSelector}, ${langSelector}` : langSelector);
      cssRuleSets.push(specificRuleSet);
    }
    if (fontSelectors.length) {
      const firstFont = fontDeclarations[sortedFontDeclarationKeys[0]];
      if (firstFont?.fontParsed) {
        const fontProps = getFontPropertyDeclarations(firstFont.fontParsed, defaultFontToken);
        const sharedSelectors = [...fontSelectors, ".is_View"].join(", ");
        cssRuleSets.push(`${sharedSelectors} {${fontProps.join("; ")}}`);
      }
    }
    return cssRuleSets;
  }
  return [];
}
function createThemeCSS(dedupedThemes, configIn) {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    let themeRuleSets = [];
    if (import_constants.isWeb) for (const {
      names,
      theme
    } of dedupedThemes) {
      const nextRules = (0, import_getThemeCSSRules.getThemeCSSRules)({
        config: configIn,
        themeName: names[0],
        names,
        theme
      });
      themeRuleSets = [...themeRuleSets, ...nextRules];
    }
    return themeRuleSets;
  }
  return [];
}
function getCSS(themeConfig, opts = {}, lastIndex) {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS && true) {
    const {
      separator = "\n",
      sinceLastCall,
      exclude
    } = opts;
    if (sinceLastCall && lastIndex.value >= 0) {
      const rules = (0, import_insertStyleRule.getAllRules)();
      const newRules = rules.slice(lastIndex.value);
      lastIndex.value = rules.length;
      return newRules.join(separator);
    }
    lastIndex.value = 0;
    const runtimeStyles = (0, import_insertStyleRule.getAllRules)().join(separator);
    if (exclude === "design-system") return runtimeStyles;
    const themeRules = exclude ? "" : themeConfig.getThemeRulesSets().join(separator);
    return `${`._ovs-contain {overscroll-behavior:contain;}
.t_unmounted .is_View, .t_unmounted .is_Text { transition: none !important; }
.is_View { display: flex; align-items: stretch; flex-direction: column; flex-basis: auto; box-sizing: border-box; min-height: 0; min-width: 0; flex-shrink: 0; }
.is_Text { display: inline; box-sizing: border-box; word-wrap: break-word; white-space: pre-wrap; margin: 0; }
@scope (.is_Text) to (.is_View) { .is_Text { white-space: inherit; word-wrap: inherit; } }
._dsp_contents {display:contents;}
._no_backdrop::backdrop {display: none;}
.is_Input::selection, .is_TextArea::selection {background-color: var(--selectionColor);}
.is_Input::placeholder, .is_TextArea::placeholder {color: var(--placeholderColor);}
:root ._pe-boxonly>* {pointer-events:none;}
:root ._pe-boxnone>* {pointer-events:auto;}
._hsb-x::-webkit-scrollbar:horizontal { display: none !important; }
._hsb-y::-webkit-scrollbar:vertical { display: none !important; }
._hsb-x { scrollbar-width: none !important; }
._hsb-y { scrollbar-width: none !important; }
${import_registerCSSVariable.autoVariables.length ? `:root{${import_registerCSSVariable.autoVariables.map(v => `--${v.name}:${v.val}`).join(";")}}` : ""}
${themeConfig.cssRuleSets.join(separator)}`}
${themeRules}
${runtimeStyles}`;
  }
  return "";
}
