// @tamagui/web/dist/cjs/helpers/defaultAnimationDriver.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var defaultAnimationDriver_exports = {};
__export(defaultAnimationDriver_exports, {
  defaultAnimationDriver: () => defaultAnimationDriver
});
module.exports = __toCommonJS(defaultAnimationDriver_exports);
const noAnimationDriver = method => {
  console.warn(`No animation driver configured. To use ${method}, you must pass \`animations\` to createTamagui. See: https://tamagui.dev/docs/core/animations`);
};
const createEmptyAnimationDriver = () => ({
  isReactNative: false,
  inputStyle: "css",
  outputStyle: "css",
  isStub: true,
  animations: {},
  useAnimations: () => noAnimationDriver("animations"),
  usePresence: () => noAnimationDriver("usePresence"),
  ResetPresence: () => noAnimationDriver("ResetPresence"),
  useAnimatedNumber: () => noAnimationDriver("useAnimatedNumber"),
  useAnimatedNumberStyle: () => noAnimationDriver("useAnimatedNumberStyle"),
  useAnimatedNumbersStyle: () => noAnimationDriver("useAnimatedNumbersStyle"),
  useAnimatedNumberReaction: () => noAnimationDriver("useAnimatedNumberReaction")
});
const defaultAnimationDriver = createEmptyAnimationDriver();
