// @tamagui/web/dist/cjs/helpers/getTokenForKey.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var getTokenForKey_exports = {};
__export(getTokenForKey_exports, {
  getLastFontFamilyToken: () => getLastFontFamilyToken,
  getTokenForKey: () => getTokenForKey,
  resolveVariableValue: () => resolveVariableValue,
  setLastFontFamilyToken: () => setLastFontFamilyToken
});
module.exports = __toCommonJS(getTokenForKey_exports);
var import_helpers = require("@tamagui/helpers");
var import_config = require("../config.native.js");
var import_createVariable = require("../createVariable.native.js");
var import_getVariantExtras = require("./getVariantExtras.native.js");
var import_normalizeColor = require("./normalizeColor.native.js");
var fontShorthand = {
  fontSize: "size",
  fontWeight: "weight"
};
var didLogMissingToken = false;
var colorKeys = import_helpers.tokenCategories.color;
var _lastFontFamilyToken = null;
function getLastFontFamilyToken() {
  return _lastFontFamilyToken;
}
function setLastFontFamilyToken(value) {
  _lastFontFamilyToken = value;
}
var getTokenForKey = function (key, value, styleProps, styleState) {
  var _staticConfig_accept;
  var resolveAs = styleProps.resolveValues || "none";
  if (resolveAs === "none") return value;
  var opacityModifier;
  if (key in colorKeys) {
    var slashIdx = value.indexOf("/");
    if (slashIdx > 0) {
      var raw = value.slice(slashIdx + 1);
      if (raw.length > 0) {
        var num = Number(raw);
        if (!Number.isNaN(num)) {
          opacityModifier = Math.max(0, Math.min(1, num / 100));
          value = value.slice(0, slashIdx);
        }
      }
    }
  }
  var {
    theme,
    conf = (0, import_config.getConfig)(),
    context,
    fontFamily,
    staticConfig
  } = styleState;
  var themeValue = theme ? theme[value] || theme[value.slice(1)] : void 0;
  var tokensParsed = conf.tokensParsed;
  var valOrVar;
  var hasSet = false;
  var customTokenAccept = staticConfig === null || staticConfig === void 0 ? void 0 : (_staticConfig_accept = staticConfig.accept) === null || _staticConfig_accept === void 0 ? void 0 : _staticConfig_accept[key];
  if (customTokenAccept) {
    var _tokensParsed_customTokenAccept;
    var val = themeValue !== null && themeValue !== void 0 ? themeValue : (_tokensParsed_customTokenAccept = tokensParsed[customTokenAccept]) === null || _tokensParsed_customTokenAccept === void 0 ? void 0 : _tokensParsed_customTokenAccept[value];
    if (val != null) {
      resolveAs = "value";
      valOrVar = val;
      hasSet = true;
    }
  }
  if (themeValue) {
    if (resolveAs === "except-theme") return value;
    valOrVar = themeValue;
    if (process.env.NODE_ENV === "development" && styleState.debug === "verbose") {
      globalThis.tamaguiAvoidTracking = true;
      console.info(` - resolving ${key} to theme value ${value} resolveAs ${resolveAs}`, valOrVar);
      globalThis.tamaguiAvoidTracking = false;
    }
    hasSet = true;
  } else {
    if (value in conf.specificTokens) {
      hasSet = true;
      valOrVar = conf.specificTokens[value];
    } else {
      switch (key) {
        case "fontFamily":
          var _fontsParsed_value;
          valOrVar = ((_fontsParsed_value = ((context === null || context === void 0 ? void 0 : context.language) ? (0, import_getVariantExtras.getFontsForLanguage)(conf.fontsParsed, context.language) : conf.fontsParsed)[value]) === null || _fontsParsed_value === void 0 ? void 0 : _fontsParsed_value.family) || value;
          setLastFontFamilyToken(value);
          hasSet = true;
          break;
        case "fontSize":
        case "lineHeight":
        case "letterSpacing":
        case "fontWeight":
          var fam = fontFamily || conf.defaultFontToken;
          if (fam) {
            var _font_;
            var fontsParsed1 = (context === null || context === void 0 ? void 0 : context.language) ? (0, import_getVariantExtras.getFontsForLanguage)(conf.fontsParsed, context.language) : conf.fontsParsed;
            var font = fontsParsed1[fam] || fontsParsed1[conf.defaultFontToken];
            valOrVar = (font === null || font === void 0 ? void 0 : (_font_ = font[fontShorthand[key] || key]) === null || _font_ === void 0 ? void 0 : _font_[value]) || value;
            hasSet = true;
          }
          break;
      }
      for (var cat in import_helpers.tokenCategories) if (key in import_helpers.tokenCategories[cat]) {
        var _tokensParsed_cat;
        var res = (_tokensParsed_cat = tokensParsed[cat]) === null || _tokensParsed_cat === void 0 ? void 0 : _tokensParsed_cat[value];
        if (res != null) {
          valOrVar = res;
          hasSet = true;
        } else if (process.env.NODE_ENV === "development") {
          if (process.env.TAMAGUI_DISABLE_MISSING_TOKEN_LOG !== "1") {
            if (!didLogMissingToken) {
              didLogMissingToken = true;
              console.groupCollapsed(`[tamagui] Warning: missing token ${key} in category ${cat} - ${value} (open for details)`);
              console.info(`Note: this could just be due to you not setting all the theme tokens Tamagui expects, which is harmless, but
                    it also often can be because you have a duplicated Tamagui in your bundle, which can cause tricky bugs.`);
              console.info(`To see if you have duplicated dependencies, in Chrome DevTools hit CMD+P and type TamaguiProvider.
                    If you see both a .cjs and a .mjs entry, it's duplicated.`);
              console.info(`You can debug that issue by opening the .mjs and .cjs files and setting a breakpoint at the top of each.`);
              console.info(`We only log this warning one time as it's sometimes harmless, to disable this log entirely set process.env.TAMAGUI_DISABLE_MISSING_TOKEN_LOG=1.`);
              console.groupEnd();
            }
          }
        }
      }
    }
    if (!hasSet) {
      var spaceVar = tokensParsed.space[value];
      if (spaceVar != null) {
        valOrVar = spaceVar;
        hasSet = true;
      }
    }
  }
  if (hasSet) {
    var out = resolveVariableValue(key, valOrVar, resolveAs);
    if (opacityModifier !== void 0 && opacityModifier < 1) {
      var _normalizeColor;
      out = (_normalizeColor = (0, import_normalizeColor.normalizeColor)(String(out), opacityModifier)) !== null && _normalizeColor !== void 0 ? _normalizeColor : out;
    }
    if (process.env.NODE_ENV === "development" && styleState.debug === "verbose") {
      globalThis.tamaguiAvoidTracking = true;
      console.info(`resolved`, resolveAs, valOrVar, out);
      globalThis.tamaguiAvoidTracking = false;
    }
    return out;
  }
  if (process.env.NODE_ENV === "development" && styleState.debug === "verbose") console.warn(`Warning: no token found for ${key}, omitting`);
};
function resolveVariableValue(key, valOrVar, resolveValues) {
  if (resolveValues === "none") return valOrVar;
  if ((0, import_createVariable.isVariable)(valOrVar)) {
    if (resolveValues === "value") return valOrVar.val;
    var get = valOrVar === null || valOrVar === void 0 ? void 0 : valOrVar.get;
    if (key !== "shadowColor") {
      if (typeof get === "function") return get(resolveValues === "web" ? "web" : void 0);
    }
    return valOrVar.val;
  }
  return valOrVar;
}
