// @tamagui/web/dist/cjs/helpers/insertStyleRule.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var insertStyleRule_exports = {};
__export(insertStyleRule_exports, {
  getAllRules: () => getAllRules,
  getAllSelectors: () => getAllSelectors,
  insertStyleRules: () => insertStyleRules,
  scanAllSheets: () => scanAllSheets,
  setNonce: () => setNonce,
  shouldInsertStyleRules: () => shouldInsertStyleRules,
  stopAccumulatingRules: () => stopAccumulatingRules,
  updateRules: () => updateRules
});
module.exports = __toCommonJS(insertStyleRule_exports);
require("@tamagui/helpers");
require("../createVariable.native.js");
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) return !!right[Symbol.hasInstance](left);else return left instanceof right;
}
var totalSelectorsInserted = /* @__PURE__ */new Map();
var allSelectors = {};
var allRules = {};
var getAllSelectors = function () {
  return allSelectors;
};
var getAllRules = function () {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) return Object.keys(allRules).sort().map(function (key) {
    return allRules[key];
  });
  return [];
};
function scanAllSheets() {
  arguments.length > 0 && arguments[0] !== void 0 && arguments[0];
  arguments.length > 1 && arguments[1];
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    if (process.env.NODE_ENV === "test") return;
    return;
  }
}
var bailAfterEnv = process.env.TAMAGUI_BAIL_AFTER_SCANNING_X_CSS_RULES;
bailAfterEnv && +bailAfterEnv;
var tamaguiSelectorRegex = /\.tm_xxt/;
function getTamaguiSelector(rule) {
  var collectThemes = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : false;
  if (_instanceof(rule, CSSStyleRule)) {
    var text = rule.selectorText;
    if (text[0] === ":" && text[1] === "r" && tamaguiSelectorRegex.test(text)) {
      var id = getIdentifierFromTamaguiSelector(
      // next.js minifies it so its in front
      text.replace(tamaguiSelectorRegex, ""));
      return collectThemes ? [id, rule, true] : [id, rule];
    }
  } else if (_instanceof(rule, CSSMediaRule)) {
    if (rule.cssRules.length > 1) return;
    return getTamaguiSelector(rule.cssRules[0]);
  }
}
var getIdentifierFromTamaguiSelector = function (selector) {
  var dotIndex = selector.indexOf(":");
  if (dotIndex > -1) return selector.slice(7, dotIndex);
  return selector.slice(7);
};
var trackAllRules = true;
function stopAccumulatingRules() {
  trackAllRules = false;
}
function updateRules(identifier, rules) {
  if (trackAllRules) allRules[identifier] = rules.join(" ");
  return true;
}
function setNonce(_) {}
function insertStyleRules(rulesToInsert) {}
var maxToInsert = process.env.TAMAGUI_INSERT_SELECTOR_TRIES ? +process.env.TAMAGUI_INSERT_SELECTOR_TRIES : 1;
function shouldInsertStyleRules(identifier) {
  if (process.env.IS_STATIC === "is_static") return true;
  var total = totalSelectorsInserted.get(identifier) || 0;
  if (process.env.NODE_ENV === "development") {
    if (total > +(process.env.TAMAGUI_STYLE_INSERTION_WARNING_LIMIT || 10)) console.warn(`Warning: inserting many CSS rules, you may be animating something and generating many CSS insertions, which can degrade performance. Instead, try using the "disableClassName" property on elements that change styles often. To disable this warning set TAMAGUI_STYLE_INSERTION_WARNING_LIMIT from 50000 to something higher`);
  }
  return total < maxToInsert;
}
