// @tamagui/web/dist/cjs/helpers/pointerEvents.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var pointerEvents_native_exports = {};
__export(pointerEvents_native_exports, {
  usePointerEvents: () => usePointerEvents
});
module.exports = __toCommonJS(pointerEvents_native_exports);
var import_react = require("react");
var import_helpers = require("@tamagui/helpers");
function usePointerEvents(props, viewProps) {
  var {
    onPointerDown,
    onPointerUp,
    onPointerMove,
    onPointerCancel,
    onPointerEnter,
    onPointerLeave
  } = props;
  var hasPointerEvents = onPointerDown || onPointerUp || onPointerMove || onPointerCancel || onPointerEnter || onPointerLeave;
  var isInsideRef = (0, import_react.useRef)(false);
  var layoutRef = (0, import_react.useRef)({
    width: 0,
    height: 0
  });
  var isCapturedRef = (0, import_react.useRef)(false);
  if (!hasPointerEvents) return;
  var createNormalizedEvent = function (e) {
    var touch = e.nativeEvent;
    var _touch_identifier;
    var normalized = {
      ...e,
      clientX: touch.pageX,
      clientY: touch.pageY,
      pageX: touch.pageX,
      pageY: touch.pageY,
      offsetX: touch.locationX,
      offsetY: touch.locationY,
      pointerType: "touch",
      pointerId: (_touch_identifier = touch.identifier) !== null && _touch_identifier !== void 0 ? _touch_identifier : 0,
      nativeEvent: touch,
      target: {
        setPointerCapture: function (_pointerId) {
          isCapturedRef.current = true;
        },
        releasePointerCapture: function (_pointerId) {
          isCapturedRef.current = false;
        }
      }
    };
    return normalized;
  };
  if (onPointerDown) {
    viewProps.onTouchStart = (0, import_helpers.composeEventHandlers)(viewProps.onTouchStart, function (e) {
      onPointerDown(createNormalizedEvent(e));
    });
  }
  if (onPointerUp) {
    viewProps.onTouchEnd = (0, import_helpers.composeEventHandlers)(viewProps.onTouchEnd, function (e) {
      isCapturedRef.current = false;
      onPointerUp(createNormalizedEvent(e));
    });
  }
  if (onPointerMove) {
    viewProps.onTouchMove = (0, import_helpers.composeEventHandlers)(viewProps.onTouchMove, function (e) {
      var {
        locationX,
        locationY
      } = e.nativeEvent;
      var {
        width,
        height
      } = layoutRef.current;
      var isInBounds = locationX >= 0 && locationX <= width && locationY >= 0 && locationY <= height;
      if (isCapturedRef.current || isInBounds) {
        onPointerMove(createNormalizedEvent(e));
      }
    });
  }
  if (onPointerCancel) {
    viewProps.onTouchCancel = (0, import_helpers.composeEventHandlers)(viewProps.onTouchCancel, function (e) {
      isCapturedRef.current = false;
      onPointerCancel(createNormalizedEvent(e));
    });
  }
  if (onPointerEnter || onPointerLeave || onPointerMove) {
    viewProps.onLayout = (0, import_helpers.composeEventHandlers)(viewProps.onLayout, function (e) {
      layoutRef.current = {
        width: e.nativeEvent.layout.width,
        height: e.nativeEvent.layout.height
      };
    });
  }
  if (onPointerEnter) {
    viewProps.onTouchStart = (0, import_helpers.composeEventHandlers)(viewProps.onTouchStart, function (e) {
      var {
        locationX,
        locationY
      } = e.nativeEvent;
      var {
        width,
        height
      } = layoutRef.current;
      if (locationX >= 0 && locationX <= width && locationY >= 0 && locationY <= height) {
        isInsideRef.current = true;
        onPointerEnter(createNormalizedEvent(e));
      }
    });
  }
  if (onPointerLeave) {
    viewProps.onTouchMove = (0, import_helpers.composeEventHandlers)(viewProps.onTouchMove, function (e) {
      var {
        locationX,
        locationY
      } = e.nativeEvent;
      var {
        width,
        height
      } = layoutRef.current;
      var isInside = locationX >= 0 && locationX <= width && locationY >= 0 && locationY <= height;
      if (isInsideRef.current && !isInside) {
        isInsideRef.current = false;
        onPointerLeave(createNormalizedEvent(e));
      }
    });
    viewProps.onTouchEnd = (0, import_helpers.composeEventHandlers)(viewProps.onTouchEnd, function (e) {
      if (isInsideRef.current) {
        isInsideRef.current = false;
        onPointerLeave(createNormalizedEvent(e));
      }
    });
  }
}
