// @tamagui/web/dist/cjs/helpers/registerCSSVariable.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var registerCSSVariable_exports = {};
__export(registerCSSVariable_exports, {
  autoVariables: () => autoVariables,
  getOrCreateMutatedVariable: () => getOrCreateMutatedVariable,
  getOrCreateVariable: () => getOrCreateVariable,
  mutatedAutoVariables: () => mutatedAutoVariables,
  registerCSSVariable: () => registerCSSVariable,
  tokensValueToVariable: () => tokensValueToVariable,
  variableToCSS: () => variableToCSS
});
module.exports = __toCommonJS(registerCSSVariable_exports);
var import_createVariable = require("../createVariable.native.js");
var registerCSSVariable = function (v) {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    tokensValueToVariable.set((0, import_createVariable.getVariableValue)(v), v);
  }
};
var variableToCSS = function (v) {
  var unitless = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : false;
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    return `--${process.env.TAMAGUI_CSS_VARIABLE_PREFIX || ""}${(0, import_createVariable.createCSSVariable)(v.name, false)}:${!unitless && typeof v.val === "number" ? `${v.val}px` : v.val}`;
  }
  return "";
};
var tokensValueToVariable = /* @__PURE__ */new Map();
var autoVarId = 0;
var autoVariables = [];
var getOrCreateVariable = function (val) {
  if (tokensValueToVariable.has(val)) {
    return tokensValueToVariable.get(val);
  }
  var name = `t${autoVarId++}`;
  var variable = `var(--${name})`;
  var v = {
    val,
    name,
    variable
  };
  tokensValueToVariable.set(val, v);
  autoVariables.push(v);
  return v;
};
var mutatedVarId = 1e4;
var mutatedAutoVariables = [];
var mutatedTokensValueToVariable = /* @__PURE__ */new Map();
var getOrCreateMutatedVariable = function (val) {
  if (mutatedTokensValueToVariable.has(val)) {
    return mutatedTokensValueToVariable.get(val);
  }
  var name = `t${mutatedVarId++}`;
  var variable = `var(--${name})`;
  var v = {
    val,
    name,
    variable
  };
  mutatedTokensValueToVariable.set(val, v);
  mutatedAutoVariables.push(v);
  return v;
};
