// @tamagui/radio-headless/dist/esm/BubbleInput.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import React from "react";
import { usePrevious } from "@tamagui/use-previous";
var BubbleInput = function (props) {
  var {
    checked,
    bubbles = true,
    control,
    isHidden,
    accentColor,
    ...inputProps
  } = props;
  var ref = React.useRef(null);
  var prevChecked = usePrevious(checked);
  React.useEffect(function () {
    var input = ref.current;
    var inputProto = window.HTMLInputElement.prototype;
    var descriptor = Object.getOwnPropertyDescriptor(inputProto, "checked");
    var setChecked = descriptor.set;
    if (prevChecked !== checked && setChecked) {
      var event = new Event("click", {
        bubbles
      });
      setChecked.call(input, checked);
      input.dispatchEvent(event);
    }
  }, [prevChecked, checked, bubbles]);
  return /* @__PURE__ */_jsx("input", {
    type: "radio",
    defaultChecked: checked,
    ...inputProps,
    tabIndex: -1,
    ref,
    "aria-hidden": isHidden,
    style: {
      ...(isHidden ? {
        // ...controlSize,
        position: "absolute",
        pointerEvents: "none",
        opacity: 0,
        margin: 0
      } : {
        appearance: "auto",
        accentColor
      }),
      ...props.style
    }
  });
};
export { BubbleInput };
