// @tamagui/radio-headless/dist/esm/useRadioGroup.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { registerFocusable } from "@tamagui/focusable";
import { composeEventHandlers } from "@tamagui/helpers";
import { useLabelContext } from "@tamagui/label";
import { useControllableState } from "@tamagui/use-controllable-state";
import { useContext, useEffect, useRef, useState } from "react";
import { BubbleInput } from "./BubbleInput.native.js";
import { getState } from "./utils.native.js";
function useRadioGroup(params) {
  var {
    value: valueProp,
    onValueChange,
    defaultValue,
    required,
    disabled,
    name,
    native,
    accentColor,
    orientation,
    ref
  } = params;
  var [value, setValue] = useControllableState({
    prop: valueProp,
    defaultProp: defaultValue,
    onChange: onValueChange
  });
  return {
    providerValue: {
      value,
      onChange: setValue,
      required,
      disabled,
      name,
      native,
      accentColor
    },
    frameAttrs: {
      role: "radiogroup",
      "aria-orientation": orientation,
      "data-disabled": disabled ? "" : void 0
    },
    rovingFocusGroupAttrs: {
      orientation,
      loop: true
    }
  };
}
var ARROW_KEYS = ["ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight"];
var useRadioGroupItem = function (params) {
  var {
    radioGroupContext,
    value,
    labelledBy: ariaLabelledby,
    disabled: itemDisabled,
    ref: refProp,
    id,
    onPress,
    onKeyDown,
    onFocus
  } = params;
  var {
    value: groupValue,
    disabled,
    required,
    onChange,
    name,
    native,
    accentColor
  } = useContext(radioGroupContext);
  var [button, setButton] = useState(null);
  var hasConsumerStoppedPropagationRef = useRef(false);
  var ref = useRef(null);
  var composedRefs = useComposedRefs(refProp, function (node) {
    return setButton(node);
  }, ref);
  var isArrowKeyPressedRef = useRef(false);
  var isFormControl = isWeb ? button ? Boolean(button.closest("form")) : true : false;
  var checked = groupValue === value;
  var labelId = useLabelContext(button);
  var labelledBy = ariaLabelledby || labelId;
  useEffect(function () {
    if (isWeb) {
      var handleKeyDown = function (event) {
        if (ARROW_KEYS.includes(event.key)) isArrowKeyPressedRef.current = true;
      };
      var handleKeyUp = function () {
        isArrowKeyPressedRef.current = false;
      };
      document.addEventListener("keydown", handleKeyDown);
      document.addEventListener("keyup", handleKeyUp);
      return function () {
        document.removeEventListener("keydown", handleKeyDown);
        document.removeEventListener("keyup", handleKeyUp);
      };
    }
  }, []);
  useEffect(function () {
    if (!id) return;
    if (disabled) return;
    return registerFocusable(id, {
      focusAndSelect: function () {
        onChange === null || onChange === void 0 || onChange(value);
      },
      focus: function () {}
    });
  }, [id, value, disabled]);
  var isDisabled = disabled || itemDisabled;
  return {
    providerValue: {
      checked
    },
    checked,
    isFormControl,
    bubbleInput: /* @__PURE__ */_jsx(BubbleInput, {
      isHidden: !native,
      control: button,
      bubbles: !hasConsumerStoppedPropagationRef.current,
      name,
      value,
      checked,
      required,
      disabled: isDisabled,
      ...(isWeb && native && {
        accentColor,
        id
      })
    }),
    native,
    frameAttrs: {
      "data-state": getState(checked),
      "data-disabled": isDisabled ? "" : void 0,
      role: "radio",
      "aria-labelledby": labelledBy,
      "aria-checked": checked,
      "aria-required": required,
      disabled: isDisabled,
      ref: composedRefs,
      ...(isWeb && {
        type: "button",
        value
      }),
      id,
      onPress: composeEventHandlers(onPress, function (event) {
        if (!checked) onChange === null || onChange === void 0 || onChange(value);
        if (isFormControl) {
          hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
          if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
        }
      }),
      ...(isWeb && {
        onKeyDown: composeEventHandlers(onKeyDown, function (event) {
          if (event.key === "Enter" || event.key === " ") {
            if (!checked) onChange === null || onChange === void 0 || onChange(value);
          }
        }),
        onFocus: composeEventHandlers(onFocus, function () {
          if (isArrowKeyPressedRef.current) {
            var _ref_current;
            (_ref_current = ref.current) === null || _ref_current === void 0 || _ref_current.click();
          }
        })
      })
    },
    rovingFocusGroupAttrs: {
      asChild: "except-style",
      focusable: !isDisabled,
      active: checked
    }
  };
};
function useRadioGroupItemIndicator(params) {
  var {
    radioGroupItemContext,
    disabled,
    ...rest
  } = params;
  var {
    checked
  } = useContext(radioGroupItemContext);
  return {
    checked,
    "data-state": getState(checked),
    "data-disabled": disabled ? "" : void 0,
    ...rest
  };
}
export { useRadioGroup, useRadioGroupItem, useRadioGroupItemIndicator };
