// @tamagui/floating/dist/esm/middleware/inner.native.js
import * as ReactDOM from "react-dom";
import { detectOverflow, offset } from "@floating-ui/react-dom";
function getArgsWithCustomFloatingHeight(state, height) {
  return {
    ...state,
    rects: {
      ...state.rects,
      floating: {
        ...state.rects.floating,
        height
      }
    }
  };
}
var inner = function (props) {
  return {
    name: "inner",
    options: props,
    async fn(state) {
      var _listRef_current;
      var {
        listRef,
        overflowRef,
        onFallbackChange,
        offset: innerOffset = 0,
        index = 0,
        minItemsVisible = 4,
        referenceOverflowThreshold = 0,
        scrollRef,
        padding = 0
      } = props;
      var {
        rects,
        elements: {
          floating
        }
      } = state;
      var item = (_listRef_current = listRef.current) === null || _listRef_current === void 0 ? void 0 : _listRef_current[index !== null && index !== void 0 ? index : 0];
      var scrollEl = (scrollRef === null || scrollRef === void 0 ? void 0 : scrollRef.current) || floating;
      var clientTop = floating.clientTop || scrollEl.clientTop;
      var floatingIsBordered = floating.clientTop !== 0;
      var scrollElIsBordered = scrollEl.clientTop !== 0;
      var floatingIsScrollEl = floating === scrollEl;
      if (!item || index == null) {
        onFallbackChange === null || onFallbackChange === void 0 ? void 0 : onFallbackChange(true);
        return {};
      }
      var nextArgs = {
        ...state,
        ...(await offset(-item.offsetTop - floating.clientTop - rects.reference.height / 2 - item.offsetHeight / 2 - innerOffset).fn(state))
      };
      var detectOverflowOptions = {
        padding
      };
      var overflow = await detectOverflow(getArgsWithCustomFloatingHeight(nextArgs, scrollEl.scrollHeight + clientTop + floating.clientTop), detectOverflowOptions);
      var refOverflow = await detectOverflow(nextArgs, {
        ...detectOverflowOptions,
        elementContext: "reference"
      });
      var diffY = Math.max(0, overflow.top);
      var nextY = nextArgs.y + diffY;
      var isScrollable = scrollEl.scrollHeight > scrollEl.clientHeight;
      var rounder = isScrollable ? function (v) {
        return v;
      } : Math.round;
      var maxHeight = rounder(Math.max(0, scrollEl.scrollHeight + (floatingIsBordered && floatingIsScrollEl || scrollElIsBordered ? clientTop * 2 : 0) - diffY - Math.max(0, overflow.bottom)));
      scrollEl.style.maxHeight = `${maxHeight}px`;
      scrollEl.scrollTop = diffY;
      if (onFallbackChange) {
        var _listRef_current1;
        var _listRef_current_length;
        var shouldFallback = scrollEl.offsetHeight < item.offsetHeight * Math.min(minItemsVisible, (_listRef_current_length = (_listRef_current1 = listRef.current) === null || _listRef_current1 === void 0 ? void 0 : _listRef_current1.length) !== null && _listRef_current_length !== void 0 ? _listRef_current_length : 0) - 1 || refOverflow.top >= -referenceOverflowThreshold || refOverflow.bottom >= -referenceOverflowThreshold;
        ReactDOM.flushSync(function () {
          return onFallbackChange(shouldFallback);
        });
      }
      if (overflowRef) {
        ;
        overflowRef.current = await detectOverflow(getArgsWithCustomFloatingHeight({
          ...nextArgs,
          y: nextY
        }, scrollEl.offsetHeight + clientTop + floating.clientTop), detectOverflowOptions);
      }
      return {
        y: nextY
      };
    }
  };
};
export { inner };
