// @tamagui/image/dist/cjs/createImage.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createImage_exports = {};
__export(createImage_exports, {
  createImage: () => createImage
});
module.exports = __toCommonJS(createImage_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_web = require("@tamagui/web");
var defaultObjectFitMap = function (objectFit) {
  switch (objectFit) {
    case "fill":
      return "stretch";
    case "none":
      return "center";
    case "scale-down":
      return "contain";
    case "contain":
      return "contain";
    default:
      return "cover";
  }
};
var defaultTransformSource = function (props) {
  var {
    src,
    source,
    width,
    height
  } = props;
  if (source) return source;
  if (src && typeof src !== "string") return src;
  return {
    uri: src,
    width,
    height
  };
};
function createImage(options) {
  var {
    Component,
    mapObjectFitToResizeMode = defaultObjectFitMap,
    resizeModePropName = "resizeMode",
    objectPositionPropName,
    transformSource = defaultTransformSource
  } = options;
  var inlinePropsSet = /* @__PURE__ */new Set(["source", "placeholder", "transition", "contentFit", "contentPosition", "cachePolicy", "recyclingKey", "allowDownscaling", "autoplay", "blurRadius", "priority", "placeholderContentFit", "responsivePolicy", "onLoadStart", "onProgress", "onLoadEnd"]);
  var StyledImage = (0, import_web.styled)(Component, {
    name: "Image"
  }, {
    inlineProps: inlinePropsSet
  });
  var ImageComponent = StyledImage.styleable(function (incomingProps, ref) {
    var props = incomingProps;
    var {
      src,
      width,
      height,
      borderRadius,
      objectFit,
      objectPosition,
      // web only props - filter out on native
      decoding,
      elementTiming,
      fetchpriority,
      isMap,
      loading,
      sizes,
      useMap,
      onLoad,
      onError,
      ...rest
    } = props;
    var resolvedWidth = typeof width === "string" && width[0] === "$" ? (0, import_web.getTokenValue)(width, "size") : width;
    var resolvedHeight = typeof height === "string" && height[0] === "$" ? (0, import_web.getTokenValue)(height, "size") : height;
    var finalSource = transformSource({
      src,
      width: resolvedWidth,
      height: resolvedHeight
    });
    var incomingStyle = Array.isArray(rest.style) ? Object.assign({}, ...rest.style.flat()) : rest.style;
    var resolvedBorderRadius = typeof borderRadius === "string" && borderRadius[0] === "$" ? (0, import_web.getTokenValue)(borderRadius, "radius") : borderRadius;
    var finalProps = {
      ...rest,
      source: finalSource,
      style: {
        ...incomingStyle,
        ...(resolvedBorderRadius !== void 0 && {
          borderRadius: resolvedBorderRadius
        }),
        ...(resolvedWidth !== void 0 && {
          width: resolvedWidth
        }),
        ...(resolvedHeight !== void 0 && {
          height: resolvedHeight
        })
      }
    };
    if (objectFit) {
      finalProps[resizeModePropName] = mapObjectFitToResizeMode(objectFit);
    }
    if (objectPositionPropName && objectPosition) {
      finalProps[objectPositionPropName] = objectPosition;
    }
    if (onLoad) {
      finalProps.onLoad = function (e) {
        var _e_nativeEvent;
        var source = (e === null || e === void 0 ? void 0 : (_e_nativeEvent = e.nativeEvent) === null || _e_nativeEvent === void 0 ? void 0 : _e_nativeEvent.source) || (e === null || e === void 0 ? void 0 : e.source) || {};
        onLoad({
          target: {
            naturalHeight: source === null || source === void 0 ? void 0 : source.height,
            naturalWidth: source === null || source === void 0 ? void 0 : source.width
          },
          type: "load"
        });
      };
    }
    if (onError) {
      finalProps.onError = function () {
        onError({
          type: "error"
        });
      };
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Component, {
      ref,
      ...finalProps
    });
  });
  var comp = Component;
  ImageComponent.getSize = comp.getSize || function () {};
  ImageComponent.getSizeWithHeaders = comp.getSizeWithHeaders || function () {};
  ImageComponent.prefetch = comp.prefetch || function () {};
  ImageComponent.prefetchWithMetadata = comp.prefetchWithMetadata || function () {};
  ImageComponent.abortPrefetch = comp.abortPrefetch || function () {};
  ImageComponent.queryCache = comp.queryCache || function () {};
  return ImageComponent;
}
