// @tamagui/alert-dialog/dist/jsx/AlertDialog.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb, useIsomorphicLayoutEffect } from "@tamagui/constants";
import { Slottable, View, createStyledContext, isTamaguiElement, styled } from "@tamagui/core";
import { Dialog, DialogClose, DialogContent, DialogDescription, DialogOverlay, DialogOverlayFrame, DialogPortal, DialogTitle, DialogTrigger, DialogWarningProvider } from "@tamagui/dialog";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { useControllableState } from "@tamagui/use-controllable-state";
import * as React from "react";
import { Alert } from "react-native-web";
import { jsx, jsxs } from "react/jsx-runtime";
const getAlertDialogScope = scope => scope;
const ROOT_NAME = "AlertDialog";
const NativeAlertDialogTriggerFrame = styled(View, {
  name: "AlertDialogTrigger"
});
const AlertDialogTrigger = NativeAlertDialogTriggerFrame.styleable(function AlertDialogTrigger2(props, forwardedRef) {
  if (props["__native"]) {
    const {
      __native,
      onPress,
      __onPress,
      ...rest
    } = props;
    return /* @__PURE__ */jsx(NativeAlertDialogTriggerFrame, {
      ...rest,
      onPress: composeEventHandlers(onPress, __onPress)
    });
  }
  const {
    scope,
    ...triggerProps
  } = props;
  return /* @__PURE__ */jsx(DialogTrigger, {
    scope: getAlertDialogScope(scope),
    ...triggerProps,
    ref: forwardedRef
  });
});
const AlertDialogPortal = function AlertDialogPortal2(props) {
  const {
    scope,
    ...portalProps
  } = props;
  return /* @__PURE__ */jsx(DialogPortal, {
    scope: getAlertDialogScope(scope),
    ...portalProps
  });
};
const AlertDialogOverlay = styled(DialogOverlayFrame, {
  name: "AlertDialogOverlay"
}).styleable(function AlertDialogOverlay2(props, forwardedRef) {
  const {
    scope,
    ...overlayProps
  } = props;
  return /* @__PURE__ */jsx(DialogOverlay, {
    scope: getAlertDialogScope(scope),
    ...overlayProps,
    ref: forwardedRef
  });
});
const CONTENT_NAME = "AlertDialogContent";
const {
  Provider: AlertDialogContextProvider,
  useStyledContext: useAlertDialogContentContext
} = createStyledContext({}, "AlertDialogContext");
const AlertDialogContent = React.forwardRef(function AlertDialogContent2(props, forwardedRef) {
  const {
    scope,
    children,
    ...contentProps
  } = props;
  const dialogScope = getAlertDialogScope(scope);
  const contentRef = React.useRef(null);
  const composedRefs = useComposedRefs(forwardedRef, contentRef);
  const cancelRef = React.useRef(null);
  return /* @__PURE__ */jsx(DialogWarningProvider, {
    contentName: CONTENT_NAME,
    titleName: TITLE_NAME,
    docsSlug: "alert-dialog",
    children: /* @__PURE__ */jsx(AlertDialogContextProvider, {
      scope,
      cancelRef,
      destructiveRef: React.useRef(null),
      children: /* @__PURE__ */jsxs(DialogContent, {
        role: "alertdialog",
        "aria-modal": true,
        scope: dialogScope,
        ...contentProps,
        ref: composedRefs,
        onOpenAutoFocus: composeEventHandlers(contentProps.onOpenAutoFocus, event => {
          event.preventDefault();
          if (isWeb) cancelRef.current?.focus({
            preventScroll: true
          });
        }),
        onPointerDownOutside: event => event.preventDefault(),
        onInteractOutside: event => event.preventDefault(),
        children: [/* @__PURE__ */jsx(Slottable, {
          children
        }), process.env.NODE_ENV === "development" && /* @__PURE__ */jsx(DescriptionWarning, {
          contentRef
        })]
      })
    })
  });
});
const TITLE_NAME = "AlertDialogTitle";
const AlertDialogTitle = styled(View, {
  name: TITLE_NAME
}).styleable(function AlertDialogTitle2(props, forwardedRef) {
  const {
    scope,
    ...titleProps
  } = props;
  return /* @__PURE__ */jsx(DialogTitle, {
    scope: getAlertDialogScope(scope),
    ...titleProps,
    ref: forwardedRef
  });
});
const DESCRIPTION_NAME = "AlertDialogDescription";
const AlertDialogDescription = styled(View, {
  name: DESCRIPTION_NAME
}).styleable(function AlertDialogDescription2(props, forwardedRef) {
  const {
    scope,
    ...descriptionProps
  } = props;
  return /* @__PURE__ */jsx(DialogDescription, {
    scope: getAlertDialogScope(scope),
    ...descriptionProps,
    ref: forwardedRef
  });
});
const AlertDialogAction = styled(View, {
  name: "AlertDialogAction"
}).styleable(function AlertDialogAction2(props, forwardedRef) {
  const {
    scope,
    ...actionProps
  } = props;
  return /* @__PURE__ */jsx(DialogClose, {
    scope: getAlertDialogScope(scope),
    ...actionProps,
    ref: forwardedRef
  });
});
const AlertDialogCancel = styled(View, {
  name: "AlertDialogCancel"
}).styleable(function AlertDialogCancel2(props, forwardedRef) {
  const {
    scope,
    ...cancelProps
  } = props;
  const {
    cancelRef
  } = useAlertDialogContentContext(scope);
  const ref = useComposedRefs(forwardedRef, cancelRef);
  return /* @__PURE__ */jsx(DialogClose, {
    scope: getAlertDialogScope(scope),
    ...cancelProps,
    ref
  });
});
const AlertDialogDestructive = styled(View, {
  name: "AlertDialogDestructive"
}).styleable(function AlertDialogDestructive2(props, forwardedRef) {
  const {
    scope,
    ...destructiveProps
  } = props;
  const {
    destructiveRef
  } = useAlertDialogContentContext(scope);
  const ref = useComposedRefs(forwardedRef, destructiveRef);
  return /* @__PURE__ */jsx(DialogClose, {
    scope: getAlertDialogScope(scope),
    ...destructiveProps,
    ref
  });
});
const DescriptionWarning = ({
  contentRef
}) => {
  if (process.env.NODE_ENV === "development") React.useEffect(() => {
    if (!isWeb) return;
    if (!document.getElementById(
    // @ts-ignore
    contentRef.current?.getAttribute("aria-describedby"))) console.warn(`\`${CONTENT_NAME}\` requires a description for the component to be accessible for screen reader users.
  
        You can add a description to the \`${CONTENT_NAME}\` by passing a \`${DESCRIPTION_NAME}\` component as a child, which also benefits sighted users by adding visible context to the dialog.
        
        Alternatively, you can use your own component as a description by assigning it an \`id\` and passing the same value to the \`aria-describedby\` prop in \`${CONTENT_NAME}\`. If the description is confusing or duplicative for sighted users, you can use the \`@radix-ui/react-visually-hidden\` primitive as a wrapper around your description component.
        
        For more information, see https://tamagui.dev/docs/components/alert-dialog`);
  }, [contentRef]);
  return null;
};
const AlertDialogInner = props => {
  const {
    scope,
    native,
    ...alertDialogProps
  } = props;
  return /* @__PURE__ */jsx(Dialog, {
    scope: getAlertDialogScope(scope),
    ...alertDialogProps,
    modal: true
  });
};
function forEachChildDeep(children, onChild) {
  for (const child of children) {
    if (!React.isValidElement(child)) continue;
    if (!onChild(child)) continue;
    const childProps = child.props;
    if (childProps.children) forEachChildDeep(React.Children.toArray(childProps.children), onChild);
  }
}
const AlertDialog = withStaticProperties(AlertDialogInner, {
  Trigger: AlertDialogTrigger,
  Portal: AlertDialogPortal,
  Overlay: AlertDialogOverlay,
  Content: AlertDialogContent,
  Action: AlertDialogAction,
  Cancel: AlertDialogCancel,
  Destructive: AlertDialogDestructive,
  Title: AlertDialogTitle,
  Description: AlertDialogDescription
});
AlertDialog.displayName = ROOT_NAME;
export { AlertDialog, AlertDialogAction, AlertDialogCancel, AlertDialogContent, AlertDialogDescription, AlertDialogDestructive, AlertDialogOverlay, AlertDialogPortal, AlertDialogTitle, AlertDialogTrigger };
