// @tamagui/toast/dist/esm/ToastState.mjs
let toastsCounter = 1;
class Observer {
  subscribers = [];
  toasts = [];
  dismissedToasts = /* @__PURE__ */new Set();
  /**
   * Subscribe to toast state changes.
   * Returns an unsubscribe function.
   */
  subscribe = subscriber => {
    this.subscribers.push(subscriber);
    return () => {
      const index = this.subscribers.indexOf(subscriber);
      if (index > -1) {
        this.subscribers.splice(index, 1);
      }
    };
  };
  /**
   * Publish a toast to all subscribers
   */
  publish = data => {
    this.subscribers.forEach(subscriber => subscriber(data));
  };
  /**
   * Add a new toast to the internal array and publish to subscribers
   */
  addToast = data => {
    this.publish(data);
    this.toasts = [...this.toasts, data];
  };
  /**
   * Create or update a toast
   */
  create = data => {
    const {
      title,
      ...rest
    } = data;
    const id = typeof data?.id === "number" || typeof data?.id === "string" && data.id.length > 0 ? data.id : toastsCounter++;
    const alreadyExists = this.toasts.find(toast2 => toast2.id === id);
    const dismissible = data.dismissible ?? true;
    if (this.dismissedToasts.has(id)) {
      this.dismissedToasts.delete(id);
    }
    if (alreadyExists) {
      this.toasts = this.toasts.map(toast2 => {
        if (toast2.id === id) {
          this.publish({
            ...toast2,
            ...data,
            id,
            title,
            dismissible
          });
          return {
            ...toast2,
            ...data,
            id,
            title,
            dismissible
          };
        }
        return toast2;
      });
    } else {
      this.addToast({
        title,
        ...rest,
        dismissible,
        id
      });
    }
    return id;
  };
  /**
   * Dismiss a toast by id, or all toasts if no id provided
   */
  dismiss = id => {
    if (id !== void 0) {
      this.dismissedToasts.add(id);
      requestAnimationFrame(() => {
        this.subscribers.forEach(subscriber => subscriber({
          id,
          dismiss: true
        }));
      });
    } else {
      this.toasts.forEach(toast2 => {
        this.subscribers.forEach(subscriber => subscriber({
          id: toast2.id,
          dismiss: true
        }));
      });
    }
    return id;
  };
  /**
   * Show a basic toast message
   */
  message = (title, data) => {
    return this.create({
      ...data,
      title,
      type: "default"
    });
  };
  /**
   * Show a success toast
   */
  success = (title, data) => {
    return this.create({
      ...data,
      title,
      type: "success"
    });
  };
  /**
   * Show an error toast
   */
  error = (title, data) => {
    return this.create({
      ...data,
      title,
      type: "error"
    });
  };
  /**
   * Show a warning toast
   */
  warning = (title, data) => {
    return this.create({
      ...data,
      title,
      type: "warning"
    });
  };
  /**
   * Show an info toast
   */
  info = (title, data) => {
    return this.create({
      ...data,
      title,
      type: "info"
    });
  };
  /**
   * Show a loading toast
   */
  loading = (title, data) => {
    return this.create({
      ...data,
      title,
      type: "loading"
    });
  };
  /**
   * Show a toast for a promise, automatically transitioning through
   * loading -> success/error states
   */
  promise = (promise, data) => {
    if (!data) {
      return;
    }
    let id = void 0;
    if (data.loading !== void 0) {
      id = this.create({
        promise,
        type: "loading",
        title: data.loading,
        description: typeof data.description !== "function" ? data.description : void 0,
        // loading toasts shouldn't auto-dismiss
        duration: Number.POSITIVE_INFINITY
      });
    }
    const p = Promise.resolve(promise instanceof Function ? promise() : promise);
    let shouldDismiss = id !== void 0;
    let result;
    const originalPromise = p.then(async response => {
      result = ["resolve", response];
      if (isHttpResponse(response) && !response.ok) {
        shouldDismiss = false;
        const errorMsg = typeof data.error === "function" ? await data.error(`HTTP error! status: ${response.status}`) : data.error;
        const description = typeof data.description === "function" ? await data.description(`HTTP error! status: ${response.status}`) : data.description;
        this.create({
          id,
          type: "error",
          title: errorMsg,
          description
        });
      } else if (data.success !== void 0) {
        shouldDismiss = false;
        const successMsg = typeof data.success === "function" ? await data.success(response) : data.success;
        const description = typeof data.description === "function" ? await data.description(response) : data.description;
        this.create({
          id,
          type: "success",
          title: successMsg,
          description
        });
      }
    }).catch(async error => {
      result = ["reject", error];
      if (data.error !== void 0) {
        shouldDismiss = false;
        const errorMsg = typeof data.error === "function" ? await data.error(error) : data.error;
        const description = typeof data.description === "function" ? await data.description(error) : data.description;
        this.create({
          id,
          type: "error",
          title: errorMsg,
          description
        });
      }
    }).finally(() => {
      if (shouldDismiss) {
        this.dismiss(id);
        id = void 0;
      }
      data.finally?.();
    });
    const unwrap = () => new Promise((resolve, reject) => originalPromise.then(() => result[0] === "reject" ? reject(result[1]) : resolve(result[1])).catch(reject));
    if (typeof id !== "string" && typeof id !== "number") {
      return {
        unwrap
      };
    } else {
      return Object.assign(id, {
        unwrap
      });
    }
  };
  /**
   * Show a custom JSX toast
   */
  custom = (jsx, data) => {
    const id = data?.id ?? toastsCounter++;
    this.create({
      jsx: jsx(id),
      ...data,
      id
    });
    return id;
  };
  /**
   * Get all active (non-dismissed) toasts
   */
  getActiveToasts = () => {
    return this.toasts.filter(toast2 => !this.dismissedToasts.has(toast2.id));
  };
  /**
   * Get full toast history
   */
  getHistory = () => {
    return this.toasts;
  };
}
function isHttpResponse(data) {
  return data && typeof data === "object" && "ok" in data && typeof data.ok === "boolean" && "status" in data && typeof data.status === "number";
}
const ToastState = new Observer();
const toastFunction = (title, data) => {
  return ToastState.create({
    ...data,
    title
  });
};
const getHistory = () => ToastState.getHistory();
const getToasts = () => ToastState.getActiveToasts();
const toast = Object.assign(toastFunction, {
  success: ToastState.success,
  error: ToastState.error,
  warning: ToastState.warning,
  info: ToastState.info,
  loading: ToastState.loading,
  promise: ToastState.promise,
  custom: ToastState.custom,
  dismiss: ToastState.dismiss,
  message: ToastState.message,
  getHistory,
  getToasts
});
export { ToastState, toast };
