// @tamagui/toast/dist/esm/Toast.mjs
import { AnimatePresence } from "@tamagui/animate-presence";
import { styled, useEvent } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { YStack } from "@tamagui/stacks";
import { SizableText } from "@tamagui/text";
import { useControllableState } from "@tamagui/use-controllable-state";
import * as React from "react";
import { ToastAnnounceExclude } from "./ToastAnnounce.mjs";
import { useToast, useToastController, useToastState } from "./ToastImperative.mjs";
import { ToastImpl, ToastImplFrame, useToastInteractiveContext } from "./ToastImpl.mjs";
import { ToastProvider } from "./ToastProvider.mjs";
import { ToastViewport } from "./ToastViewport.mjs";
import { jsx } from "react/jsx-runtime";
const TITLE_NAME = "ToastTitle";
const ToastTitle = styled(SizableText, {
  name: TITLE_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color",
        size: "$4"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const DESCRIPTION_NAME = "ToastDescription";
const ToastDescription = styled(SizableText, {
  name: DESCRIPTION_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color11",
        size: "$1"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const ACTION_NAME = "ToastAction";
const ToastAction = React.forwardRef(function ToastAction2(props, forwardedRef) {
  const {
    altText,
    ...actionProps
  } = props;
  if (!altText) return null;
  return /* @__PURE__ */jsx(ToastAnnounceExclude, {
    altText,
    asChild: true,
    children: /* @__PURE__ */jsx(ToastClose, {
      ...actionProps,
      ref: forwardedRef
    })
  });
});
ToastAction.propTypes = {
  altText(props) {
    if (!props.altText) {
      return new Error(`Missing prop \`altText\` expected on \`${ACTION_NAME}\``);
    }
    return null;
  }
};
const CLOSE_NAME = "ToastClose";
const ToastCloseFrame = styled(YStack, {
  name: CLOSE_NAME,
  render: "button"
});
const ToastClose = React.forwardRef(function ToastClose2(props, forwardedRef) {
  const {
    scope,
    ...closeProps
  } = props;
  const interactiveContext = useToastInteractiveContext(scope);
  return /* @__PURE__ */jsx(ToastAnnounceExclude, {
    asChild: true,
    children: /* @__PURE__ */jsx(ToastCloseFrame, {
      "aria-label": "Close",
      ...closeProps,
      ref: forwardedRef,
      onPress: composeEventHandlers(props.onPress, interactiveContext.onClose)
    })
  });
});
const ToastComponent = ToastImplFrame.styleable(function Toast(props, forwardedRef) {
  const {
    forceMount,
    open: openProp,
    defaultOpen,
    onOpenChange,
    ...toastProps
  } = props;
  const [open, setOpen] = useControllableState({
    prop: openProp,
    defaultProp: defaultOpen ?? true,
    onChange: onOpenChange,
    strategy: "most-recent-wins"
  });
  const currentToast = useToastState();
  const {
    hide
  } = useToastController();
  const id = React.useId();
  const onPause = useEvent(props.onPause);
  const onResume = useEvent(props.onResume);
  const isHide = currentToast?.hide === true;
  const shouldShow = (forceMount || open) && !isHide;
  return /* @__PURE__ */jsx(AnimatePresence, {
    children: shouldShow ? /* @__PURE__ */jsx(ToastImpl, {
      id,
      open,
      ...toastProps,
      ref: forwardedRef,
      onClose: () => {
        setOpen(false);
        hide();
      },
      onPause,
      onResume,
      onSwipeEnd: composeEventHandlers(props.onSwipeEnd, event => {
        setOpen(false);
      })
    }) : null
  }, id);
});
const Toast2 = withStaticProperties(ToastComponent, {
  Title: ToastTitle,
  Description: ToastDescription,
  Action: ToastAction,
  Close: ToastClose
});
export { Toast2 as Toast, ToastProvider, ToastViewport, useToast, useToastController, useToastState };
