// @tamagui/tabs/dist/jsx/createTabs.mjs
import { composeRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { Group, useGroupItem } from "@tamagui/group";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { RovingFocusGroup } from "@tamagui/roving-focus";
import { SizableContext } from "@tamagui/sizable-context";
import { useControllableState } from "@tamagui/use-controllable-state";
import { useDirection } from "@tamagui/use-direction";
import { useEvent } from "@tamagui/web";
import * as React from "react";
import { TabsProvider, useTabsContext } from "./StyledContext.mjs";
import { DefaultTabsContentFrame, DefaultTabsFrame, DefaultTabsTabFrame } from "./Tabs.mjs";
import { jsx } from "react/jsx-runtime";
function createTabs(createProps) {
  const {
    ContentFrame = DefaultTabsContentFrame,
    TabFrame = DefaultTabsTabFrame,
    TabsFrame = DefaultTabsFrame
  } = createProps;
  const TABS_CONTEXT = "TabsContext";
  const TAB_LIST_NAME = "TabsList";
  const TabsList = React.forwardRef((props, forwardedRef) => {
    const {
      __scopeTabs,
      loop = true,
      children,
      ...listProps
    } = props;
    const context = useTabsContext(__scopeTabs);
    return /* @__PURE__ */jsx(RovingFocusGroup, {
      __scopeRovingFocusGroup: __scopeTabs || TABS_CONTEXT,
      orientation: context.orientation,
      dir: context.dir,
      loop,
      asChild: true,
      children: /* @__PURE__ */jsx(Group, {
        role: "tablist",
        componentName: TAB_LIST_NAME,
        "aria-orientation": context.orientation,
        ref: forwardedRef,
        orientation: context.orientation,
        ...listProps,
        children
      })
    });
  });
  TabsList.displayName = TAB_LIST_NAME;
  const TRIGGER_NAME = "TabsTrigger";
  const TabsTrigger = TabFrame.styleable((props, forwardedRef) => {
    const {
      __scopeTabs,
      value,
      disabled = false,
      onInteraction,
      activeStyle,
      activeTheme,
      unstyled = false,
      ...triggerProps
    } = props;
    const context = useTabsContext(__scopeTabs);
    const triggerId = makeTriggerId(context.baseId, value);
    const contentId = makeContentId(context.baseId, value);
    const isSelected = value === context.value;
    const [layout, setLayout] = React.useState(null);
    const triggerRef = React.useRef(null);
    const groupItemProps = useGroupItem({
      disabled: !!disabled
    });
    React.useEffect(() => {
      context.registerTrigger();
      return () => context.unregisterTrigger();
    }, []);
    React.useEffect(() => {
      if (!triggerRef.current || !isWeb) return;
      const el = triggerRef.current;
      function getTriggerSize() {
        if (!el) return;
        setLayout({
          width: el.offsetWidth,
          height: el.offsetHeight,
          x: el.offsetLeft,
          y: el.offsetTop
        });
      }
      getTriggerSize();
      const observer = new ResizeObserver(getTriggerSize);
      observer.observe(el);
      return () => {
        observer.disconnect();
      };
    }, [context.triggersCount]);
    React.useEffect(() => {
      if (isSelected && layout) {
        onInteraction?.("select", layout);
      }
    }, [isSelected, value, layout]);
    return /* @__PURE__ */jsx(RovingFocusGroup.Item, {
      __scopeRovingFocusGroup: __scopeTabs || TABS_CONTEXT,
      asChild: true,
      focusable: !disabled,
      active: isSelected,
      children: /* @__PURE__ */jsx(TabFrame, {
        ...(!isWeb && {
          onLayout: event => {
            setLayout(event.nativeEvent.layout);
          }
        }),
        onMouseEnter: composeEventHandlers(props.onMouseEnter, () => {
          if (layout) {
            onInteraction?.("hover", layout);
          }
        }),
        onMouseLeave: composeEventHandlers(props.onMouseLeave, () => {
          onInteraction?.("hover", null);
        }),
        role: "tab",
        "aria-selected": isSelected,
        "aria-controls": contentId,
        "data-state": isSelected ? "active" : "inactive",
        "data-disabled": disabled ? "" : void 0,
        id: triggerId,
        theme: activeTheme ?? null,
        unstyled,
        ...(!unstyled && {
          size: context.size
        }),
        ...(isSelected && {
          ...(!unstyled && !activeStyle && {
            backgroundColor: "$backgroundActive"
          }),
          ...activeStyle
        }),
        ...groupItemProps,
        disabled: disabled ?? groupItemProps.disabled,
        ...triggerProps,
        ref: composeRefs(forwardedRef, triggerRef),
        onPress: composeEventHandlers(props.onPress ?? void 0, event => {
          const webChecks = !isWeb || event.button === 0 && event.ctrlKey === false;
          if (!disabled && !isSelected && webChecks) {
            context.onChange(value);
          }
        }),
        ...(isWeb && {
          onKeyDown: composeEventHandlers(props.onKeyDown, event => {
            if ([" ", "Enter"].includes(event.key)) {
              context.onChange(value);
              event.preventDefault();
            }
          }),
          onFocus: composeEventHandlers(props.onFocus, event => {
            if (layout) {
              onInteraction?.("focus", layout);
            }
            const isAutomaticActivation = context.activationMode !== "manual";
            if (!isSelected && !disabled && isAutomaticActivation) {
              context.onChange(value);
            }
          }),
          onBlur: composeEventHandlers(props.onBlur, () => {
            onInteraction?.("focus", null);
          })
        })
      })
    });
  });
  TabsTrigger.displayName = TRIGGER_NAME;
  const TabsContent = ContentFrame.styleable(function TabsContent2(props, forwardedRef) {
    const {
      __scopeTabs,
      value,
      forceMount,
      children,
      ...contentProps
    } = props;
    const context = useTabsContext(__scopeTabs);
    const isSelected = value === context.value;
    const show = forceMount || isSelected;
    const triggerId = makeTriggerId(context.baseId, value);
    const contentId = makeContentId(context.baseId, value);
    if (!show) {
      return null;
    }
    return /* @__PURE__ */jsx(ContentFrame, {
      "data-state": isSelected ? "active" : "inactive",
      "data-orientation": context.orientation,
      role: "tabpanel",
      "aria-labelledby": triggerId,
      hidden: !show,
      id: contentId,
      tabIndex: 0,
      ...contentProps,
      ref: forwardedRef,
      children
    }, value);
  });
  const TabsComponent = TabsFrame.styleable(function Tabs(props, forwardedRef) {
    const {
      __scopeTabs,
      value: valueProp,
      onValueChange,
      defaultValue,
      orientation = "horizontal",
      dir,
      activationMode = "manual",
      size = "$true",
      ...tabsProps
    } = props;
    const direction = useDirection(dir);
    const [value, setValue] = useControllableState({
      prop: valueProp,
      onChange: onValueChange,
      defaultProp: defaultValue ?? ""
    });
    const [triggersCount, setTriggersCount] = React.useState(0);
    const registerTrigger = useEvent(() => setTriggersCount(v => v + 1));
    const unregisterTrigger = useEvent(() => setTriggersCount(v => v - 1));
    return /* @__PURE__ */jsx(SizableContext.Provider, {
      size,
      children: /* @__PURE__ */jsx(TabsProvider, {
        scope: __scopeTabs,
        baseId: React.useId(),
        value,
        onChange: setValue,
        orientation,
        dir: direction,
        activationMode,
        size,
        registerTrigger,
        triggersCount,
        unregisterTrigger,
        children: /* @__PURE__ */jsx(TabsFrame, {
          direction,
          "data-orientation": orientation,
          ...tabsProps,
          ref: forwardedRef
        })
      })
    });
  });
  return withStaticProperties(TabsComponent, {
    List: TabsList,
    /**
     * @deprecated Use Tabs.Tab instead
     */
    Trigger: TabsTrigger,
    Tab: TabsTrigger,
    Content: TabsContent
  });
}
function makeTriggerId(baseId, value) {
  return `${baseId}-trigger-${value}`;
}
function makeContentId(baseId, value) {
  return `${baseId}-content-${value}`;
}
export { createTabs };
