// @tamagui/form/dist/esm/Form.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { View, createStyledContext, styled } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
var FORM_NAME = "Form";
var FormFrame = styled(View, {
  name: FORM_NAME,
  render: "form"
});
var FormContext = createStyledContext({
  onSubmit: void 0
});
var {
  useStyledContext: useFormContext,
  Provider: FormProvider
} = FormContext;
var FormTriggerFrame = styled(View, {
  name: "FormTrigger"
});
var FormTrigger = FormTriggerFrame.styleable(function (props, forwardedRef) {
  var {
    scope,
    children,
    onPress,
    ...triggerProps
  } = props;
  var context = useFormContext(scope);
  return /* @__PURE__ */_jsx(FormTriggerFrame, {
    role: "button",
    ...triggerProps,
    ref: forwardedRef,
    onPress: composeEventHandlers(onPress, context.onSubmit),
    children
  });
});
var FormComponent = FormFrame.styleable(function Form(param, ref) {
  var {
    scope,
    onSubmit,
    ...props
  } = param;
  return /* @__PURE__ */_jsx(FormProvider, {
    scope,
    onSubmit,
    children: /* @__PURE__ */_jsx(FormFrame, {
      ref,
      ...props,
      onSubmit: function (e) {
        e.preventDefault();
        onSubmit === null || onSubmit === void 0 ? void 0 : onSubmit();
      }
    })
  });
});
var Form2 = withStaticProperties(FormComponent, {
  Trigger: FormTrigger
});
export { Form2 as Form, FormContext, FormFrame, FormProvider, FormTrigger, useFormContext };
