// @tamagui/web/dist/cjs/views/TamaguiRoot.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var TamaguiRoot_exports = {};
__export(TamaguiRoot_exports, {
  TamaguiRoot: () => TamaguiRoot
});
module.exports = __toCommonJS(TamaguiRoot_exports);
var import_react = __toESM(require("react"), 1);
var import_config = require("../config.cjs");
var import_Theme = require("./Theme.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
let defaultFontClass = "";
function TamaguiRoot({
  children,
  theme,
  isRootRoot,
  passThrough,
  style
}) {
  const [mounted, setMounted] = import_react.default.useState(!isRootRoot);
  import_react.default.useEffect(() => {
    if (!mounted) {
      setMounted(true);
    }
  }, []);
  if (!defaultFontClass) {
    const config = (0, import_config.getConfig)();
    const defaultFont = config.defaultFont;
    if (defaultFont) {
      defaultFontClass = `font_${defaultFont}`;
    }
  }
  const contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)("span", {
    style,
    className: `_dsp_contents ${mounted ? "" : "t_unmounted"} ${defaultFontClass}`,
    children
  });
  if (isRootRoot) {
    return contents;
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_Theme.Theme, {
    passThrough,
    contain: true,
    forceClassName: true,
    name: theme,
    children: contents
  });
}
