// @tamagui/web/dist/cjs/views/TamaguiProvider.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var TamaguiProvider_exports = {};
__export(TamaguiProvider_exports, {
  TamaguiProvider: () => TamaguiProvider
});
module.exports = __toCommonJS(TamaguiProvider_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("@tamagui/constants");
var import_use_did_finish_ssr = require("@tamagui/use-did-finish-ssr");
var import_react = __toESM(require("react"), 1);
var import_config = require("../config.native.js");
var import_ComponentContext = require("../contexts/ComponentContext.native.js");
var import_insertStyleRule = require("../helpers/insertStyleRule.native.js");
var import_useMedia = require("../hooks/useMedia.native.js");
var import_resolveAnimationDriver = require("../helpers/resolveAnimationDriver.native.js");
var import_TamaguiRoot = require("./TamaguiRoot.native.js");
var import_ThemeProvider = require("./ThemeProvider.native.js");
var _cachedFirstKey;
var _cachedConfig;
function firstThemeKey(config) {
  if (config !== _cachedConfig) {
    _cachedConfig = config;
    _cachedFirstKey = (config === null || config === void 0 ? void 0 : config.themes) ? Object.keys(config.themes)[0] : void 0;
  }
  return _cachedFirstKey;
}
function TamaguiProvider(param) {
  var {
    children,
    disableInjectCSS,
    config,
    className,
    defaultTheme: defaultThemeProp,
    reset,
    insets
  } = param;
  var defaultTheme = defaultThemeProp || firstThemeKey(config) || "light";
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    (0, import_insertStyleRule.stopAccumulatingRules)();
    (0, import_useMedia.updateMediaListeners)();
  }, []);
  var memoizedInsets = import_react.default.useMemo(function () {
    return insets;
  }, [insets === null || insets === void 0 ? void 0 : insets.top, insets === null || insets === void 0 ? void 0 : insets.right, insets === null || insets === void 0 ? void 0 : insets.bottom, insets === null || insets === void 0 ? void 0 : insets.left]);
  var defaultAnimationDriver = import_react.default.useMemo(function () {
    return (0, import_resolveAnimationDriver.resolveAnimationDriver)(config === null || config === void 0 ? void 0 : config.animations);
  }, [config === null || config === void 0 ? void 0 : config.animations]);
  (0, import_react.useEffect)(function () {
    var _defaultAnimationDriver_onMount;
    defaultAnimationDriver === null || defaultAnimationDriver === void 0 || (_defaultAnimationDriver_onMount = defaultAnimationDriver.onMount) == null || _defaultAnimationDriver_onMount.call(defaultAnimationDriver);
  }, []);
  var contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ComponentContext.ComponentContext.Provider, {
    animationDriver: defaultAnimationDriver,
    insets: memoizedInsets,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ThemeProvider.ThemeProvider, {
      defaultTheme,
      reset,
      className,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_TamaguiRoot.TamaguiRoot, {
        theme: defaultTheme,
        isRootRoot: true,
        children
      })
    })
  });
  if ((0, import_config.getSetting)("disableSSR")) contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_use_did_finish_ssr.ClientOnly, {
    enabled: true,
    children: contents
  });
  return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
    children: [contents, false]
  });
}
