// @tamagui/web/dist/cjs/views/Theme.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Theme_exports = {};
__export(Theme_exports, {
  Theme: () => Theme,
  getThemedChildren: () => getThemedChildren
});
module.exports = __toCommonJS(Theme_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("@tamagui/constants");
var import_react = __toESM(require("react"), 1);
var import_createVariable = require("../createVariable.native.js");
var import_useTheme = require("../hooks/useTheme.native.js");
var import_useThemeState = require("../hooks/useThemeState.native.js");
var import_ThemeDebug = require("./ThemeDebug.native.js");
var Theme = /* @__PURE__ */(0, import_react.forwardRef)(function Theme2(props, ref) {
  "use no memo";

  if (props.disable) {
    return props.children;
  }
  var {
    passThrough
  } = props;
  var isRoot = !!props["_isRoot"];
  var [_, themeState] = (0, import_useTheme.useThemeWithState)(props, isRoot);
  var disableDirectChildTheme = props["disable-child-theme"];
  var finalChildren = disableDirectChildTheme ? import_react.Children.map(props.children, function (child) {
    return passThrough || ! /* @__PURE__ */(0, import_react.isValidElement)(child) ? child : /* @__PURE__ */(0, import_react.cloneElement)(child, {
      ["data-disable-theme"]: true
    });
  }) : props.children;
  if (ref) {
    try {
      import_react.default.Children.only(finalChildren);
      finalChildren = /* @__PURE__ */(0, import_react.cloneElement)(finalChildren, {
        ref
      });
    } catch (e) {}
  }
  var stateRef = (0, import_react.useRef)({
    hasEverThemed: false
  });
  return getThemedChildren(themeState, finalChildren, props, isRoot, stateRef, passThrough);
});
Theme["avoidForwardRef"] = true;
function getThemedChildren(themeState, children, props) {
  var isRoot = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : false,
    stateRef = arguments.length > 4 ? arguments[4] : void 0,
    passThrough = arguments.length > 5 && arguments[5] !== void 0 ? arguments[5] : false;
  var {
    shallow,
    forceClassName
  } = props;
  var state = stateRef.current;
  var hasEverThemed = state.hasEverThemed;
  var shouldRenderChildrenWithTheme = hasEverThemed || themeState.isNew || isRoot || (0, import_useThemeState.hasThemeUpdatingProps)(props);
  if (process.env.NODE_ENV === "development" && props.debug === "visualize") {
    children = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ThemeDebug.ThemeDebug, {
      themeState,
      themeProps: props,
      children
    });
  }
  if (!shouldRenderChildrenWithTheme) {
    return children;
  }
  children = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_useThemeState.ThemeStateContext.Provider, {
    value: themeState.id,
    children
  });
  var {
    isInverse,
    name
  } = themeState;
  var requiresExtraWrapper = isInverse || forceClassName;
  if (!state.hasEverThemed) {
    state.hasEverThemed = true;
  }
  if (requiresExtraWrapper ||
  // if the theme is exactly dark or light, its likely to change between dark/light
  // and that would require wrapping which would re-parent, so to avoid re-parenting do this
  themeState.name === "dark" || themeState.name === "light") {
    state.hasEverThemed = "wrapped";
  }
  if (shallow) {
    if (!themeState.parentId) {} else {
      var parentState = (0, import_useThemeState.getThemeState)(themeState.isNew ? themeState.id : themeState.parentId);
      if (!parentState) throw new Error(`\u203C\uFE0F010`);
      children = import_react.Children.toArray(children).map(function (child) {
        return /* @__PURE__ */(0, import_react.isValidElement)(child) ? passThrough ? child : /* @__PURE__ */(0, import_react.cloneElement)(child, void 0, /* @__PURE__ */(0, import_jsx_runtime.jsx)(Theme, {
          name: parentState.name,
          children: child.props.children
        })) : child;
      });
    }
  }
  if (process.env.NODE_ENV === "development") {
    if (!passThrough && props.debug) {
      console.warn(` getThemedChildren`, {
        requiresExtraWrapper,
        forceClassName,
        themeState,
        state,
        themeSpanProps: getThemeClassNameAndColor(themeState, props, isRoot)
      });
    }
  }
  if (forceClassName === false) {
    return children;
  }
  if (import_constants.isWeb) {
    var baseStyle = props.contain ? inertContainedStyle : inertStyle;
    var {
      className = "",
      color
    } = passThrough ? {} : getThemeClassNameAndColor(themeState, props, isRoot);
    children = /* @__PURE__ */(0, import_jsx_runtime.jsx)("span", {
      className: `${className} is_Theme`,
      style: passThrough ? baseStyle : {
        color,
        ...baseStyle
      },
      children
    });
    if (state.hasEverThemed === "wrapped") {
      var className1 = requiresExtraWrapper ? `${name.startsWith("light") ? "t_light" : name.startsWith("dark") ? "t_dark" : ""} _dsp_contents` : `_dsp_contents`;
      children = /* @__PURE__ */(0, import_jsx_runtime.jsx)("span", {
        className: className1,
        children
      });
    }
    return children;
  }
  return children;
}
var inertStyle = {
  display: "contents"
};
var inertContainedStyle = {
  display: "contents",
  contain: "strict"
};
var empty = {
  className: "",
  color: void 0
};
function getThemeClassNameAndColor(themeState, props) {
  var isRoot = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : false;
  if (!themeState.isNew && !props.forceClassName) {
    return empty;
  }
  var themeColor = (themeState === null || themeState === void 0 ? void 0 : themeState.theme) && themeState.isNew ? (0, import_createVariable.variableToString)(themeState.theme.color) : "";
  var style = themeColor ? {
    color: themeColor
  } : void 0;
  var themeClassName = themeState.name.replace(schemePrefix, "");
  var themeNameParts = themeClassName.split("_");
  var themeClasses = `t_${themeClassName}`;
  if (themeNameParts.length > 1) {
    var hierarchyClasses = [];
    for (var i = 1; i <= themeNameParts.length; i++) {
      hierarchyClasses.push(`t_${themeNameParts.slice(0, i).join("_")}`);
    }
    themeClasses = hierarchyClasses.join(" ");
  }
  var className = `${isRoot ? "" : "t_sub_theme"} ${themeClasses}`;
  return {
    color: themeColor,
    className
  };
}
var schemePrefix = /^(dark|light)_/;
