// @tamagui/use-window-dimensions/dist/esm/helpers.mjs
import { isClient } from "@tamagui/constants";
import { initialValue } from "./initialValue.mjs";
let lastSize = initialValue;
let docEl = null;
function getWindowSize() {
  docEl ||= window.document.documentElement;
  const nextSize = {
    fontScale: 1,
    height: docEl.clientHeight,
    scale: window.devicePixelRatio || 1,
    width: docEl.clientWidth
  };
  if (nextSize.height !== lastSize.height || nextSize.width !== lastSize.width || nextSize.scale !== lastSize.scale) {
    lastSize = nextSize;
    return nextSize;
  }
  return lastSize;
}
const cbs = /* @__PURE__ */new Set();
if (isClient && typeof window !== "undefined" && typeof window.addEventListener === "function") {
  let flushUpdate = function () {
    lastUpdate = Date.now();
    cbs.forEach(cb => cb(getWindowSize()));
  };
  let lastUpdate = Date.now();
  let tm;
  const USER_MAX_MS = process.env.TAMAGUI_USE_WINDOW_DIMENSIONS_MAX_UPDATE_MS;
  const updateMaxMs = USER_MAX_MS ? +USER_MAX_MS : 100;
  const onResize = () => {
    clearTimeout(tm);
    const timeSinceLast = Date.now() - lastUpdate;
    if (timeSinceLast < updateMaxMs) tm = setTimeout(() => {
      flushUpdate();
    }, updateMaxMs - timeSinceLast);else flushUpdate();
  };
  window.addEventListener("resize", onResize);
}
function subscribe(cb) {
  cbs.add(cb);
  return () => cbs.delete(cb);
}
export { getWindowSize, subscribe };
