// @tamagui/native/dist/esm/gestureState.native.js
import { createGlobalState } from "./globalState.native.js";
var state = createGlobalState(`gesture`, {
  enabled: false,
  Gesture: null,
  GestureDetector: null,
  ScrollView: null
});
function getGestureHandler() {
  return {
    get isEnabled() {
      return state.get().enabled;
    },
    get state() {
      return state.get();
    },
    set(updates) {
      Object.assign(state.get(), updates);
    },
    disable() {
      state.get().enabled = false;
    },
    createPressGesture(config) {
      var {
        Gesture
      } = state.get();
      if (!Gesture) return null;
      var _config_delayLongPress;
      var longPressDuration = (_config_delayLongPress = config.delayLongPress) !== null && _config_delayLongPress !== void 0 ? _config_delayLongPress : 500;
      var tap = Gesture.Tap().runOnJS(true).maxDuration(1e4).onBegin(function (e) {
        var _config_onPressIn;
        (_config_onPressIn = config.onPressIn) === null || _config_onPressIn === void 0 ? void 0 : _config_onPressIn.call(config, e);
      }).onEnd(function (e) {
        var _config_onPress;
        (_config_onPress = config.onPress) === null || _config_onPress === void 0 ? void 0 : _config_onPress.call(config, e);
      }).onFinalize(function (e) {
        var _config_onPressOut;
        (_config_onPressOut = config.onPressOut) === null || _config_onPressOut === void 0 ? void 0 : _config_onPressOut.call(config, e);
      });
      if (config.hitSlop) tap.hitSlop(config.hitSlop);
      if (!config.onLongPress) return tap;
      var longPress = Gesture.LongPress().runOnJS(true).minDuration(longPressDuration).onBegin(function (e) {
        var _config_onPressIn;
        return (_config_onPressIn = config.onPressIn) === null || _config_onPressIn === void 0 ? void 0 : _config_onPressIn.call(config, e);
      }).onStart(function (e) {
        var _config_onLongPress;
        return (_config_onLongPress = config.onLongPress) === null || _config_onLongPress === void 0 ? void 0 : _config_onLongPress.call(config, e);
      }).onFinalize(function (e) {
        var _config_onPressOut;
        return (_config_onPressOut = config.onPressOut) === null || _config_onPressOut === void 0 ? void 0 : _config_onPressOut.call(config, e);
      });
      if (config.hitSlop) longPress.hitSlop(config.hitSlop);
      return Gesture.Exclusive(longPress, tap);
    }
  };
}
export { getGestureHandler };
