// @tamagui/native/dist/esm/gestureState.mjs
import { createGlobalState } from "./globalState.mjs";
const state = createGlobalState(`gesture`, {
  enabled: false,
  Gesture: null,
  GestureDetector: null,
  ScrollView: null
});
function getGestureHandler() {
  return {
    get isEnabled() {
      return state.get().enabled;
    },
    get state() {
      return state.get();
    },
    set(updates) {
      Object.assign(state.get(), updates);
    },
    disable() {
      state.get().enabled = false;
    },
    createPressGesture(config) {
      const {
        Gesture
      } = state.get();
      if (!Gesture) return null;
      const longPressDuration = config.delayLongPress ?? 500;
      const tap = Gesture.Tap().runOnJS(true).maxDuration(1e4).onBegin(e => {
        config.onPressIn?.(e);
      }).onEnd(e => {
        config.onPress?.(e);
      }).onFinalize(e => {
        config.onPressOut?.(e);
      });
      if (config.hitSlop) tap.hitSlop(config.hitSlop);
      if (!config.onLongPress) return tap;
      const longPress = Gesture.LongPress().runOnJS(true).minDuration(longPressDuration).onBegin(e => config.onPressIn?.(e)).onStart(e => config.onLongPress?.(e)).onFinalize(e => config.onPressOut?.(e));
      if (config.hitSlop) longPress.hitSlop(config.hitSlop);
      return Gesture.Exclusive(longPress, tap);
    }
  };
}
export { getGestureHandler };
