// @tamagui/tabs/dist/esm/createTabs.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { composeRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { Group, useGroupItem } from "@tamagui/group";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { RovingFocusGroup } from "@tamagui/roving-focus";
import { SizableContext } from "@tamagui/sizable-context";
import { useControllableState } from "@tamagui/use-controllable-state";
import { useDirection } from "@tamagui/use-direction";
import { useEvent } from "@tamagui/web";
import * as React from "react";
import { TabsProvider, useTabsContext } from "./StyledContext.native.js";
import { DefaultTabsContentFrame, DefaultTabsFrame, DefaultTabsTabFrame } from "./Tabs.native.js";
function createTabs(createProps) {
  var {
    ContentFrame = DefaultTabsContentFrame,
    TabFrame = DefaultTabsTabFrame,
    TabsFrame = DefaultTabsFrame
  } = createProps;
  var TABS_CONTEXT = "TabsContext";
  var TAB_LIST_NAME = "TabsList";
  var TabsList = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
    var {
      __scopeTabs,
      loop = true,
      children,
      ...listProps
    } = props;
    var context = useTabsContext(__scopeTabs);
    return /* @__PURE__ */_jsx(RovingFocusGroup, {
      __scopeRovingFocusGroup: __scopeTabs || TABS_CONTEXT,
      orientation: context.orientation,
      dir: context.dir,
      loop,
      asChild: true,
      children: /* @__PURE__ */_jsx(Group, {
        role: "tablist",
        componentName: TAB_LIST_NAME,
        "aria-orientation": context.orientation,
        ref: forwardedRef,
        orientation: context.orientation,
        ...listProps,
        children
      })
    });
  });
  TabsList.displayName = TAB_LIST_NAME;
  var TRIGGER_NAME = "TabsTrigger";
  var TabsTrigger = TabFrame.styleable(function (props, forwardedRef) {
    var {
      __scopeTabs,
      value,
      disabled = false,
      onInteraction,
      activeStyle,
      activeTheme,
      unstyled = false,
      ...triggerProps
    } = props;
    var context = useTabsContext(__scopeTabs);
    var triggerId = makeTriggerId(context.baseId, value);
    var contentId = makeContentId(context.baseId, value);
    var isSelected = value === context.value;
    var [layout, setLayout] = React.useState(null);
    var triggerRef = React.useRef(null);
    var groupItemProps = useGroupItem({
      disabled: !!disabled
    });
    React.useEffect(function () {
      context.registerTrigger();
      return function () {
        return context.unregisterTrigger();
      };
    }, []);
    React.useEffect(function () {
      if (!triggerRef.current || !isWeb) return;
      var el = triggerRef.current;
      function getTriggerSize() {
        if (!el) return;
        setLayout({
          width: el.offsetWidth,
          height: el.offsetHeight,
          x: el.offsetLeft,
          y: el.offsetTop
        });
      }
      getTriggerSize();
      var observer = new ResizeObserver(getTriggerSize);
      observer.observe(el);
      return function () {
        observer.disconnect();
      };
    }, [context.triggersCount]);
    React.useEffect(function () {
      if (isSelected && layout) {
        onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("select", layout);
      }
    }, [isSelected, value, layout]);
    var _props_onPress;
    return /* @__PURE__ */_jsx(RovingFocusGroup.Item, {
      __scopeRovingFocusGroup: __scopeTabs || TABS_CONTEXT,
      asChild: true,
      focusable: !disabled,
      active: isSelected,
      children: /* @__PURE__ */_jsx(TabFrame, {
        ...(!isWeb && {
          onLayout: function (event) {
            setLayout(event.nativeEvent.layout);
          }
        }),
        onMouseEnter: composeEventHandlers(props.onMouseEnter, function () {
          if (layout) {
            onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("hover", layout);
          }
        }),
        onMouseLeave: composeEventHandlers(props.onMouseLeave, function () {
          onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("hover", null);
        }),
        role: "tab",
        "aria-selected": isSelected,
        "aria-controls": contentId,
        "data-state": isSelected ? "active" : "inactive",
        "data-disabled": disabled ? "" : void 0,
        id: triggerId,
        theme: activeTheme !== null && activeTheme !== void 0 ? activeTheme : null,
        unstyled,
        ...(!unstyled && {
          size: context.size
        }),
        ...(isSelected && {
          ...(!unstyled && !activeStyle && {
            backgroundColor: "$backgroundActive"
          }),
          ...activeStyle
        }),
        ...groupItemProps,
        disabled: disabled !== null && disabled !== void 0 ? disabled : groupItemProps.disabled,
        ...triggerProps,
        ref: composeRefs(forwardedRef, triggerRef),
        onPress: composeEventHandlers((_props_onPress = props.onPress) !== null && _props_onPress !== void 0 ? _props_onPress : void 0, function (event) {
          var webChecks = !isWeb || event.button === 0 && event.ctrlKey === false;
          if (!disabled && !isSelected && webChecks) {
            context.onChange(value);
          }
        }),
        ...(isWeb && {
          onKeyDown: composeEventHandlers(props.onKeyDown, function (event) {
            if ([" ", "Enter"].includes(event.key)) {
              context.onChange(value);
              event.preventDefault();
            }
          }),
          onFocus: composeEventHandlers(props.onFocus, function (event) {
            if (layout) {
              onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("focus", layout);
            }
            var isAutomaticActivation = context.activationMode !== "manual";
            if (!isSelected && !disabled && isAutomaticActivation) {
              context.onChange(value);
            }
          }),
          onBlur: composeEventHandlers(props.onBlur, function () {
            onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("focus", null);
          })
        })
      })
    });
  });
  TabsTrigger.displayName = TRIGGER_NAME;
  var TabsContent = ContentFrame.styleable(function TabsContent2(props, forwardedRef) {
    var {
      __scopeTabs,
      value,
      forceMount,
      children,
      ...contentProps
    } = props;
    var context = useTabsContext(__scopeTabs);
    var isSelected = value === context.value;
    var show = forceMount || isSelected;
    var triggerId = makeTriggerId(context.baseId, value);
    var contentId = makeContentId(context.baseId, value);
    if (!show) {
      return null;
    }
    return /* @__PURE__ */_jsx(ContentFrame, {
      "data-state": isSelected ? "active" : "inactive",
      "data-orientation": context.orientation,
      role: "tabpanel",
      "aria-labelledby": triggerId,
      // @ts-ignore
      hidden: !show,
      id: contentId,
      tabIndex: 0,
      ...contentProps,
      ref: forwardedRef,
      children
    }, value);
  });
  var TabsComponent = TabsFrame.styleable(function Tabs(props, forwardedRef) {
    var {
      __scopeTabs,
      value: valueProp,
      onValueChange,
      defaultValue,
      orientation = "horizontal",
      dir,
      activationMode = "manual",
      size = "$true",
      ...tabsProps
    } = props;
    var direction = useDirection(dir);
    var [value, setValue] = useControllableState({
      prop: valueProp,
      onChange: onValueChange,
      defaultProp: defaultValue !== null && defaultValue !== void 0 ? defaultValue : ""
    });
    var [triggersCount, setTriggersCount] = React.useState(0);
    var registerTrigger = useEvent(function () {
      return setTriggersCount(function (v) {
        return v + 1;
      });
    });
    var unregisterTrigger = useEvent(function () {
      return setTriggersCount(function (v) {
        return v - 1;
      });
    });
    return /* @__PURE__ */_jsx(SizableContext.Provider, {
      size,
      children: /* @__PURE__ */_jsx(TabsProvider, {
        scope: __scopeTabs,
        baseId: React.useId(),
        value,
        onChange: setValue,
        orientation,
        dir: direction,
        activationMode,
        size,
        registerTrigger,
        triggersCount,
        unregisterTrigger,
        children: /* @__PURE__ */_jsx(TabsFrame, {
          direction,
          "data-orientation": orientation,
          ...tabsProps,
          ref: forwardedRef
        })
      })
    });
  });
  return withStaticProperties(TabsComponent, {
    List: TabsList,
    /**
    * @deprecated Use Tabs.Tab instead
    */
    Trigger: TabsTrigger,
    Tab: TabsTrigger,
    Content: TabsContent
  });
}
function makeTriggerId(baseId, value) {
  return `${baseId}-trigger-${value}`;
}
function makeContentId(baseId, value) {
  return `${baseId}-content-${value}`;
}
export { createTabs };
