// @tamagui/tabs/dist/esm/Tabs.mjs
import { styled, View } from "@tamagui/core";
import { getButtonSized } from "@tamagui/get-button-sized";
import { SizableStack, ThemeableStack } from "@tamagui/stacks";
const TABS_NAME = "Tabs";
const DefaultTabsFrame = styled(SizableStack, {
  name: TABS_NAME
});
const TRIGGER_NAME = "TabsTrigger";
const DefaultTabsTabFrame = styled(View, {
  name: TRIGGER_NAME,
  role: "tab",
  variants: {
    size: {
      "...size": getButtonSized
    },
    disabled: {
      true: {
        pointerEvents: "none"
      }
    },
    unstyled: {
      false: {
        borderWidth: 0,
        backgroundColor: "$background",
        userSelect: "none",
        justifyContent: "center",
        alignItems: "center",
        flexWrap: "nowrap",
        flexDirection: "row",
        cursor: "pointer",
        pressStyle: {
          backgroundColor: "$backgroundPress"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineWidth: 2,
          outlineStyle: "solid",
          zIndex: 10
        }
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
}, {
  accept: {
    activeStyle: "style"
  }
});
const CONTENT_NAME = "TabsContent";
const DefaultTabsContentFrame = styled(ThemeableStack, {
  name: CONTENT_NAME
});
export { DefaultTabsContentFrame, DefaultTabsFrame, DefaultTabsTabFrame };
