// @tamagui/radio-group/dist/esm/createRadioGroup.native.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { isWeb, withStaticProperties } from "@tamagui/core";
import { useRadioGroup, useRadioGroupItem, useRadioGroupItemIndicator } from "@tamagui/radio-headless";
import { RovingFocusGroup } from "@tamagui/roving-focus";
import React from "react";
import { RadioGroupFrame, RadioGroupIndicatorFrame, RadioGroupItemFrame } from "./RadioGroup.native.js";
var ensureContext = function (x) {
  if (!x.context) {
    x.context = RadioGroupContext;
  }
};
var RadioGroupContext = /* @__PURE__ */React.createContext({});
var RadioGroupItemContext = /* @__PURE__ */React.createContext({
  checked: false,
  disabled: false
});
function createRadioGroup(createProps) {
  var {
    disableActiveTheme,
    Frame = RadioGroupFrame,
    Indicator = RadioGroupIndicatorFrame,
    Item = RadioGroupItemFrame
  } = createProps;
  ensureContext(Frame);
  ensureContext(Indicator);
  ensureContext(Item);
  var RadioGroupImp = Frame.styleable(function (props, ref) {
    var {
      value,
      defaultValue,
      onValueChange,
      required = false,
      disabled = false,
      name,
      native,
      accentColor,
      orientation = "vertical",
      ...rest
    } = props;
    var {
      providerValue,
      frameAttrs,
      rovingFocusGroupAttrs
    } = useRadioGroup({
      orientation,
      name,
      defaultValue,
      value,
      onValueChange,
      required,
      disabled,
      native,
      accentColor
    });
    return /* @__PURE__ */_jsx(RadioGroupContext.Provider, {
      value: providerValue,
      children: /* @__PURE__ */_jsx(RovingFocusGroup, {
        ...rovingFocusGroupAttrs,
        children: /* @__PURE__ */_jsx(RadioGroupFrame, {
          ...frameAttrs,
          ref,
          ...rest
        })
      })
    });
  });
  var RadioGroupItemImp = Item.styleable(function (props, ref) {
    var {
      value,
      labelledBy,
      onPress,
      onKeyDown,
      disabled,
      id,
      ...rest
    } = props;
    var {
      providerValue,
      bubbleInput,
      rovingFocusGroupAttrs,
      frameAttrs,
      isFormControl,
      native
    } = useRadioGroupItem({
      radioGroupContext: RadioGroupContext,
      value,
      id,
      labelledBy,
      disabled,
      onPress,
      onKeyDown
    });
    return /* @__PURE__ */_jsx(RadioGroupItemContext.Provider, {
      value: providerValue,
      children: isWeb && native ? bubbleInput : /* @__PURE__ */_jsxs(_Fragment, {
        children: [/* @__PURE__ */_jsx(RovingFocusGroup.Item, {
          ...rovingFocusGroupAttrs,
          children: /* @__PURE__ */_jsx(RadioGroupItemFrame, {
            ...frameAttrs,
            ref,
            ...rest
          })
        }), isFormControl && bubbleInput]
      })
    });
  });
  RadioGroupItemImp.displayName = "RadioGroupItem";
  var RadioIndicator = Indicator.styleable(function (props, forwardedRef) {
    var {
      forceMount,
      disabled,
      ...indicatorProps
    } = props;
    var {
      checked,
      ...useIndicatorRest
    } = useRadioGroupItemIndicator({
      radioGroupItemContext: RadioGroupItemContext,
      disabled
    });
    if (forceMount || checked) {
      return /* @__PURE__ */_jsx(Indicator, {
        ...useIndicatorRest,
        ref: forwardedRef,
        ...indicatorProps
      });
    }
    return null;
  });
  RadioIndicator.displayName = "RadioIndicator";
  var RadioGroup = withStaticProperties(RadioGroupImp, {
    Item: RadioGroupItemImp,
    Indicator: RadioIndicator
  });
  RadioGroup.displayName = "RadioGroup";
  return RadioGroup;
}
export { createRadioGroup };
