// @tamagui/animate/dist/cjs/Animate.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Animate_exports = {};
__export(Animate_exports, {
  Animate: () => Animate
});
module.exports = __toCommonJS(Animate_exports);
var import_animate_presence = require("@tamagui/animate-presence");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
function Animate({
  children,
  lazyMount,
  type,
  present,
  passThrough,
  ...props
}) {
  const [lazyMounted, setLazyMounted] = (0, import_react.useState)(lazyMount ? false : present);
  (0, import_react.useEffect)(() => {
    if (passThrough) return;
    if (!lazyMount) return;
    if (!present) return;
    (0, import_react.startTransition)(() => {
      setLazyMounted(present);
    });
  }, [lazyMount, present]);
  const mounted = !present ? false : lazyMount ? lazyMounted : present;
  if (type === "presence") {
    if (props.keepChildrenMounted) {
      return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate_presence.PresenceChild, {
        isPresent: true,
        ...(!passThrough && {
          initial: props.initial ? void 0 : false,
          onExitComplete: props.onExitComplete,
          enterVariant: props.enterVariant,
          exitVariant: props.exitVariant,
          enterExitVariant: props.enterExitVariant,
          // BUGFIX: this causes continous re-renders if keepChildrenMounted is true, see HeaderMenu
          // but since we always re-render this component on open changes this should be fine to leave off?
          presenceAffectsLayout: false,
          isPresent: present,
          custom: props.custom
        }),
        children
      });
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate_presence.AnimatePresence, {
      passThrough,
      ...props,
      children: mounted || passThrough ? children : null
    });
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children
  });
}
