// @tamagui/avatar/dist/cjs/Avatar.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Avatar_exports = {};
__export(Avatar_exports, {
  Avatar: () => Avatar,
  AvatarFallback: () => AvatarFallback,
  AvatarFallbackFrame: () => AvatarFallbackFrame,
  AvatarFrame: () => AvatarFrame,
  AvatarImage: () => AvatarImage,
  createAvatarScope: () => createAvatarScope
});
module.exports = __toCommonJS(Avatar_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@tamagui/core");
var import_create_context = require("@tamagui/create-context");
var import_helpers = require("@tamagui/helpers");
var import_image = require("@tamagui/image");
var import_shapes = require("@tamagui/shapes");
var import_stacks = require("@tamagui/stacks");
var React = __toESM(require("react"), 1);
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var AVATAR_NAME = "Avatar";
var [createAvatarContext, createAvatarScope] = (0, import_create_context.createContextScope)(AVATAR_NAME);
var [AvatarProvider, useAvatarContext] = createAvatarContext(AVATAR_NAME);
var IMAGE_NAME = "AvatarImage";
var AvatarImage = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var _getShapeSize;
  var {
    __scopeAvatar,
    src,
    source,
    onLoadingStatusChange = function () {},
    ...imageProps
  } = props;
  var context = useAvatarContext(IMAGE_NAME, __scopeAvatar);
  var [status, setStatus] = React.useState("idle");
  var shapeSize = (0, import_core.getVariableValue)((_getShapeSize = (0, import_shapes.getShapeSize)(context.size,
  // @ts-expect-error
  {
    tokens: (0, import_core.getTokens)()
  })) === null || _getShapeSize === void 0 ? void 0 : _getShapeSize.width);
  var resolvedSrc = src || (source && (typeof source === "undefined" ? "undefined" : _type_of(source)) === "object" && "uri" in source ? source.uri : source);
  React.useEffect(function () {
    if (!resolvedSrc) {
      setStatus("error");
    } else {
      setStatus("idle");
    }
  }, [resolvedSrc]);
  React.useEffect(function () {
    onLoadingStatusChange(status);
    context.onImageLoadingStatusChange(status);
  }, [status]);
  if (!resolvedSrc) {
    return null;
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_stacks.YStack, {
    fullscreen: true,
    zIndex: 1,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_image.Image, {
      position: "absolute",
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      objectFit: "cover",
      ...(typeof shapeSize === "number" && !Number.isNaN(shapeSize) && {
        width: shapeSize,
        height: shapeSize
      }),
      ...imageProps,
      // @ts-ignore
      ref: forwardedRef,
      // @ts-ignore
      src: resolvedSrc,
      // onLoadStart={() => {
      //   // setStatus('loading')
      // }}
      onError: function () {
        setStatus("error");
      },
      onLoad: function () {
        setStatus("loaded");
      }
    })
  });
});
AvatarImage.displayName = IMAGE_NAME;
var FALLBACK_NAME = "AvatarFallback";
var AvatarFallbackFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: FALLBACK_NAME,
  position: "absolute",
  fullscreen: true,
  zIndex: 0
});
var AvatarFallback = AvatarFallbackFrame.styleable(function (props, forwardedRef) {
  var {
    __scopeAvatar,
    delayMs,
    ...fallbackProps
  } = props;
  var context = useAvatarContext(FALLBACK_NAME, __scopeAvatar);
  var [canRender, setCanRender] = React.useState(delayMs === void 0);
  React.useEffect(function () {
    if (delayMs !== void 0) {
      var timerId = setTimeout(function () {
        return setCanRender(true);
      }, delayMs);
      return function () {
        return clearTimeout(timerId);
      };
    }
  }, [delayMs]);
  return canRender && context.imageLoadingStatus !== "loaded" ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(AvatarFallbackFrame, {
    ...fallbackProps,
    ref: forwardedRef
  }) : null;
});
AvatarFallback.displayName = FALLBACK_NAME;
var AvatarFrame = (0, import_core.styled)(import_shapes.Square, {
  name: AVATAR_NAME,
  position: "relative",
  overflow: "hidden"
});
var Avatar = (0, import_helpers.withStaticProperties)(/* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var {
    __scopeAvatar,
    size = "$true",
    ...avatarProps
  } = props;
  var [imageLoadingStatus, setImageLoadingStatus] = React.useState("idle");
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(AvatarProvider, {
    size,
    scope: __scopeAvatar,
    imageLoadingStatus,
    onImageLoadingStatusChange: setImageLoadingStatus,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(AvatarFrame, {
      size,
      ...avatarProps,
      ref: forwardedRef
    })
  });
}), {
  Image: AvatarImage,
  Fallback: AvatarFallback
});
Avatar.displayName = AVATAR_NAME;
