// @tamagui/floating/dist/cjs/index.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  FloatingDelayGroup: () => import_useDelayGroup.FloatingDelayGroup,
  FloatingOverrideContext: () => import_useFloating.FloatingOverrideContext,
  PopupTriggerMap: () => import_PopupTriggerMap.PopupTriggerMap,
  arrow: () => import_Floating.arrow,
  autoPlacement: () => import_Floating.autoPlacement,
  autoUpdate: () => import_Floating.autoUpdate,
  createFloatingEvents: () => import_createFloatingEvents.createFloatingEvents,
  detectOverflow: () => import_Floating.detectOverflow,
  flip: () => import_Floating.flip,
  getOverflowAncestors: () => import_Floating.getOverflowAncestors,
  hide: () => import_Floating.hide,
  inline: () => import_Floating.inline,
  inner: () => import_inner.inner,
  limitShift: () => import_Floating.limitShift,
  offset: () => import_Floating.offset,
  platform: () => import_Floating.platform,
  safePolygon: () => import_safePolygon.safePolygon,
  shift: () => import_Floating.shift,
  size: () => import_Floating.size,
  useClick: () => import_useClick.useClick,
  useDelayGroup: () => import_useDelayGroup.useDelayGroup,
  useDelayGroupContext: () => import_useDelayGroup.useDelayGroupContext,
  useFloating: () => import_useFloating.useFloating,
  useFloatingRaw: () => import_Floating2.useFloating,
  useFocus: () => import_useFocus.useFocus,
  useHover: () => import_useHover.useHover,
  useInnerOffset: () => import_useInnerOffset.useInnerOffset,
  useInteractions: () => import_useInteractions.useInteractions,
  useListNavigation: () => import_useListNavigation.useListNavigation,
  useRole: () => import_useRole.useRole,
  useTypeahead: () => import_useTypeahead.useTypeahead
});
module.exports = __toCommonJS(index_exports);
var import_Floating = require("./Floating.native.js");
var import_useFloating = require("./useFloating.native.js");
var import_Floating2 = require("./Floating.native.js");
var import_createFloatingEvents = require("./interactions/createFloatingEvents.native.js");
var import_PopupTriggerMap = require("./interactions/PopupTriggerMap.native.js");
var import_useInteractions = require("./interactions/useInteractions.native.js");
var import_useHover = require("./interactions/useHover.native.js");
var import_safePolygon = require("./interactions/safePolygon.native.js");
var import_useFocus = require("./interactions/useFocus.native.js");
var import_useRole = require("./interactions/useRole.native.js");
var import_useClick = require("./interactions/useClick.native.js");
var import_useListNavigation = require("./interactions/useListNavigation.native.js");
var import_useTypeahead = require("./interactions/useTypeahead.native.js");
var import_useInnerOffset = require("./interactions/useInnerOffset.native.js");
var import_useDelayGroup = require("./interactions/useDelayGroup.native.js");
var import_inner = require("./middleware/inner.native.js");
