// @tamagui/dismissable/dist/cjs/Dismissable.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Dismissable_exports = {};
__export(Dismissable_exports, {
  Dismissable: () => Dismissable,
  DismissableBranch: () => DismissableBranch,
  debugDismissableLayers: () => debugDismissableLayers,
  dispatchDiscreteCustomEvent: () => dispatchDiscreteCustomEvent,
  getDismissableLayerCount: () => getDismissableLayerCount,
  useDismissableLayersAbove: () => useDismissableLayersAbove,
  useHasDismissableLayers: () => useHasDismissableLayers,
  useIsInsideDismissable: () => useIsInsideDismissable
});
module.exports = __toCommonJS(Dismissable_exports);
var import_compose_refs = require("@tamagui/compose-refs");
var import_core = require("@tamagui/core");
var import_use_escape_keydown = require("@tamagui/use-escape-keydown");
var import_use_event = require("@tamagui/use-event");
var React = __toESM(require("react"), 1);
var ReactDOM = __toESM(require("react-dom"), 1);
var import_jsx_runtime = require("react/jsx-runtime");
function dispatchDiscreteCustomEvent(target, event) {
  if (target) ReactDOM.flushSync(() => target.dispatchEvent(event));
}
const DISMISSABLE_LAYER_NAME = "Dismissable";
const CONTEXT_UPDATE = "dismissable.update";
const POINTER_DOWN_OUTSIDE = "dismissable.pointerDownOutside";
const FOCUS_OUTSIDE = "dismissable.focusOutside";
let originalBodyPointerEvents;
const globalLayers = /* @__PURE__ */new Set();
const layerChangeListeners = /* @__PURE__ */new Set();
let layersWithPointerEventsDisabledCount = 0;
function notifyLayerChange() {
  for (const listener of layerChangeListeners) {
    listener();
  }
}
function getDismissableLayerCount() {
  return globalLayers.size;
}
function debugDismissableLayers() {
  const layers = Array.from(globalLayers);
  console.log("[Dismissable] Active layers:", layers.length, layers);
  return layers;
}
function useHasDismissableLayers() {
  const [count, setCount] = React.useState(() => globalLayers.size);
  React.useEffect(() => {
    setCount(globalLayers.size);
    const update = () => setCount(globalLayers.size);
    layerChangeListeners.add(update);
    return () => {
      layerChangeListeners.delete(update);
    };
  }, []);
  return count > 0;
}
const DismissableContext = React.createContext({
  layers: /* @__PURE__ */new Set(),
  layersWithOutsidePointerEventsDisabled: /* @__PURE__ */new Set(),
  branches: /* @__PURE__ */new Set()
});
function useIsInsideDismissable(ref) {
  const context = React.useContext(DismissableContext);
  const [isInside, setIsInside] = React.useState(false);
  React.useEffect(() => {
    const check = () => {
      const el = ref.current;
      if (!el) {
        setIsInside(false);
        return;
      }
      for (const layer of context.layers) {
        if (layer.contains(el)) {
          setIsInside(true);
          return;
        }
      }
      setIsInside(false);
    };
    check();
    document.addEventListener(CONTEXT_UPDATE, check);
    return () => document.removeEventListener(CONTEXT_UPDATE, check);
  }, [context.layers, ref]);
  return isInside;
}
function useDismissableLayersAbove(ref) {
  const context = React.useContext(DismissableContext);
  const [count, setCount] = React.useState(0);
  React.useEffect(() => {
    const check = () => {
      const el = ref.current;
      if (!el) {
        setCount(0);
        return;
      }
      let above = 0;
      for (const layer of context.layers) {
        if (layer.contains(el)) {
          above++;
        }
      }
      setCount(above);
    };
    check();
    document.addEventListener(CONTEXT_UPDATE, check);
    return () => document.removeEventListener(CONTEXT_UPDATE, check);
  }, [context.layers, ref]);
  return count;
}
const Dismissable = React.forwardRef((props, forwardedRef) => {
  const {
    disableOutsidePointerEvents = false,
    forceUnmount,
    onEscapeKeyDown,
    onPointerDownOutside,
    onFocusOutside,
    onInteractOutside,
    onDismiss,
    asChild,
    children,
    branches: branchesProp,
    ...layerProps
  } = props;
  const Comp = asChild ? import_core.Slot : import_core.View;
  const context = React.useContext(DismissableContext);
  const [node, setNode] = React.useState(null);
  const [, force] = React.useState({});
  const composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, node2 => setNode(node2));
  const layers = Array.from(context.layers);
  const [highestLayerWithOutsidePointerEventsDisabled] = [...context.layersWithOutsidePointerEventsDisabled].slice(-1);
  const highestLayerWithOutsidePointerEventsDisabledIndex = layers.indexOf(highestLayerWithOutsidePointerEventsDisabled);
  const index = node ? layers.indexOf(node) : -1;
  const isBodyPointerEventsDisabled = context.layersWithOutsidePointerEventsDisabled.size > 0;
  const isPointerEventsEnabled = index >= highestLayerWithOutsidePointerEventsDisabledIndex;
  const pointerDownOutside = usePointerDownOutside(event => {
    const target = event.target;
    const branches = branchesProp || context.branches;
    const isPointerDownOnBranch = [...branches].some(branch => branch.contains(target));
    if (!isPointerEventsEnabled || isPointerDownOnBranch) return;
    onPointerDownOutside?.(event);
    onInteractOutside?.(event);
    if (!event.defaultPrevented) onDismiss?.();
  });
  const focusOutside = useFocusOutside(event => {
    const target = event.target;
    const branches = branchesProp || context.branches;
    const isFocusInBranch = [...branches].some(branch => branch.contains(target));
    if (isFocusInBranch) return;
    onFocusOutside?.(event);
    onInteractOutside?.(event);
    if (!event.defaultPrevented) onDismiss?.();
  });
  const forceUnmountRef = React.useRef(forceUnmount);
  React.useEffect(() => {
    forceUnmountRef.current = forceUnmount;
  }, [forceUnmount]);
  (0, import_use_escape_keydown.useEscapeKeydown)(event => {
    if (forceUnmountRef.current) return;
    const currentLayers = Array.from(context.layers);
    const currentIndex = node ? currentLayers.indexOf(node) : -1;
    const isHighestLayer = currentIndex === currentLayers.length - 1;
    if (!isHighestLayer) return;
    onEscapeKeyDown?.(event);
    if (!event.defaultPrevented && onDismiss) {
      event.preventDefault();
      onDismiss();
    }
  });
  React.useEffect(() => {
    if (!node) return;
    if (forceUnmount) return;
    if (disableOutsidePointerEvents) {
      if (context.layersWithOutsidePointerEventsDisabled.size === 0) {
        originalBodyPointerEvents = document.body.style.pointerEvents;
        document.body.style.pointerEvents = "none";
      }
      context.layersWithOutsidePointerEventsDisabled.add(node);
      layersWithPointerEventsDisabledCount++;
    }
    context.layers.add(node);
    globalLayers.add(node);
    if (disableOutsidePointerEvents || layersWithPointerEventsDisabledCount > 0) {
      dispatchUpdate();
    }
    notifyLayerChange();
    return () => {
      if (disableOutsidePointerEvents) {
        if (context.layersWithOutsidePointerEventsDisabled.size === 1) {
          document.body.style.pointerEvents = originalBodyPointerEvents;
        }
      }
    };
  }, [node, disableOutsidePointerEvents, forceUnmount, context]);
  React.useEffect(() => {
    if (forceUnmount) return;
    return () => {
      if (!node) return;
      const hadPointerEventsDisabled = context.layersWithOutsidePointerEventsDisabled.has(node);
      context.layers.delete(node);
      context.layersWithOutsidePointerEventsDisabled.delete(node);
      globalLayers.delete(node);
      if (layersWithPointerEventsDisabledCount > 0) {
        dispatchUpdate();
      }
      notifyLayerChange();
      if (hadPointerEventsDisabled) {
        layersWithPointerEventsDisabledCount--;
      }
    };
  }, [node, context, forceUnmount]);
  React.useEffect(() => {
    const handleUpdate = () => {
      if (layersWithPointerEventsDisabledCount > 0) {
        force({});
      }
    };
    document.addEventListener(CONTEXT_UPDATE, handleUpdate);
    return () => document.removeEventListener(CONTEXT_UPDATE, handleUpdate);
  }, []);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Comp, {
    ...layerProps,
    ref: composedRefs,
    ...(!asChild && {
      display: "contents"
    }),
    pointerEvents: isBodyPointerEventsDisabled ? isPointerEventsEnabled ? "auto" : "none" : void 0,
    onFocusCapture: (0, import_core.composeEventHandlers)(props.onFocusCapture, focusOutside.onFocusCapture),
    onBlurCapture: (0, import_core.composeEventHandlers)(props.onBlurCapture, focusOutside.onBlurCapture),
    onPointerDownCapture: (0, import_core.composeEventHandlers)(props.onPointerDownCapture, pointerDownOutside.onPointerDownCapture),
    children
  });
});
Dismissable.displayName = DISMISSABLE_LAYER_NAME;
const BRANCH_NAME = "DismissableBranch";
const DismissableBranch = React.forwardRef((props, forwardedRef) => {
  const {
    branches: branchesProp,
    ...rest
  } = props;
  const context = React.useContext(DismissableContext);
  const ref = React.useRef(null);
  const composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, ref);
  React.useEffect(() => {
    const node = ref.current;
    if (!(node instanceof HTMLElement)) return;
    const branches = branchesProp || context.branches;
    if (node && branches) {
      branches.add(node);
      return () => {
        branches.delete(node);
      };
    }
  }, [branchesProp, context.branches]);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.View, {
    asChild: "except-style",
    ...rest,
    ref: composedRefs
  });
});
DismissableBranch.displayName = BRANCH_NAME;
function usePointerDownOutside(onPointerDownOutside) {
  const handlePointerDownOutside = (0, import_use_event.useEvent)(onPointerDownOutside);
  const isPointerInsideReactTreeRef = React.useRef(false);
  const handleClickRef = React.useRef(() => {});
  React.useEffect(() => {
    const handlePointerDown = event => {
      if (event.target && !isPointerInsideReactTreeRef.current) {
        let handleAndDispatchPointerDownOutsideEvent = function () {
          handleAndDispatchCustomEvent(POINTER_DOWN_OUTSIDE, handlePointerDownOutside, eventDetail, {
            discrete: true
          });
        };
        const eventDetail = {
          originalEvent: event
        };
        if (event.pointerType === "touch") {
          document.removeEventListener("click", handleClickRef.current);
          handleClickRef.current = handleAndDispatchPointerDownOutsideEvent;
          document.addEventListener("click", handleClickRef.current, {
            once: true
          });
        } else {
          handleAndDispatchPointerDownOutsideEvent();
        }
      }
      isPointerInsideReactTreeRef.current = false;
    };
    const timerId = setTimeout(() => {
      document.addEventListener("pointerdown", handlePointerDown);
    }, 0);
    return () => {
      window.clearTimeout(timerId);
      document.removeEventListener("pointerdown", handlePointerDown);
      document.removeEventListener("click", handleClickRef.current);
    };
  }, [handlePointerDownOutside]);
  return {
    // ensures we check React component tree (not just DOM tree)
    onPointerDownCapture: () => {
      isPointerInsideReactTreeRef.current = true;
    }
  };
}
function useFocusOutside(onFocusOutside) {
  const handleFocusOutside = (0, import_use_event.useEvent)(onFocusOutside);
  const isFocusInsideReactTreeRef = React.useRef(false);
  React.useEffect(() => {
    const handleFocus = event => {
      if (event.target && !isFocusInsideReactTreeRef.current) {
        const eventDetail = {
          originalEvent: event
        };
        handleAndDispatchCustomEvent(FOCUS_OUTSIDE, handleFocusOutside, eventDetail, {
          discrete: false
        });
      }
    };
    document.addEventListener("focusin", handleFocus);
    return () => document.removeEventListener("focusin", handleFocus);
  }, [handleFocusOutside]);
  return {
    onFocusCapture: () => {
      isFocusInsideReactTreeRef.current = true;
    },
    onBlurCapture: () => {
      isFocusInsideReactTreeRef.current = false;
    }
  };
}
function dispatchUpdate() {
  const event = new CustomEvent(CONTEXT_UPDATE);
  document.dispatchEvent(event);
}
function handleAndDispatchCustomEvent(name, handler, detail, {
  discrete
}) {
  const target = detail.originalEvent.target;
  const event = new CustomEvent(name, {
    bubbles: false,
    cancelable: true,
    detail
  });
  if (handler) target.addEventListener(name, handler, {
    once: true
  });
  if (discrete) {
    dispatchDiscreteCustomEvent(target, event);
  } else {
    target.dispatchEvent(event);
  }
}
