// @tamagui/select/dist/esm/context.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createStyledContext } from "@tamagui/core";
import { getPortal } from "@tamagui/native";
import { createContext } from "react";
var SelectZIndexContext = /* @__PURE__ */createContext(void 0);
var {
  Provider: SelectProvider,
  useStyledContext: useSelectContext
} = createStyledContext(null, "Select");
var {
  Provider: SelectItemParentProvider,
  useStyledContext: useSelectItemParentContext
} = createStyledContext(null, "SelectItem");
var ForwardSelectContext = function (param) {
  var {
    context,
    itemContext,
    children
  } = param;
  var portalState = getPortal().state;
  if (portalState.type === "teleport") {
    return children;
  }
  return /* @__PURE__ */_jsx(SelectProvider, {
    isInSheet: true,
    scope: context.scopeName,
    ...context,
    children: /* @__PURE__ */_jsx(SelectItemParentProvider, {
      scope: context.scopeName,
      ...itemContext,
      children
    })
  });
};
export { ForwardSelectContext, SelectItemParentProvider, SelectProvider, SelectZIndexContext, useSelectContext, useSelectItemParentContext };
