// @tamagui/select/dist/esm/SelectItem.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb, useIsomorphicLayoutEffect } from "@tamagui/constants";
import { createStyledContext } from "@tamagui/core";
import { ListItem } from "@tamagui/list-item";
import * as React from "react";
import { useSelectItemParentContext } from "./context.native.js";
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
    return !!right[Symbol.hasInstance](left);
  } else {
    return left instanceof right;
  }
}
var ITEM_NAME = "SelectItem";
var {
  Provider: SelectItemContextProvider,
  useStyledContext: useSelectItemContext
} = createStyledContext(null, ITEM_NAME);
var SelectItem = ListItem.Frame.styleable(function SelectItem2(props, forwardedRef) {
  var {
    scope,
    value,
    disabled = false,
    textValue: textValueProp,
    index,
    ...restProps
  } = props;
  var context = useSelectItemParentContext(scope);
  var {
    setSelectedIndex,
    listRef,
    setOpen,
    onChange,
    activeIndexSubscribe,
    activeIndexRef,
    valueSubscribe,
    allowMouseUpRef,
    allowSelectRef,
    setValueAtIndex,
    selectTimeoutRef,
    dataRef,
    interactions,
    shouldRenderWebNative,
    size,
    onActiveChange,
    initialValue,
    setActiveIndexFast
  } = context;
  var [isSelected, setSelected] = React.useState(initialValue === value);
  useIsomorphicLayoutEffect(function () {
    if (initialValue === value) {
      setSelectedIndex(index);
    }
  }, []);
  React.useEffect(function () {
    var handleActiveIndex = function (i) {
      if (index === i) {
        onActiveChange(value, index);
        if (isWeb) {
          requestAnimationFrame(function () {
            var _listRef_current_index;
            listRef === null || listRef === void 0 ? void 0 : (_listRef_current_index = listRef.current[index]) === null || _listRef_current_index === void 0 ? void 0 : _listRef_current_index.focus();
          });
        }
      }
    };
    var currentActiveIndex = activeIndexRef === null || activeIndexRef === void 0 ? void 0 : activeIndexRef.current;
    if (currentActiveIndex !== null && currentActiveIndex !== void 0) {
      handleActiveIndex(currentActiveIndex);
    }
    return activeIndexSubscribe(handleActiveIndex);
  }, [index]);
  React.useEffect(function () {
    return valueSubscribe(function (val) {
      setSelected(val === value);
    });
  }, [value]);
  var textId = React.useId();
  var refCallback = React.useCallback(function (node) {
    if (!isWeb) return;
    if (_instanceof(node, HTMLElement)) {
      if (listRef) {
        listRef.current[index] = node;
      }
    }
  }, [index, listRef]);
  var composedRefs = useComposedRefs(forwardedRef, refCallback);
  useIsomorphicLayoutEffect(function () {
    setValueAtIndex(index, value);
  }, [index, setValueAtIndex, value]);
  function handleSelect() {
    setSelectedIndex(index);
    onChange(value);
    setOpen(false);
  }
  var selectItemProps = React.useMemo(function () {
    return interactions ? interactions.getItemProps({
      onTouchMove() {
        allowSelectRef.current = true;
        allowMouseUpRef.current = false;
      },
      onTouchEnd() {
        allowSelectRef.current = false;
        allowMouseUpRef.current = true;
      },
      onKeyDown(event) {
        if (event.key === "Enter" || event.key === " " && !(dataRef === null || dataRef === void 0 ? void 0 : dataRef.current.typing)) {
          event.preventDefault();
          handleSelect();
        } else if (event.key === "ArrowDown" || event.key === "ArrowUp") {
          event.preventDefault();
          event.stopPropagation();
          var _listRef_current_length;
          var itemCount = (_listRef_current_length = listRef === null || listRef === void 0 ? void 0 : listRef.current.length) !== null && _listRef_current_length !== void 0 ? _listRef_current_length : 0;
          if (itemCount === 0) return;
          var nextIndex;
          if (event.key === "ArrowDown") {
            nextIndex = index + 1 >= itemCount ? 0 : index + 1;
          } else {
            nextIndex = index - 1 < 0 ? itemCount - 1 : index - 1;
          }
          setActiveIndexFast === null || setActiveIndexFast === void 0 ? void 0 : setActiveIndexFast(nextIndex);
        } else {
          allowSelectRef.current = true;
        }
      },
      onClick() {
        if (allowSelectRef.current) {
          handleSelect();
        }
      },
      onMouseUp() {
        if (!allowMouseUpRef.current) {
          allowMouseUpRef.current = true;
          allowSelectRef.current = true;
          return;
        }
        if (allowSelectRef.current) {
          handleSelect();
        }
        clearTimeout(selectTimeoutRef.current);
        selectTimeoutRef.current = setTimeout(function () {
          allowSelectRef.current = true;
        });
      }
    }) : {
      onPress: handleSelect
    };
  }, [handleSelect, index, listRef, setActiveIndexFast]);
  return /* @__PURE__ */_jsx(SelectItemContextProvider, {
    scope,
    value,
    textId: textId || "",
    isSelected,
    children: shouldRenderWebNative ? /* @__PURE__ */_jsx("option", {
      value,
      children: props.children
    }) : /* @__PURE__ */_jsx(ListItem.Frame, {
      render: "div",
      componentName: ITEM_NAME,
      ref: composedRefs,
      role: "option",
      "aria-labelledby": textId,
      "aria-selected": isSelected,
      "data-state": isSelected ? "active" : "inactive",
      "aria-disabled": disabled || void 0,
      "data-disabled": disabled ? "" : void 0,
      tabIndex: disabled ? void 0 : -1,
      ...(!props.unstyled && {
        cursor: "default",
        size,
        outlineOffset: -0.5,
        zIndex: 100,
        hoverStyle: {
          backgroundColor: "$backgroundHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress"
        },
        focusStyle: {
          backgroundColor: "$backgroundFocus"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineWidth: 1,
          outlineStyle: "solid"
        }
      }),
      ...restProps,
      ...selectItemProps
    })
  });
}, {
  disableTheme: true
});
export { SelectItem, SelectItemContextProvider, useSelectItemContext };
