// @tamagui/select/dist/esm/SelectTrigger.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useComposedRefs } from "@tamagui/compose-refs";
import { ListItem } from "@tamagui/list-item";
import * as React from "react";
import { useSelectContext, useSelectItemParentContext } from "./context.native.js";
var TRIGGER_NAME = "SelectTrigger";
var SelectTrigger = /* @__PURE__ */React.forwardRef(function SelectTrigger2(props, forwardedRef) {
  var _context_floatingContext;
  var {
    scope,
    disabled = false,
    unstyled = false,
    ...triggerProps
  } = props;
  var context = useSelectContext(scope);
  var itemParentContext = useSelectItemParentContext(scope);
  var composedRefs = useComposedRefs(forwardedRef, (_context_floatingContext = context.floatingContext) === null || _context_floatingContext === void 0 ? void 0 : _context_floatingContext.refs.setReference);
  if (itemParentContext.shouldRenderWebNative) return null;
  return /* @__PURE__ */_jsx(ListItem, {
    componentName: TRIGGER_NAME,
    unstyled,
    render: "button",
    type: "button",
    id: itemParentContext.id,
    ...(!unstyled && {
      focusVisibleStyle: {
        outlineStyle: "solid",
        outlineWidth: 2,
        outlineColor: "$outlineColor"
      },
      borderWidth: 1,
      size: itemParentContext.size
    }),
    role: "combobox",
    "aria-haspopup": "listbox",
    "aria-expanded": context.open,
    "aria-autocomplete": "none",
    dir: context.dir,
    disabled,
    "data-disabled": disabled ? "" : void 0,
    ...triggerProps,
    ref: composedRefs,
    onPress() {
      itemParentContext.setOpen(!context.open);
    }
  });
});
export { SelectTrigger };
