// @tamagui/select/dist/esm/SelectItemText.native.js
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { useComposedRefs } from "@tamagui/compose-refs";
import { styled, useIsomorphicLayoutEffect } from "@tamagui/core";
import { SizableText } from "@tamagui/text";
import * as React from "react";
import { useSelectItemParentContext } from "./context.native.js";
import { useSelectItemContext } from "./SelectItem.native.js";
var ITEM_TEXT_NAME = "SelectItemText";
var SelectItemTextFrame = styled(SizableText, {
  name: ITEM_TEXT_NAME,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        color: "$color",
        ellipsis: true
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var SelectItemText = SelectItemTextFrame.styleable(function SelectItemText2(props, forwardedRef) {
  var {
    scope,
    className,
    ...itemTextProps
  } = props;
  var itemParentContext = useSelectItemParentContext(scope);
  var ref = React.useRef(null);
  var composedRefs = useComposedRefs(forwardedRef, ref);
  var itemContext = useSelectItemContext(scope);
  var contents = React.useRef(null);
  contents.current = /* @__PURE__ */_jsx(SelectItemTextFrame, {
    className,
    size: itemParentContext.size,
    id: itemContext.textId,
    ...itemTextProps,
    ref: composedRefs
  });
  useIsomorphicLayoutEffect(function () {
    if (itemParentContext.initialValue === itemContext.value) {
      itemParentContext.setSelectedItem(contents.current);
    }
  }, []);
  useIsomorphicLayoutEffect(function () {
    return itemParentContext.valueSubscribe(function (val) {
      if (val === itemContext.value) {
        itemParentContext.setSelectedItem(contents.current);
      }
    });
  }, [itemContext.value]);
  if (itemParentContext.shouldRenderWebNative) {
    return /* @__PURE__ */_jsx(_Fragment, {
      children: props.children
    });
  }
  return /* @__PURE__ */_jsx(_Fragment, {
    children: contents.current
  });
});
export { ITEM_TEXT_NAME, SelectItemText, SelectItemTextFrame };
