// @tamagui/select/dist/esm/SelectItem.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb, useIsomorphicLayoutEffect } from "@tamagui/constants";
import { createStyledContext } from "@tamagui/core";
import { ListItem } from "@tamagui/list-item";
import * as React from "react";
import { useSelectItemParentContext } from "./context.mjs";
import { jsx } from "react/jsx-runtime";
const ITEM_NAME = "SelectItem";
const {
  Provider: SelectItemContextProvider,
  useStyledContext: useSelectItemContext
} = createStyledContext(null, ITEM_NAME);
const SelectItem = ListItem.Frame.styleable(function SelectItem2(props, forwardedRef) {
  const {
    scope,
    value,
    disabled = false,
    textValue: textValueProp,
    index,
    ...restProps
  } = props;
  const context = useSelectItemParentContext(scope);
  const {
    setSelectedIndex,
    listRef,
    setOpen,
    onChange,
    activeIndexSubscribe,
    activeIndexRef,
    valueSubscribe,
    allowMouseUpRef,
    allowSelectRef,
    setValueAtIndex,
    selectTimeoutRef,
    dataRef,
    interactions,
    shouldRenderWebNative,
    size,
    onActiveChange,
    initialValue,
    setActiveIndexFast
  } = context;
  const [isSelected, setSelected] = React.useState(initialValue === value);
  useIsomorphicLayoutEffect(() => {
    if (initialValue === value) {
      setSelectedIndex(index);
    }
  }, []);
  React.useEffect(() => {
    const handleActiveIndex = i => {
      if (index === i) {
        onActiveChange(value, index);
        if (isWeb) {
          requestAnimationFrame(() => {
            listRef?.current[index]?.focus();
          });
        }
      }
    };
    const currentActiveIndex = activeIndexRef?.current;
    if (currentActiveIndex !== null && currentActiveIndex !== void 0) {
      handleActiveIndex(currentActiveIndex);
    }
    return activeIndexSubscribe(handleActiveIndex);
  }, [index]);
  React.useEffect(() => {
    return valueSubscribe(val => {
      setSelected(val === value);
    });
  }, [value]);
  const textId = React.useId();
  const refCallback = React.useCallback(node => {
    if (!isWeb) return;
    if (node instanceof HTMLElement) {
      if (listRef) {
        listRef.current[index] = node;
      }
    }
  }, [index, listRef]);
  const composedRefs = useComposedRefs(forwardedRef, refCallback);
  useIsomorphicLayoutEffect(() => {
    setValueAtIndex(index, value);
  }, [index, setValueAtIndex, value]);
  function handleSelect() {
    setSelectedIndex(index);
    onChange(value);
    setOpen(false);
  }
  const selectItemProps = React.useMemo(() => {
    return interactions ? interactions.getItemProps({
      onTouchMove() {
        allowSelectRef.current = true;
        allowMouseUpRef.current = false;
      },
      onTouchEnd() {
        allowSelectRef.current = false;
        allowMouseUpRef.current = true;
      },
      onKeyDown(event) {
        if (event.key === "Enter" || event.key === " " && !dataRef?.current.typing) {
          event.preventDefault();
          handleSelect();
        } else if (event.key === "ArrowDown" || event.key === "ArrowUp") {
          event.preventDefault();
          event.stopPropagation();
          const itemCount = listRef?.current.length ?? 0;
          if (itemCount === 0) return;
          let nextIndex;
          if (event.key === "ArrowDown") {
            nextIndex = index + 1 >= itemCount ? 0 : index + 1;
          } else {
            nextIndex = index - 1 < 0 ? itemCount - 1 : index - 1;
          }
          setActiveIndexFast?.(nextIndex);
        } else {
          allowSelectRef.current = true;
        }
      },
      onClick() {
        if (allowSelectRef.current) {
          handleSelect();
        }
      },
      onMouseUp() {
        if (!allowMouseUpRef.current) {
          allowMouseUpRef.current = true;
          allowSelectRef.current = true;
          return;
        }
        if (allowSelectRef.current) {
          handleSelect();
        }
        clearTimeout(selectTimeoutRef.current);
        selectTimeoutRef.current = setTimeout(() => {
          allowSelectRef.current = true;
        });
      }
    }) : {
      onPress: handleSelect
    };
  }, [handleSelect, index, listRef, setActiveIndexFast]);
  return /* @__PURE__ */jsx(SelectItemContextProvider, {
    scope,
    value,
    textId: textId || "",
    isSelected,
    children: shouldRenderWebNative ? /* @__PURE__ */jsx("option", {
      value,
      children: props.children
    }) : /* @__PURE__ */jsx(ListItem.Frame, {
      render: "div",
      componentName: ITEM_NAME,
      ref: composedRefs,
      role: "option",
      "aria-labelledby": textId,
      "aria-selected": isSelected,
      "data-state": isSelected ? "active" : "inactive",
      "aria-disabled": disabled || void 0,
      "data-disabled": disabled ? "" : void 0,
      tabIndex: disabled ? void 0 : -1,
      ...(!props.unstyled && {
        cursor: "default",
        size,
        outlineOffset: -0.5,
        zIndex: 100,
        hoverStyle: {
          backgroundColor: "$backgroundHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress"
        },
        focusStyle: {
          backgroundColor: "$backgroundFocus"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineWidth: 1,
          outlineStyle: "solid"
        }
      }),
      ...restProps,
      ...selectItemProps
    })
  });
}, {
  disableTheme: true
});
export { SelectItem, SelectItemContextProvider, useSelectItemContext };
