// @tamagui/checkbox/dist/jsx/createCheckbox.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createCheckbox_exports = {};
__export(createCheckbox_exports, {
  CheckboxContext: () => CheckboxContext,
  createCheckbox: () => createCheckbox
});
module.exports = __toCommonJS(createCheckbox_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_checkbox_headless = require("@tamagui/checkbox-headless");
var import_core = require("@tamagui/core");
var import_focusable = require("@tamagui/focusable");
var import_font_size = require("@tamagui/font-size");
var import_get_token = require("@tamagui/get-token");
var import_helpers_tamagui = require("@tamagui/helpers-tamagui");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var import_react = __toESM(require("react"), 1);
var import_Checkbox = require("./Checkbox.native.js");
var import_CheckboxStyledContext = require("./CheckboxStyledContext.native.js");
var CheckboxContext = /* @__PURE__ */import_react.default.createContext({
  checked: false,
  disabled: false
});
var ensureContext = function (x) {
  if (!x.context) x.context = CheckboxContext;
};
function createCheckbox(createProps) {
  var {
    Frame = import_Checkbox.CheckboxFrame,
    Indicator = import_Checkbox.CheckboxIndicatorFrame
  } = createProps;
  ensureContext(Frame);
  ensureContext(Indicator);
  var FrameComponent = Frame.styleable(function Checkbox(_props, forwardedRef) {
    var {
      scaleSize = .45,
      sizeAdjust = 0,
      scaleIcon,
      checked: checkedProp,
      defaultChecked,
      onCheckedChange,
      native,
      unstyled = false,
      activeStyle,
      activeTheme,
      ...props
    } = _props;
    var propsActive = (0, import_core.useProps)(props);
    var styledContext = import_react.default.useContext(import_CheckboxStyledContext.CheckboxStyledContext.context);
    var adjustedSize = 0;
    var size = 0;
    if (!unstyled) {
      var _propsActive_size, _ref;
      adjustedSize = (0, import_core.getVariableValue)((0, import_get_token.getSize)((_ref = (_propsActive_size = propsActive.size) !== null && _propsActive_size !== void 0 ? _propsActive_size : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size) !== null && _ref !== void 0 ? _ref : "$true", {
        shift: sizeAdjust
      }));
      size = scaleSize ? Math.round(adjustedSize * scaleSize) : adjustedSize;
    }
    var [checked = false, setChecked] = (0, import_use_controllable_state.useControllableState)({
      prop: checkedProp,
      defaultProp: defaultChecked,
      onChange: onCheckedChange
    });
    var {
      checkboxProps,
      checkboxRef,
      bubbleInput
    } = (0, import_checkbox_headless.useCheckbox)(
    // @ts-ignore
    propsActive, [checked, setChecked], forwardedRef);
    import_react.default.useEffect(function () {
      if (!props.id) return;
      if (props.disabled) return;
      return (0, import_focusable.registerFocusable)(props.id, {
        focusAndSelect: function () {
          setChecked === null || setChecked === void 0 || setChecked(function (value) {
            return !value;
          });
        },
        focus: function () {}
      });
    }, [props.id, props.disabled]);
    if ((0, import_core.shouldRenderNativePlatform)(native) === "web") return /* @__PURE__ */(0, import_jsx_runtime.jsx)("input", {
      type: "checkbox",
      defaultChecked: (0, import_checkbox_headless.isIndeterminate)(checked) ? false : checked,
      tabIndex: -1,
      ref: checkboxRef,
      disabled: checkboxProps.disabled,
      style: {
        appearance: "auto",
        accentColor: "var(--color6)",
        ...checkboxProps.style
      }
    });
    var memoizedContext = (0, import_react.useMemo)(function () {
      return {
        checked,
        disabled: checkboxProps.disabled
      };
    }, [checked, checkboxProps.disabled]);
    var isActive = !!checked;
    var disabled = checkboxProps.disabled;
    var _propsActive_size1, _ref1, _ref2;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(CheckboxContext.Provider, {
      value: memoizedContext,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_CheckboxStyledContext.CheckboxStyledContext.Provider, {
        size: (_ref1 = (_propsActive_size1 = propsActive.size) !== null && _propsActive_size1 !== void 0 ? _propsActive_size1 : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size) !== null && _ref1 !== void 0 ? _ref1 : "$true",
        scaleIcon: (_ref2 = scaleIcon !== null && scaleIcon !== void 0 ? scaleIcon : styledContext === null || styledContext === void 0 ? void 0 : styledContext.scaleIcon) !== null && _ref2 !== void 0 ? _ref2 : 1,
        unstyled,
        active: isActive,
        disabled,
        children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(Frame, {
          render: "button",
          ref: checkboxRef,
          unstyled,
          theme: activeTheme !== null && activeTheme !== void 0 ? activeTheme : null,
          ...(import_core.isWeb && {
            type: "button"
          }),
          ...(!unstyled && {
            width: size,
            height: size,
            size
          }),
          checked,
          disabled,
          ...checkboxProps,
          ...props,
          ...(isActive && {
            ...(!unstyled && !activeStyle && {
              backgroundColor: "$backgroundActive"
            }),
            ...activeStyle
          }),
          children: propsActive.children
        }), bubbleInput]
      })
    });
  });
  var IndicatorComponent = Indicator.styleable(function (props, forwardedRef) {
    var {
      children: childrenProp,
      forceMount,
      disablePassStyles,
      unstyled = false,
      activeStyle,
      ...indicatorProps
    } = props;
    var styledContext = import_CheckboxStyledContext.CheckboxStyledContext.useStyledContext();
    var {
      active
    } = styledContext;
    var children = childrenProp;
    if (!unstyled) {
      var iconSize = (typeof styledContext.size === "number" ? styledContext.size * .65 : (0, import_font_size.getFontSize)(styledContext.size)) * styledContext.scaleIcon;
      var theme = (0, import_core.useTheme)();
      var getThemedIcon = (0, import_helpers_tamagui.useGetThemedIcon)({
        size: iconSize,
        color: theme.color
      });
      children = import_react.default.Children.toArray(childrenProp).map(function (child) {
        if (disablePassStyles || ! /* @__PURE__ */import_react.default.isValidElement(child)) return child;
        return getThemedIcon(child);
      });
    }
    var context = import_react.default.useContext(CheckboxContext);
    if (forceMount || (0, import_checkbox_headless.isIndeterminate)(context.checked) || context.checked === true) return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Indicator, {
      pointerEvents: "none",
      ...indicatorProps,
      ...(active && activeStyle),
      ref: forwardedRef,
      children
    });
    return null;
  });
  return (0, import_core.withStaticProperties)(FrameComponent, {
    Indicator: IndicatorComponent
  });
}
