// @tamagui/avatar/dist/jsx/Avatar.mjs
import { getTokens, getVariableValue, styled } from "@tamagui/core";
import { createContextScope } from "@tamagui/create-context";
import { withStaticProperties } from "@tamagui/helpers";
import { Image } from "@tamagui/image";
import { Square, getShapeSize } from "@tamagui/shapes";
import { YStack } from "@tamagui/stacks";
import * as React from "react";
import { jsx } from "react/jsx-runtime";
const AVATAR_NAME = "Avatar";
const [createAvatarContext, createAvatarScope] = createContextScope(AVATAR_NAME);
const [AvatarProvider, useAvatarContext] = createAvatarContext(AVATAR_NAME);
const IMAGE_NAME = "AvatarImage";
const AvatarImage = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeAvatar,
    src,
    source,
    onLoadingStatusChange = () => {},
    ...imageProps
  } = props;
  const context = useAvatarContext(IMAGE_NAME, __scopeAvatar);
  const [status, setStatus] = React.useState("idle");
  const shapeSize = getVariableValue(getShapeSize(context.size,
  // @ts-expect-error
  {
    tokens: getTokens()
  })?.width);
  const resolvedSrc = src || (source && typeof source === "object" && "uri" in source ? source.uri : source);
  React.useEffect(() => {
    if (!resolvedSrc) {
      setStatus("error");
    } else {
      setStatus("idle");
    }
  }, [resolvedSrc]);
  React.useEffect(() => {
    onLoadingStatusChange(status);
    context.onImageLoadingStatusChange(status);
  }, [status]);
  if (!resolvedSrc) {
    return null;
  }
  return /* @__PURE__ */jsx(YStack, {
    fullscreen: true,
    zIndex: 1,
    children: /* @__PURE__ */jsx(Image, {
      position: "absolute",
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      objectFit: "cover",
      ...(typeof shapeSize === "number" && !Number.isNaN(shapeSize) && {
        width: shapeSize,
        height: shapeSize
      }),
      ...imageProps,
      ref: forwardedRef,
      src: resolvedSrc,
      onError: () => {
        setStatus("error");
      },
      onLoad: () => {
        setStatus("loaded");
      }
    })
  });
});
AvatarImage.displayName = IMAGE_NAME;
const FALLBACK_NAME = "AvatarFallback";
const AvatarFallbackFrame = styled(YStack, {
  name: FALLBACK_NAME,
  position: "absolute",
  fullscreen: true,
  zIndex: 0
});
const AvatarFallback = AvatarFallbackFrame.styleable((props, forwardedRef) => {
  const {
    __scopeAvatar,
    delayMs,
    ...fallbackProps
  } = props;
  const context = useAvatarContext(FALLBACK_NAME, __scopeAvatar);
  const [canRender, setCanRender] = React.useState(delayMs === void 0);
  React.useEffect(() => {
    if (delayMs !== void 0) {
      const timerId = setTimeout(() => setCanRender(true), delayMs);
      return () => clearTimeout(timerId);
    }
  }, [delayMs]);
  return canRender && context.imageLoadingStatus !== "loaded" ? /* @__PURE__ */jsx(AvatarFallbackFrame, {
    ...fallbackProps,
    ref: forwardedRef
  }) : null;
});
AvatarFallback.displayName = FALLBACK_NAME;
const AvatarFrame = styled(Square, {
  name: AVATAR_NAME,
  position: "relative",
  overflow: "hidden"
});
const Avatar = withStaticProperties(React.forwardRef((props, forwardedRef) => {
  const {
    __scopeAvatar,
    size = "$true",
    ...avatarProps
  } = props;
  const [imageLoadingStatus, setImageLoadingStatus] = React.useState("idle");
  return /* @__PURE__ */jsx(AvatarProvider, {
    size,
    scope: __scopeAvatar,
    imageLoadingStatus,
    onImageLoadingStatusChange: setImageLoadingStatus,
    children: /* @__PURE__ */jsx(AvatarFrame, {
      size,
      ...avatarProps,
      ref: forwardedRef
    })
  });
}), {
  Image: AvatarImage,
  Fallback: AvatarFallback
});
Avatar.displayName = AVATAR_NAME;
export { Avatar, AvatarFallback, AvatarFallbackFrame, AvatarFrame, AvatarImage, createAvatarScope };
