// @tamagui/sheet/dist/jsx/createSheet.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createSheet_exports = {};
__export(createSheet_exports, {
  createSheet: () => createSheet
});
module.exports = __toCommonJS(createSheet_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_helpers = require("@tamagui/helpers");
var import_portal = require("@tamagui/portal");
var import_remove_scroll = require("@tamagui/remove-scroll");
var import_use_did_finish_ssr = require("@tamagui/use-did-finish-ssr");
var import_z_index_stack = require("@tamagui/z-index-stack");
var import_react = require("react");
var import_react_native = require("react-native");
var import_constants2 = require("./constants.native.js");
var import_nativeSheet = require("./nativeSheet.native.js");
var import_SheetContext = require("./SheetContext.native.js");
var import_SheetImplementationCustom = require("./SheetImplementationCustom.native.js");
var import_SheetScrollView = require("./SheetScrollView.native.js");
var import_useSheetController = require("./useSheetController.native.js");
var import_useSheetOffscreenSize = require("./useSheetOffscreenSize.native.js");
function createSheet(param) {
  var {
    Handle,
    Frame,
    Overlay
  } = param;
  var SheetHandle = Handle.styleable(function (param2, forwardedRef) {
    var {
      __scopeSheet,
      ...props
    } = param2;
    var context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_HANDLE_NAME, __scopeSheet);
    var composedRef = (0, import_compose_refs.useComposedRefs)(context.handleRef, forwardedRef);
    var wasDraggingRef = (0, import_react.useRef)(false);
    (0, import_react.useEffect)(function () {
      if (!context.scrollBridge) return;
      return context.scrollBridge.onParentDragging(function (isDragging) {
        if (isDragging) wasDraggingRef.current = true;
      });
    }, [context.scrollBridge]);
    if (context.onlyShowFrame) return null;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Handle, {
      ref: composedRef,
      onPressIn: function () {
        wasDraggingRef.current = false;
      },
      onPress: function () {
        if (wasDraggingRef.current) {
          wasDraggingRef.current = false;
          return;
        }
        var max = context.snapPoints.length + (context.dismissOnSnapToBottom ? -1 : 0);
        var nextPos = (context.position + 1) % max;
        context.setPosition(nextPos);
      },
      open: context.open,
      ...props
    });
  });
  var SheetOverlay = Overlay.styleable(function (propsIn, ref) {
    var {
      __scopeSheet,
      ...props
    } = propsIn;
    var context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_OVERLAY_NAME, __scopeSheet);
    var element = (0, import_react.useMemo)(function () {
      return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Overlay, {
        ...props,
        onPress: (0, import_helpers.composeEventHandlers)(props.onPress, context.dismissOnOverlayPress ? function () {
          context.setOpen(false);
        } : void 0)
      });
    }, [props.onPress, props.opacity, context.dismissOnOverlayPress]);
    (0, import_constants.useIsomorphicLayoutEffect)(function () {
      var _context_onOverlayComponent;
      (_context_onOverlayComponent = context.onOverlayComponent) === null || _context_onOverlayComponent === void 0 || _context_onOverlayComponent.call(context, element);
    }, [element]);
    if (context.onlyShowFrame) return null;
    return null;
  });
  var SheetFrame = Frame.styleable(function (param2, forwardedRef) {
    var {
      __scopeSheet,
      adjustPaddingForOffscreenContent,
      disableHideBottomOverflow,
      children,
      ...props
    } = param2;
    var context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_NAME, __scopeSheet);
    var {
      hasFit,
      disableRemoveScroll,
      frameSize,
      contentRef,
      open
    } = context;
    var composedContentRef = (0, import_compose_refs.useComposedRefs)(forwardedRef, contentRef);
    var offscreenSize = (0, import_useSheetOffscreenSize.useSheetOffscreenSize)(context);
    var stableFrameSize = (0, import_react.useRef)(frameSize);
    (0, import_react.useEffect)(function () {
      if (open && frameSize) stableFrameSize.current = frameSize;
    }, [open, frameSize]);
    var sheetContents = (0, import_react.useMemo)(function () {
      var shouldUseFixedHeight = hasFit && !open && stableFrameSize.current;
      return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(Frame, {
        ref: composedContentRef,
        flex: hasFit && open ? 0 : 1,
        flexBasis: hasFit ? "auto" : void 0,
        height: shouldUseFixedHeight ? stableFrameSize.current : hasFit ? void 0 : frameSize,
        pointerEvents: open ? "auto" : "none",
        "data-state": open ? "open" : "closed",
        ...props,
        children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_z_index_stack.StackZIndexContext, {
          zIndex: (0, import_portal.resolveViewZIndex)(props.zIndex),
          children
        }), adjustPaddingForOffscreenContent && /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.View, {
          "data-sheet-offscreen-pad": true,
          height: offscreenSize,
          width: "100%"
        })]
      });
    }, [open, props, frameSize, offscreenSize, adjustPaddingForOffscreenContent, hasFit]);
    return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
      children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_remove_scroll.RemoveScroll, {
        enabled: !disableRemoveScroll && context.open,
        children: sheetContents
      }), !disableHideBottomOverflow && /* @__PURE__ */(0, import_jsx_runtime.jsx)(Frame, {
        ...props,
        componentName: "SheetCover",
        children: null,
        testID: void 0,
        id: void 0,
        position: "absolute",
        bottom: "-100%",
        zIndex: -1,
        height: context.frameSize,
        left: 0,
        right: 0,
        borderWidth: 0,
        borderRadius: 0,
        shadowOpacity: 0
      })]
    });
  });
  var Sheet = /* @__PURE__ */(0, import_react.forwardRef)(function Sheet2(props, ref) {
    var hydrated = (0, import_use_did_finish_ssr.useDidFinishSSR)();
    var {
      isShowingNonSheet
    } = (0, import_useSheetController.useSheetController)();
    var SheetImplementation = import_SheetImplementationCustom.SheetImplementationCustom;
    if (props.native && import_react_native.Platform.OS === "ios") {
      var impl = (0, import_nativeSheet.getNativeSheet)("ios");
      if (impl) SheetImplementation = impl;
    }
    if (isShowingNonSheet || !hydrated) return null;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(SheetImplementation, {
      ref,
      ...props
    });
  });
  var components = {
    Frame: SheetFrame,
    Overlay: SheetOverlay,
    Handle: SheetHandle,
    ScrollView: import_SheetScrollView.SheetScrollView
  };
  var Controlled = (0, import_helpers.withStaticProperties)(Sheet, components);
  return (0, import_helpers.withStaticProperties)(Sheet, {
    ...components,
    Controlled
  });
}
