// @tamagui/sheet/dist/jsx/createSheet.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { View } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { resolveViewZIndex } from "@tamagui/portal";
import { RemoveScroll } from "@tamagui/remove-scroll";
import { useDidFinishSSR } from "@tamagui/use-did-finish-ssr";
import { StackZIndexContext } from "@tamagui/z-index-stack";
import { forwardRef, useMemo, useEffect, useRef } from "react";
import { Platform } from "react-native-web";
import { SHEET_HANDLE_NAME, SHEET_NAME, SHEET_OVERLAY_NAME } from "./constants.mjs";
import { getNativeSheet } from "./nativeSheet.mjs";
import { useSheetContext } from "./SheetContext.mjs";
import { SheetImplementationCustom } from "./SheetImplementationCustom.mjs";
import { SheetScrollView } from "./SheetScrollView.mjs";
import { useSheetController } from "./useSheetController.mjs";
import { useSheetOffscreenSize } from "./useSheetOffscreenSize.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function createSheet({
  Handle,
  Frame,
  Overlay
}) {
  const SheetHandle = Handle.styleable(({
    __scopeSheet,
    ...props
  }, forwardedRef) => {
    const context = useSheetContext(SHEET_HANDLE_NAME, __scopeSheet);
    const composedRef = useComposedRefs(context.handleRef, forwardedRef);
    const wasDraggingRef = useRef(false);
    useEffect(() => {
      if (!context.scrollBridge) return;
      return context.scrollBridge.onParentDragging(isDragging => {
        if (isDragging) wasDraggingRef.current = true;
      });
    }, [context.scrollBridge]);
    if (context.onlyShowFrame) return null;
    return /* @__PURE__ */jsx(Handle, {
      ref: composedRef,
      onPressIn: () => {
        wasDraggingRef.current = false;
      },
      onPress: () => {
        if (wasDraggingRef.current) {
          wasDraggingRef.current = false;
          return;
        }
        const max = context.snapPoints.length + (context.dismissOnSnapToBottom ? -1 : 0);
        const nextPos = (context.position + 1) % max;
        context.setPosition(nextPos);
      },
      open: context.open,
      ...props
    });
  });
  const SheetOverlay = Overlay.styleable((propsIn, ref) => {
    const {
      __scopeSheet,
      ...props
    } = propsIn;
    const context = useSheetContext(SHEET_OVERLAY_NAME, __scopeSheet);
    const element = useMemo(() => {
      return /* @__PURE__ */jsx(Overlay, {
        ...props,
        onPress: composeEventHandlers(props.onPress, context.dismissOnOverlayPress ? () => {
          context.setOpen(false);
        } : void 0)
      });
    }, [props.onPress, props.opacity, context.dismissOnOverlayPress]);
    useIsomorphicLayoutEffect(() => {
      context.onOverlayComponent?.(element);
    }, [element]);
    if (context.onlyShowFrame) return null;
    return null;
  });
  const SheetFrame = Frame.styleable(({
    __scopeSheet,
    adjustPaddingForOffscreenContent,
    disableHideBottomOverflow,
    children,
    ...props
  }, forwardedRef) => {
    const context = useSheetContext(SHEET_NAME, __scopeSheet);
    const {
      hasFit,
      disableRemoveScroll,
      frameSize,
      contentRef,
      open
    } = context;
    const composedContentRef = useComposedRefs(forwardedRef, contentRef);
    const offscreenSize = useSheetOffscreenSize(context);
    const stableFrameSize = useRef(frameSize);
    useEffect(() => {
      if (open && frameSize) stableFrameSize.current = frameSize;
    }, [open, frameSize]);
    const sheetContents = useMemo(() => {
      const shouldUseFixedHeight = hasFit && !open && stableFrameSize.current;
      return /* @__PURE__ */jsxs(Frame, {
        ref: composedContentRef,
        flex: hasFit && open ? 0 : 1,
        flexBasis: hasFit ? "auto" : void 0,
        height: shouldUseFixedHeight ? stableFrameSize.current : hasFit ? void 0 : frameSize,
        pointerEvents: open ? "auto" : "none",
        "data-state": open ? "open" : "closed",
        ...props,
        children: [/* @__PURE__ */jsx(StackZIndexContext, {
          zIndex: resolveViewZIndex(props.zIndex),
          children
        }), adjustPaddingForOffscreenContent && /* @__PURE__ */jsx(View, {
          "data-sheet-offscreen-pad": true,
          height: offscreenSize,
          width: "100%"
        })]
      });
    }, [open, props, frameSize, offscreenSize, adjustPaddingForOffscreenContent, hasFit]);
    return /* @__PURE__ */jsxs(Fragment, {
      children: [/* @__PURE__ */jsx(RemoveScroll, {
        enabled: !disableRemoveScroll && context.open,
        children: sheetContents
      }), !disableHideBottomOverflow && /* @__PURE__ */jsx(Frame, {
        ...props,
        componentName: "SheetCover",
        children: null,
        testID: void 0,
        id: void 0,
        position: "absolute",
        bottom: "-100%",
        zIndex: -1,
        height: context.frameSize,
        left: 0,
        right: 0,
        borderWidth: 0,
        borderRadius: 0,
        shadowOpacity: 0
      })]
    });
  });
  const Sheet = forwardRef(function Sheet2(props, ref) {
    const hydrated = useDidFinishSSR();
    const {
      isShowingNonSheet
    } = useSheetController();
    let SheetImplementation = SheetImplementationCustom;
    if (props.native && Platform.OS === "ios") {}
    if (isShowingNonSheet || !hydrated) return null;
    return /* @__PURE__ */jsx(SheetImplementation, {
      ref,
      ...props
    });
  });
  const components = {
    Frame: SheetFrame,
    Overlay: SheetOverlay,
    Handle: SheetHandle,
    ScrollView: SheetScrollView
  };
  const Controlled = withStaticProperties(Sheet, components);
  return withStaticProperties(Sheet, {
    ...components,
    Controlled
  });
}
export { createSheet };
