// @tamagui/sheet/dist/jsx/SheetImplementationCustom.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SheetImplementationCustom_exports = {};
__export(SheetImplementationCustom_exports, {
  SheetImplementationCustom: () => SheetImplementationCustom
});
module.exports = __toCommonJS(SheetImplementationCustom_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_adapt = require("@tamagui/adapt");
var import_animate_presence = require("@tamagui/animate-presence");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_native = require("@tamagui/native");
var import_portal = require("@tamagui/portal");
var import_react = __toESM(require("react"), 1);
var import_react_native = require("react-native");
var import_contexts = require("./contexts.native.js");
var import_GestureDetectorWrapper = require("./GestureDetectorWrapper.native.js");
var import_GestureSheetContext = require("./GestureSheetContext.native.js");
var import_helpers = require("./helpers.native.js");
var import_SheetContext = require("./SheetContext.native.js");
var import_useGestureHandlerPan = require("./useGestureHandlerPan.native.js");
var import_useKeyboardControllerSheet = require("./useKeyboardControllerSheet.native.js");
var import_useSheetOpenState = require("./useSheetOpenState.native.js");
var import_useSheetProviderProps = require("./useSheetProviderProps.native.js");
var hiddenSize = 10000.1;
var _cachedSafeAreaTop;
function getSafeAreaTopInset() {
  if (_cachedSafeAreaTop !== void 0) return _cachedSafeAreaTop;
  _cachedSafeAreaTop = (0, import_native.getSafeArea)().getInsets().top;
  return _cachedSafeAreaTop;
}
var relativeDimensionTo = import_constants.isWeb ? "window" : "screen";
var SheetImplementationCustom = /* @__PURE__ */import_react.default.forwardRef(function SheetImplementationCustom2(props, forwardedRef) {
  var parentSheet = import_react.default.useContext(import_contexts.ParentSheetContext);
  var {
    transition,
    transitionConfig: transitionConfigProp,
    modal = false,
    zIndex = parentSheet.zIndex + 1,
    moveOnKeyboardChange = false,
    unmountChildrenWhenHidden = false,
    portalProps,
    containerComponent: ContainerComponent = import_react.default.Fragment,
    onAnimationComplete
  } = props;
  var state = (0, import_useSheetOpenState.useSheetOpenState)(props);
  var [overlayComponent, setOverlayComponent] = import_react.default.useState(null);
  var providerProps = (0, import_useSheetProviderProps.useSheetProviderProps)(props, state, {
    onOverlayComponent: setOverlayComponent
  });
  var {
    frameSize,
    setFrameSize,
    snapPoints,
    snapPointsMode,
    hasFit,
    position,
    setPosition,
    scrollBridge,
    screenSize,
    setMaxContentSize,
    maxSnapPoint
  } = providerProps;
  var {
    open,
    controller,
    isHidden
  } = state;
  var openRef = import_react.default.useRef(open);
  openRef.current = open;
  var sheetRef = import_react.default.useRef(void 0);
  var ref = (0, import_compose_refs.useComposedRefs)(forwardedRef, sheetRef, providerProps.contentRef);
  var {
    animationDriver
  } = (0, import_core.useConfiguration)();
  if (!animationDriver) throw new Error(`Sheet requires an animation driver to be set`);
  var transitionConfig = function () {
    var _animationDriver_animations;
    if (transitionConfigProp) return transitionConfigProp;
    var [animationProp, animationPropConfig] = !transition ? [] : Array.isArray(transition) ? transition : [transition];
    if (animationProp && ((_animationDriver_animations = animationDriver.animations) === null || _animationDriver_animations === void 0 ? void 0 : _animationDriver_animations[animationProp])) return {
      ...animationDriver.animations[animationProp],
      ...animationPropConfig
    };
    return null;
  }();
  var [isShowingInnerSheet, setIsShowingInnerSheet] = import_react.default.useState(false);
  var shouldHideParentSheet = !import_constants.isWeb && modal && isShowingInnerSheet && (0, import_portal.needsPortalRepropagation)();
  var sheetInsideSheet = import_react.default.useContext(import_contexts.SheetInsideSheetContext);
  var onInnerSheet = import_react.default.useCallback(function (hasChild) {
    setIsShowingInnerSheet(hasChild);
  }, []);
  var stableFrameSize = import_react.default.useRef(frameSize);
  import_react.default.useEffect(function () {
    if (open && frameSize) stableFrameSize.current = frameSize;
  }, [open, frameSize]);
  var effectiveFrameSize = open ? frameSize : stableFrameSize.current || frameSize;
  var positions = import_react.default.useMemo(function () {
    return snapPoints.map(function (point) {
      return getYPositions(snapPointsMode, point, screenSize, effectiveFrameSize);
    });
  }, [screenSize, effectiveFrameSize, snapPoints, snapPointsMode]);
  var {
    keyboardHeight,
    isKeyboardVisible,
    dismissKeyboard,
    pauseKeyboardHandler,
    flushPendingHide
  } = (0, import_useKeyboardControllerSheet.useKeyboardControllerSheet)({
    enabled: !import_constants.isWeb && Boolean(moveOnKeyboardChange)
  });
  var [isDragging, setIsDragging_] = import_react.default.useState(false);
  var isDraggingRef = import_react.default.useRef(false);
  var setIsDragging = import_react.default.useCallback(function (val) {
    isDraggingRef.current = val;
    pauseKeyboardHandler.current = val;
    setIsDragging_(val);
    if (!val) flushPendingHide();
  }, [pauseKeyboardHandler, flushPendingHide]);
  var activePositionsRef = import_react.default.useRef(positions);
  var activePositions = import_react.default.useMemo(function () {
    if (isDragging || isDraggingRef.current) return activePositionsRef.current;
    var result;
    if (!isKeyboardVisible || keyboardHeight <= 0) result = positions;else {
      var safeAreaTop = import_constants.isWeb ? 0 : getSafeAreaTopInset();
      result = positions.map(function (p) {
        if (screenSize && p >= screenSize) return p;
        return Math.max(safeAreaTop, p - keyboardHeight);
      });
    }
    activePositionsRef.current = result;
    return result;
  }, [positions, isKeyboardVisible, keyboardHeight, screenSize, isDragging]);
  var {
    useAnimatedNumber,
    useAnimatedNumberStyle,
    useAnimatedNumberReaction
  } = animationDriver;
  var _animationDriver_View;
  var AnimatedView = (_animationDriver_View = animationDriver.View) !== null && _animationDriver_View !== void 0 ? _animationDriver_View : import_core.View;
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    if (!(sheetInsideSheet && open)) return;
    sheetInsideSheet(true);
    return function () {
      sheetInsideSheet(false);
    };
  }, [sheetInsideSheet, open]);
  var nextParentContext = import_react.default.useMemo(function () {
    return {
      zIndex
    };
  }, [zIndex]);
  var startPosition = (0, import_core.useDidFinishSSR)() && screenSize ? screenSize : hiddenSize;
  var animatedNumber = useAnimatedNumber(startPosition);
  var at = import_react.default.useRef(startPosition);
  var hasntMeasured = at.current === hiddenSize;
  var [disableAnimation, setDisableAnimation] = (0, import_react.useState)(hasntMeasured);
  var skipAdaptAnimation = import_react.default.useRef(false);
  if (controller === null || controller === void 0 ? void 0 : controller.skipNextAnimation) skipAdaptAnimation.current = true;
  var hasScrollView = import_react.default.useRef(false);
  var opacityFallbackTimer = import_react.default.useRef(null);
  useAnimatedNumberReaction({
    value: animatedNumber,
    hostRef: sheetRef
  }, import_react.default.useCallback(function (value) {
    at.current = value;
    scrollBridge.paneY = value;
    var minY = activePositions[0];
    var wasAtTop = scrollBridge.isAtTop;
    var nowAtTop = value <= minY + 5;
    if (wasAtTop !== nowAtTop) {
      scrollBridge.isAtTop = nowAtTop;
      if (nowAtTop) {
        var _scrollBridge_setScrollEnabled;
        if (scrollBridge.y > 0) {
          var _scrollBridge_forceScrollTo;
          (_scrollBridge_forceScrollTo = scrollBridge.forceScrollTo) === null || _scrollBridge_forceScrollTo === void 0 || _scrollBridge_forceScrollTo.call(scrollBridge, 0);
          scrollBridge.y = 0;
        }
        scrollBridge.scrollLockY = void 0;
        (_scrollBridge_setScrollEnabled = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled === void 0 || _scrollBridge_setScrollEnabled.call(scrollBridge, true);
      } else {
        var _scrollBridge_setScrollEnabled1;
        scrollBridge.scrollLockY = 0;
        (_scrollBridge_setScrollEnabled1 = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled1 === void 0 || _scrollBridge_setScrollEnabled1.call(scrollBridge, false);
      }
    }
  }, [animationDriver, activePositions]));
  function stopSpring() {
    animatedNumber.stop();
    if (scrollBridge.onFinishAnimate) {
      scrollBridge.onFinishAnimate();
      scrollBridge.onFinishAnimate = void 0;
    }
  }
  var animateTo = (0, import_core.useEvent)(function (position2, animationOverride) {
    if (frameSize === 0) return;
    var toValue = isHidden || position2 === -1 ? screenSize : activePositions[position2];
    if (at.current === toValue) return;
    at.current = toValue;
    stopSpring();
    var isOpenAnimation = position2 !== -1 && !isHidden;
    if (opacityFallbackTimer.current) {
      clearTimeout(opacityFallbackTimer.current);
      opacityFallbackTimer.current = null;
    }
    var animationCompleteCallback = function () {
      if (opacityFallbackTimer.current) {
        clearTimeout(opacityFallbackTimer.current);
        opacityFallbackTimer.current = null;
      }
      if (!isOpenAnimation && !openRef.current) setOpacity(0);
      onAnimationComplete === null || onAnimationComplete === void 0 || onAnimationComplete({
        open: isOpenAnimation
      });
    };
    if (!isOpenAnimation) opacityFallbackTimer.current = setTimeout(function () {
      opacityFallbackTimer.current = null;
      if (!openRef.current) setOpacity(0);
    }, 1e3);
    if (skipAdaptAnimation.current) {
      skipAdaptAnimation.current = false;
      animatedNumber.setValue(toValue, {
        type: "timing",
        duration: 0
      }, animationCompleteCallback);
      return;
    }
    animatedNumber.setValue(toValue, animationOverride || {
      type: "spring",
      ...transitionConfig
    }, animationCompleteCallback);
  });
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    if (hasntMeasured && screenSize && frameSize) {
      at.current = screenSize;
      animatedNumber.setValue(screenSize, {
        type: "timing",
        duration: 0
      }, function () {
        setTimeout(function () {
          setDisableAnimation(false);
        }, 10);
      });
      return;
    }
    if (disableAnimation) return;
    if (!frameSize || !screenSize || isHidden || hasntMeasured && !open) return;
    animateTo(position);
    if (position === -1) {
      scrollBridge.scrollLock = false;
      scrollBridge.scrollStartY = -1;
    }
    if (open && position >= 0) {
      var isTopPosition = position === 0;
      scrollBridge.isAtTop = isTopPosition;
      if (isTopPosition) {
        var _scrollBridge_setScrollEnabled;
        scrollBridge.scrollLockY = void 0;
        (_scrollBridge_setScrollEnabled = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled === void 0 || _scrollBridge_setScrollEnabled.call(scrollBridge, true);
      } else {
        var _scrollBridge_setScrollEnabled1;
        scrollBridge.scrollLockY = 0;
        (_scrollBridge_setScrollEnabled1 = scrollBridge.setScrollEnabled) === null || _scrollBridge_setScrollEnabled1 === void 0 || _scrollBridge_setScrollEnabled1.call(scrollBridge, false);
      }
    }
  }, [hasntMeasured, disableAnimation, isHidden, frameSize, screenSize, open, position]);
  var _props_disableDrag;
  var disableDrag = (_props_disableDrag = props.disableDrag) !== null && _props_disableDrag !== void 0 ? _props_disableDrag : controller === null || controller === void 0 ? void 0 : controller.disableDrag;
  (0, import_core.useThemeName)();
  var [blockPan, setBlockPan] = import_react.default.useState(false);
  var panResponder = import_react.default.useMemo(function () {
    if (disableDrag) return;
    if (!frameSize) return;
    if (isShowingInnerSheet) return;
    var minY = positions[0];
    scrollBridge.paneMinY = minY;
    var startY = at.current;
    function setPanning(val) {
      setIsDragging(val);
    }
    var release = function (param) {
      var {
        vy
      } = param;
      scrollBridge.setParentDragging(false);
      if (scrollBridge.scrollLock) return;
      isExternalDrag = false;
      previouslyScrolling = false;
      setPanning(false);
      var end = at.current + frameSize * vy * .2;
      var closestPoint = 0;
      var dist = Number.POSITIVE_INFINITY;
      for (var i = 0; i < positions.length; i++) {
        var position2 = positions[i];
        var curDist = end > position2 ? end - position2 : position2 - end;
        if (curDist < dist) {
          dist = curDist;
          closestPoint = i;
        }
      }
      setPosition(closestPoint);
      animateTo(closestPoint);
    };
    var finish = function (_e, state2) {
      release({
        vy: state2.vy,
        dragAt: state2.dy
      });
    };
    var previouslyScrolling = false;
    var onMoveShouldSet = function (e, param) {
      var {
        dy
      } = param;
      function getShouldSet() {
        if (e.target === providerProps.handleRef.current) return true;
        if (scrollBridge.hasScrollableContent === true) {
          if (scrollBridge.scrollLock) return false;
          var isScrolled = scrollBridge.y !== 0;
          var isDraggingUp = dy < 0;
          var isNearTop = scrollBridge.paneY - 5 <= scrollBridge.paneMinY;
          if (isScrolled) return false;
          if (isNearTop) {
            if (hasScrollView.current && isDraggingUp) return false;
          }
        }
        return Math.abs(dy) > 10;
      }
      var granted = getShouldSet();
      if (granted) scrollBridge.setParentDragging(true);
      return granted;
    };
    var grant = function () {
      setPanning(true);
      stopSpring();
      startY = at.current;
    };
    var isExternalDrag = false;
    scrollBridge.drag = function (dy) {
      if (!isExternalDrag) {
        isExternalDrag = true;
        grant();
      }
      var to = dy + startY;
      animatedNumber.setValue((0, import_helpers.resisted)(to, minY), {
        type: "direct"
      });
    };
    scrollBridge.release = release;
    scrollBridge.snapToPosition = function (positionIndex) {
      isExternalDrag = false;
      setPanning(false);
      setPosition(positionIndex);
      animateTo(positionIndex);
    };
    return import_react_native.PanResponder.create({
      onMoveShouldSetPanResponder: onMoveShouldSet,
      onPanResponderGrant: grant,
      onPanResponderMove: function (_e, param) {
        var {
          dy
        } = param;
        var toFull = dy + startY;
        var to = (0, import_helpers.resisted)(toFull, minY);
        if (to <= minY) scrollBridge.setParentDragging(false);else scrollBridge.setParentDragging(true);
        animatedNumber.setValue(to, {
          type: "direct"
        });
      },
      onPanResponderEnd: finish,
      onPanResponderTerminate: finish,
      onPanResponderRelease: finish
    });
  }, [disableDrag, isShowingInnerSheet, animateTo, frameSize, positions, setPosition]);
  import_react.default.useEffect(function () {
    if (isDragging || isHidden || !open || disableAnimation) return;
    if (!frameSize || !screenSize) return;
    animateTo(position, {
      type: "timing",
      duration: 250
    });
  }, [isKeyboardVisible, keyboardHeight]);
  var wasDragging = import_react.default.useRef(false);
  import_react.default.useEffect(function () {
    if (isDragging) {
      wasDragging.current = true;
      return;
    }
    if (!wasDragging.current) return;
    wasDragging.current = false;
    if (!frameSize || !screenSize || isHidden || !open) return;
    animateTo(position);
  }, [isDragging]);
  import_react.default.useEffect(function () {
    if (!open && isKeyboardVisible) dismissKeyboard();
  }, [open]);
  var {
    panGesture,
    panGestureRef,
    gestureHandlerEnabled
  } = (0, import_useGestureHandlerPan.useGestureHandlerPan)({
    positions: activePositions,
    frameSize,
    setPosition,
    animateTo,
    stopSpring,
    scrollBridge,
    setIsDragging,
    getCurrentPosition: function () {
      return at.current;
    },
    resisted: import_helpers.resisted,
    disableDrag,
    isShowingInnerSheet,
    setAnimatedPosition: function (val) {
      at.current = val;
      animatedNumber.setValue(val, {
        type: "direct"
      });
    },
    pauseKeyboardHandler
  });
  var handleAnimationViewLayout = (0, import_core.useEvent)(function (e) {
    var _e_nativeEvent;
    if (!open && stableFrameSize.current !== 0) return;
    var layoutHeight = (_e_nativeEvent = e.nativeEvent) === null || _e_nativeEvent === void 0 ? void 0 : _e_nativeEvent.layout.height;
    var next = modal ? Math.min(layoutHeight, import_react_native.Dimensions.get(relativeDimensionTo).height) : layoutHeight;
    if (!next) return;
    setFrameSize(next);
  });
  var handleMaxContentViewLayout = import_react.default.useCallback(function (e) {
    var _e_nativeEvent;
    var next = Math.min((_e_nativeEvent = e.nativeEvent) === null || _e_nativeEvent === void 0 ? void 0 : _e_nativeEvent.layout.height, import_react_native.Dimensions.get(relativeDimensionTo).height);
    if (!next) return;
    setMaxContentSize(next);
  }, []);
  var animatedStyle = useAnimatedNumberStyle(animatedNumber, import_react.default.useCallback(function (val) {
    "worklet";

    return {
      transform: [{
        translateY: frameSize === 0 ? hiddenSize : val
      }]
    };
  }, [frameSize]));
  var [opacity, setOpacity] = import_react.default.useState(open ? 1 : 0);
  if (open && opacity === 0) {
    setOpacity(1);
    if (opacityFallbackTimer.current) {
      clearTimeout(opacityFallbackTimer.current);
      opacityFallbackTimer.current = null;
    }
  }
  var forcedContentHeight = hasFit ? void 0 : snapPointsMode === "percent" ? `${maxSnapPoint}${import_constants.isWeb ? modal ? "dvh" : "%" : "%"}` : maxSnapPoint;
  var setHasScrollView = import_react.default.useCallback(function (val) {
    hasScrollView.current = val;
  }, []);
  var contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.LayoutMeasurementController, {
    disable: !open,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_contexts.ParentSheetContext.Provider, {
      value: nextParentContext,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_SheetContext.SheetProvider, {
        ...providerProps,
        setHasScrollView,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_GestureSheetContext.GestureSheetProvider, {
          isDragging,
          blockPan,
          setBlockPan,
          panGesture,
          panGestureRef,
          children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate_presence.AnimatePresence, {
            custom: {
              open
            },
            children: shouldHideParentSheet || !open ? null : overlayComponent
          }), snapPointsMode !== "percent" && /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_react_native.View, {
            style: {
              opacity: 0,
              position: "absolute",
              top: 0,
              left: 0,
              right: 0,
              bottom: 0,
              pointerEvents: "none"
            },
            onLayout: handleMaxContentViewLayout
          }), /* @__PURE__ */(0, import_jsx_runtime.jsx)(AnimatedView, {
            ref,
            onLayout: handleAnimationViewLayout,
            transition: isDragging || disableAnimation ? null : transition,
            disableClassName: true,
            style: [{
              position: "absolute",
              zIndex,
              width: "100%",
              height: forcedContentHeight,
              minHeight: forcedContentHeight,
              opacity: !shouldHideParentSheet ? opacity : 0,
              ...((shouldHideParentSheet || !open) && {
                pointerEvents: "none"
              })
            }, animatedStyle],
            children: gestureHandlerEnabled && panGesture ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_GestureDetectorWrapper.GestureDetectorWrapper, {
              gesture: panGesture,
              style: {
                flex: 1
              },
              children: props.children
            }) : /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_react_native.View, {
              ...(panResponder === null || panResponder === void 0 ? void 0 : panResponder.panHandlers),
              style: {
                flex: 1,
                width: "100%",
                height: "100%"
              },
              children: props.children
            })
          })]
        })
      })
    })
  });
  if ((0, import_portal.needsPortalRepropagation)()) {
    var adaptContext = (0, import_adapt.useAdaptContext)();
    contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_adapt.ProvideAdaptContext, {
      ...adaptContext,
      children: contents
    });
  }
  var shouldMountChildren = unmountChildrenWhenHidden ? !!opacity : true;
  if (modal) {
    var modalContents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_portal.Portal, {
      stackZIndex: zIndex,
      ...portalProps,
      children: shouldMountChildren && /* @__PURE__ */(0, import_jsx_runtime.jsx)(ContainerComponent, {
        children: contents
      })
    });
    if (import_constants.isWeb) return modalContents;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_contexts.SheetInsideSheetContext.Provider, {
      value: onInnerSheet,
      children: modalContents
    });
  }
  return contents;
});
function getYPositions(mode, point, screenSize, frameSize) {
  if (!screenSize || !frameSize) return 0;
  if (mode === "mixed") {
    if (typeof point === "number") return screenSize - Math.min(screenSize, Math.max(0, point));
    if (point === "fit") return screenSize - Math.min(screenSize, frameSize);
    if (point.endsWith("%")) {
      var pct = Math.min(100, Math.max(0, Number(point.slice(0, -1)))) / 100;
      if (Number.isNaN(pct)) {
        console.warn("Invalid snapPoint percentage string");
        return 0;
      }
      return Math.round(screenSize - pct * screenSize);
    }
    console.warn("Invalid snapPoint unknown value");
    return 0;
  }
  if (mode === "fit") {
    if (point === 0) return screenSize;
    return screenSize - Math.min(screenSize, frameSize);
  }
  if (mode === "constant" && typeof point === "number") return screenSize - Math.min(screenSize, Math.max(0, point));
  var pct1 = Math.min(100, Math.max(0, Number(point))) / 100;
  if (Number.isNaN(pct1)) {
    console.warn("Invalid snapPoint percentage");
    return 0;
  }
  return Math.round(screenSize - pct1 * screenSize);
}
