// @tamagui/sheet/dist/jsx/SheetScrollView.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SheetScrollView_exports = {};
__export(SheetScrollView_exports, {
  SheetScrollView: () => SheetScrollView
});
module.exports = __toCommonJS(SheetScrollView_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_compose_refs = require("@tamagui/compose-refs");
var import_core = require("@tamagui/core");
var import_scroll_view = require("@tamagui/scroll-view");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var import_react = __toESM(require("react"), 1);
var import_GestureSheetContext = require("./GestureSheetContext.native.js");
var import_gestureState = require("./gestureState.native.js");
var import_SheetContext = require("./SheetContext.native.js");
var import_useSheetScrollViewGestures = require("./useSheetScrollViewGestures.native.js");
var SHEET_SCROLL_VIEW_NAME = "SheetScrollView";
var SheetScrollView = /* @__PURE__ */import_react.default.forwardRef(function (param, ref) {
  var {
    __scopeSheet,
    children,
    onScroll,
    scrollEnabled: scrollEnabledProp,
    ...props
  } = param;
  var context = (0, import_SheetContext.useSheetContext)(SHEET_SCROLL_VIEW_NAME, __scopeSheet);
  var gestureContext = (0, import_GestureSheetContext.useGestureSheetContext)();
  var {
    scrollBridge,
    setHasScrollView
  } = context;
  var [scrollEnabled] = (0, import_use_controllable_state.useControllableState)({
    prop: scrollEnabledProp,
    defaultProp: true
  });
  var scrollRef = import_react.default.useRef(null);
  var panGestureRef = gestureContext === null || gestureContext === void 0 ? void 0 : gestureContext.panGestureRef;
  var {
    ScrollView: RNGHScrollView
  } = (0, import_gestureState.getGestureHandlerState)();
  var useRNGHScrollView = (0, import_gestureState.isGestureHandlerEnabled)() && RNGHScrollView && panGestureRef;
  var currentScrollOffset = (0, import_react.useRef)(0);
  var lockedScrollY = (0, import_react.useRef)(0);
  var setScrollEnabled = function (next, lockTo) {
    if (!next) {
      var _scrollRef_current_scrollTo, _scrollRef_current;
      var lockY = lockTo !== null && lockTo !== void 0 ? lockTo : currentScrollOffset.current;
      lockedScrollY.current = lockY;
      scrollBridge.scrollLockY = lockY;
      (_scrollRef_current = scrollRef.current) === null || _scrollRef_current === void 0 ? void 0 : (_scrollRef_current_scrollTo = _scrollRef_current.scrollTo) === null || _scrollRef_current_scrollTo === void 0 ? void 0 : _scrollRef_current_scrollTo.call(_scrollRef_current, {
        x: 0,
        y: lockY,
        animated: false
      });
    } else {
      lockedScrollY.current = currentScrollOffset.current;
      scrollBridge.scrollLockY = void 0;
    }
  };
  var forceScrollTo = function (y) {
    var _scrollRef_current_scrollTo, _scrollRef_current;
    (_scrollRef_current = scrollRef.current) === null || _scrollRef_current === void 0 ? void 0 : (_scrollRef_current_scrollTo = _scrollRef_current.scrollTo) === null || _scrollRef_current_scrollTo === void 0 ? void 0 : _scrollRef_current_scrollTo.call(_scrollRef_current, {
      x: 0,
      y,
      animated: false
    });
  };
  (0, import_react.useEffect)(function () {
    setHasScrollView(true);
    if ((0, import_gestureState.isGestureHandlerEnabled)()) {
      scrollBridge.setScrollEnabled = setScrollEnabled;
      scrollBridge.forceScrollTo = forceScrollTo;
    }
    return function () {
      setHasScrollView(false);
      scrollBridge.setScrollEnabled = void 0;
      scrollBridge.forceScrollTo = void 0;
    };
  }, []);
  var [hasScrollableContent, setHasScrollableContent] = (0, import_react.useState)(true);
  var parentHeight = (0, import_react.useRef)(0);
  var contentHeight = (0, import_react.useRef)(0);
  var updateScrollable = function () {
    if (parentHeight.current && contentHeight.current) {
      setHasScrollableContent(contentHeight.current > parentHeight.current);
    }
  };
  (0, import_react.useEffect)(function () {
    scrollBridge.hasScrollableContent = hasScrollableContent;
  }, [hasScrollableContent]);
  var gestureProps = (0, import_useSheetScrollViewGestures.useSheetScrollViewGestures)({
    scrollRef,
    scrollBridge,
    hasScrollableContent,
    scrollEnabled,
    setScrollEnabled
  });
  var contentWrapper = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.View, {
    onLayout: function (e) {
      var height = Math.floor(e.nativeEvent.layout.height);
      if (height !== contentHeight.current) {
        contentHeight.current = height;
        updateScrollable();
      }
    },
    children
  });
  if (useRNGHScrollView && RNGHScrollView && panGestureRef) {
    var RNGHComponent = RNGHScrollView;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(RNGHComponent, {
      ref: (0, import_compose_refs.composeRefs)(scrollRef, ref),
      style: {
        flex: 1
      },
      scrollEventThrottle: 1,
      scrollEnabled,
      simultaneousHandlers: [panGestureRef],
      onLayout: function (e) {
        parentHeight.current = Math.ceil(e.nativeEvent.layout.height);
        updateScrollable();
      },
      onScroll: function (e) {
        var {
          y
        } = e.nativeEvent.contentOffset;
        currentScrollOffset.current = y;
        if (scrollBridge.scrollLockY !== void 0) {
          if (y !== scrollBridge.scrollLockY) {
            var _scrollRef_current_scrollTo, _scrollRef_current;
            (_scrollRef_current = scrollRef.current) === null || _scrollRef_current === void 0 ? void 0 : (_scrollRef_current_scrollTo = _scrollRef_current.scrollTo) === null || _scrollRef_current_scrollTo === void 0 ? void 0 : _scrollRef_current_scrollTo.call(_scrollRef_current, {
              x: 0,
              y: scrollBridge.scrollLockY,
              animated: false
            });
          }
          scrollBridge.y = scrollBridge.scrollLockY;
          onScroll === null || onScroll === void 0 ? void 0 : onScroll({
            ...e,
            nativeEvent: {
              ...e.nativeEvent,
              contentOffset: {
                ...e.nativeEvent.contentOffset,
                y: scrollBridge.scrollLockY
              }
            }
          });
          return;
        }
        scrollBridge.y = y;
        if (y > 0) scrollBridge.scrollStartY = -1;
        onScroll === null || onScroll === void 0 ? void 0 : onScroll(e);
      },
      contentContainerStyle: {
        minHeight: "100%"
      },
      bounces: false,
      keyboardShouldPersistTaps: "always",
      keyboardDismissMode: "none",
      ...props,
      children: contentWrapper
    });
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_scroll_view.ScrollView, {
    onLayout: function (e) {
      parentHeight.current = Math.ceil(e.nativeEvent.layout.height);
      updateScrollable();
    },
    ref: (0, import_compose_refs.composeRefs)(scrollRef, ref),
    flex: 1,
    scrollEventThrottle: 1,
    className: "_ovs-contain",
    scrollEnabled,
    onScroll: function (e) {
      var {
        y
      } = e.nativeEvent.contentOffset;
      scrollBridge.y = y;
      if (y > 0) scrollBridge.scrollStartY = -1;
      onScroll === null || onScroll === void 0 ? void 0 : onScroll(e);
    },
    contentContainerStyle: {
      minHeight: "100%"
    },
    ...gestureProps,
    ...props,
    children: contentWrapper
  });
});
