// @tamagui/sheet/dist/jsx/SheetScrollView.mjs
import { composeRefs } from "@tamagui/compose-refs";
import { View } from "@tamagui/core";
import { ScrollView } from "@tamagui/scroll-view";
import { useControllableState } from "@tamagui/use-controllable-state";
import React, { useEffect, useRef, useState } from "react";
import { useGestureSheetContext } from "./GestureSheetContext.mjs";
import { getGestureHandlerState, isGestureHandlerEnabled } from "./gestureState.mjs";
import { useSheetContext } from "./SheetContext.mjs";
import { useSheetScrollViewGestures } from "./useSheetScrollViewGestures.mjs";
import { jsx } from "react/jsx-runtime";
const SHEET_SCROLL_VIEW_NAME = "SheetScrollView";
const SheetScrollView = React.forwardRef(({
  __scopeSheet,
  children,
  onScroll,
  scrollEnabled: scrollEnabledProp,
  ...props
}, ref) => {
  const context = useSheetContext(SHEET_SCROLL_VIEW_NAME, __scopeSheet);
  const gestureContext = useGestureSheetContext();
  const {
    scrollBridge,
    setHasScrollView
  } = context;
  const [scrollEnabled] = useControllableState({
    prop: scrollEnabledProp,
    defaultProp: true
  });
  const scrollRef = React.useRef(null);
  const panGestureRef = gestureContext?.panGestureRef;
  const {
    ScrollView: RNGHScrollView
  } = getGestureHandlerState();
  const useRNGHScrollView = isGestureHandlerEnabled() && RNGHScrollView && panGestureRef;
  const currentScrollOffset = useRef(0);
  const lockedScrollY = useRef(0);
  const setScrollEnabled = (next, lockTo) => {
    if (!next) {
      const lockY = lockTo ?? currentScrollOffset.current;
      lockedScrollY.current = lockY;
      scrollBridge.scrollLockY = lockY;
      scrollRef.current?.scrollTo?.({
        x: 0,
        y: lockY,
        animated: false
      });
    } else {
      lockedScrollY.current = currentScrollOffset.current;
      scrollBridge.scrollLockY = void 0;
    }
  };
  const forceScrollTo = y => {
    scrollRef.current?.scrollTo?.({
      x: 0,
      y,
      animated: false
    });
  };
  useEffect(() => {
    setHasScrollView(true);
    if (isGestureHandlerEnabled()) {
      scrollBridge.setScrollEnabled = setScrollEnabled;
      scrollBridge.forceScrollTo = forceScrollTo;
    }
    return () => {
      setHasScrollView(false);
      scrollBridge.setScrollEnabled = void 0;
      scrollBridge.forceScrollTo = void 0;
    };
  }, []);
  const [hasScrollableContent, setHasScrollableContent] = useState(true);
  const parentHeight = useRef(0);
  const contentHeight = useRef(0);
  const updateScrollable = () => {
    if (parentHeight.current && contentHeight.current) {
      setHasScrollableContent(contentHeight.current > parentHeight.current);
    }
  };
  useEffect(() => {
    scrollBridge.hasScrollableContent = hasScrollableContent;
  }, [hasScrollableContent]);
  const gestureProps = useSheetScrollViewGestures({
    scrollRef,
    scrollBridge,
    hasScrollableContent,
    scrollEnabled,
    setScrollEnabled
  });
  const contentWrapper = /* @__PURE__ */jsx(View, {
    onLayout: e => {
      const height = Math.floor(e.nativeEvent.layout.height);
      if (height !== contentHeight.current) {
        contentHeight.current = height;
        updateScrollable();
      }
    },
    children
  });
  if (useRNGHScrollView && RNGHScrollView && panGestureRef) {
    const RNGHComponent = RNGHScrollView;
    return /* @__PURE__ */jsx(RNGHComponent, {
      ref: composeRefs(scrollRef, ref),
      style: {
        flex: 1
      },
      scrollEventThrottle: 1,
      scrollEnabled,
      simultaneousHandlers: [panGestureRef],
      onLayout: e => {
        parentHeight.current = Math.ceil(e.nativeEvent.layout.height);
        updateScrollable();
      },
      onScroll: e => {
        const {
          y
        } = e.nativeEvent.contentOffset;
        currentScrollOffset.current = y;
        if (scrollBridge.scrollLockY !== void 0) {
          if (y !== scrollBridge.scrollLockY) {
            scrollRef.current?.scrollTo?.({
              x: 0,
              y: scrollBridge.scrollLockY,
              animated: false
            });
          }
          scrollBridge.y = scrollBridge.scrollLockY;
          onScroll?.({
            ...e,
            nativeEvent: {
              ...e.nativeEvent,
              contentOffset: {
                ...e.nativeEvent.contentOffset,
                y: scrollBridge.scrollLockY
              }
            }
          });
          return;
        }
        scrollBridge.y = y;
        if (y > 0) scrollBridge.scrollStartY = -1;
        onScroll?.(e);
      },
      contentContainerStyle: {
        minHeight: "100%"
      },
      bounces: false,
      keyboardShouldPersistTaps: "always",
      keyboardDismissMode: "none",
      ...props,
      children: contentWrapper
    });
  }
  return /* @__PURE__ */jsx(ScrollView, {
    onLayout: e => {
      parentHeight.current = Math.ceil(e.nativeEvent.layout.height);
      updateScrollable();
    },
    ref: composeRefs(scrollRef, ref),
    flex: 1,
    scrollEventThrottle: 1,
    className: "_ovs-contain",
    scrollEnabled,
    onScroll: e => {
      const {
        y
      } = e.nativeEvent.contentOffset;
      scrollBridge.y = y;
      if (y > 0) scrollBridge.scrollStartY = -1;
      onScroll?.(e);
    },
    contentContainerStyle: {
      minHeight: "100%"
    },
    ...gestureProps,
    ...props,
    children: contentWrapper
  });
});
export { SheetScrollView };
