// @tamagui/sheet/dist/jsx/useGestureHandlerPan.mjs
import { useCallback, useMemo, useRef } from "react";
import { getGestureHandlerState, isGestureHandlerEnabled } from "./gestureState.mjs";
const AT_TOP_THRESHOLD = 5;
function useGestureHandlerPan(config) {
  const {
    positions,
    frameSize,
    setPosition,
    animateTo,
    stopSpring,
    scrollBridge,
    setIsDragging,
    getCurrentPosition,
    resisted,
    disableDrag,
    isShowingInnerSheet,
    setAnimatedPosition,
    scrollGestureRef
  } = config;
  const gestureHandlerEnabled = isGestureHandlerEnabled();
  const panGestureRef = useRef(null);
  const gestureStateRef = useRef({
    startY: 0,
    // track last translation when pan was active (for position calculation after handoff)
    lastPanTranslationY: 0,
    // accumulated position offset from all pan movements
    accumulatedOffset: 0,
    // track previous translation for direction detection (like actions-sheet)
    prevTranslationY: 0,
    // track if scroll was engaged (scrollY > 0) at some point
    scrollEngaged: false,
    // positions frozen at gesture start — keyboard may dismiss during drag (input blur),
    // causing positions to revert. Frozen positions ensure stable snap calculation.
    frozenPositions: [],
    frozenMinY: 0,
    // whether pan gesture actually started (vs just a tap in onBegin)
    panStarted: false
  });
  const onStart = useCallback(() => {
    stopSpring();
  }, [stopSpring]);
  const onEnd = useCallback((closestPoint, animationOverride) => {
    setIsDragging(false);
    scrollBridge.setParentDragging(false);
    scrollBridge.setScrollEnabled?.(true);
    setPosition(closestPoint);
    animateTo(closestPoint, animationOverride);
  }, [setIsDragging, scrollBridge, setPosition, animateTo]);
  const panGesture = useMemo(() => {
    if (!gestureHandlerEnabled || disableDrag || isShowingInnerSheet || !frameSize) {
      return null;
    }
    const {
      Gesture
    } = getGestureHandlerState();
    if (!Gesture) {
      return null;
    }
    const minY = positions[0];
    const gs = gestureStateRef.current;
    const gesture = Gesture.Pan().withRef(panGestureRef).activeOffsetY([-10, 10]).failOffsetX([-20, 20]).shouldCancelWhenOutside(false).onBegin(() => {
      gs.panStarted = false;
      if (config.pauseKeyboardHandler) {
        config.pauseKeyboardHandler.current = true;
      }
      const pos = getCurrentPosition();
      const atTop = pos <= minY + AT_TOP_THRESHOLD;
      const currentScrollY = scrollBridge.y;
      gs.startY = pos;
      gs.lastPanTranslationY = 0;
      gs.accumulatedOffset = 0;
      gs.prevTranslationY = 0;
      gs.scrollEngaged = currentScrollY > 0;
      gs.frozenPositions = [...positions];
      gs.frozenMinY = minY;
      if (!atTop) {
        scrollBridge.setScrollEnabled?.(false, 0);
      }
    }).onStart(() => {
      gs.panStarted = true;
      setIsDragging(true);
      scrollBridge.initialPosition = gs.startY;
      onStart();
    }).onChange(event => {
      const {
        translationY
      } = event;
      const isSwipingDown = gs.prevTranslationY < translationY;
      const deltaY = translationY - gs.prevTranslationY;
      gs.prevTranslationY = translationY;
      const scrollY = scrollBridge.y;
      if (scrollY > 0) {
        gs.scrollEngaged = true;
      }
      const currentPos = gs.startY + gs.accumulatedOffset;
      const isCurrentlyAtTop = currentPos <= minY + AT_TOP_THRESHOLD;
      const nodeIsScrolling = scrollY > 0;
      let panHandles = false;
      const hasScrollableContent = scrollBridge.hasScrollableContent !== false;
      if (!isCurrentlyAtTop) {
        if (isSwipingDown) {
          panHandles = !nodeIsScrolling || !hasScrollableContent;
        } else {
          panHandles = true;
        }
      } else {
        if (isSwipingDown) {
          if (nodeIsScrolling && hasScrollableContent) {
            panHandles = false;
          } else if (gs.scrollEngaged && hasScrollableContent) {
            panHandles = true;
          } else {
            panHandles = true;
          }
        } else {
          if (hasScrollableContent) {
            panHandles = false;
          } else {
            panHandles = true;
          }
        }
      }
      if (panHandles) {
        const lockTo = isCurrentlyAtTop && isSwipingDown && gs.scrollEngaged ? void 0 : 0;
        scrollBridge.setScrollEnabled?.(false, lockTo);
        gs.accumulatedOffset += deltaY;
        const newPosition = resisted(gs.startY + gs.accumulatedOffset, minY);
        scrollBridge.paneY = newPosition;
        setAnimatedPosition(newPosition);
        scrollBridge.setParentDragging(newPosition > minY);
      } else {
        scrollBridge.setScrollEnabled?.(true);
      }
    }).onEnd(event => {
      const {
        velocityY
      } = event;
      const currentPos = gs.startY + gs.accumulatedOffset;
      scrollBridge.scrollLockY = void 0;
      const snapPositions = gs.frozenPositions.length > 0 ? gs.frozenPositions : positions;
      const snapMinY = gs.frozenPositions.length > 0 ? gs.frozenMinY : minY;
      if (currentPos <= snapMinY + AT_TOP_THRESHOLD && scrollBridge.y > 0) {
        onEnd(0);
        return;
      }
      const velocity = velocityY / 1e3;
      const projectedEnd = currentPos + frameSize * velocity * 0.2;
      let closestPoint = 0;
      let minDist = Number.POSITIVE_INFINITY;
      for (let i = 0; i < snapPositions.length; i++) {
        const pos = snapPositions[i];
        const dist = Math.abs(projectedEnd - pos);
        if (dist < minDist) {
          minDist = dist;
          closestPoint = i;
        }
      }
      onEnd(closestPoint);
    }).onFinalize(() => {
      scrollBridge.scrollLockY = void 0;
      if (gs.panStarted) {
        setIsDragging(false);
      } else {
        if (config.pauseKeyboardHandler) {
          config.pauseKeyboardHandler.current = false;
        }
      }
    }).runOnJS(true);
    if (scrollGestureRef?.current) {
      return gesture.simultaneousWithExternalGesture(scrollGestureRef.current);
    }
    return gesture;
  }, [gestureHandlerEnabled, disableDrag, isShowingInnerSheet, frameSize, positions, scrollBridge, getCurrentPosition, resisted, onStart, onEnd, setIsDragging, setAnimatedPosition]);
  return {
    panGesture,
    panGestureRef,
    gestureHandlerEnabled
  };
}
export { useGestureHandlerPan };
