// @tamagui/sheet/dist/jsx/useSheetProviderProps.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useSheetProviderProps_exports = {};
__export(useSheetProviderProps_exports, {
  useSheetProviderProps: () => useSheetProviderProps
});
module.exports = __toCommonJS(useSheetProviderProps_exports);
var import_react = __toESM(require("react"), 1);
var import_core = require("@tamagui/core");
var import_use_constant = require("@tamagui/use-constant");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
function useSheetProviderProps(props, state) {
  var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
  var handleRef = import_react.default.useRef(null);
  var contentRef = import_react.default.useRef(null);
  var [frameSize, setFrameSize] = import_react.default.useState(0);
  var [maxContentSize, setMaxContentSize] = import_react.default.useState(0);
  var _props_snapPointsMode;
  var snapPointsMode = (_props_snapPointsMode = props.snapPointsMode) !== null && _props_snapPointsMode !== void 0 ? _props_snapPointsMode : "percent";
  var _props_snapPoints;
  var snapPointsProp = (_props_snapPoints = props.snapPoints) !== null && _props_snapPoints !== void 0 ? _props_snapPoints : snapPointsMode === "percent" ? [80] : snapPointsMode === "constant" ? [256] : ["fit"];
  var hasFit = snapPointsProp[0] === "fit";
  var snapPoints = import_react.default.useMemo(function () {
    return props.dismissOnSnapToBottom ? [...snapPointsProp, 0] : snapPointsProp;
  }, [JSON.stringify(snapPointsProp), props.dismissOnSnapToBottom]);
  var [position_, setPositionImmediate] = (0, import_use_controllable_state.useControllableState)({
    prop: props.position,
    defaultProp: props.defaultPosition || (state.open ? 0 : -1),
    onChange: props.onPositionChange,
    strategy: "most-recent-wins"
  });
  var position = state.open === false ? -1 : position_;
  var {
    open
  } = state;
  var setPosition = import_react.default.useCallback(function (next) {
    if (props.dismissOnSnapToBottom && next === snapPoints.length - 1) {
      state.setOpen(false);
    } else {
      setPositionImmediate(next);
    }
  }, [props.dismissOnSnapToBottom, snapPoints.length, setPositionImmediate, state.setOpen]);
  if (process.env.NODE_ENV === "development") {
    if (snapPointsMode === "mixed" && snapPoints.some(function (p) {
      if (typeof p === "string") {
        if (p === "fit") {
          return false;
        }
        if (p.endsWith("%")) {
          var n = Number(p.slice(0, -1));
          return n < 0 || n > 100;
        }
        return true;
      }
      return typeof p !== "number" || p < 0;
    })) {
      console.warn('\u26A0\uFE0F Invalid snapPoint given, snapPoints must be positive numeric values, string percentages between 0-100%, or "fit" when snapPointsMode is mixed');
    }
    if (snapPointsMode === "mixed" && snapPoints.indexOf("fit") > 0) {
      console.warn('\u26A0\uFE0F Invalid snapPoint given, "fit" must be the first/largest snap point when snapPointsMode is mixed');
    }
    if (snapPointsMode === "fit" && (snapPoints.length !== (props.dismissOnSnapToBottom ? 2 : 1) || snapPoints[0] !== "fit")) {
      console.warn("\u26A0\uFE0F Invalid snapPoint given, there are no snap points when snapPointsMode is fit");
    }
    if (snapPointsMode === "constant" && snapPoints.some(function (p) {
      return typeof p !== "number" || p < 0;
    })) {
      console.warn("\u26A0\uFE0F Invalid snapPoint given, snapPoints must be positive numeric values when snapPointsMode is constant");
    }
    if (snapPointsMode === "percent" && snapPoints.some(function (p) {
      return typeof p !== "number" || p < 0 || p > 100;
    })) {
      console.warn("\u26A0\uFE0F Invalid snapPoint given, snapPoints must be numeric values between 0 and 100 when snapPointsMode is percent");
    }
  }
  if (open && props.dismissOnSnapToBottom && position === snapPoints.length - 1) {
    setPositionImmediate(0);
  }
  var shouldSetPositionOpen = open && position < 0;
  import_react.default.useEffect(function () {
    if (shouldSetPositionOpen) {
      setPosition(0);
    }
  }, [setPosition, shouldSetPositionOpen]);
  var {
    animationDriver
  } = (0, import_core.useConfiguration)();
  if (!animationDriver) {
    throw new Error(process.env.NODE_ENV === "production" ? `\u274C 008` : "Must set animations in tamagui.config.ts");
  }
  var scrollBridge = (0, import_use_constant.useConstant)(function () {
    var parentDragListeners = /* @__PURE__ */new Set();
    var bridge = {
      hasScrollableContent: false,
      enabled: false,
      y: 0,
      paneY: 0,
      paneMinY: 0,
      scrollStartY: -1,
      drag: function () {},
      release: function () {},
      scrollLock: false,
      isParentDragging: false,
      onParentDragging: function (cb) {
        parentDragListeners.add(cb);
        return function () {
          parentDragListeners.delete(cb);
        };
      },
      setParentDragging: function (val) {
        if (val !== bridge.isParentDragging) {
          bridge.isParentDragging = val;
          parentDragListeners.forEach(function (cb) {
            return cb(val);
          });
        }
      }
    };
    return bridge;
  });
  var disableRemoveScroll;
  if (props.disableRemoveScroll !== void 0) {
    disableRemoveScroll = props.disableRemoveScroll || !open || !props.modal;
  } else if (props.forceRemoveScrollEnabled !== void 0) {
    if (process.env.NODE_ENV === "development") {
      console.warn("[Sheet] forceRemoveScrollEnabled is deprecated. Use disableRemoveScroll instead.");
    }
    disableRemoveScroll = !props.forceRemoveScrollEnabled;
  } else {
    disableRemoveScroll = !open || !props.modal;
  }
  var maxSnapPoint = snapPoints[0];
  var screenSize = snapPointsMode === "percent" ? frameSize / ((typeof maxSnapPoint === "number" ? maxSnapPoint : 100) / 100) : maxContentSize;
  var _props_dismissOnOverlayPress, _props_dismissOnSnapToBottom;
  var providerProps = {
    screenSize,
    maxSnapPoint,
    disableRemoveScroll,
    scrollBridge,
    modal: !!props.modal,
    open: state.open,
    setOpen: state.setOpen,
    hidden: !!state.isHidden,
    contentRef,
    handleRef,
    frameSize,
    setFrameSize,
    dismissOnOverlayPress: (_props_dismissOnOverlayPress = props.dismissOnOverlayPress) !== null && _props_dismissOnOverlayPress !== void 0 ? _props_dismissOnOverlayPress : true,
    dismissOnSnapToBottom: (_props_dismissOnSnapToBottom = props.dismissOnSnapToBottom) !== null && _props_dismissOnSnapToBottom !== void 0 ? _props_dismissOnSnapToBottom : false,
    onOverlayComponent: options.onOverlayComponent,
    scope: props.__scopeSheet,
    hasFit,
    position,
    snapPoints,
    snapPointsMode,
    setMaxContentSize,
    setPosition,
    setPositionImmediate,
    onlyShowFrame: false
  };
  return providerProps;
}
