// @tamagui/remove-scroll/dist/esm/useDisableScroll.native.js
import { useEffect, useRef } from "react";
var canUseDOM = function () {
  return typeof window !== "undefined" && !!window.document && !!window.document.createElement;
};
var refCount = 0;
var savedScrollY = 0;
var previousStyles = null;
function isIOSSafari() {
  if (typeof navigator === "undefined") return false;
  var ua = navigator.userAgent;
  var isIOS = /iPad|iPhone|iPod/.test(ua) || navigator.platform === "MacIntel" && navigator.maxTouchPoints > 1;
  var isSafari = /^((?!chrome|android).)*safari/i.test(ua);
  return isIOS && isSafari;
}
var useDisableBodyScroll = function (enabled) {
  var wasEnabled = useRef(false);
  useEffect(function () {
    if (!canUseDOM()) {
      return;
    }
    if (enabled && !wasEnabled.current) {
      wasEnabled.current = true;
      if (++refCount === 1) {
        var html = document.documentElement;
        var body = document.body;
        savedScrollY = window.scrollY;
        previousStyles = {
          htmlOverflow: html.style.overflow,
          htmlScrollbarGutter: html.style.scrollbarGutter,
          bodyPosition: body.style.position,
          bodyTop: body.style.top,
          bodyWidth: body.style.width,
          bodyOverflow: body.style.overflow,
          bodyOverscrollBehavior: body.style.overscrollBehavior
        };
        html.style.scrollbarGutter = "stable";
        html.style.overflow = "hidden";
        body.style.overscrollBehavior = "none";
        if (isIOSSafari()) {
          body.style.position = "fixed";
          body.style.top = `-${savedScrollY}px`;
          body.style.width = "100%";
          body.style.overflow = "hidden";
        }
      }
    } else if (!enabled && wasEnabled.current) {
      wasEnabled.current = false;
      if (--refCount === 0 && previousStyles) {
        var html1 = document.documentElement;
        var body1 = document.body;
        html1.style.overflow = previousStyles.htmlOverflow;
        html1.style.scrollbarGutter = previousStyles.htmlScrollbarGutter;
        body1.style.position = previousStyles.bodyPosition;
        body1.style.top = previousStyles.bodyTop;
        body1.style.width = previousStyles.bodyWidth;
        body1.style.overflow = previousStyles.bodyOverflow;
        body1.style.overscrollBehavior = previousStyles.bodyOverscrollBehavior;
        if (savedScrollY > 0) {
          window.scrollTo(0, savedScrollY);
        }
        previousStyles = null;
        savedScrollY = 0;
      }
    }
  }, [enabled]);
  useEffect(function () {
    return function () {
      if (wasEnabled.current) {
        wasEnabled.current = false;
        if (--refCount === 0 && previousStyles) {
          var html = document.documentElement;
          var body = document.body;
          html.style.overflow = previousStyles.htmlOverflow;
          html.style.scrollbarGutter = previousStyles.htmlScrollbarGutter;
          body.style.position = previousStyles.bodyPosition;
          body.style.top = previousStyles.bodyTop;
          body.style.width = previousStyles.bodyWidth;
          body.style.overflow = previousStyles.bodyOverflow;
          body.style.overscrollBehavior = previousStyles.bodyOverscrollBehavior;
          if (savedScrollY > 0) {
            window.scrollTo(0, savedScrollY);
          }
          previousStyles = null;
          savedScrollY = 0;
        }
      }
    };
  }, []);
};
export { useDisableBodyScroll };
