// @tamagui/select/dist/cjs/context.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var context_exports = {};
__export(context_exports, {
  ForwardSelectContext: () => ForwardSelectContext,
  SelectItemParentProvider: () => SelectItemParentProvider,
  SelectProvider: () => SelectProvider,
  SelectZIndexContext: () => SelectZIndexContext,
  useSelectContext: () => useSelectContext,
  useSelectItemParentContext: () => useSelectItemParentContext
});
module.exports = __toCommonJS(context_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@tamagui/core");
var import_native = require("@tamagui/native");
var import_react = require("react");
var SelectZIndexContext = /* @__PURE__ */(0, import_react.createContext)(void 0);
var {
  Provider: SelectProvider,
  useStyledContext: useSelectContext
} = (0, import_core.createStyledContext)(null, "Select");
var {
  Provider: SelectItemParentProvider,
  useStyledContext: useSelectItemParentContext
} = (0, import_core.createStyledContext)(null, "SelectItem");
var ForwardSelectContext = function (param) {
  var {
    context,
    itemContext,
    children
  } = param;
  var portalState = (0, import_native.getPortal)().state;
  if (portalState.type === "teleport") {
    return children;
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(SelectProvider, {
    isInSheet: true,
    scope: context.scopeName,
    ...context,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(SelectItemParentProvider, {
      scope: context.scopeName,
      ...itemContext,
      children
    })
  });
};
