// @tamagui/select/dist/cjs/SelectViewport.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SelectViewport_exports = {};
__export(SelectViewport_exports, {
  SelectViewport: () => SelectViewport,
  SelectViewportFrame: () => SelectViewportFrame
});
module.exports = __toCommonJS(SelectViewport_exports);
var import_adapt = require("@tamagui/adapt");
var import_animate_presence = require("@tamagui/animate-presence");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_portal = require("@tamagui/portal");
var import_stacks = require("@tamagui/stacks");
var import_start_transition = require("@tamagui/start-transition");
var React = __toESM(require("react"), 1);
var import_constants2 = require("./constants.cjs");
var import_context = require("./context.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
const SelectViewportFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: import_constants2.VIEWPORT_NAME,
  variants: {
    unstyled: {
      false: {
        size: "$2",
        backgroundColor: "$background",
        elevate: true,
        bordered: true,
        userSelect: "none",
        outlineWidth: 0
      }
    },
    size: {
      "...size": (val, {
        tokens
      }) => {
        return {
          borderRadius: tokens.radius[val] ?? val
        };
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const needsRepropagation = (0, import_portal.needsPortalRepropagation)();
const SelectViewport = SelectViewportFrame.styleable(function SelectViewport2(props, forwardedRef) {
  const {
    scope,
    children,
    disableScroll,
    ...viewportProps
  } = props;
  const context = (0, import_context.useSelectContext)(scope);
  const itemContext = (0, import_context.useSelectItemParentContext)(scope);
  const isAdapted = (0, import_adapt.useAdaptIsActive)(context.adaptScope);
  const [lazyMounted, setLazyMounted] = React.useState(context.lazyMount ? false : true);
  React.useEffect(() => {
    if (!context.lazyMount) return;
    if (!context.open) return;
    if (lazyMounted) return;
    (0, import_start_transition.startTransition)(() => {
      setLazyMounted(true);
    });
  }, [context.lazyMount, context.open, lazyMounted]);
  const composedRefs = (0, import_compose_refs.useComposedRefs)(
  // @ts-ignore TODO react 19 type needs fix
  forwardedRef, context.floatingContext?.refs.setFloating);
  (0, import_constants.useIsomorphicLayoutEffect)(() => {
    if (context.update) {
      context.update();
    }
  }, [isAdapted]);
  (0, import_constants.useIsomorphicLayoutEffect)(() => {
    if (context.lazyMount && lazyMounted && context.open && context.update) {
      context.update();
    }
  }, [lazyMounted]);
  if (itemContext.shouldRenderWebNative) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_stacks.YStack, {
      position: "relative",
      children
    });
  }
  if (isAdapted || !import_constants.isWeb) {
    let content = children;
    if (needsRepropagation) {
      content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_context.ForwardSelectContext, {
        itemContext,
        context,
        children: content
      });
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_adapt.AdaptPortalContents, {
      scope: context.adaptScope,
      children: content
    });
  }
  if (!itemContext.interactions) {
    if (process.env.NODE_ENV === "development") {
      console.warn(`No interactions provided to Select, potentially missing Adapt`);
    }
    return null;
  }
  const {
    style,
    // remove this, it was set to "Select" always
    className,
    ...floatingProps
  } = itemContext.interactions.getFloatingProps();
  return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
    children: [!disableScroll && !props.unstyled && /* @__PURE__ */(0, import_jsx_runtime.jsx)("style", {
      dangerouslySetInnerHTML: {
        __html: selectViewportCSS
      }
    }), /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate_presence.AnimatePresence, {
      children: context.open ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(SelectViewportFrame, {
        size: itemContext.size,
        role: "presentation",
        ...viewportProps,
        ...style,
        ...floatingProps,
        ...(!props.unstyled && {
          overflowY: disableScroll ? void 0 : style.overflow ?? "auto"
        }),
        ref: composedRefs,
        children: lazyMounted ? children : null
      }, "select-viewport") : null
    }), !context.open && !(context.lazyMount && context.renderValue) && lazyMounted && /* @__PURE__ */(0, import_jsx_runtime.jsx)("div", {
      style: {
        display: "none"
      },
      children
    })]
  });
});
const selectViewportCSS = `
.is_SelectViewport {
  scrollbar-width: none;
  -webkit-overflow-scrolling: touch;
  overscroll-behavior: contain;
}

.is_SelectViewport::-webkit-scrollbar{
  display:none
}
`;
