// @tamagui/select/dist/cjs/SelectItem.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SelectItem_exports = {};
__export(SelectItem_exports, {
  SelectItem: () => SelectItem,
  SelectItemContextProvider: () => SelectItemContextProvider,
  useSelectItemContext: () => useSelectItemContext
});
module.exports = __toCommonJS(SelectItem_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_list_item = require("@tamagui/list-item");
var React = __toESM(require("react"), 1);
var import_context = require("./context.native.js");
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
    return !!right[Symbol.hasInstance](left);
  } else {
    return left instanceof right;
  }
}
var ITEM_NAME = "SelectItem";
var {
  Provider: SelectItemContextProvider,
  useStyledContext: useSelectItemContext
} = (0, import_core.createStyledContext)(null, ITEM_NAME);
var SelectItem = import_list_item.ListItem.Frame.styleable(function SelectItem2(props, forwardedRef) {
  var {
    scope,
    value,
    disabled = false,
    textValue: textValueProp,
    index,
    ...restProps
  } = props;
  var context = (0, import_context.useSelectItemParentContext)(scope);
  var {
    setSelectedIndex,
    listRef,
    setOpen,
    onChange,
    activeIndexSubscribe,
    activeIndexRef,
    valueSubscribe,
    allowMouseUpRef,
    allowSelectRef,
    setValueAtIndex,
    selectTimeoutRef,
    dataRef,
    interactions,
    shouldRenderWebNative,
    size,
    onActiveChange,
    initialValue,
    setActiveIndexFast
  } = context;
  var [isSelected, setSelected] = React.useState(initialValue === value);
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    if (initialValue === value) {
      setSelectedIndex(index);
    }
  }, []);
  React.useEffect(function () {
    var handleActiveIndex = function (i) {
      if (index === i) {
        onActiveChange(value, index);
        if (import_constants.isWeb) {
          requestAnimationFrame(function () {
            var _listRef_current_index;
            listRef === null || listRef === void 0 ? void 0 : (_listRef_current_index = listRef.current[index]) === null || _listRef_current_index === void 0 ? void 0 : _listRef_current_index.focus();
          });
        }
      }
    };
    var currentActiveIndex = activeIndexRef === null || activeIndexRef === void 0 ? void 0 : activeIndexRef.current;
    if (currentActiveIndex !== null && currentActiveIndex !== void 0) {
      handleActiveIndex(currentActiveIndex);
    }
    return activeIndexSubscribe(handleActiveIndex);
  }, [index]);
  React.useEffect(function () {
    return valueSubscribe(function (val) {
      setSelected(val === value);
    });
  }, [value]);
  var textId = React.useId();
  var refCallback = React.useCallback(function (node) {
    if (!import_constants.isWeb) return;
    if (_instanceof(node, HTMLElement)) {
      if (listRef) {
        listRef.current[index] = node;
      }
    }
  }, [index, listRef]);
  var composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, refCallback);
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    setValueAtIndex(index, value);
  }, [index, setValueAtIndex, value]);
  function handleSelect() {
    setSelectedIndex(index);
    onChange(value);
    setOpen(false);
  }
  var selectItemProps = React.useMemo(function () {
    return interactions ? interactions.getItemProps({
      onTouchMove() {
        allowSelectRef.current = true;
        allowMouseUpRef.current = false;
      },
      onTouchEnd() {
        allowSelectRef.current = false;
        allowMouseUpRef.current = true;
      },
      onKeyDown(event) {
        if (event.key === "Enter" || event.key === " " && !(dataRef === null || dataRef === void 0 ? void 0 : dataRef.current.typing)) {
          event.preventDefault();
          handleSelect();
        } else if (event.key === "ArrowDown" || event.key === "ArrowUp") {
          event.preventDefault();
          event.stopPropagation();
          var _listRef_current_length;
          var itemCount = (_listRef_current_length = listRef === null || listRef === void 0 ? void 0 : listRef.current.length) !== null && _listRef_current_length !== void 0 ? _listRef_current_length : 0;
          if (itemCount === 0) return;
          var nextIndex;
          if (event.key === "ArrowDown") {
            nextIndex = index + 1 >= itemCount ? 0 : index + 1;
          } else {
            nextIndex = index - 1 < 0 ? itemCount - 1 : index - 1;
          }
          setActiveIndexFast === null || setActiveIndexFast === void 0 ? void 0 : setActiveIndexFast(nextIndex);
        } else {
          allowSelectRef.current = true;
        }
      },
      onClick() {
        if (allowSelectRef.current) {
          handleSelect();
        }
      },
      onMouseUp() {
        if (!allowMouseUpRef.current) {
          allowMouseUpRef.current = true;
          allowSelectRef.current = true;
          return;
        }
        if (allowSelectRef.current) {
          handleSelect();
        }
        clearTimeout(selectTimeoutRef.current);
        selectTimeoutRef.current = setTimeout(function () {
          allowSelectRef.current = true;
        });
      }
    }) : {
      onPress: handleSelect
    };
  }, [handleSelect, index, listRef, setActiveIndexFast]);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(SelectItemContextProvider, {
    scope,
    value,
    textId: textId || "",
    isSelected,
    children: shouldRenderWebNative ? /* @__PURE__ */(0, import_jsx_runtime.jsx)("option", {
      value,
      children: props.children
    }) : /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_list_item.ListItem.Frame, {
      render: "div",
      componentName: ITEM_NAME,
      ref: composedRefs,
      role: "option",
      "aria-labelledby": textId,
      "aria-selected": isSelected,
      "data-state": isSelected ? "active" : "inactive",
      "aria-disabled": disabled || void 0,
      "data-disabled": disabled ? "" : void 0,
      tabIndex: disabled ? void 0 : -1,
      ...(!props.unstyled && {
        cursor: "default",
        size,
        outlineOffset: -0.5,
        zIndex: 100,
        hoverStyle: {
          backgroundColor: "$backgroundHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress"
        },
        focusStyle: {
          backgroundColor: "$backgroundFocus"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineWidth: 1,
          outlineStyle: "solid"
        }
      }),
      ...restProps,
      ...selectItemProps
    })
  });
}, {
  disableTheme: true
});
