// @tamagui/select/dist/cjs/SelectItemText.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SelectItemText_exports = {};
__export(SelectItemText_exports, {
  ITEM_TEXT_NAME: () => ITEM_TEXT_NAME,
  SelectItemText: () => SelectItemText,
  SelectItemTextFrame: () => SelectItemTextFrame
});
module.exports = __toCommonJS(SelectItemText_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_compose_refs = require("@tamagui/compose-refs");
var import_core = require("@tamagui/core");
var import_text = require("@tamagui/text");
var React = __toESM(require("react"), 1);
var import_context = require("./context.native.js");
var import_SelectItem = require("./SelectItem.native.js");
var ITEM_TEXT_NAME = "SelectItemText";
var SelectItemTextFrame = (0, import_core.styled)(import_text.SizableText, {
  name: ITEM_TEXT_NAME,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        color: "$color",
        ellipsis: true
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var SelectItemText = SelectItemTextFrame.styleable(function SelectItemText2(props, forwardedRef) {
  var {
    scope,
    className,
    ...itemTextProps
  } = props;
  var itemParentContext = (0, import_context.useSelectItemParentContext)(scope);
  var ref = React.useRef(null);
  var composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, ref);
  var itemContext = (0, import_SelectItem.useSelectItemContext)(scope);
  var contents = React.useRef(null);
  contents.current = /* @__PURE__ */(0, import_jsx_runtime.jsx)(SelectItemTextFrame, {
    className,
    size: itemParentContext.size,
    id: itemContext.textId,
    ...itemTextProps,
    ref: composedRefs
  });
  (0, import_core.useIsomorphicLayoutEffect)(function () {
    if (itemParentContext.initialValue === itemContext.value) {
      itemParentContext.setSelectedItem(contents.current);
    }
  }, []);
  (0, import_core.useIsomorphicLayoutEffect)(function () {
    return itemParentContext.valueSubscribe(function (val) {
      if (val === itemContext.value) {
        itemParentContext.setSelectedItem(contents.current);
      }
    });
  }, [itemContext.value]);
  if (itemParentContext.shouldRenderWebNative) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
      children: props.children
    });
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children: contents.current
  });
});
