// @tamagui/create-menu/dist/cjs/createNativeMenu/createNativeMenu.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createNativeMenu_exports = {};
__export(createNativeMenu_exports, {
  createNativeMenu: () => createNativeMenu
});
module.exports = __toCommonJS(createNativeMenu_exports);
var import_native = require("@tamagui/native");
var import_web = require("@tamagui/web");
var import_react = __toESM(require("react"), 1);
var import_jsx_runtime = require("react/jsx-runtime");
const MAPPED_TYPES = ["SubContent", "SubTrigger", "Content", "Sub", "Group", "CheckboxItem"];
const CONTAINER_TYPES = ["SubContent", "Content", "Sub", "Group"];
function getComponentType(displayName) {
  for (const type of MAPPED_TYPES) {
    if (displayName === type || displayName.includes(`(${type})`)) {
      return type;
    }
  }
  return null;
}
function isItemLike(props, displayName) {
  if (getComponentType(displayName)) return false;
  return "onSelect" in props || "textValue" in props;
}
function isPortalLike(displayName) {
  return displayName === "Portal" || displayName.includes("Portal");
}
const emptyStub = () => null;
function createWebStubs() {
  return {
    Menu: (0, import_web.withStaticProperties)(emptyStub, {
      Trigger: emptyStub,
      Content: emptyStub,
      Item: emptyStub,
      ItemTitle: emptyStub,
      ItemSubtitle: emptyStub,
      SubTrigger: emptyStub,
      Group: emptyStub,
      ItemIcon: emptyStub,
      Separator: emptyStub,
      CheckboxItem: emptyStub,
      ItemIndicator: emptyStub,
      ItemImage: emptyStub,
      Label: emptyStub,
      Arrow: emptyStub,
      Sub: emptyStub,
      SubContent: emptyStub,
      Preview: emptyStub,
      Portal: emptyStub,
      RadioGroup: emptyStub,
      RadioItem: emptyStub,
      Auxiliary: emptyStub
    })
  };
}
const createNativeMenu = MenuType => {
  if (import_web.isWeb) {
    return createWebStubs();
  }
  const isContextMenu = MenuType === "ContextMenu";
  const isAndroid = !import_web.isIos && !import_web.isWeb;
  let resolved = null;
  let warned = false;
  function resolve() {
    if (resolved) return resolved;
    const zeego = (0, import_native.getZeego)();
    if (!zeego.isEnabled) {
      if (!warned) {
        warned = true;
        console.warn(`Warning: Must call import '@tamagui/native/setup-zeego' at your app entry point to use native menus`);
      }
      return null;
    }
    const menu = isContextMenu ? zeego.state.ContextMenu : zeego.state.DropdownMenu;
    resolved = {
      menu,
      componentMap: {
        SubContent: menu.SubContent,
        Content: menu.Content,
        Sub: menu.Sub,
        Group: menu.Group,
        SubTrigger: menu.SubTrigger
      }
    };
    return resolved;
  }
  function transformChildren(menu, map, children, shouldReverseOnIos = false) {
    const result = [];
    import_react.default.Children.forEach(children, child => {
      if (!import_react.default.isValidElement(child)) {
        result.push(child);
        return;
      }
      const displayName = child.type?.displayName || "";
      const props = child.props;
      if (isPortalLike(displayName)) {
        const inner = transformChildren(menu, map, props.children, false);
        import_react.default.Children.forEach(inner, c => result.push(c));
        return;
      }
      if (displayName.includes("ScrollView")) {
        const inner = transformChildren(menu, map, props.children, false);
        import_react.default.Children.forEach(inner, c => result.push(c));
        return;
      }
      const componentType = getComponentType(displayName);
      if (componentType === "CheckboxItem") {
        const {
          checked,
          onCheckedChange,
          value,
          onValueChange,
          children: cbChildren,
          ...rest
        } = props;
        const finalValue = value ?? (checked ? "on" : "off");
        const finalOnValueChange = onValueChange ?? (onCheckedChange && (v => onCheckedChange(v === "on")));
        const cleanChildren = import_react.default.Children.map(cbChildren, c => {
          if (!import_react.default.isValidElement(c)) return c;
          const dn = c.type?.displayName || "";
          if (dn.includes("ItemIndicator")) return null;
          return c;
        });
        result.push(import_react.default.createElement(menu.CheckboxItem, {
          ...rest,
          key: child.key,
          value: finalValue,
          onValueChange: finalOnValueChange
        }, cleanChildren));
        return;
      }
      if (componentType) {
        const {
          children: childChildren,
          ...restProps
        } = props;
        const isContainer = CONTAINER_TYPES.includes(componentType);
        const shouldReverse = componentType === "Content" || componentType === "SubContent";
        result.push(import_react.default.createElement(map[componentType], {
          ...restProps,
          key: child.key
        }, isContainer ? transformChildren(menu, map, childChildren, shouldReverse) : childChildren));
        return;
      }
      if (isItemLike(props, displayName)) {
        const {
          children: itemChildren,
          ...itemProps
        } = props;
        result.push(import_react.default.createElement(menu.Item, {
          ...itemProps,
          key: child.key
        }, itemChildren));
        return;
      }
      result.push(child);
    });
    if (import_web.isIos && shouldReverseOnIos && !isContextMenu) {
      result.reverse();
    }
    return result;
  }
  function lazyZeego(name, displayName) {
    const Comp = props => {
      const z = resolve();
      if (!z) return null;
      return import_react.default.createElement(z.menu[name], props);
    };
    Comp.displayName = displayName || name;
    return Comp;
  }
  const Trigger = lazyZeego("Trigger");
  const Content = lazyZeego("Content");
  const Item = lazyZeego("Item");
  const ItemTitle = lazyZeego("ItemTitle");
  const ItemSubtitle = lazyZeego("ItemSubtitle");
  const ItemIcon = lazyZeego("ItemIcon");
  const ItemImage = lazyZeego("ItemImage");
  const ItemIndicator = lazyZeego("ItemIndicator");
  const Group = lazyZeego("Group");
  const Label = lazyZeego("Label");
  const Separator = lazyZeego("Separator");
  const Sub = lazyZeego("Sub");
  const SubTrigger = lazyZeego("SubTrigger");
  const SubContent = lazyZeego("SubContent");
  const Portal = ({
    children
  }) => /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children
  });
  Portal.displayName = "Portal";
  const Arrow = () => null;
  Arrow.displayName = "Arrow";
  const RadioGroup = ({
    children
  }) => /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children
  });
  RadioGroup.displayName = `${MenuType}RadioGroup`;
  const RadioItem = ({
    children
  }) => /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children
  });
  RadioItem.displayName = `${MenuType}RadioItem`;
  const CheckboxItem = () => null;
  CheckboxItem.displayName = "CheckboxItem";
  const Preview = isContextMenu ? lazyZeego("Preview", `${MenuType}Preview`) : () => null;
  Preview.displayName = `${MenuType}Preview`;
  const Auxiliary = isContextMenu ? lazyZeego("Auxiliary", `${MenuType}Auxiliary`) : () => null;
  Auxiliary.displayName = `${MenuType}Auxiliary`;
  const Menu = ({
    children,
    onOpenChange,
    onOpenWillChange
  }) => {
    const z = resolve();
    if (!z) return null;
    const rootProps = {
      onOpenChange
    };
    if (isContextMenu && onOpenWillChange) {
      rootProps.onOpenWillChange = onOpenWillChange;
    }
    const content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(z.menu.Root, {
      ...rootProps,
      children: transformChildren(z.menu, z.componentMap, children)
    });
    if (isAndroid) {
      return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_native.NativeMenuContext.Provider, {
        value: true,
        children: content
      });
    }
    return content;
  };
  Menu.displayName = MenuType;
  return {
    Menu: (0, import_web.withStaticProperties)(Menu, {
      Trigger,
      Content,
      Item,
      ItemTitle,
      ItemSubtitle,
      ItemIcon,
      ItemImage,
      ItemIndicator,
      Group,
      Label,
      Separator,
      Sub,
      SubTrigger,
      SubContent,
      CheckboxItem,
      Portal,
      RadioGroup,
      RadioItem,
      Arrow,
      Preview,
      Auxiliary
    })
  };
};
