// @tamagui/create-menu/dist/cjs/createNativeMenu/utils.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var utils_exports = {};
__export(utils_exports, {
  create: () => create,
  filterNull: () => filterNull,
  flattenChildren: () => flattenChildren,
  flattenChildrenKeyless: () => flattenChildrenKeyless,
  isInstanceOfComponent: () => isInstanceOfComponent,
  pickChildren: () => pickChildren
});
module.exports = __toCommonJS(utils_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_react = __toESM(require("react"), 1);
function flattenChildrenKeyless(children) {
  var childrenArray = import_react.default.Children.toArray(children);
  return childrenArray.reduce(function (flatChildren, child) {
    if (child.type === import_react.default.Fragment) {
      return flatChildren.concat(flattenChildren(child.props.children));
    }
    flatChildren.push(child);
    return flatChildren;
  }, []);
}
function flattenChildren(children, componentNamesToIgnore) {
  var depth = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : 0,
    keys = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : [];
  return import_react.Children.toArray(children).flatMap(function (elem) {
    if (/* @__PURE__ */(0, import_react.isValidElement)(elem)) {
      if (typeof (elem === null || elem === void 0 ? void 0 : elem.type) == "function" && (componentNamesToIgnore === null || componentNamesToIgnore === void 0 ? void 0 : componentNamesToIgnore.some(function (skipComponentName) {
        return elem.type.displayName.includes(skipComponentName);
      }))) {
        return elem.props.children;
      }
    }
    return elem;
  }).reduce(function (acc, node, nodeIndex) {
    if (node.type === import_react.default.Fragment) {
      acc.push.apply(acc, flattenChildren(node.props.children, componentNamesToIgnore, depth + 1, keys.concat(node.key || nodeIndex)));
    } else {
      if (/* @__PURE__ */(0, import_react.isValidElement)(node)) {
        acc.push(/* @__PURE__ */(0, import_react.cloneElement)(node, {
          key: keys.concat(String(node.key)).join(".")
        }));
      } else if (typeof node === "string" || typeof node === "number") {
        acc.push(node);
      }
    }
    return acc;
  }, []);
}
var pickChildren = function (_children, targetChild, componentNamesToIgnore) {
  var children = flattenChildren(_children, componentNamesToIgnore);
  var target = [];
  var withoutTargetChildren = import_react.default.Children.map(children, function (item) {
    if (! /* @__PURE__ */(0, import_react.isValidElement)(item)) return item;
    if (isInstanceOfComponent(item, targetChild)) {
      target.push(/* @__PURE__ */(0, import_react.cloneElement)(item));
      return null;
    }
    return item;
  });
  var targetChildren = target.length >= 0 ? target : void 0;
  return {
    targetChildren,
    withoutTargetChildren
  };
};
var isInstanceOfComponent = function (element, targetElement) {
  var _element_type, _element_type1;
  var matches = (element === null || element === void 0 ? void 0 : element.type) === targetElement || typeof (element === null || element === void 0 ? void 0 : element.type) == "function" && ((element === null || element === void 0 ? void 0 : (_element_type = element.type) === null || _element_type === void 0 ? void 0 : _element_type.displayName) === targetElement.displayName || (element === null || element === void 0 ? void 0 : (_element_type1 = element.type) === null || _element_type1 === void 0 ? void 0 : _element_type1.displayName) === targetElement.displayName + "Wrapper");
  return matches;
};
var filterNull = function (t) {
  return t != null;
};
var create = function (Component, displayName) {
  var MenuComponent = function (props) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Component, {
      ...props
    });
  };
  MenuComponent.displayName = displayName;
  return MenuComponent;
};
