// @tamagui/tabs/dist/cjs/createTabs.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createTabs_exports = {};
__export(createTabs_exports, {
  createTabs: () => createTabs
});
module.exports = __toCommonJS(createTabs_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_group = require("@tamagui/group");
var import_helpers = require("@tamagui/helpers");
var import_roving_focus = require("@tamagui/roving-focus");
var import_sizable_context = require("@tamagui/sizable-context");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var import_use_direction = require("@tamagui/use-direction");
var import_web = require("@tamagui/web");
var React = __toESM(require("react"), 1);
var import_StyledContext = require("./StyledContext.native.js");
var import_Tabs = require("./Tabs.native.js");
function createTabs(createProps) {
  var {
    ContentFrame = import_Tabs.DefaultTabsContentFrame,
    TabFrame = import_Tabs.DefaultTabsTabFrame,
    TabsFrame = import_Tabs.DefaultTabsFrame
  } = createProps;
  var TABS_CONTEXT = "TabsContext";
  var TAB_LIST_NAME = "TabsList";
  var TabsList = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
    var {
      __scopeTabs,
      loop = true,
      children,
      ...listProps
    } = props;
    var context = (0, import_StyledContext.useTabsContext)(__scopeTabs);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup, {
      __scopeRovingFocusGroup: __scopeTabs || TABS_CONTEXT,
      orientation: context.orientation,
      dir: context.dir,
      loop,
      asChild: true,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_group.Group, {
        role: "tablist",
        componentName: TAB_LIST_NAME,
        "aria-orientation": context.orientation,
        ref: forwardedRef,
        orientation: context.orientation,
        ...listProps,
        children
      })
    });
  });
  TabsList.displayName = TAB_LIST_NAME;
  var TRIGGER_NAME = "TabsTrigger";
  var TabsTrigger = TabFrame.styleable(function (props, forwardedRef) {
    var {
      __scopeTabs,
      value,
      disabled = false,
      onInteraction,
      activeStyle,
      activeTheme,
      unstyled = false,
      ...triggerProps
    } = props;
    var context = (0, import_StyledContext.useTabsContext)(__scopeTabs);
    var triggerId = makeTriggerId(context.baseId, value);
    var contentId = makeContentId(context.baseId, value);
    var isSelected = value === context.value;
    var [layout, setLayout] = React.useState(null);
    var triggerRef = React.useRef(null);
    var groupItemProps = (0, import_group.useGroupItem)({
      disabled: !!disabled
    });
    React.useEffect(function () {
      context.registerTrigger();
      return function () {
        return context.unregisterTrigger();
      };
    }, []);
    React.useEffect(function () {
      if (!triggerRef.current || !import_constants.isWeb) return;
      var el = triggerRef.current;
      function getTriggerSize() {
        if (!el) return;
        setLayout({
          width: el.offsetWidth,
          height: el.offsetHeight,
          x: el.offsetLeft,
          y: el.offsetTop
        });
      }
      getTriggerSize();
      var observer = new ResizeObserver(getTriggerSize);
      observer.observe(el);
      return function () {
        observer.disconnect();
      };
    }, [context.triggersCount]);
    React.useEffect(function () {
      if (isSelected && layout) {
        onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("select", layout);
      }
    }, [isSelected, value, layout]);
    var _props_onPress;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup.Item, {
      __scopeRovingFocusGroup: __scopeTabs || TABS_CONTEXT,
      asChild: true,
      focusable: !disabled,
      active: isSelected,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(TabFrame, {
        ...(!import_constants.isWeb && {
          onLayout: function (event) {
            setLayout(event.nativeEvent.layout);
          }
        }),
        onMouseEnter: (0, import_helpers.composeEventHandlers)(props.onMouseEnter, function () {
          if (layout) {
            onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("hover", layout);
          }
        }),
        onMouseLeave: (0, import_helpers.composeEventHandlers)(props.onMouseLeave, function () {
          onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("hover", null);
        }),
        role: "tab",
        "aria-selected": isSelected,
        "aria-controls": contentId,
        "data-state": isSelected ? "active" : "inactive",
        "data-disabled": disabled ? "" : void 0,
        id: triggerId,
        theme: activeTheme !== null && activeTheme !== void 0 ? activeTheme : null,
        unstyled,
        ...(!unstyled && {
          size: context.size
        }),
        ...(isSelected && {
          ...(!unstyled && !activeStyle && {
            backgroundColor: "$backgroundActive"
          }),
          ...activeStyle
        }),
        ...groupItemProps,
        disabled: disabled !== null && disabled !== void 0 ? disabled : groupItemProps.disabled,
        ...triggerProps,
        ref: (0, import_compose_refs.composeRefs)(forwardedRef, triggerRef),
        onPress: (0, import_helpers.composeEventHandlers)((_props_onPress = props.onPress) !== null && _props_onPress !== void 0 ? _props_onPress : void 0, function (event) {
          var webChecks = !import_constants.isWeb || event.button === 0 && event.ctrlKey === false;
          if (!disabled && !isSelected && webChecks) {
            context.onChange(value);
          }
        }),
        ...(import_constants.isWeb && {
          onKeyDown: (0, import_helpers.composeEventHandlers)(props.onKeyDown, function (event) {
            if ([" ", "Enter"].includes(event.key)) {
              context.onChange(value);
              event.preventDefault();
            }
          }),
          onFocus: (0, import_helpers.composeEventHandlers)(props.onFocus, function (event) {
            if (layout) {
              onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("focus", layout);
            }
            var isAutomaticActivation = context.activationMode !== "manual";
            if (!isSelected && !disabled && isAutomaticActivation) {
              context.onChange(value);
            }
          }),
          onBlur: (0, import_helpers.composeEventHandlers)(props.onBlur, function () {
            onInteraction === null || onInteraction === void 0 ? void 0 : onInteraction("focus", null);
          })
        })
      })
    });
  });
  TabsTrigger.displayName = TRIGGER_NAME;
  var TabsContent = ContentFrame.styleable(function TabsContent2(props, forwardedRef) {
    var {
      __scopeTabs,
      value,
      forceMount,
      children,
      ...contentProps
    } = props;
    var context = (0, import_StyledContext.useTabsContext)(__scopeTabs);
    var isSelected = value === context.value;
    var show = forceMount || isSelected;
    var triggerId = makeTriggerId(context.baseId, value);
    var contentId = makeContentId(context.baseId, value);
    if (!show) {
      return null;
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(ContentFrame, {
      "data-state": isSelected ? "active" : "inactive",
      "data-orientation": context.orientation,
      role: "tabpanel",
      "aria-labelledby": triggerId,
      // @ts-ignore
      hidden: !show,
      id: contentId,
      tabIndex: 0,
      ...contentProps,
      ref: forwardedRef,
      children
    }, value);
  });
  var TabsComponent = TabsFrame.styleable(function Tabs(props, forwardedRef) {
    var {
      __scopeTabs,
      value: valueProp,
      onValueChange,
      defaultValue,
      orientation = "horizontal",
      dir,
      activationMode = "manual",
      size = "$true",
      ...tabsProps
    } = props;
    var direction = (0, import_use_direction.useDirection)(dir);
    var [value, setValue] = (0, import_use_controllable_state.useControllableState)({
      prop: valueProp,
      onChange: onValueChange,
      defaultProp: defaultValue !== null && defaultValue !== void 0 ? defaultValue : ""
    });
    var [triggersCount, setTriggersCount] = React.useState(0);
    var registerTrigger = (0, import_web.useEvent)(function () {
      return setTriggersCount(function (v) {
        return v + 1;
      });
    });
    var unregisterTrigger = (0, import_web.useEvent)(function () {
      return setTriggersCount(function (v) {
        return v - 1;
      });
    });
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_sizable_context.SizableContext.Provider, {
      size,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_StyledContext.TabsProvider, {
        scope: __scopeTabs,
        baseId: React.useId(),
        value,
        onChange: setValue,
        orientation,
        dir: direction,
        activationMode,
        size,
        registerTrigger,
        triggersCount,
        unregisterTrigger,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(TabsFrame, {
          direction,
          "data-orientation": orientation,
          ...tabsProps,
          ref: forwardedRef
        })
      })
    });
  });
  return (0, import_helpers.withStaticProperties)(TabsComponent, {
    List: TabsList,
    /**
    * @deprecated Use Tabs.Tab instead
    */
    Trigger: TabsTrigger,
    Tab: TabsTrigger,
    Content: TabsContent
  });
}
function makeTriggerId(baseId, value) {
  return `${baseId}-trigger-${value}`;
}
function makeContentId(baseId, value) {
  return `${baseId}-content-${value}`;
}
