// @tamagui/v2-toast/dist/jsx/ToastProvider.mjs
import { createCollection } from "@tamagui/collection";
import { createStyledContext } from "@tamagui/core";
import { startTransition } from "@tamagui/start-transition";
import * as React from "react";
import { TOAST_CONTEXT } from "./constants.mjs";
import { ToastImperativeProvider } from "./ToastImperative.mjs";
import { jsx } from "react/jsx-runtime";
const PROVIDER_NAME = "ToastProvider";
const [Collection, useCollection] = createCollection("Toast");
const {
  Provider: ToastProviderProvider,
  useStyledContext: useToastProviderContext
} = createStyledContext(
// since we always provide this we can avoid setting here
{}, "Toast__");
const ToastProvider = props => {
  const {
    scope = TOAST_CONTEXT,
    id: providedId,
    burntOptions,
    native,
    notificationOptions,
    label = "Notification",
    duration = 5e3,
    swipeDirection = "right",
    swipeThreshold = 50,
    children
  } = props;
  const backupId = React.useId();
  const id = providedId ?? backupId;
  const [viewports, setViewports] = React.useState({});
  const [toastCount, setToastCount] = React.useState(0);
  const isFocusedToastEscapeKeyDownRef = React.useRef(false);
  const isClosePausedRef = React.useRef(false);
  const handleViewportChange = React.useCallback((name, viewport) => {
    startTransition(() => {
      setViewports(prev => ({
        ...prev,
        [name]: viewport
      }));
    });
  }, []);
  const options = React.useMemo(() => {
    return {
      duration,
      burntOptions,
      native,
      notificationOptions
    };
  }, [JSON.stringify([duration, burntOptions, native, notificationOptions])]);
  return /* @__PURE__ */jsx(Collection.Provider, {
    scope,
    children: /* @__PURE__ */jsx(ToastProviderProvider, {
      scope,
      id,
      label,
      duration,
      swipeDirection,
      swipeThreshold,
      toastCount,
      viewports,
      onViewportChange: handleViewportChange,
      onToastAdd: React.useCallback(() => {
        startTransition(() => {
          setToastCount(prevCount => prevCount + 1);
        });
      }, []),
      onToastRemove: React.useCallback(() => {
        startTransition(() => {
          setToastCount(prevCount => prevCount - 1);
        });
      }, []),
      isFocusedToastEscapeKeyDownRef,
      isClosePausedRef,
      options,
      children: /* @__PURE__ */jsx(ToastImperativeProvider, {
        options,
        children
      })
    })
  });
};
function ReprogapateToastProvider(props) {
  const {
    children,
    context
  } = props;
  return /* @__PURE__ */jsx(Collection.Provider, {
    scope: context.toastScope,
    children: /* @__PURE__ */jsx(ToastProviderProvider, {
      ...context,
      children: /* @__PURE__ */jsx(ToastImperativeProvider, {
        options: context.options,
        children
      })
    })
  });
}
ToastProvider.propTypes = {
  label(props) {
    if (props.label && typeof props.label === "string" && !props.label.trim()) {
      const error = `Invalid prop \`label\` supplied to \`${PROVIDER_NAME}\`. Expected non-empty \`string\`.`;
      return new Error(error);
    }
    return null;
  }
};
ToastProvider.displayName = PROVIDER_NAME;
export { Collection, ReprogapateToastProvider, ToastProvider, useCollection, useToastProviderContext };
