// @tamagui/v2-toast/dist/jsx/Toaster.mjs
import { AnimatePresence } from "@tamagui/animate-presence";
import { isWeb } from "@tamagui/constants";
import { Theme, View, styled, useThemeName } from "@tamagui/core";
import { Portal } from "@tamagui/portal";
import * as React from "react";
import { ToastItem } from "./ToastItem.mjs";
import { ToastState } from "./ToastState.mjs";
import { jsx } from "react/jsx-runtime";
const VISIBLE_TOASTS_AMOUNT = 4;
const VIEWPORT_OFFSET = 16;
const TOAST_GAP = 14;
const TOAST_LIFETIME = 4e3;
const TOAST_WIDTH = 356;
const ToasterFrame = styled(View, {
  name: "Toaster",
  variants: {
    unstyled: {
      false: {
        position: "fixed",
        zIndex: 1e5,
        pointerEvents: "box-none",
        maxWidth: "100%",
        // need min-height to contain absolutely positioned toasts
        // toasts will overflow upward/downward from their anchor position
        minHeight: 1
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const Toaster = React.forwardRef(function Toaster2(props, _ref) {
  const {
    position = "bottom-right",
    width = TOAST_WIDTH,
    expand = false,
    visibleToasts = VISIBLE_TOASTS_AMOUNT,
    gap = TOAST_GAP,
    duration = TOAST_LIFETIME,
    offset = VIEWPORT_OFFSET,
    hotkey = ["altKey", "KeyT"],
    swipeDirection = "right",
    swipeThreshold = 50,
    closeButton = false,
    theme: themeProp,
    icons,
    toastOptions,
    containerAriaLabel = "Notifications",
    disableNative = false,
    burntOptions,
    notificationOptions,
    className,
    style
  } = props;
  const [toasts, setToasts] = React.useState([]);
  const [heights, setHeights] = React.useState([]);
  const [expanded, setExpanded] = React.useState(false);
  const [interacting, setInteracting] = React.useState(false);
  const listRef = React.useRef(null);
  const lastFocusedElementRef = React.useRef(null);
  const isFocusWithinRef = React.useRef(false);
  React.useEffect(() => {
    return ToastState.subscribe(toast => {
      if (toast.dismiss) {
        setToasts(toasts2 => toasts2.map(t => t.id === toast.id ? {
          ...t,
          delete: true
        } : t));
        return;
      }
      setToasts(toasts2 => {
        const indexOfExistingToast = toasts2.findIndex(t => t.id === toast.id);
        if (indexOfExistingToast !== -1) {
          return [...toasts2.slice(0, indexOfExistingToast), {
            ...toasts2[indexOfExistingToast],
            ...toast
          }, ...toasts2.slice(indexOfExistingToast + 1)];
        }
        return [toast, ...toasts2];
      });
    });
  }, []);
  React.useEffect(() => {
    if (toasts.length <= 1) {
      setExpanded(false);
    }
  }, [toasts.length]);
  React.useEffect(() => {
    if (!isWeb) return;
    const handleKeyDown = event => {
      const isHotkeyPressed = hotkey.length > 0 && hotkey.every(key => event[key] || event.code === key);
      if (isHotkeyPressed) {
        setExpanded(true);
        listRef.current?.focus();
      }
      if (event.code === "Escape" && (document.activeElement === listRef.current || listRef.current?.contains(document.activeElement))) {
        setExpanded(false);
      }
    };
    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [hotkey]);
  React.useEffect(() => {
    if (!isWeb || !listRef.current) return;
    return () => {
      if (lastFocusedElementRef.current) {
        lastFocusedElementRef.current.focus({
          preventScroll: true
        });
        lastFocusedElementRef.current = null;
        isFocusWithinRef.current = false;
      }
    };
  }, []);
  const removeToast = React.useCallback(toastToRemove => {
    setToasts(toasts2 => {
      if (!toasts2.find(toast => toast.id === toastToRemove.id)?.delete) {
        ToastState.dismiss(toastToRemove.id);
      }
      return toasts2.filter(({
        id
      }) => id !== toastToRemove.id);
    });
  }, []);
  const [yPosition, xPosition] = position.split("-");
  const offsetStyles = React.useMemo(() => {
    const styles = {};
    const defaultOffset = typeof offset === "number" ? offset : VIEWPORT_OFFSET;
    const offsetObj = typeof offset === "object" ? offset : {
      top: defaultOffset,
      right: defaultOffset,
      bottom: defaultOffset,
      left: defaultOffset
    };
    if (yPosition === "top") {
      styles.top = offsetObj.top ?? defaultOffset;
    } else {
      styles.bottom = offsetObj.bottom ?? defaultOffset;
    }
    if (xPosition === "left") {
      styles.left = offsetObj.left ?? defaultOffset;
    } else if (xPosition === "right") {
      styles.right = offsetObj.right ?? defaultOffset;
    } else {
      styles.left = "50%";
      styles.transform = "translateX(-50%)";
    }
    return styles;
  }, [offset, yPosition, xPosition]);
  const currentTheme = useThemeName();
  const resolvedTheme = themeProp === "system" || !themeProp ? currentTheme?.includes("dark") ? "dark" : "light" : themeProp;
  const hotkeyLabel = hotkey.join("+").replace(/Key/g, "").replace(/Digit/g, "");
  if (toasts.length === 0) {
    return null;
  }
  const content = /* @__PURE__ */jsx(ToasterFrame, {
    ref: listRef,
    width,
    "aria-label": `${containerAriaLabel} ${hotkeyLabel}`,
    tabIndex: -1,
    "aria-live": "polite",
    "aria-relevant": "additions text",
    "aria-atomic": false,
    style: {
      ...offsetStyles,
      ...style
    },
    className,
    "data-y-position": yPosition,
    "data-x-position": xPosition,
    onMouseEnter: () => setExpanded(true),
    onMouseMove: () => setExpanded(true),
    onMouseLeave: () => {
      if (!interacting) {
        setExpanded(false);
      }
    },
    onPointerDown: () => setInteracting(true),
    onPointerUp: () => setInteracting(false),
    ...(isWeb && {
      onBlur: event => {
        if (isFocusWithinRef.current && !event.currentTarget.contains(event.relatedTarget)) {
          isFocusWithinRef.current = false;
          if (lastFocusedElementRef.current) {
            lastFocusedElementRef.current.focus({
              preventScroll: true
            });
            lastFocusedElementRef.current = null;
          }
        }
      },
      onFocus: event => {
        if (!isFocusWithinRef.current) {
          isFocusWithinRef.current = true;
          lastFocusedElementRef.current = event.relatedTarget;
        }
      }
    }),
    children: /* @__PURE__ */jsx(AnimatePresence, {
      children: toasts.map((toast, index) => {
        const isVisible = index < visibleToasts;
        const isFront = index === 0;
        return /* @__PURE__ */jsx(ToastItem, {
          toast,
          index,
          expanded: expanded || expand,
          interacting,
          position,
          visibleToasts,
          removeToast,
          heights,
          setHeights,
          duration: toast.duration ?? toastOptions?.duration ?? duration,
          gap,
          swipeDirection,
          swipeThreshold,
          closeButton: toast.closeButton ?? closeButton,
          icons,
          disableNative,
          burntOptions,
          notificationOptions
        }, toast.id);
      })
    })
  });
  if (isWeb) {
    return /* @__PURE__ */jsx(Portal, {
      children: /* @__PURE__ */jsx(Theme, {
        name: resolvedTheme,
        children: content
      })
    });
  }
  return /* @__PURE__ */jsx(Theme, {
    name: resolvedTheme,
    children: content
  });
});
Toaster.displayName = "Toaster";
export { Toaster };
