// @tamagui/v2-toast/dist/jsx/useDragGesture.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useDragGesture_native_exports = {};
__export(useDragGesture_native_exports, {
  useDragGesture: () => useDragGesture
});
module.exports = __toCommonJS(useDragGesture_native_exports);
var React = __toESM(require("react"), 1);
var import_react_native = require("react-native");
var GESTURE_GRANT_THRESHOLD = 10;
var VELOCITY_THRESHOLD = 0.11;
function getDampening(delta) {
  var factor = Math.abs(delta) / 20;
  return 1 / (1.5 + factor);
}
function shouldGrantGestureMove(dir, dx, dy) {
  if ((dir === "horizontal" || dir === "left") && dx < -GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "horizontal" || dir === "right") && dx > GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "vertical" || dir === "up") && dy < -GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "vertical" || dir === "down") && dy > GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  return false;
}
function useDragGesture(options) {
  var {
    direction,
    threshold,
    onDragStart,
    onDragEnd,
    onDragCancel,
    disabled
  } = options;
  var [dragState, setDragState] = React.useState({
    isDragging: false,
    offsetX: 0,
    offsetY: 0,
    velocityX: 0,
    velocityY: 0
  });
  var dragStartTimeRef = React.useRef(0);
  var isHorizontal = direction === "left" || direction === "right" || direction === "horizontal";
  var isVertical = direction === "up" || direction === "down" || direction === "vertical";
  var panResponder = React.useMemo(function () {
    if (disabled) return null;
    return import_react_native.PanResponder.create({
      onMoveShouldSetPanResponder: function (_e, gesture) {
        return shouldGrantGestureMove(direction, gesture.dx, gesture.dy);
      },
      onPanResponderGrant: function () {
        dragStartTimeRef.current = Date.now();
        setDragState(function (prev) {
          return {
            ...prev,
            isDragging: true
          };
        });
        onDragStart === null || onDragStart === void 0 ? void 0 : onDragStart();
      },
      onPanResponderMove: function (_e, gesture) {
        var {
          dx,
          dy
        } = gesture;
        var offsetX = 0;
        var offsetY = 0;
        if (isHorizontal) {
          if (direction === "right" && dx > 0 || direction === "left" && dx < 0 || direction === "horizontal") {
            offsetX = dx;
          } else {
            offsetX = dx * getDampening(dx);
          }
        }
        if (isVertical) {
          if (direction === "down" && dy > 0 || direction === "up" && dy < 0 || direction === "vertical") {
            offsetY = dy;
          } else {
            offsetY = dy * getDampening(dy);
          }
        }
        setDragState(function (prev) {
          return {
            ...prev,
            offsetX,
            offsetY
          };
        });
      },
      onPanResponderRelease: function (_e, gesture) {
        var {
          dx,
          dy,
          vx,
          vy
        } = gesture;
        var relevantDelta = isHorizontal ? dx : dy;
        var relevantVelocity = isHorizontal ? Math.abs(vx) : Math.abs(vy);
        var passedThreshold = Math.abs(relevantDelta) >= threshold;
        var hasVelocity = relevantVelocity > VELOCITY_THRESHOLD;
        var isCorrectDirection = false;
        if (direction === "right" && dx > 0) isCorrectDirection = true;else if (direction === "left" && dx < 0) isCorrectDirection = true;else if (direction === "horizontal" && Math.abs(dx) > Math.abs(dy)) isCorrectDirection = true;else if (direction === "down" && dy > 0) isCorrectDirection = true;else if (direction === "up" && dy < 0) isCorrectDirection = true;else if (direction === "vertical" && Math.abs(dy) > Math.abs(dx)) isCorrectDirection = true;
        var shouldDismiss = isCorrectDirection && (passedThreshold || hasVelocity);
        if (shouldDismiss) {
          onDragEnd === null || onDragEnd === void 0 ? void 0 : onDragEnd(true);
        } else {
          setDragState({
            isDragging: false,
            offsetX: 0,
            offsetY: 0,
            velocityX: 0,
            velocityY: 0
          });
          onDragCancel === null || onDragCancel === void 0 ? void 0 : onDragCancel();
        }
      },
      onPanResponderTerminate: function () {
        setDragState({
          isDragging: false,
          offsetX: 0,
          offsetY: 0,
          velocityX: 0,
          velocityY: 0
        });
        onDragCancel === null || onDragCancel === void 0 ? void 0 : onDragCancel();
      }
    });
  }, [disabled, direction, threshold, isHorizontal, isVertical, onDragStart, onDragEnd, onDragCancel]);
  var resetDrag = React.useCallback(function () {
    setDragState({
      isDragging: false,
      offsetX: 0,
      offsetY: 0,
      velocityX: 0,
      velocityY: 0
    });
  }, []);
  var _panResponder_panHandlers;
  return {
    dragState,
    gestureHandlers: (_panResponder_panHandlers = panResponder === null || panResponder === void 0 ? void 0 : panResponder.panHandlers) !== null && _panResponder_panHandlers !== void 0 ? _panResponder_panHandlers : {},
    resetDrag
  };
}
