// @tamagui/web/dist/esm/views/Theme.mjs
import { isWeb } from "@tamagui/constants";
import React, { Children, cloneElement, forwardRef, isValidElement, useRef } from "react";
import { variableToString } from "../createVariable.mjs";
import { useThemeWithState } from "../hooks/useTheme.mjs";
import { getThemeState, hasThemeUpdatingProps, ThemeStateContext } from "../hooks/useThemeState.mjs";
import { ThemeDebug } from "./ThemeDebug.mjs";
import { jsx } from "react/jsx-runtime";
const Theme = forwardRef(function Theme2(props, ref) {
  "use no memo";

  if (props.disable) {
    return props.children;
  }
  const {
    passThrough
  } = props;
  const isRoot = !!props["_isRoot"];
  const [_, themeState] = useThemeWithState(props, isRoot);
  const disableDirectChildTheme = props["disable-child-theme"];
  let finalChildren = disableDirectChildTheme ? Children.map(props.children, child => passThrough || !isValidElement(child) ? child : cloneElement(child, {
    ["data-disable-theme"]: true
  })) : props.children;
  if (ref) {
    try {
      React.Children.only(finalChildren);
      finalChildren = cloneElement(finalChildren, {
        ref
      });
    } catch {}
  }
  const stateRef = useRef({
    hasEverThemed: false
  });
  return getThemedChildren(themeState, finalChildren, props, isRoot, stateRef, passThrough);
});
Theme["avoidForwardRef"] = true;
function getThemedChildren(themeState, children, props, isRoot = false, stateRef, passThrough = false) {
  const {
    shallow,
    forceClassName
  } = props;
  const state = stateRef.current;
  let hasEverThemed = state.hasEverThemed;
  let shouldRenderChildrenWithTheme = hasEverThemed || themeState.isNew || isRoot || hasThemeUpdatingProps(props);
  if (process.env.NODE_ENV === "development" && props.debug === "visualize") {
    children = /* @__PURE__ */jsx(ThemeDebug, {
      themeState,
      themeProps: props,
      children
    });
  }
  if (!shouldRenderChildrenWithTheme) {
    return children;
  }
  children = /* @__PURE__ */jsx(ThemeStateContext.Provider, {
    value: themeState.id,
    children
  });
  const {
    isInverse,
    name
  } = themeState;
  const requiresExtraWrapper = isInverse || forceClassName;
  if (!state.hasEverThemed) {
    state.hasEverThemed = true;
  }
  if (requiresExtraWrapper ||
  // if the theme is exactly dark or light, its likely to change between dark/light
  // and that would require wrapping which would re-parent, so to avoid re-parenting do this
  themeState.name === "dark" || themeState.name === "light") {
    state.hasEverThemed = "wrapped";
  }
  if (shallow) {
    if (!themeState.parentId) {} else {
      const parentState = getThemeState(themeState.isNew ? themeState.id : themeState.parentId);
      if (!parentState) throw new Error(`\u203C\uFE0F010`);
      children = Children.toArray(children).map(child => {
        return isValidElement(child) ? passThrough ? child : cloneElement(child, void 0, /* @__PURE__ */jsx(Theme, {
          name: parentState.name,
          children: child.props.children
        })) : child;
      });
    }
  }
  if (process.env.NODE_ENV === "development") {
    if (!passThrough && props.debug) {
      console.warn(` getThemedChildren`, {
        requiresExtraWrapper,
        forceClassName,
        themeState,
        state,
        themeSpanProps: getThemeClassNameAndColor(themeState, props, isRoot)
      });
    }
  }
  if (forceClassName === false) {
    return children;
  }
  if (isWeb) {
    const baseStyle = props.contain ? inertContainedStyle : inertStyle;
    const {
      className = "",
      color
    } = passThrough ? {} : getThemeClassNameAndColor(themeState, props, isRoot);
    children = /* @__PURE__ */jsx("span", {
      className: `${className} is_Theme`,
      style: passThrough ? baseStyle : {
        color,
        ...baseStyle
      },
      children
    });
    if (state.hasEverThemed === "wrapped") {
      const className2 = requiresExtraWrapper ? `${name.startsWith("light") ? "t_light" : name.startsWith("dark") ? "t_dark" : ""} _dsp_contents` : `_dsp_contents`;
      children = /* @__PURE__ */jsx("span", {
        className: className2,
        children
      });
    }
    return children;
  }
  return children;
}
const inertStyle = {
  display: "contents"
};
const inertContainedStyle = {
  display: "contents",
  contain: "strict"
};
const empty = {
  className: "",
  color: void 0
};
function getThemeClassNameAndColor(themeState, props, isRoot = false) {
  if (!themeState.isNew && !props.forceClassName) {
    return empty;
  }
  const themeColor = themeState?.theme && themeState.isNew ? variableToString(themeState.theme.color) : "";
  const style = themeColor ? {
    color: themeColor
  } : void 0;
  const themeClassName = themeState.name.replace(schemePrefix, "");
  const themeNameParts = themeClassName.split("_");
  let themeClasses = `t_${themeClassName}`;
  if (themeNameParts.length > 1) {
    const hierarchyClasses = [];
    for (let i = 1; i <= themeNameParts.length; i++) {
      hierarchyClasses.push(`t_${themeNameParts.slice(0, i).join("_")}`);
    }
    themeClasses = hierarchyClasses.join(" ");
  }
  const className = `${isRoot ? "" : "t_sub_theme"} ${themeClasses}`;
  return {
    color: themeColor,
    className
  };
}
const schemePrefix = /^(dark|light)_/;
export { Theme, getThemedChildren };
