// @tamagui/web/dist/esm/views/Theme.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { isWeb } from "@tamagui/constants";
import React, { Children, cloneElement, forwardRef, isValidElement, useRef } from "react";
import { variableToString } from "../createVariable.native.js";
import { useThemeWithState } from "../hooks/useTheme.native.js";
import { getThemeState, hasThemeUpdatingProps, ThemeStateContext } from "../hooks/useThemeState.native.js";
import { ThemeDebug } from "./ThemeDebug.native.js";
var Theme = /* @__PURE__ */forwardRef(function Theme2(props, ref) {
  "use no memo";

  if (props.disable) {
    return props.children;
  }
  var {
    passThrough
  } = props;
  var isRoot = !!props["_isRoot"];
  var [_, themeState] = useThemeWithState(props, isRoot);
  var disableDirectChildTheme = props["disable-child-theme"];
  var finalChildren = disableDirectChildTheme ? Children.map(props.children, function (child) {
    return passThrough || ! /* @__PURE__ */isValidElement(child) ? child : /* @__PURE__ */cloneElement(child, {
      ["data-disable-theme"]: true
    });
  }) : props.children;
  if (ref) {
    try {
      React.Children.only(finalChildren);
      finalChildren = /* @__PURE__ */cloneElement(finalChildren, {
        ref
      });
    } catch (e) {}
  }
  var stateRef = useRef({
    hasEverThemed: false
  });
  return getThemedChildren(themeState, finalChildren, props, isRoot, stateRef, passThrough);
});
Theme["avoidForwardRef"] = true;
function getThemedChildren(themeState, children, props) {
  var isRoot = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : false,
    stateRef = arguments.length > 4 ? arguments[4] : void 0,
    passThrough = arguments.length > 5 && arguments[5] !== void 0 ? arguments[5] : false;
  var {
    shallow,
    forceClassName
  } = props;
  var state = stateRef.current;
  var hasEverThemed = state.hasEverThemed;
  var shouldRenderChildrenWithTheme = hasEverThemed || themeState.isNew || isRoot || hasThemeUpdatingProps(props);
  if (process.env.NODE_ENV === "development" && props.debug === "visualize") {
    children = /* @__PURE__ */_jsx(ThemeDebug, {
      themeState,
      themeProps: props,
      children
    });
  }
  if (!shouldRenderChildrenWithTheme) {
    return children;
  }
  children = /* @__PURE__ */_jsx(ThemeStateContext.Provider, {
    value: themeState.id,
    children
  });
  var {
    isInverse,
    name
  } = themeState;
  var requiresExtraWrapper = isInverse || forceClassName;
  if (!state.hasEverThemed) {
    state.hasEverThemed = true;
  }
  if (requiresExtraWrapper ||
  // if the theme is exactly dark or light, its likely to change between dark/light
  // and that would require wrapping which would re-parent, so to avoid re-parenting do this
  themeState.name === "dark" || themeState.name === "light") {
    state.hasEverThemed = "wrapped";
  }
  if (shallow) {
    if (!themeState.parentId) {} else {
      var parentState = getThemeState(themeState.isNew ? themeState.id : themeState.parentId);
      if (!parentState) throw new Error(`\u203C\uFE0F010`);
      children = Children.toArray(children).map(function (child) {
        return /* @__PURE__ */isValidElement(child) ? passThrough ? child : /* @__PURE__ */cloneElement(child, void 0, /* @__PURE__ */_jsx(Theme, {
          name: parentState.name,
          children: child.props.children
        })) : child;
      });
    }
  }
  if (process.env.NODE_ENV === "development") {
    if (!passThrough && props.debug) {
      console.warn(` getThemedChildren`, {
        requiresExtraWrapper,
        forceClassName,
        themeState,
        state,
        themeSpanProps: getThemeClassNameAndColor(themeState, props, isRoot)
      });
    }
  }
  if (forceClassName === false) {
    return children;
  }
  if (isWeb) {
    var baseStyle = props.contain ? inertContainedStyle : inertStyle;
    var {
      className = "",
      color
    } = passThrough ? {} : getThemeClassNameAndColor(themeState, props, isRoot);
    children = /* @__PURE__ */_jsx("span", {
      className: `${className} is_Theme`,
      style: passThrough ? baseStyle : {
        color,
        ...baseStyle
      },
      children
    });
    if (state.hasEverThemed === "wrapped") {
      var className1 = requiresExtraWrapper ? `${name.startsWith("light") ? "t_light" : name.startsWith("dark") ? "t_dark" : ""} _dsp_contents` : `_dsp_contents`;
      children = /* @__PURE__ */_jsx("span", {
        className: className1,
        children
      });
    }
    return children;
  }
  return children;
}
var inertStyle = {
  display: "contents"
};
var inertContainedStyle = {
  display: "contents",
  contain: "strict"
};
var empty = {
  className: "",
  color: void 0
};
function getThemeClassNameAndColor(themeState, props) {
  var isRoot = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : false;
  if (!themeState.isNew && !props.forceClassName) {
    return empty;
  }
  var themeColor = (themeState === null || themeState === void 0 ? void 0 : themeState.theme) && themeState.isNew ? variableToString(themeState.theme.color) : "";
  var style = themeColor ? {
    color: themeColor
  } : void 0;
  var themeClassName = themeState.name.replace(schemePrefix, "");
  var themeNameParts = themeClassName.split("_");
  var themeClasses = `t_${themeClassName}`;
  if (themeNameParts.length > 1) {
    var hierarchyClasses = [];
    for (var i = 1; i <= themeNameParts.length; i++) {
      hierarchyClasses.push(`t_${themeNameParts.slice(0, i).join("_")}`);
    }
    themeClasses = hierarchyClasses.join(" ");
  }
  var className = `${isRoot ? "" : "t_sub_theme"} ${themeClasses}`;
  return {
    color: themeColor,
    className
  };
}
var schemePrefix = /^(dark|light)_/;
export { Theme, getThemedChildren };
