// @tamagui/radio-headless/dist/jsx/useRadioGroup.mjs
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { registerFocusable } from "@tamagui/focusable";
import { composeEventHandlers } from "@tamagui/helpers";
import { useLabelContext } from "@tamagui/label";
import { useControllableState } from "@tamagui/use-controllable-state";
import { useContext, useEffect, useRef, useState } from "react";
import { BubbleInput } from "./BubbleInput.mjs";
import { getState } from "./utils.mjs";
import { jsx } from "react/jsx-runtime";
function useRadioGroup(params) {
  const {
    value: valueProp,
    onValueChange,
    defaultValue,
    required,
    disabled,
    name,
    native,
    accentColor,
    orientation,
    ref
  } = params;
  const [value, setValue] = useControllableState({
    prop: valueProp,
    defaultProp: defaultValue,
    onChange: onValueChange
  });
  return {
    providerValue: {
      value,
      onChange: setValue,
      required,
      disabled,
      name,
      native,
      accentColor
    },
    frameAttrs: {
      role: "radiogroup",
      "aria-orientation": orientation,
      "data-disabled": disabled ? "" : void 0
    },
    rovingFocusGroupAttrs: {
      orientation,
      loop: true
    }
  };
}
const ARROW_KEYS = ["ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight"];
const useRadioGroupItem = params => {
  const {
    radioGroupContext,
    value,
    labelledBy: ariaLabelledby,
    disabled: itemDisabled,
    ref: refProp,
    id,
    onPress,
    onKeyDown,
    onFocus
  } = params;
  const {
    value: groupValue,
    disabled,
    required,
    onChange,
    name,
    native,
    accentColor
  } = useContext(radioGroupContext);
  const [button, setButton] = useState(null);
  const hasConsumerStoppedPropagationRef = useRef(false);
  const ref = useRef(null);
  const composedRefs = useComposedRefs(refProp, node => setButton(node), ref);
  const isArrowKeyPressedRef = useRef(false);
  const isFormControl = isWeb ? button ? Boolean(button.closest("form")) : true : false;
  const checked = groupValue === value;
  const labelId = useLabelContext(button);
  const labelledBy = ariaLabelledby || labelId;
  useEffect(() => {
    if (isWeb) {
      const handleKeyDown = event => {
        if (ARROW_KEYS.includes(event.key)) isArrowKeyPressedRef.current = true;
      };
      const handleKeyUp = () => {
        isArrowKeyPressedRef.current = false;
      };
      document.addEventListener("keydown", handleKeyDown);
      document.addEventListener("keyup", handleKeyUp);
      return () => {
        document.removeEventListener("keydown", handleKeyDown);
        document.removeEventListener("keyup", handleKeyUp);
      };
    }
  }, []);
  const isDisabled = disabled || itemDisabled;
  return {
    providerValue: {
      checked
    },
    checked,
    isFormControl,
    bubbleInput: /* @__PURE__ */jsx(BubbleInput, {
      isHidden: !native,
      control: button,
      bubbles: !hasConsumerStoppedPropagationRef.current,
      name,
      value,
      checked,
      required,
      disabled: isDisabled,
      ...(isWeb && native && {
        accentColor,
        id
      })
    }),
    native,
    frameAttrs: {
      "data-state": getState(checked),
      "data-disabled": isDisabled ? "" : void 0,
      role: "radio",
      "aria-labelledby": labelledBy,
      "aria-checked": checked,
      "aria-required": required,
      disabled: isDisabled,
      ref: composedRefs,
      ...(isWeb && {
        type: "button",
        value
      }),
      id,
      onPress: composeEventHandlers(onPress, event => {
        if (!checked) onChange?.(value);
        if (isFormControl) {
          hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
          if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
        }
      }),
      ...(isWeb && {
        onKeyDown: composeEventHandlers(onKeyDown, event => {
          if (event.key === "Enter" || event.key === " ") {
            if (!checked) onChange?.(value);
          }
        }),
        onFocus: composeEventHandlers(onFocus, () => {
          if (isArrowKeyPressedRef.current) ref.current?.click();
        })
      })
    },
    rovingFocusGroupAttrs: {
      asChild: "except-style",
      focusable: !isDisabled,
      active: checked
    }
  };
};
function useRadioGroupItemIndicator(params) {
  const {
    radioGroupItemContext,
    disabled,
    ...rest
  } = params;
  const {
    checked
  } = useContext(radioGroupItemContext);
  return {
    checked,
    "data-state": getState(checked),
    "data-disabled": disabled ? "" : void 0,
    ...rest
  };
}
export { useRadioGroup, useRadioGroupItem, useRadioGroupItemIndicator };
