// @tamagui/animate-presence/dist/esm/AnimatePresence.mjs
import { useInsertionEffect } from "react";
import { useConstant } from "@tamagui/use-constant";
import { useForceUpdate } from "@tamagui/use-force-update";
import { Children, isValidElement, useContext, useMemo, useRef, useState } from "react";
import { LayoutGroupContext } from "./LayoutGroupContext.mjs";
import { PresenceChild } from "./PresenceChild.mjs";
import { Fragment, jsx } from "react/jsx-runtime";
const getChildKey = child => {
  return child.key || (() => {
    const ct = child.type;
    const defaultName = ct["displayName"] || ct["name"] || "";
    if (ct && typeof ct === "object" && "staticConfig" in ct) {
      return ct.staticConfig.componentName || defaultName;
    }
    return defaultName;
  })();
};
function onlyElements(children) {
  const filtered = [];
  Children.forEach(children, child => {
    if (isValidElement(child)) filtered.push(child);
  });
  return filtered;
}
const AnimatePresence = ({
  children,
  enterVariant,
  exitVariant,
  enterExitVariant,
  initial = true,
  onExitComplete,
  exitBeforeEnter,
  mode,
  presenceAffectsLayout = true,
  custom,
  passThrough
}) => {
  const effectiveMode = mode ?? (exitBeforeEnter ? "wait" : "sync");
  const presentChildren = useMemo(() => onlyElements(children), [children]);
  const presentKeys = presentChildren.map(getChildKey);
  const isInitialRender = useRef(true);
  const frozenCustomRef = useRef(/* @__PURE__ */new Map());
  const pendingPresentChildren = useRef(presentChildren);
  const exitComplete = useConstant(() => /* @__PURE__ */new Map());
  const [diffedChildren, setDiffedChildren] = useState(presentChildren);
  const [renderedChildren, setRenderedChildren] = useState(presentChildren);
  const forceRender = useContext(LayoutGroupContext).forceRender ?? useForceUpdate();
  if (passThrough) {
    return /* @__PURE__ */jsx(Fragment, {
      children
    });
  }
  useInsertionEffect(() => {
    isInitialRender.current = false;
    pendingPresentChildren.current = presentChildren;
    for (let i = 0; i < renderedChildren.length; i++) {
      const key = getChildKey(renderedChildren[i]);
      if (!presentKeys.includes(key)) {
        if (exitComplete.get(key) !== true) {
          exitComplete.set(key, false);
        }
      } else {
        exitComplete.delete(key);
        frozenCustomRef.current.delete(key);
      }
    }
  }, [renderedChildren, presentKeys.length, presentKeys.join("-")]);
  if (presentChildren !== diffedChildren) {
    let nextChildren = [...presentChildren];
    for (let i = 0; i < renderedChildren.length; i++) {
      const child = renderedChildren[i];
      const key = getChildKey(child);
      if (!presentKeys.includes(key)) {
        nextChildren.splice(i, 0, child);
        if (!frozenCustomRef.current.has(key)) {
          frozenCustomRef.current.set(key, custom);
        }
      }
    }
    const exitingChildren = renderedChildren.filter(child => !presentKeys.includes(getChildKey(child)));
    if (effectiveMode === "wait" && exitingChildren.length) {
      nextChildren = exitingChildren;
    }
    setRenderedChildren(onlyElements(nextChildren));
    setDiffedChildren(presentChildren);
    return null;
  }
  return /* @__PURE__ */jsx(Fragment, {
    children: renderedChildren.map(child => {
      const key = getChildKey(child);
      const isPresent = presentChildren === renderedChildren || presentKeys.includes(key);
      const onExit = () => {
        if (exitComplete.has(key)) {
          exitComplete.set(key, true);
        } else {
          return;
        }
        let isEveryExitComplete = true;
        exitComplete.forEach(isExitComplete => {
          if (!isExitComplete) isEveryExitComplete = false;
        });
        if (isEveryExitComplete) {
          forceRender?.();
          setRenderedChildren(pendingPresentChildren.current);
          onExitComplete?.();
        }
      };
      return /* @__PURE__ */jsx(PresenceChild, {
        isPresent,
        initial: !isInitialRender.current || initial ? void 0 : false,
        custom: isPresent ? custom : frozenCustomRef.current.get(key) ?? custom,
        presenceAffectsLayout,
        enterExitVariant,
        enterVariant,
        exitVariant,
        onExitComplete: isPresent ? void 0 : onExit,
        children: child
      }, key);
    })
  });
};
AnimatePresence.displayName = "AnimatePresence";
export { AnimatePresence };
