// @tamagui/menu/dist/jsx/createNonNativeMenu.mjs
import { createBaseMenu } from "@tamagui/create-menu";
import { usePopperContextSlow } from "@tamagui/popper";
import { ScrollView } from "@tamagui/scroll-view";
import { useControllableState } from "@tamagui/use-controllable-state";
import { composeEventHandlers, composeRefs, createStyledContext, isAndroid, isWeb, Slot, styled, useEvent, useIsTouchDevice, View, withStaticProperties } from "@tamagui/web";
import * as React from "react";
import { useId } from "react";
import { jsx } from "react/jsx-runtime";
const DROPDOWN_MENU_CONTEXT = "MenuContext";
function useMenuTriggerSetup(open) {
  const triggerStateSettersRef = React.useRef(/* @__PURE__ */new Map());
  const activeTriggerIdRef = React.useRef(null);
  const setActiveTrigger = useEvent(id => {
    const prevId = activeTriggerIdRef.current;
    if (prevId === id) return;
    if (prevId) {
      triggerStateSettersRef.current.get(prevId)?.(false);
    }
    activeTriggerIdRef.current = id;
    if (id && open) {
      triggerStateSettersRef.current.get(id)?.(true);
    }
  });
  const registerTrigger = useEvent((id, setOpenState) => {
    triggerStateSettersRef.current.set(id, setOpenState);
    setOpenState(activeTriggerIdRef.current === id && open);
  });
  const unregisterTrigger = useEvent(id => {
    triggerStateSettersRef.current.delete(id);
    if (activeTriggerIdRef.current === id) {
      activeTriggerIdRef.current = null;
    }
  });
  React.useEffect(() => {
    if (!open) {
      setActiveTrigger(null);
      return;
    }
    const activeId = activeTriggerIdRef.current;
    if (activeId) {
      triggerStateSettersRef.current.get(activeId)?.(true);
    }
  }, [open, setActiveTrigger]);
  return {
    setActiveTrigger,
    registerTrigger,
    unregisterTrigger
  };
}
function createNonNativeMenu(params) {
  const {
    Menu
  } = createBaseMenu(params);
  const DROPDOWN_MENU_NAME = "Menu";
  const {
    Provider: MenuProvider,
    useStyledContext: useMenuContext
  } = createStyledContext();
  const MenuComp = props => {
    const {
      scope,
      children,
      dir,
      open: openProp,
      defaultOpen,
      onOpenChange,
      modal = true,
      ...rest
    } = props;
    const triggerRef = React.useRef(null);
    const [open = false, setOpen] = useControllableState({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    const openRef = React.useRef(open);
    openRef.current = open;
    const {
      setActiveTrigger,
      registerTrigger,
      unregisterTrigger
    } = useMenuTriggerSetup(open);
    return /* @__PURE__ */jsx(MenuProvider, {
      scope,
      triggerId: useId(),
      triggerRef,
      contentId: useId(),
      openRef,
      onOpenChange: React.useCallback(nextOpen => setOpen(nextOpen), [setOpen]),
      onOpenToggle: React.useCallback(() => setOpen(prevOpen => !prevOpen), [setOpen]),
      modal,
      setActiveTrigger,
      registerTrigger,
      unregisterTrigger,
      children: /* @__PURE__ */jsx(Menu, {
        scope: scope || DROPDOWN_MENU_CONTEXT,
        open,
        onOpenChange: setOpen,
        dir,
        modal,
        ...rest,
        children
      })
    });
  };
  MenuComp.displayName = DROPDOWN_MENU_NAME;
  const TRIGGER_NAME = "MenuTrigger";
  const MenuTriggerFrame = Menu.Anchor;
  const MenuTrigger = View.styleable((props, forwardedRef) => {
    const {
      scope,
      asChild,
      children,
      disabled = false,
      onKeydown,
      ...triggerProps
    } = props;
    const context = useMenuContext(scope);
    const popperCtx = usePopperContextSlow(scope || DROPDOWN_MENU_CONTEXT);
    const Comp = asChild ? Slot : View;
    const isTouchDevice = useIsTouchDevice();
    const triggerElRef = React.useRef(null);
    const triggerId = React.useId();
    const [triggerOpen, setTriggerOpen] = React.useState(false);
    const {
      registerTrigger,
      unregisterTrigger
    } = context;
    React.useEffect(() => {
      registerTrigger(triggerId, setTriggerOpen);
      return () => unregisterTrigger(triggerId);
    }, [registerTrigger, unregisterTrigger, triggerId]);
    const activateSelf = React.useCallback(() => {
      context.setActiveTrigger(triggerId);
      const el = triggerElRef.current;
      if (el) {
        context.triggerRef.current = el;
        if (el instanceof HTMLElement) {
          popperCtx.refs?.setReference(el);
          requestAnimationFrame(() => popperCtx.update?.());
        }
      }
    }, [context, triggerId, popperCtx]);
    const pressEvent = isWeb ? isTouchDevice ? "onClick" : "onPointerDown" : "onPress";
    return /* @__PURE__ */jsx(MenuTriggerFrame, {
      asChild: true,
      componentName: TRIGGER_NAME,
      scope: scope || DROPDOWN_MENU_CONTEXT,
      children: /* @__PURE__ */jsx(Comp, {
        role: "button",
        id: context.triggerId,
        "aria-haspopup": "menu",
        "aria-expanded": triggerOpen,
        "aria-controls": triggerOpen ? context.contentId : void 0,
        "data-state": triggerOpen ? "open" : "closed",
        "data-disabled": disabled ? "" : void 0,
        "aria-disabled": disabled || void 0,
        ref: composeRefs(forwardedRef, context.triggerRef, triggerElRef),
        ...{
          [pressEvent]: composeEventHandlers(
          //@ts-ignore
          props[pressEvent], event => {
            if (!disabled) {
              if (isWeb && event instanceof PointerEvent && event.button !== 0 && event.ctrlKey === true) return;
              if (context.openRef.current) {
                context.setActiveTrigger(null);
              } else {
                activateSelf();
              }
              context.onOpenToggle();
              if (!context.openRef.current) event.preventDefault();
            }
          })
        },
        ...(isWeb && {
          onKeyDown: composeEventHandlers(onKeydown, event => {
            if (disabled) return;
            if (["Enter", " "].includes(event.key)) {
              if (context.openRef.current) {
                context.setActiveTrigger(null);
              } else {
                activateSelf();
              }
              context.onOpenToggle();
            }
            if (event.key === "ArrowDown") {
              activateSelf();
              context.onOpenChange(true);
            }
            if (["Enter", " ", "ArrowDown"].includes(event.key)) event.preventDefault();
          })
        }),
        ...triggerProps,
        children
      })
    });
  });
  MenuTrigger.displayName = TRIGGER_NAME;
  const PORTAL_NAME = "MenuPortal";
  const MenuPortal = props => {
    const {
      scope,
      children,
      ...portalProps
    } = props;
    const context = isAndroid ? useMenuContext(scope) : null;
    const content = isAndroid ? /* @__PURE__ */jsx(MenuProvider, {
      ...context,
      children
    }) : children;
    return /* @__PURE__ */jsx(Menu.Portal, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...portalProps,
      children: content
    });
  };
  MenuPortal.displayName = PORTAL_NAME;
  const CONTENT_NAME = "MenuContent";
  const MenuContent = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...contentProps
    } = props;
    const context = useMenuContext(scope);
    const hasInteractedOutsideRef = React.useRef(false);
    return /* @__PURE__ */jsx(Menu.Content, {
      id: context.contentId,
      "aria-labelledby": context.triggerId,
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...contentProps,
      ref: forwardedRef,
      onCloseAutoFocus: composeEventHandlers(props.onCloseAutoFocus, event => {
        if (!hasInteractedOutsideRef.current) {
          requestAnimationFrame(() => {
            const activeEl = document.activeElement;
            if (!activeEl || activeEl === document.body) {
              context.triggerRef.current?.focus();
            }
          });
        }
        hasInteractedOutsideRef.current = false;
        event.preventDefault();
      }),
      onInteractOutside: composeEventHandlers(props.onInteractOutside, event => {
        const originalEvent = event.detail.originalEvent;
        const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
        const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
        if (!context.modal || isRightClick) hasInteractedOutsideRef.current = true;
      }),
      style: isWeb ? {
        ...props.style,
        ...{
          "--tamagui-menu-content-transform-origin": "var(--tamagui-popper-transform-origin)",
          "--tamagui-menu-content-available-width": "var(--tamagui-popper-available-width)",
          "--tamagui-menu-content-available-height": "var(--tamagui-popper-available-height)",
          "--tamagui-menu-trigger-width": "var(--tamagui-popper-anchor-width)",
          "--tamagui-menu-trigger-height": "var(--tamagui-popper-anchor-height)"
        }
      } : props.style
    });
  });
  MenuContent.displayName = CONTENT_NAME;
  const DROPDOWN_MENU_SUB_NAME = "MenuSub";
  const MenuSub = props => {
    const {
      scope,
      children,
      open: openProp,
      onOpenChange,
      defaultOpen,
      ...rest
    } = props;
    const [open = false, setOpen] = useControllableState({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    return /* @__PURE__ */jsx(Menu.Sub, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      open,
      onOpenChange: setOpen,
      ...rest,
      children
    });
  };
  MenuSub.displayName = DROPDOWN_MENU_SUB_NAME;
  const SUB_CONTENT_NAME = "MenuSubContent";
  const MenuSubContent = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...subContentProps
    } = props;
    return /* @__PURE__ */jsx(Menu.SubContent, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...subContentProps,
      ref: forwardedRef,
      style: isWeb ? {
        ...props.style,
        ...{
          "--tamagui-menu-content-transform-origin": "var(--tamagui-popper-transform-origin)",
          "--tamagui-menu-content-available-width": "var(--tamagui-popper-available-width)",
          "--tamagui-menu-content-available-height": "var(--tamagui-popper-available-height)",
          "--tamagui-menu-trigger-width": "var(--tamagui-popper-anchor-width)",
          "--tamagui-menu-trigger-height": "var(--tamagui-popper-anchor-height)"
        }
      } : null
    });
  });
  MenuSubContent.displayName = SUB_CONTENT_NAME;
  const MenuScrollView = styled(ScrollView, {
    flexShrink: 1,
    alignSelf: "stretch",
    showsHorizontalScrollIndicator: false,
    showsVerticalScrollIndicator: false,
    "$platform-web": {
      maxHeight: "var(--tamagui-menu-content-available-height)"
    }
  });
  const Group = Menu.Group;
  const Label = Menu.Label;
  const Item = Menu.Item;
  const CheckboxItem = Menu.CheckboxItem;
  const RadioGroup = Menu.RadioGroup;
  const RadioItem = Menu.RadioItem;
  const ItemIndicator = Menu.ItemIndicator;
  const Separator = Menu.Separator;
  const Arrow = Menu.Arrow;
  const SubTrigger = Menu.SubTrigger;
  const ItemTitle = Menu.ItemTitle;
  const ItemSubtitle = Menu.ItemSubtitle;
  const ItemImage = Menu.ItemImage;
  const ItemIcon = Menu.ItemIcon;
  return withStaticProperties(MenuComp, {
    Root: MenuComp,
    Trigger: MenuTrigger,
    Portal: MenuPortal,
    Content: MenuContent,
    Group,
    Label,
    Item,
    CheckboxItem,
    RadioGroup,
    RadioItem,
    ItemIndicator,
    Separator,
    Arrow,
    Sub: MenuSub,
    SubTrigger,
    SubContent: MenuSubContent,
    ItemTitle,
    ItemSubtitle,
    ItemImage,
    ItemIcon,
    ScrollView: MenuScrollView
  });
}
export { DROPDOWN_MENU_CONTEXT, createNonNativeMenu };
