// @tamagui/spacer/dist/esm/spacedChildren.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createComponent } from "@tamagui/web";
import React from "react";
import { Spacer } from "./Spacer.native.js";
import { isUnspaced } from "./Unspaced.native.js";
function spacedChildren(props) {
  var _childrenList__type, _childrenList_;
  var {
    isZStack,
    children,
    space,
    direction,
    spaceFlex,
    separator,
    ensureKeys
  } = props;
  var hasSpace = !!(space || spaceFlex);
  var hasSeparator = !(separator === void 0 || separator === null);
  var areChildrenArray = Array.isArray(children);
  if (!ensureKeys && !(hasSpace || hasSeparator || isZStack)) {
    return children;
  }
  var childrenList = areChildrenArray ? children : React.Children.toArray(children);
  var len = childrenList.length;
  if (len <= 1 && !isZStack && !((_childrenList_ = childrenList[0]) === null || _childrenList_ === void 0 ? void 0 : (_childrenList__type = _childrenList_["type"]) === null || _childrenList__type === void 0 ? void 0 : _childrenList__type["shouldForwardSpace"])) {
    return children;
  }
  var final = [];
  var _iteratorNormalCompletion = true,
    _didIteratorError = false,
    _iteratorError = void 0;
  try {
    for (var _iterator = childrenList.entries()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var [index, child] = _step.value;
      var _child_type;
      var isEmpty = child === null || child === void 0 || Array.isArray(child) && child.length === 0;
      if (!isEmpty && /* @__PURE__ */React.isValidElement(child) && ((_child_type = child.type) === null || _child_type === void 0 ? void 0 : _child_type["shouldForwardSpace"])) {
        child = /* @__PURE__ */React.cloneElement(child, {
          space,
          spaceFlex,
          separator,
          key: child.key
        });
      }
      if (isEmpty || !child || child["key"] && !isZStack) {
        final.push(child);
      } else {
        final.push(/* @__PURE__ */_jsx(React.Fragment, {
          children: isZStack ? /* @__PURE__ */_jsx(AbsoluteFill, {
            children: child
          }) : child
        }, `${index}0t`));
      }
      if (isUnspaced(child) && index === 0) continue;
      if (isZStack) continue;
      var next = childrenList[index + 1];
      if (next && !isEmpty && !isUnspaced(next)) {
        if (separator) {
          if (hasSpace) {
            final.push(createSpacer({
              key: `_${index}_00t`,
              direction,
              space,
              spaceFlex
            }));
          }
          final.push(/* @__PURE__ */_jsx(React.Fragment, {
            children: separator
          }, `${index}03t`));
          if (hasSpace) {
            final.push(createSpacer({
              key: `_${index}01t`,
              direction,
              space,
              spaceFlex
            }));
          }
        } else {
          final.push(createSpacer({
            key: `_${index}02t`,
            direction,
            space,
            spaceFlex
          }));
        }
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return != null) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }
  if (process.env.NODE_ENV === "development") {
    if (props.debug) {
      console.info(`  Spaced children`, final, props);
    }
  }
  return final;
}
function createSpacer(param) {
  var {
    key,
    direction,
    space,
    spaceFlex
  } = param;
  return /* @__PURE__ */_jsx(Spacer, {
    // @ts-ignore
    size: space,
    // @ts-ignore
    direction,
    ...(typeof spaceFlex !== "undefined" && {
      flex: spaceFlex === true ? 1 : spaceFlex === false ? 0 : spaceFlex
    })
  }, key);
}
var AbsoluteFill = createComponent({
  defaultProps: {
    display: "flex",
    flexDirection: "column",
    position: "absolute",
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    pointerEvents: "box-none"
  }
});
export { spacedChildren };
