// @tamagui/spacer/dist/esm/spacedChildren.mjs
import { createComponent } from "@tamagui/web";
import React from "react";
import { Spacer } from "./Spacer.mjs";
import { isUnspaced } from "./Unspaced.mjs";
import { jsx } from "react/jsx-runtime";
function spacedChildren(props) {
  const {
    isZStack,
    children,
    space,
    direction,
    spaceFlex,
    separator,
    ensureKeys
  } = props;
  const hasSpace = !!(space || spaceFlex);
  const hasSeparator = !(separator === void 0 || separator === null);
  const areChildrenArray = Array.isArray(children);
  if (!ensureKeys && !(hasSpace || hasSeparator || isZStack)) {
    return children;
  }
  const childrenList = areChildrenArray ? children : React.Children.toArray(children);
  const len = childrenList.length;
  if (len <= 1 && !isZStack && !childrenList[0]?.["type"]?.["shouldForwardSpace"]) {
    return children;
  }
  const final = [];
  for (let [index, child] of childrenList.entries()) {
    const isEmpty = child === null || child === void 0 || Array.isArray(child) && child.length === 0;
    if (!isEmpty && React.isValidElement(child) && child.type?.["shouldForwardSpace"]) {
      child = React.cloneElement(child, {
        space,
        spaceFlex,
        separator,
        key: child.key
      });
    }
    if (isEmpty || !child || child["key"] && !isZStack) {
      final.push(child);
    } else {
      final.push(/* @__PURE__ */jsx(React.Fragment, {
        children: isZStack ? /* @__PURE__ */jsx(AbsoluteFill, {
          children: child
        }) : child
      }, `${index}0t`));
    }
    if (isUnspaced(child) && index === 0) continue;
    if (isZStack) continue;
    const next = childrenList[index + 1];
    if (next && !isEmpty && !isUnspaced(next)) {
      if (separator) {
        if (hasSpace) {
          final.push(createSpacer({
            key: `_${index}_00t`,
            direction,
            space,
            spaceFlex
          }));
        }
        final.push(/* @__PURE__ */jsx(React.Fragment, {
          children: separator
        }, `${index}03t`));
        if (hasSpace) {
          final.push(createSpacer({
            key: `_${index}01t`,
            direction,
            space,
            spaceFlex
          }));
        }
      } else {
        final.push(createSpacer({
          key: `_${index}02t`,
          direction,
          space,
          spaceFlex
        }));
      }
    }
  }
  if (process.env.NODE_ENV === "development") {
    if (props.debug) {
      console.info(`  Spaced children`, final, props);
    }
  }
  return final;
}
function createSpacer({
  key,
  direction,
  space,
  spaceFlex
}) {
  return /* @__PURE__ */jsx(Spacer, {
    size: space,
    direction,
    ...(typeof spaceFlex !== "undefined" && {
      flex: spaceFlex === true ? 1 : spaceFlex === false ? 0 : spaceFlex
    })
  }, key);
}
const AbsoluteFill = createComponent({
  defaultProps: {
    display: "flex",
    flexDirection: "column",
    position: "absolute",
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    pointerEvents: "box-none"
  }
});
export { spacedChildren };
