// @tamagui/cubic-bezier-animator/dist/esm/index.native.js
import { bezier } from "./cubicBezier.native.js";
function animate(param) {
  var start = null;
  var easing = param.cubicBezier ? bezier(...param.cubicBezier) : function (v) {
    return v;
  };
  var {
    x: fromX,
    y: fromY,
    scaleX: fromScaleX,
    scaleY: fromScaleY
  } = param.from;
  var {
    x: toX,
    y: toY,
    scaleX: toScaleX,
    scaleY: toScaleY
  } = param.to;
  function frame(timestamp) {
    if (!start) start = timestamp;
    var progress = timestamp - start;
    var x = toX !== void 0 ? fromX + (toX - fromX) * easing(progress / param.duration) : void 0;
    var y = toY !== void 0 ? fromY + (toY - fromY) * easing(progress / param.duration) : void 0;
    var scaleX = toScaleX !== void 0 ? fromScaleX + (toScaleX - fromScaleX) * easing(progress / param.duration) : void 0;
    var scaleY = toScaleY !== void 0 ? fromScaleY + (toScaleY - fromScaleY) * easing(progress / param.duration) : void 0;
    param.onUpdate({
      x,
      y,
      scaleX,
      scaleY
    });
    if (progress < param.duration) {
      requestAnimationFrame(frame);
    }
  }
  requestAnimationFrame(frame);
}
export { animate };
