// @tamagui/cubic-bezier-animator/dist/esm/index.js
import { bezier } from "./cubicBezier.mjs";
function animate(param) {
  let start = null;
  const easing = param.cubicBezier ? bezier(...param.cubicBezier) : v => v;
  const {
    x: fromX,
    y: fromY,
    scaleX: fromScaleX,
    scaleY: fromScaleY
  } = param.from;
  const {
    x: toX,
    y: toY,
    scaleX: toScaleX,
    scaleY: toScaleY
  } = param.to;
  function frame(timestamp) {
    if (!start) start = timestamp;
    const progress = timestamp - start;
    const x = toX !== void 0 ? fromX + (toX - fromX) * easing(progress / param.duration) : void 0;
    const y = toY !== void 0 ? fromY + (toY - fromY) * easing(progress / param.duration) : void 0;
    const scaleX = toScaleX !== void 0 ? fromScaleX + (toScaleX - fromScaleX) * easing(progress / param.duration) : void 0;
    const scaleY = toScaleY !== void 0 ? fromScaleY + (toScaleY - fromScaleY) * easing(progress / param.duration) : void 0;
    param.onUpdate({
      x,
      y,
      scaleX,
      scaleY
    });
    if (progress < param.duration) {
      requestAnimationFrame(frame);
    }
  }
  requestAnimationFrame(frame);
}
export { animate };
