// @tamagui/checkbox/dist/esm/createCheckbox.native.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { isIndeterminate, useCheckbox } from "@tamagui/checkbox-headless";
import { getVariableValue, isWeb, shouldRenderNativePlatform, useProps, useTheme, withStaticProperties } from "@tamagui/core";
import { registerFocusable } from "@tamagui/focusable";
import { getFontSize } from "@tamagui/font-size";
import { getSize } from "@tamagui/get-token";
import { useGetThemedIcon } from "@tamagui/helpers-tamagui";
import { useControllableState } from "@tamagui/use-controllable-state";
import React, { useMemo } from "react";
import { CheckboxFrame, CheckboxIndicatorFrame } from "./Checkbox.native.js";
import { CheckboxStyledContext } from "./CheckboxStyledContext.native.js";
var CheckboxContext = /* @__PURE__ */React.createContext({
  checked: false,
  disabled: false
});
var ensureContext = function (x) {
  if (!x.context) x.context = CheckboxContext;
};
function createCheckbox(createProps) {
  var {
    Frame = CheckboxFrame,
    Indicator = CheckboxIndicatorFrame
  } = createProps;
  ensureContext(Frame);
  ensureContext(Indicator);
  return withStaticProperties(Frame.styleable(function Checkbox(_props, forwardedRef) {
    var {
      scaleSize = .45,
      sizeAdjust = 0,
      scaleIcon,
      checked: checkedProp,
      defaultChecked,
      onCheckedChange,
      native,
      unstyled = false,
      activeStyle,
      activeTheme,
      ...props
    } = _props;
    var propsActive = useProps(props);
    var styledContext = React.useContext(CheckboxStyledContext.context);
    var adjustedSize = 0;
    var size = 0;
    if (!unstyled) {
      var _propsActive_size, _ref;
      adjustedSize = getVariableValue(getSize((_ref = (_propsActive_size = propsActive.size) !== null && _propsActive_size !== void 0 ? _propsActive_size : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size) !== null && _ref !== void 0 ? _ref : "$true", {
        shift: sizeAdjust
      }));
      size = scaleSize ? Math.round(adjustedSize * scaleSize) : adjustedSize;
    }
    var [checked = false, setChecked] = useControllableState({
      prop: checkedProp,
      defaultProp: defaultChecked,
      onChange: onCheckedChange
    });
    var {
      checkboxProps,
      checkboxRef,
      bubbleInput
    } = useCheckbox(
    // @ts-ignore
    propsActive, [checked, setChecked], forwardedRef);
    React.useEffect(function () {
      if (!props.id) return;
      if (props.disabled) return;
      return registerFocusable(props.id, {
        focusAndSelect: function () {
          setChecked === null || setChecked === void 0 || setChecked(function (value) {
            return !value;
          });
        },
        focus: function () {}
      });
    }, [props.id, props.disabled]);
    if (shouldRenderNativePlatform(native) === "web") return /* @__PURE__ */_jsx("input", {
      type: "checkbox",
      defaultChecked: isIndeterminate(checked) ? false : checked,
      tabIndex: -1,
      ref: checkboxRef,
      disabled: checkboxProps.disabled,
      style: {
        appearance: "auto",
        accentColor: "var(--color6)",
        ...checkboxProps.style
      }
    });
    var memoizedContext = useMemo(function () {
      return {
        checked,
        disabled: checkboxProps.disabled
      };
    }, [checked, checkboxProps.disabled]);
    var isActive = !!checked;
    var disabled = checkboxProps.disabled;
    var _propsActive_size1, _ref1, _ref2;
    return /* @__PURE__ */_jsx(CheckboxContext.Provider, {
      value: memoizedContext,
      children: /* @__PURE__ */_jsxs(CheckboxStyledContext.Provider, {
        size: (_ref1 = (_propsActive_size1 = propsActive.size) !== null && _propsActive_size1 !== void 0 ? _propsActive_size1 : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size) !== null && _ref1 !== void 0 ? _ref1 : "$true",
        scaleIcon: (_ref2 = scaleIcon !== null && scaleIcon !== void 0 ? scaleIcon : styledContext === null || styledContext === void 0 ? void 0 : styledContext.scaleIcon) !== null && _ref2 !== void 0 ? _ref2 : 1,
        unstyled,
        active: isActive,
        disabled,
        children: [/* @__PURE__ */_jsx(Frame, {
          render: "button",
          ref: checkboxRef,
          unstyled,
          theme: activeTheme !== null && activeTheme !== void 0 ? activeTheme : null,
          ...(isWeb && {
            type: "button"
          }),
          ...(!unstyled && {
            width: size,
            height: size,
            size
          }),
          checked,
          disabled,
          ...checkboxProps,
          ...props,
          ...(isActive && {
            ...(!unstyled && !activeStyle && {
              backgroundColor: "$backgroundActive"
            }),
            ...activeStyle
          }),
          children: propsActive.children
        }), bubbleInput]
      })
    });
  }), {
    Indicator: Indicator.styleable(function (props, forwardedRef) {
      var {
        children: childrenProp,
        forceMount,
        disablePassStyles,
        unstyled = false,
        activeStyle,
        ...indicatorProps
      } = props;
      var styledContext = CheckboxStyledContext.useStyledContext();
      var {
        active
      } = styledContext;
      var children = childrenProp;
      if (!unstyled) {
        var getThemedIcon = useGetThemedIcon({
          size: (typeof styledContext.size === "number" ? styledContext.size * .65 : getFontSize(styledContext.size)) * styledContext.scaleIcon,
          color: useTheme().color
        });
        children = React.Children.toArray(childrenProp).map(function (child) {
          if (disablePassStyles || ! /* @__PURE__ */React.isValidElement(child)) return child;
          return getThemedIcon(child);
        });
      }
      var context = React.useContext(CheckboxContext);
      if (forceMount || isIndeterminate(context.checked) || context.checked === true) return /* @__PURE__ */_jsx(Indicator, {
        pointerEvents: "none",
        ...indicatorProps,
        ...(active && activeStyle),
        ref: forwardedRef,
        children
      });
      return null;
    })
  });
}
export { CheckboxContext, createCheckbox };
