// @tamagui/checkbox/dist/esm/createCheckbox.mjs
import { isIndeterminate, useCheckbox } from "@tamagui/checkbox-headless";
import { getVariableValue, isWeb, shouldRenderNativePlatform, useProps, useTheme, withStaticProperties } from "@tamagui/core";
import { registerFocusable } from "@tamagui/focusable";
import { getFontSize } from "@tamagui/font-size";
import { getSize } from "@tamagui/get-token";
import { useGetThemedIcon } from "@tamagui/helpers-tamagui";
import { useControllableState } from "@tamagui/use-controllable-state";
import React, { useMemo } from "react";
import { CheckboxFrame, CheckboxIndicatorFrame } from "./Checkbox.mjs";
import { CheckboxStyledContext } from "./CheckboxStyledContext.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
const CheckboxContext = React.createContext({
  checked: false,
  disabled: false
});
const ensureContext = x => {
  if (!x.context) x.context = CheckboxContext;
};
function createCheckbox(createProps) {
  const {
    Frame = CheckboxFrame,
    Indicator = CheckboxIndicatorFrame
  } = createProps;
  ensureContext(Frame);
  ensureContext(Indicator);
  return withStaticProperties(Frame.styleable(function Checkbox(_props, forwardedRef) {
    const {
      scaleSize = .45,
      sizeAdjust = 0,
      scaleIcon,
      checked: checkedProp,
      defaultChecked,
      onCheckedChange,
      native,
      unstyled = false,
      activeStyle,
      activeTheme,
      ...props
    } = _props;
    const propsActive = useProps(props);
    const styledContext = React.useContext(CheckboxStyledContext.context);
    let adjustedSize = 0;
    let size = 0;
    if (!unstyled) {
      adjustedSize = getVariableValue(getSize(propsActive.size ?? styledContext?.size ?? "$true", {
        shift: sizeAdjust
      }));
      size = scaleSize ? Math.round(adjustedSize * scaleSize) : adjustedSize;
    }
    const [checked = false, setChecked] = useControllableState({
      prop: checkedProp,
      defaultProp: defaultChecked,
      onChange: onCheckedChange
    });
    const {
      checkboxProps,
      checkboxRef,
      bubbleInput
    } = useCheckbox(
    // @ts-ignore
    propsActive, [checked, setChecked], forwardedRef);
    if (shouldRenderNativePlatform(native) === "web") return /* @__PURE__ */jsx("input", {
      type: "checkbox",
      defaultChecked: isIndeterminate(checked) ? false : checked,
      tabIndex: -1,
      ref: checkboxRef,
      disabled: checkboxProps.disabled,
      style: {
        appearance: "auto",
        accentColor: "var(--color6)",
        ...checkboxProps.style
      }
    });
    const memoizedContext = useMemo(() => ({
      checked,
      disabled: checkboxProps.disabled
    }), [checked, checkboxProps.disabled]);
    const isActive = !!checked;
    const disabled = checkboxProps.disabled;
    return /* @__PURE__ */jsx(CheckboxContext.Provider, {
      value: memoizedContext,
      children: /* @__PURE__ */jsxs(CheckboxStyledContext.Provider, {
        size: propsActive.size ?? styledContext?.size ?? "$true",
        scaleIcon: scaleIcon ?? styledContext?.scaleIcon ?? 1,
        unstyled,
        active: isActive,
        disabled,
        children: [/* @__PURE__ */jsx(Frame, {
          render: "button",
          ref: checkboxRef,
          unstyled,
          theme: activeTheme ?? null,
          ...(isWeb && {
            type: "button"
          }),
          ...(!unstyled && {
            width: size,
            height: size,
            size
          }),
          checked,
          disabled,
          ...checkboxProps,
          ...props,
          ...(isActive && {
            ...(!unstyled && !activeStyle && {
              backgroundColor: "$backgroundActive"
            }),
            ...activeStyle
          }),
          children: propsActive.children
        }), bubbleInput]
      })
    });
  }), {
    Indicator: Indicator.styleable((props, forwardedRef) => {
      const {
        children: childrenProp,
        forceMount,
        disablePassStyles,
        unstyled = false,
        activeStyle,
        ...indicatorProps
      } = props;
      const styledContext = CheckboxStyledContext.useStyledContext();
      const {
        active
      } = styledContext;
      let children = childrenProp;
      if (!unstyled) {
        const getThemedIcon = useGetThemedIcon({
          size: (typeof styledContext.size === "number" ? styledContext.size * .65 : getFontSize(styledContext.size)) * styledContext.scaleIcon,
          color: useTheme().color
        });
        children = React.Children.toArray(childrenProp).map(child => {
          if (disablePassStyles || !React.isValidElement(child)) return child;
          return getThemedIcon(child);
        });
      }
      const context = React.useContext(CheckboxContext);
      if (forceMount || isIndeterminate(context.checked) || context.checked === true) return /* @__PURE__ */jsx(Indicator, {
        pointerEvents: "none",
        ...indicatorProps,
        ...(active && activeStyle),
        ref: forwardedRef,
        children
      });
      return null;
    })
  });
}
export { CheckboxContext, createCheckbox };
