// @tamagui/v2-toast/dist/esm/ToastImperative.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { isAndroid, isIos, isWeb } from "@tamagui/core";
import React from "react";
import { createNativeToast } from "./createNativeToast.native.js";
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var ToastContext = /* @__PURE__ */React.createContext({});
var ToastCurrentContext = /* @__PURE__ */React.createContext(null);
var useToastController = function () {
  return React.useContext(ToastContext);
};
var useToastState = function () {
  return React.useContext(ToastCurrentContext);
};
var useToast = function () {
  return {
    ...useToastController(),
    currentToast: useToastState()
  };
};
var ToastImperativeProvider = function (param) {
  var {
    children,
    options
  } = param;
  var counterRef = React.useRef(0);
  var [toast, setToast] = React.useState(null);
  var [lastNativeToastRef, setLastNativeToastRef] = React.useState(null);
  var show = React.useCallback(function (title, showOptions) {
    var _showOptions_native;
    var native = (_showOptions_native = showOptions === null || showOptions === void 0 ? void 0 : showOptions.native) !== null && _showOptions_native !== void 0 ? _showOptions_native : options.native;
    var isWebNative = Array.isArray(native) ? native.includes("web") : native === "web";
    var isMobileNative = Array.isArray(native) ? native.includes("mobile") : native === "mobile";
    var isAndroidNative = isMobileNative || (Array.isArray(native) ? native.includes("android") : native === "android");
    var isIosNative = isMobileNative || (Array.isArray(native) ? native.includes("ios") : native === "ios");
    var isHandledNatively = native === true || isWeb && isWebNative || !isWeb && isMobileNative || isAndroid && isAndroidNative || isIos && isIosNative;
    if (isHandledNatively) {
      var nativeToastResult = createNativeToast(title, showOptions !== null && showOptions !== void 0 ? showOptions : {});
      if ((typeof nativeToastResult === "undefined" ? "undefined" : _type_of(nativeToastResult)) === "object" && nativeToastResult.nativeToastRef) {
        setLastNativeToastRef(nativeToastResult.nativeToastRef);
      }
    }
    counterRef.current++;
    var _showOptions_viewportName;
    setToast({
      ...(showOptions === null || showOptions === void 0 ? void 0 : showOptions.customData),
      ...showOptions,
      viewportName: (_showOptions_viewportName = showOptions === null || showOptions === void 0 ? void 0 : showOptions.viewportName) !== null && _showOptions_viewportName !== void 0 ? _showOptions_viewportName : "default",
      title,
      id: counterRef.current.toString(),
      isHandledNatively
    });
    return true;
  }, [setToast, JSON.stringify(options.native || null)]);
  var hide = React.useCallback(function () {
    lastNativeToastRef === null || lastNativeToastRef === void 0 ? void 0 : lastNativeToastRef.close();
    setToast(function (prev) {
      if (!prev) return null;
      return {
        ...prev,
        hide: true
      };
    });
    setTimeout(function () {
      setToast(null);
    }, 100);
  }, [setToast, lastNativeToastRef]);
  var contextValue = React.useMemo(function () {
    return {
      show,
      hide,
      nativeToast: lastNativeToastRef,
      options
    };
  }, [show, hide, lastNativeToastRef, JSON.stringify(options || null)]);
  return /* @__PURE__ */_jsx(ToastContext.Provider, {
    value: contextValue,
    children: /* @__PURE__ */_jsx(ToastCurrentContext.Provider, {
      value: toast,
      children
    })
  });
};
export { ToastImperativeProvider, useToast, useToastController, useToastState };
