// @tamagui/v2-toast/dist/esm/ToastImpl.mjs
import { useIsPresent } from "@tamagui/animate-presence";
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { View, Theme, createStyledContext, styled, useConfiguration, useEvent, useThemeName } from "@tamagui/core";
import { Dismissable } from "@tamagui/dismissable";
import { composeEventHandlers } from "@tamagui/helpers";
import { PortalItem } from "@tamagui/portal";
import { YStack } from "@tamagui/stacks";
import * as React from "react";
import { PanResponder } from "react-native-web";
import { TOAST_NAME } from "./constants.mjs";
import { ToastAnnounce } from "./ToastAnnounce.mjs";
import { Collection, useToastProviderContext } from "./ToastProvider.mjs";
import { VIEWPORT_PAUSE, VIEWPORT_RESUME } from "./ToastViewport.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
const ToastImplFrame = styled(YStack, {
  name: "ToastImpl",
  focusable: true,
  variants: {
    unstyled: {
      false: {
        focusStyle: {
          outlineStyle: "solid",
          outlineWidth: 2,
          outlineColor: "$outlineColor"
        },
        backgroundColor: "$color6",
        borderRadius: "$4",
        paddingHorizontal: "$4",
        paddingVertical: "$3",
        marginHorizontal: "auto",
        marginVertical: "$1",
        elevation: "$3"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const {
  Provider: ToastInteractiveProvider,
  useStyledContext: useToastInteractiveContext
} = createStyledContext({
  onClose() {}
});
const ToastImpl = React.forwardRef((props, forwardedRef) => {
  const {
    scope,
    type = "foreground",
    duration: durationProp,
    open,
    onClose,
    onEscapeKeyDown,
    onPause,
    onResume,
    onSwipeStart,
    onSwipeMove,
    onSwipeCancel,
    onSwipeEnd,
    viewportName = "default",
    ...toastProps
  } = props;
  const isPresent = useIsPresent();
  const context = useToastProviderContext(scope);
  const [node, setNode] = React.useState(null);
  const composedRefs = useComposedRefs(forwardedRef, setNode);
  const duration = durationProp || context.duration;
  const closeTimerStartTimeRef = React.useRef(0);
  const closeTimerRemainingTimeRef = React.useRef(duration);
  const closeTimerRef = React.useRef(0);
  const {
    onToastAdd,
    onToastRemove
  } = context;
  const viewport = React.useMemo(() => {
    return context.viewports[viewportName];
  }, [context.viewports, viewportName]);
  const handleClose = useEvent(() => {
    if (!isPresent) {
      return;
    }
    if (isWeb) {
      const isFocusInToast = node?.contains(document.activeElement);
      if (isFocusInToast) viewport?.focus();
    }
    onClose();
  });
  const startTimer = React.useCallback(duration2 => {
    if (!duration2 || duration2 === Number.POSITIVE_INFINITY) return;
    clearTimeout(closeTimerRef.current);
    closeTimerStartTimeRef.current = (/* @__PURE__ */new Date()).getTime();
    closeTimerRef.current = setTimeout(handleClose, duration2);
  }, [handleClose]);
  const handleResume = React.useCallback(() => {
    startTimer(closeTimerRemainingTimeRef.current);
    onResume?.();
  }, [onResume, startTimer]);
  const handlePause = React.useCallback(() => {
    const elapsedTime = (/* @__PURE__ */new Date()).getTime() - closeTimerStartTimeRef.current;
    closeTimerRemainingTimeRef.current = closeTimerRemainingTimeRef.current - elapsedTime;
    window.clearTimeout(closeTimerRef.current);
    onPause?.();
  }, [onPause]);
  React.useEffect(() => {
    if (!isWeb) return;
    if (viewport) {
      viewport.addEventListener(VIEWPORT_PAUSE, handlePause);
      viewport.addEventListener(VIEWPORT_RESUME, handleResume);
      return () => {
        viewport.removeEventListener(VIEWPORT_PAUSE, handlePause);
        viewport.removeEventListener(VIEWPORT_RESUME, handleResume);
      };
    }
  }, [viewport, duration, onPause, onResume, startTimer]);
  React.useEffect(() => {
    if (open && !context.isClosePausedRef.current) {
      startTimer(duration);
    }
  }, [open, duration, context.isClosePausedRef, startTimer]);
  React.useEffect(() => {
    onToastAdd();
    return () => onToastRemove();
  }, [onToastAdd, onToastRemove]);
  const announceTextContent = React.useMemo(() => {
    if (!isWeb) return null;
    return node ? getAnnounceTextContent(node) : null;
  }, [node]);
  const isHorizontalSwipe = ["left", "right", "horizontal"].includes(context.swipeDirection);
  const {
    animationDriver
  } = useConfiguration();
  if (!animationDriver) {
    throw new Error("Must set animations in tamagui.config.ts");
  }
  const {
    useAnimatedNumber,
    useAnimatedNumberStyle
  } = animationDriver;
  const animatedNumber = useAnimatedNumber(0);
  const AnimatedView = animationDriver["NumberView"] ?? animationDriver.View ?? View;
  const animatedStyles = useAnimatedNumberStyle(animatedNumber, val => {
    "worklet";

    return {
      transform: [isHorizontalSwipe ? {
        translateX: val
      } : {
        translateY: val
      }]
    };
  });
  const panResponder = React.useMemo(() => {
    return PanResponder.create({
      onMoveShouldSetPanResponder: (e, gesture) => {
        const shouldMove = shouldGrantGestureMove(context.swipeDirection, gesture);
        if (shouldMove) {
          onSwipeStart?.(e);
          return true;
        }
        return false;
      },
      onPanResponderGrant: e => {
        if (!isWeb) {
          handlePause?.();
        }
      },
      onPanResponderMove: (e, gesture) => {
        const {
          x,
          y
        } = getGestureDistance(context.swipeDirection, gesture);
        const delta = {
          x,
          y
        };
        animatedNumber.setValue(isHorizontalSwipe ? x : y, {
          type: "direct"
        });
        if (isDeltaInDirection(delta, context.swipeDirection, context.swipeThreshold)) {
          onSwipeEnd?.(e);
        }
        onSwipeMove?.(e);
      },
      onPanResponderEnd: (e, {
        dx,
        dy
      }) => {
        if (!isDeltaInDirection({
          x: dx,
          y: dy
        }, context.swipeDirection, context.swipeThreshold)) {
          if (!isWeb) {
            handleResume?.();
          }
          onSwipeCancel?.(e);
          animatedNumber.setValue(0, {
            type: "spring"
          });
        }
      }
    });
  }, [handlePause, handleResume]);
  const themeName = useThemeName();
  return /* @__PURE__ */jsxs(Fragment, {
    children: [announceTextContent && /* @__PURE__ */jsx(ToastAnnounce, {
      scope,
      role: "status",
      "aria-live": type === "foreground" ? "assertive" : "polite",
      "aria-atomic": true,
      children: announceTextContent
    }), /* @__PURE__ */jsx(PortalItem, {
      hostName: viewportName ?? "default",
      children: /* @__PURE__ */jsx(ToastInteractiveProvider, {
        scope,
        onClose: () => {
          handleClose();
        },
        children: /* @__PURE__ */jsx(Dismissable, {
          onEscapeKeyDown: composeEventHandlers(onEscapeKeyDown, () => {
            if (!context.isFocusedToastEscapeKeyDownRef.current) {
              handleClose();
            }
            context.isFocusedToastEscapeKeyDownRef.current = false;
          }),
          children: /* @__PURE__ */jsx(Theme, {
            contain: true,
            forceClassName: true,
            name: themeName,
            children: /* @__PURE__ */jsx(AnimatedView, {
              ...panResponder?.panHandlers,
              style: [{
                margin: "auto"
              }, animatedStyles],
              children: /* @__PURE__ */jsx(Collection.ItemSlot, {
                scope: context.toastScope,
                children: /* @__PURE__ */jsx(ToastImplFrame, {
                  role: "status",
                  "aria-live": "off",
                  "aria-atomic": true,
                  "data-state": open ? "open" : "closed",
                  "data-swipe-direction": context.swipeDirection,
                  pointerEvents: "auto",
                  "$platform-web": {
                    touchAction: "none",
                    userSelect: "none"
                  },
                  ...toastProps,
                  ref: composedRefs,
                  ...(isWeb && {
                    onKeyDown: composeEventHandlers(props.onKeyDown, event => {
                      if (event.key !== "Escape") return;
                      onEscapeKeyDown?.(event);
                      if (!event.defaultPrevented) {
                        context.isFocusedToastEscapeKeyDownRef.current = true;
                        handleClose();
                      }
                    })
                  })
                })
              })
            })
          })
        })
      }, props.id)
    })]
  });
});
ToastImpl.propTypes = {
  type(props) {
    if (props.type && !["foreground", "background"].includes(props.type)) {
      const error = `Invalid prop \`type\` supplied to \`${TOAST_NAME}\`. Expected \`foreground | background\`.`;
      return new Error(error);
    }
    return null;
  }
};
const isDeltaInDirection = (delta, direction, threshold = 0) => {
  const deltaX = Math.abs(delta.x);
  const deltaY = Math.abs(delta.y);
  const isDeltaX = deltaX > deltaY;
  if (direction === "left" || direction === "right" || direction === "horizontal") {
    return isDeltaX && deltaX > threshold;
  }
  return !isDeltaX && deltaY > threshold;
};
function getAnnounceTextContent(container) {
  if (!isWeb) return "";
  const textContent = [];
  const childNodes = Array.from(container.childNodes);
  childNodes.forEach(node => {
    if (node.nodeType === node.TEXT_NODE && node.textContent) textContent.push(node.textContent);
    if (isHTMLElement(node)) {
      const isHidden = node.ariaHidden || node.hidden || node.style.display === "none";
      const isExcluded = node.dataset.toastAnnounceExclude === "";
      if (!isHidden) {
        if (isExcluded) {
          const altText = node.dataset.toastAnnounceAlt;
          if (altText) textContent.push(altText);
        } else {
          textContent.push(...getAnnounceTextContent(node));
        }
      }
    }
  });
  return textContent;
}
function isHTMLElement(node) {
  return node.nodeType === node.ELEMENT_NODE;
}
const GESTURE_GRANT_THRESHOLD = 10;
const shouldGrantGestureMove = (dir, {
  dx,
  dy
}) => {
  if ((dir === "horizontal" || dir === "left") && dx < -GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "horizontal" || dir === "right") && dx > GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "vertical" || dir === "up") && dy > -GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "vertical" || dir === "down") && dy < GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  return false;
};
const getGestureDistance = (dir, {
  dx,
  dy
}) => {
  let y = 0;
  let x = 0;
  if (dir === "horizontal") x = dx;else if (dir === "left") x = Math.min(0, dx);else if (dir === "right") x = Math.max(0, dx);else if (dir === "vertical") y = dy;else if (dir === "up") y = Math.min(0, dy);else if (dir === "down") y = Math.max(0, dy);
  return {
    x,
    y
  };
};
export { ToastImpl, ToastImplFrame, useToastInteractiveContext };
