// @tamagui/v2-toast/dist/esm/ToastImperative.mjs
import { isAndroid, isIos, isWeb } from "@tamagui/core";
import React from "react";
import { createNativeToast } from "./createNativeToast.mjs";
import { jsx } from "react/jsx-runtime";
const ToastContext = React.createContext({});
const ToastCurrentContext = React.createContext(null);
const useToastController = () => {
  return React.useContext(ToastContext);
};
const useToastState = () => {
  return React.useContext(ToastCurrentContext);
};
const useToast = () => {
  return {
    ...useToastController(),
    currentToast: useToastState()
  };
};
const ToastImperativeProvider = ({
  children,
  options
}) => {
  const counterRef = React.useRef(0);
  const [toast, setToast] = React.useState(null);
  const [lastNativeToastRef, setLastNativeToastRef] = React.useState(null);
  const show = React.useCallback((title, showOptions) => {
    const native = showOptions?.native ?? options.native;
    const isWebNative = Array.isArray(native) ? native.includes("web") : native === "web";
    const isMobileNative = Array.isArray(native) ? native.includes("mobile") : native === "mobile";
    const isAndroidNative = isMobileNative || (Array.isArray(native) ? native.includes("android") : native === "android");
    const isIosNative = isMobileNative || (Array.isArray(native) ? native.includes("ios") : native === "ios");
    const isHandledNatively = native === true || isWeb && isWebNative || !isWeb && isMobileNative || isAndroid && isAndroidNative || isIos && isIosNative;
    if (isHandledNatively) {
      const nativeToastResult = createNativeToast(title, showOptions ?? {});
      if (typeof nativeToastResult === "object" && nativeToastResult.nativeToastRef) {
        setLastNativeToastRef(nativeToastResult.nativeToastRef);
      }
    }
    counterRef.current++;
    setToast({
      ...showOptions?.customData,
      ...showOptions,
      viewportName: showOptions?.viewportName ?? "default",
      title,
      id: counterRef.current.toString(),
      isHandledNatively
    });
    return true;
  }, [setToast, JSON.stringify(options.native || null)]);
  const hide = React.useCallback(() => {
    lastNativeToastRef?.close();
    setToast(prev => {
      if (!prev) return null;
      return {
        ...prev,
        hide: true
      };
    });
    setTimeout(() => {
      setToast(null);
    }, 100);
  }, [setToast, lastNativeToastRef]);
  const contextValue = React.useMemo(() => {
    return {
      show,
      hide,
      nativeToast: lastNativeToastRef,
      options
    };
  }, [show, hide, lastNativeToastRef, JSON.stringify(options || null)]);
  return /* @__PURE__ */jsx(ToastContext.Provider, {
    value: contextValue,
    children: /* @__PURE__ */jsx(ToastCurrentContext.Provider, {
      value: toast,
      children
    })
  });
};
export { ToastImperativeProvider, useToast, useToastController, useToastState };
