// @tamagui/v2-toast/dist/esm/ToastState.native.js
function _class_call_check(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}
function _define_property(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }
  return obj;
}
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
    return !!right[Symbol.hasInstance](left);
  } else {
    return left instanceof right;
  }
}
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var toastsCounter = 1;
var Observer = function Observer2() {
  "use strict";

  var _this = this;
  _class_call_check(this, Observer2);
  _define_property(this, "subscribers", []);
  _define_property(this, "toasts", []);
  _define_property(this, "dismissedToasts", /* @__PURE__ */new Set());
  _define_property(this, "subscribe", function (subscriber) {
    _this.subscribers.push(subscriber);
    return function () {
      var index = _this.subscribers.indexOf(subscriber);
      if (index > -1) {
        _this.subscribers.splice(index, 1);
      }
    };
  });
  _define_property(this, "publish", function (data) {
    _this.subscribers.forEach(function (subscriber) {
      return subscriber(data);
    });
  });
  _define_property(this, "addToast", function (data) {
    _this.publish(data);
    _this.toasts = [..._this.toasts, data];
  });
  _define_property(this, "create", function (data) {
    var {
      title,
      ...rest
    } = data;
    var id = typeof (data === null || data === void 0 ? void 0 : data.id) === "number" || typeof (data === null || data === void 0 ? void 0 : data.id) === "string" && data.id.length > 0 ? data.id : toastsCounter++;
    var alreadyExists = _this.toasts.find(function (toast2) {
      return toast2.id === id;
    });
    var _data_dismissible;
    var dismissible = (_data_dismissible = data.dismissible) !== null && _data_dismissible !== void 0 ? _data_dismissible : true;
    if (_this.dismissedToasts.has(id)) {
      _this.dismissedToasts.delete(id);
    }
    if (alreadyExists) {
      _this.toasts = _this.toasts.map(function (toast2) {
        if (toast2.id === id) {
          _this.publish({
            ...toast2,
            ...data,
            id,
            title,
            dismissible
          });
          return {
            ...toast2,
            ...data,
            id,
            title,
            dismissible
          };
        }
        return toast2;
      });
    } else {
      _this.addToast({
        title,
        ...rest,
        dismissible,
        id
      });
    }
    return id;
  });
  _define_property(this, "dismiss", function (id) {
    if (id !== void 0) {
      _this.dismissedToasts.add(id);
      requestAnimationFrame(function () {
        _this.subscribers.forEach(function (subscriber) {
          return subscriber({
            id,
            dismiss: true
          });
        });
      });
    } else {
      _this.toasts.forEach(function (toast2) {
        _this.subscribers.forEach(function (subscriber) {
          return subscriber({
            id: toast2.id,
            dismiss: true
          });
        });
      });
    }
    return id;
  });
  _define_property(this, "message", function (title, data) {
    return _this.create({
      ...data,
      title,
      type: "default"
    });
  });
  _define_property(this, "success", function (title, data) {
    return _this.create({
      ...data,
      title,
      type: "success"
    });
  });
  _define_property(this, "error", function (title, data) {
    return _this.create({
      ...data,
      title,
      type: "error"
    });
  });
  _define_property(this, "warning", function (title, data) {
    return _this.create({
      ...data,
      title,
      type: "warning"
    });
  });
  _define_property(this, "info", function (title, data) {
    return _this.create({
      ...data,
      title,
      type: "info"
    });
  });
  _define_property(this, "loading", function (title, data) {
    return _this.create({
      ...data,
      title,
      type: "loading"
    });
  });
  _define_property(this, "promise", function (promise, data) {
    if (!data) {
      return;
    }
    var id = void 0;
    if (data.loading !== void 0) {
      id = _this.create({
        promise,
        type: "loading",
        title: data.loading,
        description: typeof data.description !== "function" ? data.description : void 0,
        // loading toasts shouldn't auto-dismiss
        duration: Number.POSITIVE_INFINITY
      });
    }
    var p = Promise.resolve(_instanceof(promise, Function) ? promise() : promise);
    var shouldDismiss = id !== void 0;
    var result;
    var originalPromise = p.then(async function (response) {
      result = ["resolve", response];
      if (isHttpResponse(response) && !response.ok) {
        shouldDismiss = false;
        var errorMsg = typeof data.error === "function" ? await data.error(`HTTP error! status: ${response.status}`) : data.error;
        var description = typeof data.description === "function" ? await data.description(`HTTP error! status: ${response.status}`) : data.description;
        _this.create({
          id,
          type: "error",
          title: errorMsg,
          description
        });
      } else if (data.success !== void 0) {
        shouldDismiss = false;
        var successMsg = typeof data.success === "function" ? await data.success(response) : data.success;
        var description1 = typeof data.description === "function" ? await data.description(response) : data.description;
        _this.create({
          id,
          type: "success",
          title: successMsg,
          description: description1
        });
      }
    }).catch(async function (error) {
      result = ["reject", error];
      if (data.error !== void 0) {
        shouldDismiss = false;
        var errorMsg = typeof data.error === "function" ? await data.error(error) : data.error;
        var description = typeof data.description === "function" ? await data.description(error) : data.description;
        _this.create({
          id,
          type: "error",
          title: errorMsg,
          description
        });
      }
    }).finally(function () {
      var _data_finally;
      if (shouldDismiss) {
        _this.dismiss(id);
        id = void 0;
      }
      (_data_finally = data.finally) === null || _data_finally === void 0 ? void 0 : _data_finally.call(data);
    });
    var unwrap = function () {
      return new Promise(function (resolve, reject) {
        return originalPromise.then(function () {
          return result[0] === "reject" ? reject(result[1]) : resolve(result[1]);
        }).catch(reject);
      });
    };
    if (typeof id !== "string" && typeof id !== "number") {
      return {
        unwrap
      };
    } else {
      return Object.assign(id, {
        unwrap
      });
    }
  });
  _define_property(this, "custom", function (jsx, data) {
    var _data_id;
    var id = (_data_id = data === null || data === void 0 ? void 0 : data.id) !== null && _data_id !== void 0 ? _data_id : toastsCounter++;
    _this.create({
      jsx: jsx(id),
      ...data,
      id
    });
    return id;
  });
  _define_property(this, "getActiveToasts", function () {
    return _this.toasts.filter(function (toast2) {
      return !_this.dismissedToasts.has(toast2.id);
    });
  });
  _define_property(this, "getHistory", function () {
    return _this.toasts;
  });
};
function isHttpResponse(data) {
  return data && (typeof data === "undefined" ? "undefined" : _type_of(data)) === "object" && "ok" in data && typeof data.ok === "boolean" && "status" in data && typeof data.status === "number";
}
var ToastState = new Observer();
var toastFunction = function (title, data) {
  return ToastState.create({
    ...data,
    title
  });
};
var getHistory = function () {
  return ToastState.getHistory();
};
var getToasts = function () {
  return ToastState.getActiveToasts();
};
var toast = Object.assign(toastFunction, {
  success: ToastState.success,
  error: ToastState.error,
  warning: ToastState.warning,
  info: ToastState.info,
  loading: ToastState.loading,
  promise: ToastState.promise,
  custom: ToastState.custom,
  dismiss: ToastState.dismiss,
  message: ToastState.message,
  getHistory,
  getToasts
});
export { ToastState, toast };
