// @tamagui/v2-toast/dist/esm/Toast.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { AnimatePresence } from "@tamagui/animate-presence";
import { styled, useEvent } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { YStack } from "@tamagui/stacks";
import { SizableText } from "@tamagui/text";
import { useControllableState } from "@tamagui/use-controllable-state";
import * as React from "react";
import { ToastAnnounceExclude } from "./ToastAnnounce.native.js";
import { useToast, useToastController, useToastState } from "./ToastImperative.native.js";
import { ToastImpl, ToastImplFrame, useToastInteractiveContext } from "./ToastImpl.native.js";
import { ToastProvider } from "./ToastProvider.native.js";
import { ToastViewport } from "./ToastViewport.native.js";
var TITLE_NAME = "ToastTitle";
var ToastTitle = styled(SizableText, {
  name: TITLE_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color",
        size: "$4"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var DESCRIPTION_NAME = "ToastDescription";
var ToastDescription = styled(SizableText, {
  name: DESCRIPTION_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color11",
        size: "$1"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var ACTION_NAME = "ToastAction";
var ToastAction = /* @__PURE__ */React.forwardRef(function ToastAction2(props, forwardedRef) {
  var {
    altText,
    ...actionProps
  } = props;
  if (!altText) return null;
  return /* @__PURE__ */_jsx(ToastAnnounceExclude, {
    altText,
    asChild: true,
    children: /* @__PURE__ */_jsx(ToastClose, {
      ...actionProps,
      ref: forwardedRef
    })
  });
});
ToastAction.propTypes = {
  altText(props) {
    if (!props.altText) {
      return new Error(`Missing prop \`altText\` expected on \`${ACTION_NAME}\``);
    }
    return null;
  }
};
var CLOSE_NAME = "ToastClose";
var ToastCloseFrame = styled(YStack, {
  name: CLOSE_NAME,
  render: "button"
});
var ToastClose = /* @__PURE__ */React.forwardRef(function ToastClose2(props, forwardedRef) {
  var {
    scope,
    ...closeProps
  } = props;
  var interactiveContext = useToastInteractiveContext(scope);
  return /* @__PURE__ */_jsx(ToastAnnounceExclude, {
    asChild: true,
    children: /* @__PURE__ */_jsx(ToastCloseFrame, {
      "aria-label": "Close",
      ...closeProps,
      ref: forwardedRef,
      onPress: composeEventHandlers(props.onPress, interactiveContext.onClose)
    })
  });
});
var ToastComponent = ToastImplFrame.styleable(function Toast(props, forwardedRef) {
  var {
    forceMount,
    open: openProp,
    defaultOpen,
    onOpenChange,
    ...toastProps
  } = props;
  var [open, setOpen] = useControllableState({
    prop: openProp,
    defaultProp: defaultOpen !== null && defaultOpen !== void 0 ? defaultOpen : true,
    onChange: onOpenChange,
    strategy: "most-recent-wins"
  });
  var currentToast = useToastState();
  var {
    hide
  } = useToastController();
  var id = React.useId();
  var onPause = useEvent(props.onPause);
  var onResume = useEvent(props.onResume);
  var isHide = (currentToast === null || currentToast === void 0 ? void 0 : currentToast.hide) === true;
  var shouldShow = (forceMount || open) && !isHide;
  return /* @__PURE__ */_jsx(AnimatePresence, {
    children: shouldShow ? /* @__PURE__ */_jsx(ToastImpl, {
      id,
      open,
      ...toastProps,
      ref: forwardedRef,
      onClose: function () {
        setOpen(false);
        hide();
      },
      onPause,
      onResume,
      onSwipeEnd: composeEventHandlers(props.onSwipeEnd, function (event) {
        setOpen(false);
      })
    }) : null
  }, id);
});
var Toast2 = withStaticProperties(ToastComponent, {
  Title: ToastTitle,
  Description: ToastDescription,
  Action: ToastAction,
  Close: ToastClose
});
export { Toast2 as Toast, ToastProvider, ToastViewport, useToast, useToastController, useToastState };
