// @tamagui/use-debounce/dist/esm/index.native.js
import * as React from "react";
function debounce(func, wait, leading) {
  var timeout;
  var isCancelled = false;
  function debounced() {
    var _this = this;
    isCancelled = false;
    var args = arguments;
    if (leading && !timeout) {
      func.apply(this, args);
    }
    clearTimeout(timeout);
    timeout = setTimeout(function () {
      timeout = null;
      if (!(leading || isCancelled)) {
        func.apply(_this, args);
      }
      isCancelled = false;
    }, wait);
  }
  debounced.cancel = function () {
    isCancelled = true;
  };
  return debounced;
}
var defaultOpts = {
  leading: false
};
function useDebounce(fn, wait) {
  var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : defaultOpts,
    mountArgs = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : [fn];
  var dbEffect = React.useRef(null);
  React.useEffect(function () {
    return function () {
      var _dbEffect_current;
      (_dbEffect_current = dbEffect.current) === null || _dbEffect_current === void 0 ? void 0 : _dbEffect_current.cancel();
    };
  }, []);
  return React.useMemo(function () {
    dbEffect.current = debounce(fn, wait, options.leading);
    return dbEffect.current;
  }, [options.leading, ...mountArgs]);
}
function useDebounceValue(val) {
  var amt = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 0;
  var [state, setState] = React.useState(val);
  React.useEffect(function () {
    var tm = setTimeout(function () {
      setState(function (prev) {
        if (prev === val) return prev;
        return val;
      });
    }, amt);
    return function () {
      clearTimeout(tm);
    };
  }, [val]);
  return state;
}
export { debounce, useDebounce, useDebounceValue };
