// @tamagui/toast/dist/cjs/ToastImperative.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var ToastImperative_exports = {};
__export(ToastImperative_exports, {
  ToastImperativeProvider: () => ToastImperativeProvider,
  useToast: () => useToast,
  useToastController: () => useToastController,
  useToastState: () => useToastState
});
module.exports = __toCommonJS(ToastImperative_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@tamagui/core");
var import_react = __toESM(require("react"), 1);
var import_createNativeToast = require("./createNativeToast.native.js");
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var ToastContext = /* @__PURE__ */import_react.default.createContext({});
var ToastCurrentContext = /* @__PURE__ */import_react.default.createContext(null);
var useToastController = function () {
  return import_react.default.useContext(ToastContext);
};
var useToastState = function () {
  return import_react.default.useContext(ToastCurrentContext);
};
var useToast = function () {
  return {
    ...useToastController(),
    currentToast: useToastState()
  };
};
var ToastImperativeProvider = function (param) {
  var {
    children,
    options
  } = param;
  var counterRef = import_react.default.useRef(0);
  var [toast, setToast] = import_react.default.useState(null);
  var [lastNativeToastRef, setLastNativeToastRef] = import_react.default.useState(null);
  var show = import_react.default.useCallback(function (title, showOptions) {
    var _showOptions_native;
    var native = (_showOptions_native = showOptions === null || showOptions === void 0 ? void 0 : showOptions.native) !== null && _showOptions_native !== void 0 ? _showOptions_native : options.native;
    var isWebNative = Array.isArray(native) ? native.includes("web") : native === "web";
    var isMobileNative = Array.isArray(native) ? native.includes("mobile") : native === "mobile";
    var isAndroidNative = isMobileNative || (Array.isArray(native) ? native.includes("android") : native === "android");
    var isIosNative = isMobileNative || (Array.isArray(native) ? native.includes("ios") : native === "ios");
    var isHandledNatively = native === true || import_core.isWeb && isWebNative || !import_core.isWeb && isMobileNative || import_core.isAndroid && isAndroidNative || import_core.isIos && isIosNative;
    if (isHandledNatively) {
      var nativeToastResult = (0, import_createNativeToast.createNativeToast)(title, showOptions !== null && showOptions !== void 0 ? showOptions : {});
      if ((typeof nativeToastResult === "undefined" ? "undefined" : _type_of(nativeToastResult)) === "object" && nativeToastResult.nativeToastRef) {
        setLastNativeToastRef(nativeToastResult.nativeToastRef);
      }
    }
    counterRef.current++;
    var _showOptions_viewportName;
    setToast({
      ...(showOptions === null || showOptions === void 0 ? void 0 : showOptions.customData),
      ...showOptions,
      viewportName: (_showOptions_viewportName = showOptions === null || showOptions === void 0 ? void 0 : showOptions.viewportName) !== null && _showOptions_viewportName !== void 0 ? _showOptions_viewportName : "default",
      title,
      id: counterRef.current.toString(),
      isHandledNatively
    });
    return true;
  }, [setToast, JSON.stringify(options.native || null)]);
  var hide = import_react.default.useCallback(function () {
    lastNativeToastRef === null || lastNativeToastRef === void 0 ? void 0 : lastNativeToastRef.close();
    setToast(function (prev) {
      if (!prev) return null;
      return {
        ...prev,
        hide: true
      };
    });
    setTimeout(function () {
      setToast(null);
    }, 100);
  }, [setToast, lastNativeToastRef]);
  var contextValue = import_react.default.useMemo(function () {
    return {
      show,
      hide,
      nativeToast: lastNativeToastRef,
      options
    };
  }, [show, hide, lastNativeToastRef, JSON.stringify(options || null)]);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(ToastContext.Provider, {
    value: contextValue,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(ToastCurrentContext.Provider, {
      value: toast,
      children
    })
  });
};
