// @tamagui/toast/dist/cjs/ToastImpl.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var ToastImpl_exports = {};
__export(ToastImpl_exports, {
  ToastImpl: () => ToastImpl,
  ToastImplFrame: () => ToastImplFrame,
  useToastInteractiveContext: () => useToastInteractiveContext
});
module.exports = __toCommonJS(ToastImpl_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_animate_presence = require("@tamagui/animate-presence");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_dismissable = require("@tamagui/dismissable");
var import_helpers = require("@tamagui/helpers");
var import_portal = require("@tamagui/portal");
var import_stacks = require("@tamagui/stacks");
var React = __toESM(require("react"), 1);
var import_constants2 = require("./constants.native.js");
var import_ToastAnnounce = require("./ToastAnnounce.native.js");
var import_ToastProvider = require("./ToastProvider.native.js");
var import_ToastViewport = require("./ToastViewport.native.js");
var getPanResponder = function () {
  if (import_constants.isWeb) return null;
  return require("react-native").PanResponder;
};
var ToastImplFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: "ToastImpl",
  tabIndex: 0,
  variants: {
    unstyled: {
      false: {
        focusStyle: {
          outlineStyle: "solid",
          outlineWidth: 2,
          outlineColor: "$outlineColor"
        },
        backgroundColor: "$color6",
        borderRadius: "$4",
        paddingHorizontal: "$4",
        paddingVertical: "$3",
        marginHorizontal: "auto",
        marginVertical: "$1",
        elevation: "$3"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var {
  Provider: ToastInteractiveProvider,
  useStyledContext: useToastInteractiveContext
} = (0, import_core.createStyledContext)({
  onClose() {}
});
var ToastImpl = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var {
    scope,
    type = "foreground",
    duration: durationProp,
    open,
    onClose,
    onEscapeKeyDown,
    onPause,
    onResume,
    onSwipeStart,
    onSwipeMove,
    onSwipeCancel,
    onSwipeEnd,
    viewportName = "default",
    ...toastProps
  } = props;
  var isPresent = (0, import_animate_presence.useIsPresent)();
  var context = (0, import_ToastProvider.useToastProviderContext)(scope);
  var [node, setNode] = React.useState(null);
  var composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, setNode);
  var duration = durationProp || context.duration;
  var closeTimerStartTimeRef = React.useRef(0);
  var closeTimerRemainingTimeRef = React.useRef(duration);
  var closeTimerRef = React.useRef(0);
  var {
    onToastAdd,
    onToastRemove
  } = context;
  var viewport = React.useMemo(function () {
    return context.viewports[viewportName];
  }, [context.viewports, viewportName]);
  var handleClose = (0, import_core.useEvent)(function () {
    if (!isPresent) {
      return;
    }
    if (import_constants.isWeb) {
      var isFocusInToast = node === null || node === void 0 ? void 0 : node.contains(document.activeElement);
      if (isFocusInToast) viewport === null || viewport === void 0 ? void 0 : viewport.focus();
    }
    onClose();
  });
  var startTimer = React.useCallback(function (duration2) {
    if (!duration2 || duration2 === Number.POSITIVE_INFINITY) return;
    clearTimeout(closeTimerRef.current);
    closeTimerStartTimeRef.current = (/* @__PURE__ */new Date()).getTime();
    closeTimerRef.current = setTimeout(handleClose, duration2);
  }, [handleClose]);
  var handleResume = React.useCallback(function () {
    startTimer(closeTimerRemainingTimeRef.current);
    onResume === null || onResume === void 0 ? void 0 : onResume();
  }, [onResume, startTimer]);
  var handlePause = React.useCallback(function () {
    var elapsedTime = (/* @__PURE__ */new Date()).getTime() - closeTimerStartTimeRef.current;
    closeTimerRemainingTimeRef.current = closeTimerRemainingTimeRef.current - elapsedTime;
    window.clearTimeout(closeTimerRef.current);
    onPause === null || onPause === void 0 ? void 0 : onPause();
  }, [onPause]);
  React.useEffect(function () {
    if (!import_constants.isWeb) return;
    if (viewport) {
      viewport.addEventListener(import_ToastViewport.VIEWPORT_PAUSE, handlePause);
      viewport.addEventListener(import_ToastViewport.VIEWPORT_RESUME, handleResume);
      return function () {
        viewport.removeEventListener(import_ToastViewport.VIEWPORT_PAUSE, handlePause);
        viewport.removeEventListener(import_ToastViewport.VIEWPORT_RESUME, handleResume);
      };
    }
  }, [viewport, duration, onPause, onResume, startTimer]);
  React.useEffect(function () {
    if (open && !context.isClosePausedRef.current) {
      startTimer(duration);
    }
  }, [open, duration, context.isClosePausedRef, startTimer]);
  React.useEffect(function () {
    onToastAdd();
    return function () {
      return onToastRemove();
    };
  }, [onToastAdd, onToastRemove]);
  var announceTextContent = React.useMemo(function () {
    if (!import_constants.isWeb) return null;
    return node ? getAnnounceTextContent(node) : null;
  }, [node]);
  var isHorizontalSwipe = ["left", "right", "horizontal"].includes(context.swipeDirection);
  var {
    animationDriver
  } = (0, import_core.useConfiguration)();
  if (!animationDriver) {
    throw new Error("Must set animations in tamagui.config.ts");
  }
  var {
    useAnimatedNumber,
    useAnimatedNumberStyle
  } = animationDriver;
  var animatedNumber = useAnimatedNumber(0);
  var _animationDriver_NumberView, _ref;
  var AnimatedView = (_ref = (_animationDriver_NumberView = animationDriver["NumberView"]) !== null && _animationDriver_NumberView !== void 0 ? _animationDriver_NumberView : animationDriver.View) !== null && _ref !== void 0 ? _ref : import_core.View;
  var animatedStyles = useAnimatedNumberStyle(animatedNumber, function (val) {
    "worklet";

    return {
      transform: [isHorizontalSwipe ? {
        translateX: val
      } : {
        translateY: val
      }]
    };
  });
  var panResponder = React.useMemo(function () {
    var PanResponder = getPanResponder();
    if (!PanResponder) return null;
    return PanResponder.create({
      onMoveShouldSetPanResponder: function (e, gesture) {
        var shouldMove = shouldGrantGestureMove(context.swipeDirection, gesture);
        if (shouldMove) {
          onSwipeStart === null || onSwipeStart === void 0 ? void 0 : onSwipeStart(e);
          return true;
        }
        return false;
      },
      onPanResponderGrant: function (e) {
        if (!import_constants.isWeb) {
          handlePause === null || handlePause === void 0 ? void 0 : handlePause();
        }
      },
      onPanResponderMove: function (e, gesture) {
        var {
          x,
          y
        } = getGestureDistance(context.swipeDirection, gesture);
        var delta = {
          x,
          y
        };
        animatedNumber.setValue(isHorizontalSwipe ? x : y, {
          type: "direct"
        });
        if (isDeltaInDirection(delta, context.swipeDirection, context.swipeThreshold)) {
          onSwipeEnd === null || onSwipeEnd === void 0 ? void 0 : onSwipeEnd(e);
        }
        onSwipeMove === null || onSwipeMove === void 0 ? void 0 : onSwipeMove(e);
      },
      onPanResponderEnd: function (e, param) {
        var {
          dx,
          dy
        } = param;
        if (!isDeltaInDirection({
          x: dx,
          y: dy
        }, context.swipeDirection, context.swipeThreshold)) {
          if (!import_constants.isWeb) {
            handleResume === null || handleResume === void 0 ? void 0 : handleResume();
          }
          onSwipeCancel === null || onSwipeCancel === void 0 ? void 0 : onSwipeCancel(e);
          animatedNumber.setValue(0, {
            type: "spring"
          });
        }
      }
    });
  }, [handlePause, handleResume]);
  var themeName = (0, import_core.useThemeName)();
  return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
    children: [announceTextContent && /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ToastAnnounce.ToastAnnounce, {
      scope,
      // Toasts are always role=status to avoid stuttering issues with role=alert in SRs.
      // biome-ignore lint/a11y/useSemanticElements: <explanation>
      role: "status",
      "aria-live": type === "foreground" ? "assertive" : "polite",
      "aria-atomic": true,
      children: announceTextContent
    }), /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_portal.PortalItem, {
      hostName: viewportName !== null && viewportName !== void 0 ? viewportName : "default",
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(ToastInteractiveProvider, {
        scope,
        onClose: function () {
          handleClose();
        },
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_dismissable.Dismissable, {
          // asChild
          onEscapeKeyDown: (0, import_helpers.composeEventHandlers)(onEscapeKeyDown, function () {
            if (!context.isFocusedToastEscapeKeyDownRef.current) {
              handleClose();
            }
            context.isFocusedToastEscapeKeyDownRef.current = false;
          }),
          children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.Theme, {
            contain: true,
            forceClassName: true,
            name: themeName,
            children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(AnimatedView, {
              ...(panResponder === null || panResponder === void 0 ? void 0 : panResponder.panHandlers),
              style: [{
                margin: "auto"
              }, animatedStyles],
              children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_ToastProvider.Collection.ItemSlot, {
                scope: context.toastScope,
                children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(ToastImplFrame, {
                  // Ensure toasts are announced as status list or status when focused
                  role: "status",
                  "aria-live": "off",
                  "aria-atomic": true,
                  "data-state": open ? "open" : "closed",
                  "data-swipe-direction": context.swipeDirection,
                  pointerEvents: "auto",
                  "$platform-web": {
                    touchAction: "none",
                    userSelect: "none"
                  },
                  ...toastProps,
                  ref: composedRefs,
                  ...(import_constants.isWeb && {
                    onKeyDown: (0, import_helpers.composeEventHandlers)(props.onKeyDown, function (event) {
                      if (event.key !== "Escape") return;
                      onEscapeKeyDown === null || onEscapeKeyDown === void 0 ? void 0 : onEscapeKeyDown(event);
                      if (!event.defaultPrevented) {
                        context.isFocusedToastEscapeKeyDownRef.current = true;
                        handleClose();
                      }
                    })
                  })
                })
              })
            })
          })
        })
      }, props.id)
    })]
  });
});
ToastImpl.propTypes = {
  type(props) {
    if (props.type && !["foreground", "background"].includes(props.type)) {
      var error = `Invalid prop \`type\` supplied to \`${import_constants2.TOAST_NAME}\`. Expected \`foreground | background\`.`;
      return new Error(error);
    }
    return null;
  }
};
var isDeltaInDirection = function (delta, direction) {
  var threshold = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : 0;
  var deltaX = Math.abs(delta.x);
  var deltaY = Math.abs(delta.y);
  var isDeltaX = deltaX > deltaY;
  if (direction === "left" || direction === "right" || direction === "horizontal") {
    return isDeltaX && deltaX > threshold;
  }
  return !isDeltaX && deltaY > threshold;
};
function getAnnounceTextContent(container) {
  if (!import_constants.isWeb) return "";
  var textContent = [];
  var childNodes = Array.from(container.childNodes);
  childNodes.forEach(function (node) {
    if (node.nodeType === node.TEXT_NODE && node.textContent) textContent.push(node.textContent);
    if (isHTMLElement(node)) {
      var isHidden = node.ariaHidden || node.hidden || node.style.display === "none";
      var isExcluded = node.dataset.toastAnnounceExclude === "";
      if (!isHidden) {
        if (isExcluded) {
          var altText = node.dataset.toastAnnounceAlt;
          if (altText) textContent.push(altText);
        } else {
          textContent.push(...getAnnounceTextContent(node));
        }
      }
    }
  });
  return textContent;
}
function isHTMLElement(node) {
  return node.nodeType === node.ELEMENT_NODE;
}
var GESTURE_GRANT_THRESHOLD = 10;
var shouldGrantGestureMove = function (dir, param) {
  var {
    dx,
    dy
  } = param;
  if ((dir === "horizontal" || dir === "left") && dx < -GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "horizontal" || dir === "right") && dx > GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "vertical" || dir === "up") && dy > -GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  if ((dir === "vertical" || dir === "down") && dy < GESTURE_GRANT_THRESHOLD) {
    return true;
  }
  return false;
};
var getGestureDistance = function (dir, param) {
  var {
    dx,
    dy
  } = param;
  var y = 0;
  var x = 0;
  if (dir === "horizontal") x = dx;else if (dir === "left") x = Math.min(0, dx);else if (dir === "right") x = Math.max(0, dx);else if (dir === "vertical") y = dy;else if (dir === "up") y = Math.min(0, dy);else if (dir === "down") y = Math.max(0, dy);
  return {
    x,
    y
  };
};
