// @tamagui/toast/dist/cjs/useAnimatedDragGesture.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useAnimatedDragGesture_native_exports = {};
__export(useAnimatedDragGesture_native_exports, {
  useAnimatedDragGesture: () => useAnimatedDragGesture
});
module.exports = __toCommonJS(useAnimatedDragGesture_native_exports);
var import_native = require("@tamagui/native");
var React = __toESM(require("react"), 1);
var import_react_native = require("react-native");
var VELOCITY_THRESHOLD = 0.11;
var GESTURE_GRANT_THRESHOLD = 10;
function resisted(delta) {
  var maxResist = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 25;
  if (delta >= 0) return delta;
  var pastBoundary = Math.abs(delta);
  var resistedDistance = Math.sqrt(pastBoundary) * 2;
  return -Math.min(resistedDistance, maxResist);
}
var EXIT_DRAG_CAP = 80;
function cappedExit(delta) {
  if (Math.abs(delta) <= EXIT_DRAG_CAP) return delta;
  var sign = delta > 0 ? 1 : -1;
  var overshoot = Math.abs(delta) - EXIT_DRAG_CAP;
  return sign * (EXIT_DRAG_CAP + Math.sqrt(overshoot) * 2);
}
function computeOffset(direction, dx, dy) {
  var offsetX = 0;
  var offsetY = 0;
  if (direction === "right") {
    offsetX = dx > 0 ? cappedExit(dx) : resisted(dx);
  } else if (direction === "left") {
    offsetX = dx < 0 ? cappedExit(dx) : -resisted(-dx);
  } else if (direction === "down") {
    offsetY = dy > 0 ? cappedExit(dy) : resisted(dy);
  } else if (direction === "up") {
    offsetY = dy < 0 ? cappedExit(dy) : -resisted(-dy);
  } else if (direction === "horizontal") {
    offsetX = cappedExit(dx);
  } else if (direction === "vertical") {
    offsetY = cappedExit(dy);
  }
  return {
    offsetX,
    offsetY
  };
}
function computeExitDirection(direction, dx, dy) {
  if (direction === "right" && dx > 0) return "right";
  if (direction === "left" && dx < 0) return "left";
  if (direction === "horizontal") {
    if (Math.abs(dx) > Math.abs(dy)) return dx > 0 ? "right" : "left";
  }
  if (direction === "down" && dy > 0) return "down";
  if (direction === "up" && dy < 0) return "up";
  if (direction === "vertical") {
    if (Math.abs(dy) > Math.abs(dx)) return dy > 0 ? "down" : "up";
  }
  return null;
}
function shouldGrantGestureMove(dir, dx, dy) {
  var absDx = Math.abs(dx);
  var absDy = Math.abs(dy);
  if ((dir === "horizontal" || dir === "left" || dir === "right") && absDx > GESTURE_GRANT_THRESHOLD && absDx > absDy) {
    return true;
  }
  if ((dir === "vertical" || dir === "up" || dir === "down") && absDy > GESTURE_GRANT_THRESHOLD && absDy > absDx) {
    return true;
  }
  return false;
}
function useAnimatedDragGesture(options) {
  var {
    direction,
    threshold,
    disabled
  } = options;
  var [isDragging, setIsDragging] = React.useState(false);
  var isHorizontal = direction === "left" || direction === "right" || direction === "horizontal";
  var gestureRef = React.useRef(null);
  var onDragMoveRef = React.useRef(options.onDragMove);
  var onDragStartRef = React.useRef(options.onDragStart);
  var onDismissRef = React.useRef(options.onDismiss);
  var onCancelRef = React.useRef(options.onCancel);
  onDragMoveRef.current = options.onDragMove;
  onDragStartRef.current = options.onDragStart;
  onDismissRef.current = options.onDismiss;
  onCancelRef.current = options.onCancel;
  var rnghEnabled = (0, import_native.getGestureHandler)().isEnabled;
  var gesture = React.useMemo(function () {
    if (!rnghEnabled || disabled) return null;
    var {
      Gesture
    } = (0, import_native.getGestureHandler)().state;
    if (!Gesture) return null;
    var pan = Gesture.Pan().withRef(gestureRef).shouldCancelWhenOutside(false).runOnJS(true);
    if (isHorizontal) {
      pan.activeOffsetX([-10, 10]);
      pan.failOffsetY([-20, 20]);
    } else {
      pan.activeOffsetY([-10, 10]);
      pan.failOffsetX([-20, 20]);
    }
    pan.onStart(function () {
      var _onDragStartRef_current;
      setIsDragging(true);
      (_onDragStartRef_current = onDragStartRef.current) === null || _onDragStartRef_current === void 0 ? void 0 : _onDragStartRef_current.call(onDragStartRef);
    }).onChange(function (event) {
      var {
        offsetX,
        offsetY
      } = computeOffset(direction, event.translationX, event.translationY);
      onDragMoveRef.current(offsetX, offsetY);
    }).onEnd(function (event) {
      var dx = event.translationX;
      var dy = event.translationY;
      var relevantDelta = isHorizontal ? dx : dy;
      var relevantVelocity = isHorizontal ? Math.abs(event.velocityX / 1e3) : Math.abs(event.velocityY / 1e3);
      var passedThreshold = Math.abs(relevantDelta) >= threshold;
      var hasVelocity = relevantVelocity > VELOCITY_THRESHOLD;
      var exitDirection = computeExitDirection(direction, dx, dy);
      var shouldDismiss = exitDirection && (passedThreshold || hasVelocity);
      setIsDragging(false);
      if (shouldDismiss && exitDirection) {
        onDismissRef.current(exitDirection, relevantVelocity);
      } else {
        onCancelRef.current();
      }
    }).onFinalize(function () {
      setIsDragging(false);
    });
    return pan;
  }, [disabled, direction, threshold, isHorizontal, rnghEnabled]);
  var panResponder = React.useMemo(function () {
    if (rnghEnabled || disabled) return null;
    return import_react_native.PanResponder.create({
      onMoveShouldSetPanResponder: function (_e, g) {
        return shouldGrantGestureMove(direction, g.dx, g.dy);
      },
      onMoveShouldSetPanResponderCapture: function (_e, g) {
        return shouldGrantGestureMove(direction, g.dx, g.dy);
      },
      onPanResponderTerminationRequest: function () {
        return false;
      },
      onPanResponderGrant: function () {
        var _onDragStartRef_current;
        setIsDragging(true);
        (_onDragStartRef_current = onDragStartRef.current) === null || _onDragStartRef_current === void 0 ? void 0 : _onDragStartRef_current.call(onDragStartRef);
      },
      onPanResponderMove: function (_e, g) {
        var {
          offsetX,
          offsetY
        } = computeOffset(direction, g.dx, g.dy);
        onDragMoveRef.current(offsetX, offsetY);
      },
      onPanResponderRelease: function (_e, g) {
        var {
          dx,
          dy,
          vx,
          vy
        } = g;
        var relevantDelta = isHorizontal ? dx : dy;
        var relevantVelocity = isHorizontal ? Math.abs(vx) : Math.abs(vy);
        var passedThreshold = Math.abs(relevantDelta) >= threshold;
        var hasVelocity = relevantVelocity > VELOCITY_THRESHOLD;
        var exitDirection = computeExitDirection(direction, dx, dy);
        var shouldDismiss = exitDirection && (passedThreshold || hasVelocity);
        setIsDragging(false);
        if (shouldDismiss && exitDirection) {
          onDismissRef.current(exitDirection, relevantVelocity);
        } else {
          onCancelRef.current();
        }
      },
      onPanResponderTerminate: function () {
        setIsDragging(false);
        onCancelRef.current();
      }
    });
  }, [disabled, direction, threshold, isHorizontal, rnghEnabled]);
  var _panResponder_panHandlers;
  return {
    isDragging,
    gestureHandlers: (_panResponder_panHandlers = panResponder === null || panResponder === void 0 ? void 0 : panResponder.panHandlers) !== null && _panResponder_panHandlers !== void 0 ? _panResponder_panHandlers : {},
    gesture
  };
}
