// @tamagui/toast/dist/cjs/useAnimatedDragGesture.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useAnimatedDragGesture_exports = {};
__export(useAnimatedDragGesture_exports, {
  useAnimatedDragGesture: () => useAnimatedDragGesture
});
module.exports = __toCommonJS(useAnimatedDragGesture_exports);
var React = __toESM(require("react"), 1);
const VELOCITY_THRESHOLD = 0.11;
function resisted(delta, maxResist = 25) {
  if (delta >= 0) return delta;
  const pastBoundary = Math.abs(delta);
  const resistedDistance = Math.sqrt(pastBoundary) * 2;
  return -Math.min(resistedDistance, maxResist);
}
const EXIT_DRAG_CAP = 80;
function cappedExit(delta) {
  if (Math.abs(delta) <= EXIT_DRAG_CAP) return delta;
  const sign = delta > 0 ? 1 : -1;
  const overshoot = Math.abs(delta) - EXIT_DRAG_CAP;
  return sign * (EXIT_DRAG_CAP + Math.sqrt(overshoot) * 2);
}
function useAnimatedDragGesture(options) {
  const {
    direction,
    threshold,
    disabled,
    expanded,
    onDragMove,
    onDragStart,
    onDismiss,
    onCancel
  } = options;
  const [isDragging, setIsDragging] = React.useState(false);
  const dragStartRef = React.useRef(null);
  const lockedDirectionRef = React.useRef(null);
  const captureElementRef = React.useRef(null);
  const isHorizontal = direction === "left" || direction === "right" || direction === "horizontal";
  const isVertical = direction === "up" || direction === "down" || direction === "vertical";
  const preventSelectRef = React.useRef(null);
  function startPreventingSelection() {
    if (typeof document === "undefined") return;
    window.getSelection()?.removeAllRanges();
    const handler = e => e.preventDefault();
    preventSelectRef.current = handler;
    document.addEventListener("selectstart", handler, true);
  }
  function stopPreventingSelection() {
    if (preventSelectRef.current) {
      document.removeEventListener("selectstart", preventSelectRef.current, true);
      preventSelectRef.current = null;
    }
  }
  const cleanup = React.useCallback(() => {
    dragStartRef.current = null;
    lockedDirectionRef.current = null;
    setIsDragging(false);
    stopPreventingSelection();
  }, []);
  React.useEffect(() => {
    return () => {
      if (dragStartRef.current) {
        if (captureElementRef.current && dragStartRef.current.pointerId) {
          try {
            captureElementRef.current.releasePointerCapture(dragStartRef.current.pointerId);
          } catch {}
        }
        cleanup();
      }
    };
  }, [cleanup]);
  const handlePointerDown = React.useCallback(event => {
    if (disabled) return;
    if (event.button !== 0) return;
    const target = event.target;
    if (target.closest('button, a, input, textarea, select, [role="button"]')) {
      return;
    }
    const hasSelection = (window.getSelection()?.toString().length ?? 0) > 0;
    if (hasSelection) return;
    const captureElement = event.currentTarget;
    captureElementRef.current = captureElement;
    captureElement.setPointerCapture(event.pointerId);
    dragStartRef.current = {
      startX: event.clientX,
      startY: event.clientY,
      startTime: Date.now(),
      pointerId: event.pointerId
    };
    startPreventingSelection();
    setIsDragging(true);
    onDragStart?.();
  }, [disabled, onDragStart]);
  const handlePointerMove = React.useCallback(event => {
    if (!dragStartRef.current || disabled) return;
    const deltaX = event.clientX - dragStartRef.current.startX;
    const deltaY = event.clientY - dragStartRef.current.startY;
    if (!lockedDirectionRef.current && (Math.abs(deltaX) > 1 || Math.abs(deltaY) > 1)) {
      lockedDirectionRef.current = Math.abs(deltaX) > Math.abs(deltaY) ? "x" : "y";
    }
    let offsetX = 0;
    let offsetY = 0;
    if (direction === "right") {
      offsetX = deltaX > 0 ? cappedExit(deltaX) : resisted(deltaX);
    } else if (direction === "left") {
      offsetX = deltaX < 0 ? cappedExit(deltaX) : -resisted(-deltaX);
    } else if (direction === "down") {
      offsetY = deltaY > 0 ? cappedExit(deltaY) : resisted(deltaY);
    } else if (direction === "up") {
      offsetY = deltaY < 0 ? cappedExit(deltaY) : -resisted(-deltaY);
    } else if (direction === "horizontal") {
      offsetX = cappedExit(deltaX);
    } else if (direction === "vertical") {
      offsetY = cappedExit(deltaY);
    }
    onDragMove(offsetX, offsetY);
  }, [disabled, direction, expanded, isHorizontal, isVertical, onDragMove]);
  const handlePointerUp = React.useCallback(event => {
    if (!dragStartRef.current || disabled) return;
    const deltaX = event.clientX - dragStartRef.current.startX;
    const deltaY = event.clientY - dragStartRef.current.startY;
    const timeTaken = Date.now() - dragStartRef.current.startTime;
    const velocityX = Math.abs(deltaX) / timeTaken;
    const velocityY = Math.abs(deltaY) / timeTaken;
    const lockedDirection = lockedDirectionRef.current;
    const isLockedToWrongAxis = lockedDirection === "y" && isHorizontal || lockedDirection === "x" && isVertical;
    const relevantDelta = isHorizontal ? deltaX : deltaY;
    const relevantVelocity = isHorizontal ? velocityX : velocityY;
    const passedThreshold = Math.abs(relevantDelta) >= threshold;
    const hasVelocity = relevantVelocity > VELOCITY_THRESHOLD;
    let exitDirection = null;
    if (!isLockedToWrongAxis) {
      if (direction === "right" && deltaX > 0) exitDirection = "right";else if (direction === "left" && deltaX < 0) exitDirection = "left";else if (direction === "horizontal") {
        if (Math.abs(deltaX) > Math.abs(deltaY)) {
          exitDirection = deltaX > 0 ? "right" : "left";
        }
      } else if (direction === "down" && deltaY > 0) exitDirection = "down";else if (direction === "up" && deltaY < 0) exitDirection = "up";else if (direction === "vertical") {
        if (Math.abs(deltaY) > Math.abs(deltaX)) {
          exitDirection = deltaY > 0 ? "down" : "up";
        }
      }
    }
    const shouldDismiss = exitDirection && (passedThreshold || hasVelocity);
    if (captureElementRef.current) {
      try {
        captureElementRef.current.releasePointerCapture(event.pointerId);
      } catch {}
    }
    cleanup();
    if (shouldDismiss && exitDirection) {
      onDismiss(exitDirection, relevantVelocity);
    } else {
      onCancel();
    }
  }, [disabled, direction, threshold, isHorizontal, isVertical, onDismiss, onCancel, cleanup]);
  const handlePointerCancel = React.useCallback(event => {
    if (captureElementRef.current) {
      try {
        captureElementRef.current.releasePointerCapture(event.pointerId);
      } catch {}
    }
    cleanup();
    onCancel();
  }, [onCancel, cleanup]);
  const handleLostPointerCapture = React.useCallback(() => {
    if (dragStartRef.current) {
      cleanup();
      onCancel();
    }
  }, [onCancel, cleanup]);
  const gestureHandlers = {
    onPointerDown: handlePointerDown,
    onPointerMove: handlePointerMove,
    onPointerUp: handlePointerUp,
    onPointerCancel: handlePointerCancel,
    onLostPointerCapture: handleLostPointerCapture
  };
  return {
    isDragging,
    gestureHandlers,
    gesture: null
  };
}
