// @tamagui/progress/dist/esm/Progress.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { getVariableValue, isWeb, styled } from "@tamagui/core";
import { createContextScope } from "@tamagui/create-context";
import { getSize } from "@tamagui/get-token";
import { withStaticProperties } from "@tamagui/helpers";
import { YStack } from "@tamagui/stacks";
import { useState } from "react";
var PROGRESS_NAME = "Progress";
var [createProgressContext, createProgressScope] = createContextScope(PROGRESS_NAME);
var [ProgressProvider, useProgressContext] = createProgressContext(PROGRESS_NAME);
var INDICATOR_NAME = "ProgressIndicator";
var ProgressIndicatorFrame = styled(YStack, {
  name: INDICATOR_NAME,
  variants: {
    unstyled: {
      false: {
        height: "100%",
        width: "100%",
        backgroundColor: "$background"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var ProgressIndicator = ProgressIndicatorFrame.styleable(function ProgressIndicator2(props, forwardedRef) {
  var {
    __scopeProgress,
    transition,
    ...indicatorProps
  } = props;
  var context = useProgressContext(INDICATOR_NAME, __scopeProgress);
  var _context_value;
  var progressRatio = ((_context_value = context.value) !== null && _context_value !== void 0 ? _context_value : 0) / context.max;
  var x;
  if (isWeb) {
    x = `${-100 + progressRatio * 50}%`;
  } else {
    var baseWidth = context.width || 0;
    x = Math.ceil(-baseWidth * (2 - progressRatio));
  }
  var _context_value1;
  return /* @__PURE__ */_jsx(ProgressIndicatorFrame, {
    "data-state": getProgressState(context.value, context.max),
    "data-value": (_context_value1 = context.value) !== null && _context_value1 !== void 0 ? _context_value1 : void 0,
    "data-max": context.max,
    x,
    width: "200%",
    ...(!props.unstyled && {
      animateOnly: ["transform"],
      // on native, hide until we have width measurement
      ...(!isWeb && context.width === 0 && {
        opacity: 0
      })
    }),
    ...indicatorProps,
    ref: forwardedRef,
    transition: !isWeb && !context.width ? null : transition
  });
});
function defaultGetValueLabel(value, max) {
  return `${Math.round(value / max * 100)}%`;
}
function getProgressState(value, maxValue) {
  return value == null ? "indeterminate" : value === maxValue ? "complete" : "loading";
}
function isNumber(value) {
  return typeof value === "number";
}
function isValidMaxNumber(max) {
  return isNumber(max) && !Number.isNaN(max) && max > 0;
}
function isValidValueNumber(value, max) {
  return isNumber(value) && !Number.isNaN(value) && value <= max && value >= 0;
}
var DEFAULT_MAX = 100;
var ProgressFrame = styled(YStack, {
  name: "Progress",
  variants: {
    unstyled: {
      false: {
        borderRadius: 1e5,
        overflow: "hidden",
        backgroundColor: "$background"
      }
    },
    size: {
      "...size": function (val) {
        var size = Math.round(getVariableValue(getSize(val)) * 0.25);
        return {
          height: size,
          minWidth: getVariableValue(size) * 20,
          width: "100%"
        };
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var Progress = withStaticProperties(ProgressFrame.styleable(function Progress2(props, forwardedRef) {
  var {
    // @ts-expect-error
    __scopeProgress,
    value: valueProp,
    max: maxProp,
    getValueLabel = defaultGetValueLabel,
    size = "$true",
    ...progressProps
  } = props;
  var max = isValidMaxNumber(maxProp) ? maxProp : DEFAULT_MAX;
  var value = isValidValueNumber(valueProp, max) ? Math.round(valueProp) : null;
  var valueLabel = isNumber(value) ? getValueLabel(value, max) : void 0;
  var [width, setWidth] = useState(0);
  return /* @__PURE__ */_jsx(ProgressProvider, {
    scope: __scopeProgress,
    value,
    max,
    width,
    children: /* @__PURE__ */_jsx(ProgressFrame, {
      "aria-valuemax": max,
      "aria-valuemin": 0,
      "aria-valuenow": isNumber(value) ? value : void 0,
      "aria-valuetext": valueLabel,
      // @ts-ignore
      role: "progressbar",
      "data-state": getProgressState(value, max),
      "data-value": value !== null && value !== void 0 ? value : void 0,
      "data-max": max,
      ...(progressProps.unstyled !== true && {
        size
      }),
      ...progressProps,
      ...(!isWeb && {
        onLayout: function (e) {
          var _progressProps_onLayout;
          var newWidth = Math.round(e.nativeEvent.layout.width);
          if (newWidth !== width) {
            setWidth(newWidth);
          }
          (_progressProps_onLayout = progressProps.onLayout) === null || _progressProps_onLayout === void 0 ? void 0 : _progressProps_onLayout.call(progressProps, e);
        }
      }),
      ref: forwardedRef
    })
  });
}), {
  Indicator: ProgressIndicator
});
export { Progress, ProgressFrame, ProgressIndicator, ProgressIndicatorFrame, createProgressScope };
