// @tamagui/tooltip/dist/cjs/Tooltip.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Tooltip_exports = {};
__export(Tooltip_exports, {
  Tooltip: () => Tooltip2,
  TooltipGroup: () => TooltipGroup,
  closeOpenTooltips: () => closeOpenTooltips
});
module.exports = __toCommonJS(Tooltip_exports);
var import_polyfill_dev = require("@tamagui/polyfill-dev");
var import_floating = require("@tamagui/floating");
var import_core = require("@tamagui/core");
var import_floating2 = require("@tamagui/floating");
var import_get_token = require("@tamagui/get-token");
var import_helpers = require("@tamagui/helpers");
var import_popover = require("@tamagui/popover");
var import_popper = require("@tamagui/popper");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var React = __toESM(require("react"), 1);
var import_jsx_runtime = require("react/jsx-runtime");
const TOOLTIP_SCOPE = "";
const ALWAYS_DISABLE_TOOLTIP = {
  focus: true,
  "remove-scroll": true
  // it's nice to hit escape to hide a tooltip
  // dismiss: true
};
const TooltipContent = import_popper.PopperContentFrame.styleable((props, ref) => {
  const preventAnimation = React.useContext(PreventTooltipAnimationContext);
  const zIndexFromContext = React.useContext(TooltipZIndexContext);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_popover.PopoverContent, {
    scope: props.scope || TOOLTIP_SCOPE,
    alwaysDisable: ALWAYS_DISABLE_TOOLTIP,
    ...(!props.unstyled && {
      backgroundColor: "$background",
      alignItems: "center",
      pointerEvents: "none",
      size: "$true"
    }),
    ref,
    ...(zIndexFromContext !== void 0 && {
      zIndex: zIndexFromContext
    }),
    ...props,
    ...(preventAnimation && {
      transition: null
    })
  });
}, {
  staticConfig: {
    componentName: "Tooltip"
  }
});
const TooltipArrow = React.forwardRef((props, ref) => {
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_popover.PopoverArrow, {
    scope: props.scope || TOOLTIP_SCOPE,
    componentName: "Tooltip",
    ref,
    ...props
  });
});
const PreventTooltipAnimationContext = React.createContext(false);
const TooltipZIndexContext = React.createContext(void 0);
const TooltipGroup = ({
  children,
  delay,
  preventAnimation = false,
  timeoutMs
}) => {
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(PreventTooltipAnimationContext.Provider, {
    value: preventAnimation,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_floating.FloatingDelayGroup, {
      timeoutMs,
      delay: React.useMemo(() => delay, [JSON.stringify(delay)]),
      children
    })
  });
};
const setOpens = /* @__PURE__ */new Set();
const closeOpenTooltips = () => {
  setOpens.forEach(x => x(false));
};
const TooltipComponent = React.forwardRef(function Tooltip(props, ref) {
  "use no memo";

  const {
    children,
    delay: delayProp,
    restMs: restMsProp,
    onOpenChange: onOpenChangeProp,
    focus,
    open: openProp,
    disableAutoCloseOnScroll,
    zIndex,
    scope = TOOLTIP_SCOPE,
    ...restProps
  } = props;
  const triggerRef = React.useRef(null);
  const [hasCustomAnchor, setHasCustomAnchor] = React.useState(false);
  const {
    delay: delayGroup,
    setCurrentId
  } = (0, import_floating.useDelayGroupContext)();
  const delay = delayProp !== void 0 ? delayProp : delayGroup ?? 400;
  const restMs = restMsProp ?? (typeof delay === "number" ? delay : 0);
  const [open, setOpen] = (0, import_use_controllable_state.useControllableState)({
    prop: openProp,
    defaultProp: false,
    onChange: onOpenChangeProp
  });
  const id = props.groupId;
  const onOpenChange = (0, import_core.useEvent)(open2 => {
    if (open2) {
      setCurrentId(id);
    }
    setOpen(open2);
  });
  React.useEffect(() => {
    if (!open) return;
    if (disableAutoCloseOnScroll) return;
    if (typeof document === "undefined") return;
    const closeIt = () => {
      setOpen(false);
    };
    setOpens.add(setOpen);
    document.documentElement.addEventListener("scroll", closeIt);
    return () => {
      setOpens.delete(setOpen);
      document.documentElement.removeEventListener("scroll", closeIt);
    };
  }, [open, disableAutoCloseOnScroll]);
  const floatingContext = (0, import_popover.useFloatingContext)({
    open,
    setOpen: onOpenChange,
    disable: false,
    disableFocus: false,
    hoverable: true,
    role: "tooltip",
    focus,
    groupId: id,
    delay,
    restMs
  });
  const onCustomAnchorAdd = React.useCallback(() => setHasCustomAnchor(true), []);
  const onCustomAnchorRemove = React.useCallback(() => setHasCustomAnchor(false), []);
  const contentId = React.useId();
  const smallerSize = props.unstyled ? null : (0, import_get_token.getSize)("$true", {
    shift: -2,
    bounds: [0]
  });
  const content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_floating2.FloatingOverrideContext.Provider, {
    value: floatingContext,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_popper.Popper, {
      scope,
      size: smallerSize?.key,
      allowFlip: true,
      stayInFrame: true,
      open,
      ...restProps,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_popover.PopoverContextProvider, {
        scope,
        contentId,
        triggerRef,
        open,
        onOpenChange: setOpen,
        onOpenToggle: voidFn,
        hasCustomAnchor,
        onCustomAnchorAdd,
        onCustomAnchorRemove,
        children
      })
    })
  });
  if (zIndex !== void 0) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(TooltipZIndexContext.Provider, {
      value: zIndex,
      children: content
    });
  }
  return content;
});
const TooltipTrigger = React.forwardRef(function TooltipTrigger2(props, ref) {
  const {
    scope,
    ...rest
  } = props;
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_popover.PopoverTrigger, {
    ...rest,
    scope: scope || TOOLTIP_SCOPE,
    ref
  });
});
const TooltipAnchor = React.forwardRef(function TooltipAnchor2(props, ref) {
  const {
    scope,
    ...rest
  } = props;
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_popover.PopoverAnchor, {
    ...rest,
    scope: scope || TOOLTIP_SCOPE,
    ref
  });
});
const Tooltip2 = (0, import_helpers.withStaticProperties)(TooltipComponent, {
  Anchor: TooltipAnchor,
  Arrow: TooltipArrow,
  Content: TooltipContent,
  Trigger: TooltipTrigger
});
const voidFn = () => {};
