// @tamagui/floating/dist/esm/interactions/useDelayGroup.mjs
import * as React from "react";
const DelayGroupContext = React.createContext({
  currentId: null,
  setCurrentId: () => {},
  delay: 0,
  timeoutMs: 0,
  initialDelay: 0
});
function useDelayGroupContext() {
  return React.useContext(DelayGroupContext);
}
function FloatingDelayGroup({
  children,
  delay,
  timeoutMs = 0
}) {
  const [currentId, setCurrentIdRaw] = React.useState(null);
  const timeoutRef = React.useRef(void 0);
  const setCurrentId = React.useCallback(id => {
    clearTimeout(timeoutRef.current);
    if (id == null && timeoutMs > 0) {
      timeoutRef.current = setTimeout(() => {
        setCurrentIdRaw(null);
      }, timeoutMs);
    } else {
      setCurrentIdRaw(id);
    }
  }, [timeoutMs]);
  React.useEffect(() => {
    return () => clearTimeout(timeoutRef.current);
  }, []);
  const value = React.useMemo(() => ({
    currentId,
    setCurrentId,
    delay,
    timeoutMs,
    initialDelay: delay
  }), [currentId, setCurrentId, delay, timeoutMs]);
  return React.createElement(DelayGroupContext.Provider, {
    value
  }, children);
}
function useDelayGroup(context, options = {}) {
  const {
    id
  } = options;
  const groupContext = React.useContext(DelayGroupContext);
  React.useEffect(() => {
    if (!context.open && groupContext.currentId === id) {
      groupContext.setCurrentId(null);
    }
  }, [context.open, id]);
  React.useEffect(() => {
    if (groupContext.currentId != null && groupContext.currentId !== id && context.open) {
      context.onOpenChange(false);
    }
  }, [groupContext.currentId, id, context.open]);
  if (groupContext.currentId != null) {
    return {
      delay: {
        open: 1,
        close: getClose(groupContext.initialDelay)
      },
      currentId: groupContext.currentId
    };
  }
  return {
    delay: groupContext.initialDelay,
    currentId: groupContext.currentId
  };
}
function getClose(delay) {
  if (typeof delay === "number") return delay;
  return delay?.close ?? 0;
}
export { FloatingDelayGroup, useDelayGroup, useDelayGroupContext };
