// @tamagui/floating/dist/esm/interactions/useDelayGroup.native.js
import * as React from "react";
var DelayGroupContext = React.createContext({
  currentId: null,
  setCurrentId: function () {},
  delay: 0,
  timeoutMs: 0,
  initialDelay: 0
});
function useDelayGroupContext() {
  return React.useContext(DelayGroupContext);
}
function FloatingDelayGroup(param) {
  var {
    children,
    delay,
    timeoutMs = 0
  } = param;
  var [currentId, setCurrentIdRaw] = React.useState(null);
  var timeoutRef = React.useRef(void 0);
  var setCurrentId = React.useCallback(function (id) {
    clearTimeout(timeoutRef.current);
    if (id == null && timeoutMs > 0) {
      timeoutRef.current = setTimeout(function () {
        setCurrentIdRaw(null);
      }, timeoutMs);
    } else {
      setCurrentIdRaw(id);
    }
  }, [timeoutMs]);
  React.useEffect(function () {
    return function () {
      return clearTimeout(timeoutRef.current);
    };
  }, []);
  var value = React.useMemo(function () {
    return {
      currentId,
      setCurrentId,
      delay,
      timeoutMs,
      initialDelay: delay
    };
  }, [currentId, setCurrentId, delay, timeoutMs]);
  return React.createElement(DelayGroupContext.Provider, {
    value
  }, children);
}
function useDelayGroup(context) {
  var options = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
  var {
    id
  } = options;
  var groupContext = React.useContext(DelayGroupContext);
  React.useEffect(function () {
    if (!context.open && groupContext.currentId === id) {
      groupContext.setCurrentId(null);
    }
  }, [context.open, id]);
  React.useEffect(function () {
    if (groupContext.currentId != null && groupContext.currentId !== id && context.open) {
      context.onOpenChange(false);
    }
  }, [groupContext.currentId, id, context.open]);
  if (groupContext.currentId != null) {
    return {
      delay: {
        open: 1,
        close: getClose(groupContext.initialDelay)
      },
      currentId: groupContext.currentId
    };
  }
  return {
    delay: groupContext.initialDelay,
    currentId: groupContext.currentId
  };
}
function getClose(delay) {
  if (typeof delay === "number") return delay;
  var _delay_close;
  return (_delay_close = delay === null || delay === void 0 ? void 0 : delay.close) !== null && _delay_close !== void 0 ? _delay_close : 0;
}
export { FloatingDelayGroup, useDelayGroup, useDelayGroupContext };
