// @tamagui/floating/dist/esm/interactions/useFocus.native.js
import { useEffect, useMemo, useRef } from "react";
import { activeElement, clearTimeoutIfSet, contains, getDocument, getTarget, isElement, isHTMLElement, isMac, isSafari, isTypeableElement, matchesFocusVisible } from "./utils.native.js";
function isMacSafari() {
  return isMac() && isSafari();
}
function useFocus(context) {
  var props = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
  var {
    open,
    onOpenChange,
    events,
    dataRef,
    elements
  } = context;
  var {
    enabled = true,
    visibleOnly = true
  } = props;
  var blockFocusRef = useRef(false);
  var timeoutRef = useRef(-1);
  var keyboardModalityRef = useRef(true);
  useEffect(function () {
    if (!enabled) return;
    var win = getDocument(elements.domReference).defaultView || window;
    function onBlur() {
      if (!open && isHTMLElement(elements.domReference) && elements.domReference === activeElement(getDocument(elements.domReference))) {
        blockFocusRef.current = true;
      }
    }
    function onKeyDown() {
      keyboardModalityRef.current = true;
    }
    function onPointerDown() {
      keyboardModalityRef.current = false;
    }
    win.addEventListener("blur", onBlur);
    if (isMacSafari()) {
      win.addEventListener("keydown", onKeyDown, true);
      win.addEventListener("pointerdown", onPointerDown, true);
    }
    return function () {
      win.removeEventListener("blur", onBlur);
      if (isMacSafari()) {
        win.removeEventListener("keydown", onKeyDown, true);
        win.removeEventListener("pointerdown", onPointerDown, true);
      }
    };
  }, [elements.domReference, open, enabled]);
  useEffect(function () {
    if (!enabled) return;
    if (!events) return;
    function handleOpenChange(param) {
      var {
        reason
      } = param;
      if (reason === "reference-press" || reason === "escape-key") {
        blockFocusRef.current = true;
      }
    }
    events.on("openchange", handleOpenChange);
    return function () {
      events.off("openchange", handleOpenChange);
    };
  }, [events, enabled]);
  useEffect(function () {
    return function () {
      clearTimeoutIfSet(timeoutRef);
    };
  }, []);
  var reference = useMemo(function () {
    return {
      onMouseLeave() {
        blockFocusRef.current = false;
      },
      onFocus(event) {
        if (blockFocusRef.current) return;
        var target = getTarget(event.nativeEvent);
        if (visibleOnly && isElement(target)) {
          if (isMacSafari() && !event.relatedTarget) {
            if (!keyboardModalityRef.current && !isTypeableElement(target)) {
              return;
            }
          } else if (!matchesFocusVisible(target)) {
            return;
          }
        }
        onOpenChange(true, event.nativeEvent, "focus");
      },
      onBlur(event) {
        blockFocusRef.current = false;
        var relatedTarget = event.relatedTarget;
        var nativeEvent = event.nativeEvent;
        timeoutRef.current = window.setTimeout(function () {
          var activeEl = activeElement(elements.domReference ? elements.domReference.ownerDocument : document);
          if (!relatedTarget && activeEl === elements.domReference) return;
          if (contains(context.refs.floating.current, activeEl) || contains(elements.domReference, activeEl)) {
            return;
          }
          onOpenChange(false, nativeEvent, "focus");
        });
      }
    };
  }, [context.refs.floating, elements.domReference, onOpenChange, visibleOnly]);
  return useMemo(function () {
    return enabled ? {
      reference
    } : {};
  }, [enabled, reference]);
}
export { useFocus };
