// @tamagui/floating/dist/esm/interactions/useInnerOffset.native.js
import * as React from "react";
import * as ReactDOM from "react-dom";
import { useEvent } from "@tamagui/use-event";
function useInnerOffset(context, props) {
  var {
    open,
    elements
  } = context;
  var {
    enabled = true,
    overflowRef,
    scrollRef,
    onChange: unstable_onChange
  } = props;
  var onChange = useEvent(unstable_onChange);
  var controlledScrollingRef = React.useRef(false);
  var prevScrollTopRef = React.useRef(null);
  var initialOverflowRef = React.useRef(null);
  React.useEffect(function () {
    if (!enabled) return;
    function onWheel(e) {
      if (e.ctrlKey || !el || overflowRef.current == null) {
        return;
      }
      var dY = e.deltaY;
      var isAtTop = overflowRef.current.top >= -0.5;
      var isAtBottom = overflowRef.current.bottom >= -0.5;
      var remainingScroll = el.scrollHeight - el.clientHeight;
      var sign = dY < 0 ? -1 : 1;
      var method = dY < 0 ? "max" : "min";
      if (el.scrollHeight <= el.clientHeight) {
        return;
      }
      if (!isAtTop && dY > 0 || !isAtBottom && dY < 0) {
        e.preventDefault();
        ReactDOM.flushSync(function () {
          onChange(function (d) {
            return d + Math[method](dY, remainingScroll * sign);
          });
        });
      } else if (/firefox/i.test(navigator.userAgent)) {
        el.scrollTop += dY;
      }
    }
    var el = (scrollRef === null || scrollRef === void 0 ? void 0 : scrollRef.current) || elements.floating;
    if (open && el) {
      el.addEventListener("wheel", onWheel);
      requestAnimationFrame(function () {
        prevScrollTopRef.current = el.scrollTop;
        if (overflowRef.current != null) {
          initialOverflowRef.current = {
            ...overflowRef.current
          };
        }
      });
      return function () {
        prevScrollTopRef.current = null;
        initialOverflowRef.current = null;
        el.removeEventListener("wheel", onWheel);
      };
    }
  }, [enabled, open, elements.floating, overflowRef, scrollRef, onChange]);
  var floating = React.useMemo(function () {
    return {
      onKeyDown() {
        controlledScrollingRef.current = true;
      },
      onWheel() {
        controlledScrollingRef.current = false;
      },
      onPointerMove() {
        controlledScrollingRef.current = false;
      },
      onScroll() {
        var el = (scrollRef === null || scrollRef === void 0 ? void 0 : scrollRef.current) || elements.floating;
        if (!overflowRef.current || !el || !controlledScrollingRef.current) {
          return;
        }
        if (prevScrollTopRef.current !== null) {
          var scrollDiff = el.scrollTop - prevScrollTopRef.current;
          if (overflowRef.current.bottom < -0.5 && scrollDiff < -1 || overflowRef.current.top < -0.5 && scrollDiff > 1) {
            ReactDOM.flushSync(function () {
              return onChange(function (d) {
                return d + scrollDiff;
              });
            });
          }
        }
        requestAnimationFrame(function () {
          prevScrollTopRef.current = el.scrollTop;
        });
      }
    };
  }, [elements.floating, onChange, overflowRef, scrollRef]);
  return React.useMemo(function () {
    return enabled ? {
      floating
    } : {};
  }, [enabled, floating]);
}
export { useInnerOffset };
