// @tamagui/floating/dist/esm/interactions/useTypeahead.native.js
import { useLayoutEffect, useMemo, useRef } from "react";
import { useEvent } from "@tamagui/use-event";
import { clearTimeoutIfSet, stopEvent } from "./utils.native.js";
function useTypeahead(context, props) {
  var {
    open,
    dataRef
  } = context;
  var {
    listRef,
    activeIndex,
    onMatch: unstable_onMatch,
    onTypingChange: unstable_onTypingChange,
    enabled = true,
    findMatch = null,
    resetMs = 750,
    ignoreKeys = [],
    selectedIndex = null
  } = props;
  var timeoutIdRef = useRef(-1);
  var stringRef = useRef("");
  var _ref;
  var prevIndexRef = useRef((_ref = selectedIndex !== null && selectedIndex !== void 0 ? selectedIndex : activeIndex) !== null && _ref !== void 0 ? _ref : -1);
  var matchIndexRef = useRef(null);
  var onMatch = useEvent(unstable_onMatch || function () {});
  var onTypingChange = useEvent(unstable_onTypingChange || function () {});
  var findMatchRef = useRef(findMatch);
  findMatchRef.current = findMatch;
  var ignoreKeysRef = useRef(ignoreKeys);
  ignoreKeysRef.current = ignoreKeys;
  useLayoutEffect(function () {
    if (open) {
      clearTimeoutIfSet(timeoutIdRef);
      matchIndexRef.current = null;
      stringRef.current = "";
    }
  }, [open]);
  useLayoutEffect(function () {
    if (open && stringRef.current === "") {
      var _ref2;
      prevIndexRef.current = (_ref2 = selectedIndex !== null && selectedIndex !== void 0 ? selectedIndex : activeIndex) !== null && _ref2 !== void 0 ? _ref2 : -1;
    }
  }, [open, selectedIndex, activeIndex]);
  var setTypingChange = function (value) {
    if (value) {
      if (!dataRef.current.typing) {
        dataRef.current.typing = value;
        onTypingChange(value);
      }
    } else {
      if (dataRef.current.typing) {
        dataRef.current.typing = value;
        onTypingChange(value);
      }
    }
  };
  var onKeyDown = function (event) {
    function getMatchingIndex(list, orderedList, string) {
      var str = findMatchRef.current ? findMatchRef.current(orderedList, string) : orderedList.find(function (text) {
        return (text === null || text === void 0 ? void 0 : text.toLocaleLowerCase().indexOf(string.toLocaleLowerCase())) === 0;
      });
      return str ? list.indexOf(str) : -1;
    }
    var listContent = listRef.current;
    if (stringRef.current.length > 0 && stringRef.current[0] !== " ") {
      if (getMatchingIndex(listContent, listContent, stringRef.current) === -1) {
        setTypingChange(false);
      } else if (event.key === " ") {
        stopEvent(event);
      }
    }
    if (listContent == null || ignoreKeysRef.current.includes(event.key) ||
    // character key
    event.key.length !== 1 ||
    // modifier key
    event.ctrlKey || event.metaKey || event.altKey) {
      return;
    }
    if (open && event.key !== " ") {
      stopEvent(event);
      setTypingChange(true);
    }
    var allowRapidSuccessionOfFirstLetter = listContent.every(function (text) {
      var _text_, _text_1;
      return text ? ((_text_ = text[0]) === null || _text_ === void 0 ? void 0 : _text_.toLocaleLowerCase()) !== ((_text_1 = text[1]) === null || _text_1 === void 0 ? void 0 : _text_1.toLocaleLowerCase()) : true;
    });
    if (allowRapidSuccessionOfFirstLetter && stringRef.current === event.key) {
      stringRef.current = "";
      prevIndexRef.current = matchIndexRef.current;
    }
    stringRef.current += event.key;
    clearTimeoutIfSet(timeoutIdRef);
    timeoutIdRef.current = window.setTimeout(function () {
      stringRef.current = "";
      prevIndexRef.current = matchIndexRef.current;
      setTypingChange(false);
    }, resetMs);
    var prevIndex = prevIndexRef.current;
    var index = getMatchingIndex(listContent, [...listContent.slice((prevIndex || 0) + 1), ...listContent.slice(0, (prevIndex || 0) + 1)], stringRef.current);
    if (index !== -1) {
      onMatch(index);
      matchIndexRef.current = index;
    } else if (event.key !== " ") {
      stringRef.current = "";
      setTypingChange(false);
    }
  };
  var reference = useMemo(function () {
    return {
      onKeyDown
    };
  }, [open, enabled]);
  var floating = useMemo(function () {
    return {
      onKeyDown,
      onKeyUp(event) {
        if (event.key === " ") {
          setTypingChange(false);
        }
      }
    };
  }, [open, enabled]);
  return useMemo(function () {
    return enabled ? {
      reference,
      floating
    } : {};
  }, [enabled, reference, floating]);
}
export { useTypeahead };
