// @tamagui/popover/dist/cjs/useFloatingContext.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useFloatingContext_exports = {};
__export(useFloatingContext_exports, {
  useFloatingContext: () => useFloatingContext
});
module.exports = __toCommonJS(useFloatingContext_exports);
var import_react = __toESM(require("react"), 1);
var import_floating = require("@tamagui/floating");
const useFloatingContext = ({
  open,
  setOpen,
  disable,
  disableFocus,
  hoverable,
  role: roleProp = "dialog",
  focus: focusProp,
  groupId,
  delay: delayProp,
  restMs: restMsProp
}) => {
  "use no memo";

  const openRef = import_react.default.useRef(open);
  openRef.current = open;
  const hoverableRef = import_react.default.useRef(hoverable);
  hoverableRef.current = hoverable;
  const disableRef = import_react.default.useRef(disable);
  disableRef.current = disable;
  const disableFocusRef = import_react.default.useRef(disableFocus);
  disableFocusRef.current = disableFocus;
  const roleRef = import_react.default.useRef(roleProp);
  roleRef.current = roleProp;
  const focusRef = import_react.default.useRef(focusProp);
  focusRef.current = focusProp;
  const groupIdRef = import_react.default.useRef(groupId);
  groupIdRef.current = groupId;
  const delayRef = import_react.default.useRef(delayProp);
  delayRef.current = delayProp;
  const restMsRef = import_react.default.useRef(restMsProp);
  restMsRef.current = restMsProp;
  const events = import_react.default.useMemo(() => (0, import_floating.createFloatingEvents)(), []);
  const triggerElements = import_react.default.useMemo(() => new import_floating.PopupTriggerMap(), []);
  import_react.default.useEffect(() => {
    events.emit("openchange", {
      open
    });
  }, [open, events]);
  return import_react.default.useCallback(props => {
    const onTriggerRef = import_react.default.useRef(false);
    const restTimerRef = import_react.default.useRef(void 0);
    const graceRef = import_react.default.useRef(void 0);
    const pendingCloseRef = import_react.default.useRef(false);
    const isOverFloatingRef = import_react.default.useRef(false);
    const handleCloseActiveRef = import_react.default.useRef(false);
    import_react.default.useEffect(() => {
      return () => {
        clearTimeout(restTimerRef.current);
        clearTimeout(graceRef.current);
      };
    }, []);
    const onOpenChange = (val, event) => {
      if (val && event?.type === "mouseenter") {
        return;
      }
      if (!val && onTriggerRef.current && (event?.type === "mousemove" || event?.type === "mouseleave")) {
        pendingCloseRef.current = true;
        return;
      }
      const type = event?.type === "mousemove" || event?.type === "mouseenter" || event?.type === "mouseleave" ? "hover" : "press";
      setOpen(val, type);
    };
    const floating = (0, import_floating.useFloatingRaw)({
      ...props,
      open: openRef.current
    });
    const currentHoverable = hoverableRef.current;
    const dataRef = import_react.default.useRef({});
    dataRef.current.placement = floating.placement;
    const floatingRefs = floating.refs;
    const nullRef = {
      current: null
    };
    const interactionContext = {
      open: openRef.current,
      onOpenChange,
      refs: {
        reference: floatingRefs?.reference || nullRef,
        floating: floatingRefs?.floating || nullRef,
        domReference: floatingRefs?.reference || nullRef
      },
      elements: {
        get reference() {
          return floatingRefs?.reference?.current || null;
        },
        get floating() {
          return floatingRefs?.floating?.current || null;
        },
        get domReference() {
          return floatingRefs?.reference?.current || null;
        }
      },
      dataRef,
      events,
      triggerElements,
      handleCloseActiveRef
    };
    const {
      delay: groupDelay,
      currentId: groupCurrentId
    } = (0, import_floating.useDelayGroup)(interactionContext, {
      id: groupIdRef.current
    });
    const isInActiveGroup = groupIdRef.current && groupCurrentId != null && typeof groupDelay === "object";
    let delay;
    let restMs;
    if (isInActiveGroup) {
      delay = groupDelay;
      restMs = 0;
    } else if (delayRef.current !== void 0) {
      delay = delayRef.current;
      restMs = restMsRef.current ?? 0;
    } else {
      delay = currentHoverable && typeof currentHoverable === "object" ? currentHoverable.delay ?? 0 : 0;
      restMs = currentHoverable && typeof currentHoverable === "object" ? currentHoverable.restMs ?? 0 : 0;
    }
    const currentRole = roleRef.current;
    const currentFocus = focusRef.current;
    const {
      getReferenceProps,
      getFloatingProps: getFloatingPropsInner
    } = (0, import_floating.useInteractions)([currentHoverable ? (0, import_floating.useHover)(interactionContext, {
      enabled: !disableRef.current && !!currentHoverable,
      delay,
      restMs,
      handleClose: (0, import_floating.safePolygon)({
        requireIntent: true,
        buffer: 1,
        __debug: typeof window !== "undefined" && new URLSearchParams(window.location.search).get("debug") === "safePolygon"
      }),
      ...(typeof currentHoverable === "object" && currentHoverable)
    }) : (0, import_floating.useHover)(interactionContext, {
      enabled: false
    }), (0, import_floating.useFocus)(interactionContext, {
      enabled: !disableRef.current && !disableFocusRef.current,
      visibleOnly: true,
      ...currentFocus
    }), (0, import_floating.useRole)(interactionContext, {
      role: currentRole
    })]);
    const getFloatingProps = currentHoverable ? userProps => {
      const merged = getFloatingPropsInner(userProps);
      const origEnter = merged.onMouseEnter;
      const origLeave = merged.onMouseLeave;
      return {
        ...merged,
        onMouseEnter: e => {
          isOverFloatingRef.current = true;
          origEnter?.(e);
        },
        onMouseLeave: e => {
          isOverFloatingRef.current = false;
          origLeave?.(e);
        }
      };
    } : getFloatingPropsInner;
    const openDelay = typeof delay === "number" ? delay : delay?.open ?? 0;
    const closeDelay = typeof delay === "number" ? delay : delay?.close ?? 0;
    const setOpenWithDelay = () => {
      clearTimeout(restTimerRef.current);
      if (restMs && !openDelay) {
        restTimerRef.current = setTimeout(() => {
          setOpen(true, "hover");
        }, restMs);
      } else if (openDelay) {
        restTimerRef.current = setTimeout(() => {
          setOpen(true, "hover");
        }, openDelay);
      } else {
        setOpen(true, "hover");
      }
    };
    return {
      ...floating,
      open: openRef.current,
      triggerElements,
      getReferenceProps,
      getFloatingProps,
      onHoverReference: currentHoverable ? _event => {
        clearTimeout(graceRef.current);
        onTriggerRef.current = true;
        pendingCloseRef.current = false;
        clearTimeout(restTimerRef.current);
        if (openRef.current) return;
        setOpenWithDelay();
      } : void 0,
      onLeaveReference: currentHoverable ? () => {
        clearTimeout(restTimerRef.current);
        clearTimeout(graceRef.current);
        graceRef.current = setTimeout(() => {
          onTriggerRef.current = false;
          if (pendingCloseRef.current) {
            pendingCloseRef.current = false;
            setOpen(false, "hover");
            return;
          }
          if (openRef.current) {
            graceRef.current = setTimeout(() => {
              if (openRef.current && !isOverFloatingRef.current && !handleCloseActiveRef.current) {
                setOpen(false, "hover");
              }
            }, Math.max(250, closeDelay));
          }
        }, 40);
      } : void 0
    };
  }, [setOpen]);
};
