// @tamagui/button/dist/cjs/Button.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Button_exports = {};
__export(Button_exports, {
  Button: () => Button,
  ButtonContext: () => ButtonContext
});
module.exports = __toCommonJS(Button_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_font_size = require("@tamagui/font-size");
var import_get_button_sized = require("@tamagui/get-button-sized");
var import_helpers_tamagui = require("@tamagui/helpers-tamagui");
var import_stacks = require("@tamagui/stacks");
var import_text = require("@tamagui/text");
var import_web = require("@tamagui/web");
var import_react = require("react");
var context = (0, import_web.createStyledContext)({
  size: void 0,
  variant: void 0,
  color: void 0,
  elevation: void 0
});
var Frame = (0, import_web.styled)(import_web.View, {
  context,
  name: "Button",
  role: "button",
  render: /* @__PURE__ */(0, import_jsx_runtime.jsx)("button", {
    type: "button"
  }),
  tabIndex: 0,
  variants: {
    unstyled: {
      false: {
        size: "$true",
        justifyContent: "center",
        alignItems: "center",
        flexWrap: "nowrap",
        flexDirection: "row",
        cursor: "pointer",
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "transparent",
        hoverStyle: {
          backgroundColor: "$backgroundHover",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress",
          borderColor: "$borderColorHover"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineStyle: "solid",
          outlineWidth: 2
        }
      }
    },
    variant: {
      outlined: process.env.TAMAGUI_HEADLESS === "1" ? {} : {
        backgroundColor: "transparent",
        borderWidth: 1,
        borderColor: "$borderColor",
        hoverStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorPress"
        }
      }
    },
    circular: import_stacks.themeableVariants.circular,
    chromeless: import_stacks.themeableVariants.chromeless,
    size: {
      "...size": function (val, extras) {
        var buttonStyle = (0, import_get_button_sized.getButtonSized)(val, extras);
        var gap = (0, import_web.getTokenValue)(val);
        return {
          ...buttonStyle,
          gap
        };
      },
      ":number": function (val, extras) {
        var buttonStyle = (0, import_get_button_sized.getButtonSized)(val, extras);
        var gap = val * 0.4;
        return {
          ...buttonStyle,
          gap
        };
      }
    },
    elevation: {
      "...size": import_stacks.getElevation,
      ":number": import_stacks.getElevation
    },
    disabled: {
      true: {
        pointerEvents: "none",
        // @ts-ignore
        "aria-disabled": true
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var Text = (0, import_web.styled)(import_text.SizableText, {
  context,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        cursor: "pointer",
        // flexGrow 1 leads to inconsistent native style where text pushes to start of view
        flexGrow: 0,
        flexShrink: 1,
        ellipsis: true,
        color: "$color"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var Icon = function (props) {
  var {
    children,
    scaleIcon = 1,
    size
  } = props;
  var styledContext = context.useStyledContext();
  if (!styledContext) {
    throw new Error("Button.Icon must be used within a Button");
  }
  var sizeToken = size !== null && size !== void 0 ? size : styledContext.size;
  var iconColor = (0, import_helpers_tamagui.useCurrentColor)(styledContext.color);
  var iconSize = (typeof sizeToken === "number" ? sizeToken * 0.5 : (0, import_font_size.getFontSize)(sizeToken)) * scaleIcon;
  return (0, import_helpers_tamagui.getIcon)(children, {
    size: iconSize,
    color: iconColor
  });
};
var ButtonContext = (0, import_web.createStyledContext)({
  size: void 0,
  variant: void 0,
  color: void 0
});
var ButtonComponent = Frame.styleable(function (propsIn, ref) {
  var isNested = (0, import_react.useContext)(import_stacks.ButtonNestingContext);
  var processedProps = (0, import_web.useProps)(propsIn, {
    noNormalize: true,
    noExpand: true
  });
  var {
    children,
    iconSize,
    icon,
    iconAfter,
    scaleIcon = 1,
    ...props
  } = processedProps;
  var size = propsIn.size || (propsIn.unstyled ? void 0 : "$true");
  var styledContext = context.useStyledContext();
  var iconColor = (0, import_helpers_tamagui.useCurrentColor)(styledContext === null || styledContext === void 0 ? void 0 : styledContext.color);
  var _ref;
  var finalSize = (_ref = iconSize !== null && iconSize !== void 0 ? iconSize : size) !== null && _ref !== void 0 ? _ref : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size;
  var iconSizeNumber = (typeof finalSize === "number" ? finalSize * 0.5 : (0, import_font_size.getFontSize)(finalSize)) * scaleIcon;
  var [themedIcon, themedIconAfter] = [icon, iconAfter].map(function (icon2) {
    if (!icon2) return null;
    return (0, import_helpers_tamagui.getIcon)(icon2, {
      size: iconSizeNumber,
      color: iconColor
    });
  });
  var wrappedChildren = (0, import_text.wrapChildrenInText)(Text, {
    children
  }, {
    unstyled: process.env.TAMAGUI_HEADLESS === "1",
    size: finalSize !== null && finalSize !== void 0 ? finalSize : styledContext === null || styledContext === void 0 ? void 0 : styledContext.size
  });
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_stacks.ButtonNestingContext.Provider, {
    value: true,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsxs)(Frame, {
      ref,
      ...props,
      ...(isNested && {
        render: "span"
      }),
      // Pass resolved size to circular variant when no explicit size provided
      ...(props.circular && !propsIn.size && {
        size
      }),
      tabIndex: 0,
      children: [themedIcon, wrappedChildren, themedIconAfter]
    })
  });
});
var Button = (0, import_web.withStaticProperties)(ButtonComponent, {
  Apply: context.Provider,
  Frame,
  Text,
  Icon
});
