// @tamagui/toggle-group/dist/esm/ToggleGroup.mjs
import { isWeb } from "@tamagui/constants";
import { registerFocusable } from "@tamagui/focusable";
import { withStaticProperties } from "@tamagui/helpers";
import { RovingFocusGroup } from "@tamagui/roving-focus";
import { useControllableState } from "@tamagui/use-controllable-state";
import { useDirection } from "@tamagui/use-direction";
import { createStyledContext, styled, View } from "@tamagui/web";
import React from "react";
import { Toggle, ToggleFrame } from "./Toggle.mjs";
import { context as ToggleContext } from "./context.mjs";
import { jsx } from "react/jsx-runtime";
const TOGGLE_GROUP_NAME = "ToggleGroup";
const TOGGLE_GROUP_ITEM_NAME = "ToggleGroupItem";
const TOGGLE_GROUP_CONTEXT = "ToggleGroup";
const {
  Provider: ToggleGroupItemProvider
} = createStyledContext();
const {
  Provider: ToggleGroupContext,
  useStyledContext: useToggleGroupContext
} = createStyledContext({});
const ToggleGroupItem = ToggleFrame.styleable((props, forwardedRef) => {
  const valueContext = useToggleGroupValueContext(props.__scopeToggleGroup);
  const context = useToggleGroupContext(props.__scopeToggleGroup);
  const toggleContext = ToggleContext.useStyledContext(props.__scopeToggleGroup);
  const active = valueContext?.value.includes(props.value);
  const color = props.color || toggleContext.color;
  const disabled = context.disabled || props.disabled || false;
  const inner = /* @__PURE__ */jsx(ToggleGroupItemImpl, {
    ref: forwardedRef,
    tabIndex: disabled ? -1 : 0,
    ...props,
    active,
    disabled
  });
  return /* @__PURE__ */jsx(ToggleGroupItemProvider, {
    scope: props.__scopeToggleGroup,
    children: /* @__PURE__ */jsx(ToggleContext.Provider, {
      color,
      active,
      children: context.rovingFocus ? /* @__PURE__ */jsx(RovingFocusGroup.Item, {
        asChild: "except-style",
        __scopeRovingFocusGroup: props.__scopeToggleGroup || TOGGLE_GROUP_CONTEXT,
        focusable: !disabled,
        active,
        children: inner
      }) : inner
    })
  });
});
ToggleGroupItem.displayName = TOGGLE_GROUP_ITEM_NAME;
const ToggleGroupItemImpl = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeToggleGroup,
    value,
    ...itemProps
  } = props;
  const valueContext = useToggleGroupValueContext(__scopeToggleGroup);
  const singleProps = {
    "aria-pressed": void 0
  };
  const typeProps = valueContext.type === "single" ? singleProps : void 0;
  return /* @__PURE__ */jsx(Toggle, {
    ...typeProps,
    ...itemProps,
    ref: forwardedRef,
    onActiveChange: pressed => {
      if (pressed) {
        valueContext.onItemActivate(value);
      } else {
        valueContext.onItemDeactivate(value);
      }
    }
  });
});
const ToggleGroup = withStaticProperties(React.forwardRef((props, forwardedRef) => {
  const {
    type,
    ...toggleGroupProps
  } = props;
  if (!isWeb) {
    React.useEffect(() => {
      if (!props.id) return;
      return registerFocusable(props.id, {
        focus: () => {}
      });
    }, [props.id]);
  }
  if (type === "single") {
    const singleProps = toggleGroupProps;
    return /* @__PURE__ */jsx(ToggleGroupImplSingle, {
      ...singleProps,
      ref: forwardedRef
    });
  }
  if (type === "multiple") {
    const multipleProps = toggleGroupProps;
    return /* @__PURE__ */jsx(ToggleGroupImplMultiple, {
      ...multipleProps,
      ref: forwardedRef
    });
  }
  throw new Error(`Missing prop \`type\` expected on \`${TOGGLE_GROUP_NAME}\``);
}), {
  Item: ToggleGroupItem
});
ToggleGroup.displayName = TOGGLE_GROUP_NAME;
const {
  Provider: ToggleGroupValueProvider,
  useStyledContext: useToggleGroupValueContext
} = createStyledContext();
const ToggleGroupImplSingle = React.forwardRef((props, forwardedRef) => {
  const {
    value: valueProp,
    defaultValue,
    onValueChange = () => {},
    disableDeactivation = false,
    children,
    ...toggleGroupSingleProps
  } = props;
  const [value, setValue] = useControllableState({
    prop: valueProp,
    defaultProp: defaultValue,
    onChange: onValueChange
  });
  return /* @__PURE__ */jsx(ToggleGroupValueProvider, {
    scope: props.__scopeToggleGroup,
    type: "single",
    value: value ? [value] : [],
    defaultValue: value,
    onItemActivate: setValue,
    onItemDeactivate: React.useCallback(() => disableDeactivation ? null : setValue(""), [setValue, disableDeactivation]),
    children: /* @__PURE__ */jsx(ToggleGroupImpl, {
      ...toggleGroupSingleProps,
      ref: forwardedRef,
      children
    })
  });
});
const ToggleGroupImplMultiple = React.forwardRef((props, forwardedRef) => {
  const {
    value: valueProp,
    defaultValue,
    onValueChange = () => {},
    disableDeactivation,
    children,
    ...toggleGroupMultipleProps
  } = props;
  const [value = [], setValue] = useControllableState({
    prop: valueProp,
    defaultProp: defaultValue,
    onChange: onValueChange
  });
  const handleButtonActivate = React.useCallback(itemValue => setValue((prevValue = []) => [...prevValue, itemValue]), [setValue]);
  const handleButtonDeactivate = React.useCallback(itemValue => setValue((prevValue = []) => prevValue.filter(value2 => value2 !== itemValue)), [setValue]);
  return /* @__PURE__ */jsx(ToggleGroupValueProvider, {
    scope: props.__scopeToggleGroup,
    type: "multiple",
    value,
    defaultValue: value,
    onItemActivate: handleButtonActivate,
    onItemDeactivate: handleButtonDeactivate,
    children: /* @__PURE__ */jsx(ToggleGroupImpl, {
      ...toggleGroupMultipleProps,
      ref: forwardedRef,
      children
    })
  });
});
const ToggleGroupFrame = styled(View, {
  name: TOGGLE_GROUP_NAME
});
const ToggleGroupImpl = ToggleGroupFrame.styleable((props, forwardedRef) => {
  const {
    __scopeToggleGroup,
    disabled = false,
    orientation = "horizontal",
    dir,
    rovingFocus = true,
    loop = true,
    color,
    ...toggleGroupProps
  } = props;
  const direction = useDirection(dir);
  const content = /* @__PURE__ */jsx(ToggleGroupFrame, {
    role: "group",
    ref: forwardedRef,
    "data-disabled": disabled ? "" : void 0,
    ...toggleGroupProps
  });
  return /* @__PURE__ */jsx(ToggleGroupContext, {
    scope: __scopeToggleGroup,
    rovingFocus,
    disabled,
    children: /* @__PURE__ */jsx(ToggleContext.Provider, {
      color,
      children: rovingFocus ? /* @__PURE__ */jsx(RovingFocusGroup, {
        asChild: "except-style",
        __scopeRovingFocusGroup: __scopeToggleGroup || TOGGLE_GROUP_CONTEXT,
        orientation,
        dir: direction,
        loop,
        children: content
      }) : content
    })
  });
});
export { ToggleGroup };
