// @tamagui/toggle-group/dist/esm/Toggle.mjs
import { composeEventHandlers } from "@tamagui/helpers";
import { useControllableState } from "@tamagui/use-controllable-state";
import { styled, View } from "@tamagui/web";
import * as React from "react";
import { context } from "./context.mjs";
import { jsx } from "react/jsx-runtime";
const NAME = "Toggle";
const ToggleFrame = styled(View, {
  name: NAME,
  render: "button",
  context,
  variants: {
    unstyled: {
      false: {
        size: "$true",
        alignItems: "center",
        justifyContent: "center",
        display: "flex",
        backgroundColor: "$background",
        borderColor: "$borderColor",
        borderWidth: 1,
        margin: -1,
        hoverStyle: {
          backgroundColor: "$backgroundHover",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "$backgroundPress",
          borderColor: "$borderColorPress"
        },
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineWidth: 2,
          outlineStyle: "solid",
          zIndex: 10
        }
      }
    },
    size: {
      "...size": (val, {
        tokens
      }) => {
        if (!val) return;
        return {
          width: tokens.size[val],
          height: tokens.size[val]
        };
      },
      ":number": val => ({
        width: val,
        height: val
      })
    },
    defaultActiveStyle: {
      true: {
        backgroundColor: "$backgroundActive",
        hoverStyle: {
          backgroundColor: "$backgroundActive"
        },
        focusStyle: {
          backgroundColor: "$backgroundActive"
        }
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
}, {
  accept: {
    activeStyle: "style"
  }
});
const Toggle = React.forwardRef(function Toggle2(props, forwardedRef) {
  const {
    active: activeProp,
    activeStyle,
    defaultActive = false,
    onActiveChange,
    activeTheme,
    unstyled = false,
    ...buttonProps
  } = props;
  const [active = false, setActive] = useControllableState({
    prop: activeProp,
    onChange: onActiveChange,
    defaultProp: defaultActive
  });
  return /* @__PURE__ */jsx(ToggleFrame, {
    theme: activeTheme ?? null,
    "aria-pressed": active,
    "data-state": active ? "on" : "off",
    "data-disabled": props.disabled ? "" : void 0,
    unstyled,
    ...(active && !activeStyle && !unstyled && {
      defaultActiveStyle: true
    }),
    ...(active && activeStyle && {
      ...activeStyle,
      hoverStyle: activeStyle,
      focusStyle: activeStyle
    }),
    ...buttonProps,
    ref: forwardedRef,
    onPress: composeEventHandlers(props.onPress ?? void 0, () => {
      if (!props.disabled) {
        setActive(prev => !prev);
      }
    })
  });
});
export { Toggle, ToggleFrame };
