// @tamagui/remove-scroll/dist/cjs/useDisableScroll.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useDisableScroll_exports = {};
__export(useDisableScroll_exports, {
  useDisableBodyScroll: () => useDisableBodyScroll
});
module.exports = __toCommonJS(useDisableScroll_exports);
var import_react = require("react");
var canUseDOM = function () {
  return typeof window !== "undefined" && !!window.document && !!window.document.createElement;
};
var refCount = 0;
var savedScrollY = 0;
var previousStyles = null;
function isIOSSafari() {
  if (typeof navigator === "undefined") return false;
  var ua = navigator.userAgent;
  var isIOS = /iPad|iPhone|iPod/.test(ua) || navigator.platform === "MacIntel" && navigator.maxTouchPoints > 1;
  var isSafari = /^((?!chrome|android).)*safari/i.test(ua);
  return isIOS && isSafari;
}
var useDisableBodyScroll = function (enabled) {
  var wasEnabled = (0, import_react.useRef)(false);
  (0, import_react.useEffect)(function () {
    if (!canUseDOM()) {
      return;
    }
    if (enabled && !wasEnabled.current) {
      wasEnabled.current = true;
      if (++refCount === 1) {
        var html = document.documentElement;
        var body = document.body;
        savedScrollY = window.scrollY;
        previousStyles = {
          htmlOverflow: html.style.overflow,
          htmlScrollbarGutter: html.style.scrollbarGutter,
          bodyPosition: body.style.position,
          bodyTop: body.style.top,
          bodyWidth: body.style.width,
          bodyOverflow: body.style.overflow,
          bodyOverscrollBehavior: body.style.overscrollBehavior
        };
        html.style.scrollbarGutter = "stable";
        html.style.overflow = "hidden";
        body.style.overscrollBehavior = "none";
        if (isIOSSafari()) {
          body.style.position = "fixed";
          body.style.top = `-${savedScrollY}px`;
          body.style.width = "100%";
          body.style.overflow = "hidden";
        }
      }
    } else if (!enabled && wasEnabled.current) {
      wasEnabled.current = false;
      if (--refCount === 0 && previousStyles) {
        var html1 = document.documentElement;
        var body1 = document.body;
        html1.style.overflow = previousStyles.htmlOverflow;
        html1.style.scrollbarGutter = previousStyles.htmlScrollbarGutter;
        body1.style.position = previousStyles.bodyPosition;
        body1.style.top = previousStyles.bodyTop;
        body1.style.width = previousStyles.bodyWidth;
        body1.style.overflow = previousStyles.bodyOverflow;
        body1.style.overscrollBehavior = previousStyles.bodyOverscrollBehavior;
        if (savedScrollY > 0) {
          window.scrollTo(0, savedScrollY);
        }
        previousStyles = null;
        savedScrollY = 0;
      }
    }
  }, [enabled]);
  (0, import_react.useEffect)(function () {
    return function () {
      if (wasEnabled.current) {
        wasEnabled.current = false;
        if (--refCount === 0 && previousStyles) {
          var html = document.documentElement;
          var body = document.body;
          html.style.overflow = previousStyles.htmlOverflow;
          html.style.scrollbarGutter = previousStyles.htmlScrollbarGutter;
          body.style.position = previousStyles.bodyPosition;
          body.style.top = previousStyles.bodyTop;
          body.style.width = previousStyles.bodyWidth;
          body.style.overflow = previousStyles.bodyOverflow;
          body.style.overscrollBehavior = previousStyles.bodyOverscrollBehavior;
          if (savedScrollY > 0) {
            window.scrollTo(0, savedScrollY);
          }
          previousStyles = null;
          savedScrollY = 0;
        }
      }
    };
  }, []);
};
