// @tamagui/image/dist/jsx/v1/Image.mjs
import React from "react";
import { isWeb } from "@tamagui/constants";
import { styled, usePropsAndStyle } from "@tamagui/web";
import { Image as RNImage } from "react-native-web";
import { jsx } from "react/jsx-runtime";
const StyledImage = styled(RNImage, {
  name: "Image"
});
let hasWarned = false;
const Image = StyledImage.styleable((inProps, ref) => {
  const [props, style] = usePropsAndStyle(inProps);
  const {
    src,
    source,
    objectFit,
    ...rest
  } = props;
  if (process.env.NODE_ENV === "development") {
    if (typeof src === "string") {
      if (typeof props.width === "string" && props.width[0] !== "$" || typeof props.height === "string" && props.height[0] !== "$") {
        if (!hasWarned) {
          hasWarned = true;
          console.warn(`React Native expects a numerical width/height. If you want to use a percent you must define the "source" prop with width, height, and uri.`);
        }
      }
    }
  }
  let finalSource = typeof src === "string" ? {
    uri: src,
    ...(isWeb && {
      width: props.width || style?.width,
      height: props.height || style?.height
    })
  } : source ?? src;
  if (finalSource && typeof finalSource === "object") {
    if (process.env.NODE_ENV === "development" && process.env.TAMAGUI_IMAGE_CHECK_ERROR) React.useEffect(() => {
      async function run() {
        if (typeof src === "string") try {
          await fetch(src).then(res => res.text());
        } catch (err) {
          console.error(`Error loading image: ${src}`, {
            props
          });
        }
      }
      run();
    }, [src]);
    if (finalSource["default"]) finalSource = finalSource["default"];
  }
  return /* @__PURE__ */jsx(RNImage, {
    resizeMode: objectFit,
    ref,
    source: finalSource,
    style,
    ...rest
  });
});
Image.getSize = RNImage.getSize;
Image.getSizeWithHeaders = RNImage.getSizeWithHeaders;
Image.prefetch = RNImage.prefetch;
Image.prefetchWithMetadata = RNImage.prefetchWithMetadata;
Image.abortPrefetch = RNImage.abortPrefetch;
Image.queryCache = RNImage.queryCache;
export { Image };
