// @tamagui/input/dist/jsx/shared.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var shared_exports = {};
__export(shared_exports, {
  INPUT_NAME: () => INPUT_NAME,
  defaultStyles: () => defaultStyles,
  inputSizeVariant: () => inputSizeVariant,
  styledBody: () => styledBody,
  textAreaSizeVariant: () => textAreaSizeVariant
});
module.exports = __toCommonJS(shared_exports);
var import_core = require("@tamagui/core");
var import_core2 = require("@tamagui/core");
var import_get_button_sized = require("@tamagui/get-button-sized");
var import_get_font_sized = require("@tamagui/get-font-sized");
var import_get_token = require("@tamagui/get-token");
var defaultStyles = {
  size: "$true",
  fontFamily: "$body",
  borderWidth: 1,
  outlineWidth: 0,
  color: "$color",
  ...(import_core2.isWeb ? {
    tabIndex: 0
  } : {
    focusable: true
  }),
  borderColor: "$borderColor",
  backgroundColor: "$background",
  // this fixes a flex bug where it overflows container
  minWidth: 0,
  hoverStyle: {
    borderColor: "$borderColorHover"
  },
  focusStyle: {
    borderColor: "$borderColorFocus"
  },
  focusVisibleStyle: {
    outlineColor: "$outlineColor",
    outlineWidth: 2,
    outlineStyle: "solid"
  }
};
var inputSizeVariant = function () {
  var val = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "$true",
    extras = arguments.length > 1 ? arguments[1] : void 0;
  if (extras.props.tag === "textarea" || extras.props.rows > 1 || extras.props.multiline || extras.props.numberOfLines > 1) {
    return textAreaSizeVariant(val, extras);
  }
  var buttonStyles = (0, import_get_button_sized.getButtonSized)(val, extras);
  var paddingHorizontal = (0, import_get_token.getSpace)(val, {
    shift: -1,
    bounds: [2]
  });
  var fontStyle = (0, import_get_font_sized.getFontSized)(val, extras);
  if (!import_core2.isWeb && fontStyle) {
    delete fontStyle["lineHeight"];
  }
  return {
    ...fontStyle,
    ...buttonStyles,
    paddingHorizontal
  };
};
var textAreaSizeVariant = function () {
  var val = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "$true",
    extras = arguments.length > 1 ? arguments[1] : void 0;
  var {
    props
  } = extras;
  var buttonStyles = (0, import_get_button_sized.getButtonSized)(val, extras);
  var fontStyle = (0, import_get_font_sized.getFontSized)(val, extras);
  var _props_rows;
  var lines = (_props_rows = props.rows) !== null && _props_rows !== void 0 ? _props_rows : props.numberOfLines;
  var height = typeof lines === "number" ? lines * (0, import_core2.getVariableValue)(fontStyle.lineHeight) : "auto";
  if (!import_core2.isWeb && fontStyle) {
    delete fontStyle["lineHeight"];
  }
  var paddingVertical = (0, import_get_token.getSpace)(val, {
    shift: -2,
    bounds: [2]
  });
  var paddingHorizontal = (0, import_get_token.getSpace)(val, {
    shift: -1,
    bounds: [2]
  });
  return {
    ...buttonStyles,
    ...fontStyle,
    paddingVertical,
    paddingHorizontal,
    height
  };
};
var INPUT_NAME = "Input";
var styledBody = [{
  name: INPUT_NAME,
  render: "input",
  variants: {
    unstyled: {
      false: defaultStyles
    },
    size: {
      "...size": inputSizeVariant
    },
    disabled: {
      true: {}
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
}, {
  isInput: true,
  accept: {
    placeholderTextColor: "color",
    selectionColor: "color",
    cursorColor: "color",
    selectionHandleColor: "color",
    underlineColorAndroid: "color"
  },
  validStyles: import_core.Text.staticConfig.validStyles
}];
