// @tamagui/core/dist/cjs/index.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  LayoutMeasurementController: () => import_use_element_layout2.LayoutMeasurementController,
  TamaguiProvider: () => TamaguiProvider,
  Text: () => Text,
  View: () => View,
  createTamagui: () => createTamagui,
  registerLayoutNode: () => import_use_element_layout2.registerLayoutNode,
  setOnLayoutStrategy: () => import_use_element_layout2.setOnLayoutStrategy
});
module.exports = __toCommonJS(index_exports);
var import_jsx_runtime = require("react/jsx-runtime");
__reExport(index_exports, require("@tamagui/web"), module.exports);
var import_react_native_media_driver = require("@tamagui/react-native-media-driver");
var import_constants = require("@tamagui/constants");
var import_use_element_layout = require("@tamagui/use-element-layout");
var import_web = require("@tamagui/web");
var import_createOptimizedView = require("./createOptimizedView.native.js");
var import_getBaseViews = require("./getBaseViews.native.js");
var import_use_element_layout2 = require("@tamagui/use-element-layout");
__reExport(index_exports, require("./reactNativeTypes.native.js"), module.exports);
var TamaguiProvider = function (props) {
  (0, import_web.useIsomorphicLayoutEffect)(function () {
    (0, import_use_element_layout.enable)();
  }, []);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_web.TamaguiProvider, {
    ...props
  });
};
var createTamagui = function (conf) {
  if (!import_constants.isWeb) {
    if (conf.media) conf.media = (0, import_react_native_media_driver.createMedia)(conf.media);
  }
  return (0, import_web.createTamagui)(conf);
};
var baseViews = (0, import_getBaseViews.getBaseViews)();
(0, import_web.setupHooks)({
  getBaseViews: import_getBaseViews.getBaseViews,
  setElementProps: function (node) {},
  usePropsTransform(elementType, propsIn, stateRef, willHydrate) {},
  useChildren(elementType, children, viewProps) {
    if (process.env.NODE_ENV === "test") return;
    if (elementType === baseViews.View && baseViews.TextAncestor) return (0, import_createOptimizedView.createOptimizedView)(children, viewProps, baseViews);
  }
});
var View = import_web.View;
var Text = import_web.Text;
