// @tamagui/core/dist/cjs/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  LayoutMeasurementController: () => import_use_element_layout2.LayoutMeasurementController,
  TamaguiProvider: () => TamaguiProvider,
  Text: () => Text,
  View: () => View,
  createTamagui: () => createTamagui,
  registerLayoutNode: () => import_use_element_layout2.registerLayoutNode,
  setOnLayoutStrategy: () => import_use_element_layout2.setOnLayoutStrategy
});
module.exports = __toCommonJS(index_exports);
__reExport(index_exports, require("@tamagui/web"), module.exports);
var import_react_native_media_driver = require("@tamagui/react-native-media-driver");
var import_constants = require("@tamagui/constants");
var import_use_element_layout = require("@tamagui/use-element-layout");
var import_web = require("@tamagui/web");
require("./createOptimizedView.cjs");
var import_getBaseViews = require("./getBaseViews.cjs");
var import_use_element_layout2 = require("@tamagui/use-element-layout");
__reExport(index_exports, require("./reactNativeTypes.cjs"), module.exports);
var import_jsx_runtime = require("react/jsx-runtime");
const TamaguiProvider = props => {
  (0, import_web.useIsomorphicLayoutEffect)(() => {
    (0, import_use_element_layout.enable)();
  }, []);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_web.TamaguiProvider, {
    ...props
  });
};
const createTamagui = conf => {
  if (!import_constants.isWeb) {
    if (conf.media) conf.media = (0, import_react_native_media_driver.createMedia)(conf.media);
  }
  return (0, import_web.createTamagui)(conf);
};
(0, import_getBaseViews.getBaseViews)();
(0, import_web.setupHooks)({
  getBaseViews: import_getBaseViews.getBaseViews,
  setElementProps: node => {
    if (node && !node["measure"]) {
      node.measure ||= (0, import_use_element_layout.createMeasure)(node);
      node.measureInWindow ||= (0, import_use_element_layout.createMeasureInWindow)(node);
      node.measureLayout ||= (0, import_use_element_layout.createMeasureLayout)(node);
    }
  },
  usePropsTransform(elementType, propsIn, stateRef, willHydrate) {
    {
      const isDOM = typeof elementType === "string";
      const {
        onMoveShouldSetResponder,
        onMoveShouldSetResponderCapture,
        onResponderEnd,
        onResponderGrant,
        onResponderMove,
        onResponderReject,
        onResponderRelease,
        onResponderStart,
        onResponderTerminate,
        onResponderTerminationRequest,
        onScrollShouldSetResponder,
        onScrollShouldSetResponderCapture,
        onSelectionChangeShouldSetResponder,
        onSelectionChangeShouldSetResponderCapture,
        onStartShouldSetResponder,
        onStartShouldSetResponderCapture,
        collapsable,
        focusable,
        accessible,
        accessibilityDisabled,
        onLayout,
        hrefAttrs,
        ...plainDOMProps
      } = propsIn;
      if (willHydrate || isDOM) (0, import_use_element_layout.useElementLayout)(stateRef, !isDOM ? void 0 : onLayout);
      if (isDOM) {
        if (plainDOMProps.href && hrefAttrs) {
          const {
            download,
            rel,
            target
          } = hrefAttrs;
          if (download != null) plainDOMProps.download = download;
          if (rel) plainDOMProps.rel = rel;
          if (typeof target === "string") plainDOMProps.target = target.charAt(0) !== "_" ? `_${target}` : target;
        }
        return plainDOMProps;
      }
    }
  }
});
const View = import_web.View;
const Text = import_web.Text;
