// @tamagui/core/dist/esm/index.native.js
import { jsx as _jsx } from "react/jsx-runtime";
export * from "@tamagui/web";
import { createMedia } from "@tamagui/react-native-media-driver";
import { isWeb } from "@tamagui/constants";
import { createMeasure, createMeasureInWindow, createMeasureLayout, enable, useElementLayout } from "@tamagui/use-element-layout";
import { TamaguiProvider as WebTamaguiProvider, Text as WebText, View as WebView, createTamagui as createTamaguiWeb, setupHooks, useIsomorphicLayoutEffect } from "@tamagui/web";
import { createOptimizedView } from "./createOptimizedView.native.js";
import { getBaseViews } from "./getBaseViews.native.js";
import { LayoutMeasurementController, registerLayoutNode, setOnLayoutStrategy } from "@tamagui/use-element-layout";
export * from "./reactNativeTypes.native.js";
var TamaguiProvider = function (props) {
  useIsomorphicLayoutEffect(function () {
    enable();
  }, []);
  return /* @__PURE__ */_jsx(WebTamaguiProvider, {
    ...props
  });
};
var createTamagui = function (conf) {
  if (!isWeb) {
    if (conf.media) conf.media = createMedia(conf.media);
  }
  return createTamaguiWeb(conf);
};
var baseViews = getBaseViews();
setupHooks({
  getBaseViews,
  setElementProps: function (node) {},
  usePropsTransform(elementType, propsIn, stateRef, willHydrate) {},
  useChildren(elementType, children, viewProps) {
    if (process.env.NODE_ENV === "test") return;
    if (elementType === baseViews.View && baseViews.TextAncestor) return createOptimizedView(children, viewProps, baseViews);
  }
});
var View = WebView;
var Text = WebText;
export { LayoutMeasurementController, TamaguiProvider, Text, View, createTamagui, registerLayoutNode, setOnLayoutStrategy };
