// @tamagui/core/dist/esm/index.js
export * from "@tamagui/web";
import { createMedia } from "@tamagui/react-native-media-driver";
import { isWeb } from "@tamagui/constants";
import { createMeasure, createMeasureInWindow, createMeasureLayout, enable, useElementLayout } from "@tamagui/use-element-layout";
import { TamaguiProvider as WebTamaguiProvider, Text as WebText, View as WebView, createTamagui as createTamaguiWeb, setupHooks, useIsomorphicLayoutEffect } from "@tamagui/web";
import { createOptimizedView } from "./createOptimizedView.mjs";
import { getBaseViews } from "./getBaseViews.mjs";
import { LayoutMeasurementController, registerLayoutNode, setOnLayoutStrategy } from "@tamagui/use-element-layout";
export * from "./reactNativeTypes.mjs";
import { jsx } from "react/jsx-runtime";
const TamaguiProvider = props => {
  useIsomorphicLayoutEffect(() => {
    enable();
  }, []);
  return /* @__PURE__ */jsx(WebTamaguiProvider, {
    ...props
  });
};
const createTamagui = conf => {
  if (!isWeb) {
    if (conf.media) conf.media = createMedia(conf.media);
  }
  return createTamaguiWeb(conf);
};
getBaseViews();
setupHooks({
  getBaseViews,
  setElementProps: node => {
    if (node && !node["measure"]) {
      node.measure ||= createMeasure(node);
      node.measureInWindow ||= createMeasureInWindow(node);
      node.measureLayout ||= createMeasureLayout(node);
    }
  },
  usePropsTransform(elementType, propsIn, stateRef, willHydrate) {
    {
      const isDOM = typeof elementType === "string";
      const {
        onMoveShouldSetResponder,
        onMoveShouldSetResponderCapture,
        onResponderEnd,
        onResponderGrant,
        onResponderMove,
        onResponderReject,
        onResponderRelease,
        onResponderStart,
        onResponderTerminate,
        onResponderTerminationRequest,
        onScrollShouldSetResponder,
        onScrollShouldSetResponderCapture,
        onSelectionChangeShouldSetResponder,
        onSelectionChangeShouldSetResponderCapture,
        onStartShouldSetResponder,
        onStartShouldSetResponderCapture,
        collapsable,
        focusable,
        accessible,
        accessibilityDisabled,
        onLayout,
        hrefAttrs,
        ...plainDOMProps
      } = propsIn;
      if (willHydrate || isDOM) useElementLayout(stateRef, !isDOM ? void 0 : onLayout);
      if (isDOM) {
        if (plainDOMProps.href && hrefAttrs) {
          const {
            download,
            rel,
            target
          } = hrefAttrs;
          if (download != null) plainDOMProps.download = download;
          if (rel) plainDOMProps.rel = rel;
          if (typeof target === "string") plainDOMProps.target = target.charAt(0) !== "_" ? `_${target}` : target;
        }
        return plainDOMProps;
      }
    }
  }
});
const View = WebView;
const Text = WebText;
export { LayoutMeasurementController, TamaguiProvider, Text, View, createTamagui, registerLayoutNode, setOnLayoutStrategy };
