// @tamagui/dialog/dist/esm/Dialog.native.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { Adapt, AdaptParent, AdaptPortalContents, ProvideAdaptContext, useAdaptContext, useAdaptIsActive } from "@tamagui/adapt";
import { Animate } from "@tamagui/animate";
import { composeRefs, useComposedRefs } from "@tamagui/compose-refs";
import { isWeb, useIsomorphicLayoutEffect } from "@tamagui/constants";
import { createStyledContext, getExpandedShorthand, LayoutMeasurementController, styled, Theme, useThemeName, View } from "@tamagui/core";
import { createContext } from "@tamagui/create-context";
import { Dismissable } from "@tamagui/dismissable";
import { FocusScope, FocusScopeController } from "@tamagui/focus-scope";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { needsPortalRepropagation, Portal, PortalItem, resolveViewZIndex } from "@tamagui/portal";
import { RemoveScroll } from "@tamagui/remove-scroll";
import { SheetController } from "@tamagui/sheet/controller";
import { ButtonNestingContext, ThemeableStack, YStack } from "@tamagui/stacks";
import { H2, Paragraph } from "@tamagui/text";
import { useControllableState } from "@tamagui/use-controllable-state";
import { StackZIndexContext } from "@tamagui/z-index-stack";
import * as React from "react";
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
    return !!right[Symbol.hasInstance](left);
  } else {
    return left instanceof right;
  }
}
var DialogContext = createStyledContext(
// since we always provide this we can avoid setting here
{}, "Dialog__");
var {
  useStyledContext: useDialogContext,
  Provider: DialogProvider
} = DialogContext;
var DialogTriggerFrame = styled(View, {
  name: "DialogTrigger"
});
var DialogTrigger = DialogTriggerFrame.styleable(function DialogTrigger2(props, forwardedRef) {
  var {
    scope,
    ...triggerProps
  } = props;
  var isInsideButton = React.useContext(ButtonNestingContext);
  var context = useDialogContext(scope);
  var composedTriggerRef = useComposedRefs(forwardedRef, context.triggerRef);
  return /* @__PURE__ */_jsx(ButtonNestingContext.Provider, {
    value: true,
    children: /* @__PURE__ */_jsx(DialogTriggerFrame, {
      render: isInsideButton ? "span" : "button",
      "aria-haspopup": "dialog",
      "aria-expanded": context.open,
      "aria-controls": context.contentId,
      "data-state": getState(context.open),
      ...triggerProps,
      ref: composedTriggerRef,
      onPress: composeEventHandlers(props.onPress, context.onOpenToggle)
    })
  });
});
var DialogPortalFrame = styled(YStack, {
  pointerEvents: "none",
  render: "dialog",
  variants: {
    unstyled: {
      false: {
        alignItems: "center",
        justifyContent: "center",
        fullscreen: true,
        "$platform-web": {
          // undo dialog styles
          borderWidth: 0,
          backgroundColor: "transparent",
          color: "inherit",
          maxInlineSize: "none",
          margin: 0,
          width: "auto",
          height: "auto",
          // ensure always in frame and right height
          maxHeight: "100vh",
          position: "fixed",
          // ensure dialog inherits stacking context from portal wrapper
          zIndex: 1
        }
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var needsRepropagation = needsPortalRepropagation();
var DialogPortalItem = function (param) {
  var {
    context,
    children
  } = param;
  var themeName = useThemeName();
  var isAdapted = useAdaptIsActive(context.adaptScope);
  var adaptContext = useAdaptContext(context.adaptScope);
  var content = /* @__PURE__ */_jsx(Theme, {
    name: themeName,
    children
  });
  if (needsRepropagation) {
    content = /* @__PURE__ */_jsx(ProvideAdaptContext, {
      ...adaptContext,
      children: /* @__PURE__ */_jsx(DialogProvider, {
        ...context,
        children: content
      })
    });
  }
  return isAdapted ? /* @__PURE__ */_jsx(AdaptPortalContents, {
    scope: context.adaptScope,
    children: content
  }) : context.modal ? /* @__PURE__ */_jsx(PortalItem, {
    hostName: context.modal ? "root" : context.adaptScope,
    children: content
  }) : content;
};
var DialogPortal = /* @__PURE__ */React.forwardRef(function (props, forwardRef) {
  var {
    scope,
    forceMount,
    children,
    ...frameProps
  } = props;
  var dialogRef = React.useRef(null);
  var ref = composeRefs(dialogRef, forwardRef);
  var context = useDialogContext(scope);
  var keepMounted = forceMount || context.keepChildrenMounted;
  var isAdapted = useAdaptIsActive(context.adaptScope);
  var [isFullyHidden, setIsFullyHidden] = React.useState(!context.open);
  if (context.open && isFullyHidden) {
    setIsFullyHidden(false);
  }
  var isVisible = context.open ? true : !isFullyHidden;
  if (isWeb) {
    useIsomorphicLayoutEffect(function () {
      var node = dialogRef.current;
      if (!_instanceof(node, HTMLDialogElement)) return;
      if (isVisible) {
        var _node_show;
        (_node_show = node.show) === null || _node_show === void 0 ? void 0 : _node_show.call(node);
      } else {
        var _node_close;
        (_node_close = node.close) === null || _node_close === void 0 ? void 0 : _node_close.call(node);
      }
    }, [isVisible]);
  }
  var onAnimationCompleteRef = React.useRef(context.onAnimationComplete);
  onAnimationCompleteRef.current = context.onAnimationComplete;
  var handleExitComplete = React.useCallback(function () {
    var _onAnimationCompleteRef_current;
    setIsFullyHidden(true);
    (_onAnimationCompleteRef_current = onAnimationCompleteRef.current) === null || _onAnimationCompleteRef_current === void 0 ? void 0 : _onAnimationCompleteRef_current.call(onAnimationCompleteRef, {
      open: false
    });
  }, []);
  React.useEffect(function () {
    if (context.open && !isAdapted && onAnimationCompleteRef.current) {
      var tm = setTimeout(function () {
        var _onAnimationCompleteRef_current;
        (_onAnimationCompleteRef_current = onAnimationCompleteRef.current) === null || _onAnimationCompleteRef_current === void 0 ? void 0 : _onAnimationCompleteRef_current.call(onAnimationCompleteRef, {
          open: true
        });
      }, 350);
      return function () {
        return clearTimeout(tm);
      };
    }
  }, [context.open, isAdapted]);
  var zIndex = getExpandedShorthand("zIndex", props);
  var contents = /* @__PURE__ */_jsx(StackZIndexContext, {
    zIndex: resolveViewZIndex(zIndex),
    children: /* @__PURE__ */_jsx(Animate, {
      type: "presence",
      present: Boolean(context.open),
      keepChildrenMounted: Boolean(keepMounted),
      onExitComplete: handleExitComplete,
      passThrough: isAdapted,
      children
    })
  });
  var framedContents = isFullyHidden && !keepMounted && !isAdapted ? null : /* @__PURE__ */_jsx(LayoutMeasurementController, {
    disable: !context.open,
    children: /* @__PURE__ */_jsx(DialogPortalFrame, {
      ref,
      ...(isWeb && context.open && {
        "aria-modal": true
      }),
      pointerEvents: context.open ? "auto" : "none",
      ...frameProps,
      className: `_no_backdrop ` + (frameProps.className || ""),
      children: contents
    })
  });
  if (isWeb) {
    return /* @__PURE__ */_jsx(Portal, {
      zIndex,
      // set to 100000 which ensures dialogs are above most fixed UI (headers, navs)
      // this makes sure its above typical stacking contexts
      stackZIndex: 1e5,
      passThrough: isAdapted,
      children: /* @__PURE__ */_jsx(PassthroughTheme, {
        passThrough: isAdapted,
        children: framedContents
      })
    });
  }
  return isAdapted ? framedContents : /* @__PURE__ */_jsx(DialogPortalItem, {
    context,
    children: framedContents
  });
});
var PassthroughTheme = function (param) {
  var {
    children,
    passThrough
  } = param;
  var themeName = useThemeName();
  return /* @__PURE__ */_jsx(Theme, {
    passThrough,
    name: themeName,
    forceClassName: true,
    children
  });
};
var OVERLAY_NAME = "DialogOverlay";
var DialogOverlayFrame = styled(YStack, {
  name: OVERLAY_NAME,
  zIndex: 1,
  variants: {
    open: {
      true: {
        pointerEvents: "auto"
      },
      false: {
        pointerEvents: "none"
      }
    },
    unstyled: {
      false: {
        fullscreen: true,
        position: "absolute",
        backgroundColor: "$background",
        pointerEvents: "auto"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var DialogOverlay = DialogOverlayFrame.styleable(function DialogOverlay2(param, forwardedRef) {
  var {
    scope,
    ...props
  } = param;
  var context = useDialogContext(scope);
  var {
    forceMount = context.forceMount,
    ...overlayProps
  } = props;
  var isAdapted = useAdaptIsActive(context.adaptScope);
  if (!forceMount) {
    if (!context.modal || isAdapted) {
      return null;
    }
  }
  return /* @__PURE__ */_jsx(DialogOverlayFrame, {
    "data-state": getState(context.open),
    // TODO: this will be apply for v2
    // onPress={() => {
    //   // if the overlay is pressed, close the dialog
    //   context.onOpenChange(false)
    // }}
    // We re-enable pointer-events prevented by `Dialog.Content` to allow scrolling the overlay.
    pointerEvents: context.open ? "auto" : "none",
    ...overlayProps,
    ref: forwardedRef
  });
});
var CONTENT_NAME = "DialogContent";
var DialogContentFrame = styled(ThemeableStack, {
  name: CONTENT_NAME,
  zIndex: 2,
  variants: {
    size: {
      "...size": function (val, extras) {
        return {};
      }
    },
    unstyled: {
      false: {
        position: "relative",
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "$borderColor",
        padding: "$true",
        borderRadius: "$true",
        elevate: true,
        // Ensure content receives pointer events (fixes React 19 + display:contents inheritance)
        pointerEvents: "auto"
      }
    }
  },
  defaultVariants: {
    size: "$true",
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var DialogContent = DialogContentFrame.styleable(function DialogContent2(param, forwardedRef) {
  var {
    scope,
    ...props
  } = param;
  var context = useDialogContext(scope);
  var contents = /* @__PURE__ */_jsx(_Fragment, {
    children: context.modal ? /* @__PURE__ */_jsx(DialogContentModal, {
      context,
      ...props,
      ref: forwardedRef
    }) : /* @__PURE__ */_jsx(DialogContentNonModal, {
      context,
      ...props,
      ref: forwardedRef
    })
  });
  if (!isWeb || context.disableRemoveScroll) {
    return contents;
  }
  return /* @__PURE__ */_jsx(RemoveScroll, {
    enabled: context.open,
    children: /* @__PURE__ */_jsx("div", {
      "data-remove-scroll-container": true,
      className: "_dsp_contents",
      children: contents
    })
  });
});
var DialogContentModal = /* @__PURE__ */React.forwardRef(function (param, forwardedRef) {
  var {
    children,
    context,
    ...props
  } = param;
  var contentRef = React.useRef(null);
  var composedRefs = useComposedRefs(forwardedRef, context.contentRef, contentRef);
  return /* @__PURE__ */_jsx(DialogContentImpl, {
    ...props,
    context,
    ref: composedRefs,
    // we make sure focus isn't trapped once `DialogContent` has been closed
    // (closed !== unmounted when animating out)
    trapFocus: context.open,
    disableOutsidePointerEvents: true,
    onCloseAutoFocus: composeEventHandlers(props.onCloseAutoFocus, function (event) {
      var _context_triggerRef_current;
      event.preventDefault();
      (_context_triggerRef_current = context.triggerRef.current) === null || _context_triggerRef_current === void 0 ? void 0 : _context_triggerRef_current.focus();
    }),
    onPointerDownOutside: composeEventHandlers(props.onPointerDownOutside, function (event) {
      var originalEvent = event["detail"].originalEvent;
      var ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
      var isRightClick = originalEvent.button === 2 || ctrlLeftClick;
      if (isRightClick) event.preventDefault();
    }),
    // When focus is trapped, a `focusout` event may still happen.
    // We make sure we don't trigger our `onDismiss` in such case.
    onFocusOutside: composeEventHandlers(props.onFocusOutside, function (event) {
      return event.preventDefault();
    }),
    ...(!props.unstyled && {
      outlineStyle: "none"
    }),
    children
  });
});
var DialogContentNonModal = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var hasInteractedOutsideRef = React.useRef(false);
  return /* @__PURE__ */_jsx(DialogContentImpl, {
    ...props,
    ref: forwardedRef,
    trapFocus: false,
    disableOutsidePointerEvents: false,
    onCloseAutoFocus: function (event) {
      var _props_onCloseAutoFocus;
      (_props_onCloseAutoFocus = props.onCloseAutoFocus) === null || _props_onCloseAutoFocus === void 0 ? void 0 : _props_onCloseAutoFocus.call(props, event);
      if (!event.defaultPrevented) {
        if (!hasInteractedOutsideRef.current) {
          var _props_context_triggerRef_current;
          (_props_context_triggerRef_current = props.context.triggerRef.current) === null || _props_context_triggerRef_current === void 0 ? void 0 : _props_context_triggerRef_current.focus();
        }
        event.preventDefault();
      }
      hasInteractedOutsideRef.current = false;
    },
    onInteractOutside: function (event) {
      var _props_onInteractOutside;
      (_props_onInteractOutside = props.onInteractOutside) === null || _props_onInteractOutside === void 0 ? void 0 : _props_onInteractOutside.call(props, event);
      if (!event.defaultPrevented) hasInteractedOutsideRef.current = true;
      var target = event.target;
      var trigger = props.context.triggerRef.current;
      if (!_instanceof(trigger, HTMLElement)) return;
      var targetIsTrigger = trigger.contains(target);
      if (targetIsTrigger) event.preventDefault();
    }
  });
});
var DialogContentImpl = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var {
    trapFocus,
    onOpenAutoFocus,
    onCloseAutoFocus,
    disableOutsidePointerEvents,
    onEscapeKeyDown,
    onPointerDownOutside,
    onFocusOutside,
    onInteractOutside,
    context,
    ...contentProps
  } = props;
  var contentRef = React.useRef(null);
  var composedRefs = useComposedRefs(forwardedRef, contentRef);
  var isAdapted = useAdaptIsActive(context.adaptScope);
  if (isAdapted) {
    if (!isWeb && !context.open) {
      return null;
    }
    return /* @__PURE__ */_jsx(DialogPortalItem, {
      context,
      children: contentProps.children
    });
  }
  var contents = /* @__PURE__ */_jsx(DialogContentFrame, {
    ref: composedRefs,
    id: context.contentId,
    role: "dialog",
    "aria-modal": context.modal,
    "aria-describedby": context.descriptionId,
    "aria-labelledby": context.titleId,
    "data-state": getState(context.open),
    // allow clicking through content during exit animation
    pointerEvents: context.open ? "auto" : "none",
    ...contentProps
  });
  if (!isWeb) {
    return contents;
  }
  return /* @__PURE__ */_jsxs(_Fragment, {
    children: [/* @__PURE__ */_jsx(Dismissable, {
      disableOutsidePointerEvents: context.open && disableOutsidePointerEvents,
      forceUnmount: !context.open,
      onEscapeKeyDown,
      onPointerDownOutside,
      onFocusOutside,
      onInteractOutside,
      onDismiss: function () {
        var _context_onOpenChange;
        return context === null || context === void 0 ? void 0 : (_context_onOpenChange = context.onOpenChange) === null || _context_onOpenChange === void 0 ? void 0 : _context_onOpenChange.call(context, false);
      },
      children: /* @__PURE__ */_jsx(FocusScope, {
        loop: true,
        enabled: context.open,
        trapped: trapFocus,
        onMountAutoFocus: onOpenAutoFocus,
        forceUnmount: !context.open,
        onUnmountAutoFocus: onCloseAutoFocus,
        children: contents
      })
    }), process.env.NODE_ENV === "development" && /* @__PURE__ */_jsxs(_Fragment, {
      children: [/* @__PURE__ */_jsx(TitleWarning, {
        titleId: context.titleId
      }), /* @__PURE__ */_jsx(DescriptionWarning, {
        contentRef,
        descriptionId: context.descriptionId
      })]
    })]
  });
});
var DialogTitleFrame = styled(H2, {
  name: "DialogTitle"
});
var DialogTitle = DialogTitleFrame.styleable(function DialogTitle2(props, forwardedRef) {
  var {
    scope,
    ...titleProps
  } = props;
  var context = useDialogContext(scope);
  return /* @__PURE__ */_jsx(DialogTitleFrame, {
    id: context.titleId,
    ...titleProps,
    ref: forwardedRef
  });
});
var DialogDescriptionFrame = styled(Paragraph, {
  name: "DialogDescription"
});
var DialogDescription = DialogDescriptionFrame.styleable(function DialogDescription2(props, forwardedRef) {
  var {
    scope,
    ...descriptionProps
  } = props;
  var context = useDialogContext(scope);
  return /* @__PURE__ */_jsx(DialogDescriptionFrame, {
    id: context.descriptionId,
    ...descriptionProps,
    ref: forwardedRef
  });
});
var CLOSE_NAME = "DialogClose";
var DialogCloseFrame = styled(View, {
  name: CLOSE_NAME,
  render: "button"
});
var DialogClose = DialogCloseFrame.styleable(function (props, forwardedRef) {
  var {
    scope,
    displayWhenAdapted,
    ...closeProps
  } = props;
  var context = useDialogContext(scope);
  var isAdapted = useAdaptIsActive(context.adaptScope);
  var isInsideButton = React.useContext(ButtonNestingContext);
  if (isAdapted && !displayWhenAdapted) {
    return null;
  }
  return /* @__PURE__ */_jsx(DialogCloseFrame, {
    "aria-label": "Dialog Close",
    render: isInsideButton ? "span" : "button",
    ...closeProps,
    ref: forwardedRef,
    onPress: composeEventHandlers(props.onPress, function () {
      context.onOpenChange(false);
    })
  });
});
function getState(open) {
  return open ? "open" : "closed";
}
var TITLE_WARNING_NAME = "DialogTitleWarning";
var [DialogWarningProvider, useWarningContext] = createContext(TITLE_WARNING_NAME, {
  contentName: CONTENT_NAME,
  titleName: "DialogTitle",
  docsSlug: "dialog"
});
var TitleWarning = function (param) {
  var {
    titleId
  } = param;
  if (process.env.NODE_ENV === "development") {
    var titleWarningContext = useWarningContext(TITLE_WARNING_NAME);
    var MESSAGE = `\`${titleWarningContext.contentName}\` wants a \`${titleWarningContext.titleName}\` to be accessible. If you want to hide the \`${titleWarningContext.titleName}\`, wrap it with <VisuallyHidden />.`;
    React.useEffect(function () {
      if (!isWeb) return;
      if (titleId) {
        var hasTitle = document.getElementById(titleId);
        if (!hasTitle) {
          console.warn(MESSAGE);
        }
      }
    }, [MESSAGE, titleId]);
  }
  return null;
};
var DESCRIPTION_WARNING_NAME = "DialogDescriptionWarning";
var DescriptionWarning = function (param) {
  var {
    contentRef,
    descriptionId
  } = param;
  if (process.env.NODE_ENV === "development") {
    var descriptionWarningContext = useWarningContext(DESCRIPTION_WARNING_NAME);
    var MESSAGE = `Warning: Missing \`Description\` or \`aria-describedby={undefined}\` for {${descriptionWarningContext.contentName}}.`;
    React.useEffect(function () {
      if (!isWeb) return;
      var contentNode = contentRef.current;
      if (!_instanceof(contentNode, HTMLElement)) {
        return;
      }
      var describedById = contentNode.getAttribute("aria-describedby");
      if (descriptionId && describedById) {
        var hasDescription = document.getElementById(descriptionId);
        if (!hasDescription) {
          console.warn(MESSAGE);
        }
      }
    }, [MESSAGE, contentRef, descriptionId]);
  }
  return null;
};
var Dialog = withStaticProperties(/* @__PURE__ */React.forwardRef(function Dialog2(props, ref) {
  var {
    scope = "",
    children,
    open: openProp,
    defaultOpen = false,
    onOpenChange,
    modal = true,
    keepChildrenMounted,
    disableRemoveScroll = false,
    onAnimationComplete
  } = props;
  var baseId = React.useId();
  var dialogId = `Dialog-${scope}-${baseId}`;
  var contentId = `${dialogId}-content`;
  var titleId = `${dialogId}-title`;
  var descriptionId = `${dialogId}-description`;
  var triggerRef = React.useRef(null);
  var contentRef = React.useRef(null);
  var [open, setOpen] = useControllableState({
    prop: openProp,
    defaultProp: defaultOpen,
    onChange: onOpenChange
  });
  var onOpenToggle = React.useCallback(function () {
    setOpen(function (prevOpen) {
      return !prevOpen;
    });
  }, [setOpen]);
  var adaptScope = `DialogAdapt${scope}`;
  var context = {
    dialogScope: scope,
    adaptScope,
    triggerRef,
    contentRef,
    contentId,
    titleId,
    descriptionId,
    open,
    onOpenChange: setOpen,
    onOpenToggle,
    modal,
    keepChildrenMounted,
    disableRemoveScroll,
    onAnimationComplete
  };
  React.useImperativeHandle(ref, function () {
    return {
      open: setOpen
    };
  }, [setOpen]);
  return /* @__PURE__ */_jsx(AdaptParent, {
    scope: adaptScope,
    portal: {
      forwardProps: props
    },
    children: /* @__PURE__ */_jsx(DialogProvider, {
      scope,
      ...context,
      children: /* @__PURE__ */_jsx(DialogSheetController, {
        onOpenChange: setOpen,
        scope,
        children
      })
    })
  });
}), {
  Trigger: DialogTrigger,
  Portal: DialogPortal,
  Overlay: DialogOverlay,
  Content: DialogContent,
  Title: DialogTitle,
  Description: DialogDescription,
  Close: DialogClose,
  FocusScope: FocusScopeController,
  Adapt
});
var DialogSheetController = function (props) {
  var context = useDialogContext(props.scope);
  var isAdapted = useAdaptIsActive(context.adaptScope);
  return /* @__PURE__ */_jsx(SheetController, {
    onOpenChange: function (val) {
      if (isAdapted) {
        var _props_onOpenChange;
        (_props_onOpenChange = props.onOpenChange) === null || _props_onOpenChange === void 0 ? void 0 : _props_onOpenChange.call(props, val);
      }
    },
    open: context.open,
    hidden: !isAdapted,
    children: props.children
  });
};
export { Dialog, DialogClose, DialogContent, DialogContext, DialogDescription, DialogOverlay, DialogOverlayFrame, DialogPortal, DialogPortalFrame, DialogProvider, DialogTitle, DialogTrigger, DialogWarningProvider, useDialogContext };
