// @tamagui/use-window-dimensions/dist/cjs/helpers.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var helpers_exports = {};
__export(helpers_exports, {
  getWindowSize: () => getWindowSize,
  subscribe: () => subscribe
});
module.exports = __toCommonJS(helpers_exports);
var import_constants = require("@tamagui/constants");
let lastSize = require("./initialValue.cjs").initialValue;
let docEl = null;
function getWindowSize() {
  docEl ||= window.document.documentElement;
  const nextSize = {
    fontScale: 1,
    height: docEl.clientHeight,
    scale: window.devicePixelRatio || 1,
    width: docEl.clientWidth
  };
  if (nextSize.height !== lastSize.height || nextSize.width !== lastSize.width || nextSize.scale !== lastSize.scale) {
    lastSize = nextSize;
    return nextSize;
  }
  return lastSize;
}
const cbs = /* @__PURE__ */new Set();
if (import_constants.isClient && typeof window !== "undefined" && typeof window.addEventListener === "function") {
  let flushUpdate = function () {
    lastUpdate = Date.now();
    cbs.forEach(cb => cb(getWindowSize()));
  };
  let lastUpdate = Date.now();
  let tm;
  const USER_MAX_MS = process.env.TAMAGUI_USE_WINDOW_DIMENSIONS_MAX_UPDATE_MS;
  const updateMaxMs = USER_MAX_MS ? +USER_MAX_MS : 100;
  const onResize = () => {
    clearTimeout(tm);
    const timeSinceLast = Date.now() - lastUpdate;
    if (timeSinceLast < updateMaxMs) tm = setTimeout(() => {
      flushUpdate();
    }, updateMaxMs - timeSinceLast);else flushUpdate();
  };
  window.addEventListener("resize", onResize);
}
function subscribe(cb) {
  cbs.add(cb);
  return () => cbs.delete(cb);
}
