// @tamagui/context-menu/dist/jsx/createNonNativeContextMenu.mjs
import { createBaseMenu } from "@tamagui/create-menu";
import { useControllableState } from "@tamagui/use-controllable-state";
import { composeEventHandlers, composeRefs, createStyledContext, isAndroid, isWeb, Slot, View, withStaticProperties } from "@tamagui/web";
import React, { useId } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
const CONTEXTMENU_CONTEXT = "ContextMenuContext";
function createNonNativeContextMenu(params) {
  const {
    Menu
  } = createBaseMenu(params);
  const CONTEXT_MENU_NAME = "ContextMenu";
  const {
    Provider: ContextMenuProvider,
    useStyledContext: useContextMenuContext
  } = createStyledContext();
  const ContextMenuComp = props => {
    const {
      scope,
      children,
      onOpenChange,
      dir,
      modal = true,
      ...rest
    } = props;
    const [open, setOpen] = React.useState(false);
    const triggerRef = React.useRef(null);
    const handleOpenChange = React.useCallback((open2, event) => {
      onOpenChange?.(open2, event);
      if (event?.defaultPrevented) return;
      setOpen(open2);
    }, [onOpenChange]);
    return /* @__PURE__ */jsx(ContextMenuProvider, {
      scope,
      triggerId: useId(),
      triggerRef,
      contentId: useId(),
      open,
      onOpenChange: handleOpenChange,
      modal,
      children: /* @__PURE__ */jsx(Menu, {
        scope: scope || CONTEXTMENU_CONTEXT,
        dir,
        open,
        onOpenChange: handleOpenChange,
        modal,
        ...rest,
        children
      })
    });
  };
  ContextMenuComp.displayName = CONTEXT_MENU_NAME;
  const TRIGGER_NAME = "ContextMenuTrigger";
  const ContextMenuTrigger = View.styleable((props, forwardedRef) => {
    const {
      scope,
      style,
      disabled = false,
      asChild,
      children,
      ...triggerProps
    } = props;
    const context = useContextMenuContext(scope);
    const pointRef = React.useRef({
      x: 0,
      y: 0
    });
    const virtualRef = React.useMemo(() => ({
      current: {
        getBoundingClientRect: () => {
          if (isWeb) {
            const scrollX = window.scrollX || document.documentElement.scrollLeft;
            const scrollY = window.scrollY || document.documentElement.scrollTop;
            return DOMRect.fromRect({
              width: 0,
              height: 0,
              x: pointRef.current.x - scrollX,
              y: pointRef.current.y - scrollY
            });
          }
          return {
            width: 0,
            height: 0,
            top: 0,
            left: 0,
            ...pointRef.current
          };
        },
        ...(!isWeb && {
          measure: c => c(pointRef.current.x, pointRef.current.y, 0, 0),
          measureInWindow: c => c(pointRef.current.x, pointRef.current.y, 0, 0)
        })
      }
    }), [pointRef.current.x, pointRef.current.y]);
    const longPressTimerRef = React.useRef(0);
    const clearLongPress = React.useCallback(() => window.clearTimeout(longPressTimerRef.current), []);
    const createOpenChangeEvent = () => {
      let defaultPrevented = false;
      return {
        get defaultPrevented() {
          return defaultPrevented;
        },
        preventDefault() {
          defaultPrevented = true;
        }
      };
    };
    const handleOpen = event => {
      if (isWeb && (event instanceof MouseEvent || event instanceof PointerEvent)) {
        pointRef.current = {
          x: event.clientX,
          y: event.clientY
        };
      } else {
        pointRef.current = {
          x: event.nativeEvent.pageX,
          y: event.nativeEvent.pageY
        };
      }
      const openChangeEvent = createOpenChangeEvent();
      context.onOpenChange(true, openChangeEvent);
      return openChangeEvent;
    };
    React.useEffect(() => clearLongPress, [clearLongPress]);
    React.useEffect(() => void (disabled && clearLongPress()), [disabled, clearLongPress]);
    const Comp = asChild ? Slot : View;
    return /* @__PURE__ */jsxs(Fragment, {
      children: [/* @__PURE__ */jsx(Menu.Anchor, {
        scope: scope || CONTEXTMENU_CONTEXT,
        virtualRef
      }), /* @__PURE__ */jsx(Comp, {
        render: "span",
        componentName: TRIGGER_NAME,
        id: context.triggerId,
        "data-state": context.open ? "open" : "closed",
        "data-disabled": disabled ? "" : void 0,
        ...triggerProps,
        ref: composeRefs(forwardedRef, context.triggerRef),
        style: isWeb ? {
          WebkitTouchCallout: "none",
          ...style
        } : null,
        ...(isWeb && {
          onContextMenu: disabled ? props.onContextMenu : composeEventHandlers(props.onContextMenu, event => {
            clearLongPress();
            const openChangeEvent = handleOpen(event);
            if (!openChangeEvent.defaultPrevented) {
              event.preventDefault();
            }
          }),
          onPointerDown: disabled ? props.onPointerDown : composeEventHandlers(props.onPointerDown, event => {
            if (event.pointerType === "mouse") return;
            clearLongPress();
            longPressTimerRef.current = window.setTimeout(() => handleOpen(event), 700);
          }),
          onPointerMove: disabled ? props.onPointerMove : composeEventHandlers(props.onPointerMove, event => {
            if (event.pointerType === "mouse") return;
            clearLongPress();
          }),
          onPointerCancel: disabled ? props.onPointerCancel : composeEventHandlers(props.onPointerCancel, event => {
            if (event.pointerType === "mouse") return;
            clearLongPress();
          }),
          onPointerUp: disabled ? props.onPointerUp : composeEventHandlers(props.onPointerUp, event => {
            if (event.pointerType === "mouse") return;
            clearLongPress();
          })
        }),
        ...(!isWeb && {
          onLongPress: disabled ? props.onLongPress : composeEventHandlers(props.onLongPress, event => {
            clearLongPress();
            const openChangeEvent = handleOpen(event);
            if (!openChangeEvent.defaultPrevented) {
              event.preventDefault();
            }
          })
        }),
        children
      })]
    });
  });
  ContextMenuTrigger.displayName = TRIGGER_NAME;
  const PORTAL_NAME = "ContextMenuPortal";
  const ContextMenuPortal = props => {
    const {
      scope,
      children,
      ...portalProps
    } = props;
    const context = isAndroid ? useContextMenuContext(scope) : null;
    const content = isAndroid ? /* @__PURE__ */jsx(ContextMenuProvider, {
      ...context,
      children
    }) : children;
    return /* @__PURE__ */jsx(Menu.Portal, {
      scope: scope || CONTEXTMENU_CONTEXT,
      ...portalProps,
      children: content
    });
  };
  ContextMenuPortal.displayName = PORTAL_NAME;
  const CONTENT_NAME = "ContextMenuContent";
  const ContextMenuContent = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...contentProps
    } = props;
    const context = useContextMenuContext(scope);
    const hasInteractedOutsideRef = React.useRef(false);
    return /* @__PURE__ */jsx(Menu.Content, {
      id: context.contentId,
      "aria-labelledby": context.triggerId,
      scope: scope || CONTEXTMENU_CONTEXT,
      ...contentProps,
      ref: forwardedRef,
      onCloseAutoFocus: composeEventHandlers(props.onCloseAutoFocus, event => {
        if (!hasInteractedOutsideRef.current) context.triggerRef.current?.focus();
        hasInteractedOutsideRef.current = false;
        event.preventDefault();
      }),
      onInteractOutside: composeEventHandlers(props.onInteractOutside, event => {
        const originalEvent = event.detail.originalEvent;
        const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
        const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
        if (!context.modal || isRightClick) hasInteractedOutsideRef.current = true;
      })
    });
  });
  ContextMenuContent.displayName = CONTENT_NAME;
  const ITEM_NAME = "ContextMenuItem";
  const ContextMenuItem = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...itemProps
    } = props;
    return /* @__PURE__ */jsx(Menu.Item, {
      componentName: ITEM_NAME,
      scope: scope || CONTEXTMENU_CONTEXT,
      ...itemProps,
      ref: forwardedRef
    });
  });
  ContextMenuItem.displayName = ITEM_NAME;
  const CHECKBOX_ITEM_NAME = "ContextMenuCheckboxItem";
  const ContextMenuCheckboxItem = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...checkboxItemProps
    } = props;
    return /* @__PURE__ */jsx(Menu.CheckboxItem, {
      componentName: CHECKBOX_ITEM_NAME,
      scope: scope || CONTEXTMENU_CONTEXT,
      ...checkboxItemProps,
      ref: forwardedRef
    });
  });
  ContextMenuCheckboxItem.displayName = CHECKBOX_ITEM_NAME;
  const RADIO_GROUP_NAME = "ContextMenuRadioGroup";
  const ContextMenuRadioGroup = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...radioGroupProps
    } = props;
    return /* @__PURE__ */jsx(Menu.RadioGroup, {
      scope: scope || CONTEXTMENU_CONTEXT,
      ...radioGroupProps,
      ref: forwardedRef
    });
  });
  ContextMenuRadioGroup.displayName = RADIO_GROUP_NAME;
  const RADIO_ITEM_NAME = "ContextMenuRadioItem";
  const ContextMenuRadioItem = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...radioItemProps
    } = props;
    return /* @__PURE__ */jsx(Menu.RadioItem, {
      componentName: RADIO_ITEM_NAME,
      scope: scope || CONTEXTMENU_CONTEXT,
      ...radioItemProps,
      ref: forwardedRef
    });
  });
  ContextMenuRadioItem.displayName = RADIO_ITEM_NAME;
  const INDICATOR_NAME = "ContextMenuItemIndicator";
  const ContextMenuItemIndicator = Menu.ItemIndicator.styleable((props, forwardedRef) => {
    const {
      scope,
      ...itemIndicatorProps
    } = props;
    return /* @__PURE__ */jsx(Menu.ItemIndicator, {
      componentName: INDICATOR_NAME,
      scope: scope || CONTEXTMENU_CONTEXT,
      ...itemIndicatorProps,
      ref: forwardedRef
    });
  });
  ContextMenuItemIndicator.displayName = INDICATOR_NAME;
  const SUB_NAME = "ContextMenuSub";
  const ContextMenuSub = props => {
    const {
      scope,
      children,
      onOpenChange,
      open: openProp,
      defaultOpen,
      ...rest
    } = props;
    const [open, setOpen] = useControllableState({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    return /* @__PURE__ */jsx(Menu.Sub, {
      scope: scope || CONTEXTMENU_CONTEXT,
      open,
      onOpenChange: setOpen,
      ...rest,
      children
    });
  };
  ContextMenuSub.displayName = SUB_NAME;
  const SUB_TRIGGER_NAME = "ContextMenuSubTrigger";
  const ContextMenuSubTrigger = View.styleable((props, forwardedRef) => {
    const {
      scope,
      ...subTriggerProps
    } = props;
    return /* @__PURE__ */jsx(Menu.SubTrigger, {
      componentName: SUB_TRIGGER_NAME,
      scope: scope || CONTEXTMENU_CONTEXT,
      ...subTriggerProps,
      ref: forwardedRef
    });
  });
  ContextMenuSubTrigger.displayName = SUB_TRIGGER_NAME;
  const SUB_CONTENT_NAME = "ContextMenuSubContent";
  const ContextMenuSubContent = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...subContentProps
    } = props;
    return /* @__PURE__ */jsx(Menu.SubContent, {
      scope: scope || CONTEXTMENU_CONTEXT,
      ...subContentProps,
      ref: forwardedRef,
      style: isWeb ? {
        ...props.style,
        ...{
          "--tamagui-context-menu-content-transform-origin": "var(--tamagui-popper-transform-origin)",
          "--tamagui-context-menu-content-available-width": "var(--tamagui-popper-available-width)",
          "--tamagui-context-menu-content-available-height": "var(--tamagui-popper-available-height)",
          "--tamagui-context-menu-trigger-width": "var(--tamagui-popper-anchor-width)",
          "--tamagui-context-menu-trigger-height": "var(--tamagui-popper-anchor-height)"
        }
      } : null
    });
  });
  ContextMenuSubContent.displayName = SUB_CONTENT_NAME;
  const ARROW_NAME = "ContextMenuArrow";
  const ContextMenuArrow = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...arrowProps
    } = props;
    return /* @__PURE__ */jsx(Menu.Arrow, {
      componentName: ARROW_NAME,
      scope: scope || CONTEXTMENU_CONTEXT,
      ...arrowProps,
      ref: forwardedRef
    });
  });
  ContextMenuArrow.displayName = ARROW_NAME;
  const ContextMenuGroup = Menu.Group;
  const ContextMenuLabel = Menu.Label;
  const ContextMenuSeparator = Menu.Separator;
  const ContextMenuItemTitle = Menu.ItemTitle;
  const ContextMenuItemSubTitle = Menu.ItemSubtitle;
  const ContextMenuItemImage = Menu.ItemImage;
  const ContextMenuItemIcon = Menu.ItemIcon;
  const ContextMenuPreview = () => null;
  return withStaticProperties(ContextMenuComp, {
    Root: ContextMenuComp,
    Trigger: ContextMenuTrigger,
    Portal: ContextMenuPortal,
    Content: ContextMenuContent,
    Group: ContextMenuGroup,
    Label: ContextMenuLabel,
    Item: ContextMenuItem,
    CheckboxItem: ContextMenuCheckboxItem,
    RadioGroup: ContextMenuRadioGroup,
    RadioItem: ContextMenuRadioItem,
    ItemIndicator: ContextMenuItemIndicator,
    Separator: ContextMenuSeparator,
    Arrow: ContextMenuArrow,
    Sub: ContextMenuSub,
    SubTrigger: ContextMenuSubTrigger,
    SubContent: ContextMenuSubContent,
    ItemTitle: ContextMenuItemTitle,
    ItemSubtitle: ContextMenuItemSubTitle,
    ItemIcon: ContextMenuItemIcon,
    ItemImage: ContextMenuItemImage,
    Preview: ContextMenuPreview
  });
}
export { CONTEXTMENU_CONTEXT, createNonNativeContextMenu };
