// @tamagui/input/dist/esm/Input.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import React from "react";
import { TextInput } from "react-native";
import { styled } from "@tamagui/core";
import { registerFocusable } from "@tamagui/focusable";
import { useNativeInputRef } from "@tamagui/element";
import { styledBody } from "./shared.native.js";
var StyledInput = styled(TextInput, styledBody[0], styledBody[1]);
var Input = StyledInput.styleable(function (props, forwardedRef) {
  var {
    // Web props we need to convert
    type,
    disabled,
    readOnly,
    id,
    rows,
    autoComplete,
    enterKeyHint,
    // Callbacks
    onChange,
    onInput,
    onKeyDown,
    onChangeText,
    onSubmitEditing,
    onSelectionChange,
    onEndEditing,
    onContentSizeChange,
    onScroll,
    onKeyPress: onKeyPressProp,
    selection,
    // Native-only props (pass through directly)
    keyboardAppearance,
    returnKeyType: returnKeyTypeProp,
    submitBehavior,
    blurOnSubmit,
    caretHidden,
    contextMenuHidden,
    selectTextOnFocus,
    secureTextEntry: secureTextEntryProp,
    maxFontSizeMultiplier,
    allowFontScaling,
    multiline: multilineProp,
    keyboardType: keyboardTypeProp,
    inputMode: inputModeProp,
    autoCapitalize: autoCapitalizeProp,
    autoCorrect: autoCorrectProp,
    autoFocusNative,
    textContentType,
    // iOS-only props
    clearButtonMode,
    clearTextOnFocus,
    enablesReturnKeyAutomatically,
    dataDetectorTypes,
    scrollEnabled,
    passwordRules,
    rejectResponderTermination,
    spellCheck,
    lineBreakStrategyIOS,
    lineBreakModeIOS,
    smartInsertDelete,
    inputAccessoryViewID,
    inputAccessoryViewButtonLabel,
    disableKeyboardShortcuts,
    // Android-only props
    importantForAutofill,
    disableFullscreenUI,
    inlineImageLeft,
    inlineImagePadding,
    returnKeyLabel,
    textBreakStrategy,
    textAlignVertical,
    verticalAlign,
    showSoftInputOnFocus,
    numberOfLines: numberOfLinesProp,
    // Web-only props to filter out
    // @ts-ignore
    dirname,
    min,
    max,
    minLength,
    multiple,
    name,
    pattern,
    required,
    step,
    render,
    ...rest
  } = props;
  var {
    ref,
    composedRef
  } = useNativeInputRef(forwardedRef);
  var secureTextEntry = secureTextEntryProp !== null && secureTextEntryProp !== void 0 ? secureTextEntryProp : false;
  var keyboardType = keyboardTypeProp !== null && keyboardTypeProp !== void 0 ? keyboardTypeProp : "default";
  var inputMode = inputModeProp;
  if (!secureTextEntryProp && !keyboardTypeProp && !inputModeProp) {
    switch (type) {
      case "password":
        secureTextEntry = true;
        break;
      case "email":
        keyboardType = "email-address";
        inputMode = "email";
        break;
      case "tel":
        keyboardType = "phone-pad";
        inputMode = "tel";
        break;
      case "number":
        keyboardType = "numeric";
        inputMode = "numeric";
        break;
      case "url":
        keyboardType = "url";
        inputMode = "url";
        break;
      case "search":
        inputMode = "search";
        break;
    }
  }
  var returnKeyType = returnKeyTypeProp;
  if (!returnKeyType && enterKeyHint) {
    switch (enterKeyHint) {
      case "done":
        returnKeyType = "done";
        break;
      case "go":
        returnKeyType = "go";
        break;
      case "next":
        returnKeyType = "next";
        break;
      case "search":
        returnKeyType = "search";
        break;
      case "send":
        returnKeyType = "send";
        break;
    }
  }
  var multiline = multilineProp !== null && multilineProp !== void 0 ? multilineProp : render === "textarea" || rows && rows > 1;
  var numberOfLines = numberOfLinesProp !== null && numberOfLinesProp !== void 0 ? numberOfLinesProp : rows;
  var autoCorrect = autoCorrectProp === "on" ? true : autoCorrectProp === "off" ? false : autoCorrectProp;
  var autoCapitalize = autoCapitalizeProp === "on" ? "sentences" : autoCapitalizeProp === "off" ? "none" : autoCapitalizeProp;
  React.useEffect(function () {
    if (!id || disabled) return;
    return registerFocusable(id, {
      focusAndSelect: function () {
        var _ref_current;
        return (_ref_current = ref.current) === null || _ref_current === void 0 ? void 0 : _ref_current.focus();
      },
      focus: function () {
        var _ref_current;
        return (_ref_current = ref.current) === null || _ref_current === void 0 ? void 0 : _ref_current.focus();
      }
    });
  }, [id, disabled]);
  var handleChangeText = function (text) {
    onChangeText === null || onChangeText === void 0 ? void 0 : onChangeText(text);
    if (onChange) {
      onChange({
        target: {
          value: text
        },
        type: "change"
      });
    }
    if (onInput) {
      onInput({
        target: {
          value: text
        },
        type: "input"
      });
    }
  };
  var handleKeyPress = function (e) {
    onKeyPressProp === null || onKeyPressProp === void 0 ? void 0 : onKeyPressProp(e);
    if (onKeyDown) {
      var {
        key
      } = e.nativeEvent;
      if (key === "Backspace" || key === "Enter" || key.length === 1) {
        onKeyDown({
          key,
          type: "keydown"
        });
      }
    }
  };
  var handleSubmitEditing = function (e) {
    if (onKeyDown) {
      onKeyDown({
        key: "Enter",
        type: "keydown"
      });
    }
    onSubmitEditing === null || onSubmitEditing === void 0 ? void 0 : onSubmitEditing(e);
  };
  var handleSelectionChange = function (e) {
    onSelectionChange === null || onSelectionChange === void 0 ? void 0 : onSelectionChange(e);
  };
  var finalProps = {
    ...rest,
    editable: !disabled && !readOnly,
    secureTextEntry,
    keyboardType,
    keyboardAppearance,
    inputMode,
    returnKeyType,
    multiline,
    numberOfLines,
    selection,
    autoComplete,
    autoFocus: autoFocusNative,
    // callbacks
    onChangeText: handleChangeText,
    onKeyPress: onKeyPressProp || onKeyDown ? handleKeyPress : void 0,
    onSubmitEditing: onKeyDown || onSubmitEditing ? handleSubmitEditing : void 0,
    onSelectionChange: onSelectionChange ? handleSelectionChange : void 0,
    onEndEditing,
    onContentSizeChange,
    onScroll,
    // cross-platform native props
    submitBehavior,
    blurOnSubmit,
    caretHidden,
    contextMenuHidden,
    selectTextOnFocus,
    maxFontSizeMultiplier,
    allowFontScaling,
    autoCapitalize,
    autoCorrect,
    textContentType,
    // iOS-only props
    clearButtonMode,
    clearTextOnFocus,
    enablesReturnKeyAutomatically,
    dataDetectorTypes,
    scrollEnabled,
    passwordRules,
    rejectResponderTermination,
    spellCheck,
    lineBreakStrategyIOS,
    lineBreakModeIOS,
    smartInsertDelete,
    inputAccessoryViewID,
    inputAccessoryViewButtonLabel,
    disableKeyboardShortcuts,
    // Android-only props
    importantForAutofill,
    disableFullscreenUI,
    inlineImageLeft,
    inlineImagePadding,
    returnKeyLabel,
    textBreakStrategy,
    textAlignVertical,
    verticalAlign,
    showSoftInputOnFocus
  };
  return /* @__PURE__ */_jsx(StyledInput, {
    ref: composedRef,
    ...finalProps
  });
});
export { Input };
