// @tamagui/input/dist/esm/shared.native.js
import { Text } from "@tamagui/core";
import { getVariableValue, isWeb } from "@tamagui/core";
import { getButtonSized } from "@tamagui/get-button-sized";
import { getFontSized } from "@tamagui/get-font-sized";
import { getSpace } from "@tamagui/get-token";
var defaultStyles = {
  size: "$true",
  fontFamily: "$body",
  borderWidth: 1,
  outlineWidth: 0,
  color: "$color",
  ...(isWeb ? {
    tabIndex: 0
  } : {
    focusable: true
  }),
  borderColor: "$borderColor",
  backgroundColor: "$background",
  // this fixes a flex bug where it overflows container
  minWidth: 0,
  hoverStyle: {
    borderColor: "$borderColorHover"
  },
  focusStyle: {
    borderColor: "$borderColorFocus"
  },
  focusVisibleStyle: {
    outlineColor: "$outlineColor",
    outlineWidth: 2,
    outlineStyle: "solid"
  }
};
var inputSizeVariant = function () {
  var val = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "$true",
    extras = arguments.length > 1 ? arguments[1] : void 0;
  if (extras.props.tag === "textarea" || extras.props.rows > 1 || extras.props.multiline || extras.props.numberOfLines > 1) {
    return textAreaSizeVariant(val, extras);
  }
  var buttonStyles = getButtonSized(val, extras);
  var paddingHorizontal = getSpace(val, {
    shift: -1,
    bounds: [2]
  });
  var fontStyle = getFontSized(val, extras);
  if (!isWeb && fontStyle) {
    delete fontStyle["lineHeight"];
  }
  return {
    ...fontStyle,
    ...buttonStyles,
    paddingHorizontal
  };
};
var textAreaSizeVariant = function () {
  var val = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "$true",
    extras = arguments.length > 1 ? arguments[1] : void 0;
  var {
    props
  } = extras;
  var buttonStyles = getButtonSized(val, extras);
  var fontStyle = getFontSized(val, extras);
  var _props_rows;
  var lines = (_props_rows = props.rows) !== null && _props_rows !== void 0 ? _props_rows : props.numberOfLines;
  var height = typeof lines === "number" ? lines * getVariableValue(fontStyle.lineHeight) : "auto";
  if (!isWeb && fontStyle) {
    delete fontStyle["lineHeight"];
  }
  var paddingVertical = getSpace(val, {
    shift: -2,
    bounds: [2]
  });
  var paddingHorizontal = getSpace(val, {
    shift: -1,
    bounds: [2]
  });
  return {
    ...buttonStyles,
    ...fontStyle,
    paddingVertical,
    paddingHorizontal,
    height
  };
};
var INPUT_NAME = "Input";
var styledBody = [{
  name: INPUT_NAME,
  render: "input",
  variants: {
    unstyled: {
      false: defaultStyles
    },
    size: {
      "...size": inputSizeVariant
    },
    disabled: {
      true: {}
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
}, {
  isInput: true,
  accept: {
    placeholderTextColor: "color",
    selectionColor: "color",
    cursorColor: "color",
    selectionHandleColor: "color",
    underlineColorAndroid: "color"
  },
  validStyles: Text.staticConfig.validStyles
}];
export { INPUT_NAME, defaultStyles, inputSizeVariant, styledBody, textAreaSizeVariant };
