// @tamagui/web/dist/esm/helpers/defaultAnimationDriver.native.js
var noAnimationDriver = function (method) {
  console.warn(`No animation driver configured. To use ${method}, you must pass \`animations\` to createTamagui. See: https://tamagui.dev/docs/core/animations`);
};
var createEmptyAnimationDriver = function () {
  return {
    isReactNative: false,
    inputStyle: "css",
    outputStyle: "css",
    isStub: true,
    animations: {},
    useAnimations: function () {
      return noAnimationDriver("animations");
    },
    usePresence: function () {
      return noAnimationDriver("usePresence");
    },
    ResetPresence: function () {
      return noAnimationDriver("ResetPresence");
    },
    useAnimatedNumber: function () {
      return noAnimationDriver("useAnimatedNumber");
    },
    useAnimatedNumberStyle: function () {
      return noAnimationDriver("useAnimatedNumberStyle");
    },
    useAnimatedNumbersStyle: function () {
      return noAnimationDriver("useAnimatedNumbersStyle");
    },
    useAnimatedNumberReaction: function () {
      return noAnimationDriver("useAnimatedNumberReaction");
    }
  };
};
var defaultAnimationDriver = createEmptyAnimationDriver();
export { defaultAnimationDriver };
