// @tamagui/web/dist/esm/helpers/expandStyle.mjs
import { isAndroid, isWeb } from "@tamagui/constants";
import { getSetting } from "../config.mjs";
import { parseBorderShorthand } from "./parseBorderShorthand.mjs";
import { parseOutlineShorthand } from "./parseOutlineShorthand.mjs";
const neg1Flex = [["flexGrow", 0], ["flexShrink", 1], ["flexBasis", "auto"]];
function expandStyle(key, value) {
  if (key === "flex") {
    if (value === -1) return neg1Flex;
    return [["flexGrow", value], ["flexShrink", 1], ["flexBasis", getSetting("styleCompat") === "legacy" ? "auto" : 0]];
  }
  switch (key) {
    case "writingDirection":
      return [["direction", value]];
    case "backdropFilter":
      return [["backdropFilter", value], ["WebkitBackdropFilter", value]];
  }
  if (key in EXPANSIONS) return EXPANSIONS[key].map(k => [k, value]);
}
const all = ["Top", "Right", "Bottom", "Left"];
const horiz = ["Right", "Left"];
const vert = ["Top", "Bottom"];
const xy = ["X", "Y"];
const EXPANSIONS = {
  borderColor: ["TopColor", "RightColor", "BottomColor", "LeftColor"],
  borderRadius: ["TopLeftRadius", "TopRightRadius", "BottomRightRadius", "BottomLeftRadius"],
  borderWidth: ["TopWidth", "RightWidth", "BottomWidth", "LeftWidth"],
  margin: all,
  marginHorizontal: horiz,
  marginVertical: vert,
  padding: all,
  paddingHorizontal: horiz,
  paddingVertical: vert,
  ...(isWeb && {
    borderStyle: ["TopStyle", "RightStyle", "BottomStyle", "LeftStyle"],
    overflow: xy,
    overscrollBehavior: xy
  })
};
for (const parent in EXPANSIONS) {
  const prefix = parent.slice(0, /[A-Z]/.exec(parent)?.index ?? parent.length);
  EXPANSIONS[parent] = EXPANSIONS[parent].map(k => `${prefix}${k}`);
}
export { expandStyle };
