// @tamagui/web/dist/esm/helpers/insertStyleRule.mjs
import { StyleObjectIdentifier, StyleObjectRules } from "@tamagui/helpers";
import { createVariable } from "../createVariable.mjs";
const scannedCache = /* @__PURE__ */new WeakMap();
const totalSelectorsInserted = /* @__PURE__ */new Map();
const allSelectors = {};
const allRules = {};
const getAllSelectors = () => allSelectors;
const getAllRules = () => {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) return Object.keys(allRules).sort().map(key => allRules[key]);
  return [];
};
let lastScannedSheets = null;
function scanAllSheets(collectThemes = false, tokens) {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    if (process.env.NODE_ENV === "test") return;
    let themes;
    const sheets = document.styleSheets || [];
    const prev = lastScannedSheets;
    const current = new Set(sheets);
    for (const sheet2 of current) if (sheet2) {
      const out = updateSheetStyles(sheet2, false, collectThemes, tokens);
      if (out) themes = out;
    }
    lastScannedSheets = current;
    if (prev) {
      for (const sheet2 of prev) if (sheet2 && !current.has(sheet2)) updateSheetStyles(sheet2, true);
    }
    return themes;
  }
}
function trackInsertedStyle(id) {
  const next = (totalSelectorsInserted.get(id) || 0) + 1;
  totalSelectorsInserted.set(id, next);
  return next;
}
const bailAfterEnv = process.env.TAMAGUI_BAIL_AFTER_SCANNING_X_CSS_RULES;
const bailAfter = bailAfterEnv ? +bailAfterEnv : 8e3;
function updateSheetStyles(sheet2, remove = false, collectThemes = false, tokens) {
  let rules;
  try {
    rules = sheet2.cssRules;
    if (!rules) return;
  } catch {
    return;
  }
  const firstSelector = getTamaguiSelector(rules[0], collectThemes)?.[0];
  const lastSelector = getTamaguiSelector(rules[rules.length - 1], collectThemes)?.[0];
  const cacheKey = `${rules.length}${firstSelector}${lastSelector}`;
  const lastScanned = scannedCache.get(sheet2);
  if (!remove) {
    if (lastScanned === cacheKey) return;
  }
  const len = rules.length;
  let fails = 0;
  let dedupedThemes;
  const nameToTheme = {};
  for (let i = 0; i < len; i++) {
    const rule = rules[i];
    if (!(rule instanceof CSSStyleRule)) continue;
    const response = getTamaguiSelector(rule, collectThemes);
    if (response) fails = 0;else {
      fails++;
      if (fails > bailAfter) return;
      continue;
    }
    const [identifier, cssRule, isTheme] = response;
    if (isTheme) {
      const deduped = addThemesFromCSS(cssRule, tokens);
      if (deduped) {
        for (const name of deduped.names) if (nameToTheme[name]) {
          Object.apply(nameToTheme[name], deduped.theme);
          deduped.names = deduped.names.filter(x => x !== name);
        } else nameToTheme[name] = deduped.theme;
        dedupedThemes ||= [];
        dedupedThemes.push(deduped);
      }
      continue;
    }
  }
  scannedCache.set(sheet2, cacheKey);
  return dedupedThemes;
}
let colorVarToVal;
let rootComputedStyle = null;
function addThemesFromCSS(cssStyleRule, tokens) {
  const selectors = cssStyleRule.selectorText.split(",");
  if (!selectors.length) return;
  if (tokens?.color && !colorVarToVal) {
    colorVarToVal = {};
    for (const key in tokens.color) {
      const token = tokens.color[key];
      if (token) colorVarToVal[token.name] = token.val;
    }
  }
  const rules = (cssStyleRule.cssText || "").slice(cssStyleRule.selectorText.length + 2, -1).split(";");
  const values = {};
  for (const rule of rules) {
    const sepI = rule.indexOf(":");
    if (sepI === -1) continue;
    const varIndex = rule.indexOf("--");
    let key = rule.slice(varIndex === -1 ? 0 : varIndex + 2, sepI);
    if (process.env.TAMAGUI_CSS_VARIABLE_PREFIX) key = key.replace(process.env.TAMAGUI_CSS_VARIABLE_PREFIX, "");
    const val = rule.slice(sepI + 2);
    let value;
    if (val[0] === "v" && val.startsWith("var(")) {
      const varName = val.slice(6, -1);
      const tokenVal = colorVarToVal?.[varName];
      if (tokenVal) value = tokenVal;else {
        rootComputedStyle ||= getComputedStyle(document.body);
        value = rootComputedStyle.getPropertyValue("--" + varName);
      }
    } else value = val;
    values[key] = createVariable({
      key,
      name: key,
      val: value
    }, true);
  }
  const names = /* @__PURE__ */new Set();
  for (const selector of selectors) {
    if (selector === " .tm_xxt") continue;
    const lastThemeSelectorIndex = selector.lastIndexOf(".t_");
    const name = selector.slice(lastThemeSelectorIndex).slice(3);
    const [schemeChar] = selector[lastThemeSelectorIndex - 5];
    const scheme = schemeChar === "d" ? "dark" : schemeChar === "i" ? "light" : "";
    const themeName = scheme && scheme !== name ? `${scheme}_${name}` : name;
    if (!themeName || themeName === "light_dark" || themeName === "dark_light") continue;
    names.add(themeName);
  }
  return {
    names: [...names],
    theme: values
  };
}
const tamaguiSelectorRegex = /\.tm_xxt/;
function getTamaguiSelector(rule, collectThemes = false) {
  if (rule instanceof CSSStyleRule) {
    const text = rule.selectorText;
    if (text[0] === ":" && text[1] === "r" && tamaguiSelectorRegex.test(text)) {
      const id = getIdentifierFromTamaguiSelector(
      // next.js minifies it so its in front
      text.replace(tamaguiSelectorRegex, ""));
      return collectThemes ? [id, rule, true] : [id, rule];
    }
  } else if (rule instanceof CSSMediaRule) {
    if (rule.cssRules.length > 1) return;
    return getTamaguiSelector(rule.cssRules[0]);
  }
}
const getIdentifierFromTamaguiSelector = selector => {
  const dotIndex = selector.indexOf(":");
  if (dotIndex > -1) return selector.slice(7, dotIndex);
  return selector.slice(7);
};
let sheet = null;
let trackAllRules = true;
function stopAccumulatingRules() {
  trackAllRules = false;
}
function updateRules(identifier, rules) {
  if (trackAllRules) allRules[identifier] = rules.join(" ");
  return true;
}
let nonce = "";
function setNonce(_) {
  nonce = _;
}
function insertStyleRules(rulesToInsert) {
  if (!sheet && document.head) {
    const styleTag = document.createElement("style");
    styleTag.id = "_tamagui-styles";
    if (nonce) styleTag.nonce = nonce;
    sheet = document.head.appendChild(styleTag).sheet;
  }
  if (!sheet) return;
  for (const key in rulesToInsert) {
    const styleObject = rulesToInsert[key];
    const identifier = styleObject[StyleObjectIdentifier];
    if (!shouldInsertStyleRules(identifier)) continue;
    const rules = styleObject[StyleObjectRules];
    allSelectors[identifier] = rules.join("\n");
    trackInsertedStyle(identifier);
    updateRules(identifier, rules);
    try {
      for (const rule of rules) sheet.insertRule(rule, sheet.cssRules.length);
    } catch (err) {
      if (process.env.NODE_ENV === "production") console.error(`Error inserting style rule`, rules);
    }
  }
}
const maxToInsert = process.env.TAMAGUI_INSERT_SELECTOR_TRIES ? +process.env.TAMAGUI_INSERT_SELECTOR_TRIES : 1;
function shouldInsertStyleRules(identifier) {
  if (process.env.IS_STATIC === "is_static") return true;
  const total = totalSelectorsInserted.get(identifier) || 0;
  if (process.env.NODE_ENV === "development") {
    if (total > +(process.env.TAMAGUI_STYLE_INSERTION_WARNING_LIMIT || 10)) console.warn(`Warning: inserting many CSS rules, you may be animating something and generating many CSS insertions, which can degrade performance. Instead, try using the "disableClassName" property on elements that change styles often. To disable this warning set TAMAGUI_STYLE_INSERTION_WARNING_LIMIT from 50000 to something higher`);
  }
  return total < maxToInsert;
}
export { getAllRules, getAllSelectors, insertStyleRules, scanAllSheets, setNonce, shouldInsertStyleRules, stopAccumulatingRules, updateRules };
