// @tamagui/web/dist/esm/helpers/pointerEvents.native.js
import { useRef } from "react";
import { composeEventHandlers } from "@tamagui/helpers";
function usePointerEvents(props, viewProps) {
  var {
    onPointerDown,
    onPointerUp,
    onPointerMove,
    onPointerCancel,
    onPointerEnter,
    onPointerLeave
  } = props;
  var hasPointerEvents = onPointerDown || onPointerUp || onPointerMove || onPointerCancel || onPointerEnter || onPointerLeave;
  var isInsideRef = useRef(false);
  var layoutRef = useRef({
    width: 0,
    height: 0
  });
  var isCapturedRef = useRef(false);
  if (!hasPointerEvents) return;
  var createNormalizedEvent = function (e) {
    var touch = e.nativeEvent;
    var _touch_identifier;
    var normalized = {
      ...e,
      clientX: touch.pageX,
      clientY: touch.pageY,
      pageX: touch.pageX,
      pageY: touch.pageY,
      offsetX: touch.locationX,
      offsetY: touch.locationY,
      pointerType: "touch",
      pointerId: (_touch_identifier = touch.identifier) !== null && _touch_identifier !== void 0 ? _touch_identifier : 0,
      nativeEvent: touch,
      target: {
        setPointerCapture: function (_pointerId) {
          isCapturedRef.current = true;
        },
        releasePointerCapture: function (_pointerId) {
          isCapturedRef.current = false;
        }
      }
    };
    return normalized;
  };
  if (onPointerDown) {
    viewProps.onTouchStart = composeEventHandlers(viewProps.onTouchStart, function (e) {
      onPointerDown(createNormalizedEvent(e));
    });
  }
  if (onPointerUp) {
    viewProps.onTouchEnd = composeEventHandlers(viewProps.onTouchEnd, function (e) {
      isCapturedRef.current = false;
      onPointerUp(createNormalizedEvent(e));
    });
  }
  if (onPointerMove) {
    viewProps.onTouchMove = composeEventHandlers(viewProps.onTouchMove, function (e) {
      var {
        locationX,
        locationY
      } = e.nativeEvent;
      var {
        width,
        height
      } = layoutRef.current;
      var isInBounds = locationX >= 0 && locationX <= width && locationY >= 0 && locationY <= height;
      if (isCapturedRef.current || isInBounds) {
        onPointerMove(createNormalizedEvent(e));
      }
    });
  }
  if (onPointerCancel) {
    viewProps.onTouchCancel = composeEventHandlers(viewProps.onTouchCancel, function (e) {
      isCapturedRef.current = false;
      onPointerCancel(createNormalizedEvent(e));
    });
  }
  if (onPointerEnter || onPointerLeave || onPointerMove) {
    viewProps.onLayout = composeEventHandlers(viewProps.onLayout, function (e) {
      layoutRef.current = {
        width: e.nativeEvent.layout.width,
        height: e.nativeEvent.layout.height
      };
    });
  }
  if (onPointerEnter) {
    viewProps.onTouchStart = composeEventHandlers(viewProps.onTouchStart, function (e) {
      var {
        locationX,
        locationY
      } = e.nativeEvent;
      var {
        width,
        height
      } = layoutRef.current;
      if (locationX >= 0 && locationX <= width && locationY >= 0 && locationY <= height) {
        isInsideRef.current = true;
        onPointerEnter(createNormalizedEvent(e));
      }
    });
  }
  if (onPointerLeave) {
    viewProps.onTouchMove = composeEventHandlers(viewProps.onTouchMove, function (e) {
      var {
        locationX,
        locationY
      } = e.nativeEvent;
      var {
        width,
        height
      } = layoutRef.current;
      var isInside = locationX >= 0 && locationX <= width && locationY >= 0 && locationY <= height;
      if (isInsideRef.current && !isInside) {
        isInsideRef.current = false;
        onPointerLeave(createNormalizedEvent(e));
      }
    });
    viewProps.onTouchEnd = composeEventHandlers(viewProps.onTouchEnd, function (e) {
      if (isInsideRef.current) {
        isInsideRef.current = false;
        onPointerLeave(createNormalizedEvent(e));
      }
    });
  }
}
export { usePointerEvents };
