// @tamagui/web/dist/esm/helpers/registerCSSVariable.native.js
import { createCSSVariable, getVariableValue } from "../createVariable.native.js";
var registerCSSVariable = function (v) {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    tokensValueToVariable.set(getVariableValue(v), v);
  }
};
var variableToCSS = function (v) {
  var unitless = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : false;
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    return `--${process.env.TAMAGUI_CSS_VARIABLE_PREFIX || ""}${createCSSVariable(v.name, false)}:${!unitless && typeof v.val === "number" ? `${v.val}px` : v.val}`;
  }
  return "";
};
var tokensValueToVariable = /* @__PURE__ */new Map();
var autoVarId = 0;
var autoVariables = [];
var getOrCreateVariable = function (val) {
  if (tokensValueToVariable.has(val)) {
    return tokensValueToVariable.get(val);
  }
  var name = `t${autoVarId++}`;
  var variable = `var(--${name})`;
  var v = {
    val,
    name,
    variable
  };
  tokensValueToVariable.set(val, v);
  autoVariables.push(v);
  return v;
};
var mutatedVarId = 1e4;
var mutatedAutoVariables = [];
var mutatedTokensValueToVariable = /* @__PURE__ */new Map();
var getOrCreateMutatedVariable = function (val) {
  if (mutatedTokensValueToVariable.has(val)) {
    return mutatedTokensValueToVariable.get(val);
  }
  var name = `t${mutatedVarId++}`;
  var variable = `var(--${name})`;
  var v = {
    val,
    name,
    variable
  };
  mutatedTokensValueToVariable.set(val, v);
  mutatedAutoVariables.push(v);
  return v;
};
export { autoVariables, getOrCreateMutatedVariable, getOrCreateVariable, mutatedAutoVariables, registerCSSVariable, tokensValueToVariable, variableToCSS };
