// @tamagui/web/dist/esm/helpers/registerCSSVariable.mjs
import { createCSSVariable, getVariableValue } from "../createVariable.mjs";
const registerCSSVariable = v => {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    tokensValueToVariable.set(getVariableValue(v), v);
  }
};
const variableToCSS = (v, unitless = false) => {
  if (!process.env.TAMAGUI_DID_OUTPUT_CSS) {
    return `--${process.env.TAMAGUI_CSS_VARIABLE_PREFIX || ""}${createCSSVariable(v.name, false)}:${!unitless && typeof v.val === "number" ? `${v.val}px` : v.val}`;
  }
  return "";
};
const tokensValueToVariable = /* @__PURE__ */new Map();
let autoVarId = 0;
const autoVariables = [];
const getOrCreateVariable = val => {
  if (tokensValueToVariable.has(val)) {
    return tokensValueToVariable.get(val);
  }
  const name = `t${autoVarId++}`;
  const variable = `var(--${name})`;
  const v = {
    val,
    name,
    variable
  };
  tokensValueToVariable.set(val, v);
  autoVariables.push(v);
  return v;
};
let mutatedVarId = 1e4;
const mutatedAutoVariables = [];
const mutatedTokensValueToVariable = /* @__PURE__ */new Map();
const getOrCreateMutatedVariable = val => {
  if (mutatedTokensValueToVariable.has(val)) {
    return mutatedTokensValueToVariable.get(val);
  }
  const name = `t${mutatedVarId++}`;
  const variable = `var(--${name})`;
  const v = {
    val,
    name,
    variable
  };
  mutatedTokensValueToVariable.set(val, v);
  mutatedAutoVariables.push(v);
  return v;
};
export { autoVariables, getOrCreateMutatedVariable, getOrCreateVariable, mutatedAutoVariables, registerCSSVariable, tokensValueToVariable, variableToCSS };
