// @tamagui/menu/dist/cjs/createNonNativeMenu.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createNonNativeMenu_exports = {};
__export(createNonNativeMenu_exports, {
  DROPDOWN_MENU_CONTEXT: () => DROPDOWN_MENU_CONTEXT,
  createNonNativeMenu: () => createNonNativeMenu
});
module.exports = __toCommonJS(createNonNativeMenu_exports);
var import_create_menu = require("@tamagui/create-menu");
var import_popper = require("@tamagui/popper");
var import_scroll_view = require("@tamagui/scroll-view");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var import_web = require("@tamagui/web");
var React = __toESM(require("react"), 1);
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
const DROPDOWN_MENU_CONTEXT = "MenuContext";
function useMenuTriggerSetup(open) {
  const triggerStateSettersRef = React.useRef(/* @__PURE__ */new Map());
  const activeTriggerIdRef = React.useRef(null);
  const setActiveTrigger = (0, import_web.useEvent)(id => {
    const prevId = activeTriggerIdRef.current;
    if (prevId === id) return;
    if (prevId) {
      triggerStateSettersRef.current.get(prevId)?.(false);
    }
    activeTriggerIdRef.current = id;
    if (id && open) {
      triggerStateSettersRef.current.get(id)?.(true);
    }
  });
  const registerTrigger = (0, import_web.useEvent)((id, setOpenState) => {
    triggerStateSettersRef.current.set(id, setOpenState);
    setOpenState(activeTriggerIdRef.current === id && open);
  });
  const unregisterTrigger = (0, import_web.useEvent)(id => {
    triggerStateSettersRef.current.delete(id);
    if (activeTriggerIdRef.current === id) {
      activeTriggerIdRef.current = null;
    }
  });
  React.useEffect(() => {
    if (!open) {
      setActiveTrigger(null);
      return;
    }
    const activeId = activeTriggerIdRef.current;
    if (activeId) {
      triggerStateSettersRef.current.get(activeId)?.(true);
    }
  }, [open, setActiveTrigger]);
  return {
    setActiveTrigger,
    registerTrigger,
    unregisterTrigger
  };
}
function createNonNativeMenu(params) {
  const {
    Menu
  } = (0, import_create_menu.createBaseMenu)(params);
  const DROPDOWN_MENU_NAME = "Menu";
  const {
    Provider: MenuProvider,
    useStyledContext: useMenuContext
  } = (0, import_web.createStyledContext)();
  const MenuComp = props => {
    const {
      scope,
      children,
      dir,
      open: openProp,
      defaultOpen,
      onOpenChange,
      modal = true,
      ...rest
    } = props;
    const triggerRef = React.useRef(null);
    const [open = false, setOpen] = (0, import_use_controllable_state.useControllableState)({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    const openRef = React.useRef(open);
    openRef.current = open;
    const {
      setActiveTrigger,
      registerTrigger,
      unregisterTrigger
    } = useMenuTriggerSetup(open);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuProvider, {
      scope,
      triggerId: (0, import_react.useId)(),
      triggerRef,
      contentId: (0, import_react.useId)(),
      openRef,
      onOpenChange: React.useCallback(nextOpen => setOpen(nextOpen), [setOpen]),
      onOpenToggle: React.useCallback(() => setOpen(prevOpen => !prevOpen), [setOpen]),
      modal,
      setActiveTrigger,
      registerTrigger,
      unregisterTrigger,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(Menu, {
        scope: scope || DROPDOWN_MENU_CONTEXT,
        open,
        onOpenChange: setOpen,
        dir,
        modal,
        ...rest,
        children
      })
    });
  };
  MenuComp.displayName = DROPDOWN_MENU_NAME;
  const TRIGGER_NAME = "MenuTrigger";
  const MenuTriggerFrame = Menu.Anchor;
  const MenuTrigger = import_web.View.styleable((props, forwardedRef) => {
    const {
      scope,
      asChild,
      children,
      disabled = false,
      onKeydown,
      ...triggerProps
    } = props;
    const context = useMenuContext(scope);
    const popperCtx = (0, import_popper.usePopperContextSlow)(scope || DROPDOWN_MENU_CONTEXT);
    const Comp = asChild ? import_web.Slot : import_web.View;
    const isTouchDevice = (0, import_web.useIsTouchDevice)();
    const triggerElRef = React.useRef(null);
    const triggerId = React.useId();
    const [triggerOpen, setTriggerOpen] = React.useState(false);
    const {
      registerTrigger,
      unregisterTrigger
    } = context;
    React.useEffect(() => {
      registerTrigger(triggerId, setTriggerOpen);
      return () => unregisterTrigger(triggerId);
    }, [registerTrigger, unregisterTrigger, triggerId]);
    const activateSelf = React.useCallback(() => {
      context.setActiveTrigger(triggerId);
      const el = triggerElRef.current;
      if (el) {
        context.triggerRef.current = el;
        if (el instanceof HTMLElement) {
          popperCtx.refs?.setReference(el);
          requestAnimationFrame(() => popperCtx.update?.());
        }
      }
    }, [context, triggerId, popperCtx]);
    const pressEvent = import_web.isWeb ? isTouchDevice ? "onClick" : "onPointerDown" : "onPress";
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuTriggerFrame, {
      asChild: true,
      componentName: TRIGGER_NAME,
      scope: scope || DROPDOWN_MENU_CONTEXT,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(Comp, {
        role: "button",
        id: context.triggerId,
        "aria-haspopup": "menu",
        "aria-expanded": triggerOpen,
        "aria-controls": triggerOpen ? context.contentId : void 0,
        "data-state": triggerOpen ? "open" : "closed",
        "data-disabled": disabled ? "" : void 0,
        "aria-disabled": disabled || void 0,
        ref: (0, import_web.composeRefs)(forwardedRef, context.triggerRef, triggerElRef),
        ...{
          [pressEvent]: (0, import_web.composeEventHandlers)(
          //@ts-ignore
          props[pressEvent], event => {
            if (!disabled) {
              if (import_web.isWeb && event instanceof PointerEvent && event.button !== 0 && event.ctrlKey === true) return;
              if (context.openRef.current) {
                context.setActiveTrigger(null);
              } else {
                activateSelf();
              }
              context.onOpenToggle();
              if (!context.openRef.current) event.preventDefault();
            }
          })
        },
        ...(import_web.isWeb && {
          onKeyDown: (0, import_web.composeEventHandlers)(onKeydown, event => {
            if (disabled) return;
            if (["Enter", " "].includes(event.key)) {
              if (context.openRef.current) {
                context.setActiveTrigger(null);
              } else {
                activateSelf();
              }
              context.onOpenToggle();
            }
            if (event.key === "ArrowDown") {
              activateSelf();
              context.onOpenChange(true);
            }
            if (["Enter", " ", "ArrowDown"].includes(event.key)) event.preventDefault();
          })
        }),
        ...triggerProps,
        children
      })
    });
  });
  MenuTrigger.displayName = TRIGGER_NAME;
  const PORTAL_NAME = "MenuPortal";
  const MenuPortal = props => {
    const {
      scope,
      children,
      ...portalProps
    } = props;
    const context = import_web.isAndroid ? useMenuContext(scope) : null;
    const content = import_web.isAndroid ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuProvider, {
      ...context,
      children
    }) : children;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Menu.Portal, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...portalProps,
      children: content
    });
  };
  MenuPortal.displayName = PORTAL_NAME;
  const CONTENT_NAME = "MenuContent";
  const MenuContent = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...contentProps
    } = props;
    const context = useMenuContext(scope);
    const hasInteractedOutsideRef = React.useRef(false);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Menu.Content, {
      id: context.contentId,
      "aria-labelledby": context.triggerId,
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...contentProps,
      ref: forwardedRef,
      onCloseAutoFocus: (0, import_web.composeEventHandlers)(props.onCloseAutoFocus, event => {
        if (!hasInteractedOutsideRef.current) {
          requestAnimationFrame(() => {
            const activeEl = document.activeElement;
            if (!activeEl || activeEl === document.body) {
              context.triggerRef.current?.focus();
            }
          });
        }
        hasInteractedOutsideRef.current = false;
        event.preventDefault();
      }),
      onInteractOutside: (0, import_web.composeEventHandlers)(props.onInteractOutside, event => {
        const originalEvent = event.detail.originalEvent;
        const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
        const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
        if (!context.modal || isRightClick) hasInteractedOutsideRef.current = true;
      }),
      style: import_web.isWeb ? {
        ...props.style,
        ...{
          "--tamagui-menu-content-transform-origin": "var(--tamagui-popper-transform-origin)",
          "--tamagui-menu-content-available-width": "var(--tamagui-popper-available-width)",
          "--tamagui-menu-content-available-height": "var(--tamagui-popper-available-height)",
          "--tamagui-menu-trigger-width": "var(--tamagui-popper-anchor-width)",
          "--tamagui-menu-trigger-height": "var(--tamagui-popper-anchor-height)"
        }
      } : props.style
    });
  });
  MenuContent.displayName = CONTENT_NAME;
  const DROPDOWN_MENU_SUB_NAME = "MenuSub";
  const MenuSub = props => {
    const {
      scope,
      children,
      open: openProp,
      onOpenChange,
      defaultOpen,
      ...rest
    } = props;
    const [open = false, setOpen] = (0, import_use_controllable_state.useControllableState)({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Menu.Sub, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      open,
      onOpenChange: setOpen,
      ...rest,
      children
    });
  };
  MenuSub.displayName = DROPDOWN_MENU_SUB_NAME;
  const SUB_CONTENT_NAME = "MenuSubContent";
  const MenuSubContent = React.forwardRef((props, forwardedRef) => {
    const {
      scope,
      ...subContentProps
    } = props;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Menu.SubContent, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...subContentProps,
      ref: forwardedRef,
      style: import_web.isWeb ? {
        ...props.style,
        ...{
          "--tamagui-menu-content-transform-origin": "var(--tamagui-popper-transform-origin)",
          "--tamagui-menu-content-available-width": "var(--tamagui-popper-available-width)",
          "--tamagui-menu-content-available-height": "var(--tamagui-popper-available-height)",
          "--tamagui-menu-trigger-width": "var(--tamagui-popper-anchor-width)",
          "--tamagui-menu-trigger-height": "var(--tamagui-popper-anchor-height)"
        }
      } : null
    });
  });
  MenuSubContent.displayName = SUB_CONTENT_NAME;
  const MenuScrollView = (0, import_web.styled)(import_scroll_view.ScrollView, {
    flexShrink: 1,
    alignSelf: "stretch",
    showsHorizontalScrollIndicator: false,
    showsVerticalScrollIndicator: false,
    "$platform-web": {
      maxHeight: "var(--tamagui-menu-content-available-height)"
    }
  });
  const Group = Menu.Group;
  const Label = Menu.Label;
  const Item = Menu.Item;
  const CheckboxItem = Menu.CheckboxItem;
  const RadioGroup = Menu.RadioGroup;
  const RadioItem = Menu.RadioItem;
  const ItemIndicator = Menu.ItemIndicator;
  const Separator = Menu.Separator;
  const Arrow = Menu.Arrow;
  const SubTrigger = Menu.SubTrigger;
  const ItemTitle = Menu.ItemTitle;
  const ItemSubtitle = Menu.ItemSubtitle;
  const ItemImage = Menu.ItemImage;
  const ItemIcon = Menu.ItemIcon;
  return (0, import_web.withStaticProperties)(MenuComp, {
    Root: MenuComp,
    Trigger: MenuTrigger,
    Portal: MenuPortal,
    Content: MenuContent,
    Group,
    Label,
    Item,
    CheckboxItem,
    RadioGroup,
    RadioItem,
    ItemIndicator,
    Separator,
    Arrow,
    Sub: MenuSub,
    SubTrigger,
    SubContent: MenuSubContent,
    ItemTitle,
    ItemSubtitle,
    ItemImage,
    ItemIcon,
    ScrollView: MenuScrollView
  });
}
