// @tamagui/menu/dist/esm/createNonNativeMenu.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createBaseMenu } from "@tamagui/create-menu";
import { usePopperContextSlow } from "@tamagui/popper";
import { ScrollView } from "@tamagui/scroll-view";
import { useControllableState } from "@tamagui/use-controllable-state";
import { composeEventHandlers, composeRefs, createStyledContext, isAndroid, isWeb, Slot, styled, useEvent, useIsTouchDevice, View, withStaticProperties } from "@tamagui/web";
import * as React from "react";
import { useId } from "react";
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
    return !!right[Symbol.hasInstance](left);
  } else {
    return left instanceof right;
  }
}
var DROPDOWN_MENU_CONTEXT = "MenuContext";
function useMenuTriggerSetup(open) {
  var triggerStateSettersRef = React.useRef(/* @__PURE__ */new Map());
  var activeTriggerIdRef = React.useRef(null);
  var setActiveTrigger = useEvent(function (id) {
    var prevId = activeTriggerIdRef.current;
    if (prevId === id) return;
    if (prevId) {
      var _triggerStateSettersRef_current_get;
      (_triggerStateSettersRef_current_get = triggerStateSettersRef.current.get(prevId)) === null || _triggerStateSettersRef_current_get === void 0 ? void 0 : _triggerStateSettersRef_current_get(false);
    }
    activeTriggerIdRef.current = id;
    if (id && open) {
      var _triggerStateSettersRef_current_get1;
      (_triggerStateSettersRef_current_get1 = triggerStateSettersRef.current.get(id)) === null || _triggerStateSettersRef_current_get1 === void 0 ? void 0 : _triggerStateSettersRef_current_get1(true);
    }
  });
  var registerTrigger = useEvent(function (id, setOpenState) {
    triggerStateSettersRef.current.set(id, setOpenState);
    setOpenState(activeTriggerIdRef.current === id && open);
  });
  var unregisterTrigger = useEvent(function (id) {
    triggerStateSettersRef.current.delete(id);
    if (activeTriggerIdRef.current === id) {
      activeTriggerIdRef.current = null;
    }
  });
  React.useEffect(function () {
    if (!open) {
      setActiveTrigger(null);
      return;
    }
    var activeId = activeTriggerIdRef.current;
    if (activeId) {
      var _triggerStateSettersRef_current_get;
      (_triggerStateSettersRef_current_get = triggerStateSettersRef.current.get(activeId)) === null || _triggerStateSettersRef_current_get === void 0 ? void 0 : _triggerStateSettersRef_current_get(true);
    }
  }, [open, setActiveTrigger]);
  return {
    setActiveTrigger,
    registerTrigger,
    unregisterTrigger
  };
}
function createNonNativeMenu(params) {
  var {
    Menu
  } = createBaseMenu(params);
  var DROPDOWN_MENU_NAME = "Menu";
  var {
    Provider: MenuProvider,
    useStyledContext: useMenuContext
  } = createStyledContext();
  var MenuComp = function (props) {
    var {
      scope,
      children,
      dir,
      open: openProp,
      defaultOpen,
      onOpenChange,
      modal = true,
      ...rest
    } = props;
    var triggerRef = React.useRef(null);
    var [open = false, setOpen] = useControllableState({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    var openRef = React.useRef(open);
    openRef.current = open;
    var {
      setActiveTrigger,
      registerTrigger,
      unregisterTrigger
    } = useMenuTriggerSetup(open);
    return /* @__PURE__ */_jsx(MenuProvider, {
      scope,
      triggerId: useId(),
      triggerRef,
      contentId: useId(),
      openRef,
      onOpenChange: React.useCallback(function (nextOpen) {
        return setOpen(nextOpen);
      }, [setOpen]),
      onOpenToggle: React.useCallback(function () {
        return setOpen(function (prevOpen) {
          return !prevOpen;
        });
      }, [setOpen]),
      modal,
      setActiveTrigger,
      registerTrigger,
      unregisterTrigger,
      children: /* @__PURE__ */_jsx(Menu, {
        scope: scope || DROPDOWN_MENU_CONTEXT,
        open,
        onOpenChange: setOpen,
        dir,
        modal,
        ...rest,
        children
      })
    });
  };
  MenuComp.displayName = DROPDOWN_MENU_NAME;
  var TRIGGER_NAME = "MenuTrigger";
  var MenuTriggerFrame = Menu.Anchor;
  var MenuTrigger = View.styleable(function (props, forwardedRef) {
    var {
      scope,
      asChild,
      children,
      disabled = false,
      onKeydown,
      ...triggerProps
    } = props;
    var context = useMenuContext(scope);
    var popperCtx = usePopperContextSlow(scope || DROPDOWN_MENU_CONTEXT);
    var Comp = asChild ? Slot : View;
    var isTouchDevice = useIsTouchDevice();
    var triggerElRef = React.useRef(null);
    var triggerId = React.useId();
    var [triggerOpen, setTriggerOpen] = React.useState(false);
    var {
      registerTrigger,
      unregisterTrigger
    } = context;
    React.useEffect(function () {
      registerTrigger(triggerId, setTriggerOpen);
      return function () {
        return unregisterTrigger(triggerId);
      };
    }, [registerTrigger, unregisterTrigger, triggerId]);
    var activateSelf = React.useCallback(function () {
      context.setActiveTrigger(triggerId);
      var el = triggerElRef.current;
      if (el) {
        context.triggerRef.current = el;
        if (_instanceof(el, HTMLElement)) {
          var _popperCtx_refs;
          (_popperCtx_refs = popperCtx.refs) === null || _popperCtx_refs === void 0 ? void 0 : _popperCtx_refs.setReference(el);
          requestAnimationFrame(function () {
            var _popperCtx_update;
            return (_popperCtx_update = popperCtx.update) === null || _popperCtx_update === void 0 ? void 0 : _popperCtx_update.call(popperCtx);
          });
        }
      }
    }, [context, triggerId, popperCtx]);
    var pressEvent = isWeb ? isTouchDevice ? "onClick" : "onPointerDown" : "onPress";
    return /* @__PURE__ */_jsx(MenuTriggerFrame, {
      asChild: true,
      componentName: TRIGGER_NAME,
      scope: scope || DROPDOWN_MENU_CONTEXT,
      children: /* @__PURE__ */_jsx(Comp, {
        role: "button",
        id: context.triggerId,
        "aria-haspopup": "menu",
        "aria-expanded": triggerOpen,
        "aria-controls": triggerOpen ? context.contentId : void 0,
        "data-state": triggerOpen ? "open" : "closed",
        "data-disabled": disabled ? "" : void 0,
        "aria-disabled": disabled || void 0,
        ref: composeRefs(forwardedRef, context.triggerRef, triggerElRef),
        [pressEvent]: composeEventHandlers(
        //@ts-ignore
        props[pressEvent], function (event) {
          if (!disabled) {
            if (isWeb && _instanceof(event, PointerEvent) && event.button !== 0 && event.ctrlKey === true) return;
            if (context.openRef.current) {
              context.setActiveTrigger(null);
            } else {
              activateSelf();
            }
            context.onOpenToggle();
            if (!context.openRef.current) event.preventDefault();
          }
        }),
        ...(isWeb && {
          onKeyDown: composeEventHandlers(onKeydown, function (event) {
            if (disabled) return;
            if (["Enter", " "].includes(event.key)) {
              if (context.openRef.current) {
                context.setActiveTrigger(null);
              } else {
                activateSelf();
              }
              context.onOpenToggle();
            }
            if (event.key === "ArrowDown") {
              activateSelf();
              context.onOpenChange(true);
            }
            if (["Enter", " ", "ArrowDown"].includes(event.key)) event.preventDefault();
          })
        }),
        ...triggerProps,
        children
      })
    });
  });
  MenuTrigger.displayName = TRIGGER_NAME;
  var PORTAL_NAME = "MenuPortal";
  var MenuPortal = function (props) {
    var {
      scope,
      children,
      ...portalProps
    } = props;
    var context = isAndroid ? useMenuContext(scope) : null;
    var content = isAndroid ? /* @__PURE__ */_jsx(MenuProvider, {
      ...context,
      children
    }) : children;
    return /* @__PURE__ */_jsx(Menu.Portal, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...portalProps,
      children: content
    });
  };
  MenuPortal.displayName = PORTAL_NAME;
  var CONTENT_NAME = "MenuContent";
  var MenuContent = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
    var {
      scope,
      ...contentProps
    } = props;
    var context = useMenuContext(scope);
    var hasInteractedOutsideRef = React.useRef(false);
    return /* @__PURE__ */_jsx(Menu.Content, {
      id: context.contentId,
      "aria-labelledby": context.triggerId,
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...contentProps,
      ref: forwardedRef,
      onCloseAutoFocus: composeEventHandlers(props.onCloseAutoFocus, function (event) {
        if (!hasInteractedOutsideRef.current) {
          requestAnimationFrame(function () {
            var activeEl = document.activeElement;
            if (!activeEl || activeEl === document.body) {
              var _context_triggerRef_current;
              (_context_triggerRef_current = context.triggerRef.current) === null || _context_triggerRef_current === void 0 ? void 0 : _context_triggerRef_current.focus();
            }
          });
        }
        hasInteractedOutsideRef.current = false;
        event.preventDefault();
      }),
      onInteractOutside: composeEventHandlers(props.onInteractOutside, function (event) {
        var originalEvent = event.detail.originalEvent;
        var ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
        var isRightClick = originalEvent.button === 2 || ctrlLeftClick;
        if (!context.modal || isRightClick) hasInteractedOutsideRef.current = true;
      }),
      style: isWeb ? {
        ...props.style,
        ...{
          "--tamagui-menu-content-transform-origin": "var(--tamagui-popper-transform-origin)",
          "--tamagui-menu-content-available-width": "var(--tamagui-popper-available-width)",
          "--tamagui-menu-content-available-height": "var(--tamagui-popper-available-height)",
          "--tamagui-menu-trigger-width": "var(--tamagui-popper-anchor-width)",
          "--tamagui-menu-trigger-height": "var(--tamagui-popper-anchor-height)"
        }
      } : props.style
    });
  });
  MenuContent.displayName = CONTENT_NAME;
  var DROPDOWN_MENU_SUB_NAME = "MenuSub";
  var MenuSub = function (props) {
    var {
      scope,
      children,
      open: openProp,
      onOpenChange,
      defaultOpen,
      ...rest
    } = props;
    var [open = false, setOpen] = useControllableState({
      prop: openProp,
      defaultProp: defaultOpen,
      onChange: onOpenChange
    });
    return /* @__PURE__ */_jsx(Menu.Sub, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      open,
      onOpenChange: setOpen,
      ...rest,
      children
    });
  };
  MenuSub.displayName = DROPDOWN_MENU_SUB_NAME;
  var SUB_CONTENT_NAME = "MenuSubContent";
  var MenuSubContent = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
    var {
      scope,
      ...subContentProps
    } = props;
    return /* @__PURE__ */_jsx(Menu.SubContent, {
      scope: scope || DROPDOWN_MENU_CONTEXT,
      ...subContentProps,
      ref: forwardedRef,
      style: isWeb ? {
        ...props.style,
        ...{
          "--tamagui-menu-content-transform-origin": "var(--tamagui-popper-transform-origin)",
          "--tamagui-menu-content-available-width": "var(--tamagui-popper-available-width)",
          "--tamagui-menu-content-available-height": "var(--tamagui-popper-available-height)",
          "--tamagui-menu-trigger-width": "var(--tamagui-popper-anchor-width)",
          "--tamagui-menu-trigger-height": "var(--tamagui-popper-anchor-height)"
        }
      } : null
    });
  });
  MenuSubContent.displayName = SUB_CONTENT_NAME;
  var MenuScrollView = styled(ScrollView, {
    flexShrink: 1,
    alignSelf: "stretch",
    showsHorizontalScrollIndicator: false,
    showsVerticalScrollIndicator: false,
    "$platform-web": {
      maxHeight: "var(--tamagui-menu-content-available-height)"
    }
  });
  var Group = Menu.Group;
  var Label = Menu.Label;
  var Item = Menu.Item;
  var CheckboxItem = Menu.CheckboxItem;
  var RadioGroup = Menu.RadioGroup;
  var RadioItem = Menu.RadioItem;
  var ItemIndicator = Menu.ItemIndicator;
  var Separator = Menu.Separator;
  var Arrow = Menu.Arrow;
  var SubTrigger = Menu.SubTrigger;
  var ItemTitle = Menu.ItemTitle;
  var ItemSubtitle = Menu.ItemSubtitle;
  var ItemImage = Menu.ItemImage;
  var ItemIcon = Menu.ItemIcon;
  return withStaticProperties(MenuComp, {
    Root: MenuComp,
    Trigger: MenuTrigger,
    Portal: MenuPortal,
    Content: MenuContent,
    Group,
    Label,
    Item,
    CheckboxItem,
    RadioGroup,
    RadioItem,
    ItemIndicator,
    Separator,
    Arrow,
    Sub: MenuSub,
    SubTrigger,
    SubContent: MenuSubContent,
    ItemTitle,
    ItemSubtitle,
    ItemImage,
    ItemIcon,
    ScrollView: MenuScrollView
  });
}
export { DROPDOWN_MENU_CONTEXT, createNonNativeMenu };
