// @tamagui/menu/dist/esm/Menu.native.js
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { createNativeMenu, withNativeMenu } from "@tamagui/create-menu";
import { isWeb, withStaticProperties } from "@tamagui/web";
import React from "react";
import { DROPDOWN_MENU_CONTEXT, createNonNativeMenu } from "./createNonNativeMenu.native.js";
function createMenu(params) {
  var {
    Menu: NativeMenuRoot
  } = createNativeMenu("Menu");
  var NonNativeMenu = createNonNativeMenu(params);
  var COMMON_PARAMS = {
    isRoot: false,
    scope: DROPDOWN_MENU_CONTEXT
  };
  var MenuComp = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Root,
    NativeComponent: NativeMenuRoot,
    isRoot: true
  });
  var Trigger = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Trigger,
    NativeComponent: NativeMenuRoot.Trigger
  });
  var Portal = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Portal,
    NativeComponent: React.Fragment
  });
  var Content = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Content,
    NativeComponent: NativeMenuRoot.Content
  });
  var Group = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Group,
    NativeComponent: NativeMenuRoot.Group
  });
  var Label = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Label,
    NativeComponent: NativeMenuRoot.Label
  });
  var Item = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Item,
    NativeComponent: NativeMenuRoot.Item
  });
  var ItemTitle = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.ItemTitle,
    NativeComponent: NativeMenuRoot.ItemTitle
  });
  var ItemSubtitle = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.ItemSubtitle,
    NativeComponent: NativeMenuRoot.ItemSubtitle
  });
  var ItemIcon = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.ItemIcon,
    NativeComponent: NativeMenuRoot.ItemIcon
  });
  var ItemImage = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.ItemImage,
    NativeComponent: NativeMenuRoot.ItemImage
  });
  var CheckboxItem = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.CheckboxItem,
    NativeComponent: NativeMenuRoot.CheckboxItem
  });
  var RadioGroup = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.RadioGroup,
    NativeComponent: function (param) {
      var {
        children
      } = param;
      return /* @__PURE__ */_jsx(_Fragment, {
        children
      });
    }
  });
  var RadioItem = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.RadioItem,
    NativeComponent: function (param) {
      var {
        children
      } = param;
      return /* @__PURE__ */_jsx(_Fragment, {
        children
      });
    }
  });
  var ItemIndicator = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.ItemIndicator,
    NativeComponent: NativeMenuRoot.ItemIndicator
  });
  var Separator = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Separator,
    NativeComponent: NativeMenuRoot.Separator
  });
  var Arrow = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Arrow,
    NativeComponent: NativeMenuRoot.Arrow
  });
  var Sub = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.Sub,
    NativeComponent: NativeMenuRoot.Sub
  });
  var SubTrigger = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.SubTrigger,
    NativeComponent: NativeMenuRoot.SubTrigger
  });
  var SubContent = withNativeMenu({
    ...COMMON_PARAMS,
    Component: NonNativeMenu.SubContent,
    NativeComponent: NativeMenuRoot.SubContent
  });
  var NativeScrollView = function (param) {
    var {
      children
    } = param;
    return /* @__PURE__ */_jsx(_Fragment, {
      children
    });
  };
  NativeScrollView.displayName = "MenuScrollView";
  var ScrollView = isWeb ? NonNativeMenu.ScrollView : NativeScrollView;
  var Menu = withStaticProperties(MenuComp, {
    Trigger,
    Portal,
    Content,
    Group,
    Label,
    Item,
    CheckboxItem,
    RadioGroup,
    RadioItem,
    ItemIndicator,
    Separator,
    Arrow,
    Sub,
    SubTrigger,
    SubContent,
    ItemTitle,
    ItemSubtitle,
    ItemIcon,
    ItemImage,
    ScrollView
  });
  return Menu;
}
export { createMenu };
