// @tamagui/portal/dist/esm/GorhomPortalItem.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { useEvent } from "@tamagui/core";
import { getPortal, NativePortal } from "@tamagui/native";
import { useEffect, useId } from "react";
import { usePortal } from "./GorhomPortal.native.js";
var GorhomPortalItem = function (props) {
  var {
    name: _providedName,
    hostName = "root",
    handleOnMount: _providedHandleOnMount,
    handleOnUnmount: _providedHandleOnUnmount,
    handleOnUpdate: _providedHandleOnUpdate,
    children,
    passThrough
  } = props;
  var portalState = getPortal().state;
  if (portalState.type === "teleport") {
    if (passThrough) {
      return children;
    }
    return /* @__PURE__ */_jsx(NativePortal, {
      hostName,
      children
    });
  }
  return /* @__PURE__ */_jsx(GorhomPortalItemFallback, {
    name: _providedName,
    hostName,
    handleOnMount: _providedHandleOnMount,
    handleOnUnmount: _providedHandleOnUnmount,
    handleOnUpdate: _providedHandleOnUpdate,
    passThrough,
    children
  });
};
var GorhomPortalItemFallback = function (props) {
  var {
    name: _providedName,
    hostName,
    handleOnMount: _providedHandleOnMount,
    handleOnUnmount: _providedHandleOnUnmount,
    handleOnUpdate: _providedHandleOnUpdate,
    children,
    passThrough
  } = props;
  var {
    addPortal: addUpdatePortal,
    removePortal
  } = usePortal(hostName);
  var id = useId();
  var name = _providedName || id;
  var handleOnMount = useEvent(function () {
    if (_providedHandleOnMount) {
      _providedHandleOnMount(function () {
        return addUpdatePortal(name, children);
      });
    } else {
      addUpdatePortal(name, children);
    }
  });
  var handleOnUnmount = useEvent(function () {
    if (_providedHandleOnUnmount) {
      _providedHandleOnUnmount(function () {
        return removePortal(name);
      });
    } else {
      removePortal(name);
    }
  });
  var handleOnUpdate = useEvent(function () {
    if (_providedHandleOnUpdate) {
      _providedHandleOnUpdate(function () {
        return addUpdatePortal(name, children);
      });
    } else {
      addUpdatePortal(name, children);
    }
  });
  useIsomorphicLayoutEffect(function () {
    if (passThrough) return;
    handleOnMount();
    return function () {
      handleOnUnmount();
    };
  }, []);
  useEffect(function () {
    if (passThrough) return;
    handleOnUpdate();
  }, [children]);
  return passThrough ? children : null;
};
export { GorhomPortalItem };
