// @tamagui/portal/dist/esm/GorhomPortal.native.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { getPortal, NativePortalHost, NativePortalProvider } from "@tamagui/native";
import { startTransition } from "@tamagui/start-transition";
import React, { createContext, memo, useCallback, useContext, useMemo, useReducer } from "react";
import { allPortalHosts, portalListeners } from "./constants.native.js";
var ACTIONS = /* @__PURE__ */function (ACTIONS2) {
  ACTIONS2[ACTIONS2["REGISTER_HOST"] = 0] = "REGISTER_HOST";
  ACTIONS2[ACTIONS2["DEREGISTER_HOST"] = 1] = "DEREGISTER_HOST";
  ACTIONS2[ACTIONS2["ADD_UPDATE_PORTAL"] = 2] = "ADD_UPDATE_PORTAL";
  ACTIONS2[ACTIONS2["REMOVE_PORTAL"] = 3] = "REMOVE_PORTAL";
  return ACTIONS2;
}(ACTIONS || {});
var INITIAL_STATE = {};
var registerHost = function (state, hostName) {
  if (!(hostName in state)) state[hostName] = [];
  return state;
};
var deregisterHost = function (state, hostName) {
  delete state[hostName];
  return state;
};
var addUpdatePortal = function (state, hostName, portalName, node) {
  if (!(hostName in state)) state = registerHost(state, hostName);
  var index = state[hostName].findIndex(function (item) {
    return item.name === portalName;
  });
  if (index !== -1) state[hostName][index].node = node;else state[hostName].push({
    name: portalName,
    node
  });
  return state;
};
var removePortal = function (state, hostName, portalName) {
  if (!(hostName in state)) {
    if (process.env.NODE_ENV === "development") console.info(`Failed to remove portal '${portalName}', '${hostName}' was not registered!`);
    return state;
  }
  var index = state[hostName].findIndex(function (item) {
    return item.name === portalName;
  });
  if (index !== -1) state[hostName].splice(index, 1);
  return state;
};
var reducer = function (state, action) {
  var {
    type
  } = action;
  switch (type) {
    case 0:
      return registerHost({
        ...state
      }, action.hostName);
    case 1:
      return deregisterHost({
        ...state
      }, action.hostName);
    case 2:
      return addUpdatePortal({
        ...state
      }, action.hostName, action.portalName, action.node);
    case 3:
      return removePortal({
        ...state
      }, action.hostName, action.portalName);
    default:
      return state;
  }
};
var PortalStateContext = /* @__PURE__ */createContext(null);
var PortalDispatchContext = /* @__PURE__ */createContext(null);
var PortalProviderActiveContext = /* @__PURE__ */createContext(false);
var usePortalState = function (hostName) {
  var state = useContext(PortalStateContext);
  if (state === null) throw new Error("'PortalStateContext' cannot be null, please add 'PortalProvider' to the root component.");
  return state[hostName] || [];
};
var usePortal = function () {
  var hostName = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "root";
  var dispatch = useContext(PortalDispatchContext);
  if (dispatch === null) throw new Error("'PortalDispatchContext' cannot be null, please add 'PortalProvider' to the root component.");
  var registerHost2 = useCallback(function () {
    dispatch({
      type: 0,
      hostName
    });
  }, []);
  var deregisterHost2 = useCallback(function () {
    dispatch({
      type: 1,
      hostName
    });
  }, []);
  var addUpdatePortal2 = useCallback(function (name, node) {
    dispatch({
      type: 2,
      hostName,
      portalName: name,
      node
    });
  }, []);
  return {
    registerHost: registerHost2,
    deregisterHost: deregisterHost2,
    addPortal: addUpdatePortal2,
    updatePortal: addUpdatePortal2,
    removePortal: useCallback(function (name) {
      dispatch({
        type: 3,
        hostName,
        portalName: name
      });
    }, [])
  };
};
var PortalProviderComponent = function (param) {
  var {
    rootHostName = "root",
    shouldAddRootHost = true,
    children
  } = param;
  var isAlreadyInProvider = useContext(PortalProviderActiveContext);
  if (process.env.NODE_ENV === "development") {
    if (isAlreadyInProvider && shouldAddRootHost) console.warn(`[tamagui] Nested PortalProvider with shouldAddRootHost detected. This causes hydration mismatches. TamaguiProvider from 'tamagui' already includes PortalProvider - remove the explicit PortalProvider wrapper or set shouldAddRootHost={false}.`);
  }
  var [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  var transitionDispatch = useMemo(function () {
    var next = function (value) {
      startTransition(function () {
        dispatch(value);
      });
    };
    return next;
  }, [dispatch]);
  var portalState = getPortal().state;
  var content = /* @__PURE__ */_jsx(PortalProviderActiveContext.Provider, {
    value: true,
    children: /* @__PURE__ */_jsx(PortalDispatchContext.Provider, {
      value: transitionDispatch,
      children: /* @__PURE__ */_jsxs(PortalStateContext.Provider, {
        value: state,
        children: [children, shouldAddRootHost && /* @__PURE__ */_jsx(PortalHost, {
          name: rootHostName
        })]
      })
    })
  });
  if (portalState.type === "teleport") return /* @__PURE__ */_jsx(NativePortalProvider, {
    children: content
  });
  return content;
};
var PortalProvider = /* @__PURE__ */memo(PortalProviderComponent);
PortalProvider.displayName = "PortalProvider";
var defaultRenderer = function (children) {
  return /* @__PURE__ */_jsx(_Fragment, {
    children
  });
};
var PortalHost = /* @__PURE__ */memo(function PortalHost2(props) {
  if (getPortal().state.type === "teleport") return /* @__PURE__ */_jsx(NativePortalHost, {
    name: props.name
  });
  return /* @__PURE__ */_jsx(PortalHostNonNative, {
    ...props
  });
});
function PortalHostNonNative(props) {
  var {
    name,
    forwardProps,
    render = defaultRenderer
  } = props;
  var state = usePortalState(name);
  var {
    registerHost: registerHost2,
    deregisterHost: deregisterHost2
  } = usePortal(props.name);
  useIsomorphicLayoutEffect(function () {
    registerHost2();
    return function () {
      deregisterHost2();
    };
  }, []);
  if (forwardProps) return render(state.map(function (item) {
    var next = item.node;
    var {
      children,
      ...restForwardProps
    } = forwardProps;
    if (forwardProps) return React.Children.map(next, function (child) {
      return /* @__PURE__ */React.isValidElement(child) ? /* @__PURE__ */React.cloneElement(child, {
        key: child.key,
        ...restForwardProps
      }) : child;
    });
    return next;
  }));
  return render(state.map(function (item) {
    return item.node;
  }));
}
export { ACTIONS, INITIAL_STATE, PortalHost, PortalProvider, usePortal };
