// @tamagui/portal/dist/esm/GorhomPortal.mjs
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { getPortal, NativePortalHost, NativePortalProvider } from "@tamagui/native";
import { startTransition } from "@tamagui/start-transition";
import React, { createContext, memo, useCallback, useContext, useMemo, useReducer } from "react";
import { allPortalHosts, portalListeners } from "./constants.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var ACTIONS = /* @__PURE__ */(ACTIONS2 => {
  ACTIONS2[ACTIONS2["REGISTER_HOST"] = 0] = "REGISTER_HOST";
  ACTIONS2[ACTIONS2["DEREGISTER_HOST"] = 1] = "DEREGISTER_HOST";
  ACTIONS2[ACTIONS2["ADD_UPDATE_PORTAL"] = 2] = "ADD_UPDATE_PORTAL";
  ACTIONS2[ACTIONS2["REMOVE_PORTAL"] = 3] = "REMOVE_PORTAL";
  return ACTIONS2;
})(ACTIONS || {});
const INITIAL_STATE = {};
const registerHost = (state, hostName) => {
  if (!(hostName in state)) state[hostName] = [];
  return state;
};
const deregisterHost = (state, hostName) => {
  delete state[hostName];
  return state;
};
const addUpdatePortal = (state, hostName, portalName, node) => {
  if (!(hostName in state)) state = registerHost(state, hostName);
  const index = state[hostName].findIndex(item => item.name === portalName);
  if (index !== -1) state[hostName][index].node = node;else state[hostName].push({
    name: portalName,
    node
  });
  return state;
};
const removePortal = (state, hostName, portalName) => {
  if (!(hostName in state)) {
    if (process.env.NODE_ENV === "development") console.info(`Failed to remove portal '${portalName}', '${hostName}' was not registered!`);
    return state;
  }
  const index = state[hostName].findIndex(item => item.name === portalName);
  if (index !== -1) state[hostName].splice(index, 1);
  return state;
};
const reducer = (state, action) => {
  const {
    type
  } = action;
  switch (type) {
    case 0:
      return registerHost({
        ...state
      }, action.hostName);
    case 1:
      return deregisterHost({
        ...state
      }, action.hostName);
    case 2:
      return addUpdatePortal({
        ...state
      }, action.hostName, action.portalName, action.node);
    case 3:
      return removePortal({
        ...state
      }, action.hostName, action.portalName);
    default:
      return state;
  }
};
const PortalStateContext = createContext(null);
const PortalDispatchContext = createContext(null);
const PortalProviderActiveContext = createContext(false);
const usePortal = (hostName = "root") => {
  const dispatch = useContext(PortalDispatchContext);
  if (dispatch === null) throw new Error("'PortalDispatchContext' cannot be null, please add 'PortalProvider' to the root component.");
  const registerHost2 = useCallback(() => {
    dispatch({
      type: 0,
      hostName
    });
  }, []);
  const deregisterHost2 = useCallback(() => {
    dispatch({
      type: 1,
      hostName
    });
  }, []);
  const addUpdatePortal2 = useCallback((name, node) => {
    dispatch({
      type: 2,
      hostName,
      portalName: name,
      node
    });
  }, []);
  return {
    registerHost: registerHost2,
    deregisterHost: deregisterHost2,
    addPortal: addUpdatePortal2,
    updatePortal: addUpdatePortal2,
    removePortal: useCallback(name => {
      dispatch({
        type: 3,
        hostName,
        portalName: name
      });
    }, [])
  };
};
const PortalProviderComponent = ({
  rootHostName = "root",
  shouldAddRootHost = true,
  children
}) => {
  const isAlreadyInProvider = useContext(PortalProviderActiveContext);
  if (process.env.NODE_ENV === "development") {
    if (isAlreadyInProvider && shouldAddRootHost) console.warn(`[tamagui] Nested PortalProvider with shouldAddRootHost detected. This causes hydration mismatches. TamaguiProvider from 'tamagui' already includes PortalProvider - remove the explicit PortalProvider wrapper or set shouldAddRootHost={false}.`);
  }
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const transitionDispatch = useMemo(() => {
    const next = value => {
      startTransition(() => {
        dispatch(value);
      });
    };
    return next;
  }, [dispatch]);
  const portalState = getPortal().state;
  const content = /* @__PURE__ */jsx(PortalProviderActiveContext.Provider, {
    value: true,
    children: /* @__PURE__ */jsx(PortalDispatchContext.Provider, {
      value: transitionDispatch,
      children: /* @__PURE__ */jsxs(PortalStateContext.Provider, {
        value: state,
        children: [children, shouldAddRootHost && /* @__PURE__ */jsx(PortalHost, {
          name: rootHostName
        })]
      })
    })
  });
  if (portalState.type === "teleport") return /* @__PURE__ */jsx(NativePortalProvider, {
    children: content
  });
  return content;
};
const PortalProvider = memo(PortalProviderComponent);
PortalProvider.displayName = "PortalProvider";
const PortalHost = memo(function PortalHost2(props) {
  return /* @__PURE__ */jsx(PortalHostWeb, {
    ...props
  });
});
function PortalHostWeb(props) {
  useIsomorphicLayoutEffect(() => {
    return () => {
      allPortalHosts.delete(props.name);
    };
  }, [props.name]);
  return /* @__PURE__ */jsx("div", {
    style: {
      display: "contents"
    },
    ref: node => {
      if (node) {
        allPortalHosts.set(props.name, node);
        portalListeners[props.name]?.forEach(x => x(node));
      }
    }
  });
}
export { ACTIONS, INITIAL_STATE, PortalHost, PortalProvider, usePortal };
