// @tamagui/portal/dist/esm/Portal.mjs
import { TamaguiRoot, useDidFinishSSR, useThemeName } from "@tamagui/web";
import { useStackedZIndex, ZIndexHardcodedContext } from "@tamagui/z-index-stack";
import * as React from "react";
import { createPortal } from "react-dom";
import { getStackedZIndexProps } from "./helpers.mjs";
import { jsx } from "react/jsx-runtime";
const Portal = React.memo(propsIn => {
  const {
    children,
    passThrough,
    style,
    open
  } = propsIn;
  const themeName = useThemeName();
  const didHydrate = useDidFinishSSR();
  const zIndex = useStackedZIndex(getStackedZIndexProps(propsIn));
  if (passThrough) {
    return children;
  }
  if (!didHydrate) {
    return null;
  }
  return createPortal(/* @__PURE__ */jsx(TamaguiRoot, {
    theme: themeName,
    style: {
      zIndex,
      position: "fixed",
      inset: 0,
      contain: "strict",
      pointerEvents: open ? "auto" : "none",
      // prevent mobile browser from scrolling/moving this fixed element
      touchAction: "none",
      display: "flex",
      ...style
    },
    children: /* @__PURE__ */jsx(ZIndexHardcodedContext.Provider, {
      value: zIndex,
      children
    })
  }), globalThis.document?.body);
});
export { Portal };
