// @tamagui/constants/dist/esm/constants.mjs
import { useEffect, useLayoutEffect } from "react";
const isWeb = true;
const isBrowser = typeof window !== "undefined";
const isServer = !isBrowser;
const isClient = isBrowser;
const isWindowDefined = isBrowser;
const useIsomorphicLayoutEffect = isServer ? useEffect : useLayoutEffect;
const isChrome = typeof navigator !== "undefined" && /Chrome/.test(navigator.userAgent || "");
const isWebTouchable = isClient && ("ontouchstart" in window || navigator.maxTouchPoints > 0);
const isTouchable = isWebTouchable;
const isAndroid = process.env.TEST_NATIVE_PLATFORM === "android" || process.env.TEST_NATIVE_PLATFORM === "androidtv";
const isIos = process.env.TEST_NATIVE_PLATFORM === "ios" || process.env.TEST_NATIVE_PLATFORM === "tvos";
const isTV = process.env.TEST_NATIVE_PLATFORM === "androidtv" || process.env.TEST_NATIVE_PLATFORM === "tvos";
const currentPlatform = "web";
export { currentPlatform, isAndroid, isBrowser, isChrome, isClient, isIos, isServer, isTV, isTouchable, isWeb, isWebTouchable, isWindowDefined, useIsomorphicLayoutEffect };
