// @tamagui/constants/dist/esm/constants.native.js
import { useLayoutEffect } from "react";
import { Platform } from "react-native";
var isWeb = false;
var isBrowser = false;
var isServer = false;
var isClient = true;
var isWindowDefined = false;
var useIsomorphicLayoutEffect = useLayoutEffect;
var isChrome = false;
var isWebTouchable = false;
var isTouchable = true;
var isAndroid = (Platform === null || Platform === void 0 ? void 0 : Platform.OS) === "android" || process.env.TEST_NATIVE_PLATFORM === "android" || process.env.TEST_NATIVE_PLATFORM === "androidtv";
var isIos = (Platform === null || Platform === void 0 ? void 0 : Platform.OS) === "ios" || process.env.TEST_NATIVE_PLATFORM === "ios" || process.env.TEST_NATIVE_PLATFORM === "tvos";
var isTV = (Platform === null || Platform === void 0 ? void 0 : Platform.isTV) || process.env.TEST_NATIVE_PLATFORM === "androidtv" || process.env.TEST_NATIVE_PLATFORM === "tvos";
var currentPlatform = ((Platform === null || Platform === void 0 ? void 0 : Platform.OS) ? {
  ios: "ios",
  android: "android"
}[Platform.OS] : void 0) || "native";
export { currentPlatform, isAndroid, isBrowser, isChrome, isClient, isIos, isServer, isTV, isTouchable, isWeb, isWebTouchable, isWindowDefined, useIsomorphicLayoutEffect };
