// @tamagui/animations-react-native/dist/cjs/createAnimations.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createAnimations_exports = {};
__export(createAnimations_exports, {
  AnimatedText: () => AnimatedText,
  AnimatedView: () => AnimatedView,
  createAnimations: () => createAnimations,
  useAnimatedNumber: () => useAnimatedNumber,
  useAnimatedNumberReaction: () => useAnimatedNumberReaction,
  useAnimatedNumberStyle: () => useAnimatedNumberStyle,
  useAnimatedNumbersStyle: () => useAnimatedNumbersStyle
});
module.exports = __toCommonJS(createAnimations_exports);
var import_animation_helpers = require("@tamagui/animation-helpers");
var import_constants = require("@tamagui/constants");
var import_use_presence = require("@tamagui/use-presence");
var import_web = require("@tamagui/web");
var import_react = __toESM(require("react"), 1);
var import_react_native = require("react-native");
function _type_of(obj) {
  "@swc/helpers - typeof";

  return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var isFabric = !import_constants.isWeb && typeof global !== "undefined" && !!global.__nativeFabricUIManager;
var resolveDynamicValue = function (value, isDark) {
  if (value && (typeof value === "undefined" ? "undefined" : _type_of(value)) === "object" && "dynamic" in value) {
    var dynamicValue = isDark ? value.dynamic.dark : value.dynamic.light;
    return dynamicValue;
  }
  return value;
};
var animatedStyleKey = {
  transform: true,
  opacity: true
};
var colorStyleKey = {
  backgroundColor: true,
  color: true,
  borderColor: true,
  borderLeftColor: true,
  borderRightColor: true,
  borderTopColor: true,
  borderBottomColor: true
};
var costlyToAnimateStyleKey = {
  borderRadius: true,
  borderTopLeftRadius: true,
  borderTopRightRadius: true,
  borderBottomLeftRadius: true,
  borderBottomRightRadius: true,
  borderWidth: true,
  borderLeftWidth: true,
  borderRightWidth: true,
  borderTopWidth: true,
  borderBottomWidth: true,
  ...colorStyleKey
};
var AnimatedView = import_react_native.Animated.View;
var AnimatedText = import_react_native.Animated.Text;
function useAnimatedNumber(initial) {
  var state = import_react.default.useRef(null);
  if (!state.current) {
    state.current = {
      composite: null,
      val: new import_react_native.Animated.Value(initial),
      strategy: {
        type: "spring"
      }
    };
  }
  return {
    getInstance() {
      return state.current.val;
    },
    getValue() {
      return state.current.val["_value"];
    },
    stop() {
      var _state_current_composite;
      (_state_current_composite = state.current.composite) === null || _state_current_composite === void 0 ? void 0 : _state_current_composite.stop();
      state.current.composite = null;
    },
    setValue(next) {
      var {
          type,
          ...config
        } = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {
          type: "spring"
        },
        onFinish = arguments.length > 2 ? arguments[2] : void 0;
      var val = state.current.val;
      var handleFinish = onFinish ? function (param) {
        var {
          finished
        } = param;
        return finished ? onFinish() : null;
      } : void 0;
      if (type === "direct") {
        val.setValue(next);
      } else if (type === "spring") {
        var _state_current_composite;
        (_state_current_composite = state.current.composite) === null || _state_current_composite === void 0 ? void 0 : _state_current_composite.stop();
        var composite = import_react_native.Animated.spring(val, {
          ...config,
          toValue: next,
          useNativeDriver: isFabric
        });
        composite.start(handleFinish);
        state.current.composite = composite;
      } else {
        var _state_current_composite1;
        (_state_current_composite1 = state.current.composite) === null || _state_current_composite1 === void 0 ? void 0 : _state_current_composite1.stop();
        var composite1 = import_react_native.Animated.timing(val, {
          ...config,
          toValue: next,
          useNativeDriver: isFabric
        });
        composite1.start(handleFinish);
        state.current.composite = composite1;
      }
    }
  };
}
var useAnimatedNumberReaction = function (param, onValue) {
  var {
    value
  } = param;
  var onChange = (0, import_web.useEvent)(function (current) {
    onValue(current.value);
  });
  import_react.default.useEffect(function () {
    var id = value.getInstance().addListener(onChange);
    return function () {
      value.getInstance().removeListener(id);
    };
  }, [value, onChange]);
};
var useAnimatedNumberStyle = function (value, getStyle) {
  return getStyle(value.getInstance());
};
var useAnimatedNumbersStyle = function (vals, getStyle) {
  return getStyle(...vals.map(function (v) {
    return v.getInstance();
  }));
};
function createAnimations(animations, options) {
  var _options_useNativeDriver;
  var nativeDriver = (_options_useNativeDriver = options === null || options === void 0 ? void 0 : options.useNativeDriver) !== null && _options_useNativeDriver !== void 0 ? _options_useNativeDriver : isFabric;
  return {
    isReactNative: true,
    inputStyle: "value",
    outputStyle: "inline",
    avoidReRenders: true,
    animations,
    needsCustomComponent: true,
    View: AnimatedView,
    Text: AnimatedText,
    useAnimatedNumber,
    useAnimatedNumberReaction,
    useAnimatedNumberStyle,
    useAnimatedNumbersStyle,
    usePresence: import_use_presence.usePresence,
    ResetPresence: import_use_presence.ResetPresence,
    useAnimations: function (param) {
      var {
        props,
        onDidAnimate,
        style,
        componentState,
        presence,
        useStyleEmitter
      } = param;
      var _themeState_name;
      var isDisabled = import_constants.isWeb && componentState.unmounted === true;
      var isExiting = (presence === null || presence === void 0 ? void 0 : presence[0]) === false;
      var sendExitComplete = presence === null || presence === void 0 ? void 0 : presence[1];
      var [, themeState] = (0, import_web.useThemeWithState)({});
      var isDark = (themeState === null || themeState === void 0 ? void 0 : themeState.scheme) === "dark" || (themeState === null || themeState === void 0 ? void 0 : (_themeState_name = themeState.name) === null || _themeState_name === void 0 ? void 0 : _themeState_name.startsWith("dark"));
      var animateStyles = import_react.default.useRef({});
      var animatedTranforms = import_react.default.useRef([]);
      var animationsState = import_react.default.useRef(/* @__PURE__ */new WeakMap());
      var exitCycleIdRef = import_react.default.useRef(0);
      var exitCompletedRef = import_react.default.useRef(false);
      var wasExitingRef = import_react.default.useRef(false);
      var justStartedExiting = isExiting && !wasExitingRef.current;
      var justStoppedExiting = !isExiting && wasExitingRef.current;
      if (justStartedExiting) {
        exitCycleIdRef.current++;
        exitCompletedRef.current = false;
      }
      if (justStoppedExiting) {
        exitCycleIdRef.current++;
      }
      var animateOnly = props.animateOnly || [];
      var hasTransitionOnly = !!props.animateOnly;
      var isEntering = !!componentState.unmounted;
      var wasEnteringRef = import_react.default.useRef(isEntering);
      var justFinishedEntering = wasEnteringRef.current && !isEntering;
      import_react.default.useEffect(function () {
        wasEnteringRef.current = isEntering;
      });
      var args = [JSON.stringify(style), componentState, isExiting, !!onDidAnimate, isDark, justFinishedEntering, hasTransitionOnly];
      var res = import_react.default.useMemo(function () {
        var runners = [];
        var completions = [];
        var animationState = isExiting ? "exit" : isEntering || justFinishedEntering ? "enter" : "default";
        var nonAnimatedStyle = {};
        for (var key in style) {
          var rawVal = style[key];
          var val = resolveDynamicValue(rawVal, isDark);
          if (val === void 0) continue;
          if (isDisabled) {
            continue;
          }
          if (animatedStyleKey[key] == null && !costlyToAnimateStyleKey[key]) {
            nonAnimatedStyle[key] = val;
            continue;
          }
          if (hasTransitionOnly && !animateOnly.includes(key)) {
            nonAnimatedStyle[key] = val;
            continue;
          }
          if (key !== "transform") {
            animateStyles.current[key] = update(key, animateStyles.current[key], val);
            continue;
          }
          if (!val) continue;
          if (typeof val === "string") {
            console.warn(`Warning: Tamagui can't animate string transforms yet!`);
            continue;
          }
          var _iteratorNormalCompletion = true,
            _didIteratorError = false,
            _iteratorError = void 0;
          try {
            for (var _iterator = val.entries()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
              var [index, transform] = _step.value;
              var _animatedTranforms_current_index;
              if (!transform) continue;
              var tkey = Object.keys(transform)[0];
              var currentTransform = (_animatedTranforms_current_index = animatedTranforms.current[index]) === null || _animatedTranforms_current_index === void 0 ? void 0 : _animatedTranforms_current_index[tkey];
              animatedTranforms.current[index] = {
                [tkey]: update(tkey, currentTransform, transform[tkey])
              };
              animatedTranforms.current = [...animatedTranforms.current];
            }
          } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
          } finally {
            try {
              if (!_iteratorNormalCompletion && _iterator.return != null) {
                _iterator.return();
              }
            } finally {
              if (_didIteratorError) {
                throw _iteratorError;
              }
            }
          }
        }
        var animatedTransformStyle = animatedTranforms.current.length > 0 ? {
          transform: animatedTranforms.current.map(function (r) {
            var _animationsState_current_get;
            var key2 = Object.keys(r)[0];
            var val2 = ((_animationsState_current_get = animationsState.current.get(r[key2])) === null || _animationsState_current_get === void 0 ? void 0 : _animationsState_current_get.interpolation) || r[key2];
            return {
              [key2]: val2
            };
          })
        } : {};
        var animatedStyle = {
          ...Object.fromEntries(Object.entries(animateStyles.current).map(function (param2) {
            var [k, v] = param2;
            var _animationsState_current_get;
            return [k, ((_animationsState_current_get = animationsState.current.get(v)) === null || _animationsState_current_get === void 0 ? void 0 : _animationsState_current_get.interpolation) || v];
          })),
          ...animatedTransformStyle
        };
        return {
          runners,
          completions,
          style: [nonAnimatedStyle, animatedStyle]
        };
        function update(key2, animated, valIn) {
          var isColorStyleKey = colorStyleKey[key2];
          var [val2, type] = isColorStyleKey ? [0, void 0] : getValue(valIn);
          var animateToValue = val2;
          var value = animated || new import_react_native.Animated.Value(val2);
          var curInterpolation = animationsState.current.get(value);
          var interpolateArgs;
          if (type) {
            var _curInterpolation_current;
            interpolateArgs = getInterpolated((_curInterpolation_current = curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.current) !== null && _curInterpolation_current !== void 0 ? _curInterpolation_current : value["_value"], val2, type);
            animationsState.current.set(value, {
              interpolation: value.interpolate(interpolateArgs),
              current: val2
            });
          }
          if (isColorStyleKey) {
            animateToValue = (curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.animateToValue) ? 0 : 1;
            interpolateArgs = getColorInterpolated(curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.current,
            // valIn is the next color
            valIn, animateToValue);
            animationsState.current.set(value, {
              current: valIn,
              interpolation: value.interpolate(interpolateArgs),
              animateToValue: (curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.animateToValue) ? 0 : 1
            });
          }
          if (value) {
            var animationConfig = getAnimationConfig(key2, animations, props.transition, animationState);
            var resolve;
            var promise = new Promise(function (res2) {
              resolve = res2;
            });
            completions.push(promise);
            runners.push(function () {
              value.stopAnimation();
              function getAnimation() {
                return import_react_native.Animated[animationConfig.type || "spring"](value, {
                  toValue: animateToValue,
                  useNativeDriver: nativeDriver,
                  ...animationConfig
                });
              }
              var animation = animationConfig.delay ? import_react_native.Animated.sequence([import_react_native.Animated.delay(animationConfig.delay), getAnimation()]) : getAnimation();
              animation.start(function (param2) {
                var {
                  finished
                } = param2;
                if (finished || isExiting) {
                  resolve();
                }
              });
            });
          }
          if (process.env.NODE_ENV === "development") {
            if (props["debug"] === "verbose") {
              console.info(" \u{1F4A0} animate", key2, `from (${value["_value"]}) to`, valIn, `(${val2})`, "type", type, "interpolate", interpolateArgs);
            }
          }
          return value;
        }
      }, args);
      import_react.default.useEffect(function () {
        wasExitingRef.current = isExiting;
      });
      (0, import_constants.useIsomorphicLayoutEffect)(function () {
        res.runners.forEach(function (r) {
          return r();
        });
        var cycleId = exitCycleIdRef.current;
        if (res.completions.length === 0) {
          onDidAnimate === null || onDidAnimate === void 0 ? void 0 : onDidAnimate();
          if (isExiting && !exitCompletedRef.current) {
            exitCompletedRef.current = true;
            sendExitComplete === null || sendExitComplete === void 0 ? void 0 : sendExitComplete();
          }
          return;
        }
        var cancel = false;
        Promise.all(res.completions).then(function () {
          if (cancel) return;
          if (isExiting && cycleId !== exitCycleIdRef.current) return;
          if (isExiting && exitCompletedRef.current) return;
          onDidAnimate === null || onDidAnimate === void 0 ? void 0 : onDidAnimate();
          if (isExiting) {
            exitCompletedRef.current = true;
            sendExitComplete === null || sendExitComplete === void 0 ? void 0 : sendExitComplete();
          }
        });
        return function () {
          cancel = true;
        };
      }, args);
      useStyleEmitter === null || useStyleEmitter === void 0 ? void 0 : useStyleEmitter(function (nextStyle) {
        for (var key in nextStyle) {
          var rawVal = nextStyle[key];
          var val = resolveDynamicValue(rawVal, isDark);
          if (val === void 0) continue;
          if (key === "transform" && Array.isArray(val)) {
            var _iteratorNormalCompletion = true,
              _didIteratorError = false,
              _iteratorError = void 0;
            try {
              for (var _iterator = val.entries()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                var [index, transform] = _step.value;
                var _animatedTranforms_current_index;
                if (!transform) continue;
                var tkey = Object.keys(transform)[0];
                var currentTransform = (_animatedTranforms_current_index = animatedTranforms.current[index]) === null || _animatedTranforms_current_index === void 0 ? void 0 : _animatedTranforms_current_index[tkey];
                animatedTranforms.current[index] = {
                  [tkey]: update(tkey, currentTransform, transform[tkey])
                };
              }
            } catch (err) {
              _didIteratorError = true;
              _iteratorError = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion && _iterator.return != null) {
                  _iterator.return();
                }
              } finally {
                if (_didIteratorError) {
                  throw _iteratorError;
                }
              }
            }
          } else if (animatedStyleKey[key] != null || costlyToAnimateStyleKey[key]) {
            animateStyles.current[key] = update(key, animateStyles.current[key], val);
          }
        }
        res.runners.forEach(function (r) {
          return r();
        });
        function update(key2, animated, valIn) {
          var isColor = colorStyleKey[key2];
          var [numVal, type] = isColor ? [0, void 0] : getValue(valIn);
          var animateToValue = numVal;
          var value = animated || new import_react_native.Animated.Value(numVal);
          var curInterpolation = animationsState.current.get(value);
          if (type) {
            var _curInterpolation_current;
            animationsState.current.set(value, {
              interpolation: value.interpolate(getInterpolated((_curInterpolation_current = curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.current) !== null && _curInterpolation_current !== void 0 ? _curInterpolation_current : value["_value"], numVal, type)),
              current: numVal
            });
          }
          if (isColor) {
            animateToValue = (curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.animateToValue) ? 0 : 1;
            animationsState.current.set(value, {
              current: valIn,
              interpolation: value.interpolate(getColorInterpolated(curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.current, valIn, animateToValue)),
              animateToValue: (curInterpolation === null || curInterpolation === void 0 ? void 0 : curInterpolation.animateToValue) ? 0 : 1
            });
          }
          var animationConfig = getAnimationConfig(key2, animations, props.transition, "default");
          res.runners.push(function () {
            value.stopAnimation();
            var anim = import_react_native.Animated[animationConfig.type || "spring"](value, {
              toValue: animateToValue,
              useNativeDriver: nativeDriver,
              ...animationConfig
            });
            (animationConfig.delay ? import_react_native.Animated.sequence([import_react_native.Animated.delay(animationConfig.delay), anim]) : anim).start();
          });
          return value;
        }
      });
      if (process.env.NODE_ENV === "development") {
        if (props["debug"] === "verbose") {
          console.info(`Animated`, {
            response: res,
            inputStyle: style,
            isExiting
          });
        }
      }
      return res;
    }
  };
}
function getColorInterpolated(currentColor, nextColor, animateToValue) {
  var inputRange = [0, 1];
  var outputRange = [currentColor ? currentColor : nextColor, nextColor];
  if (animateToValue === 0) {
    outputRange.reverse();
  }
  return {
    inputRange,
    outputRange
  };
}
function getInterpolated(current, next) {
  var postfix = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : "deg";
  if (next === current) {
    current = next - 1e-9;
  }
  var inputRange = [current, next];
  var outputRange = [`${current}${postfix}`, `${next}${postfix}`];
  if (next < current) {
    inputRange.reverse();
    outputRange.reverse();
  }
  return {
    inputRange,
    outputRange
  };
}
function getAnimationConfig(key, animations, transition) {
  var animationState = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : "default";
  var normalized = (0, import_animation_helpers.normalizeTransition)(transition);
  var shortKey = transformShorthands[key];
  var _normalized_properties_key;
  var propAnimation = (_normalized_properties_key = normalized.properties[key]) !== null && _normalized_properties_key !== void 0 ? _normalized_properties_key : normalized.properties[shortKey];
  var animationType = null;
  var extraConf = {};
  if (typeof propAnimation === "string") {
    animationType = propAnimation;
  } else if (propAnimation && (typeof propAnimation === "undefined" ? "undefined" : _type_of(propAnimation)) === "object") {
    animationType = propAnimation.type || (0, import_animation_helpers.getEffectiveAnimation)(normalized, animationState);
    extraConf = propAnimation;
  } else {
    animationType = (0, import_animation_helpers.getEffectiveAnimation)(normalized, animationState);
  }
  if (normalized.delay && !extraConf.delay) {
    extraConf = {
      ...extraConf,
      delay: normalized.delay
    };
  }
  var found = animationType ? animations[animationType] : {};
  return {
    ...found,
    // Apply global spring config overrides (from transition={['bouncy', { stiffness: 1000 }]})
    ...normalized.config,
    // Property-specific config takes highest precedence
    ...extraConf
  };
}
var transformShorthands = {
  x: "translateX",
  y: "translateY",
  translateX: "x",
  translateY: "y"
};
function getValue(input) {
  var isColor = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : false;
  if (typeof input !== "string") {
    return [input];
  }
  var _input_match;
  var [_, number, after] = (_input_match = input.match(/([-0-9]+)(deg|%|px)/)) !== null && _input_match !== void 0 ? _input_match : [];
  return [+number, after];
}
