// @tamagui/button/dist/jsx/v1/Button.mjs
import { getFontSize } from "@tamagui/font-size";
import { getButtonSized } from "@tamagui/get-button-sized";
import { withStaticProperties } from "@tamagui/helpers";
import { useGetThemedIcon } from "@tamagui/helpers-tamagui";
import { ButtonNestingContext, ThemeableStack } from "@tamagui/stacks";
import { SizableText, wrapChildrenInText } from "@tamagui/text";
import { createStyledContext, getVariableValue, styled, useProps } from "@tamagui/web";
import { useContext } from "react";
import { spacedChildren } from "@tamagui/spacer";
import { jsx } from "react/jsx-runtime";
const ButtonContext = createStyledContext({
  // keeping these here means they work with styled() passing down color to text
  color: void 0,
  ellipsis: void 0,
  fontFamily: void 0,
  fontSize: void 0,
  fontStyle: void 0,
  fontWeight: void 0,
  letterSpacing: void 0,
  maxFontSizeMultiplier: void 0,
  size: void 0,
  textAlign: void 0,
  variant: void 0
});
const BUTTON_NAME = "Button";
const ButtonFrame = styled(ThemeableStack, {
  name: BUTTON_NAME,
  render: "button",
  context: ButtonContext,
  role: "button",
  focusable: true,
  variants: {
    unstyled: {
      false: {
        size: "$true",
        justifyContent: "center",
        alignItems: "center",
        flexWrap: "nowrap",
        flexDirection: "row",
        cursor: "pointer",
        hoverTheme: true,
        pressTheme: true,
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "transparent",
        focusVisibleStyle: {
          outlineColor: "$outlineColor",
          outlineStyle: "solid",
          outlineWidth: 2
        }
      }
    },
    variant: {
      outlined: {
        backgroundColor: "transparent",
        borderWidth: 2,
        borderColor: "$borderColor",
        hoverStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorHover"
        },
        pressStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorPress"
        },
        focusVisibleStyle: {
          backgroundColor: "transparent",
          borderColor: "$borderColorFocus"
        }
      }
    },
    size: {
      "...size": getButtonSized,
      ":number": getButtonSized
    },
    disabled: {
      true: {
        pointerEvents: "none"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const ButtonText = styled(SizableText, {
  name: "Button",
  context: ButtonContext,
  variants: {
    unstyled: {
      false: {
        userSelect: "none",
        cursor: "pointer",
        // flexGrow 1 leads to inconsistent native style where text pushes to start of view
        flexGrow: 0,
        flexShrink: 1,
        ellipsis: true,
        color: "$color"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const ButtonIcon = props => {
  const {
    children,
    scaleIcon = 1
  } = props;
  const {
    size,
    color
  } = useContext(ButtonContext);
  const iconSize = (typeof size === "number" ? size * 0.5 : getFontSize(size)) * scaleIcon;
  const getThemedIcon = useGetThemedIcon({
    size: iconSize,
    color
  });
  return getThemedIcon(children);
};
const ButtonComponent = ButtonFrame.styleable(function Button(props, ref) {
  const {
    props: buttonProps
  } = useButton(props);
  return /* @__PURE__ */jsx(ButtonFrame, {
    "data-disable-theme": true,
    ...buttonProps,
    ref
  });
});
const Button2 = withStaticProperties(ButtonComponent, {
  Text: ButtonText,
  Icon: ButtonIcon
});
function useButton({
  textProps,
  ...propsIn
}, {
  Text = Button2.Text
} = {
  Text: Button2.Text
}) {
  const isNested = useContext(ButtonNestingContext);
  const propsActive = useProps(propsIn, {
    noNormalize: true,
    noExpand: true
  });
  const {
    icon,
    iconAfter,
    gap,
    spaceFlex,
    scaleIcon = 1,
    scaleSpace = 0.66,
    noTextWrap,
    fontFamily,
    fontSize,
    fontWeight,
    fontStyle,
    letterSpacing,
    render,
    ellipsis,
    maxFontSizeMultiplier,
    ...restProps
  } = propsActive;
  const size = propsActive.size || (propsActive.unstyled ? void 0 : "$true");
  const color = propsActive.color;
  const iconSize = (typeof size === "number" ? size * 0.5 : getFontSize(size, {
    font: fontFamily?.[0] === "$" ? fontFamily : void 0
  })) * scaleIcon;
  const getThemedIcon = useGetThemedIcon({
    size: iconSize,
    color
  });
  const [themedIcon, themedIconAfter] = [icon, iconAfter].map(getThemedIcon);
  const spaceSize = gap ?? getVariableValue(iconSize) * scaleSpace;
  const contents = noTextWrap ? [propsIn.children] : wrapChildrenInText(Text, {
    children: propsIn.children,
    fontFamily,
    fontSize,
    textProps,
    fontWeight,
    fontStyle,
    letterSpacing,
    ellipsis,
    maxFontSizeMultiplier
  }, Text === ButtonText && propsActive.unstyled !== true ? {
    unstyled: process.env.TAMAGUI_HEADLESS === "1",
    size
  } : void 0);
  const inner = spacedChildren({
    // a bit arbitrary but scaling to font size is necessary so long as button does
    space: spaceSize,
    spaceFlex,
    ensureKeys: true,
    direction: propsActive.flexDirection === "column" || propsActive.flexDirection === "column-reverse" ? "vertical" : "horizontal",
    // for keys to stay the same we keep indices as similar a possible
    // so even if icons are undefined we still pass them
    children: [themedIcon, ...contents, themedIconAfter]
  });
  const props = {
    size,
    ...(propsIn.disabled && {
      // in rnw - false still has keyboard tabIndex, undefined = not actually focusable
      focusable: void 0,
      // even with tabIndex unset, it will keep focusVisibleStyle on web so disable it here
      focusVisibleStyle: {
        borderColor: "$background"
      }
    }),
    // fixes SSR issue + DOM nesting issue of not allowing button in button
    render: render ?? (isNested ? "span" :
    // defaults to <a /> when accessibilityRole = link
    // see https://github.com/tamagui/tamagui/issues/505
    propsActive.accessibilityRole === "link" || propsActive.role === "link" ? "a" : "button"),
    ...restProps,
    children: /* @__PURE__ */jsx(ButtonNestingContext.Provider, {
      value: true,
      children: inner
    }),
    // forces it to be a runtime pressStyle so it passes through context text colors
    disableClassName: true
  };
  return {
    spaceSize,
    isNested,
    props
  };
}
export { Button2 as Button, ButtonContext, ButtonFrame, ButtonIcon, ButtonText, useButton };
