// @tamagui/create-menu/dist/cjs/createBaseMenu.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createBaseMenu_exports = {};
__export(createBaseMenu_exports, {
  createBaseMenu: () => createBaseMenu
});
module.exports = __toCommonJS(createBaseMenu_exports);
var import_animate = require("@tamagui/animate");
var import_animate_presence = require("@tamagui/animate-presence");
var import_collection = require("@tamagui/collection");
var import_dismissable = require("@tamagui/dismissable");
var import_focus_guard = require("@tamagui/focus-guard");
var import_focus_scope = require("@tamagui/focus-scope");
var PopperPrimitive = __toESM(require("@tamagui/popper"), 1);
var import_portal = require("@tamagui/portal");
var import_remove_scroll = require("@tamagui/remove-scroll");
var import_roving_focus = require("@tamagui/roving-focus");
var import_use_callback_ref = require("@tamagui/use-callback-ref");
var import_use_direction = require("@tamagui/use-direction");
var import_web = require("@tamagui/web");
var React = __toESM(require("react"), 1);
var import_react = require("react");
var import_MenuPredefined = require("./MenuPredefined.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function whenMouse(handler) {
  return event => event.pointerType === "mouse" ? handler(event) : void 0;
}
const SELECTION_KEYS = ["Enter", " "];
const FIRST_KEYS = ["ArrowDown", "PageUp", "Home"];
const LAST_KEYS = ["ArrowUp", "PageDown", "End"];
const FIRST_LAST_KEYS = [...FIRST_KEYS, ...LAST_KEYS];
const SUB_OPEN_KEYS = {
  ltr: [...SELECTION_KEYS, "ArrowRight"],
  rtl: [...SELECTION_KEYS, "ArrowLeft"]
};
const SUB_CLOSE_KEYS = {
  ltr: ["ArrowLeft"],
  rtl: ["ArrowRight"]
};
const MENU_NAME = "Menu";
const [Collection, useCollection] = (0, import_collection.createCollection)(MENU_NAME);
const {
  Provider: MenuProvider,
  useStyledContext: useMenuContext
} = (0, import_web.createStyledContext)();
const {
  Provider: MenuRootProvider,
  useStyledContext: useMenuRootContext
} = (0, import_web.createStyledContext)();
const MENU_CONTEXT = "MenuContext";
function createBaseMenu({
  Item: _Item = import_MenuPredefined.MenuPredefined.MenuItem,
  Title: _Title = import_MenuPredefined.MenuPredefined.Title,
  SubTitle: _SubTitle = import_MenuPredefined.MenuPredefined.SubTitle,
  Image: _Image = import_MenuPredefined.MenuPredefined.MenuImage,
  Icon: _Icon = import_MenuPredefined.MenuPredefined.MenuIcon,
  Indicator: _Indicator = import_MenuPredefined.MenuPredefined.MenuIndicator,
  Separator: _Separator = import_MenuPredefined.MenuPredefined.MenuSeparator,
  MenuGroup: _MenuGroup = import_MenuPredefined.MenuPredefined.MenuGroup,
  Label: _Label = import_MenuPredefined.MenuPredefined.MenuLabel
}) {
  const MenuComp = props => {
    const direction = (0, import_use_direction.useDirection)(props.dir);
    const defaultPlacement = direction === "rtl" ? "bottom-end" : "bottom-start";
    const {
      scope = MENU_CONTEXT,
      open = false,
      children,
      dir,
      onOpenChange,
      modal = true,
      allowFlip = {
        padding: 10
      },
      stayInFrame = {
        padding: 10
      },
      placement = defaultPlacement,
      resize = true,
      offset = 10,
      ...rest
    } = props;
    const [content, setContent] = React.useState(null);
    const isUsingKeyboardRef = React.useRef(false);
    const handleOpenChange = (0, import_use_callback_ref.useCallbackRef)(onOpenChange);
    if (import_web.isWeb) {
      React.useEffect(() => {
        const handleKeyDown = () => {
          isUsingKeyboardRef.current = true;
          document.addEventListener("pointerdown", handlePointer, {
            capture: true,
            once: true
          });
          document.addEventListener("pointermove", handlePointer, {
            capture: true,
            once: true
          });
        };
        const handlePointer = () => isUsingKeyboardRef.current = false;
        document.addEventListener("keydown", handleKeyDown, {
          capture: true
        });
        return () => {
          document.removeEventListener("keydown", handleKeyDown, {
            capture: true
          });
          document.removeEventListener("pointerdown", handlePointer, {
            capture: true
          });
          document.removeEventListener("pointermove", handlePointer, {
            capture: true
          });
        };
      }, []);
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(PopperPrimitive.Popper, {
      scope,
      open,
      placement,
      allowFlip,
      stayInFrame,
      resize,
      offset,
      ...rest,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuProvider, {
        scope,
        open,
        onOpenChange: handleOpenChange,
        content,
        onContentChange: setContent,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuRootProvider, {
          scope,
          open,
          onClose: React.useCallback(() => handleOpenChange(false), [handleOpenChange]),
          isUsingKeyboardRef,
          dir: direction,
          modal,
          children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuSubProvider, {
            scope,
            children
          })
        })
      })
    });
  };
  const RepropagateMenuAndMenuRootProvider = props => {
    const {
      scope = MENU_CONTEXT,
      menuContext,
      rootContext,
      popperContext,
      menuSubContext,
      children
    } = props;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(PopperPrimitive.PopperProvider, {
      ...popperContext,
      scope,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuProvider, {
        scope,
        ...menuContext,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuRootProvider, {
          scope,
          ...rootContext,
          children: menuSubContext ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuSubProvider, {
            scope,
            ...menuSubContext,
            children
          }) : children
        })
      })
    });
  };
  MenuComp.displayName = MENU_NAME;
  const ANCHOR_NAME = "MenuAnchor";
  const MenuAnchor = props => {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(PopperPrimitive.PopperAnchor, {
      scope: MENU_CONTEXT,
      ...props
    });
  };
  MenuAnchor.displayName = ANCHOR_NAME;
  const PORTAL_NAME = "MenuPortal";
  const {
    Provider: PortalProvider,
    useStyledContext: usePortalContext
  } = (0, import_web.createStyledContext)(void 0, "Portal");
  const MenuPortal = props => {
    const {
      scope = MENU_CONTEXT,
      forceMount,
      zIndex,
      children
    } = props;
    const menuContext = useMenuContext(scope);
    const rootContext = useMenuRootContext(scope);
    const popperContext = PopperPrimitive.usePopperContext(scope);
    const menuSubContext = useMenuSubContext(scope);
    const themeName = (0, import_web.useThemeName)();
    const themedChildren = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_web.Theme, {
      forceClassName: true,
      name: themeName,
      children
    });
    const content = (0, import_portal.needsPortalRepropagation)() ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(RepropagateMenuAndMenuRootProvider, {
      menuContext,
      rootContext,
      popperContext,
      menuSubContext,
      scope,
      children: themedChildren
    }) : themedChildren;
    const isPresent = forceMount || rootContext.open && menuContext.open;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate.Animate, {
      type: "presence",
      present: isPresent,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_portal.Portal, {
        stackZIndex: true,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
          children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(PortalProvider, {
            scope,
            forceMount,
            children: /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_web.View, {
              zIndex: zIndex || 100,
              inset: 0,
              position: "absolute",
              children: [!!menuContext.open && !import_web.isWeb && /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_web.View, {
                inset: 0,
                position: "absolute",
                onPress: () => menuContext.onOpenChange(!menuContext.open)
              }), content]
            })
          })
        })
      })
    });
  };
  MenuPortal.displayName = PORTAL_NAME;
  const CONTENT_NAME = "MenuContent";
  const {
    Provider: MenuContentProvider,
    useStyledContext: useMenuContentContext
  } = (0, import_web.createStyledContext)();
  const MenuContentFrame = (0, import_web.styled)(PopperPrimitive.PopperContentFrame, {
    name: CONTENT_NAME
  });
  const MenuContent = MenuContentFrame.styleable((props, forwardedRef) => {
    const scope = props.scope || MENU_CONTEXT;
    const portalContext = usePortalContext(scope);
    const {
      forceMount = portalContext.forceMount,
      ...contentProps
    } = props;
    const rootContext = useMenuRootContext(scope);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.Provider, {
      scope,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.Slot, {
        scope,
        children: rootContext.modal ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuRootContentModal, {
          ...contentProps,
          ref: forwardedRef
        }) : /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuRootContentNonModal, {
          ...contentProps,
          ref: forwardedRef
        })
      })
    });
  });
  const MenuRootContentModal = React.forwardRef((props, forwardedRef) => {
    const scope = props.scope || MENU_CONTEXT;
    const context = useMenuContext(scope);
    const ref = React.useRef(null);
    const composedRefs = (0, import_web.useComposedRefs)(forwardedRef, ref);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuContentImpl, {
      ...props,
      scope,
      ref: composedRefs,
      trapFocus: context.open,
      disableOutsidePointerEvents: context.open,
      disableOutsideScroll: false,
      onFocusOutside: (0, import_web.composeEventHandlers)(props.onFocusOutside, event => event.preventDefault(), {
        checkDefaultPrevented: false
      }),
      onDismiss: () => context.onOpenChange(false)
    });
  });
  const MenuRootContentNonModal = React.forwardRef((props, forwardedRef) => {
    const scope = props.scope || MENU_CONTEXT;
    const context = useMenuContext(scope);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuContentImpl, {
      ...props,
      scope,
      ref: forwardedRef,
      trapFocus: false,
      disableOutsidePointerEvents: false,
      disableOutsideScroll: false,
      onDismiss: () => context.onOpenChange(false)
    });
  });
  const MenuContentImpl = React.forwardRef((props, forwardedRef) => {
    const {
      scope = MENU_CONTEXT,
      loop = false,
      trapFocus,
      onOpenAutoFocus,
      onCloseAutoFocus,
      disableOutsidePointerEvents,
      onEntryFocus,
      onEscapeKeyDown,
      onPointerDownOutside,
      onFocusOutside,
      onInteractOutside,
      onDismiss,
      disableOutsideScroll,
      disableDismissOnScroll = false,
      unstyled = process.env.TAMAGUI_HEADLESS === "1",
      ...contentProps
    } = props;
    const context = useMenuContext(scope);
    const rootContext = useMenuRootContext(scope);
    const getItems = useCollection(scope);
    const [currentItemId, setCurrentItemId] = React.useState(null);
    const contentRef = React.useRef(null);
    const focusableContentRef = React.useRef(null);
    const composedRefs = (0, import_web.useComposedRefs)(forwardedRef, contentRef, context.onContentChange);
    const timerRef = React.useRef(0);
    const searchRef = React.useRef("");
    const pointerGraceTimerRef = React.useRef(0);
    const pointerGraceIntentRef = React.useRef(null);
    const pointerDirRef = React.useRef("right");
    const lastPointerXRef = React.useRef(0);
    const handleTypeaheadSearch = key => {
      const search = searchRef.current + key;
      const items = getItems().filter(item => !item.disabled);
      const currentItem = document.activeElement;
      const currentMatch = items.find(item => item.ref.current === currentItem)?.textValue;
      const values = items.map(item => item.textValue);
      const nextMatch = getNextMatch(values, search, currentMatch);
      const newItem = items.find(item => item.textValue === nextMatch)?.ref.current;
      (function updateSearch(value) {
        searchRef.current = value;
        clearTimeout(timerRef.current);
        if (value !== "") timerRef.current = setTimeout(() => updateSearch(""), 1e3);
      })(search);
      if (newItem) {
        setTimeout(() => newItem.focus());
      }
    };
    React.useEffect(() => {
      return () => clearTimeout(timerRef.current);
    }, []);
    React.useEffect(() => {
      if (!import_web.isWeb || !context.open) return;
      const frame = requestAnimationFrame(() => {
        const container = contentRef.current;
        const el = container?.querySelector("[data-tamagui-menu-content]");
        if (el) focusableContentRef.current = el;
      });
      return () => cancelAnimationFrame(frame);
    }, [context.open]);
    React.useEffect(() => {
      if (!import_web.isWeb || disableDismissOnScroll || !context.open) return;
      const handleScroll = event => {
        const target = event.target;
        if (contentRef.current?.contains(target)) return;
        onDismiss?.();
      };
      window.addEventListener("scroll", handleScroll, {
        capture: true,
        passive: true
      });
      return () => {
        window.removeEventListener("scroll", handleScroll, {
          capture: true
        });
      };
    }, [disableDismissOnScroll, context.open, onDismiss]);
    if (import_web.isWeb) {
      (0, import_focus_guard.useFocusGuards)();
    }
    const isPointerMovingToSubmenu = React.useCallback(event => {
      const isMovingTowards = pointerDirRef.current === pointerGraceIntentRef.current?.side;
      const inArea = isPointerInGraceArea(event, pointerGraceIntentRef.current?.area);
      return isMovingTowards && inArea;
    }, []);
    const content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(PopperPrimitive.PopperContent, {
      role: "menu",
      tabIndex: -1,
      unstyled,
      ...(!unstyled && {
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "$borderColor",
        outlineWidth: 0,
        minWidth: 180
      }),
      "aria-orientation": "vertical",
      "data-state": getOpenState(context.open),
      "data-tamagui-menu-content": "",
      dir: rootContext.dir,
      scope: scope || MENU_CONTEXT,
      ...contentProps,
      ref: composedRefs,
      className: contentProps.transition ? void 0 : contentProps.className,
      ...(import_web.isWeb ? {
        onKeyDown: (0, import_web.composeEventHandlers)(contentProps.onKeyDown, event => {
          const target = event.target;
          const isKeyDownInside = target.closest("[data-tamagui-menu-content]") === event.currentTarget;
          const isModifierKey = event.ctrlKey || event.altKey || event.metaKey;
          const isCharacterKey = event.key.length === 1;
          if (isKeyDownInside) {
            if (event.key === "Tab") event.preventDefault();
            if (!isModifierKey && isCharacterKey) handleTypeaheadSearch(event.key);
          }
          const isOnContentFrame = event.target.hasAttribute("data-tamagui-menu-content");
          if (!isKeyDownInside || !isOnContentFrame) return;
          if (!FIRST_LAST_KEYS.includes(event.key)) return;
          event.preventDefault();
          const items = getItems().filter(item => !item.disabled);
          const candidateNodes = items.map(item => item.ref.current);
          if (LAST_KEYS.includes(event.key)) candidateNodes.reverse();
          focusFirst(candidateNodes, {
            focusVisible: true
          });
        }),
        // TODO
        // @ts-ignore
        onBlur: (0, import_web.composeEventHandlers)(props.onBlur, event => {
          if (!event.currentTarget?.contains(event.target)) {
            clearTimeout(timerRef.current);
            searchRef.current = "";
          }
        }),
        // TODO
        onPointerMove: (0, import_web.composeEventHandlers)(props.onPointerMove, event => {
          if (event.pointerType !== "mouse") return;
          const target = event.target;
          const pointerXHasChanged = lastPointerXRef.current !== event.clientX;
          if (event.currentTarget?.contains(target) && pointerXHasChanged) {
            const newDir = event.clientX > lastPointerXRef.current ? "right" : "left";
            pointerDirRef.current = newDir;
            lastPointerXRef.current = event.clientX;
          }
        })
      } : {})
    });
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuContentProvider, {
      scope,
      searchRef,
      onItemEnter: React.useCallback(event => {
        if (isPointerMovingToSubmenu(event)) event.preventDefault();
      }, [isPointerMovingToSubmenu]),
      onItemLeave: React.useCallback(event => {
        if (isPointerMovingToSubmenu(event)) return;
        focusableContentRef.current?.focus();
        setCurrentItemId(null);
      }, [isPointerMovingToSubmenu]),
      onTriggerLeave: React.useCallback(event => {
        if (isPointerMovingToSubmenu(event)) event.preventDefault();
      }, [isPointerMovingToSubmenu]),
      pointerGraceTimerRef,
      onPointerGraceIntentChange: React.useCallback(intent => {
        pointerGraceIntentRef.current = intent;
      }, []),
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_remove_scroll.RemoveScroll, {
        enabled: disableOutsideScroll,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_focus_scope.FocusScope, {
          asChild: false,
          trapped: trapFocus,
          onMountAutoFocus: (0, import_web.composeEventHandlers)(onOpenAutoFocus, event => {
            event.preventDefault();
            const content2 = document.querySelector("[data-tamagui-menu-content]");
            content2?.focus({
              preventScroll: true
            });
          }),
          onUnmountAutoFocus: onCloseAutoFocus,
          children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_dismissable.Dismissable, {
            disableOutsidePointerEvents,
            onEscapeKeyDown,
            onPointerDownOutside,
            onFocusOutside,
            onInteractOutside,
            onDismiss,
            asChild: true,
            children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup, {
              asChild: true,
              __scopeRovingFocusGroup: scope || MENU_CONTEXT,
              dir: rootContext.dir,
              orientation: "vertical",
              loop,
              currentTabStopId: currentItemId,
              onCurrentTabStopIdChange: setCurrentItemId,
              onEntryFocus: (0, import_web.composeEventHandlers)(onEntryFocus, event => {
                if (!rootContext.isUsingKeyboardRef.current) {
                  event.preventDefault();
                }
              }),
              children: content
            })
          })
        })
      })
    });
  });
  MenuContent.displayName = CONTENT_NAME;
  const ITEM_NAME = "MenuItem";
  const ITEM_SELECT = "menu.itemSelect";
  const MenuItem = _Item.styleable((props, forwardedRef) => {
    const {
      disabled = false,
      onSelect,
      preventCloseOnSelect,
      children,
      scope = MENU_CONTEXT,
      // filter out native-only props that shouldn't reach the DOM
      // @ts-ignore
      destructive,
      // @ts-ignore
      hidden,
      // @ts-ignore
      androidIconName,
      // @ts-ignore
      iosIconName,
      ...itemProps
    } = props;
    const ref = React.useRef(null);
    const rootContext = useMenuRootContext(scope);
    const contentContext = useMenuContentContext(scope);
    const composedRefs = (0, import_web.useComposedRefs)(forwardedRef, ref);
    const isPointerDownRef = React.useRef(false);
    const handleSelect = () => {
      const menuItem = ref.current;
      if (!disabled && menuItem) {
        if (import_web.isWeb) {
          const menuItemEl = menuItem;
          const itemSelectEvent = new CustomEvent(ITEM_SELECT, {
            bubbles: true,
            cancelable: true
          });
          menuItemEl.addEventListener(ITEM_SELECT, event => onSelect?.(event), {
            once: true
          });
          (0, import_dismissable.dispatchDiscreteCustomEvent)(menuItemEl, itemSelectEvent);
          if (itemSelectEvent.defaultPrevented || preventCloseOnSelect) {
            isPointerDownRef.current = false;
          } else {
            rootContext.onClose();
          }
        } else {
          onSelect?.({
            target: menuItem
          });
          isPointerDownRef.current = false;
          if (!preventCloseOnSelect) {
            rootContext.onClose();
          }
        }
      }
    };
    const content = typeof children === "string" ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_web.Text, {
      children
    }) : children;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuItemImpl, {
      outlineStyle: "none",
      ...itemProps,
      scope,
      ref: composedRefs,
      disabled,
      onPress: (0, import_web.composeEventHandlers)(props.onPress, handleSelect),
      onPointerDown: event => {
        props.onPointerDown?.(event);
        isPointerDownRef.current = true;
      },
      onPointerUp: (0, import_web.composeEventHandlers)(props.onPointerUp, event => {
        if (import_web.isWeb) {
          if (!isPointerDownRef.current) event.currentTarget?.click();
        }
      }),
      ...(import_web.isWeb ? {
        onKeyDown: (0, import_web.composeEventHandlers)(props.onKeyDown, event => {
          const isTypingAhead = contentContext.searchRef.current !== "";
          if (disabled || isTypingAhead && event.key === " ") return;
          if (SELECTION_KEYS.includes(event.key)) {
            event.currentTarget?.click();
            event.preventDefault();
          }
        })
      } : {}),
      children: content
    });
  });
  const MenuItemImpl = React.forwardRef((props, forwardedRef) => {
    const {
      scope = MENU_CONTEXT,
      disabled = false,
      textValue,
      unstyled = process.env.TAMAGUI_HEADLESS === "1",
      ...itemProps
    } = props;
    const contentContext = useMenuContentContext(scope);
    const ref = React.useRef(null);
    const composedRefs = (0, import_web.useComposedRefs)(forwardedRef, ref);
    const [isFocused, setIsFocused] = React.useState(false);
    const [textContent, setTextContent] = React.useState("");
    if (import_web.isWeb) {
      React.useEffect(() => {
        const menuItem = ref.current;
        if (menuItem) {
          setTextContent((menuItem.textContent ?? "").trim());
        }
      }, [itemProps.children]);
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.ItemSlot, {
      scope,
      disabled,
      textValue: textValue ?? textContent,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup.Item, {
        asChild: true,
        __scopeRovingFocusGroup: scope,
        focusable: !disabled,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(_Item, {
          unstyled,
          componentName: ITEM_NAME,
          role: "menuitem",
          "data-highlighted": isFocused ? "" : void 0,
          "aria-disabled": disabled || void 0,
          "data-disabled": disabled ? "" : void 0,
          ...itemProps,
          ref: composedRefs,
          onPointerMove: (0, import_web.composeEventHandlers)(props.onPointerMove, event => {
            if (event.pointerType !== "mouse") return;
            if (disabled) {
              contentContext.onItemLeave(event);
            } else {
              contentContext.onItemEnter(event);
              if (!event.defaultPrevented) {
                const item = event.currentTarget;
                item.focus({
                  preventScroll: true,
                  focusVisible: false
                });
              }
            }
          }),
          onPointerLeave: (0, import_web.composeEventHandlers)(props.onPointerLeave, event => {
            contentContext.onItemLeave(event);
          }),
          onFocus: (0, import_web.composeEventHandlers)(props.onFocus, () => setIsFocused(true)),
          onBlur: (0, import_web.composeEventHandlers)(props.onBlur, () => setIsFocused(false))
        })
      })
    });
  });
  MenuItem.displayName = ITEM_NAME;
  const ITEM_TITLE_NAME = "MenuItemTitle";
  const MenuItemTitle = _Title.styleable((props, forwardedRef) => {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(_Title, {
      ...props,
      ref: forwardedRef
    });
  });
  MenuItemTitle.displayName = ITEM_TITLE_NAME;
  const ITEM_SUB_TITLE_NAME = "MenuItemSubTitle";
  const MenuItemSubTitle = _SubTitle.styleable((props, forwardedRef) => {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(_SubTitle, {
      ...props,
      ref: forwardedRef
    });
  });
  MenuItemSubTitle.displayName = ITEM_SUB_TITLE_NAME;
  const ITEM_IMAGE = "MenuItemImage";
  const MenuItemImage = React.forwardRef((props, forwardedRef) => {
    const {
      // @ts-ignore - native menu ios config
      ios,
      // @ts-ignore
      androidIconName,
      // @ts-ignore
      iosIconName,
      ...rest
    } = props;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(_Image, {
      ...rest,
      ref: forwardedRef
    });
  });
  MenuItemImage.displayName = ITEM_IMAGE;
  const ITEM_ICON = "MenuItemIcon";
  const MenuItemIcon = _Icon.styleable((props, forwardedRef) => {
    const {
      // @ts-ignore
      ios,
      // @ts-ignore
      android,
      // @ts-ignore
      androidIconName,
      // @ts-ignore
      iosIconName,
      ...rest
    } = props;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(_Icon, {
      ...rest,
      ref: forwardedRef
    });
  });
  MenuItemIcon.displayName = ITEM_ICON;
  const CHECKBOX_ITEM_NAME = "MenuCheckboxItem";
  const MenuCheckboxItem = _Item.styleable((props, forwardedRef) => {
    const {
      checked = false,
      onCheckedChange,
      scope = MENU_CONTEXT,
      // filter out native-only props
      // @ts-ignore - native menu value state
      value,
      // @ts-ignore - native menu value change handler
      onValueChange,
      ...checkboxItemProps
    } = props;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(ItemIndicatorProvider, {
      scope,
      checked,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuItem, {
        componentName: CHECKBOX_ITEM_NAME,
        role: import_web.isWeb ? "menuitemcheckbox" : "menuitem",
        "aria-checked": isIndeterminate(checked) ? "mixed" : checked,
        ...checkboxItemProps,
        scope,
        ref: forwardedRef,
        "data-state": getCheckedState(checked),
        onSelect: (0, import_web.composeEventHandlers)(checkboxItemProps.onSelect, () => onCheckedChange?.(isIndeterminate(checked) ? true : !checked), {
          checkDefaultPrevented: false
        })
      })
    });
  });
  MenuCheckboxItem.displayName = CHECKBOX_ITEM_NAME;
  const RADIO_GROUP_NAME = "MenuRadioGroup";
  const {
    Provider: RadioGroupProvider,
    useStyledContext: useRadioGroupContext
  } = (0, import_web.createStyledContext)();
  const MenuRadioGroup = _MenuGroup.styleable((props, forwardedRef) => {
    const {
      value,
      onValueChange,
      scope = MENU_CONTEXT,
      ...groupProps
    } = props;
    const handleValueChange = (0, import_use_callback_ref.useCallbackRef)(onValueChange);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(RadioGroupProvider, {
      scope,
      value,
      onValueChange: handleValueChange,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(_MenuGroup, {
        componentName: RADIO_GROUP_NAME,
        ...groupProps,
        ref: forwardedRef
      })
    });
  });
  MenuRadioGroup.displayName = RADIO_GROUP_NAME;
  const RADIO_ITEM_NAME = "MenuRadioItem";
  const MenuRadioItem = _Item.styleable((props, forwardedRef) => {
    const {
      value,
      scope = MENU_CONTEXT,
      ...radioItemProps
    } = props;
    const context = useRadioGroupContext(scope);
    const checked = value === context.value;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(ItemIndicatorProvider, {
      scope,
      checked,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuItem, {
        componentName: RADIO_ITEM_NAME,
        ...radioItemProps,
        scope,
        "aria-checked": checked,
        ref: forwardedRef,
        role: import_web.isWeb ? "menuitemradio" : "menuitem",
        "data-state": getCheckedState(checked),
        onSelect: (0, import_web.composeEventHandlers)(radioItemProps.onSelect, () => context.onValueChange?.(value), {
          checkDefaultPrevented: false
        })
      })
    });
  });
  MenuRadioItem.displayName = RADIO_ITEM_NAME;
  const ITEM_INDICATOR_NAME = "MenuItemIndicator";
  const {
    Provider: ItemIndicatorProvider,
    useStyledContext: useItemIndicatorContext
  } = (0, import_web.createStyledContext)();
  const MenuItemIndicator = _Indicator.styleable((props, forwardedRef) => {
    const {
      scope = MENU_CONTEXT,
      forceMount,
      ...itemIndicatorProps
    } = props;
    const indicatorContext = useItemIndicatorContext(scope);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate_presence.AnimatePresence, {
      children: forceMount || isIndeterminate(indicatorContext.checked) || indicatorContext.checked === true ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(_Indicator, {
        componentName: ITEM_INDICATOR_NAME,
        render: "span",
        ...itemIndicatorProps,
        ref: forwardedRef,
        "data-state": getCheckedState(indicatorContext.checked)
      }) : null
    });
  });
  MenuItemIndicator.displayName = ITEM_INDICATOR_NAME;
  const MenuArrow = React.forwardRef(function MenuArrow2(props, forwardedRef) {
    const {
      scope = MENU_CONTEXT,
      unstyled = process.env.TAMAGUI_HEADLESS === "1",
      ...rest
    } = props;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(PopperPrimitive.PopperArrow, {
      scope,
      componentName: "PopperArrow",
      unstyled,
      ...(!unstyled && {
        backgroundColor: "$background"
      }),
      ...rest,
      ref: forwardedRef
    });
  });
  const SUB_NAME = "MenuSub";
  const {
    Provider: MenuSubProvider,
    useStyledContext: useMenuSubContext
  } = (0, import_web.createStyledContext)();
  const MenuSub = props => {
    const isTouchDevice = (0, import_web.useIsTouchDevice)();
    const {
      scope = MENU_CONTEXT
    } = props;
    const rootContext = useMenuRootContext(scope);
    const parentPopperContext = PopperPrimitive.usePopperContext(scope);
    const parentSide = parentPopperContext.placement?.split("-")[0];
    const isNestedSubmenu = parentSide === "left" || parentSide === "right";
    const defaultPlacement = isTouchDevice ? "bottom" : isNestedSubmenu ? `${parentSide}-start` : rootContext.dir === "rtl" ? "left-start" : "right-start";
    const {
      children,
      open = false,
      onOpenChange,
      allowFlip: allowFlipProp = {
        padding: 10
      },
      stayInFrame = {
        padding: 10
      },
      placement = defaultPlacement,
      ...rest
    } = props;
    const allowFlip = React.useMemo(() => {
      if (!isNestedSubmenu || typeof allowFlipProp === "boolean") return allowFlipProp;
      if (allowFlipProp.fallbackPlacements) return allowFlipProp;
      const side = placement.split("-")[0];
      const align = placement.split("-")[1] || "start";
      const otherAlign = align === "start" ? "end" : "start";
      if (side === "left" || side === "right") {
        const oppositeSide = side === "right" ? "left" : "right";
        return {
          ...(typeof allowFlipProp === "object" ? allowFlipProp : {}),
          fallbackPlacements: [`${side}-${otherAlign}`, `${oppositeSide}-${align}`, `${oppositeSide}-${otherAlign}`]
        };
      }
      return allowFlipProp;
    }, [isNestedSubmenu, allowFlipProp, placement]);
    const parentMenuContext = useMenuContext(scope);
    const [trigger, setTrigger] = React.useState(null);
    const [content, setContent] = React.useState(null);
    const handleOpenChange = (0, import_use_callback_ref.useCallbackRef)(onOpenChange);
    React.useEffect(() => {
      if (parentMenuContext.open === false) handleOpenChange(false);
      return () => handleOpenChange(false);
    }, [parentMenuContext.open, handleOpenChange]);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(PopperPrimitive.Popper, {
      open,
      placement,
      allowFlip,
      stayInFrame,
      ...rest,
      scope,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuProvider, {
        scope,
        open,
        onOpenChange: handleOpenChange,
        content,
        onContentChange: setContent,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuSubProvider, {
          scope,
          contentId: (0, import_react.useId)(),
          triggerId: (0, import_react.useId)(),
          trigger,
          onTriggerChange: setTrigger,
          children
        })
      })
    });
  };
  MenuSub.displayName = SUB_NAME;
  const SUB_TRIGGER_NAME = "MenuSubTrigger";
  const MenuSubTrigger = React.forwardRef((props, forwardedRef) => {
    const scope = props.scope || MENU_CONTEXT;
    const context = useMenuContext(scope);
    const rootContext = useMenuRootContext(scope);
    const subContext = useMenuSubContext(scope);
    const contentContext = useMenuContentContext(scope);
    const popperContext = PopperPrimitive.usePopperContext(scope);
    const openTimerRef = React.useRef(null);
    const {
      pointerGraceTimerRef,
      onPointerGraceIntentChange
    } = contentContext;
    const effectiveDir = rootContext.dir;
    const clearOpenTimer = React.useCallback(() => {
      if (openTimerRef.current) window.clearTimeout(openTimerRef.current);
      openTimerRef.current = null;
    }, []);
    React.useEffect(() => clearOpenTimer, [clearOpenTimer]);
    React.useEffect(() => {
      const pointerGraceTimer = pointerGraceTimerRef.current;
      return () => {
        window.clearTimeout(pointerGraceTimer);
        onPointerGraceIntentChange(null);
      };
    }, [pointerGraceTimerRef, onPointerGraceIntentChange]);
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuAnchor, {
      componentName: SUB_TRIGGER_NAME,
      asChild: "except-style",
      scope,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuItemImpl, {
        id: subContext.triggerId,
        "aria-haspopup": "menu",
        "aria-expanded": context.open,
        "aria-controls": subContext.contentId,
        "data-state": getOpenState(context.open),
        outlineStyle: "none",
        ...props,
        ref: (0, import_web.composeRefs)(forwardedRef, subContext.onTriggerChange),
        onPress: event => {
          props.onPress?.(event);
          if (props.disabled || event.defaultPrevented) return;
          if (import_web.isWeb) {
            event.currentTarget.focus();
          }
          if (!context.open) context.onOpenChange(true);
        },
        onPointerMove: (0, import_web.composeEventHandlers)(props.onPointerMove,
        // @ts-ignore
        whenMouse(event => {
          contentContext.onItemEnter(event);
          if (event.defaultPrevented) return;
          if (!props.disabled && !context.open && !openTimerRef.current) {
            contentContext.onPointerGraceIntentChange(null);
            openTimerRef.current = window.setTimeout(() => {
              context.onOpenChange(true);
              clearOpenTimer();
            }, 100);
          }
        })),
        onPointerLeave: (0, import_web.composeEventHandlers)(props.onPointerLeave, eventIn => {
          const event = eventIn;
          clearOpenTimer();
          const contentRect = context.content?.getBoundingClientRect();
          if (contentRect) {
            const contentEl = context.content;
            const sideEl = contentEl?.dataset?.side ? contentEl : contentEl?.querySelector("[data-side]");
            const side = sideEl?.dataset?.side || "right";
            const rightSide = side === "right";
            const bleed = rightSide ? -5 : 5;
            const contentNearEdge = contentRect[rightSide ? "left" : "right"];
            const contentFarEdge = contentRect[rightSide ? "right" : "left"];
            const polygon = {
              area: [
              // Apply a bleed on clientX to ensure that our exit point is
              // consistently within polygon bounds
              {
                x: event.clientX + bleed,
                y: event.clientY
              }, {
                x: contentNearEdge,
                y: contentRect.top
              }, {
                x: contentFarEdge,
                y: contentRect.top
              }, {
                x: contentFarEdge,
                y: contentRect.bottom
              }, {
                x: contentNearEdge,
                y: contentRect.bottom
              }],
              side
            };
            contentContext.onPointerGraceIntentChange(polygon);
            window.clearTimeout(pointerGraceTimerRef.current);
            pointerGraceTimerRef.current = window.setTimeout(() => contentContext.onPointerGraceIntentChange(null), 300);
          } else if (import_web.isWeb && subContext.trigger) {
            const triggerEl = subContext.trigger;
            const triggerRect = triggerEl?.getBoundingClientRect();
            if (triggerRect) {
              const placementSide = popperContext.placement?.split("-")[0];
              const side = placementSide === "left" || placementSide === "right" ? placementSide : rootContext.dir === "rtl" ? "left" : "right";
              const rightSide = side === "right";
              const bleed = rightSide ? -5 : 5;
              const nearEdge = rightSide ? triggerRect.right + 4 : triggerRect.left - 4;
              const farEdge = rightSide ? nearEdge + 200 : nearEdge - 200;
              const polygon = {
                area: [{
                  x: event.clientX + bleed,
                  y: event.clientY
                }, {
                  x: nearEdge,
                  y: triggerRect.top - 50
                }, {
                  x: farEdge,
                  y: triggerRect.top - 50
                }, {
                  x: farEdge,
                  y: triggerRect.bottom + 50
                }, {
                  x: nearEdge,
                  y: triggerRect.bottom + 50
                }],
                side
              };
              contentContext.onPointerGraceIntentChange(polygon);
              window.clearTimeout(pointerGraceTimerRef.current);
              pointerGraceTimerRef.current = window.setTimeout(() => contentContext.onPointerGraceIntentChange(null), 300);
            }
          } else {
            contentContext.onTriggerLeave(event);
            if (event.defaultPrevented) return;
            contentContext.onPointerGraceIntentChange(null);
          }
        }),
        ...(import_web.isWeb ? {
          onKeyDown: (0, import_web.composeEventHandlers)(props.onKeyDown, event => {
            const isTypingAhead = contentContext.searchRef.current !== "";
            if (props.disabled || isTypingAhead && event.key === " ") return;
            const willOpen = SUB_OPEN_KEYS[effectiveDir].includes(event.key);
            if (willOpen) {
              if (context.open && context.content) {
                const contentEl = context.content;
                const firstItem = contentEl.querySelector?.('[role="menuitem"]:not([data-disabled])');
                if (firstItem) {
                  firstItem.focus({
                    focusVisible: true
                  });
                  event.preventDefault();
                  return;
                }
              }
              const triggerEl = event.currentTarget;
              popperContext.refs?.setReference(triggerEl);
              context.onOpenChange(true);
              requestAnimationFrame(() => {
                popperContext.update?.();
              });
              context.content?.focus({
                focusVisible: true
              });
              event.preventDefault();
            }
          })
        } : null)
      })
    });
  });
  MenuSubTrigger.displayName = SUB_TRIGGER_NAME;
  const SUB_CONTENT_NAME = "MenuSubContent";
  const MenuSubContentFrame = (0, import_web.styled)(PopperPrimitive.PopperContentFrame, {
    name: SUB_CONTENT_NAME
  });
  const MenuSubContent = MenuSubContentFrame.styleable((props, forwardedRef) => {
    const scope = props.scope || MENU_CONTEXT;
    const portalContext = usePortalContext(scope);
    const {
      forceMount = portalContext.forceMount,
      ...subContentProps
    } = props;
    const context = useMenuContext(scope);
    const rootContext = useMenuRootContext(scope);
    const subContext = useMenuSubContext(scope);
    const popperContext = PopperPrimitive.usePopperContext(scope);
    const ref = React.useRef(null);
    const composedRefs = (0, import_web.useComposedRefs)(forwardedRef, ref);
    const placementSide = popperContext.placement?.split("-")[0];
    const dataSide = placementSide === "left" || placementSide === "right" ? placementSide : rootContext.dir === "rtl" ? "left" : "right";
    const effectiveDir = rootContext.dir;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.Provider, {
      scope,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.Slot, {
        scope,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(MenuContentImpl, {
          id: subContext.contentId,
          "aria-labelledby": subContext.triggerId,
          ...subContentProps,
          ref: composedRefs,
          "data-side": dataSide,
          disableOutsidePointerEvents: false,
          disableOutsideScroll: false,
          trapFocus: false,
          onOpenAutoFocus: event => {
            if (rootContext.isUsingKeyboardRef.current) {
              const root = ref.current;
              const content = root?.querySelector?.("[data-tamagui-menu-content]");
              (content || root)?.focus({
                preventScroll: true
              });
            }
            event.preventDefault();
          },
          onCloseAutoFocus: event => event.preventDefault(),
          onFocusOutside: (0, import_web.composeEventHandlers)(props.onFocusOutside, event => {
            if (event.target !== subContext.trigger) context.onOpenChange(false);
          }),
          onEscapeKeyDown: (0, import_web.composeEventHandlers)(props.onEscapeKeyDown, event => {
            context.onOpenChange(false);
            subContext.trigger?.focus({
              focusVisible: true
            });
            event.preventDefault();
          }),
          ...(import_web.isWeb ? {
            onKeyDown: (0, import_web.composeEventHandlers)(props.onKeyDown, event => {
              const isKeyDownInside = event.currentTarget.contains(event.target);
              const isCloseKey = SUB_CLOSE_KEYS[effectiveDir].includes(event.key);
              if (isKeyDownInside && isCloseKey) {
                context.onOpenChange(false);
                subContext.trigger?.focus({
                  focusVisible: true
                });
                event.preventDefault();
              }
            })
          } : null)
        })
      })
    });
  });
  MenuSubContent.displayName = SUB_CONTENT_NAME;
  const Anchor = MenuAnchor;
  const Portal = MenuPortal;
  const Content = MenuContent;
  const Group = _MenuGroup.styleable((props, ref) => {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(_MenuGroup, {
      ...props,
      ref
    });
  });
  Group.displayName = "MenuGroup";
  const Label = _Label.styleable((props, ref) => {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(_Label, {
      ...props,
      ref
    });
  });
  Label.displayName = "MenuLabel";
  const Item = MenuItem;
  const CheckboxItem = MenuCheckboxItem;
  const RadioGroup = MenuRadioGroup;
  const RadioItem = MenuRadioItem;
  const ItemIndicator = MenuItemIndicator;
  const Separator = _Separator.styleable((props, ref) => {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(_Separator, {
      ...props,
      ref
    });
  });
  Separator.displayName = "MenuSeparator";
  const Arrow = MenuArrow;
  const Sub = MenuSub;
  const SubTrigger = MenuSubTrigger;
  const SubContent = MenuSubContent;
  const ItemTitle = MenuItemTitle;
  const ItemSubtitle = MenuItemSubTitle;
  const ItemImage = MenuItemImage;
  const ItemIcon = MenuItemIcon;
  const Menu = (0, import_web.withStaticProperties)(MenuComp, {
    Anchor,
    Portal,
    Content,
    Group,
    Label,
    Item,
    CheckboxItem,
    RadioGroup,
    RadioItem,
    ItemIndicator,
    Separator,
    Arrow,
    Sub,
    SubTrigger,
    SubContent,
    ItemTitle,
    ItemSubtitle,
    ItemImage,
    ItemIcon
  });
  return {
    Menu
  };
}
function getOpenState(open) {
  return open ? "open" : "closed";
}
function isIndeterminate(checked) {
  return checked === "indeterminate";
}
function getCheckedState(checked) {
  return isIndeterminate(checked) ? "indeterminate" : checked ? "checked" : "unchecked";
}
function focusFirst(candidates, options) {
  const PREVIOUSLY_FOCUSED_ELEMENT = document.activeElement;
  for (const candidate of candidates) {
    if (candidate === PREVIOUSLY_FOCUSED_ELEMENT) return;
    candidate.focus({
      preventScroll: true,
      focusVisible: options?.focusVisible
    });
    if (document.activeElement !== PREVIOUSLY_FOCUSED_ELEMENT) return;
  }
}
function wrapArray(array, startIndex) {
  return array.map((_, index) => array[(startIndex + index) % array.length]);
}
function getNextMatch(values, search, currentMatch) {
  const isRepeated = search.length > 1 && Array.from(search).every(char => char === search[0]);
  const normalizedSearch = isRepeated ? search[0] : search;
  const currentMatchIndex = currentMatch ? values.indexOf(currentMatch) : -1;
  let wrappedValues = wrapArray(values, Math.max(currentMatchIndex, 0));
  const excludeCurrentMatch = normalizedSearch.length === 1;
  if (excludeCurrentMatch) wrappedValues = wrappedValues.filter(v => v !== currentMatch);
  const nextMatch = wrappedValues.find(value => value.toLowerCase().startsWith(normalizedSearch.toLowerCase()));
  return nextMatch !== currentMatch ? nextMatch : void 0;
}
function isPointInPolygon(point, polygon) {
  const {
    x,
    y
  } = point;
  let inside = false;
  for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    const xi = polygon[i].x;
    const yi = polygon[i].y;
    const xj = polygon[j].x;
    const yj = polygon[j].y;
    const intersect = yi > y !== yj > y && x < (xj - xi) * (y - yi) / (yj - yi) + xi;
    if (intersect) inside = !inside;
  }
  return inside;
}
function isPointerInGraceArea(event, area) {
  if (!area) return false;
  const cursorPos = {
    x: event.clientX,
    y: event.clientY
  };
  return isPointInPolygon(cursorPos, area);
}
