// @tamagui/switch/dist/esm/createSwitch.native.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { composeEventHandlers, getVariableValue, isWeb, View, withStaticProperties } from "@tamagui/core";
import { useSwitch } from "@tamagui/switch-headless";
import { useControllableState } from "@tamagui/use-controllable-state";
import * as React from "react";
import { SwitchStyledContext } from "./StyledContext.native.js";
import { SwitchFrame as DefaultSwitchFrame, SwitchThumb } from "./Switch.native.js";
import { useSwitchNative } from "./useSwitchNative.native.js";
function createSwitch(createProps) {
  var {
    Frame = DefaultSwitchFrame,
    Thumb = SwitchThumb
  } = createProps;
  if (process.env.NODE_ENV === "development") {
    if (Frame !== DefaultSwitchFrame && Frame.staticConfig.context && Frame.staticConfig.context !== SwitchStyledContext || Thumb !== SwitchThumb && Thumb.staticConfig.context && Thumb.staticConfig.context !== SwitchStyledContext) {
      console.warn(`Warning: createSwitch() needs to control context to pass checked state from Frame to Thumb, any custom context passed will be overridden.`);
    }
  }
  Frame.staticConfig.context = SwitchStyledContext;
  Thumb.staticConfig.context = SwitchStyledContext;
  var SwitchThumbComponent = Thumb.styleable(function SwitchThumb2(props, forwardedRef) {
    var {
      size: sizeProp,
      unstyled: unstyledProp,
      activeStyle,
      ...thumbProps
    } = props;
    var styledContext = SwitchStyledContext.useStyledContext();
    var {
      unstyled: unstyledContext,
      size: sizeContext,
      active,
      disabled,
      frameWidth = 0
    } = styledContext;
    var _ref;
    var unstyled = process.env.TAMAGUI_HEADLESS === "1" ? true : (_ref = unstyledProp !== null && unstyledProp !== void 0 ? unstyledProp : unstyledContext) !== null && _ref !== void 0 ? _ref : false;
    var _ref1;
    var size = (_ref1 = sizeProp !== null && sizeProp !== void 0 ? sizeProp : sizeContext) !== null && _ref1 !== void 0 ? _ref1 : "$true";
    var initialChecked = React.useRef(active).current;
    var initialWidth = getVariableValue(props.width || size, "size");
    var [thumbWidth, setThumbWidth] = React.useState(typeof initialWidth === "number" ? initialWidth : 0);
    var distance = frameWidth - thumbWidth;
    var x = initialChecked ? active ? 0 : -distance : active ? distance : 0;
    return /* @__PURE__ */_jsx(Thumb, {
      ref: forwardedRef,
      unstyled,
      ...(unstyled === false && {
        size
      }),
      alignSelf: initialChecked ? "flex-end" : "flex-start",
      x,
      onLayout: composeEventHandlers(props.onLayout, function (e) {
        var next = e.nativeEvent.layout.width;
        setThumbWidth(next);
      }),
      disabled,
      ...thumbProps,
      ...(active && activeStyle)
    });
  });
  var SwitchComponent = Frame.styleable(function SwitchFrame(_props, forwardedRef) {
    var {
      native,
      nativeProps,
      checked: checkedProp,
      defaultChecked,
      onCheckedChange,
      activeStyle,
      unstyled: unstyledProp,
      activeTheme: activeThemeProp,
      ...props
    } = _props;
    var [checked, setChecked] = useControllableState({
      prop: checkedProp,
      defaultProp: defaultChecked || false,
      onChange: onCheckedChange,
      transition: true
    });
    var styledContext = React.useContext(SwitchStyledContext.context);
    var [frameWidth, setFrameInnerWidth] = React.useState(0);
    var {
      switchProps,
      bubbleInput,
      switchRef
    } = useSwitch(props, [checked, setChecked],
    // @ts-ignore TODO tamagui react 19 type error
    forwardedRef);
    var nativeSwitch = useSwitchNative({
      id: props.id,
      disabled: props.disabled,
      native,
      nativeProps,
      checked,
      setChecked
    });
    if (nativeSwitch) {
      return nativeSwitch;
    }
    var disabled = props.disabled;
    var handleLayout = function (e) {
      var next = e.nativeEvent.layout.width;
      if (next !== frameWidth) {
        setFrameInnerWidth(next);
      }
    };
    var _styledContext_unstyled, _ref;
    var unstyled = (_ref = (_styledContext_unstyled = styledContext.unstyled) !== null && _styledContext_unstyled !== void 0 ? _styledContext_unstyled : unstyledProp) !== null && _ref !== void 0 ? _ref : false;
    var _styledContext_size, _ref1, _styledContext_size1, _ref2;
    return /* @__PURE__ */_jsxs(_Fragment, {
      children: [/* @__PURE__ */_jsx(SwitchStyledContext.Provider, {
        size: (_ref1 = (_styledContext_size = styledContext.size) !== null && _styledContext_size !== void 0 ? _styledContext_size : props.size) !== null && _ref1 !== void 0 ? _ref1 : "$true",
        unstyled,
        active: checked,
        disabled,
        frameWidth,
        children: /* @__PURE__ */_jsx(Frame, {
          ref: switchRef,
          render: "button",
          // activeThemeProp should be a theme name string like "active", not a style object
          theme: activeThemeProp !== null && activeThemeProp !== void 0 ? activeThemeProp : null,
          ...(isWeb && {
            type: "button"
          }),
          ...(!unstyled && {
            size: (_ref2 = (_styledContext_size1 = styledContext.size) !== null && _styledContext_size1 !== void 0 ? _styledContext_size1 : props.size) !== null && _ref2 !== void 0 ? _ref2 : "$true"
          }),
          unstyled,
          ...props,
          ...switchProps,
          disabled,
          ...(checked && {
            ...(!unstyled && !activeStyle && {
              backgroundColor: "$backgroundActive"
            }),
            ...activeStyle
          }),
          children: /* @__PURE__ */_jsx(View, {
            alignSelf: "stretch",
            flex: 1,
            onLayout: handleLayout,
            children: props.children
          })
        })
      }), bubbleInput]
    });
  }, {
    disableTheme: true
  });
  return withStaticProperties(SwitchComponent, {
    Thumb: SwitchThumbComponent
  });
}
export { createSwitch };
