// @tamagui/web/dist/esm/config.mjs
import { isWeb } from "@tamagui/constants";
import { MISSING_THEME_MESSAGE } from "./constants/constants.mjs";
let conf;
let setConfigCalledByThisInstance = false;
const haventCalledErrorMessage = process.env.NODE_ENV === "development" ? `
Haven't called createTamagui yet. ${MISSING_THEME_MESSAGE}
` : `\u274C Error 001`;
const getConfigFromGlobalOrLocal = () => {
  if (conf) {
    return conf;
  }
  if (globalThis.__tamaguiConfig) {
    if (process.env.NODE_ENV === "development" && !globalThis.__tamaguiHasWarnedGlobalFallback && !globalThis.__tamaguiPendingCheck) {
      globalThis.__tamaguiPendingCheck = true;
      setTimeout(() => {
        if (!setConfigCalledByThisInstance && !globalThis.__tamaguiHasWarnedGlobalFallback) {
          globalThis.__tamaguiHasWarnedGlobalFallback = true;
          console.warn(`\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F

Tamagui: Using global config fallback. This may indicate duplicate tamagui instances (e.g., from Vite SSR bundling). This is handled automatically, but likely causes issues!

\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F`);
        }
        globalThis.__tamaguiPendingCheck = false;
      }, 500);
    }
    return globalThis.__tamaguiConfig;
  }
  return null;
};
const getSetting = key => {
  const config = getConfigFromGlobalOrLocal();
  if (process.env.NODE_ENV === "development") {
    if (!config) throw new Error(haventCalledErrorMessage);
  }
  return config.settings[key] ??
  // @ts-expect-error
  config[key];
};
const setConfig = next => {
  setConfigCalledByThisInstance = true;
  conf = next;
  globalThis.__tamaguiConfig = next;
};
const setConfigFont = (name, font, fontParsed) => {
  const config = getConfigFromGlobalOrLocal();
  if (process.env.NODE_ENV === "development") {
    if (!config) throw new Error(haventCalledErrorMessage);
  }
  config.fonts[name] = font;
  config.fontsParsed[`$${name}`] = fontParsed;
};
const getConfig = () => {
  const config = getConfigFromGlobalOrLocal();
  if (!config) {
    throw new Error(process.env.NODE_ENV !== "production" ? `Missing tamagui config, you either have a duplicate config, or haven't set it up. Be sure createTamagui is called before rendering. Also, make sure all of your tamagui dependencies are on the same version (\`tamagui\`, \`@tamagui/package-name\`, etc.) not just in your package.json, but in your lockfile.` : "Err0");
  }
  return config;
};
const getConfigMaybe = () => {
  return getConfigFromGlobalOrLocal();
};
let tokensMerged;
function setTokens(_) {
  tokensMerged = _;
}
const getTokens = ({
  prefixed
} = {}) => {
  const config = getConfigFromGlobalOrLocal();
  if (process.env.NODE_ENV === "development") {
    if (!config) throw new Error(haventCalledErrorMessage);
  }
  const {
    tokens,
    tokensParsed
  } = config;
  if (prefixed === false) return tokens;
  if (prefixed === true) return tokensParsed;
  return tokensMerged;
};
const getTokenObject = (value, group) => {
  const config = getConfigFromGlobalOrLocal();
  return config.specificTokens[value] ?? (group ? tokensMerged[group]?.[value] : tokensMerged[Object.keys(tokensMerged).find(cat => tokensMerged[cat][value]) || ""]?.[value]);
};
const getToken = (value, group, useVariable = isWeb) => {
  const token = getTokenObject(value, group);
  return useVariable ? token?.variable : token?.val;
};
const getTokenValue = (value, group) => {
  if (value === "unset" || value === "auto") return;
  return getToken(value, group, false);
};
const useTokens = getTokens;
const getThemes = () => getConfigFromGlobalOrLocal().themes;
const updateConfig = (key, value) => {
  const config = getConfigFromGlobalOrLocal();
  Object.assign(config[key], value);
};
const getFont = name => {
  const conf2 = getConfig();
  return conf2.fontsParsed[name] ?? Object.entries(conf2.fontsParsed).find(([k]) => conf2.fontsParsed[k]?.family?.["val"] === name)?.[1];
};
let devConfig;
function setupDev(conf2) {
  if (process.env.NODE_ENV === "development") {
    devConfig = conf2;
  }
}
function loadAnimationDriver(name, driver) {
  const config = getConfigFromGlobalOrLocal();
  if (!config) {
    if (process.env.NODE_ENV === "development") {
      console.warn("loadAnimationDriver called before createTamagui");
    }
    return;
  }
  if (config.animations && !("default" in config.animations)) {
    ;
    config.animations = {
      default: config.animations
    };
  }
  if (config.animations) {
    ;
    config.animations[name] = driver;
  } else {
    ;
    config.animations = {
      default: driver,
      [name]: driver
    };
  }
}
export { devConfig, getConfig, getConfigMaybe, getFont, getSetting, getThemes, getToken, getTokenObject, getTokenValue, getTokens, loadAnimationDriver, setConfig, setConfigFont, setTokens, setupDev, updateConfig, useTokens };
