// @tamagui/web/dist/esm/config.native.js
import { isWeb } from "@tamagui/constants";
import { MISSING_THEME_MESSAGE } from "./constants/constants.native.js";
var conf;
var setConfigCalledByThisInstance = false;
var haventCalledErrorMessage = process.env.NODE_ENV === "development" ? `
Haven't called createTamagui yet. ${MISSING_THEME_MESSAGE}
` : `\u274C Error 001`;
var getConfigFromGlobalOrLocal = function () {
  if (conf) {
    return conf;
  }
  if (globalThis.__tamaguiConfig) {
    if (process.env.NODE_ENV === "development" && !globalThis.__tamaguiHasWarnedGlobalFallback && !globalThis.__tamaguiPendingCheck) {
      globalThis.__tamaguiPendingCheck = true;
      setTimeout(function () {
        if (!setConfigCalledByThisInstance && !globalThis.__tamaguiHasWarnedGlobalFallback) {
          globalThis.__tamaguiHasWarnedGlobalFallback = true;
          console.warn(`\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F

Tamagui: Using global config fallback. This may indicate duplicate tamagui instances (e.g., from Vite SSR bundling). This is handled automatically, but likely causes issues!

\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F\u26A0\uFE0F`);
        }
        globalThis.__tamaguiPendingCheck = false;
      }, 500);
    }
    return globalThis.__tamaguiConfig;
  }
  return null;
};
var getSetting = function (key) {
  var config = getConfigFromGlobalOrLocal();
  if (process.env.NODE_ENV === "development") {
    if (!config) throw new Error(haventCalledErrorMessage);
  }
  var _config_settings_key;
  return (_config_settings_key = config.settings[key]) !== null && _config_settings_key !== void 0 ? _config_settings_key :
  // @ts-expect-error
  config[key];
};
var setConfig = function (next) {
  setConfigCalledByThisInstance = true;
  conf = next;
  globalThis.__tamaguiConfig = next;
};
var setConfigFont = function (name, font, fontParsed) {
  var config = getConfigFromGlobalOrLocal();
  if (process.env.NODE_ENV === "development") {
    if (!config) throw new Error(haventCalledErrorMessage);
  }
  config.fonts[name] = font;
  config.fontsParsed[`$${name}`] = fontParsed;
};
var getConfig = function () {
  var config = getConfigFromGlobalOrLocal();
  if (!config) {
    throw new Error(process.env.NODE_ENV !== "production" ? `Missing tamagui config, you either have a duplicate config, or haven't set it up. Be sure createTamagui is called before rendering. Also, make sure all of your tamagui dependencies are on the same version (\`tamagui\`, \`@tamagui/package-name\`, etc.) not just in your package.json, but in your lockfile.` : "Err0");
  }
  return config;
};
var getConfigMaybe = function () {
  return getConfigFromGlobalOrLocal();
};
var tokensMerged;
function setTokens(_) {
  tokensMerged = _;
}
var getTokens = function () {
  var {
    prefixed
  } = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : (
  /**
  * Force either with $ or without $ prefix
  */
  {});
  var config = getConfigFromGlobalOrLocal();
  if (process.env.NODE_ENV === "development") {
    if (!config) throw new Error(haventCalledErrorMessage);
  }
  var {
    tokens,
    tokensParsed
  } = config;
  if (prefixed === false) return tokens;
  if (prefixed === true) return tokensParsed;
  return tokensMerged;
};
var getTokenObject = function (value, group) {
  var _tokensMerged_group, _tokensMerged_;
  var config = getConfigFromGlobalOrLocal();
  var _config_specificTokens_value;
  return (_config_specificTokens_value = config.specificTokens[value]) !== null && _config_specificTokens_value !== void 0 ? _config_specificTokens_value : group ? (_tokensMerged_group = tokensMerged[group]) === null || _tokensMerged_group === void 0 ? void 0 : _tokensMerged_group[value] : (_tokensMerged_ = tokensMerged[Object.keys(tokensMerged).find(function (cat) {
    return tokensMerged[cat][value];
  }) || ""]) === null || _tokensMerged_ === void 0 ? void 0 : _tokensMerged_[value];
};
var getToken = function (value, group) {
  var useVariable = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : isWeb;
  var token = getTokenObject(value, group);
  return useVariable ? token === null || token === void 0 ? void 0 : token.variable : token === null || token === void 0 ? void 0 : token.val;
};
var getTokenValue = function (value, group) {
  if (value === "unset" || value === "auto") return;
  return getToken(value, group, false);
};
var useTokens = getTokens;
var getThemes = function () {
  return getConfigFromGlobalOrLocal().themes;
};
var updateConfig = function (key, value) {
  var config = getConfigFromGlobalOrLocal();
  Object.assign(config[key], value);
};
var getFont = function (name) {
  var _Object_entries_find;
  var conf2 = getConfig();
  var _conf_fontsParsed_name;
  return (_conf_fontsParsed_name = conf2.fontsParsed[name]) !== null && _conf_fontsParsed_name !== void 0 ? _conf_fontsParsed_name : (_Object_entries_find = Object.entries(conf2.fontsParsed).find(function (param) {
    var [k] = param;
    var _conf_fontsParsed_k_family, _conf_fontsParsed_k;
    return ((_conf_fontsParsed_k = conf2.fontsParsed[k]) === null || _conf_fontsParsed_k === void 0 ? void 0 : (_conf_fontsParsed_k_family = _conf_fontsParsed_k.family) === null || _conf_fontsParsed_k_family === void 0 ? void 0 : _conf_fontsParsed_k_family["val"]) === name;
  })) === null || _Object_entries_find === void 0 ? void 0 : _Object_entries_find[1];
};
var devConfig;
function setupDev(conf2) {
  if (process.env.NODE_ENV === "development") {
    devConfig = conf2;
  }
}
function loadAnimationDriver(name, driver) {
  var config = getConfigFromGlobalOrLocal();
  if (!config) {
    if (process.env.NODE_ENV === "development") {
      console.warn("loadAnimationDriver called before createTamagui");
    }
    return;
  }
  if (config.animations && !("default" in config.animations)) {
    ;
    config.animations = {
      default: config.animations
    };
  }
  if (config.animations) {
    ;
    config.animations[name] = driver;
  } else {
    ;
    config.animations = {
      default: driver,
      [name]: driver
    };
  }
}
export { devConfig, getConfig, getConfigMaybe, getFont, getSetting, getThemes, getToken, getTokenObject, getTokenValue, getTokens, loadAnimationDriver, setConfig, setConfigFont, setTokens, setupDev, updateConfig, useTokens };
