// @tamagui/group/dist/jsx/Group.mjs
import { mergeSlotStyleProps, styled } from "@tamagui/core";
import { createContextScope } from "@tamagui/create-context";
import { withStaticProperties } from "@tamagui/helpers";
import { YStack } from "@tamagui/stacks";
import React from "react";
import { useIndex, useIndexedChildren } from "./useIndexedChildren.mjs";
import { jsx } from "react/jsx-runtime";
const GROUP_NAME = "Group";
const [createGroupContext, createGroupScope] = createContextScope(GROUP_NAME);
const [GroupProvider, useGroupContext] = createGroupContext(GROUP_NAME);
const GroupFrame = styled(YStack, {
  name: "GroupFrame",
  variants: {
    unstyled: {
      false: {
        size: "$true"
      }
    },
    size: (val, {
      tokens
    }) => {
      const borderRadius = tokens.radius[val] ?? val ?? tokens.radius["$true"];
      return {
        borderRadius
      };
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
function createGroup(verticalDefault) {
  return withStaticProperties(GroupFrame.styleable((props, ref) => {
    const {
      __scopeGroup,
      children: childrenProp,
      orientation = verticalDefault ? "vertical" : "horizontal",
      disabled,
      ...restProps
    } = props;
    const vertical = orientation === "vertical";
    const indexedChildren = useIndexedChildren(React.Children.toArray(childrenProp));
    return /* @__PURE__ */jsx(GroupProvider, {
      vertical,
      disabled,
      scope: __scopeGroup,
      children: /* @__PURE__ */jsx(GroupFrame, {
        ref,
        flexDirection: vertical ? "column" : "row",
        ...restProps,
        children: indexedChildren
      })
    });
  }), {
    Item: GroupItem
  });
}
function GroupItem(props) {
  const {
    __scopeGroup,
    children,
    forcePlacement,
    ...forwardedProps
  } = props;
  const context = useGroupContext("GroupItem", __scopeGroup);
  const treeIndex = useIndex();
  if (!treeIndex) {
    throw Error("<Group.Item/> should only be used within a <Group/>");
  }
  if (!React.isValidElement(children)) {
    return children;
  }
  const isFirst = forcePlacement === "first" || forcePlacement !== "last" && treeIndex.index === 0;
  const isLast = forcePlacement === "last" || forcePlacement !== "first" && treeIndex.index === treeIndex.maxIndex;
  const radiusStyles = getZeroedRadius(isFirst, isLast, context.vertical);
  const groupProps = {
    ...forwardedProps,
    ...radiusStyles
  };
  if (context.disabled != null) {
    groupProps.disabled = children.props.disabled ?? context.disabled;
  }
  const mergedProps = mergeSlotStyleProps(groupProps, children.props);
  return React.cloneElement(children, mergedProps);
}
const useGroupItem = (childrenProps, forcePlacement, __scopeGroup) => {
  const treeIndex = useIndex();
  const context = useGroupContext("GroupItem", __scopeGroup);
  if (!treeIndex) {
    throw Error("useGroupItem should only be used within a <Group/>");
  }
  const isFirst = forcePlacement === "first" || forcePlacement !== "last" && treeIndex.index === 0;
  const isLast = forcePlacement === "last" || forcePlacement !== "first" && treeIndex.index === treeIndex.maxIndex;
  const radiusStyles = getZeroedRadius(isFirst, isLast, context.vertical);
  return {
    disabled: childrenProps.disabled ?? context.disabled,
    ...radiusStyles
  };
};
const Group = createGroup(true);
const YGroup = Group;
const XGroup = createGroup(false);
function getZeroedRadius(isFirst, isLast, vertical) {
  if (vertical) {
    return {
      ...(isFirst ? null : {
        borderTopLeftRadius: 0,
        borderTopRightRadius: 0
      }),
      ...(isLast ? null : {
        borderBottomLeftRadius: 0,
        borderBottomRightRadius: 0
      })
    };
  }
  return {
    ...(isFirst ? null : {
      borderTopLeftRadius: 0,
      borderBottomLeftRadius: 0
    }),
    ...(isLast ? null : {
      borderTopRightRadius: 0,
      borderBottomRightRadius: 0
    })
  };
}
export { Group, GroupFrame, XGroup, YGroup, createGroupScope, useGroupItem };
