// @tamagui/group/dist/jsx/useIndexedChildren.mjs
import * as React from "react";
import { jsx } from "react/jsx-runtime";
const MaxIndexContext = React.createContext([]);
const IndexContext = React.createContext(null);
function useIndex() {
  const maxIndexPath = React.useContext(MaxIndexContext);
  const indexPathString = React.useContext(IndexContext);
  return React.useMemo(() => {
    if (indexPathString === null) {
      return null;
    }
    const indexPath = parseIndexPath(indexPathString);
    const maxIndex = maxIndexPath[maxIndexPath.length - 1];
    const index = indexPath[indexPath.length - 1];
    return {
      maxIndex,
      maxIndexPath,
      index,
      indexPath,
      indexPathString,
      isFirst: index === 0,
      isLast: index === maxIndex,
      isEven: index % 2 === 0,
      isOdd: Math.abs(index % 2) === 1
    };
  }, [maxIndexPath, indexPathString]);
}
function useIndexedChildren(children) {
  const parentMaxIndexPath = React.useContext(MaxIndexContext);
  const indexPathString = React.useContext(IndexContext);
  const childrenCount = React.Children.count(children);
  const maxIndexPath = React.useMemo(() => parentMaxIndexPath.concat(childrenCount - 1), [childrenCount]);
  return /* @__PURE__ */jsx(MaxIndexContext.Provider, {
    value: maxIndexPath,
    children: React.Children.map(children, (child, index) => React.isValidElement(child) ? /* @__PURE__ */jsx(IndexContext.Provider, {
      value: indexPathString ? `${indexPathString}.${index.toString()}` : index.toString(),
      children: child
    }, child.key) : child)
  });
}
function parseIndexPath(indexPathString) {
  return indexPathString.split(".").map(index => Number.parseInt(index, 10));
}
export { useIndex, useIndexedChildren };
