// @tamagui/image/dist/esm/createImage.mjs
import { getTokenValue, styled } from "@tamagui/web";
import { jsx } from "react/jsx-runtime";
const defaultObjectFitMap = objectFit => {
  switch (objectFit) {
    case "fill":
      return "stretch";
    case "none":
      return "center";
    case "scale-down":
      return "contain";
    case "contain":
      return "contain";
    default:
      return "cover";
  }
};
const defaultTransformSource = props => {
  const {
    src,
    source,
    width,
    height
  } = props;
  if (source) return source;
  if (src && typeof src !== "string") return src;
  return {
    uri: src,
    width,
    height
  };
};
function createImage(options) {
  const {
    Component,
    mapObjectFitToResizeMode = defaultObjectFitMap,
    resizeModePropName = "resizeMode",
    objectPositionPropName,
    transformSource = defaultTransformSource
  } = options;
  const inlinePropsSet = /* @__PURE__ */new Set(["source", "placeholder", "transition", "contentFit", "contentPosition", "cachePolicy", "recyclingKey", "allowDownscaling", "autoplay", "blurRadius", "priority", "placeholderContentFit", "responsivePolicy", "onLoadStart", "onProgress", "onLoadEnd"]);
  const StyledImage = styled(Component, {
    name: "Image"
  }, {
    inlineProps: inlinePropsSet
  });
  const ImageComponent = StyledImage.styleable((incomingProps, ref) => {
    const props = incomingProps;
    const {
      src,
      width,
      height,
      borderRadius,
      objectFit,
      objectPosition,
      // web only props - filter out on native
      decoding,
      elementTiming,
      fetchpriority,
      isMap,
      loading,
      sizes,
      useMap,
      onLoad,
      onError,
      ...rest
    } = props;
    const resolvedWidth = typeof width === "string" && width[0] === "$" ? getTokenValue(width, "size") : width;
    const resolvedHeight = typeof height === "string" && height[0] === "$" ? getTokenValue(height, "size") : height;
    const finalSource = transformSource({
      src,
      width: resolvedWidth,
      height: resolvedHeight
    });
    const incomingStyle = Array.isArray(rest.style) ? Object.assign({}, ...rest.style.flat()) : rest.style;
    const resolvedBorderRadius = typeof borderRadius === "string" && borderRadius[0] === "$" ? getTokenValue(borderRadius, "radius") : borderRadius;
    const finalProps = {
      ...rest,
      source: finalSource,
      style: {
        ...incomingStyle,
        ...(resolvedBorderRadius !== void 0 && {
          borderRadius: resolvedBorderRadius
        }),
        ...(resolvedWidth !== void 0 && {
          width: resolvedWidth
        }),
        ...(resolvedHeight !== void 0 && {
          height: resolvedHeight
        })
      }
    };
    if (objectFit) {
      finalProps[resizeModePropName] = mapObjectFitToResizeMode(objectFit);
    }
    if (objectPositionPropName && objectPosition) {
      finalProps[objectPositionPropName] = objectPosition;
    }
    if (onLoad) {
      finalProps.onLoad = e => {
        const source = e?.nativeEvent?.source || e?.source || {};
        onLoad({
          target: {
            naturalHeight: source?.height,
            naturalWidth: source?.width
          },
          type: "load"
        });
      };
    }
    if (onError) {
      finalProps.onError = () => {
        onError({
          type: "error"
        });
      };
    }
    return /* @__PURE__ */jsx(Component, {
      ref,
      ...finalProps
    });
  });
  const comp = Component;
  ImageComponent.getSize = comp.getSize || (() => {});
  ImageComponent.getSizeWithHeaders = comp.getSizeWithHeaders || (() => {});
  ImageComponent.prefetch = comp.prefetch || (() => {});
  ImageComponent.prefetchWithMetadata = comp.prefetchWithMetadata || (() => {});
  ImageComponent.abortPrefetch = comp.abortPrefetch || (() => {});
  ImageComponent.queryCache = comp.queryCache || (() => {});
  return ImageComponent;
}
export { createImage };
