// @tamagui/timer/dist/esm/index.native.js
function timer() {
  var runs = 0;
  var typesOfRuns = /* @__PURE__ */new Set();
  var timings = {};
  function print() {
    var typeRuns = runs / typesOfRuns.size;
    var totalTime = 0;
    var out = [`Ran ${typeRuns} per-type, ${runs} total`, ...[...typesOfRuns].map(function (name) {
      if (name.endsWith("(ignore)")) {
        return;
      }
      var avg = `avg ${`${timings[name] / typeRuns}`.slice(0, 9).padEnd(9)}ms`;
      var total = timings[name];
      totalTime += total;
      return `${name.slice(0, 30).padStart(31)} | ${avg} | total ${total}ms`;
    }), `                                    total ${totalTime}ms`].join("\n");
    console.info(out);
    return out;
  }
  return {
    start(opts) {
      var _opts_quiet;
      var quiet = (_opts_quiet = opts === null || opts === void 0 ? void 0 : opts.quiet) !== null && _opts_quiet !== void 0 ? _opts_quiet : true;
      function time(strings) {
        for (var _len = arguments.length, vars = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          vars[_key - 1] = arguments[_key];
        }
        var _timings, _tag;
        var elapsed = performance.now() - start;
        var tag = templateToString(strings, ...vars);
        typesOfRuns.add(tag);
        runs++;
        var _;
        (_ = (_timings = timings)[_tag = tag]) !== null && _ !== void 0 ? _ : _timings[_tag] = 0;
        timings[tag] += elapsed;
        if (!quiet) {
          var result = "";
          strings.forEach(function (str, i) {
            result += `${str}${i === strings.length - 1 ? "" : vars[i]}`;
          });
          console.info(`${`${elapsed}ms`.slice(0, 6).padStart(7)} |`, result);
        }
        start = performance.now();
      }
      var start = performance.now();
      time["print"] = print;
      return time;
    },
    profile() {
      return {
        timings,
        runs
      };
    },
    print
  };
}
function templateToString(strings) {
  for (var _len = arguments.length, vars = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    vars[_key - 1] = arguments[_key];
  }
  return strings.reduce(function (result, str, i) {
    return result + str + (vars[i] !== void 0 ? vars[i] : "");
  }, "");
}
export { timer };
