// @tamagui/sheet/dist/cjs/createSheet.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createSheet_exports = {};
__export(createSheet_exports, {
  createSheet: () => createSheet
});
module.exports = __toCommonJS(createSheet_exports);
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_helpers = require("@tamagui/helpers");
var import_portal = require("@tamagui/portal");
var import_remove_scroll = require("@tamagui/remove-scroll");
var import_use_did_finish_ssr = require("@tamagui/use-did-finish-ssr");
var import_z_index_stack = require("@tamagui/z-index-stack");
var import_react = require("react");
var import_react_native = require("react-native-web");
var import_constants2 = require("./constants.cjs");
require("./nativeSheet.cjs");
var import_SheetContext = require("./SheetContext.cjs");
var import_SheetImplementationCustom = require("./SheetImplementationCustom.cjs");
var import_SheetScrollView = require("./SheetScrollView.cjs");
var import_useSheetController = require("./useSheetController.cjs");
var import_useSheetOffscreenSize = require("./useSheetOffscreenSize.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function createSheet({
  Handle,
  Frame,
  Overlay
}) {
  const SheetHandle = Handle.styleable(({
    __scopeSheet,
    ...props
  }, forwardedRef) => {
    const context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_HANDLE_NAME, __scopeSheet);
    const composedRef = (0, import_compose_refs.useComposedRefs)(context.handleRef, forwardedRef);
    const wasDraggingRef = (0, import_react.useRef)(false);
    (0, import_react.useEffect)(() => {
      if (!context.scrollBridge) return;
      return context.scrollBridge.onParentDragging(isDragging => {
        if (isDragging) wasDraggingRef.current = true;
      });
    }, [context.scrollBridge]);
    if (context.onlyShowFrame) return null;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Handle, {
      ref: composedRef,
      onPressIn: () => {
        wasDraggingRef.current = false;
      },
      onPress: () => {
        if (wasDraggingRef.current) {
          wasDraggingRef.current = false;
          return;
        }
        const max = context.snapPoints.length + (context.dismissOnSnapToBottom ? -1 : 0);
        const nextPos = (context.position + 1) % max;
        context.setPosition(nextPos);
      },
      open: context.open,
      ...props
    });
  });
  const SheetOverlay = Overlay.styleable((propsIn, ref) => {
    const {
      __scopeSheet,
      ...props
    } = propsIn;
    const context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_OVERLAY_NAME, __scopeSheet);
    const element = (0, import_react.useMemo)(() => {
      return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Overlay, {
        ...props,
        onPress: (0, import_helpers.composeEventHandlers)(props.onPress, context.dismissOnOverlayPress ? () => {
          context.setOpen(false);
        } : void 0)
      });
    }, [props.onPress, props.opacity, context.dismissOnOverlayPress]);
    (0, import_constants.useIsomorphicLayoutEffect)(() => {
      context.onOverlayComponent?.(element);
    }, [element]);
    if (context.onlyShowFrame) return null;
    return null;
  });
  const SheetFrame = Frame.styleable(({
    __scopeSheet,
    adjustPaddingForOffscreenContent,
    disableHideBottomOverflow,
    children,
    ...props
  }, forwardedRef) => {
    const context = (0, import_SheetContext.useSheetContext)(import_constants2.SHEET_NAME, __scopeSheet);
    const {
      hasFit,
      disableRemoveScroll,
      frameSize,
      contentRef,
      open
    } = context;
    const composedContentRef = (0, import_compose_refs.useComposedRefs)(forwardedRef, contentRef);
    const offscreenSize = (0, import_useSheetOffscreenSize.useSheetOffscreenSize)(context);
    const stableFrameSize = (0, import_react.useRef)(frameSize);
    (0, import_react.useEffect)(() => {
      if (open && frameSize) stableFrameSize.current = frameSize;
    }, [open, frameSize]);
    const sheetContents = (0, import_react.useMemo)(() => {
      const shouldUseFixedHeight = hasFit && !open && stableFrameSize.current;
      return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(Frame, {
        ref: composedContentRef,
        flex: hasFit && open ? 0 : 1,
        flexBasis: hasFit ? "auto" : void 0,
        height: shouldUseFixedHeight ? stableFrameSize.current : hasFit ? void 0 : frameSize,
        pointerEvents: open ? "auto" : "none",
        "data-state": open ? "open" : "closed",
        ...props,
        children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_z_index_stack.StackZIndexContext, {
          zIndex: (0, import_portal.resolveViewZIndex)(props.zIndex),
          children
        }), adjustPaddingForOffscreenContent && /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.View, {
          "data-sheet-offscreen-pad": true,
          height: offscreenSize,
          width: "100%"
        })]
      });
    }, [open, props, frameSize, offscreenSize, adjustPaddingForOffscreenContent, hasFit]);
    return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
      children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_remove_scroll.RemoveScroll, {
        enabled: !disableRemoveScroll && context.open,
        children: sheetContents
      }), !disableHideBottomOverflow && /* @__PURE__ */(0, import_jsx_runtime.jsx)(Frame, {
        ...props,
        componentName: "SheetCover",
        children: null,
        testID: void 0,
        id: void 0,
        position: "absolute",
        bottom: "-100%",
        zIndex: -1,
        height: context.frameSize,
        left: 0,
        right: 0,
        borderWidth: 0,
        borderRadius: 0,
        shadowOpacity: 0
      })]
    });
  });
  const Sheet = (0, import_react.forwardRef)(function Sheet2(props, ref) {
    const hydrated = (0, import_use_did_finish_ssr.useDidFinishSSR)();
    const {
      isShowingNonSheet
    } = (0, import_useSheetController.useSheetController)();
    let SheetImplementation = import_SheetImplementationCustom.SheetImplementationCustom;
    if (props.native && import_react_native.Platform.OS === "ios") {}
    if (isShowingNonSheet || !hydrated) return null;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(SheetImplementation, {
      ref,
      ...props
    });
  });
  const components = {
    Frame: SheetFrame,
    Overlay: SheetOverlay,
    Handle: SheetHandle,
    ScrollView: import_SheetScrollView.SheetScrollView
  };
  const Controlled = (0, import_helpers.withStaticProperties)(Sheet, components);
  return (0, import_helpers.withStaticProperties)(Sheet, {
    ...components,
    Controlled
  });
}
