// @tamagui/sheet/dist/cjs/useKeyboardControllerSheet.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useKeyboardControllerSheet_exports = {};
__export(useKeyboardControllerSheet_exports, {
  useKeyboardControllerSheet: () => useKeyboardControllerSheet
});
module.exports = __toCommonJS(useKeyboardControllerSheet_exports);
var import_react = require("react");
const noop = () => {};
function useKeyboardControllerSheet(_options) {
  const pauseKeyboardHandler = (0, import_react.useRef)(false);
  return {
    keyboardControllerEnabled: false,
    keyboardHeight: 0,
    isKeyboardVisible: false,
    dismissKeyboard: noop,
    pauseKeyboardHandler,
    flushPendingHide: noop
  };
}
