// @tamagui/stacks/dist/cjs/Stacks.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Stacks_exports = {};
__export(Stacks_exports, {
  XStack: () => XStack,
  YStack: () => YStack,
  ZStack: () => ZStack,
  fullscreenStyle: () => fullscreenStyle
});
module.exports = __toCommonJS(Stacks_exports);
var import_core = require("@tamagui/core");
var import_getElevation = require("./getElevation.native.js");
var fullscreenStyle = {
  position: "absolute",
  inset: 0
};
var variants = {
  fullscreen: {
    true: fullscreenStyle
  },
  elevation: {
    "...size": import_getElevation.getElevation,
    ":number": import_getElevation.getElevation
  }
};
var YStack = (0, import_core.styled)(import_core.View, {
  flexDirection: "column",
  variants
});
YStack["displayName"] = "YStack";
var XStack = (0, import_core.styled)(import_core.View, {
  flexDirection: "row",
  variants
});
XStack["displayName"] = "XStack";
var ZStack = (0, import_core.styled)(YStack, {
  position: "relative"
}, {
  neverFlatten: true,
  isZStack: true
});
ZStack["displayName"] = "ZStack";
