// @tamagui/stacks/dist/cjs/ThemeableStack.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var ThemeableStack_exports = {};
__export(ThemeableStack_exports, {
  ThemeableStack: () => ThemeableStack,
  themeableVariants: () => themeableVariants
});
module.exports = __toCommonJS(ThemeableStack_exports);
var import_core = require("@tamagui/core");
var import_Stacks = require("./Stacks.cjs");
var import_variants = require("./variants.cjs");
const chromelessStyle = {
  backgroundColor: "transparent",
  borderColor: "transparent",
  shadowColor: "transparent",
  hoverStyle: {
    borderColor: "transparent"
  }
};
const themeableVariants = {
  circular: import_variants.circular,
  elevate: import_variants.elevate,
  bordered: {
    true: import_variants.bordered
  },
  transparent: {
    true: {
      backgroundColor: "transparent"
    }
  },
  chromeless: {
    true: chromelessStyle,
    all: {
      ...chromelessStyle,
      hoverStyle: chromelessStyle,
      pressStyle: chromelessStyle,
      focusStyle: chromelessStyle
    }
  }
};
const ThemeableStack = (0, import_core.styled)(import_Stacks.YStack, {
  variants: themeableVariants
});
