// @tamagui/progress/dist/jsx/Progress.mjs
import { getVariableValue, isWeb, styled } from "@tamagui/core";
import { createContextScope } from "@tamagui/create-context";
import { getSize } from "@tamagui/get-token";
import { withStaticProperties } from "@tamagui/helpers";
import { YStack } from "@tamagui/stacks";
import { useState } from "react";
import { jsx } from "react/jsx-runtime";
const PROGRESS_NAME = "Progress";
const [createProgressContext, createProgressScope] = createContextScope(PROGRESS_NAME);
const [ProgressProvider, useProgressContext] = createProgressContext(PROGRESS_NAME);
const INDICATOR_NAME = "ProgressIndicator";
const ProgressIndicatorFrame = styled(YStack, {
  name: INDICATOR_NAME,
  variants: {
    unstyled: {
      false: {
        height: "100%",
        width: "100%",
        backgroundColor: "$background"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const ProgressIndicator = ProgressIndicatorFrame.styleable(function ProgressIndicator2(props, forwardedRef) {
  const {
    __scopeProgress,
    transition,
    ...indicatorProps
  } = props;
  const context = useProgressContext(INDICATOR_NAME, __scopeProgress);
  const progressRatio = (context.value ?? 0) / context.max;
  let x;
  if (isWeb) {
    x = `${-100 + progressRatio * 50}%`;
  } else {
    const baseWidth = context.width || 0;
    x = Math.ceil(-baseWidth * (2 - progressRatio));
  }
  return /* @__PURE__ */jsx(ProgressIndicatorFrame, {
    "data-state": getProgressState(context.value, context.max),
    "data-value": context.value ?? void 0,
    "data-max": context.max,
    x,
    width: "200%",
    ...(!props.unstyled && {
      animateOnly: ["transform"],
      // on native, hide until we have width measurement
      ...(!isWeb && context.width === 0 && {
        opacity: 0
      })
    }),
    ...indicatorProps,
    ref: forwardedRef,
    transition: !isWeb && !context.width ? null : transition
  });
});
function defaultGetValueLabel(value, max) {
  return `${Math.round(value / max * 100)}%`;
}
function getProgressState(value, maxValue) {
  return value == null ? "indeterminate" : value === maxValue ? "complete" : "loading";
}
function isNumber(value) {
  return typeof value === "number";
}
function isValidMaxNumber(max) {
  return isNumber(max) && !Number.isNaN(max) && max > 0;
}
function isValidValueNumber(value, max) {
  return isNumber(value) && !Number.isNaN(value) && value <= max && value >= 0;
}
const DEFAULT_MAX = 100;
const ProgressFrame = styled(YStack, {
  name: "Progress",
  variants: {
    unstyled: {
      false: {
        borderRadius: 1e5,
        overflow: "hidden",
        backgroundColor: "$background"
      }
    },
    size: {
      "...size": val => {
        const size = Math.round(getVariableValue(getSize(val)) * 0.25);
        return {
          height: size,
          minWidth: getVariableValue(size) * 20,
          width: "100%"
        };
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const Progress = withStaticProperties(ProgressFrame.styleable(function Progress2(props, forwardedRef) {
  const {
    // @ts-expect-error
    __scopeProgress,
    value: valueProp,
    max: maxProp,
    getValueLabel = defaultGetValueLabel,
    size = "$true",
    ...progressProps
  } = props;
  const max = isValidMaxNumber(maxProp) ? maxProp : DEFAULT_MAX;
  const value = isValidValueNumber(valueProp, max) ? Math.round(valueProp) : null;
  const valueLabel = isNumber(value) ? getValueLabel(value, max) : void 0;
  const [width, setWidth] = useState(0);
  return /* @__PURE__ */jsx(ProgressProvider, {
    scope: __scopeProgress,
    value,
    max,
    width,
    children: /* @__PURE__ */jsx(ProgressFrame, {
      "aria-valuemax": max,
      "aria-valuemin": 0,
      "aria-valuenow": isNumber(value) ? value : void 0,
      "aria-valuetext": valueLabel,
      role: "progressbar",
      "data-state": getProgressState(value, max),
      "data-value": value ?? void 0,
      "data-max": max,
      ...(progressProps.unstyled !== true && {
        size
      }),
      ...progressProps,
      ...(!isWeb && {
        onLayout: e => {
          const newWidth = Math.round(e.nativeEvent.layout.width);
          if (newWidth !== width) {
            setWidth(newWidth);
          }
          progressProps.onLayout?.(e);
        }
      }),
      ref: forwardedRef
    })
  });
}), {
  Indicator: ProgressIndicator
});
export { Progress, ProgressFrame, ProgressIndicator, ProgressIndicatorFrame, createProgressScope };
