// @tamagui/form/dist/jsx/Form.mjs
import { View, createStyledContext, styled } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { jsx } from "react/jsx-runtime";
const FORM_NAME = "Form";
const FormFrame = styled(View, {
  name: FORM_NAME,
  render: "form"
});
const FormContext = createStyledContext({
  onSubmit: void 0
});
const {
  useStyledContext: useFormContext,
  Provider: FormProvider
} = FormContext;
const FormTriggerFrame = styled(View, {
  name: "FormTrigger"
});
const FormTrigger = FormTriggerFrame.styleable((props, forwardedRef) => {
  const {
    scope,
    children,
    onPress,
    ...triggerProps
  } = props;
  const context = useFormContext(scope);
  return /* @__PURE__ */jsx(FormTriggerFrame, {
    role: "button",
    ...triggerProps,
    ref: forwardedRef,
    onPress: composeEventHandlers(onPress, context.onSubmit),
    children
  });
});
const FormComponent = FormFrame.styleable(function Form({
  scope,
  onSubmit,
  ...props
}, ref) {
  return /* @__PURE__ */jsx(FormProvider, {
    scope,
    onSubmit,
    children: /* @__PURE__ */jsx(FormFrame, {
      ref,
      ...props,
      onSubmit: e => {
        e.preventDefault();
        onSubmit?.();
      }
    })
  });
});
const Form2 = withStaticProperties(FormComponent, {
  Trigger: FormTrigger
});
export { Form2 as Form, FormContext, FormFrame, FormProvider, FormTrigger, useFormContext };
