// @tamagui/slider/dist/cjs/SliderImpl.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var SliderImpl_exports = {};
__export(SliderImpl_exports, {
  SliderFrame: () => SliderFrame,
  SliderImpl: () => SliderImpl
});
module.exports = __toCommonJS(SliderImpl_exports);
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_get_token = require("@tamagui/get-token");
var import_stacks = require("@tamagui/stacks");
var React = __toESM(require("react"), 1);
var import_react_native = require("react-native-web");
var import_constants2 = require("./constants.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
const SliderFrame = (0, import_core.styled)(import_stacks.YStack, {
  position: "relative",
  variants: {
    orientation: {
      horizontal: {},
      vertical: {}
    },
    size: (val, extras) => {
      if (!val) {
        return;
      }
      const orientation = extras.props["orientation"];
      const size = Math.round((0, import_core.getVariableValue)((0, import_get_token.getSize)(val)) / 6);
      if (orientation === "horizontal") {
        return {
          height: size,
          borderRadius: size,
          justifyContent: "center"
        };
      }
      return {
        width: size,
        borderRadius: size,
        alignItems: "center"
      };
    }
  }
});
const SliderImpl = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeSlider,
    onSlideStart,
    onSlideMove,
    onSlideEnd,
    onHomeKeyDown,
    onEndKeyDown,
    onStepKeyDown,
    children,
    ...sliderProps
  } = props;
  const context = (0, import_constants2.useSliderContext)(__scopeSlider);
  const handleResponderGrant = React.useCallback(event => {
    props.onResponderGrant?.(event);
    const target = event.target;
    const thumbIndex = context.thumbs.get(target);
    const isStartingOnThumb = thumbIndex !== void 0;
    if (import_constants.isWeb && target instanceof HTMLElement) {
      if (context.thumbs.has(target)) {
        target.focus();
      }
    }
    if (!import_constants.isWeb && isStartingOnThumb) {
      context.valueIndexToChangeRef.current = thumbIndex;
    }
    onSlideStart(event, isStartingOnThumb ? "thumb" : "track");
  }, [context, onSlideStart, props.onResponderGrant]);
  const handleResponderMove = React.useCallback(event => {
    props.onResponderMove?.(event);
    event.stopPropagation();
    onSlideMove(event);
  }, [onSlideMove, props.onResponderMove]);
  const handleResponderRelease = React.useCallback(event => {
    props.onResponderRelease?.(event);
    onSlideEnd(event);
  }, [onSlideEnd, props.onResponderRelease]);
  return (
    // wrap with plain RN View for responder events - tamagui views no longer handle responder events on web
    /* @__PURE__ */
    (0, import_jsx_runtime.jsx)(SliderFrame, {
      size: "$4",
      ref: forwardedRef,
      ...sliderProps,
      "data-orientation": sliderProps.orientation,
      ...(import_constants.isWeb && {
        onKeyDown: event => {
          if (event.key === "Home") {
            onHomeKeyDown(event);
            event.preventDefault();
          } else if (event.key === "End") {
            onEndKeyDown(event);
            event.preventDefault();
          } else if (import_constants2.PAGE_KEYS.concat(import_constants2.ARROW_KEYS).includes(event.key)) {
            onStepKeyDown(event);
            event.preventDefault();
          }
        }
      }),
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_react_native.View, {
        onMoveShouldSetResponderCapture: () => true,
        onMoveShouldSetResponder: () => true,
        onStartShouldSetResponder: () => true,
        onResponderTerminationRequest: () => false,
        onResponderGrant: handleResponderGrant,
        onResponderMove: handleResponderMove,
        onResponderRelease: handleResponderRelease,
        style: {
          inset: 0,
          position: "absolute"
        },
        children
      })
    })
  );
});
