// @tamagui/checkbox-headless/dist/cjs/useCheckbox.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useCheckbox_exports = {};
__export(useCheckbox_exports, {
  useCheckbox: () => useCheckbox
});
module.exports = __toCommonJS(useCheckbox_exports);
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_helpers = require("@tamagui/helpers");
var import_label = require("@tamagui/label");
var import_react = __toESM(require("react"), 1);
var import_BubbleInput = require("./BubbleInput.cjs");
var import_utils = require("./utils.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function useCheckbox(props, [checked, setChecked], ref) {
  const {
    labelledBy: ariaLabelledby,
    name,
    required,
    disabled,
    value = "on",
    onCheckedChange,
    ...checkboxProps
  } = props;
  const [button, setButton] = import_react.default.useState(null);
  const composedRefs = (0, import_compose_refs.useComposedRefs)(ref, setButton);
  const hasConsumerStoppedPropagationRef = import_react.default.useRef(false);
  const isFormControl = import_constants.isWeb ? button ? Boolean(button.closest("form")) : true : false;
  const labelId = (0, import_label.useLabelContext)(button);
  const labelledBy = ariaLabelledby || labelId;
  const parentKeyDown = props.onKeyDown;
  const handleKeyDown = (0, import_react.useMemo)(() => (0, import_helpers.composeEventHandlers)(parentKeyDown, event => {
    if (event.key === "Enter") event.preventDefault();
  }), [parentKeyDown]);
  const handlePress = (0, import_react.useMemo)(() => (0, import_helpers.composeEventHandlers)(props.onPress, event => {
    setChecked(prevChecked => (0, import_utils.isIndeterminate)(prevChecked) ? true : !prevChecked);
    if (isFormControl && "isPropagationStopped" in event) {
      hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
      if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
    }
  }), [isFormControl]);
  return {
    bubbleInput: import_constants.isWeb && isFormControl ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_BubbleInput.BubbleInput, {
      isHidden: true,
      control: button,
      bubbles: !hasConsumerStoppedPropagationRef.current,
      name,
      value,
      checked,
      required,
      disabled
    }) : null,
    checkboxRef: composedRefs,
    checkboxProps: {
      role: "checkbox",
      "aria-labelledby": labelledBy,
      "aria-checked": (0, import_utils.isIndeterminate)(checked) ? "mixed" : checked,
      ...checkboxProps,
      ...(import_constants.isWeb && {
        type: "button",
        value,
        "data-state": (0, import_utils.getState)(checked),
        "data-disabled": disabled ? "" : void 0,
        disabled,
        onKeyDown: disabled ? void 0 : handleKeyDown
      }),
      onPress: disabled ? void 0 : handlePress
    }
  };
}
