// @tamagui/toast/dist/jsx/ToastAnnounce.mjs
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { View, Text, styled, useEvent } from "@tamagui/core";
import { Portal } from "@tamagui/portal";
import { startTransition } from "@tamagui/start-transition";
import { VisuallyHidden } from "@tamagui/visually-hidden";
import * as React from "react";
import { useToastProviderContext } from "./ToastProvider.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
const ToastAnnounceExcludeFrame = styled(View, {
  name: "ToastAnnounceExclude"
});
const ToastAnnounceExclude = React.forwardRef((props, forwardedRef) => {
  const {
    altText,
    ...announceExcludeProps
  } = props;
  return /* @__PURE__ */jsx(ToastAnnounceExcludeFrame, {
    "data-toast-announce-exclude": "",
    "data-toast-announce-alt": altText || void 0,
    ...announceExcludeProps,
    ref: forwardedRef
  });
});
const ToastAnnounce = props => {
  const {
    scope,
    children,
    ...announceProps
  } = props;
  const context = useToastProviderContext(scope);
  const [renderAnnounceText, setRenderAnnounceText] = React.useState(false);
  const [isAnnounced, setIsAnnounced] = React.useState(false);
  useNextFrame(() => {
    startTransition(() => {
      setRenderAnnounceText(true);
    });
  });
  React.useEffect(() => {
    const timer = setTimeout(() => setIsAnnounced(true), 1e3);
    return () => clearTimeout(timer);
  }, []);
  return isAnnounced ? null : /* @__PURE__ */jsx(Portal, {
    children: /* @__PURE__ */jsx(VisuallyHidden, {
      ...announceProps,
      children: renderAnnounceText && /* @__PURE__ */jsxs(Text, {
        children: [context.label, " ", children]
      })
    })
  });
};
function useNextFrame(callback = () => {}) {
  const fn = useEvent(callback);
  useIsomorphicLayoutEffect(() => {
    let raf1 = 0;
    let raf2 = 0;
    raf1 = requestAnimationFrame(() => {
      raf2 = requestAnimationFrame(fn);
    });
    return () => {
      cancelAnimationFrame(raf1);
      cancelAnimationFrame(raf2);
    };
  }, [fn]);
}
export { ToastAnnounce, ToastAnnounceExclude };
