// @tamagui/radio-group/dist/cjs/createRadioGroup.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var createRadioGroup_exports = {};
__export(createRadioGroup_exports, {
  createRadioGroup: () => createRadioGroup
});
module.exports = __toCommonJS(createRadioGroup_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@tamagui/core");
var import_radio_headless = require("@tamagui/radio-headless");
var import_roving_focus = require("@tamagui/roving-focus");
var import_react = __toESM(require("react"), 1);
var import_RadioGroup = require("./RadioGroup.native.js");
var ensureContext = function (x) {
  if (!x.context) {
    x.context = RadioGroupContext;
  }
};
var RadioGroupContext = /* @__PURE__ */import_react.default.createContext({});
var RadioGroupItemContext = /* @__PURE__ */import_react.default.createContext({
  checked: false,
  disabled: false
});
function createRadioGroup(createProps) {
  var {
    disableActiveTheme,
    Frame = import_RadioGroup.RadioGroupFrame,
    Indicator = import_RadioGroup.RadioGroupIndicatorFrame,
    Item = import_RadioGroup.RadioGroupItemFrame
  } = createProps;
  ensureContext(Frame);
  ensureContext(Indicator);
  ensureContext(Item);
  var RadioGroupImp = Frame.styleable(function (props, ref) {
    var {
      value,
      defaultValue,
      onValueChange,
      required = false,
      disabled = false,
      name,
      native,
      accentColor,
      orientation = "vertical",
      ...rest
    } = props;
    var {
      providerValue,
      frameAttrs,
      rovingFocusGroupAttrs
    } = (0, import_radio_headless.useRadioGroup)({
      orientation,
      name,
      defaultValue,
      value,
      onValueChange,
      required,
      disabled,
      native,
      accentColor
    });
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(RadioGroupContext.Provider, {
      value: providerValue,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup, {
        ...rovingFocusGroupAttrs,
        children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_RadioGroup.RadioGroupFrame, {
          ...frameAttrs,
          ref,
          ...rest
        })
      })
    });
  });
  var RadioGroupItemImp = Item.styleable(function (props, ref) {
    var {
      value,
      labelledBy,
      onPress,
      onKeyDown,
      disabled,
      id,
      ...rest
    } = props;
    var {
      providerValue,
      bubbleInput,
      rovingFocusGroupAttrs,
      frameAttrs,
      isFormControl,
      native
    } = (0, import_radio_headless.useRadioGroupItem)({
      radioGroupContext: RadioGroupContext,
      value,
      id,
      labelledBy,
      disabled,
      onPress,
      onKeyDown
    });
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(RadioGroupItemContext.Provider, {
      value: providerValue,
      children: import_core.isWeb && native ? bubbleInput : /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
        children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_roving_focus.RovingFocusGroup.Item, {
          ...rovingFocusGroupAttrs,
          children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_RadioGroup.RadioGroupItemFrame, {
            ...frameAttrs,
            ref,
            ...rest
          })
        }), isFormControl && bubbleInput]
      })
    });
  });
  RadioGroupItemImp.displayName = "RadioGroupItem";
  var RadioIndicator = Indicator.styleable(function (props, forwardedRef) {
    var {
      forceMount,
      disabled,
      ...indicatorProps
    } = props;
    var {
      checked,
      ...useIndicatorRest
    } = (0, import_radio_headless.useRadioGroupItemIndicator)({
      radioGroupItemContext: RadioGroupItemContext,
      disabled
    });
    if (forceMount || checked) {
      return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Indicator, {
        ...useIndicatorRest,
        ref: forwardedRef,
        ...indicatorProps
      });
    }
    return null;
  });
  RadioIndicator.displayName = "RadioIndicator";
  var RadioGroup = (0, import_core.withStaticProperties)(RadioGroupImp, {
    Item: RadioGroupItemImp,
    Indicator: RadioIndicator
  });
  RadioGroup.displayName = "RadioGroup";
  return RadioGroup;
}
