// @tamagui/floating/dist/cjs/interactions/useClick.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useClick_exports = {};
__export(useClick_exports, {
  useClick: () => useClick
});
module.exports = __toCommonJS(useClick_exports);
var import_react = require("react");
var import_utils = require("./utils.native.js");
function isButtonTarget(event) {
  return (0, import_utils.isHTMLElement)(event.target) && event.target.tagName === "BUTTON";
}
function isAnchorTarget(event) {
  return (0, import_utils.isHTMLElement)(event.target) && event.target.tagName === "A";
}
function isSpaceIgnored(element) {
  return (0, import_utils.isTypeableElement)(element);
}
function useClick(context) {
  var props = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
  var {
    open,
    onOpenChange,
    dataRef,
    elements: {
      domReference
    }
  } = context;
  var {
    enabled = true,
    event: eventOption = "click",
    toggle = true,
    ignoreMouse = false,
    keyboardHandlers = true,
    stickIfOpen = true
  } = props;
  var pointerTypeRef = (0, import_react.useRef)(void 0);
  var didKeyDownRef = (0, import_react.useRef)(false);
  var reference = (0, import_react.useMemo)(function () {
    return {
      onPointerDown(event) {
        pointerTypeRef.current = event.pointerType;
      },
      onMouseDown(event) {
        var pointerType = pointerTypeRef.current;
        if (event.button !== 0) return;
        if (eventOption === "click") return;
        if ((0, import_utils.isMouseLikePointerType)(pointerType, true) && ignoreMouse) return;
        if (open && toggle && (dataRef.current.openEvent && stickIfOpen ? dataRef.current.openEvent.type === "mousedown" : true)) {
          onOpenChange(false, event.nativeEvent || event, "click");
        } else {
          event.preventDefault();
          onOpenChange(true, event.nativeEvent || event, "click");
        }
      },
      onClick(event) {
        var pointerType = pointerTypeRef.current;
        if (eventOption === "mousedown" && pointerTypeRef.current) {
          pointerTypeRef.current = void 0;
          return;
        }
        if ((0, import_utils.isMouseLikePointerType)(pointerType, true) && ignoreMouse) return;
        if (open && toggle && (dataRef.current.openEvent && stickIfOpen ? dataRef.current.openEvent.type === "click" : true)) {
          onOpenChange(false, event.nativeEvent || event, "click");
        } else {
          onOpenChange(true, event.nativeEvent || event, "click");
        }
      },
      onKeyDown(event) {
        pointerTypeRef.current = void 0;
        if (event.defaultPrevented || !keyboardHandlers || isButtonTarget(event)) {
          return;
        }
        if (event.key === " " && !isSpaceIgnored(domReference)) {
          event.preventDefault();
          didKeyDownRef.current = true;
        }
        if (isAnchorTarget(event)) {
          return;
        }
        if (event.key === "Enter") {
          if (open && toggle) {
            onOpenChange(false, event.nativeEvent || event, "click");
          } else {
            onOpenChange(true, event.nativeEvent || event, "click");
          }
        }
      },
      onKeyUp(event) {
        if (event.defaultPrevented || !keyboardHandlers || isButtonTarget(event) || isSpaceIgnored(domReference)) {
          return;
        }
        if (event.key === " " && didKeyDownRef.current) {
          didKeyDownRef.current = false;
          if (open && toggle) {
            onOpenChange(false, event.nativeEvent || event, "click");
          } else {
            onOpenChange(true, event.nativeEvent || event, "click");
          }
        }
      }
    };
  }, [dataRef, domReference, eventOption, ignoreMouse, keyboardHandlers, onOpenChange, open, stickIfOpen, toggle]);
  return (0, import_react.useMemo)(function () {
    return enabled ? {
      reference
    } : {};
  }, [enabled, reference]);
}
