// @tamagui/floating/dist/cjs/interactions/useDelayGroup.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useDelayGroup_exports = {};
__export(useDelayGroup_exports, {
  FloatingDelayGroup: () => FloatingDelayGroup,
  useDelayGroup: () => useDelayGroup,
  useDelayGroupContext: () => useDelayGroupContext
});
module.exports = __toCommonJS(useDelayGroup_exports);
var React = __toESM(require("react"), 1);
const DelayGroupContext = React.createContext({
  currentId: null,
  setCurrentId: () => {},
  delay: 0,
  timeoutMs: 0,
  initialDelay: 0
});
function useDelayGroupContext() {
  return React.useContext(DelayGroupContext);
}
function FloatingDelayGroup({
  children,
  delay,
  timeoutMs = 0
}) {
  const [currentId, setCurrentIdRaw] = React.useState(null);
  const timeoutRef = React.useRef(void 0);
  const setCurrentId = React.useCallback(id => {
    clearTimeout(timeoutRef.current);
    if (id == null && timeoutMs > 0) {
      timeoutRef.current = setTimeout(() => {
        setCurrentIdRaw(null);
      }, timeoutMs);
    } else {
      setCurrentIdRaw(id);
    }
  }, [timeoutMs]);
  React.useEffect(() => {
    return () => clearTimeout(timeoutRef.current);
  }, []);
  const value = React.useMemo(() => ({
    currentId,
    setCurrentId,
    delay,
    timeoutMs,
    initialDelay: delay
  }), [currentId, setCurrentId, delay, timeoutMs]);
  return React.createElement(DelayGroupContext.Provider, {
    value
  }, children);
}
function useDelayGroup(context, options = {}) {
  const {
    id
  } = options;
  const groupContext = React.useContext(DelayGroupContext);
  React.useEffect(() => {
    if (!context.open && groupContext.currentId === id) {
      groupContext.setCurrentId(null);
    }
  }, [context.open, id]);
  React.useEffect(() => {
    if (groupContext.currentId != null && groupContext.currentId !== id && context.open) {
      context.onOpenChange(false);
    }
  }, [groupContext.currentId, id, context.open]);
  if (groupContext.currentId != null) {
    return {
      delay: {
        open: 1,
        close: getClose(groupContext.initialDelay)
      },
      currentId: groupContext.currentId
    };
  }
  return {
    delay: groupContext.initialDelay,
    currentId: groupContext.currentId
  };
}
function getClose(delay) {
  if (typeof delay === "number") return delay;
  return delay?.close ?? 0;
}
