// @tamagui/floating/dist/cjs/interactions/useFocus.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useFocus_exports = {};
__export(useFocus_exports, {
  useFocus: () => useFocus
});
module.exports = __toCommonJS(useFocus_exports);
var import_react = require("react");
var import_utils = require("./utils.native.js");
function isMacSafari() {
  return (0, import_utils.isMac)() && (0, import_utils.isSafari)();
}
function useFocus(context) {
  var props = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
  var {
    open,
    onOpenChange,
    events,
    dataRef,
    elements
  } = context;
  var {
    enabled = true,
    visibleOnly = true
  } = props;
  var blockFocusRef = (0, import_react.useRef)(false);
  var timeoutRef = (0, import_react.useRef)(-1);
  var keyboardModalityRef = (0, import_react.useRef)(true);
  (0, import_react.useEffect)(function () {
    if (!enabled) return;
    var win = (0, import_utils.getDocument)(elements.domReference).defaultView || window;
    function onBlur() {
      if (!open && (0, import_utils.isHTMLElement)(elements.domReference) && elements.domReference === (0, import_utils.activeElement)((0, import_utils.getDocument)(elements.domReference))) {
        blockFocusRef.current = true;
      }
    }
    function onKeyDown() {
      keyboardModalityRef.current = true;
    }
    function onPointerDown() {
      keyboardModalityRef.current = false;
    }
    win.addEventListener("blur", onBlur);
    if (isMacSafari()) {
      win.addEventListener("keydown", onKeyDown, true);
      win.addEventListener("pointerdown", onPointerDown, true);
    }
    return function () {
      win.removeEventListener("blur", onBlur);
      if (isMacSafari()) {
        win.removeEventListener("keydown", onKeyDown, true);
        win.removeEventListener("pointerdown", onPointerDown, true);
      }
    };
  }, [elements.domReference, open, enabled]);
  (0, import_react.useEffect)(function () {
    if (!enabled) return;
    if (!events) return;
    function handleOpenChange(param) {
      var {
        reason
      } = param;
      if (reason === "reference-press" || reason === "escape-key") {
        blockFocusRef.current = true;
      }
    }
    events.on("openchange", handleOpenChange);
    return function () {
      events.off("openchange", handleOpenChange);
    };
  }, [events, enabled]);
  (0, import_react.useEffect)(function () {
    return function () {
      (0, import_utils.clearTimeoutIfSet)(timeoutRef);
    };
  }, []);
  var reference = (0, import_react.useMemo)(function () {
    return {
      onMouseLeave() {
        blockFocusRef.current = false;
      },
      onFocus(event) {
        if (blockFocusRef.current) return;
        var target = (0, import_utils.getTarget)(event.nativeEvent);
        if (visibleOnly && (0, import_utils.isElement)(target)) {
          if (isMacSafari() && !event.relatedTarget) {
            if (!keyboardModalityRef.current && !(0, import_utils.isTypeableElement)(target)) {
              return;
            }
          } else if (!(0, import_utils.matchesFocusVisible)(target)) {
            return;
          }
        }
        onOpenChange(true, event.nativeEvent, "focus");
      },
      onBlur(event) {
        blockFocusRef.current = false;
        var relatedTarget = event.relatedTarget;
        var nativeEvent = event.nativeEvent;
        timeoutRef.current = window.setTimeout(function () {
          var activeEl = (0, import_utils.activeElement)(elements.domReference ? elements.domReference.ownerDocument : document);
          if (!relatedTarget && activeEl === elements.domReference) return;
          if ((0, import_utils.contains)(context.refs.floating.current, activeEl) || (0, import_utils.contains)(elements.domReference, activeEl)) {
            return;
          }
          onOpenChange(false, nativeEvent, "focus");
        });
      }
    };
  }, [context.refs.floating, elements.domReference, onOpenChange, visibleOnly]);
  return (0, import_react.useMemo)(function () {
    return enabled ? {
      reference
    } : {};
  }, [enabled, reference]);
}
