// @tamagui/floating/dist/cjs/interactions/useFocus.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useFocus_exports = {};
__export(useFocus_exports, {
  useFocus: () => useFocus
});
module.exports = __toCommonJS(useFocus_exports);
var import_react = require("react");
var import_utils = require("./utils.cjs");
function isMacSafari() {
  return (0, import_utils.isMac)() && (0, import_utils.isSafari)();
}
function useFocus(context, props = {}) {
  const {
    open,
    onOpenChange,
    events,
    dataRef,
    elements
  } = context;
  const {
    enabled = true,
    visibleOnly = true
  } = props;
  const blockFocusRef = (0, import_react.useRef)(false);
  const timeoutRef = (0, import_react.useRef)(-1);
  const keyboardModalityRef = (0, import_react.useRef)(true);
  (0, import_react.useEffect)(() => {
    if (!enabled) return;
    const win = (0, import_utils.getDocument)(elements.domReference).defaultView || window;
    function onBlur() {
      if (!open && (0, import_utils.isHTMLElement)(elements.domReference) && elements.domReference === (0, import_utils.activeElement)((0, import_utils.getDocument)(elements.domReference))) {
        blockFocusRef.current = true;
      }
    }
    function onKeyDown() {
      keyboardModalityRef.current = true;
    }
    function onPointerDown() {
      keyboardModalityRef.current = false;
    }
    win.addEventListener("blur", onBlur);
    if (isMacSafari()) {
      win.addEventListener("keydown", onKeyDown, true);
      win.addEventListener("pointerdown", onPointerDown, true);
    }
    return () => {
      win.removeEventListener("blur", onBlur);
      if (isMacSafari()) {
        win.removeEventListener("keydown", onKeyDown, true);
        win.removeEventListener("pointerdown", onPointerDown, true);
      }
    };
  }, [elements.domReference, open, enabled]);
  (0, import_react.useEffect)(() => {
    if (!enabled) return;
    if (!events) return;
    function handleOpenChange({
      reason
    }) {
      if (reason === "reference-press" || reason === "escape-key") {
        blockFocusRef.current = true;
      }
    }
    events.on("openchange", handleOpenChange);
    return () => {
      events.off("openchange", handleOpenChange);
    };
  }, [events, enabled]);
  (0, import_react.useEffect)(() => {
    return () => {
      (0, import_utils.clearTimeoutIfSet)(timeoutRef);
    };
  }, []);
  const reference = (0, import_react.useMemo)(() => ({
    onMouseLeave() {
      blockFocusRef.current = false;
    },
    onFocus(event) {
      if (blockFocusRef.current) return;
      const target = (0, import_utils.getTarget)(event.nativeEvent);
      if (visibleOnly && (0, import_utils.isElement)(target)) {
        if (isMacSafari() && !event.relatedTarget) {
          if (!keyboardModalityRef.current && !(0, import_utils.isTypeableElement)(target)) {
            return;
          }
        } else if (!(0, import_utils.matchesFocusVisible)(target)) {
          return;
        }
      }
      onOpenChange(true, event.nativeEvent, "focus");
    },
    onBlur(event) {
      blockFocusRef.current = false;
      const relatedTarget = event.relatedTarget;
      const nativeEvent = event.nativeEvent;
      timeoutRef.current = window.setTimeout(() => {
        const activeEl = (0, import_utils.activeElement)(elements.domReference ? elements.domReference.ownerDocument : document);
        if (!relatedTarget && activeEl === elements.domReference) return;
        if ((0, import_utils.contains)(context.refs.floating.current, activeEl) || (0, import_utils.contains)(elements.domReference, activeEl)) {
          return;
        }
        onOpenChange(false, nativeEvent, "focus");
      });
    }
  }), [context.refs.floating, elements.domReference, onOpenChange, visibleOnly]);
  return (0, import_react.useMemo)(() => enabled ? {
    reference
  } : {}, [enabled, reference]);
}
