// @tamagui/floating/dist/cjs/interactions/useHover.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useHover_exports = {};
__export(useHover_exports, {
  getDelay: () => getDelay,
  useHover: () => useHover
});
module.exports = __toCommonJS(useHover_exports);
var React = __toESM(require("react"), 1);
var import_use_event = require("@tamagui/use-event");
var import_utils = require("./utils.cjs");
const safePolygonIdentifier = "data-floating-ui-safe-polygon";
function getDelay(value, prop, pointerType) {
  if (pointerType && !(0, import_utils.isMouseLikePointerType)(pointerType)) {
    return 0;
  }
  if (typeof value === "number") {
    return value;
  }
  return value?.[prop];
}
function useHover(context, props = {}) {
  const {
    open,
    onOpenChange,
    dataRef,
    events,
    elements
  } = context;
  const {
    enabled = true,
    delay = 0,
    handleClose = null,
    mouseOnly = false,
    restMs = 0,
    move = true
  } = props;
  const handleCloseRef = React.useRef(handleClose);
  handleCloseRef.current = handleClose;
  const delayRef = React.useRef(delay);
  delayRef.current = delay;
  const openRef = React.useRef(open);
  openRef.current = open;
  const restMsRef = React.useRef(restMs);
  restMsRef.current = restMs;
  const stableOnOpenChange = (0, import_use_event.useEvent)(onOpenChange);
  const pointerTypeRef = React.useRef(void 0);
  const timeoutRef = React.useRef(-1);
  const handlerRef = React.useRef(void 0);
  const restTimeoutRef = React.useRef(-1);
  const blockMouseMoveRef = React.useRef(true);
  const performedPointerEventsMutationRef = React.useRef(false);
  const unbindMouseMoveRef = React.useRef(() => {});
  const restTimeoutPendingRef = React.useRef(false);
  const isHoverOpen = (0, import_use_event.useEvent)(() => {
    const type = dataRef.current.openEvent?.type;
    return type?.includes("mouse") && type !== "mousedown";
  });
  React.useEffect(() => {
    if (!enabled) return;
    if (!events) return;
    function onOpenChange2({
      open: open2
    }) {
      if (!open2) {
        (0, import_utils.clearTimeoutIfSet)(timeoutRef);
        (0, import_utils.clearTimeoutIfSet)(restTimeoutRef);
        blockMouseMoveRef.current = true;
        restTimeoutPendingRef.current = false;
      }
    }
    events.on("openchange", onOpenChange2);
    return () => {
      events.off("openchange", onOpenChange2);
    };
  }, [enabled, events]);
  const closeWithDelay = (0, import_use_event.useEvent)((event, runElseBranch = true, reason = "hover") => {
    const closeDelay = getDelay(delayRef.current, "close", pointerTypeRef.current);
    if (closeDelay && !handlerRef.current) {
      (0, import_utils.clearTimeoutIfSet)(timeoutRef);
      timeoutRef.current = window.setTimeout(() => stableOnOpenChange(false, event, reason), closeDelay);
    } else if (runElseBranch) {
      (0, import_utils.clearTimeoutIfSet)(timeoutRef);
      stableOnOpenChange(false, event, reason);
    }
  });
  const cleanupMouseMoveHandler = (0, import_use_event.useEvent)(() => {
    unbindMouseMoveRef.current();
    handlerRef.current = void 0;
    if (context.handleCloseActiveRef) {
      context.handleCloseActiveRef.current = false;
    }
  });
  const clearPointerEvents = (0, import_use_event.useEvent)(() => {
    if (performedPointerEventsMutationRef.current) {
      const body = (0, import_utils.getDocument)(elements.floating).body;
      body.style.pointerEvents = "";
      body.removeAttribute(safePolygonIdentifier);
      performedPointerEventsMutationRef.current = false;
    }
  });
  const isClickLikeOpenEvent = (0, import_use_event.useEvent)(() => {
    return dataRef.current.openEvent ? ["click", "mousedown"].includes(dataRef.current.openEvent.type) : false;
  });
  React.useEffect(() => {
    if (!enabled) return;
    function onReferenceMouseEnter(event) {
      (0, import_utils.clearTimeoutIfSet)(timeoutRef);
      blockMouseMoveRef.current = false;
      if (mouseOnly && !(0, import_utils.isMouseLikePointerType)(pointerTypeRef.current) || restMsRef.current > 0 && !getDelay(delayRef.current, "open")) {
        return;
      }
      const openDelay = getDelay(delayRef.current, "open", pointerTypeRef.current);
      if (openDelay) {
        timeoutRef.current = window.setTimeout(() => {
          if (!openRef.current) {
            stableOnOpenChange(true, event, "hover");
          }
        }, openDelay);
      } else if (!open) {
        stableOnOpenChange(true, event, "hover");
      }
    }
    function onReferenceMouseLeave(event) {
      if (isClickLikeOpenEvent()) {
        clearPointerEvents();
        return;
      }
      if (context.triggerElements?.hasElement(event.relatedTarget)) {
        return;
      }
      unbindMouseMoveRef.current();
      const doc = (0, import_utils.getDocument)(elements.floating);
      (0, import_utils.clearTimeoutIfSet)(restTimeoutRef);
      restTimeoutPendingRef.current = false;
      if (handleCloseRef.current) {
        if (!open) {
          (0, import_utils.clearTimeoutIfSet)(timeoutRef);
        }
        const placement = dataRef.current.placement || "bottom";
        const reference2 = elements.domReference;
        const floating = elements.floating;
        if (!reference2 || !floating) return;
        handlerRef.current = handleCloseRef.current({
          x: event.clientX,
          y: event.clientY,
          placement,
          elements: {
            reference: reference2,
            floating,
            domReference: reference2
          },
          onClose() {
            if (context.handleCloseActiveRef) {
              context.handleCloseActiveRef.current = false;
            }
            clearPointerEvents();
            cleanupMouseMoveHandler();
            if (!isClickLikeOpenEvent()) {
              closeWithDelay(event, true, "safe-polygon");
            }
          }
        });
        if (context.handleCloseActiveRef) {
          context.handleCloseActiveRef.current = true;
        }
        const handler = handlerRef.current;
        doc.addEventListener("mousemove", handler);
        unbindMouseMoveRef.current = () => {
          doc.removeEventListener("mousemove", handler);
        };
        return;
      }
      const shouldClose = pointerTypeRef.current === "touch" ? !(0, import_utils.contains)(elements.floating, event.relatedTarget) : true;
      if (shouldClose) {
        closeWithDelay(event);
      }
    }
    function onScrollMouseLeave(event) {
      if (isClickLikeOpenEvent()) return;
      if (context.triggerElements?.hasElement(event.relatedTarget)) {
        return;
      }
      const placement = dataRef.current.placement || "bottom";
      const reference2 = elements.domReference;
      const floating = elements.floating;
      if (!reference2 || !floating) return;
      handleCloseRef.current?.({
        x: event.clientX,
        y: event.clientY,
        placement,
        elements: {
          reference: reference2,
          floating,
          domReference: reference2
        },
        onClose() {
          clearPointerEvents();
          cleanupMouseMoveHandler();
          if (!isClickLikeOpenEvent()) {
            closeWithDelay(event);
          }
        }
      })(event);
    }
    function onFloatingMouseEnter() {
      (0, import_utils.clearTimeoutIfSet)(timeoutRef);
    }
    function onFloatingMouseLeave(event) {
      if (isClickLikeOpenEvent()) return;
      if (context.triggerElements?.hasElement(event.relatedTarget)) {
        return;
      }
      closeWithDelay(event, false);
    }
    if ((0, import_utils.isElement)(elements.domReference)) {
      const reference2 = elements.domReference;
      const floating = elements.floating;
      if (open) {
        reference2.addEventListener("mouseleave", onScrollMouseLeave);
      }
      if (move) {
        reference2.addEventListener("mousemove", onReferenceMouseEnter, {
          once: true
        });
      }
      reference2.addEventListener("mouseenter", onReferenceMouseEnter);
      reference2.addEventListener("mouseleave", onReferenceMouseLeave);
      if (floating) {
        floating.addEventListener("mouseleave", onScrollMouseLeave);
        floating.addEventListener("mouseenter", onFloatingMouseEnter);
        floating.addEventListener("mouseleave", onFloatingMouseLeave);
      }
      return () => {
        if (open) {
          reference2.removeEventListener("mouseleave", onScrollMouseLeave);
        }
        if (move) {
          reference2.removeEventListener("mousemove", onReferenceMouseEnter);
        }
        reference2.removeEventListener("mouseenter", onReferenceMouseEnter);
        reference2.removeEventListener("mouseleave", onReferenceMouseLeave);
        if (floating) {
          floating.removeEventListener("mouseleave", onScrollMouseLeave);
          floating.removeEventListener("mouseenter", onFloatingMouseEnter);
          floating.removeEventListener("mouseleave", onFloatingMouseLeave);
        }
        cleanupMouseMoveHandler();
      };
    }
  }, [elements, enabled, context, mouseOnly, move, open, dataRef]);
  React.useLayoutEffect(() => {
    if (!enabled) return;
    if (open && handleCloseRef.current?.__options?.blockPointerEvents && isHoverOpen()) {
      performedPointerEventsMutationRef.current = true;
      const floatingEl = elements.floating;
      if ((0, import_utils.isElement)(elements.domReference) && floatingEl) {
        const body = (0, import_utils.getDocument)(elements.floating).body;
        body.setAttribute(safePolygonIdentifier, "");
        const ref = elements.domReference;
        body.style.pointerEvents = "none";
        ref.style.pointerEvents = "auto";
        floatingEl.style.pointerEvents = "auto";
        return () => {
          body.style.pointerEvents = "";
          ref.style.pointerEvents = "";
          floatingEl.style.pointerEvents = "";
        };
      }
    }
  }, [enabled, open, elements, isHoverOpen]);
  React.useLayoutEffect(() => {
    if (!open) {
      pointerTypeRef.current = void 0;
      restTimeoutPendingRef.current = false;
      cleanupMouseMoveHandler();
      clearPointerEvents();
    }
  }, [open]);
  React.useEffect(() => {
    return () => {
      cleanupMouseMoveHandler();
      (0, import_utils.clearTimeoutIfSet)(timeoutRef);
      (0, import_utils.clearTimeoutIfSet)(restTimeoutRef);
      clearPointerEvents();
    };
  }, [enabled, elements.domReference]);
  const reference = React.useMemo(() => {
    function setPointerRef(event) {
      pointerTypeRef.current = event.pointerType;
    }
    return {
      onPointerDown: setPointerRef,
      onPointerEnter: setPointerRef,
      onMouseMove(event) {
        const {
          nativeEvent
        } = event;
        function handleMouseMove() {
          if (!blockMouseMoveRef.current && !openRef.current) {
            stableOnOpenChange(true, nativeEvent, "hover");
          }
        }
        if (mouseOnly && !(0, import_utils.isMouseLikePointerType)(pointerTypeRef.current)) {
          return;
        }
        if (open || restMsRef.current === 0) {
          return;
        }
        if (restTimeoutPendingRef.current && event.movementX ** 2 + event.movementY ** 2 < 2) {
          return;
        }
        (0, import_utils.clearTimeoutIfSet)(restTimeoutRef);
        if (pointerTypeRef.current === "touch") {
          handleMouseMove();
        } else {
          restTimeoutPendingRef.current = true;
          restTimeoutRef.current = window.setTimeout(handleMouseMove, restMsRef.current);
        }
      }
    };
  }, [mouseOnly, open]);
  return React.useMemo(() => enabled ? {
    reference
  } : {}, [enabled, reference]);
}
