// @tamagui/floating/dist/cjs/interactions/utils.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var utils_exports = {};
__export(utils_exports, {
  activeElement: () => activeElement,
  clearTimeoutIfSet: () => clearTimeoutIfSet,
  contains: () => contains,
  enqueueFocus: () => enqueueFocus,
  findNonDisabledListIndex: () => findNonDisabledListIndex,
  getDocument: () => getDocument,
  getMaxListIndex: () => getMaxListIndex,
  getMinListIndex: () => getMinListIndex,
  getTarget: () => getTarget,
  isElement: () => isElement,
  isHTMLElement: () => isHTMLElement,
  isIndexOutOfListBounds: () => isIndexOutOfListBounds,
  isListIndexDisabled: () => isListIndexDisabled,
  isMac: () => isMac,
  isMouseLikePointerType: () => isMouseLikePointerType,
  isSafari: () => isSafari,
  isTypeableCombobox: () => isTypeableCombobox,
  isTypeableElement: () => isTypeableElement,
  isVirtualClick: () => isVirtualClick,
  isVirtualPointerEvent: () => isVirtualPointerEvent,
  matchesFocusVisible: () => matchesFocusVisible,
  stopEvent: () => stopEvent
});
module.exports = __toCommonJS(utils_exports);
function getDocument(node) {
  return node?.ownerDocument || document;
}
function contains(parent, child) {
  if (!parent || !child) return false;
  const rootNode = child.getRootNode?.();
  if (parent.contains(child)) return true;
  if (rootNode && isShadowRoot(rootNode)) {
    let next = child;
    while (next) {
      if (parent === next) return true;
      next = next.parentNode || next.host;
    }
  }
  return false;
}
function isShadowRoot(node) {
  return node instanceof ShadowRoot;
}
function getTarget(event) {
  if ("composedPath" in event) {
    return event.composedPath()[0];
  }
  return event.target;
}
function activeElement(doc) {
  let el = doc.activeElement;
  while (el?.shadowRoot?.activeElement != null) {
    el = el.shadowRoot.activeElement;
  }
  return el;
}
function isHTMLElement(value) {
  return value instanceof HTMLElement;
}
function isElement(value) {
  return value instanceof Element;
}
const TYPEABLE_SELECTOR = "input:not([type='hidden']):not([disabled]),[contenteditable]:not([contenteditable='false']),textarea:not([disabled])";
function isTypeableElement(element) {
  return isHTMLElement(element) && element.matches(TYPEABLE_SELECTOR);
}
function isTypeableCombobox(element) {
  if (!element) return false;
  return element.getAttribute("role") === "combobox" && isTypeableElement(element);
}
function getPlatform() {
  const uaData = navigator.userAgentData;
  if (uaData?.platform) return uaData.platform;
  return navigator.platform;
}
function getUserAgent() {
  const uaData = navigator.userAgentData;
  if (uaData && Array.isArray(uaData.brands)) {
    return uaData.brands.map(({
      brand,
      version
    }) => `${brand}/${version}`).join(" ");
  }
  return navigator.userAgent;
}
function isSafari() {
  return /apple/i.test(navigator.vendor);
}
function isMac() {
  return getPlatform().toLowerCase().startsWith("mac") && !navigator.maxTouchPoints;
}
function isJSDOM() {
  return getUserAgent().includes("jsdom/");
}
function matchesFocusVisible(element) {
  if (!element || isJSDOM()) return true;
  try {
    return element.matches(":focus-visible");
  } catch {
    return true;
  }
}
function isMouseLikePointerType(pointerType, strict) {
  const values = ["mouse", "pen"];
  if (!strict) {
    values.push("", void 0);
  }
  return values.includes(pointerType);
}
function clearTimeoutIfSet(timeoutRef) {
  if (timeoutRef.current !== -1) {
    clearTimeout(timeoutRef.current);
    timeoutRef.current = -1;
  }
}
function stopEvent(event) {
  event.preventDefault();
  event.stopPropagation();
}
function isVirtualClick(event) {
  if (event.mozInputSource === 0 && event.isTrusted) return true;
  if (isAndroid() && event.pointerType) {
    return event.type === "click" && event.buttons === 1;
  }
  return event.detail === 0 && !event.pointerType;
}
function isVirtualPointerEvent(event) {
  if (isJSDOM()) return false;
  return !isAndroid() && event.width === 0 && event.height === 0 || isAndroid() && event.width === 1 && event.height === 1 && event.pressure === 0 && event.detail === 0 && event.pointerType === "mouse" || event.width < 1 && event.height < 1 && event.pressure === 0 && event.detail === 0 && event.pointerType === "touch";
}
function isAndroid() {
  const re = /android/i;
  return re.test(getPlatform()) || re.test(getUserAgent());
}
let rafId = 0;
function enqueueFocus(el, options = {}) {
  const {
    preventScroll = false,
    cancelPrevious = true,
    sync = false
  } = options;
  cancelPrevious && cancelAnimationFrame(rafId);
  const exec = () => el?.focus({
    preventScroll
  });
  if (sync) {
    exec();
  } else {
    rafId = requestAnimationFrame(exec);
  }
}
function isListIndexDisabled(listRef, index, disabledIndices) {
  if (typeof disabledIndices === "function") return disabledIndices(index);
  if (disabledIndices) return disabledIndices.includes(index);
  const element = listRef.current[index];
  return element == null || element.hasAttribute("disabled") || element.getAttribute("aria-disabled") === "true";
}
function findNonDisabledListIndex(listRef, {
  startingIndex = -1,
  decrement = false,
  disabledIndices,
  amount = 1
} = {}) {
  let index = startingIndex;
  do {
    index += decrement ? -amount : amount;
  } while (index >= 0 && index <= listRef.current.length - 1 && isListIndexDisabled(listRef, index, disabledIndices));
  return index;
}
function getMinListIndex(listRef, disabledIndices) {
  return findNonDisabledListIndex(listRef, {
    disabledIndices
  });
}
function getMaxListIndex(listRef, disabledIndices) {
  return findNonDisabledListIndex(listRef, {
    decrement: true,
    startingIndex: listRef.current.length,
    disabledIndices
  });
}
function isIndexOutOfListBounds(listRef, index) {
  return index < 0 || index >= listRef.current.length;
}
