// @tamagui/slider/dist/esm/SliderImpl.mjs
import { isWeb } from "@tamagui/constants";
import { getVariableValue, styled } from "@tamagui/core";
import { getSize } from "@tamagui/get-token";
import { YStack } from "@tamagui/stacks";
import * as React from "react";
import { View } from "react-native-web";
import { ARROW_KEYS, PAGE_KEYS, useSliderContext } from "./constants.mjs";
import { jsx } from "react/jsx-runtime";
const SliderFrame = styled(YStack, {
  position: "relative",
  variants: {
    orientation: {
      horizontal: {},
      vertical: {}
    },
    size: (val, extras) => {
      if (!val) {
        return;
      }
      const orientation = extras.props["orientation"];
      const size = Math.round(getVariableValue(getSize(val)) / 6);
      if (orientation === "horizontal") {
        return {
          height: size,
          borderRadius: size,
          justifyContent: "center"
        };
      }
      return {
        width: size,
        borderRadius: size,
        alignItems: "center"
      };
    }
  }
});
const SliderImpl = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeSlider,
    onSlideStart,
    onSlideMove,
    onSlideEnd,
    onHomeKeyDown,
    onEndKeyDown,
    onStepKeyDown,
    children,
    ...sliderProps
  } = props;
  const context = useSliderContext(__scopeSlider);
  const handleResponderGrant = React.useCallback(event => {
    props.onResponderGrant?.(event);
    const target = event.target;
    const thumbIndex = context.thumbs.get(target);
    const isStartingOnThumb = thumbIndex !== void 0;
    if (isWeb && target instanceof HTMLElement) {
      if (context.thumbs.has(target)) {
        target.focus();
      }
    }
    if (!isWeb && isStartingOnThumb) {
      context.valueIndexToChangeRef.current = thumbIndex;
    }
    onSlideStart(event, isStartingOnThumb ? "thumb" : "track");
  }, [context, onSlideStart, props.onResponderGrant]);
  const handleResponderMove = React.useCallback(event => {
    props.onResponderMove?.(event);
    event.stopPropagation();
    onSlideMove(event);
  }, [onSlideMove, props.onResponderMove]);
  const handleResponderRelease = React.useCallback(event => {
    props.onResponderRelease?.(event);
    onSlideEnd(event);
  }, [onSlideEnd, props.onResponderRelease]);
  return (
    // wrap with plain RN View for responder events - tamagui views no longer handle responder events on web
    /* @__PURE__ */
    jsx(SliderFrame, {
      size: "$4",
      ref: forwardedRef,
      ...sliderProps,
      "data-orientation": sliderProps.orientation,
      ...(isWeb && {
        onKeyDown: event => {
          if (event.key === "Home") {
            onHomeKeyDown(event);
            event.preventDefault();
          } else if (event.key === "End") {
            onEndKeyDown(event);
            event.preventDefault();
          } else if (PAGE_KEYS.concat(ARROW_KEYS).includes(event.key)) {
            onStepKeyDown(event);
            event.preventDefault();
          }
        }
      }),
      children: /* @__PURE__ */jsx(View, {
        onMoveShouldSetResponderCapture: () => true,
        onMoveShouldSetResponder: () => true,
        onStartShouldSetResponder: () => true,
        onResponderTerminationRequest: () => false,
        onResponderGrant: handleResponderGrant,
        onResponderMove: handleResponderMove,
        onResponderRelease: handleResponderRelease,
        style: {
          inset: 0,
          position: "absolute"
        },
        children
      })
    })
  );
});
export { SliderFrame, SliderImpl };
