// @tamagui/slider/dist/esm/SliderImpl.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { isWeb } from "@tamagui/constants";
import { getVariableValue, styled } from "@tamagui/core";
import { getSize } from "@tamagui/get-token";
import { YStack } from "@tamagui/stacks";
import * as React from "react";
import { View } from "react-native";
import { ARROW_KEYS, PAGE_KEYS, useSliderContext } from "./constants.native.js";
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
    return !!right[Symbol.hasInstance](left);
  } else {
    return left instanceof right;
  }
}
var SliderFrame = styled(YStack, {
  position: "relative",
  variants: {
    orientation: {
      horizontal: {},
      vertical: {}
    },
    size: function (val, extras) {
      if (!val) {
        return;
      }
      var orientation = extras.props["orientation"];
      var size = Math.round(getVariableValue(getSize(val)) / 6);
      if (orientation === "horizontal") {
        return {
          height: size,
          borderRadius: size,
          justifyContent: "center"
        };
      }
      return {
        width: size,
        borderRadius: size,
        alignItems: "center"
      };
    }
  }
});
var SliderImpl = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var {
    __scopeSlider,
    onSlideStart,
    onSlideMove,
    onSlideEnd,
    onHomeKeyDown,
    onEndKeyDown,
    onStepKeyDown,
    children,
    ...sliderProps
  } = props;
  var context = useSliderContext(__scopeSlider);
  var handleResponderGrant = React.useCallback(function (event) {
    var _props_onResponderGrant;
    (_props_onResponderGrant = props.onResponderGrant) === null || _props_onResponderGrant === void 0 ? void 0 : _props_onResponderGrant.call(props, event);
    var target = event.target;
    var thumbIndex = context.thumbs.get(target);
    var isStartingOnThumb = thumbIndex !== void 0;
    if (isWeb && _instanceof(target, HTMLElement)) {
      if (context.thumbs.has(target)) {
        target.focus();
      }
    }
    if (!isWeb && isStartingOnThumb) {
      context.valueIndexToChangeRef.current = thumbIndex;
    }
    onSlideStart(event, isStartingOnThumb ? "thumb" : "track");
  }, [context, onSlideStart, props.onResponderGrant]);
  var handleResponderMove = React.useCallback(function (event) {
    var _props_onResponderMove;
    (_props_onResponderMove = props.onResponderMove) === null || _props_onResponderMove === void 0 ? void 0 : _props_onResponderMove.call(props, event);
    event.stopPropagation();
    onSlideMove(event);
  }, [onSlideMove, props.onResponderMove]);
  var handleResponderRelease = React.useCallback(function (event) {
    var _props_onResponderRelease;
    (_props_onResponderRelease = props.onResponderRelease) === null || _props_onResponderRelease === void 0 ? void 0 : _props_onResponderRelease.call(props, event);
    onSlideEnd(event);
  }, [onSlideEnd, props.onResponderRelease]);
  return (
    // wrap with plain RN View for responder events - tamagui views no longer handle responder events on web
    /* @__PURE__ */
    _jsx(SliderFrame, {
      size: "$4",
      ref: forwardedRef,
      ...sliderProps,
      "data-orientation": sliderProps.orientation,
      ...(isWeb && {
        onKeyDown: function (event) {
          if (event.key === "Home") {
            onHomeKeyDown(event);
            event.preventDefault();
          } else if (event.key === "End") {
            onEndKeyDown(event);
            event.preventDefault();
          } else if (PAGE_KEYS.concat(ARROW_KEYS).includes(event.key)) {
            onStepKeyDown(event);
            event.preventDefault();
          }
        }
      }),
      children: /* @__PURE__ */_jsx(View, {
        onMoveShouldSetResponderCapture: function () {
          return true;
        },
        onMoveShouldSetResponder: function () {
          return true;
        },
        onStartShouldSetResponder: function () {
          return true;
        },
        onResponderTerminationRequest: function () {
          return false;
        },
        onResponderGrant: handleResponderGrant,
        onResponderMove: handleResponderMove,
        onResponderRelease: handleResponderRelease,
        style: {
          inset: 0,
          position: "absolute"
        },
        children
      })
    })
  );
});
export { SliderFrame, SliderImpl };
