// @tamagui/elements/dist/jsx/index.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var index_exports = {};
__export(index_exports, {
  Article: () => Article,
  Aside: () => Aside,
  Footer: () => Footer,
  Header: () => Header,
  Main: () => Main,
  Nav: () => Nav,
  Section: () => Section
});
module.exports = __toCommonJS(index_exports);
var import_core = require("@tamagui/core");
var Section = (0, import_core.styled)(import_core.View, {
  name: "Section",
  render: "section",
  flexDirection: "column",
  role: "region"
});
var Article = (0, import_core.styled)(import_core.View, {
  name: "Article",
  render: "article",
  flexDirection: "column"
});
var Main = (0, import_core.styled)(import_core.View, {
  name: "Main",
  render: "main",
  flexDirection: "column"
});
var Header = (0, import_core.styled)(import_core.View, {
  name: "Header",
  render: "header",
  role: "banner",
  flexDirection: "column"
});
var Aside = (0, import_core.styled)(import_core.View, {
  name: "Aside",
  render: "aside",
  flexDirection: "column"
});
var Footer = (0, import_core.styled)(import_core.View, {
  name: "Footer",
  render: "footer",
  flexDirection: "column"
});
var Nav = (0, import_core.styled)(import_core.View, {
  name: "Nav",
  render: "nav",
  flexDirection: "column"
});
