// @tamagui/radio-group/dist/jsx/createRadioGroup.mjs
import { isWeb, withStaticProperties } from "@tamagui/core";
import { useRadioGroup, useRadioGroupItem, useRadioGroupItemIndicator } from "@tamagui/radio-headless";
import { RovingFocusGroup } from "@tamagui/roving-focus";
import React from "react";
import { RadioGroupFrame, RadioGroupIndicatorFrame, RadioGroupItemFrame } from "./RadioGroup.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
const ensureContext = x => {
  if (!x.context) {
    x.context = RadioGroupContext;
  }
};
const RadioGroupContext = React.createContext({});
const RadioGroupItemContext = React.createContext({
  checked: false,
  disabled: false
});
function createRadioGroup(createProps) {
  const {
    disableActiveTheme,
    Frame = RadioGroupFrame,
    Indicator = RadioGroupIndicatorFrame,
    Item = RadioGroupItemFrame
  } = createProps;
  ensureContext(Frame);
  ensureContext(Indicator);
  ensureContext(Item);
  const RadioGroupImp = Frame.styleable((props, ref) => {
    const {
      value,
      defaultValue,
      onValueChange,
      required = false,
      disabled = false,
      name,
      native,
      accentColor,
      orientation = "vertical",
      ...rest
    } = props;
    const {
      providerValue,
      frameAttrs,
      rovingFocusGroupAttrs
    } = useRadioGroup({
      orientation,
      name,
      defaultValue,
      value,
      onValueChange,
      required,
      disabled,
      native,
      accentColor
    });
    return /* @__PURE__ */jsx(RadioGroupContext.Provider, {
      value: providerValue,
      children: /* @__PURE__ */jsx(RovingFocusGroup, {
        ...rovingFocusGroupAttrs,
        children: /* @__PURE__ */jsx(RadioGroupFrame, {
          ...frameAttrs,
          ref,
          ...rest
        })
      })
    });
  });
  const RadioGroupItemImp = Item.styleable((props, ref) => {
    const {
      value,
      labelledBy,
      onPress,
      onKeyDown,
      disabled,
      id,
      ...rest
    } = props;
    const {
      providerValue,
      bubbleInput,
      rovingFocusGroupAttrs,
      frameAttrs,
      isFormControl,
      native
    } = useRadioGroupItem({
      radioGroupContext: RadioGroupContext,
      value,
      id,
      labelledBy,
      disabled,
      onPress,
      onKeyDown
    });
    return /* @__PURE__ */jsx(RadioGroupItemContext.Provider, {
      value: providerValue,
      children: isWeb && native ? bubbleInput : /* @__PURE__ */jsxs(Fragment, {
        children: [/* @__PURE__ */jsx(RovingFocusGroup.Item, {
          ...rovingFocusGroupAttrs,
          children: /* @__PURE__ */jsx(RadioGroupItemFrame, {
            ...frameAttrs,
            ref,
            ...rest
          })
        }), isFormControl && bubbleInput]
      })
    });
  });
  RadioGroupItemImp.displayName = "RadioGroupItem";
  const RadioIndicator = Indicator.styleable((props, forwardedRef) => {
    const {
      forceMount,
      disabled,
      ...indicatorProps
    } = props;
    const {
      checked,
      ...useIndicatorRest
    } = useRadioGroupItemIndicator({
      radioGroupItemContext: RadioGroupItemContext,
      disabled
    });
    if (forceMount || checked) {
      return /* @__PURE__ */jsx(Indicator, {
        ...useIndicatorRest,
        ref: forwardedRef,
        ...indicatorProps
      });
    }
    return null;
  });
  RadioIndicator.displayName = "RadioIndicator";
  const RadioGroup = withStaticProperties(RadioGroupImp, {
    Item: RadioGroupItemImp,
    Indicator: RadioIndicator
  });
  RadioGroup.displayName = "RadioGroup";
  return RadioGroup;
}
export { createRadioGroup };
