// @tamagui/adapt/dist/esm/Adapt.mjs
import { isAndroid, isIos, isTouchable, isWeb, useIsomorphicLayoutEffect } from "@tamagui/constants";
import { createStyledContext, useMedia } from "@tamagui/core";
import { withStaticProperties } from "@tamagui/helpers";
import { getPortal } from "@tamagui/native";
import { PortalHost, PortalItem } from "@tamagui/portal";
import { StackZIndexContext } from "@tamagui/z-index-stack";
import React, { createContext, useContext, useId, useMemo } from "react";
import { Fragment, jsx } from "react/jsx-runtime";
function createAdaptChildrenStore() {
  let children = null;
  const listeners = /* @__PURE__ */new Set();
  return {
    set(c) {
      children = c;
      for (const l of listeners) l();
    },
    get: () => children,
    subscribe(callback) {
      listeners.add(callback);
      return () => listeners.delete(callback);
    }
  };
}
const AdaptChildrenStoreContext = createContext(null);
const AdaptContext = createStyledContext({
  Contents: null,
  scopeName: "",
  portalName: "",
  platform: null,
  setPlatform: x => {},
  when: null,
  setWhen: () => {}
});
const LastAdaptContextScope = createContext("");
const ProvideAdaptContext = ({
  children,
  ...context
}) => {
  const scope = context.scopeName || "";
  const lastScope = useContext(LastAdaptContextScope);
  return /* @__PURE__ */jsx(LastAdaptContextScope.Provider, {
    value: lastScope || context.lastScope || "",
    children: /* @__PURE__ */jsx(AdaptContext.Provider, {
      scope,
      lastScope: lastScope || context.lastScope,
      ...context,
      children
    })
  });
};
const useAdaptContext = scope => {
  const lastScope = useContext(LastAdaptContextScope);
  const adaptScope = scope ?? lastScope;
  return AdaptContext.useStyledContext(adaptScope);
};
const AdaptPortals = /* @__PURE__ */new Map();
const AdaptParent = ({
  children,
  Contents,
  scope,
  portal
}) => {
  const id = useId();
  const portalName = `AdaptPortal${scope}${id}`;
  const childrenStoreRef = React.useRef(null);
  if (!childrenStoreRef.current) childrenStoreRef.current = createAdaptChildrenStore();
  const isTeleport = false;
  const FinalContents = useMemo(() => {
    if (Contents) return Contents;
    if (AdaptPortals.has(portalName)) return AdaptPortals.get(portalName);
    const element = () => {
      return /* @__PURE__ */jsx(PortalHost, {
        name: portalName,
        forwardProps: typeof portal === "boolean" ? void 0 : portal?.forwardProps
      }, id);
    };
    AdaptPortals.set(portalName, element);
    return element;
  }, [portalName, Contents, isTeleport]);
  useIsomorphicLayoutEffect(() => {
    AdaptPortals.set(portalName, FinalContents);
    return () => {
      AdaptPortals.delete(portalName);
    };
  }, [portalName, isTeleport]);
  const [when, setWhen] = React.useState(null);
  const [platform, setPlatform] = React.useState(null);
  return /* @__PURE__ */jsx(AdaptChildrenStoreContext.Provider, {
    value: childrenStoreRef.current,
    children: /* @__PURE__ */jsx(LastAdaptContextScope.Provider, {
      value: scope,
      children: /* @__PURE__ */jsx(ProvideAdaptContext, {
        Contents: FinalContents,
        when,
        platform,
        setPlatform,
        setWhen,
        portalName,
        scopeName: scope,
        children
      })
    })
  });
};
const AdaptContents = ({
  scope,
  ...rest
}) => {
  const context = useAdaptContext(scope);
  if (!context?.Contents) throw new Error(process.env.NODE_ENV === "production" ? `tamagui.dev/docs/intro/errors#warning-002` : `You're rendering a Tamagui <Adapt /> component without nesting it inside a parent that is able to adapt.`);
  return React.createElement(context.Contents, {
    ...rest,
    key: `stable`
  });
};
AdaptContents.shouldForwardSpace = true;
const Adapt = withStaticProperties(function Adapt2(props) {
  const {
    platform,
    when,
    children,
    scope
  } = props;
  const context = useAdaptContext(scope);
  const enabled = useAdaptIsActiveGiven(props);
  useIsomorphicLayoutEffect(() => {
    context?.setWhen?.(when || enabled);
    context?.setPlatform?.(platform || null);
  }, [when, platform, enabled, context.setWhen, context.setPlatform]);
  useIsomorphicLayoutEffect(() => {
    return () => {
      context?.setWhen?.(null);
      context?.setPlatform?.(null);
    };
  }, []);
  let output;
  if (typeof children === "function") {
    const Component = context?.Contents;
    output = children(Component ? /* @__PURE__ */jsx(Component, {}) : null);
  } else output = children;
  return /* @__PURE__ */jsx(StackZIndexContext, {
    children: !enabled ? null : output
  });
}, {
  Contents: AdaptContents
});
const AdaptPortalContents = props => {
  const isActive = useAdaptIsActive(props.scope);
  const {
    portalName
  } = useAdaptContext(props.scope);
  const childrenStore = useContext(AdaptChildrenStoreContext);
  if (!isWeb && getPortal().state.type === "teleport" && childrenStore) return /* @__PURE__ */jsx(AdaptPortalTeleport, {
    isActive,
    store: childrenStore,
    children: props.children
  });
  return /* @__PURE__ */jsx(PortalItem, {
    passThrough: !isActive,
    hostName: portalName,
    children: props.children
  });
};
function AdaptPortalTeleport({
  isActive,
  store,
  children
}) {
  useIsomorphicLayoutEffect(() => {
    if (!isActive) return;
    store.set(children);
    return () => store.set(null);
  });
  return isActive ? null : /* @__PURE__ */jsx(Fragment, {
    children
  });
}
const useAdaptIsActiveGiven = ({
  when,
  platform
}) => {
  const media = useMedia();
  if (when == null && platform == null) return false;
  if (when === true) return true;
  let enabled = false;
  if (platform === "touch") enabled = isTouchable;else if (platform === "native") enabled = !isWeb;else if (platform === "web") enabled = isWeb;else if (platform === "ios") enabled = isIos;else if (platform === "android") enabled = isAndroid;
  if (platform && enabled == false) return false;
  if (when && typeof when === "string") enabled = media[when];
  return enabled;
};
const useAdaptIsActive = scope => {
  return useAdaptIsActiveGiven(useAdaptContext(scope));
};
export { Adapt, AdaptContents, AdaptContext, AdaptParent, AdaptPortalContents, ProvideAdaptContext, useAdaptContext, useAdaptIsActive };
