// @tamagui/roving-focus/dist/cjs/RovingFocusGroup.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var RovingFocusGroup_exports = {};
__export(RovingFocusGroup_exports, {
  RovingFocusGroup: () => RovingFocusGroup,
  RovingFocusGroupItem: () => RovingFocusGroupItem
});
module.exports = __toCommonJS(RovingFocusGroup_exports);
var import_collection = require("@tamagui/collection");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_helpers = require("@tamagui/helpers");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var import_use_direction = require("@tamagui/use-direction");
var React = __toESM(require("react"), 1);
var import_jsx_runtime = require("react/jsx-runtime");
const ENTRY_FOCUS = "rovingFocusGroup.onEntryFocus";
const EVENT_OPTIONS = {
  bubbles: false,
  cancelable: true
};
const RovingFocusGroupImpl = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeRovingFocusGroup,
    orientation,
    loop = false,
    dir,
    currentTabStopId: currentTabStopIdProp,
    defaultCurrentTabStopId,
    onCurrentTabStopIdChange,
    onEntryFocus,
    asChild,
    ...groupProps
  } = props;
  const ref = React.useRef(null);
  const composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, ref);
  const direction = (0, import_use_direction.useDirection)(dir);
  const [currentTabStopId = null, setCurrentTabStopId] = (0, import_use_controllable_state.useControllableState)({
    prop: currentTabStopIdProp,
    defaultProp: defaultCurrentTabStopId ?? null,
    onChange: onCurrentTabStopIdChange
  });
  const [isTabbingBackOut, setIsTabbingBackOut] = React.useState(false);
  const handleEntryFocus = (0, import_core.useEvent)(onEntryFocus);
  const getItems = useCollection(__scopeRovingFocusGroup || ROVING_FOCUS_GROUP_CONTEXT);
  const isClickFocusRef = React.useRef(false);
  const [focusableItemsCount, setFocusableItemsCount] = React.useState(0);
  const Comp = asChild ? import_core.Slot : import_core.View;
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(RovingFocusProvider, {
    scope: __scopeRovingFocusGroup,
    orientation,
    dir: direction,
    loop,
    currentTabStopId,
    onItemFocus: React.useCallback(tabStopId => setCurrentTabStopId(tabStopId), [setCurrentTabStopId]),
    onItemShiftTab: React.useCallback(() => setIsTabbingBackOut(true), []),
    onFocusableItemAdd: React.useCallback(() => setFocusableItemsCount(prevCount => prevCount + 1), []),
    onFocusableItemRemove: React.useCallback(() => setFocusableItemsCount(prevCount => prevCount - 1), []),
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(Comp, {
      tabIndex: isTabbingBackOut || focusableItemsCount === 0 ? -1 : 0,
      "data-orientation": orientation,
      ...groupProps,
      ref: composedRefs,
      outlineStyle: "none",
      onMouseDown: (0, import_helpers.composeEventHandlers)(props.onMouseDown, () => {
        isClickFocusRef.current = true;
      }),
      onFocus: (0, import_helpers.composeEventHandlers)(props.onFocus, event => {
        const isKeyboardFocus = !isClickFocusRef.current;
        if (event.target === event.currentTarget && isKeyboardFocus && !isTabbingBackOut) {
          const entryFocusEvent = new CustomEvent(ENTRY_FOCUS, EVENT_OPTIONS);
          handleEntryFocus(entryFocusEvent);
          if (!entryFocusEvent.defaultPrevented) {
            const items = getItems().filter(item => item.focusable);
            const activeItem = items.find(item => item.active);
            const currentItem = items.find(item => item.id === currentTabStopId);
            const candidateItems = [activeItem, currentItem, ...items].filter(Boolean);
            const candidateNodes = candidateItems.map(item => item.ref.current);
            focusFirst(candidateNodes, {
              focusVisible: false
            });
          }
        }
        isClickFocusRef.current = false;
      }),
      onBlur: (0, import_helpers.composeEventHandlers)(props.onBlur, () => setIsTabbingBackOut(false))
    })
  });
});
const ITEM_NAME = "RovingFocusGroupItem";
const RovingFocusGroupItem = React.forwardRef((props, forwardedRef) => {
  const {
    __scopeRovingFocusGroup,
    focusable = true,
    active = false,
    tabStopId,
    ...itemProps
  } = props;
  const autoId = React.useId();
  const id = tabStopId || autoId;
  const context = useRovingFocusContext(__scopeRovingFocusGroup);
  const isCurrentTabStop = context.currentTabStopId === id;
  const getItems = useCollection(__scopeRovingFocusGroup || ROVING_FOCUS_GROUP_CONTEXT);
  const {
    onFocusableItemAdd,
    onFocusableItemRemove
  } = context;
  React.useEffect(() => {
    if (focusable) {
      onFocusableItemAdd();
      return () => onFocusableItemRemove();
    }
  }, [focusable, onFocusableItemAdd, onFocusableItemRemove]);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.ItemSlot, {
    scope: __scopeRovingFocusGroup || ROVING_FOCUS_GROUP_CONTEXT,
    id,
    focusable,
    active,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.View, {
      tabIndex: focusable ? 0 : -1,
      "data-orientation": context.orientation,
      ...itemProps,
      ref: forwardedRef,
      onMouseDown: (0, import_helpers.composeEventHandlers)(props.onMouseDown, event => {
        if (!focusable) event.preventDefault();else context.onItemFocus(id);
      }),
      onFocus: (0, import_helpers.composeEventHandlers)(props.onFocus, () => context.onItemFocus(id)),
      ...(import_constants.isWeb && {
        onKeyDown: (0, import_helpers.composeEventHandlers)(props.onKeyDown, event => {
          if (event.key === "Tab" && event.shiftKey) {
            context.onItemShiftTab();
            return;
          }
          if (event.target !== event.currentTarget) return;
          const focusIntent = getFocusIntent(event, context.orientation, context.dir);
          if (focusIntent !== void 0) {
            event.preventDefault();
            const items = getItems().filter(item => item.focusable);
            let candidateNodes = items.map(item => item.ref.current);
            if (focusIntent === "last") candidateNodes.reverse();else if (focusIntent === "prev" || focusIntent === "next") {
              if (focusIntent === "prev") candidateNodes.reverse();
              const currentIndex = candidateNodes.indexOf(event.currentTarget);
              candidateNodes = context.loop ? wrapArray(candidateNodes, currentIndex + 1) : candidateNodes.slice(currentIndex + 1);
            }
            setTimeout(() => focusFirst(candidateNodes, {
              focusVisible: true
            }));
          }
        })
      })
    })
  });
});
RovingFocusGroupItem.displayName = ITEM_NAME;
const GROUP_NAME = "RovingFocusGroup";
const [Collection, useCollection] = (0, import_collection.createCollection)(GROUP_NAME);
const {
  Provider: RovingFocusProvider,
  useStyledContext: useRovingFocusContext
} = (0, import_core.createStyledContext)();
const ROVING_FOCUS_GROUP_CONTEXT = "RovingFocusGroupContext";
const RovingFocusGroup = (0, import_helpers.withStaticProperties)(React.forwardRef((props, forwardedRef) => {
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.Provider, {
    scope: props.__scopeRovingFocusGroup || ROVING_FOCUS_GROUP_CONTEXT,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(Collection.Slot, {
      scope: props.__scopeRovingFocusGroup || ROVING_FOCUS_GROUP_CONTEXT,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(RovingFocusGroupImpl, {
        ...props,
        ref: forwardedRef
      })
    })
  });
}), {
  Item: RovingFocusGroupItem
});
RovingFocusGroup.displayName = GROUP_NAME;
const MAP_KEY_TO_FOCUS_INTENT = {
  ArrowLeft: "prev",
  ArrowUp: "prev",
  ArrowRight: "next",
  ArrowDown: "next",
  PageUp: "first",
  Home: "first",
  PageDown: "last",
  End: "last"
};
function getDirectionAwareKey(key, dir) {
  if (dir !== "rtl") return key;
  return key === "ArrowLeft" ? "ArrowRight" : key === "ArrowRight" ? "ArrowLeft" : key;
}
function getFocusIntent(event, orientation, dir) {
  const key = getDirectionAwareKey(event.key, dir);
  if (orientation === "vertical" && ["ArrowLeft", "ArrowRight"].includes(key)) return void 0;
  if (orientation === "horizontal" && ["ArrowUp", "ArrowDown"].includes(key)) return void 0;
  return MAP_KEY_TO_FOCUS_INTENT[key];
}
function focusFirst(candidates, options) {
  const PREVIOUSLY_FOCUSED_ELEMENT = document.activeElement;
  for (const candidate of candidates) {
    if (candidate === PREVIOUSLY_FOCUSED_ELEMENT) return;
    candidate.focus({
      focusVisible: options?.focusVisible
    });
    if (document.activeElement !== PREVIOUSLY_FOCUSED_ELEMENT) return;
  }
}
function wrapArray(array, startIndex) {
  return array.map((_, index) => array[(startIndex + index) % array.length]);
}
