// @tamagui/linear-gradient/dist/cjs/LinearGradient.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var LinearGradient_exports = {};
__export(LinearGradient_exports, {
  LinearGradient: () => LinearGradient
});
module.exports = __toCommonJS(LinearGradient_exports);
var import_core = require("@tamagui/core");
var import_stacks = require("@tamagui/stacks");
var import_linear_gradient = require("./linear-gradient.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
const LinearGradientFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: "LinearGradient",
  overflow: "hidden",
  position: "relative"
});
const LinearGradient = LinearGradientFrame.styleable((propsIn, ref) => {
  const props = (0, import_core.useProps)(propsIn);
  const {
    start,
    end,
    colors: colorsProp,
    locations,
    children,
    ...stackProps
  } = props;
  const theme = (0, import_core.useTheme)();
  let colors = props.colors?.map(c => {
    return theme[c]?.get("web") ?? c;
  }) || [];
  if (process.env.NODE_ENV !== "production") {
    if (colors.some(c => {
      const normalized = (0, import_core.normalizeColor)(c);
      if (!normalized || normalized.startsWith("$")) {
        return true;
      }
    })) {
      console.error(`LinearGradient: "colors" prop contains invalid color tokens: ${colors} fallback to default colors: ["#000", "#fff"]`);
      colors = ["#000", "#fff"];
    }
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(LinearGradientFrame, {
    ref,
    ...stackProps,
    children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_linear_gradient.LinearGradient, {
      start,
      end,
      colors,
      locations,
      style: gradientStyle
    }), children]
  });
});
const gradientStyle = {
  position: "absolute",
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  zIndex: 0
};
