// @tamagui/linear-gradient/dist/cjs/linear-gradient.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var linear_gradient_exports = {};
__export(linear_gradient_exports, {
  LinearGradient: () => LinearGradient
});
module.exports = __toCommonJS(linear_gradient_exports);
var import_core = require("@tamagui/core");
var React = __toESM(require("react"), 1);
var import_react_native = require("react-native-web");
var import_jsx_runtime = require("react/jsx-runtime");
function needsDimensionAwareAngle(start, end) {
  if (!start && !end) return false;
  const startX = start?.[0] ?? 0.5;
  const endX = end?.[0] ?? 0.5;
  return startX !== endX;
}
function LinearGradient({
  colors,
  locations,
  start,
  end,
  ...props
}) {
  const normalizedStart = start ? Array.isArray(start) ? start : [start.x, start.y] : void 0;
  const normalizedEnd = end ? Array.isArray(end) ? end : [end.x, end.y] : void 0;
  const needsLayout = needsDimensionAwareAngle(normalizedStart, normalizedEnd);
  const [{
    height,
    width
  }, setLayout] = React.useState({
    height: 1,
    width: 1
  });
  const linearGradientBackgroundImage = React.useMemo(() => {
    return getLinearGradientBackgroundImage(
    // @ts-expect-error ok
    colors, locations, normalizedStart, normalizedEnd, width, height);
  }, [colors, locations, normalizedStart, normalizedEnd, width, height]);
  if (!needsLayout) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_react_native.View, {
      ...props,
      onLayout: props.onLayout,
      style: [props.style,
      // @ts-ignore: [ts] Property 'backgroundImage' does not exist on type 'ViewStyle'.
      {
        backgroundImage: linearGradientBackgroundImage
      }]
    });
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_react_native.View, {
    ...props,
    style: [props.style,
    // @ts-ignore: [ts] Property 'backgroundImage' does not exist on type 'ViewStyle'.
    {
      backgroundImage: linearGradientBackgroundImage
    }],
    onLayout: event => {
      const {
        width: width2,
        height: height2
      } = event.nativeEvent.layout;
      setLayout(oldLayout => {
        if (width2 !== oldLayout.width || height2 !== oldLayout.height) {
          return {
            height: height2,
            width: width2
          };
        }
        return oldLayout;
      });
      if (props.onLayout) {
        props.onLayout(event);
      }
    }
  });
}
function getLinearGradientBackgroundImage(colors, locations, startPoint, endPoint, width = 1, height = 1) {
  const gradientColors = calculateGradientColors(
  // @ts-expect-error TODO fix numbers
  colors, locations);
  const angle = calculatePseudoAngle(width, height, startPoint, endPoint);
  return `linear-gradient(${angle}deg, ${gradientColors.join(", ")})`;
}
function calculatePseudoAngle(width, height, startPoint, endPoint) {
  const getControlPoints = () => {
    let correctedStartPoint = [0, 0];
    if (Array.isArray(startPoint)) {
      correctedStartPoint = [startPoint[0] != null ? startPoint[0] : 0, startPoint[1] != null ? startPoint[1] : 0];
    }
    let correctedEndPoint = [0, 1];
    if (Array.isArray(endPoint)) {
      correctedEndPoint = [endPoint[0] != null ? endPoint[0] : 0, endPoint[1] != null ? endPoint[1] : 1];
    }
    return [correctedStartPoint, correctedEndPoint];
  };
  const [start, end] = getControlPoints();
  start[0] *= width;
  end[0] *= width;
  start[1] *= height;
  end[1] *= height;
  const py = end[1] - start[1];
  const px = end[0] - start[0];
  return 90 + Math.atan2(py, px) * 180 / Math.PI;
}
function calculateGradientColors(colors, locations) {
  return colors.map((color, index) => {
    const output = (0, import_core.normalizeColor)(color);
    if (locations && locations[index] !== void 0) {
      const location = Math.max(0, Math.min(1, locations[index]));
      const percentage = location * 100;
      return `${output} ${percentage}%`;
    }
    return output;
  });
}
