// @tamagui/shapes/dist/jsx/Square.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Square_exports = {};
__export(Square_exports, {
  Square: () => Square
});
module.exports = __toCommonJS(Square_exports);
var import_stacks = require("@tamagui/stacks");
var import_web = require("@tamagui/web");
var import_getShapeSize = require("./getShapeSize.native.js");
var Square = (0, import_web.styled)(import_stacks.ThemeableStack, {
  name: "Square",
  alignItems: "center",
  justifyContent: "center",
  variants: {
    size: {
      "...size": import_getShapeSize.getShapeSize,
      ":number": import_getShapeSize.getShapeSize
    }
  }
}, {
  memo: true
});
