// @tamagui/dialog/dist/jsx/Dialog.mjs
import { Adapt, AdaptParent, AdaptPortalContents, ProvideAdaptContext, useAdaptContext, useAdaptIsActive } from "@tamagui/adapt";
import { Animate } from "@tamagui/animate";
import { composeRefs, useComposedRefs } from "@tamagui/compose-refs";
import { isWeb, useIsomorphicLayoutEffect } from "@tamagui/constants";
import { createStyledContext, getExpandedShorthand, LayoutMeasurementController, styled, Theme, useThemeName, View } from "@tamagui/core";
import { createContext } from "@tamagui/create-context";
import { Dismissable } from "@tamagui/dismissable";
import { FocusScope, FocusScopeController } from "@tamagui/focus-scope";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { needsPortalRepropagation, Portal, PortalItem, resolveViewZIndex } from "@tamagui/portal";
import { RemoveScroll } from "@tamagui/remove-scroll";
import { SheetController } from "@tamagui/sheet/controller";
import { ButtonNestingContext, ThemeableStack, YStack } from "@tamagui/stacks";
import { H2, Paragraph } from "@tamagui/text";
import { useControllableState } from "@tamagui/use-controllable-state";
import { StackZIndexContext } from "@tamagui/z-index-stack";
import * as React from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
const DialogContext = createStyledContext(
// since we always provide this we can avoid setting here
{}, "Dialog__");
const {
  useStyledContext: useDialogContext,
  Provider: DialogProvider
} = DialogContext;
const DialogTriggerFrame = styled(View, {
  name: "DialogTrigger"
});
const DialogTrigger = DialogTriggerFrame.styleable(function DialogTrigger2(props, forwardedRef) {
  const {
    scope,
    ...triggerProps
  } = props;
  const isInsideButton = React.useContext(ButtonNestingContext);
  const context = useDialogContext(scope);
  const composedTriggerRef = useComposedRefs(forwardedRef, context.triggerRef);
  return /* @__PURE__ */jsx(ButtonNestingContext.Provider, {
    value: true,
    children: /* @__PURE__ */jsx(DialogTriggerFrame, {
      render: isInsideButton ? "span" : "button",
      "aria-haspopup": "dialog",
      "aria-expanded": context.open,
      "aria-controls": context.contentId,
      "data-state": getState(context.open),
      ...triggerProps,
      ref: composedTriggerRef,
      onPress: composeEventHandlers(props.onPress, context.onOpenToggle)
    })
  });
});
const DialogPortalFrame = styled(YStack, {
  pointerEvents: "none",
  render: "dialog",
  variants: {
    unstyled: {
      false: {
        alignItems: "center",
        justifyContent: "center",
        fullscreen: true,
        "$platform-web": {
          // undo dialog styles
          borderWidth: 0,
          backgroundColor: "transparent",
          color: "inherit",
          maxInlineSize: "none",
          margin: 0,
          width: "auto",
          height: "auto",
          // ensure always in frame and right height
          maxHeight: "100vh",
          position: "fixed",
          // ensure dialog inherits stacking context from portal wrapper
          zIndex: 1
        }
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const needsRepropagation = needsPortalRepropagation();
const DialogPortalItem = ({
  context,
  children
}) => {
  const themeName = useThemeName();
  const isAdapted = useAdaptIsActive(context.adaptScope);
  const adaptContext = useAdaptContext(context.adaptScope);
  let content = /* @__PURE__ */jsx(Theme, {
    name: themeName,
    children
  });
  if (needsRepropagation) {
    content = /* @__PURE__ */jsx(ProvideAdaptContext, {
      ...adaptContext,
      children: /* @__PURE__ */jsx(DialogProvider, {
        ...context,
        children: content
      })
    });
  }
  return isAdapted ? /* @__PURE__ */jsx(AdaptPortalContents, {
    scope: context.adaptScope,
    children: content
  }) : context.modal ? /* @__PURE__ */jsx(PortalItem, {
    hostName: context.modal ? "root" : context.adaptScope,
    children: content
  }) : content;
};
const DialogPortal = React.forwardRef((props, forwardRef) => {
  const {
    scope,
    forceMount,
    children,
    ...frameProps
  } = props;
  const dialogRef = React.useRef(null);
  const ref = composeRefs(dialogRef, forwardRef);
  const context = useDialogContext(scope);
  const keepMounted = forceMount || context.keepChildrenMounted;
  const isAdapted = useAdaptIsActive(context.adaptScope);
  const [isFullyHidden, setIsFullyHidden] = React.useState(!context.open);
  if (context.open && isFullyHidden) {
    setIsFullyHidden(false);
  }
  const isVisible = context.open ? true : !isFullyHidden;
  if (isWeb) {
    useIsomorphicLayoutEffect(() => {
      const node = dialogRef.current;
      if (!(node instanceof HTMLDialogElement)) return;
      if (isVisible) {
        node.show?.();
      } else {
        node.close?.();
      }
    }, [isVisible]);
  }
  const onAnimationCompleteRef = React.useRef(context.onAnimationComplete);
  onAnimationCompleteRef.current = context.onAnimationComplete;
  const handleExitComplete = React.useCallback(() => {
    setIsFullyHidden(true);
    onAnimationCompleteRef.current?.({
      open: false
    });
  }, []);
  React.useEffect(() => {
    if (context.open && !isAdapted && onAnimationCompleteRef.current) {
      const tm = setTimeout(() => {
        onAnimationCompleteRef.current?.({
          open: true
        });
      }, 350);
      return () => clearTimeout(tm);
    }
  }, [context.open, isAdapted]);
  const zIndex = getExpandedShorthand("zIndex", props);
  const contents = /* @__PURE__ */jsx(StackZIndexContext, {
    zIndex: resolveViewZIndex(zIndex),
    children: /* @__PURE__ */jsx(Animate, {
      type: "presence",
      present: Boolean(context.open),
      keepChildrenMounted: Boolean(keepMounted),
      onExitComplete: handleExitComplete,
      passThrough: isAdapted,
      children
    })
  });
  const framedContents = isFullyHidden && !keepMounted && !isAdapted ? null : /* @__PURE__ */jsx(LayoutMeasurementController, {
    disable: !context.open,
    children: /* @__PURE__ */jsx(DialogPortalFrame, {
      ref,
      ...(isWeb && context.open && {
        "aria-modal": true
      }),
      pointerEvents: context.open ? "auto" : "none",
      ...frameProps,
      className: `_no_backdrop ` + (frameProps.className || ""),
      children: contents
    })
  });
  if (isWeb) {
    return /* @__PURE__ */jsx(Portal, {
      zIndex,
      stackZIndex: 1e5,
      passThrough: isAdapted,
      children: /* @__PURE__ */jsx(PassthroughTheme, {
        passThrough: isAdapted,
        children: framedContents
      })
    });
  }
  return isAdapted ? framedContents : /* @__PURE__ */jsx(DialogPortalItem, {
    context,
    children: framedContents
  });
});
const PassthroughTheme = ({
  children,
  passThrough
}) => {
  const themeName = useThemeName();
  return /* @__PURE__ */jsx(Theme, {
    passThrough,
    name: themeName,
    forceClassName: true,
    children
  });
};
const OVERLAY_NAME = "DialogOverlay";
const DialogOverlayFrame = styled(YStack, {
  name: OVERLAY_NAME,
  zIndex: 1,
  variants: {
    open: {
      true: {
        pointerEvents: "auto"
      },
      false: {
        pointerEvents: "none"
      }
    },
    unstyled: {
      false: {
        fullscreen: true,
        position: "absolute",
        backgroundColor: "$background",
        pointerEvents: "auto"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const DialogOverlay = DialogOverlayFrame.styleable(function DialogOverlay2({
  scope,
  ...props
}, forwardedRef) {
  const context = useDialogContext(scope);
  const {
    forceMount = context.forceMount,
    ...overlayProps
  } = props;
  const isAdapted = useAdaptIsActive(context.adaptScope);
  if (!forceMount) {
    if (!context.modal || isAdapted) {
      return null;
    }
  }
  return /* @__PURE__ */jsx(DialogOverlayFrame, {
    "data-state": getState(context.open),
    pointerEvents: context.open ? "auto" : "none",
    ...overlayProps,
    ref: forwardedRef
  });
});
const CONTENT_NAME = "DialogContent";
const DialogContentFrame = styled(ThemeableStack, {
  name: CONTENT_NAME,
  zIndex: 2,
  variants: {
    size: {
      "...size": (val, extras) => {
        return {};
      }
    },
    unstyled: {
      false: {
        position: "relative",
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "$borderColor",
        padding: "$true",
        borderRadius: "$true",
        elevate: true,
        // Ensure content receives pointer events (fixes React 19 + display:contents inheritance)
        pointerEvents: "auto"
      }
    }
  },
  defaultVariants: {
    size: "$true",
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
const DialogContent = DialogContentFrame.styleable(function DialogContent2({
  scope,
  ...props
}, forwardedRef) {
  const context = useDialogContext(scope);
  const contents = /* @__PURE__ */jsx(Fragment, {
    children: context.modal ? /* @__PURE__ */jsx(DialogContentModal, {
      context,
      ...props,
      ref: forwardedRef
    }) : /* @__PURE__ */jsx(DialogContentNonModal, {
      context,
      ...props,
      ref: forwardedRef
    })
  });
  if (!isWeb || context.disableRemoveScroll) {
    return contents;
  }
  return /* @__PURE__ */jsx(RemoveScroll, {
    enabled: context.open,
    children: /* @__PURE__ */jsx("div", {
      "data-remove-scroll-container": true,
      className: "_dsp_contents",
      children: contents
    })
  });
});
const DialogContentModal = React.forwardRef(({
  children,
  context,
  ...props
}, forwardedRef) => {
  const contentRef = React.useRef(null);
  const composedRefs = useComposedRefs(forwardedRef, context.contentRef, contentRef);
  return /* @__PURE__ */jsx(DialogContentImpl, {
    ...props,
    context,
    ref: composedRefs,
    trapFocus: context.open,
    disableOutsidePointerEvents: true,
    onCloseAutoFocus: composeEventHandlers(props.onCloseAutoFocus, event => {
      event.preventDefault();
      context.triggerRef.current?.focus();
    }),
    onPointerDownOutside: composeEventHandlers(props.onPointerDownOutside, event => {
      const originalEvent = event["detail"].originalEvent;
      const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
      const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
      if (isRightClick) event.preventDefault();
    }),
    onFocusOutside: composeEventHandlers(props.onFocusOutside, event => event.preventDefault()),
    ...(!props.unstyled && {
      outlineStyle: "none"
    }),
    children
  });
});
const DialogContentNonModal = React.forwardRef((props, forwardedRef) => {
  const hasInteractedOutsideRef = React.useRef(false);
  return /* @__PURE__ */jsx(DialogContentImpl, {
    ...props,
    ref: forwardedRef,
    trapFocus: false,
    disableOutsidePointerEvents: false,
    onCloseAutoFocus: event => {
      props.onCloseAutoFocus?.(event);
      if (!event.defaultPrevented) {
        if (!hasInteractedOutsideRef.current) {
          props.context.triggerRef.current?.focus();
        }
        event.preventDefault();
      }
      hasInteractedOutsideRef.current = false;
    },
    onInteractOutside: event => {
      props.onInteractOutside?.(event);
      if (!event.defaultPrevented) hasInteractedOutsideRef.current = true;
      const target = event.target;
      const trigger = props.context.triggerRef.current;
      if (!(trigger instanceof HTMLElement)) return;
      const targetIsTrigger = trigger.contains(target);
      if (targetIsTrigger) event.preventDefault();
    }
  });
});
const DialogContentImpl = React.forwardRef((props, forwardedRef) => {
  const {
    trapFocus,
    onOpenAutoFocus,
    onCloseAutoFocus,
    disableOutsidePointerEvents,
    onEscapeKeyDown,
    onPointerDownOutside,
    onFocusOutside,
    onInteractOutside,
    context,
    ...contentProps
  } = props;
  const contentRef = React.useRef(null);
  const composedRefs = useComposedRefs(forwardedRef, contentRef);
  const isAdapted = useAdaptIsActive(context.adaptScope);
  if (isAdapted) {
    if (!isWeb && !context.open) {
      return null;
    }
    return /* @__PURE__ */jsx(DialogPortalItem, {
      context,
      children: contentProps.children
    });
  }
  const contents = /* @__PURE__ */jsx(DialogContentFrame, {
    ref: composedRefs,
    id: context.contentId,
    role: "dialog",
    "aria-modal": context.modal,
    "aria-describedby": context.descriptionId,
    "aria-labelledby": context.titleId,
    "data-state": getState(context.open),
    pointerEvents: context.open ? "auto" : "none",
    ...contentProps
  });
  if (!isWeb) {
    return contents;
  }
  return /* @__PURE__ */jsxs(Fragment, {
    children: [/* @__PURE__ */jsx(Dismissable, {
      disableOutsidePointerEvents: context.open && disableOutsidePointerEvents,
      forceUnmount: !context.open,
      onEscapeKeyDown,
      onPointerDownOutside,
      onFocusOutside,
      onInteractOutside,
      onDismiss: () => context?.onOpenChange?.(false),
      children: /* @__PURE__ */jsx(FocusScope, {
        loop: true,
        enabled: context.open,
        trapped: trapFocus,
        onMountAutoFocus: onOpenAutoFocus,
        forceUnmount: !context.open,
        onUnmountAutoFocus: onCloseAutoFocus,
        children: contents
      })
    }), process.env.NODE_ENV === "development" && /* @__PURE__ */jsxs(Fragment, {
      children: [/* @__PURE__ */jsx(TitleWarning, {
        titleId: context.titleId
      }), /* @__PURE__ */jsx(DescriptionWarning, {
        contentRef,
        descriptionId: context.descriptionId
      })]
    })]
  });
});
const DialogTitleFrame = styled(H2, {
  name: "DialogTitle"
});
const DialogTitle = DialogTitleFrame.styleable(function DialogTitle2(props, forwardedRef) {
  const {
    scope,
    ...titleProps
  } = props;
  const context = useDialogContext(scope);
  return /* @__PURE__ */jsx(DialogTitleFrame, {
    id: context.titleId,
    ...titleProps,
    ref: forwardedRef
  });
});
const DialogDescriptionFrame = styled(Paragraph, {
  name: "DialogDescription"
});
const DialogDescription = DialogDescriptionFrame.styleable(function DialogDescription2(props, forwardedRef) {
  const {
    scope,
    ...descriptionProps
  } = props;
  const context = useDialogContext(scope);
  return /* @__PURE__ */jsx(DialogDescriptionFrame, {
    id: context.descriptionId,
    ...descriptionProps,
    ref: forwardedRef
  });
});
const CLOSE_NAME = "DialogClose";
const DialogCloseFrame = styled(View, {
  name: CLOSE_NAME,
  render: "button"
});
const DialogClose = DialogCloseFrame.styleable((props, forwardedRef) => {
  const {
    scope,
    displayWhenAdapted,
    ...closeProps
  } = props;
  const context = useDialogContext(scope);
  const isAdapted = useAdaptIsActive(context.adaptScope);
  const isInsideButton = React.useContext(ButtonNestingContext);
  if (isAdapted && !displayWhenAdapted) {
    return null;
  }
  return /* @__PURE__ */jsx(DialogCloseFrame, {
    "aria-label": "Dialog Close",
    render: isInsideButton ? "span" : "button",
    ...closeProps,
    ref: forwardedRef,
    onPress: composeEventHandlers(props.onPress, () => {
      context.onOpenChange(false);
    })
  });
});
function getState(open) {
  return open ? "open" : "closed";
}
const TITLE_WARNING_NAME = "DialogTitleWarning";
const [DialogWarningProvider, useWarningContext] = createContext(TITLE_WARNING_NAME, {
  contentName: CONTENT_NAME,
  titleName: "DialogTitle",
  docsSlug: "dialog"
});
const TitleWarning = ({
  titleId
}) => {
  if (process.env.NODE_ENV === "development") {
    const titleWarningContext = useWarningContext(TITLE_WARNING_NAME);
    const MESSAGE = `\`${titleWarningContext.contentName}\` wants a \`${titleWarningContext.titleName}\` to be accessible. If you want to hide the \`${titleWarningContext.titleName}\`, wrap it with <VisuallyHidden />.`;
    React.useEffect(() => {
      if (!isWeb) return;
      if (titleId) {
        const hasTitle = document.getElementById(titleId);
        if (!hasTitle) {
          console.warn(MESSAGE);
        }
      }
    }, [MESSAGE, titleId]);
  }
  return null;
};
const DESCRIPTION_WARNING_NAME = "DialogDescriptionWarning";
const DescriptionWarning = ({
  contentRef,
  descriptionId
}) => {
  if (process.env.NODE_ENV === "development") {
    const descriptionWarningContext = useWarningContext(DESCRIPTION_WARNING_NAME);
    const MESSAGE = `Warning: Missing \`Description\` or \`aria-describedby={undefined}\` for {${descriptionWarningContext.contentName}}.`;
    React.useEffect(() => {
      if (!isWeb) return;
      const contentNode = contentRef.current;
      if (!(contentNode instanceof HTMLElement)) {
        return;
      }
      const describedById = contentNode.getAttribute("aria-describedby");
      if (descriptionId && describedById) {
        const hasDescription = document.getElementById(descriptionId);
        if (!hasDescription) {
          console.warn(MESSAGE);
        }
      }
    }, [MESSAGE, contentRef, descriptionId]);
  }
  return null;
};
const Dialog = withStaticProperties(React.forwardRef(function Dialog2(props, ref) {
  const {
    scope = "",
    children,
    open: openProp,
    defaultOpen = false,
    onOpenChange,
    modal = true,
    keepChildrenMounted,
    disableRemoveScroll = false,
    onAnimationComplete
  } = props;
  const baseId = React.useId();
  const dialogId = `Dialog-${scope}-${baseId}`;
  const contentId = `${dialogId}-content`;
  const titleId = `${dialogId}-title`;
  const descriptionId = `${dialogId}-description`;
  const triggerRef = React.useRef(null);
  const contentRef = React.useRef(null);
  const [open, setOpen] = useControllableState({
    prop: openProp,
    defaultProp: defaultOpen,
    onChange: onOpenChange
  });
  const onOpenToggle = React.useCallback(() => {
    setOpen(prevOpen => !prevOpen);
  }, [setOpen]);
  const adaptScope = `DialogAdapt${scope}`;
  const context = {
    dialogScope: scope,
    adaptScope,
    triggerRef,
    contentRef,
    contentId,
    titleId,
    descriptionId,
    open,
    onOpenChange: setOpen,
    onOpenToggle,
    modal,
    keepChildrenMounted,
    disableRemoveScroll,
    onAnimationComplete
  };
  React.useImperativeHandle(ref, () => ({
    open: setOpen
  }), [setOpen]);
  return /* @__PURE__ */jsx(AdaptParent, {
    scope: adaptScope,
    portal: {
      forwardProps: props
    },
    children: /* @__PURE__ */jsx(DialogProvider, {
      scope,
      ...context,
      children: /* @__PURE__ */jsx(DialogSheetController, {
        onOpenChange: setOpen,
        scope,
        children
      })
    })
  });
}), {
  Trigger: DialogTrigger,
  Portal: DialogPortal,
  Overlay: DialogOverlay,
  Content: DialogContent,
  Title: DialogTitle,
  Description: DialogDescription,
  Close: DialogClose,
  FocusScope: FocusScopeController,
  Adapt
});
const DialogSheetController = props => {
  const context = useDialogContext(props.scope);
  const isAdapted = useAdaptIsActive(context.adaptScope);
  return /* @__PURE__ */jsx(SheetController, {
    onOpenChange: val => {
      if (isAdapted) {
        props.onOpenChange?.(val);
      }
    },
    open: context.open,
    hidden: !isAdapted,
    children: props.children
  });
};
export { Dialog, DialogClose, DialogContent, DialogContext, DialogDescription, DialogOverlay, DialogOverlayFrame, DialogPortal, DialogPortalFrame, DialogProvider, DialogTitle, DialogTrigger, DialogWarningProvider, useDialogContext };
