// @tamagui/dialog/dist/jsx/Dialog.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Dialog_exports = {};
__export(Dialog_exports, {
  Dialog: () => Dialog,
  DialogClose: () => DialogClose,
  DialogContent: () => DialogContent,
  DialogContext: () => DialogContext,
  DialogDescription: () => DialogDescription,
  DialogOverlay: () => DialogOverlay,
  DialogOverlayFrame: () => DialogOverlayFrame,
  DialogPortal: () => DialogPortal,
  DialogPortalFrame: () => DialogPortalFrame,
  DialogProvider: () => DialogProvider,
  DialogTitle: () => DialogTitle,
  DialogTrigger: () => DialogTrigger,
  DialogWarningProvider: () => DialogWarningProvider,
  useDialogContext: () => useDialogContext
});
module.exports = __toCommonJS(Dialog_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_adapt = require("@tamagui/adapt");
var import_animate = require("@tamagui/animate");
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_create_context = require("@tamagui/create-context");
var import_dismissable = require("@tamagui/dismissable");
var import_focus_scope = require("@tamagui/focus-scope");
var import_helpers = require("@tamagui/helpers");
var import_portal = require("@tamagui/portal");
var import_remove_scroll = require("@tamagui/remove-scroll");
var import_controller = require("@tamagui/sheet/controller");
var import_stacks = require("@tamagui/stacks");
var import_text = require("@tamagui/text");
var import_use_controllable_state = require("@tamagui/use-controllable-state");
var import_z_index_stack = require("@tamagui/z-index-stack");
var React = __toESM(require("react"), 1);
function _instanceof(left, right) {
  if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
    return !!right[Symbol.hasInstance](left);
  } else {
    return left instanceof right;
  }
}
var DialogContext = (0, import_core.createStyledContext)(
// since we always provide this we can avoid setting here
{}, "Dialog__");
var {
  useStyledContext: useDialogContext,
  Provider: DialogProvider
} = DialogContext;
var DialogTriggerFrame = (0, import_core.styled)(import_core.View, {
  name: "DialogTrigger"
});
var DialogTrigger = DialogTriggerFrame.styleable(function DialogTrigger2(props, forwardedRef) {
  var {
    scope,
    ...triggerProps
  } = props;
  var isInsideButton = React.useContext(import_stacks.ButtonNestingContext);
  var context = useDialogContext(scope);
  var composedTriggerRef = (0, import_compose_refs.useComposedRefs)(forwardedRef, context.triggerRef);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_stacks.ButtonNestingContext.Provider, {
    value: true,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogTriggerFrame, {
      render: isInsideButton ? "span" : "button",
      "aria-haspopup": "dialog",
      "aria-expanded": context.open,
      "aria-controls": context.contentId,
      "data-state": getState(context.open),
      ...triggerProps,
      ref: composedTriggerRef,
      onPress: (0, import_helpers.composeEventHandlers)(props.onPress, context.onOpenToggle)
    })
  });
});
var DialogPortalFrame = (0, import_core.styled)(import_stacks.YStack, {
  pointerEvents: "none",
  render: "dialog",
  variants: {
    unstyled: {
      false: {
        alignItems: "center",
        justifyContent: "center",
        fullscreen: true,
        "$platform-web": {
          // undo dialog styles
          borderWidth: 0,
          backgroundColor: "transparent",
          color: "inherit",
          maxInlineSize: "none",
          margin: 0,
          width: "auto",
          height: "auto",
          // ensure always in frame and right height
          maxHeight: "100vh",
          position: "fixed",
          // ensure dialog inherits stacking context from portal wrapper
          zIndex: 1
        }
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var needsRepropagation = (0, import_portal.needsPortalRepropagation)();
var DialogPortalItem = function (param) {
  var {
    context,
    children
  } = param;
  var themeName = (0, import_core.useThemeName)();
  var isAdapted = (0, import_adapt.useAdaptIsActive)(context.adaptScope);
  var adaptContext = (0, import_adapt.useAdaptContext)(context.adaptScope);
  var content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.Theme, {
    name: themeName,
    children
  });
  if (needsRepropagation) {
    content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_adapt.ProvideAdaptContext, {
      ...adaptContext,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogProvider, {
        ...context,
        children: content
      })
    });
  }
  return isAdapted ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_adapt.AdaptPortalContents, {
    scope: context.adaptScope,
    children: content
  }) : context.modal ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_portal.PortalItem, {
    hostName: context.modal ? "root" : context.adaptScope,
    children: content
  }) : content;
};
var DialogPortal = /* @__PURE__ */React.forwardRef(function (props, forwardRef) {
  var {
    scope,
    forceMount,
    children,
    ...frameProps
  } = props;
  var dialogRef = React.useRef(null);
  var ref = (0, import_compose_refs.composeRefs)(dialogRef, forwardRef);
  var context = useDialogContext(scope);
  var keepMounted = forceMount || context.keepChildrenMounted;
  var isAdapted = (0, import_adapt.useAdaptIsActive)(context.adaptScope);
  var [isFullyHidden, setIsFullyHidden] = React.useState(!context.open);
  if (context.open && isFullyHidden) {
    setIsFullyHidden(false);
  }
  var isVisible = context.open ? true : !isFullyHidden;
  if (import_constants.isWeb) {
    (0, import_constants.useIsomorphicLayoutEffect)(function () {
      var node = dialogRef.current;
      if (!_instanceof(node, HTMLDialogElement)) return;
      if (isVisible) {
        var _node_show;
        (_node_show = node.show) === null || _node_show === void 0 ? void 0 : _node_show.call(node);
      } else {
        var _node_close;
        (_node_close = node.close) === null || _node_close === void 0 ? void 0 : _node_close.call(node);
      }
    }, [isVisible]);
  }
  var onAnimationCompleteRef = React.useRef(context.onAnimationComplete);
  onAnimationCompleteRef.current = context.onAnimationComplete;
  var handleExitComplete = React.useCallback(function () {
    var _onAnimationCompleteRef_current;
    setIsFullyHidden(true);
    (_onAnimationCompleteRef_current = onAnimationCompleteRef.current) === null || _onAnimationCompleteRef_current === void 0 ? void 0 : _onAnimationCompleteRef_current.call(onAnimationCompleteRef, {
      open: false
    });
  }, []);
  React.useEffect(function () {
    if (context.open && !isAdapted && onAnimationCompleteRef.current) {
      var tm = setTimeout(function () {
        var _onAnimationCompleteRef_current;
        (_onAnimationCompleteRef_current = onAnimationCompleteRef.current) === null || _onAnimationCompleteRef_current === void 0 ? void 0 : _onAnimationCompleteRef_current.call(onAnimationCompleteRef, {
          open: true
        });
      }, 350);
      return function () {
        return clearTimeout(tm);
      };
    }
  }, [context.open, isAdapted]);
  var zIndex = (0, import_core.getExpandedShorthand)("zIndex", props);
  var contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_z_index_stack.StackZIndexContext, {
    zIndex: (0, import_portal.resolveViewZIndex)(zIndex),
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_animate.Animate, {
      type: "presence",
      present: Boolean(context.open),
      keepChildrenMounted: Boolean(keepMounted),
      onExitComplete: handleExitComplete,
      passThrough: isAdapted,
      children
    })
  });
  var framedContents = isFullyHidden && !keepMounted && !isAdapted ? null : /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.LayoutMeasurementController, {
    disable: !context.open,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogPortalFrame, {
      ref,
      ...(import_constants.isWeb && context.open && {
        "aria-modal": true
      }),
      pointerEvents: context.open ? "auto" : "none",
      ...frameProps,
      className: `_no_backdrop ` + (frameProps.className || ""),
      children: contents
    })
  });
  if (import_constants.isWeb) {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_portal.Portal, {
      zIndex,
      // set to 100000 which ensures dialogs are above most fixed UI (headers, navs)
      // this makes sure its above typical stacking contexts
      stackZIndex: 1e5,
      passThrough: isAdapted,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(PassthroughTheme, {
        passThrough: isAdapted,
        children: framedContents
      })
    });
  }
  return isAdapted ? framedContents : /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogPortalItem, {
    context,
    children: framedContents
  });
});
var PassthroughTheme = function (param) {
  var {
    children,
    passThrough
  } = param;
  var themeName = (0, import_core.useThemeName)();
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.Theme, {
    passThrough,
    name: themeName,
    forceClassName: true,
    children
  });
};
var OVERLAY_NAME = "DialogOverlay";
var DialogOverlayFrame = (0, import_core.styled)(import_stacks.YStack, {
  name: OVERLAY_NAME,
  zIndex: 1,
  variants: {
    open: {
      true: {
        pointerEvents: "auto"
      },
      false: {
        pointerEvents: "none"
      }
    },
    unstyled: {
      false: {
        fullscreen: true,
        position: "absolute",
        backgroundColor: "$background",
        pointerEvents: "auto"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var DialogOverlay = DialogOverlayFrame.styleable(function DialogOverlay2(param, forwardedRef) {
  var {
    scope,
    ...props
  } = param;
  var context = useDialogContext(scope);
  var {
    forceMount = context.forceMount,
    ...overlayProps
  } = props;
  var isAdapted = (0, import_adapt.useAdaptIsActive)(context.adaptScope);
  if (!forceMount) {
    if (!context.modal || isAdapted) {
      return null;
    }
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogOverlayFrame, {
    "data-state": getState(context.open),
    // TODO: this will be apply for v2
    // onPress={() => {
    //   // if the overlay is pressed, close the dialog
    //   context.onOpenChange(false)
    // }}
    // We re-enable pointer-events prevented by `Dialog.Content` to allow scrolling the overlay.
    pointerEvents: context.open ? "auto" : "none",
    ...overlayProps,
    ref: forwardedRef
  });
});
var CONTENT_NAME = "DialogContent";
var DialogContentFrame = (0, import_core.styled)(import_stacks.ThemeableStack, {
  name: CONTENT_NAME,
  zIndex: 2,
  variants: {
    size: {
      "...size": function (val, extras) {
        return {};
      }
    },
    unstyled: {
      false: {
        position: "relative",
        backgroundColor: "$background",
        borderWidth: 1,
        borderColor: "$borderColor",
        padding: "$true",
        borderRadius: "$true",
        elevate: true,
        // Ensure content receives pointer events (fixes React 19 + display:contents inheritance)
        pointerEvents: "auto"
      }
    }
  },
  defaultVariants: {
    size: "$true",
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var DialogContent = DialogContentFrame.styleable(function DialogContent2(param, forwardedRef) {
  var {
    scope,
    ...props
  } = param;
  var context = useDialogContext(scope);
  var contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children: context.modal ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogContentModal, {
      context,
      ...props,
      ref: forwardedRef
    }) : /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogContentNonModal, {
      context,
      ...props,
      ref: forwardedRef
    })
  });
  if (!import_constants.isWeb || context.disableRemoveScroll) {
    return contents;
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_remove_scroll.RemoveScroll, {
    enabled: context.open,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)("div", {
      "data-remove-scroll-container": true,
      className: "_dsp_contents",
      children: contents
    })
  });
});
var DialogContentModal = /* @__PURE__ */React.forwardRef(function (param, forwardedRef) {
  var {
    children,
    context,
    ...props
  } = param;
  var contentRef = React.useRef(null);
  var composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, context.contentRef, contentRef);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogContentImpl, {
    ...props,
    context,
    ref: composedRefs,
    // we make sure focus isn't trapped once `DialogContent` has been closed
    // (closed !== unmounted when animating out)
    trapFocus: context.open,
    disableOutsidePointerEvents: true,
    onCloseAutoFocus: (0, import_helpers.composeEventHandlers)(props.onCloseAutoFocus, function (event) {
      var _context_triggerRef_current;
      event.preventDefault();
      (_context_triggerRef_current = context.triggerRef.current) === null || _context_triggerRef_current === void 0 ? void 0 : _context_triggerRef_current.focus();
    }),
    onPointerDownOutside: (0, import_helpers.composeEventHandlers)(props.onPointerDownOutside, function (event) {
      var originalEvent = event["detail"].originalEvent;
      var ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
      var isRightClick = originalEvent.button === 2 || ctrlLeftClick;
      if (isRightClick) event.preventDefault();
    }),
    // When focus is trapped, a `focusout` event may still happen.
    // We make sure we don't trigger our `onDismiss` in such case.
    onFocusOutside: (0, import_helpers.composeEventHandlers)(props.onFocusOutside, function (event) {
      return event.preventDefault();
    }),
    ...(!props.unstyled && {
      outlineStyle: "none"
    }),
    children
  });
});
var DialogContentNonModal = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var hasInteractedOutsideRef = React.useRef(false);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogContentImpl, {
    ...props,
    ref: forwardedRef,
    trapFocus: false,
    disableOutsidePointerEvents: false,
    onCloseAutoFocus: function (event) {
      var _props_onCloseAutoFocus;
      (_props_onCloseAutoFocus = props.onCloseAutoFocus) === null || _props_onCloseAutoFocus === void 0 ? void 0 : _props_onCloseAutoFocus.call(props, event);
      if (!event.defaultPrevented) {
        if (!hasInteractedOutsideRef.current) {
          var _props_context_triggerRef_current;
          (_props_context_triggerRef_current = props.context.triggerRef.current) === null || _props_context_triggerRef_current === void 0 ? void 0 : _props_context_triggerRef_current.focus();
        }
        event.preventDefault();
      }
      hasInteractedOutsideRef.current = false;
    },
    onInteractOutside: function (event) {
      var _props_onInteractOutside;
      (_props_onInteractOutside = props.onInteractOutside) === null || _props_onInteractOutside === void 0 ? void 0 : _props_onInteractOutside.call(props, event);
      if (!event.defaultPrevented) hasInteractedOutsideRef.current = true;
      var target = event.target;
      var trigger = props.context.triggerRef.current;
      if (!_instanceof(trigger, HTMLElement)) return;
      var targetIsTrigger = trigger.contains(target);
      if (targetIsTrigger) event.preventDefault();
    }
  });
});
var DialogContentImpl = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
  var {
    trapFocus,
    onOpenAutoFocus,
    onCloseAutoFocus,
    disableOutsidePointerEvents,
    onEscapeKeyDown,
    onPointerDownOutside,
    onFocusOutside,
    onInteractOutside,
    context,
    ...contentProps
  } = props;
  var contentRef = React.useRef(null);
  var composedRefs = (0, import_compose_refs.useComposedRefs)(forwardedRef, contentRef);
  var isAdapted = (0, import_adapt.useAdaptIsActive)(context.adaptScope);
  if (isAdapted) {
    if (!import_constants.isWeb && !context.open) {
      return null;
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogPortalItem, {
      context,
      children: contentProps.children
    });
  }
  var contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogContentFrame, {
    ref: composedRefs,
    id: context.contentId,
    role: "dialog",
    "aria-modal": context.modal,
    "aria-describedby": context.descriptionId,
    "aria-labelledby": context.titleId,
    "data-state": getState(context.open),
    // allow clicking through content during exit animation
    pointerEvents: context.open ? "auto" : "none",
    ...contentProps
  });
  if (!import_constants.isWeb) {
    return contents;
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
    children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(import_dismissable.Dismissable, {
      disableOutsidePointerEvents: context.open && disableOutsidePointerEvents,
      forceUnmount: !context.open,
      onEscapeKeyDown,
      onPointerDownOutside,
      onFocusOutside,
      onInteractOutside,
      onDismiss: function () {
        var _context_onOpenChange;
        return context === null || context === void 0 ? void 0 : (_context_onOpenChange = context.onOpenChange) === null || _context_onOpenChange === void 0 ? void 0 : _context_onOpenChange.call(context, false);
      },
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_focus_scope.FocusScope, {
        loop: true,
        enabled: context.open,
        trapped: trapFocus,
        onMountAutoFocus: onOpenAutoFocus,
        forceUnmount: !context.open,
        onUnmountAutoFocus: onCloseAutoFocus,
        children: contents
      })
    }), process.env.NODE_ENV === "development" && /* @__PURE__ */(0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, {
      children: [/* @__PURE__ */(0, import_jsx_runtime.jsx)(TitleWarning, {
        titleId: context.titleId
      }), /* @__PURE__ */(0, import_jsx_runtime.jsx)(DescriptionWarning, {
        contentRef,
        descriptionId: context.descriptionId
      })]
    })]
  });
});
var DialogTitleFrame = (0, import_core.styled)(import_text.H2, {
  name: "DialogTitle"
});
var DialogTitle = DialogTitleFrame.styleable(function DialogTitle2(props, forwardedRef) {
  var {
    scope,
    ...titleProps
  } = props;
  var context = useDialogContext(scope);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogTitleFrame, {
    id: context.titleId,
    ...titleProps,
    ref: forwardedRef
  });
});
var DialogDescriptionFrame = (0, import_core.styled)(import_text.Paragraph, {
  name: "DialogDescription"
});
var DialogDescription = DialogDescriptionFrame.styleable(function DialogDescription2(props, forwardedRef) {
  var {
    scope,
    ...descriptionProps
  } = props;
  var context = useDialogContext(scope);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogDescriptionFrame, {
    id: context.descriptionId,
    ...descriptionProps,
    ref: forwardedRef
  });
});
var CLOSE_NAME = "DialogClose";
var DialogCloseFrame = (0, import_core.styled)(import_core.View, {
  name: CLOSE_NAME,
  render: "button"
});
var DialogClose = DialogCloseFrame.styleable(function (props, forwardedRef) {
  var {
    scope,
    displayWhenAdapted,
    ...closeProps
  } = props;
  var context = useDialogContext(scope);
  var isAdapted = (0, import_adapt.useAdaptIsActive)(context.adaptScope);
  var isInsideButton = React.useContext(import_stacks.ButtonNestingContext);
  if (isAdapted && !displayWhenAdapted) {
    return null;
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogCloseFrame, {
    "aria-label": "Dialog Close",
    render: isInsideButton ? "span" : "button",
    ...closeProps,
    ref: forwardedRef,
    onPress: (0, import_helpers.composeEventHandlers)(props.onPress, function () {
      context.onOpenChange(false);
    })
  });
});
function getState(open) {
  return open ? "open" : "closed";
}
var TITLE_WARNING_NAME = "DialogTitleWarning";
var [DialogWarningProvider, useWarningContext] = (0, import_create_context.createContext)(TITLE_WARNING_NAME, {
  contentName: CONTENT_NAME,
  titleName: "DialogTitle",
  docsSlug: "dialog"
});
var TitleWarning = function (param) {
  var {
    titleId
  } = param;
  if (process.env.NODE_ENV === "development") {
    var titleWarningContext = useWarningContext(TITLE_WARNING_NAME);
    var MESSAGE = `\`${titleWarningContext.contentName}\` wants a \`${titleWarningContext.titleName}\` to be accessible. If you want to hide the \`${titleWarningContext.titleName}\`, wrap it with <VisuallyHidden />.`;
    React.useEffect(function () {
      if (!import_constants.isWeb) return;
      if (titleId) {
        var hasTitle = document.getElementById(titleId);
        if (!hasTitle) {
          console.warn(MESSAGE);
        }
      }
    }, [MESSAGE, titleId]);
  }
  return null;
};
var DESCRIPTION_WARNING_NAME = "DialogDescriptionWarning";
var DescriptionWarning = function (param) {
  var {
    contentRef,
    descriptionId
  } = param;
  if (process.env.NODE_ENV === "development") {
    var descriptionWarningContext = useWarningContext(DESCRIPTION_WARNING_NAME);
    var MESSAGE = `Warning: Missing \`Description\` or \`aria-describedby={undefined}\` for {${descriptionWarningContext.contentName}}.`;
    React.useEffect(function () {
      if (!import_constants.isWeb) return;
      var contentNode = contentRef.current;
      if (!_instanceof(contentNode, HTMLElement)) {
        return;
      }
      var describedById = contentNode.getAttribute("aria-describedby");
      if (descriptionId && describedById) {
        var hasDescription = document.getElementById(descriptionId);
        if (!hasDescription) {
          console.warn(MESSAGE);
        }
      }
    }, [MESSAGE, contentRef, descriptionId]);
  }
  return null;
};
var Dialog = (0, import_helpers.withStaticProperties)(/* @__PURE__ */React.forwardRef(function Dialog2(props, ref) {
  var {
    scope = "",
    children,
    open: openProp,
    defaultOpen = false,
    onOpenChange,
    modal = true,
    keepChildrenMounted,
    disableRemoveScroll = false,
    onAnimationComplete
  } = props;
  var baseId = React.useId();
  var dialogId = `Dialog-${scope}-${baseId}`;
  var contentId = `${dialogId}-content`;
  var titleId = `${dialogId}-title`;
  var descriptionId = `${dialogId}-description`;
  var triggerRef = React.useRef(null);
  var contentRef = React.useRef(null);
  var [open, setOpen] = (0, import_use_controllable_state.useControllableState)({
    prop: openProp,
    defaultProp: defaultOpen,
    onChange: onOpenChange
  });
  var onOpenToggle = React.useCallback(function () {
    setOpen(function (prevOpen) {
      return !prevOpen;
    });
  }, [setOpen]);
  var adaptScope = `DialogAdapt${scope}`;
  var context = {
    dialogScope: scope,
    adaptScope,
    triggerRef,
    contentRef,
    contentId,
    titleId,
    descriptionId,
    open,
    onOpenChange: setOpen,
    onOpenToggle,
    modal,
    keepChildrenMounted,
    disableRemoveScroll,
    onAnimationComplete
  };
  React.useImperativeHandle(ref, function () {
    return {
      open: setOpen
    };
  }, [setOpen]);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_adapt.AdaptParent, {
    scope: adaptScope,
    portal: {
      forwardProps: props
    },
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogProvider, {
      scope,
      ...context,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(DialogSheetController, {
        onOpenChange: setOpen,
        scope,
        children
      })
    })
  });
}), {
  Trigger: DialogTrigger,
  Portal: DialogPortal,
  Overlay: DialogOverlay,
  Content: DialogContent,
  Title: DialogTitle,
  Description: DialogDescription,
  Close: DialogClose,
  FocusScope: import_focus_scope.FocusScopeController,
  Adapt: import_adapt.Adapt
});
var DialogSheetController = function (props) {
  var context = useDialogContext(props.scope);
  var isAdapted = (0, import_adapt.useAdaptIsActive)(context.adaptScope);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_controller.SheetController, {
    onOpenChange: function (val) {
      if (isAdapted) {
        var _props_onOpenChange;
        (_props_onOpenChange = props.onOpenChange) === null || _props_onOpenChange === void 0 ? void 0 : _props_onOpenChange.call(props, val);
      }
    },
    open: context.open,
    hidden: !isAdapted,
    children: props.children
  });
};
