// @tamagui/context-menu/dist/cjs/ContextMenu.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var ContextMenu_exports = {};
__export(ContextMenu_exports, {
  createContextMenu: () => createContextMenu
});
module.exports = __toCommonJS(ContextMenu_exports);
var import_create_menu = require("@tamagui/create-menu");
var import_web = require("@tamagui/web");
var import_react = __toESM(require("react"), 1);
var import_createNonNativeContextMenu = require("./createNonNativeContextMenu.native.js");
var COMMON_PARAMS = {
  isRoot: false,
  scope: import_createNonNativeContextMenu.CONTEXTMENU_CONTEXT
};
function createContextMenu(param) {
  var {
    Menu: NativeMenuRoot
  } = (0, import_create_menu.createNativeMenu)("ContextMenu");
  var NonNativeContextMenu = (0, import_createNonNativeContextMenu.createNonNativeContextMenu)(param);
  var ContextMenuComp = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Root,
    NativeComponent: NativeMenuRoot,
    isRoot: true
  });
  var Trigger = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Trigger,
    NativeComponent: NativeMenuRoot.Trigger
  });
  var Portal = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Portal,
    NativeComponent: import_react.default.Fragment
  });
  var Content = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Content,
    NativeComponent: NativeMenuRoot.Content
  });
  var Preview = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Preview,
    NativeComponent: NativeMenuRoot.Preview
  });
  var Group = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Group,
    NativeComponent: NativeMenuRoot.Group
  });
  var Label = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Label,
    NativeComponent: NativeMenuRoot.Label
  });
  var Item = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Item,
    NativeComponent: NativeMenuRoot.Item
  });
  var ItemTitle = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemTitle,
    NativeComponent: NativeMenuRoot.ItemTitle
  });
  var ItemSubtitle = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemSubtitle,
    NativeComponent: NativeMenuRoot.ItemSubtitle
  });
  var ItemIcon = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemIcon,
    NativeComponent: NativeMenuRoot.ItemIcon
  });
  var ItemImage = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemImage,
    NativeComponent: NativeMenuRoot.ItemImage
  });
  var CheckboxItem = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.CheckboxItem,
    NativeComponent: NativeMenuRoot.CheckboxItem
  });
  var RadioGroup = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.RadioGroup,
    NativeComponent: function () {
      return null;
    }
  });
  var RadioItem = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.RadioItem,
    NativeComponent: function (param2) {
      var {
        children
      } = param2;
      return children;
    }
  });
  var ItemIndicator = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.ItemIndicator,
    NativeComponent: NativeMenuRoot.ItemIndicator
  });
  var Separator = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Separator,
    NativeComponent: NativeMenuRoot.Separator
  });
  var Arrow = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Arrow,
    NativeComponent: NativeMenuRoot.Arrow
  });
  var Sub = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.Sub,
    NativeComponent: NativeMenuRoot.Sub
  });
  var SubTrigger = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.SubTrigger,
    NativeComponent: NativeMenuRoot.SubTrigger
  });
  var SubContent = (0, import_create_menu.withNativeMenu)({
    ...COMMON_PARAMS,
    Component: NonNativeContextMenu.SubContent,
    NativeComponent: NativeMenuRoot.SubContent
  });
  var ContextMenu = (0, import_web.withStaticProperties)(ContextMenuComp, {
    Trigger,
    Portal,
    Content,
    Group,
    Label,
    Item,
    CheckboxItem,
    RadioGroup,
    RadioItem,
    ItemIndicator,
    Separator,
    Arrow,
    Sub,
    // cast to React.FC to avoid TS error
    SubTrigger,
    SubContent,
    ItemTitle,
    ItemSubtitle,
    // cast to React.FC to avoid TS error
    ItemIcon,
    ItemImage,
    Preview
  });
  return ContextMenu;
}
