// @tamagui/switch-headless/dist/cjs/useSwitch.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var useSwitch_exports = {};
__export(useSwitch_exports, {
  useSwitch: () => useSwitch
});
module.exports = __toCommonJS(useSwitch_exports);
var import_compose_refs = require("@tamagui/compose-refs");
var import_constants = require("@tamagui/constants");
var import_helpers = require("@tamagui/helpers");
var import_label = require("@tamagui/label");
var import_use_previous = require("@tamagui/use-previous");
var React = __toESM(require("react"), 1);
var import_jsx_runtime = require("react/jsx-runtime");
function getState(checked) {
  return checked ? "checked" : "unchecked";
}
const BubbleInput = props => {
  const {
    control,
    checked,
    bubbles = true,
    ...inputProps
  } = props;
  const ref = React.useRef(null);
  const prevChecked = (0, import_use_previous.usePrevious)(checked);
  React.useEffect(() => {
    const input = ref.current;
    const inputProto = window.HTMLInputElement.prototype;
    const setChecked = Object.getOwnPropertyDescriptor(inputProto, "checked").set;
    if (prevChecked !== checked && setChecked) {
      const event = new Event("click", {
        bubbles
      });
      setChecked.call(input, checked);
      input.dispatchEvent(event);
    }
  }, [prevChecked, checked, bubbles]);
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)("input", {
    type: "checkbox",
    "aria-hidden": true,
    defaultChecked: checked,
    ...inputProps,
    tabIndex: -1,
    ref,
    style: {
      ...props.style,
      position: "absolute",
      pointerEvents: "none",
      opacity: 0,
      margin: 0
    }
  });
};
function useSwitch(props, [checked, setChecked], ref) {
  {
    const {
      disabled,
      name,
      value,
      required
    } = props;
    const hasConsumerStoppedPropagationRef = React.useRef(false);
    const [button, setButton] = React.useState(null);
    const composedRefs = (0, import_compose_refs.useComposedRefs)(ref, setButton);
    const isFormControl = import_constants.isWeb ? button ? Boolean(button.closest("form")) : true : false;
    const labelId = (0, import_label.useLabelContext)(button);
    const ariaLabelledBy = props["aria-labelledby"] || props.labeledBy || labelId;
    return {
      switchProps: {
        role: "switch",
        "aria-checked": checked,
        ...(import_constants.isWeb ? {
          tabIndex: disabled ? void 0 : 0,
          "data-state": getState(checked),
          "data-disabled": disabled ? "" : void 0,
          disabled
        } : {}),
        "aria-labelledby": ariaLabelledBy,
        onPress: (0, import_helpers.composeEventHandlers)(props.onPress, event => {
          setChecked(prevChecked => !prevChecked);
          if (import_constants.isWeb && isFormControl) {
            hasConsumerStoppedPropagationRef.current = event.isPropagationStopped();
            if (!hasConsumerStoppedPropagationRef.current) event.stopPropagation();
          }
        })
      },
      switchRef: composedRefs,
      bubbleInput: import_constants.isWeb && isFormControl ? /* @__PURE__ */(0, import_jsx_runtime.jsx)(BubbleInput, {
        control: button,
        bubbles: !hasConsumerStoppedPropagationRef.current,
        name,
        value,
        checked,
        required,
        disabled,
        style: {
          transform: "translateX(-100%)"
        }
      }) : null
    };
  }
}
