// @tamagui/portal/dist/cjs/GorhomPortalItem.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var GorhomPortalItem_native_exports = {};
__export(GorhomPortalItem_native_exports, {
  GorhomPortalItem: () => GorhomPortalItem
});
module.exports = __toCommonJS(GorhomPortalItem_native_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("@tamagui/constants");
var import_core = require("@tamagui/core");
var import_native = require("@tamagui/native");
var import_react = require("react");
var import_GorhomPortal = require("./GorhomPortal.native.js");
var GorhomPortalItem = function (props) {
  var {
    name: _providedName,
    hostName = "root",
    handleOnMount: _providedHandleOnMount,
    handleOnUnmount: _providedHandleOnUnmount,
    handleOnUpdate: _providedHandleOnUpdate,
    children,
    passThrough
  } = props;
  var portalState = (0, import_native.getPortal)().state;
  if (portalState.type === "teleport") {
    if (passThrough) {
      return children;
    }
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_native.NativePortal, {
      hostName,
      children
    });
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(GorhomPortalItemFallback, {
    name: _providedName,
    hostName,
    handleOnMount: _providedHandleOnMount,
    handleOnUnmount: _providedHandleOnUnmount,
    handleOnUpdate: _providedHandleOnUpdate,
    passThrough,
    children
  });
};
var GorhomPortalItemFallback = function (props) {
  var {
    name: _providedName,
    hostName,
    handleOnMount: _providedHandleOnMount,
    handleOnUnmount: _providedHandleOnUnmount,
    handleOnUpdate: _providedHandleOnUpdate,
    children,
    passThrough
  } = props;
  var {
    addPortal: addUpdatePortal,
    removePortal
  } = (0, import_GorhomPortal.usePortal)(hostName);
  var id = (0, import_react.useId)();
  var name = _providedName || id;
  var handleOnMount = (0, import_core.useEvent)(function () {
    if (_providedHandleOnMount) {
      _providedHandleOnMount(function () {
        return addUpdatePortal(name, children);
      });
    } else {
      addUpdatePortal(name, children);
    }
  });
  var handleOnUnmount = (0, import_core.useEvent)(function () {
    if (_providedHandleOnUnmount) {
      _providedHandleOnUnmount(function () {
        return removePortal(name);
      });
    } else {
      removePortal(name);
    }
  });
  var handleOnUpdate = (0, import_core.useEvent)(function () {
    if (_providedHandleOnUpdate) {
      _providedHandleOnUpdate(function () {
        return addUpdatePortal(name, children);
      });
    } else {
      addUpdatePortal(name, children);
    }
  });
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    if (passThrough) return;
    handleOnMount();
    return function () {
      handleOnUnmount();
    };
  }, []);
  (0, import_react.useEffect)(function () {
    if (passThrough) return;
    handleOnUpdate();
  }, [children]);
  return passThrough ? children : null;
};
