// @tamagui/portal/dist/cjs/GorhomPortal.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
// If the importer is in node compatibility mode or this is not an ESM
// file that has been converted to a CommonJS file using a Babel-
// compatible transform (i.e. "__esModule" has not been set), then set
// "default" to the CommonJS "module.exports" for node compatibility.
isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
  value: mod,
  enumerable: true
}) : target, mod));
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var GorhomPortal_exports = {};
__export(GorhomPortal_exports, {
  ACTIONS: () => ACTIONS,
  INITIAL_STATE: () => INITIAL_STATE,
  PortalHost: () => PortalHost,
  PortalProvider: () => PortalProvider,
  usePortal: () => usePortal
});
module.exports = __toCommonJS(GorhomPortal_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_constants = require("@tamagui/constants");
var import_native = require("@tamagui/native");
var import_start_transition = require("@tamagui/start-transition");
var import_react = __toESM(require("react"), 1);
require("./constants.native.js");
var ACTIONS = /* @__PURE__ */function (ACTIONS2) {
  ACTIONS2[ACTIONS2["REGISTER_HOST"] = 0] = "REGISTER_HOST";
  ACTIONS2[ACTIONS2["DEREGISTER_HOST"] = 1] = "DEREGISTER_HOST";
  ACTIONS2[ACTIONS2["ADD_UPDATE_PORTAL"] = 2] = "ADD_UPDATE_PORTAL";
  ACTIONS2[ACTIONS2["REMOVE_PORTAL"] = 3] = "REMOVE_PORTAL";
  return ACTIONS2;
}(ACTIONS || {});
var INITIAL_STATE = {};
var registerHost = function (state, hostName) {
  if (!(hostName in state)) state[hostName] = [];
  return state;
};
var deregisterHost = function (state, hostName) {
  delete state[hostName];
  return state;
};
var addUpdatePortal = function (state, hostName, portalName, node) {
  if (!(hostName in state)) state = registerHost(state, hostName);
  var index = state[hostName].findIndex(function (item) {
    return item.name === portalName;
  });
  if (index !== -1) state[hostName][index].node = node;else state[hostName].push({
    name: portalName,
    node
  });
  return state;
};
var removePortal = function (state, hostName, portalName) {
  if (!(hostName in state)) {
    if (process.env.NODE_ENV === "development") console.info(`Failed to remove portal '${portalName}', '${hostName}' was not registered!`);
    return state;
  }
  var index = state[hostName].findIndex(function (item) {
    return item.name === portalName;
  });
  if (index !== -1) state[hostName].splice(index, 1);
  return state;
};
var reducer = function (state, action) {
  var {
    type
  } = action;
  switch (type) {
    case 0:
      return registerHost({
        ...state
      }, action.hostName);
    case 1:
      return deregisterHost({
        ...state
      }, action.hostName);
    case 2:
      return addUpdatePortal({
        ...state
      }, action.hostName, action.portalName, action.node);
    case 3:
      return removePortal({
        ...state
      }, action.hostName, action.portalName);
    default:
      return state;
  }
};
var PortalStateContext = /* @__PURE__ */(0, import_react.createContext)(null);
var PortalDispatchContext = /* @__PURE__ */(0, import_react.createContext)(null);
var PortalProviderActiveContext = /* @__PURE__ */(0, import_react.createContext)(false);
var usePortalState = function (hostName) {
  var state = (0, import_react.useContext)(PortalStateContext);
  if (state === null) throw new Error("'PortalStateContext' cannot be null, please add 'PortalProvider' to the root component.");
  return state[hostName] || [];
};
var usePortal = function () {
  var hostName = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "root";
  var dispatch = (0, import_react.useContext)(PortalDispatchContext);
  if (dispatch === null) throw new Error("'PortalDispatchContext' cannot be null, please add 'PortalProvider' to the root component.");
  var registerHost2 = (0, import_react.useCallback)(function () {
    dispatch({
      type: 0,
      hostName
    });
  }, []);
  var deregisterHost2 = (0, import_react.useCallback)(function () {
    dispatch({
      type: 1,
      hostName
    });
  }, []);
  var addUpdatePortal2 = (0, import_react.useCallback)(function (name, node) {
    dispatch({
      type: 2,
      hostName,
      portalName: name,
      node
    });
  }, []);
  return {
    registerHost: registerHost2,
    deregisterHost: deregisterHost2,
    addPortal: addUpdatePortal2,
    updatePortal: addUpdatePortal2,
    removePortal: (0, import_react.useCallback)(function (name) {
      dispatch({
        type: 3,
        hostName,
        portalName: name
      });
    }, [])
  };
};
var PortalProviderComponent = function (param) {
  var {
    rootHostName = "root",
    shouldAddRootHost = true,
    children
  } = param;
  var isAlreadyInProvider = (0, import_react.useContext)(PortalProviderActiveContext);
  if (process.env.NODE_ENV === "development") {
    if (isAlreadyInProvider && shouldAddRootHost) console.warn(`[tamagui] Nested PortalProvider with shouldAddRootHost detected. This causes hydration mismatches. TamaguiProvider from 'tamagui' already includes PortalProvider - remove the explicit PortalProvider wrapper or set shouldAddRootHost={false}.`);
  }
  var [state, dispatch] = (0, import_react.useReducer)(reducer, INITIAL_STATE);
  var transitionDispatch = (0, import_react.useMemo)(function () {
    var next = function (value) {
      (0, import_start_transition.startTransition)(function () {
        dispatch(value);
      });
    };
    return next;
  }, [dispatch]);
  var portalState = (0, import_native.getPortal)().state;
  var content = /* @__PURE__ */(0, import_jsx_runtime.jsx)(PortalProviderActiveContext.Provider, {
    value: true,
    children: /* @__PURE__ */(0, import_jsx_runtime.jsx)(PortalDispatchContext.Provider, {
      value: transitionDispatch,
      children: /* @__PURE__ */(0, import_jsx_runtime.jsxs)(PortalStateContext.Provider, {
        value: state,
        children: [children, shouldAddRootHost && /* @__PURE__ */(0, import_jsx_runtime.jsx)(PortalHost, {
          name: rootHostName
        })]
      })
    })
  });
  if (portalState.type === "teleport") return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_native.NativePortalProvider, {
    children: content
  });
  return content;
};
var PortalProvider = /* @__PURE__ */(0, import_react.memo)(PortalProviderComponent);
PortalProvider.displayName = "PortalProvider";
var defaultRenderer = function (children) {
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, {
    children
  });
};
var PortalHost = /* @__PURE__ */(0, import_react.memo)(function PortalHost2(props) {
  if ((0, import_native.getPortal)().state.type === "teleport") return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_native.NativePortalHost, {
    name: props.name
  });
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(PortalHostNonNative, {
    ...props
  });
});
function PortalHostNonNative(props) {
  var {
    name,
    forwardProps,
    render = defaultRenderer
  } = props;
  var state = usePortalState(name);
  var {
    registerHost: registerHost2,
    deregisterHost: deregisterHost2
  } = usePortal(props.name);
  (0, import_constants.useIsomorphicLayoutEffect)(function () {
    registerHost2();
    return function () {
      deregisterHost2();
    };
  }, []);
  if (forwardProps) return render(state.map(function (item) {
    var next = item.node;
    var {
      children,
      ...restForwardProps
    } = forwardProps;
    if (forwardProps) return import_react.default.Children.map(next, function (child) {
      return /* @__PURE__ */import_react.default.isValidElement(child) ? /* @__PURE__ */import_react.default.cloneElement(child, {
        key: child.key,
        ...restForwardProps
      }) : child;
    });
    return next;
  }));
  return render(state.map(function (item) {
    return item.node;
  }));
}
