// @tamagui/portal/dist/cjs/Portal.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all) __defProp(target, name, {
    get: all[name],
    enumerable: true
  });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from)) if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
  }
  return to;
};
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: true
}), mod);
var Portal_native_exports = {};
__export(Portal_native_exports, {
  Portal: () => Portal
});
module.exports = __toCommonJS(Portal_native_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_core = require("@tamagui/core");
var import_native = require("@tamagui/native");
var import_z_index_stack = require("@tamagui/z-index-stack");
var import_GorhomPortalItem = require("./GorhomPortalItem.native.js");
var import_helpers = require("./helpers.native.js");
var Portal = function (propsIn) {
  var zIndex = (0, import_z_index_stack.useStackedZIndex)((0, import_helpers.getStackedZIndexProps)(propsIn));
  var {
    children,
    passThrough
  } = propsIn;
  var contents = /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_core.View, {
    pointerEvents: "box-none",
    position: "absolute",
    inset: 0,
    maxWidth: "100%",
    zIndex,
    passThrough,
    children
  });
  var portalState = (0, import_native.getPortal)().state;
  if (portalState.type === "teleport") {
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_native.NativePortal, {
      hostName: "root",
      children: contents
    });
  }
  return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_GorhomPortalItem.GorhomPortalItem, {
    passThrough,
    hostName: "root",
    children: contents
  });
};
