// @tamagui/label/dist/esm/Label.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useComposedRefs } from "@tamagui/compose-refs";
import { isWeb } from "@tamagui/constants";
import { createContext } from "@tamagui/create-context";
import { focusFocusable } from "@tamagui/focusable";
import { getButtonSized } from "@tamagui/get-button-sized";
import { getFontSized } from "@tamagui/get-font-sized";
import { SizableText } from "@tamagui/text";
import { styled } from "@tamagui/web";
import * as React from "react";
var NAME = "Label";
var [LabelProvider, useLabelContextImpl] = createContext(NAME, {
  id: void 0,
  controlRef: {
    current: null
  }
});
var LabelFrame = styled(SizableText, {
  name: "Label",
  render: "label",
  variants: {
    unstyled: {
      false: {
        size: "$true",
        color: "$color",
        backgroundColor: "transparent",
        display: "flex",
        alignItems: "center",
        userSelect: "none",
        cursor: "default",
        pressStyle: {
          color: "$colorPress"
        }
      }
    },
    size: {
      "...size": function (val, extras) {
        var buttonStyle = getButtonSized(val, extras);
        var buttonHeight = buttonStyle === null || buttonStyle === void 0 ? void 0 : buttonStyle.height;
        var fontStyle = getFontSized(val, extras);
        return {
          ...fontStyle,
          lineHeight: buttonHeight ? extras.tokens.size[buttonHeight] : void 0
        };
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
var Label = LabelFrame.styleable(function Label2(props, forwardedRef) {
  var {
    htmlFor,
    id: idProp,
    ...labelProps
  } = props;
  var controlRef = React.useRef(null);
  var ref = React.useRef(null);
  var composedRefs = useComposedRefs(forwardedRef, ref);
  var backupId = React.useId();
  var id = idProp !== null && idProp !== void 0 ? idProp : backupId;
  if (isWeb) {
    React.useEffect(function () {
      if (htmlFor) {
        var element = document.getElementById(htmlFor);
        var label = ref.current;
        if (label && element) {
          var getAriaLabel = function () {
            return element.getAttribute("aria-labelledby");
          };
          var ariaLabelledBy = [id, getAriaLabel()].filter(Boolean).join(" ");
          element.setAttribute("aria-labelledby", ariaLabelledBy);
          controlRef.current = element;
          return function () {
            var _getAriaLabel;
            if (!id) return;
            var ariaLabelledBy2 = (_getAriaLabel = getAriaLabel()) === null || _getAriaLabel === void 0 ? void 0 : _getAriaLabel.replace(id, "");
            if (ariaLabelledBy2 === "") {
              element.removeAttribute("aria-labelledby");
            } else if (ariaLabelledBy2) {
              element.setAttribute("aria-labelledby", ariaLabelledBy2);
            }
          };
        }
      }
    }, [id, htmlFor]);
  }
  return /* @__PURE__ */_jsx(LabelProvider, {
    id,
    controlRef,
    children: /* @__PURE__ */_jsx(LabelFrame, {
      id,
      // @ts-ignore
      htmlFor,
      ...labelProps,
      ref: composedRefs,
      onMouseDown: function (event) {
        var _props_onMouseDown;
        (_props_onMouseDown = props.onMouseDown) === null || _props_onMouseDown === void 0 ? void 0 : _props_onMouseDown.call(props, event);
        if (!event.defaultPrevented && event.detail > 1) {
          event.preventDefault();
        }
      },
      onPress: function (event) {
        var _props_onPress;
        (_props_onPress = props.onPress) === null || _props_onPress === void 0 ? void 0 : _props_onPress.call(props, event);
        if (isWeb) {
          if (htmlFor || !controlRef.current || event.defaultPrevented) return;
          var isClickingControl = controlRef.current.contains(event.target);
          var isUserClick = event.isTrusted === true;
          if (!isClickingControl && isUserClick) {
            controlRef.current.click();
            controlRef.current.focus();
          }
        } else {
          if (props.htmlFor) {
            focusFocusable(props.htmlFor);
          }
        }
      }
    })
  });
});
var useLabelContext = function (element) {
  var context = useLabelContextImpl("LabelConsumer");
  var {
    controlRef
  } = context;
  React.useEffect(function () {
    if (element) controlRef.current = element;
  }, [element, controlRef]);
  return context.id;
};
export { Label, LabelFrame, useLabelContext };
